{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1336fe2c",
   "metadata": {},
   "source": [
    "# UBS-seq Pipeline\n",
    "## Basic Workflow"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4f9191a",
   "metadata": {},
   "source": [
    "**C. Bryan Daniels**\n",
    "\n",
    "**2/1/2024**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1153fdea-42ab-4a16-9376-aae40c035695",
   "metadata": {},
   "source": [
    "## Project: UBS-seq Basic Pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1999833",
   "metadata": {},
   "source": [
    "The purpose of this project is run a minimally viable UBS-seq pipline. For simplicity, it will run several single-end samples, mapping only to the genome. The core steps of the pipeline are:\n",
    "- cut_apapter\n",
    "- quality_control\n",
    "- align2ref\n",
    "- sort2ref\n",
    "- dedupe\n",
    "- filter->all_multi_unique\n",
    "- call_peaks\n",
    "- select_groups\n",
    "- analysis_and_annotation\n",
    "\n",
    "This pipeline is based upon the paper by [Qing Dai, etal](https://doi.org/10.1038/s41587-023-02034-w) and the UBS-seq pipeline developed by [Chang Ye](https://github.com/y9c/m5C-UBSseq)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db5933a4-a627-451d-abe7-7e68204c1d47",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd477b9e-5c7d-4c3c-8125-89644f367a8d",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### The logic for the Pipeline is defined through a series of Steps using dirs to save intermediate results\n",
    "1. For each **Step** in the pipeline a dir will be created and labeled **Step** and will contain all files created by that **Step**\n",
    "2. Within a **Step**, **in_path** and **out_path** will generically refer to the prior and current **Step**\n",
    "3. Within each **Step**, the appropriate processes will occur. Generally this involves processing files from **in_path** and saving to **out_path**\n",
    "4. **Abbreviated filenames** should not change through the pipeline (suffixes will reflect current file formats). The dir name should reflect the **Step**, not the filename.\n",
    "6. The function **mkpath(step)** returns a path for a dir **Step**. It will create a dir if need be, but not overwrite an existing dir\n",
    "8. The function **fname(path,sample,suffix)** returns a file name without actually creating the file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13e28624-07ae-4a54-ba80-dfde893059bc",
   "metadata": {},
   "source": [
    "#### Environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "41e148d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys, subprocess\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt \n",
    "from pathlib import Path\n",
    "from IPython.display import display, HTML\n",
    "from snakemake import load_configfile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "42cc293d-6fe9-45fc-a0cd-59b7dfffb18b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fname(path, base, sufix):\n",
    "    'Return a path and suffix complete filename'\n",
    "    return Path(path)/f\"{base}.{sufix}\"\n",
    "\n",
    "def mkpath(path):\n",
    "    'Return dir path name; creates (not over-writting) a new dir within the pwd'\n",
    "    path = Path(path)\n",
    "    if not os.path.exists(path): os.makedirs(path)\n",
    "    return path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "daf19bdc-0368-4ecf-98cf-785917a77577",
   "metadata": {},
   "source": [
    "#### Path References"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0e9cbdfe-97bf-4c0d-9e86-ef93ad062c5d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Skip for now \n",
    "# config = load_configfile(home_path/\"config.yaml\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b02f60de-c362-4078-beba-82c870de2c78",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Repo home; Assuming script is run within 'workspace/' Edit otherwise!\n",
    "home_path = Path.cwd()/'..' # "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9aca73fd-ff33-45c1-a011-49464e7a699c",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_path = home_path/'reference'\n",
    "data_path = home_path/'data'\n",
    "workspace_path = home_path/'workspace'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dbfc6989-8a84-4b33-ae5c-ee507af9442f",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_fa = ref_path/'genome/Homo_sapiens.GRCh38.genome.fa'\n",
    "genome_idx = ref_path/'index/hist3n/Homo_sapiens.GRCh38.genome'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ae9c5015",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add to shell PATH\n",
    "os.environ['PATH'] = f\"{str(home_path)}:\" + os.environ['PATH'] # home_path\n",
    "os.environ['PATH'] = '/home/cdaniels/bin/homer:' + os.environ['PATH'] # homer\n",
    "os.environ['PATH'] = '/home/cdaniels/bin/hisat-3n:' + os.environ['PATH'] # hisat-3n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "889270cf-bf90-4da3-90fe-397d8d0b57f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "32"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Number of cores                                                                                                                                                                                               \n",
    "nc = get_ipython().getoutput('nproc')                                                                                                                                                                           \n",
    "nc = int(nc[0])                                                                                                                                                                                                 \n",
    "nc  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8499228-734a-4c2c-ad0a-deab43ae64c7",
   "metadata": {},
   "source": [
    "#### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "46c1ab4f-937a-4b6a-a09d-a34462da366b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def nlines(file):\n",
    "    'Returns fast linecout (fast)'\n",
    "    result = subprocess.run(['wc', '-l', file], stdout=subprocess.PIPE)\n",
    "    n = int(result.stdout.split()[0])\n",
    "    return n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "ed22df72-94d2-412d-bf77-67e25b974acc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def nseqs(bam_fastq):\n",
    "    'Returns number of sequences in bam, sam, fasta or fastq file'\n",
    "    n = !samtools view -c {bam_fastq}\n",
    "    return int(n[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "db7b2623-6c8a-465f-97af-1cd3e358849b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def samples_string(samples,path,suffix='bam'):\n",
    "    'Returns a space delimited string of sample files'\n",
    "    return \" \".join([str(fname(path,sample,suffix)) for sample in samples])    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "f9ffe35a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_table(ds1, ds2, ds1_name, ds2_name, y_label=None, xs_labels=None, table_label=None):\n",
    "    'Makes a bar graph comparing two datasets and their corresponding names'\n",
    "    assert(len(ds1)==len(ds2))\n",
    "    n = np.arange(len(ds1))\n",
    "    width = 0.35\n",
    "    fig, ax = plt.subplots()\n",
    "    rects1 = ax.bar(n - width/2, ds1, width, label=ds1_name)\n",
    "    rects2 = ax.bar(n + width/2, ds2, width,label=ds2_name)\n",
    "    ax.set_ylabel(y_label)\n",
    "    ax.set_title(table_label)\n",
    "    ax.set_xticks(n)\n",
    "    ax.set_xticklabels(range(1,len(ds1)+1)) if xs_labels == None else ax.set_xticklabels(xs_labels)\n",
    "    ax.legend()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a7b59797-ffd0-4d1c-947e-45f07e301a2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_histogram(ds, ds_name, table_label=None, y_label=\"Frequency\", density=True):\n",
    "    'Makes a histogram for a dataset and its name'\n",
    "    fig, ax = plt.subplots()\n",
    "    ax.hist(ds, density=density)\n",
    "    ax.set_ylabel(y_label)\n",
    "    ax.set_xlabel(ds_name)\n",
    "    if table_label is not None:\n",
    "        ax.set_title(table_label)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a21c1f30-2663-4aa7-9e81-fd90a96b92e5",
   "metadata": {},
   "source": [
    "## Step: select_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf0c6fbb-a4b9-4a0c-97d2-80646efb627e",
   "metadata": {},
   "source": [
    "#### Select Samples and Filenames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d73f04c9-8803-4e6f-8336-a53a40d78bab",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "out_path = mkpath('samples')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1245b4d6-d147-423a-8d74-038e689e960e",
   "metadata": {},
   "source": [
    "#### Project definitions for treated samples, control samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "67bbf37a-e8c4-4f9f-a267-251bed3f485f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# short sample names\n",
    "treated = [\"t1\",\"t2\"] # Edit\n",
    "control = [] # Edit\n",
    "samples = treated + control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "32c7d8e1-0b1b-495e-af19-7de635060e12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['t1', 't2']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a1738eec-e76c-4104-858a-5319359818a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sample filenames\n",
    "samples_fn = [data_path/'test0_R1.fq.gz', data_path/'test1_R1.fq.gz', data_path/'test2_R1.fq.gz']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f98ef3fd-e6fc-442d-9147-d2ab1dc04074",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'t1': PosixPath('/home/cdaniels/uofc_data/ubs_seq/UBS-seq_basic/workspace/../data/test0_R1.fq.gz'),\n",
       " 't2': PosixPath('/home/cdaniels/uofc_data/ubs_seq/UBS-seq_basic/workspace/../data/test1_R1.fq.gz')}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# dict\n",
    "s2fn = {name: fname for (name,fname) in zip(samples,samples_fn)};\n",
    "s2fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "4c84cd58-777e-4645-bac1-5a3c549ad8dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#for (s,fn) in s2fn.items():\n",
    "#    ! ln -s {data_path/fn} {fname(out_path,s,\"fq.gz\")}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4f8d16b",
   "metadata": {},
   "source": [
    "```\n",
    "# 3 mapping the reads to human genome\n",
    "# using the unaligned read from #2\n",
    "# https://www.gencodegenes.org/human/\n",
    "# Release 39\n",
    "# https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_39/GRCh38.primary_assembly.genome.fa.gz\n",
    "#\n",
    "#Comprehensive gene annotation\n",
    "# https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_39/gencode.v39.annotation.gtf.gz\n",
    "# Need to crate a STAR Index\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b919096-f2bd-4baa-b629-a3ea09e6075d",
   "metadata": {},
   "source": [
    "##### Download human genome and corresponding annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12e6cdcc",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#!wget -q -O {genome_path}/GRCh38.primary_assembly.genome.fa.gz https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_39/GRCh38.primary_assembly.genome.fa.gz\n",
    "#!gunzip {genome_path}/GRCh38.primary_assembly.genome.fa.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7295c25b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!wget -q -O {genome_path}/gencode.v39.annotation.gtf.gz https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_39/gencode.v39.annotation.gtf.gz\n",
    "#!gunzip {genome_path}/gencode.v39.annotation.gtf.gz "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d77ef355",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_fasta      = genome_path/\"GRCh38.primary_assembly.genome.fa\"\n",
    "genome_annotation = genome_path/\"gencode.v39.annotation.gtf\"\n",
    "genome_bed12      = genome_path/\"gencode.v39.annotation.bed12\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a151df26",
   "metadata": {},
   "source": [
    "##### Generate human genome index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "c45ca391",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#! STAR  --runMode genomeGenerate \\\n",
    "#        --runThreadN {nc} \\\n",
    "#        --readFilesCommand zcat \\\n",
    "#        --genomeDir {genome_index} \\\n",
    "#        --genomeFastaFiles {genome_fasta} \\\n",
    "#        --sjdbGTFfile {genome_annotation} \\\n",
    "#        --sjdbOverhang 80 "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6658fe03-cdac-4749-9b06-20d9d094b69c",
   "metadata": {},
   "source": [
    "##### Generate human genome annotations"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "15dea431-d6fb-4c4e-bd3a-5cd171b4316b",
   "metadata": {},
   "source": [
    "Both of the following commands create **bed12** files that can be used by Homer. **gffread** files are two-orderes of magnitude and are much faster. Accordingly, it is used by default. *I'm not sure if the larger **bed12** created by **gtfbed** has advantages. It is saved for such a purpose*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32c2ee09-2c56-4198-afc3-c402b9abc29b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use by defalut\n",
    "#! gffread --bed -o {genome_bed12} {genome_annotation}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "bdc7e109-2d64-4020-a31d-06358bbe46d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Alternative, larger and slower and better??\n",
    "#! gtf2bed < {genome_annotation} > {genome_bed12}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "1739d29c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 28G\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 1.2K Oct  3 09:30 chrLength.txt\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 3.2K Oct  3 09:30 chrNameLength.txt\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 2.0K Oct  3 09:30 chrName.txt\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 2.1K Oct  3 09:30 chrStart.txt\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  53M Oct  3 09:30 exonGeTrInfo.tab\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  22M Oct  3 09:30 exonInfo.tab\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 2.5M Oct  3 09:30 geneInfo.tab\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 3.0G Oct  3 09:46 Genome\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  966 Oct  3 09:46 genomeParameters.txt\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  50K Oct  3 09:46 Log.out\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  24G Oct  3 09:46 SA\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 1.5G Oct  3 09:46 SAindex\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  12M Oct  3 09:45 sjdbInfo.txt\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  12M Oct  3 09:30 sjdbList.fromGTF.out.tab\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 9.8M Oct  3 09:45 sjdbList.out.tab\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  16M Oct  3 09:30 transcriptInfo.tab\n"
     ]
    }
   ],
   "source": [
    "!ls -lh {genome_index}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c8cd20c-e5e6-44b4-8c06-f0ffb7f6051b",
   "metadata": {},
   "source": [
    "#### Only need to create rRNA index once"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "1c5ee74b",
   "metadata": {},
   "outputs": [],
   "source": [
    "rRNA_path  = mkpath(ref_path/\"rRNA\")\n",
    "rRNA_index = rRNA_path/\"index/\" # dir/basename"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30ab6740",
   "metadata": {},
   "source": [
    "```\n",
    "# 2 filtering the reads aligned to the rRNA and keep the unaligned reads\n",
    "# human rRNA sequence\n",
    "# NR_003285.3 Homo sapiens RNA, 5.8S ribosomal N5 (RNA5-8SN5), ribosomal RNA\n",
    "# NR_003286.4 Homo sapiens RNA, 18S ribosomal N5 (RNA18SN5), ribosomal RNA\n",
    "# NR_003287.4 Homo sapiens RNA, 28S ribosomal N5 (RNA28SN5), ribosomal RNA\n",
    "# NR_023363.1 Homo sapiens RNA, 5S ribosomal 1 (RNA5S1), ribosomal RNA\n",
    "\n",
    "# using bowtie2\n",
    "# https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml\n",
    "# 2.1 building bowtie2 index\n",
    "# https://bowtie-bio.sourceforge.net/bowtie2/manual.shtml#indexing-a-reference-genome\n",
    "\n",
    "bowtie2-build rRNA_Sequence.fa OutIndexFileName \n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "20d9a8dc",
   "metadata": {},
   "source": [
    "    \n",
    "   Given an `accession number` and the `name` download the corresponding fasta file from NCBI."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "ae485fae",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_ncbi_accession(number, name, path = \".\"):\n",
    "    os.system(\n",
    "        f'wget -q \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nucleotide&id={number}&rettype=fasta&retmode=text\" -O {path}/{name}.fa')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "307dab24",
   "metadata": {},
   "source": [
    "##### Download rRNA fasta sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "1d8424b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "rRNA_num = [\"NR_003285.3\",\"NR_003286.4\",\"NR_003287.4\",\"NR_023363.1\"]\n",
    "rRNA_name= [\"RNA5-8SN5\",\"RNA18SN5\",\"RNA28SN5\",\"RNA5S1\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "id": "4650770e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#for num, name in zip(rRNA_num, rRNA_name):\n",
    "#    ! {get_ncbi_accession(num,name,rRNA_path)} &> /dev/null"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "10dcbc3b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/cdaniels/uofc_data/ref/rRNA/RNA18SN5.fa\n",
      "/home/cdaniels/uofc_data/ref/rRNA/RNA28SN5.fa\n",
      "/home/cdaniels/uofc_data/ref/rRNA/RNA5-8SN5.fa\n",
      "/home/cdaniels/uofc_data/ref/rRNA/RNA5S1.fa\n"
     ]
    }
   ],
   "source": [
    "! ls {rRNA_path}/*.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d1afa35",
   "metadata": {},
   "source": [
    "##### Use bowtie2-build to create an rRNA index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 337,
   "id": "395fb9f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bowtie 2 version 2.5.1 by Ben Langmead (langmea@cs.jhu.edu, www.cs.jhu.edu/~langmea)\n",
      "Usage: bowtie2-build [options]* <reference_in> <bt2_index_base>\n",
      "    reference_in            comma-separated list of files with ref sequences\n",
      "    bt2_index_base          write bt2 data to files with this dir/basename\n",
      "*** Bowtie 2 indexes will work with Bowtie v1.2.3 and later. ***\n",
      "Options:\n",
      "    -f                      reference files are Fasta (default)\n",
      "    -c                      reference sequences given on cmd line (as\n",
      "                            <reference_in>)\n",
      "    --large-index           force generated index to be 'large', even if ref\n",
      "                            has fewer than 4 billion nucleotides\n",
      "    --debug                 use the debug binary; slower, assertions enabled\n",
      "    --sanitized             use sanitized binary; slower, uses ASan and/or UBSan\n",
      "    --verbose               log the issued command\n",
      "    -a/--noauto             disable automatic -p/--bmax/--dcv memory-fitting\n",
      "    -p/--packed             use packed strings internally; slower, less memory\n",
      "    --bmax <int>            max bucket sz for blockwise suffix-array builder\n",
      "    --bmaxdivn <int>        max bucket sz as divisor of ref len (default: 4)\n",
      "    --dcv <int>             diff-cover period for blockwise (default: 1024)\n",
      "    --nodc                  disable diff-cover (algorithm becomes quadratic)\n",
      "    -r/--noref              don't build .3/.4 index files\n",
      "    -3/--justref            just build .3/.4 index files\n",
      "    -o/--offrate <int>      SA is sampled every 2^<int> BWT chars (default: 5)\n",
      "    -t/--ftabchars <int>    # of chars consumed in initial lookup (default: 10)\n",
      "    --threads <int>         # of threads\n",
      "    --seed <int>            seed for random number generator\n",
      "    -q/--quiet              verbose output (for debugging)\n",
      "    --h/--help              print this message and quit\n",
      "    --version               print version information and quit\n"
     ]
    }
   ],
   "source": [
    "!bowtie2-build -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "f9bd7d3d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#!bowtie2-build $(ls {rRNA_path}/*.fa | paste -sd \",\" -) {rRNA_index}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec9af419",
   "metadata": {},
   "source": [
    "Downsize by selecting `N` samples. When ready to use all, set `N = .9999`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "1aa4914d",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_path = mkpath(\"orig\")\n",
    "out_path = mkpath(\"downsize\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "897a8c8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "N = .9999"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32d1b82f",
   "metadata": {},
   "source": [
    "Copy **N** random records (each 4 lines long) from each to new files in `downsize` with the command `seqtk sample`. There are other useful `seqtk` commands."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "9395b6d3",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Usage:   seqtk sample [-2] [-s seed=11] <in.fa> <frac>|<number>\n",
      "\n",
      "Options: -s INT       RNG seed [11]\n",
      "         -2           2-pass mode: twice as slow but with much reduced memory\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!seqtk sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "dfba9339",
   "metadata": {},
   "outputs": [],
   "source": [
    "for sample in samples:\n",
    "    ! seqtk sample {fname(in_path,sample, \"fq\")} {N} > {fname(out_path,sample, \"fq\")}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e11be53",
   "metadata": {},
   "source": [
    "Check that the files look correct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "25371f62",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 13G\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  2.4G Oct 13 23:47 a4_1.fq\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  708M Oct 13 23:47 a4_2.fq\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  1.6G Oct 13 23:47 a4_3.fq\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  1.1G Oct 13 23:47 a4_4.fq\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  2.3G Oct 13 23:47 no_1.fq\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  2.6G Oct 13 23:48 no_2.fq\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 1007M Oct 13 23:48 no_3.fq\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  1.3G Oct 13 23:48 no_4.fq\n"
     ]
    }
   ],
   "source": [
    "!ls -lh {out_path}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c7307e9-bcb8-4515-be28-da83ddd2649a",
   "metadata": {},
   "source": [
    "How many sequence reads do we have per Sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "113b16b9-601a-4b74-83e7-12d25a9af329",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "downsize/a4_1.fq: 13,144,772\n",
      "downsize/a4_2.fq: 3,907,314\n",
      "downsize/a4_3.fq: 8,779,388\n",
      "downsize/a4_4.fq: 5,636,274\n",
      "downsize/no_1.fq: 12,845,154\n",
      "downsize/no_2.fq: 14,555,266\n",
      "downsize/no_3.fq: 5,530,889\n",
      "downsize/no_4.fq: 7,068,411\n"
     ]
    }
   ],
   "source": [
    "# Total Reads per Samples\n",
    "files = [fname(out_path,sample, \"fq\") for sample in samples]\n",
    "for f in files:\n",
    "    ns = nseqs(f)\n",
    "    print(f\"{f}: {ns:,}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51157b74",
   "metadata": {},
   "source": [
    "## Step: adapter_trimmed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1731c31-b9b1-4a09-b964-3dece49e1910",
   "metadata": {},
   "source": [
    "#### Trim the adapter and downstream sequence as well as trimmng lower quality downstream sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8b898be1",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_path = mkpath(\"samples\")\n",
    "out_path = mkpath(\"trimmed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "632eaa99",
   "metadata": {},
   "outputs": [],
   "source": [
    "#adapter = \"AGATCGGAAGAGCACACGTCTGAACTCCAG\"\n",
    "adapter =  \"AGATCGGAAGAGCACACGTCT\"\n",
    "barcode3 = \"ATCACG\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e73ebe78-4020-4894-8c0c-3239aee02330",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "t1.fq.gz  t2.fq.gz\n"
     ]
    }
   ],
   "source": [
    "!ls {in_path}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca715978-8029-4cfe-a74c-765eb8ed882f",
   "metadata": {},
   "source": [
    "#### cutadapt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa3acb07-1ef4-46b8-9d0f-9362af98a5c8",
   "metadata": {},
   "source": [
    "https://cutadapt.readthedocs.io/en/stable/guide.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "bbb14bd9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cutadapt version 4.4\n",
      "\n",
      "Copyright (C) 2010-2022 Marcel Martin <marcel.martin@scilifelab.se>\n",
      "\n",
      "cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. All reads from input.fastq will be written to\n",
      "output.fastq with the adapter sequence removed. Adapter matching is\n",
      "error-tolerant. Multiple adapter sequences can be given (use further -a\n",
      "options), but only the best-matching adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Run \"cutadapt --help\" to see all command-line options.\n",
      "See https://cutadapt.readthedocs.io/ for full documentation.\n",
      "\n",
      "Options:\n",
      "  -h, --help            Show this help message and exit\n",
      "  --version             Show version number and exit\n",
      "  --debug               Print debug log. Use twice to also print DP matrices\n",
      "  -j CORES, --cores CORES\n",
      "                        Number of CPU cores to use. Use 0 to auto-detect.\n",
      "                        Default: 1\n",
      "\n",
      "Finding adapters:\n",
      "  Parameters -a, -g, -b specify adapters to be removed from each read (or from\n",
      "  R1 if data is paired-end. If specified multiple times, only the best\n",
      "  matching adapter is trimmed (but see the --times option). Use notation\n",
      "  'file:FILE' to read adapter sequences from a FASTA file.\n",
      "\n",
      "  -a ADAPTER, --adapter ADAPTER\n",
      "                        Sequence of an adapter ligated to the 3' end (paired\n",
      "                        data: of the first read). The adapter and subsequent\n",
      "                        bases are trimmed. If a '$' character is appended\n",
      "                        ('anchoring'), the adapter is only found if it is a\n",
      "                        suffix of the read.\n",
      "  -g ADAPTER, --front ADAPTER\n",
      "                        Sequence of an adapter ligated to the 5' end (paired\n",
      "                        data: of the first read). The adapter and any preceding\n",
      "                        bases are trimmed. Partial matches at the 5' end are\n",
      "                        allowed. If a '^' character is prepended ('anchoring'),\n",
      "                        the adapter is only found if it is a prefix of the read.\n",
      "  -b ADAPTER, --anywhere ADAPTER\n",
      "                        Sequence of an adapter that may be ligated to the 5' or\n",
      "                        3' end (paired data: of the first read). Both types of\n",
      "                        matches as described under -a and -g are allowed. If the\n",
      "                        first base of the read is part of the match, the\n",
      "                        behavior is as with -g, otherwise as with -a. This\n",
      "                        option is mostly for rescuing failed library\n",
      "                        preparations - do not use if you know which end your\n",
      "                        adapter was ligated to!\n",
      "  -e E, --error-rate E, --errors E\n",
      "                        Maximum allowed error rate (if 0 <= E < 1), or absolute\n",
      "                        number of errors for full-length adapter match (if E is\n",
      "                        an integer >= 1). Error rate = no. of errors divided by\n",
      "                        length of matching region. Default: 0.1 (10%)\n",
      "  --no-indels           Allow only mismatches in alignments. Default: allow both\n",
      "                        mismatches and indels\n",
      "  -n COUNT, --times COUNT\n",
      "                        Remove up to COUNT adapters from each read. Default: 1\n",
      "  -O MINLENGTH, --overlap MINLENGTH\n",
      "                        Require MINLENGTH overlap between read and adapter for\n",
      "                        an adapter to be found. Default: 3\n",
      "  --match-read-wildcards\n",
      "                        Interpret IUPAC wildcards in reads. Default: False\n",
      "  -N, --no-match-adapter-wildcards\n",
      "                        Do not interpret IUPAC wildcards in adapters.\n",
      "  --action {trim,retain,mask,lowercase,none}\n",
      "                        What to do if a match was found. trim: trim adapter and\n",
      "                        up- or downstream sequence; retain: trim, but retain\n",
      "                        adapter; mask: replace with 'N' characters; lowercase:\n",
      "                        convert to lowercase; none: leave unchanged. Default:\n",
      "                        trim\n",
      "  --rc, --revcomp       Check both the read and its reverse complement for\n",
      "                        adapter matches. If match is on reverse-complemented\n",
      "                        version, output that one. Default: check only read\n",
      "\n",
      "Additional read modifications:\n",
      "  -u LEN, --cut LEN     Remove LEN bases from each read (or R1 if paired; use -U\n",
      "                        option for R2). If LEN is positive, remove bases from\n",
      "                        the beginning. If LEN is negative, remove bases from the\n",
      "                        end. Can be used twice if LENs have different signs.\n",
      "                        Applied *before* adapter trimming.\n",
      "  --nextseq-trim 3'CUTOFF\n",
      "                        NextSeq-specific quality trimming (each read). Trims\n",
      "                        also dark cycles appearing as high-quality G bases.\n",
      "  -q [5'CUTOFF,]3'CUTOFF, --quality-cutoff [5'CUTOFF,]3'CUTOFF\n",
      "                        Trim low-quality bases from 5' and/or 3' ends of each\n",
      "                        read before adapter removal. Applied to both reads if\n",
      "                        data is paired. If one value is given, only the 3' end\n",
      "                        is trimmed. If two comma-separated cutoffs are given,\n",
      "                        the 5' end is trimmed with the first cutoff, the 3' end\n",
      "                        with the second.\n",
      "  --quality-base N      Assume that quality values in FASTQ are encoded as\n",
      "                        ascii(quality + N). This needs to be set to 64 for some\n",
      "                        old Illumina FASTQ files. Default: 33\n",
      "  --poly-a              Trim poly-A tails\n",
      "  --length LENGTH, -l LENGTH\n",
      "                        Shorten reads to LENGTH. Positive values remove bases at\n",
      "                        the end while negative ones remove bases at the\n",
      "                        beginning. This and the following modifications are\n",
      "                        applied after adapter trimming.\n",
      "  --trim-n              Trim N's on ends of reads.\n",
      "  --length-tag TAG      Search for TAG followed by a decimal number in the\n",
      "                        description field of the read. Replace the decimal\n",
      "                        number with the correct length of the trimmed read. For\n",
      "                        example, use --length-tag 'length=' to correct fields\n",
      "                        like 'length=123'.\n",
      "  --strip-suffix STRIP_SUFFIX\n",
      "                        Remove this suffix from read names if present. Can be\n",
      "                        given multiple times.\n",
      "  -x PREFIX, --prefix PREFIX\n",
      "                        Add this prefix to read names. Use {name} to insert the\n",
      "                        name of the matching adapter.\n",
      "  -y SUFFIX, --suffix SUFFIX\n",
      "                        Add this suffix to read names; can also include {name}\n",
      "  --rename TEMPLATE     Rename reads using TEMPLATE containing variables such as\n",
      "                        {id}, {adapter_name} etc. (see documentation)\n",
      "  --zero-cap, -z        Change negative quality values to zero.\n",
      "\n",
      "Filtering of processed reads:\n",
      "  Filters are applied after above read modifications. Paired-end reads are\n",
      "  always discarded pairwise (see also --pair-filter).\n",
      "\n",
      "  -m LEN[:LEN2], --minimum-length LEN[:LEN2]\n",
      "                        Discard reads shorter than LEN. Default: 0\n",
      "  -M LEN[:LEN2], --maximum-length LEN[:LEN2]\n",
      "                        Discard reads longer than LEN. Default: no limit\n",
      "  --max-n COUNT         Discard reads with more than COUNT 'N' bases. If COUNT\n",
      "                        is a number between 0 and 1, it is interpreted as a\n",
      "                        fraction of the read length.\n",
      "  --max-expected-errors ERRORS, --max-ee ERRORS\n",
      "                        Discard reads whose expected number of errors (computed\n",
      "                        from quality values) exceeds ERRORS.\n",
      "  --discard-trimmed, --discard\n",
      "                        Discard reads that contain an adapter. Use also -O to\n",
      "                        avoid discarding too many randomly matching reads.\n",
      "  --discard-untrimmed, --trimmed-only\n",
      "                        Discard reads that do not contain an adapter.\n",
      "  --discard-casava      Discard reads that did not pass CASAVA filtering (header\n",
      "                        has :Y:).\n",
      "\n",
      "Output:\n",
      "  --quiet               Print only error messages.\n",
      "  --report {full,minimal}\n",
      "                        Which type of report to print: 'full' or 'minimal'.\n",
      "                        Default: full\n",
      "  --json FILE           Dump report in JSON format to FILE\n",
      "  -o FILE, --output FILE\n",
      "                        Write trimmed reads to FILE. FASTQ or FASTA format is\n",
      "                        chosen depending on input. Summary report is sent to\n",
      "                        standard output. Use '{name}' for demultiplexing (see\n",
      "                        docs). Default: write to standard output\n",
      "  --fasta               Output FASTA to standard output even on FASTQ input.\n",
      "  -Z                    Use compression level 1 for gzipped output files\n",
      "                        (faster, but uses more space)\n",
      "  --info-file FILE      Write information about each read and its adapter\n",
      "                        matches into FILE. See the documentation for the file\n",
      "                        format.\n",
      "  -r FILE, --rest-file FILE\n",
      "                        When the adapter matches in the middle of a read, write\n",
      "                        the rest (after the adapter) to FILE.\n",
      "  --wildcard-file FILE  When the adapter has N wildcard bases, write adapter\n",
      "                        bases matching wildcard positions to FILE. (Inaccurate\n",
      "                        with indels.)\n",
      "  --too-short-output FILE\n",
      "                        Write reads that are too short (according to length\n",
      "                        specified by -m) to FILE. Default: discard reads\n",
      "  --too-long-output FILE\n",
      "                        Write reads that are too long (according to length\n",
      "                        specified by -M) to FILE. Default: discard reads\n",
      "  --untrimmed-output FILE\n",
      "                        Write reads that do not contain any adapter to FILE.\n",
      "                        Default: output to same file as trimmed reads\n",
      "\n",
      "Paired-end options:\n",
      "  The -A/-G/-B/-U/-Q options work like their lowercase counterparts, but are\n",
      "  applied to R2 (second read in pair)\n",
      "\n",
      "  -A ADAPTER            3' adapter to be removed from R2\n",
      "  -G ADAPTER            5' adapter to be removed from R2\n",
      "  -B ADAPTER            5'/3 adapter to be removed from R2\n",
      "  -U LENGTH             Remove LENGTH bases from R2\n",
      "  -Q [5'CUTOFF,]3'CUTOFF\n",
      "                        Quality-trimming cutoff for R2. Default: same as for R1\n",
      "  -p FILE, --paired-output FILE\n",
      "                        Write R2 to FILE.\n",
      "  --pair-adapters       Treat adapters given with -a/-A etc. as pairs. Either\n",
      "                        both or none are removed from each read pair.\n",
      "  --pair-filter {any,both,first}\n",
      "                        Which of the reads in a paired-end read have to match\n",
      "                        the filtering criterion in order for the pair to be\n",
      "                        filtered. Default: any\n",
      "  --interleaved         Read and/or write interleaved paired-end reads.\n",
      "  --untrimmed-paired-output FILE\n",
      "                        Write second read in a pair to this FILE when no adapter\n",
      "                        was found. Use with --untrimmed-output. Default: output\n",
      "                        to same file as trimmed reads\n",
      "  --too-short-paired-output FILE\n",
      "                        Write second read in a pair to this file if pair is too\n",
      "                        short.\n",
      "  --too-long-paired-output FILE\n",
      "                        Write second read in a pair to this file if pair is too\n",
      "                        long.\n"
     ]
    }
   ],
   "source": [
    "! cutadapt --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd81c901-05a6-4b85-9dc2-2c2201af649d",
   "metadata": {},
   "source": [
    "##### paramaters:\n",
    "- --action=trim, trim adapter and downstream sequence\n",
    "- --nextseq-trim=20, is used to trim these low-quality tails of 'G's by introducing a variable quality threshold"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7715b93c-a9e6-4e60-9189-2cba431867ff",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "```\n",
    "cutadapt -j {threads} \\\n",
    "            -n 2 \\\n",
    "            -a \"{params.barcode3}{params.adapter3};e=0.15;o=6;anywhere;\" \\\n",
    "            --untrimmed-output={output.fastq_untrimmed} \\\n",
    "            -o - {input} 2>{output.report1} | \\\n",
    "        cutadapt -j {threads} \\\n",
    "            -u 5 -u -5 \\\n",
    "            --rename='{{id}}_{{cut_prefix}}{{cut_suffix}} {{comment}}' \\\n",
    "            --max-n=0 \\\n",
    "            -q 15 \\\n",
    "            --nextseq-trim=15 \\\n",
    "            -m 20 \\\n",
    "            --too-short-output={output.fastq_tooshort} \\\n",
    "            -o {output.fastq_cut} - >{output.report2}\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "c556375c-0e45-402f-b5fe-6418606482ca",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done           00:00:01       935,632 reads @   1.5 µs/read;  40.28 M reads/minute\n",
      "Done           00:00:01     1,365,245 reads @   1.3 µs/read;  46.48 M reads/minute\n"
     ]
    }
   ],
   "source": [
    "for sample in samples:\n",
    "    !cutadapt -j 0 --nextseq-trim=15 --action=trim -a '{barcode3}{adapter};e=0.15;o=6;anywhere;'\\\n",
    "            -n 2 -u 5 -u -5 --max-n=0 -q 15 -m 20 \\\n",
    "            --rename='{{id}}_{{cut_prefix}}{{cut_suffix}} {{comment}}'\\\n",
    "            --too-short-output={fname(out_path,sample,\"fastq_tooshort\")} \\\n",
    "            -o {fname(out_path,sample,\"fq.gz\")}  \\\n",
    "            {fname(in_path,sample,\"fq.gz\")} > {fname(out_path,sample,\"log\")}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0fd27b4-6fb6-43ec-abda-908fe49df9f8",
   "metadata": {},
   "source": [
    "#### Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0deddf8f-d88b-4b1f-832b-328714f045f3",
   "metadata": {},
   "source": [
    "Take a look at a FASTQ file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "e07c75da-35fd-4c4e-b07a-3696595806f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('samples/t1.fq.gz')"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_fn  = fname(in_path,treated[0],'fq.gz')\n",
    "out_fn = fname(out_path,treated[0],'fq.gz')\n",
    "in_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "c450ef4a-6b0d-4ded-a5f1-d387be292b48",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@SRR23538290.50 50 length=121\n",
      "GGAAATTGTGGTAATTTTAGAGTTAATATATGTTGATGGGTGTTGATTTTTTTTGATGGTATCACGAGATCGGAAGAGCACACGTCTGAACTCCAGTCACCACTCAATCGCGTATTCCGTC\n",
      "+\n",
      "FFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,FFFFF,FFF:F\n",
      "@SRR23538290.144 144 length=121\n",
      "TTTTTTTGGATATTAGGAAAAAATTTTGTAGAGAGAGTAAAAAATTTACTAGTATCACGAGATCGGAAGAGCACACGTCTGAACTCCAGTCACCACTCAATCTCGTATGCCGTCTTCTGCT\n",
      "+\n",
      "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFF:,FF,F:FFFFFF\n",
      "@SRR23538290.276 276 length=121\n",
      "CCAGGAATTATAGTAATGTTTAGTTTATTGAGTTAAATGTTTGTCGGATGTGTTTGTGAGGGGATGAGGATGATAAGATTTTGTTTAGTATCACGAGATCGGAAGAGCACACGTCTGAACT\n",
      "+\n",
      "FF:,FFF:,FFF,FFF:FFFFFFFFFFFF,F:F:F:F,F,:,FFF,,F,F,,:,:,,FFFFF:FF:F,,F,,FF,F:,,,F,F,F:F::F::FFFFFFF:FFFF:FF,FFFFFFFFF,FFF\n",
      "@SRR23538290.379 379 length=121\n",
      "CCTCAAGAGTTTTAATGTTTTTAATGTAGGTGATGATTGGGAAAACACATCACGAGATCGGAAGAGCACACGTCTGAACTCCAGTCACCACTCAATCTCGTATGCCGTCTTCTTCTTGAAA\n",
      "+\n",
      "FFF,:,FFF:FFFFFFFFFFFFFFF:FF:FFFFFFFFFFF:FFFFFF:FFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFF::,:F:FFF,,,,F,F:,FF\n",
      "\n",
      "gzip: stdout: Broken pipe\n"
     ]
    }
   ],
   "source": [
    "! zcat {in_fn}|head -16  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "bbff0c14-bd75-4988-a235-07400fa4a89b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def show_adapter(reads):\n",
    "    for read in reads.split('\\n'):\n",
    "        read = read.replace(adapter, f'<span style=\"color: blue;\">{barcode3}{adapter}</span>')\n",
    "        display(HTML(read))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b131dc87-f706-4aa6-b82d-8fec51b0b7c2",
   "metadata": {},
   "source": [
    "Look for adapter in untrimmed reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "c6560944-596a-443c-9f53-d07613ceedbd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "GGAAATTGTGGTAATTTTAGAGTTAATATATGTTGATGGGTGTTGATTTTTTTTGATGGTATCACG<span style=\"color: blue;\">ATCACGAGATCGGAAGAGCACACGTCT</span>GAACTCCAGTCACCACTCAATCGCGTATTCCGTC"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTTTTTTGGATATTAGGAAAAAATTTTGTAGAGAGAGTAAAAAATTTACTAGTATCACG<span style=\"color: blue;\">ATCACGAGATCGGAAGAGCACACGTCT</span>GAACTCCAGTCACCACTCAATCTCGTATGCCGTCTTCTGCT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "CCAGGAATTATAGTAATGTTTAGTTTATTGAGTTAAATGTTTGTCGGATGTGTTTGTGAGGGGATGAGGATGATAAGATTTTGTTTAGTATCACG<span style=\"color: blue;\">ATCACGAGATCGGAAGAGCACACGTCT</span>GAACT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "CCTCAAGAGTTTTAATGTTTTTAATGTAGGTGATGATTGGGAAAACACATCACG<span style=\"color: blue;\">ATCACGAGATCGGAAGAGCACACGTCT</span>GAACTCCAGTCACCACTCAATCTCGTATGCCGTCTTCTTCTTGAAA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "CAGATTAAAATTAATTTGAAATTTAATCATAAAGATATAGGAAAATAAAACTTATTATCTAAATACAAAAATAAATCGAAAATCACGAGATCTGAATATCACACTTCTGAACTCCAGTCAC"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "ACTTTGGTGTTTTATGTTGGGGAGGATGATGAGGAGGATGATGATTTTAATTAGAATCACATATCACG<span style=\"color: blue;\">ATCACGAGATCGGAAGAGCACACGTCT</span>GAACTCCAGTCACCACTCAATCTCGTATGCCG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "CCCAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACAAAAAACCAAGAGGGAAAGACAAACAGAAAAAAACAAAAACCCAAAAAAACAGAAA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTCAAGTAATGAAGGGTAAGATGAAGTGAAAGGTAAAGAATTGTGTGAGGGTGGGATTGTTTATTGAGTAGTCGCCATCACG<span style=\"color: blue;\">ATCACGAGATCGGAAGAGCACACGTCT</span>GAACTCCAGTCACCACTC"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "CTGTAAAAATAATGTGGTATTTATGAGTTATTTTTTATAGGGTAGTTGTATCACG<span style=\"color: blue;\">ATCACGAGATCGGAAGAGCACACGTCT</span>GAACTCCAGTCACCACTCAATCTCGTATGCCGTCTTCTGCTTGAA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "reads = ! zcat {in_fn} | head -36  | seqtk seq -A |grep -v '>'\n",
    "reads = ('\\n').join(reads)\n",
    "show_adapter(reads)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fad1e27a-a3a6-4bfe-9920-203e65654575",
   "metadata": {},
   "source": [
    "Verify that adapters and all downstream elements of reads have been trimmed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "65a7f174-35a8-4956-bdf3-eb86888a902a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "TTGTGGTAATTTTAGAGTTAATATATGTTGATGGGTGTTGATTTTTTTTGATGGT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTGGATATTAGGAAAAAATTTTGTAGAGAGAGTAAAAAATTTACTAGT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "AATTATAGTAATGTTTAGTTTATTGAGTTAAATGTTTGTCGGATGTGTTTGTGAGGGGATGAGGATGATAAGATTTTGTTTAGT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "AGAGTTTTAATGTTTTTAATGTAGGTGATGATTGGGAAAACAC"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TAAAATTAATTTGAAATTTAATCATAAAGATATAGGAAAATAAAACTTATTATCTAAATACAAAAATAAATCGAAA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "GGTGTTTTATGTTGGGGAGGATGATGAGGAGGATGATGATTTTAATTAGAATCACAT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACAAAAAACCAAGAGGGAAAGACAAACAGAAAAAAACAAAAACCCAAAAAAA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "GTAATGAAGGGTAAGATGAAGTGAAAGGTAAAGAATTGTGTGAGGGTGGGATTGTTTATTGAGTAGTCGCC"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "AAAATAATGTGGTATTTATGAGTTATTTTTTATAGGGTAGTTGT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "reads = !zcat {out_fn}| head -36  | seqtk seq -A |grep -v '>'\n",
    "reads = ('\\n').join(reads)\n",
    "show_adapter(reads)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22988563-df9e-471e-ad5c-e2b34d86f94f",
   "metadata": {},
   "source": [
    "Verify that adapters and all downstream elements of reads have been trimmed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47d80e80-4f86-421a-bd13-f1b66cb48c48",
   "metadata": {},
   "source": [
    "No reads should have been deleted with cutadapt. Veryify that number of reads before and after cutadapt are the same."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "e03f142f-9d37-4d6c-965d-c17053979e68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[935632, 1365245]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Untrimmed\n",
    "files = [fname(in_path,sample, \"fq.gz\") for sample in samples]\n",
    "ins = [nseqs(f) for f in files]\n",
    "ins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "172a1fac-5cf4-4d31-a7d1-25cc250689f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[930121, 1354899]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Adapter Trimmed\n",
    "files = [fname(out_path,sample, \"fq.gz\") for sample in samples]\n",
    "ins = [nseqs(f) for f in files]\n",
    "ins"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba63237d-b9da-4e19-9b66-0c523efadc17",
   "metadata": {},
   "source": [
    "Compare the sum of read lengths in the original and adapter trimmed files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "9004da50-f983-496a-8ec4-7e8a45047d22",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[114147104, 166559890]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "files = [fname(in_path,sample, \"fq.gz\") for sample in samples]\n",
    "res = []\n",
    "for f in files:\n",
    "    n = !seqtk seq -A {f}|grep -v '>'|wc -c\n",
    "    res.append(int(n[0]))\n",
    "ins = res\n",
    "ins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "fb5d33e9-c511-41f2-b8d4-3f7e606fa4ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[53971150, 71959521]"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "files = [fname(out_path,sample, \"fq.gz\") for sample in samples]\n",
    "res = []\n",
    "for f in files:\n",
    "    n = !seqtk seq -A {f}|grep -v '>'|wc -c\n",
    "    res.append(int(n[0]))\n",
    "outs = res\n",
    "outs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "34da7817",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjcAAAGxCAYAAACeKZf2AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8g+/7EAAAACXBIWXMAAA9hAAAPYQGoP6dpAABEhUlEQVR4nO3df3zN9f//8fux38PGJhuabfJr3iJvi0bym0ZKCe+3an6H1DAqS35MfkQRJer9jpbe0lSIElbefiTJsFLkRzbzYyO82/zc2F7fP/rufDo2nMO2Yy+36+VyLped53m+Xq/H62znnPter+fzdSyGYRgCAAAwiTLOLgAAAKAoEW4AAICpEG4AAICpEG4AAICpEG4AAICpEG4AAICpEG4AAICpEG4AAICpEG4AAICpEG5gSlu3btWjjz6q6tWry8PDQwEBAYqIiNDIkSOdXdoNe/nll1W9enW5urqqQoUKBR5PTU2VxWKx65aamnrNbZ0/f14TJkzQ+vXrb7je/Hri4+Pt6pd/K1OmjCpWrKi2bdtq7dq1N7z9mxUSEqI+ffrc9HrWr19vs38uLi4KCAhQ9+7dtWfPnpsv1A6l+fXQp08fhYSEOLsMlDKuzi4AKGpffvmlHn74YbVq1UrTp09XlSpVlJ6erqSkJH388ceaMWOGs0t02Oeff67JkydrzJgxioyMlIeHR4E+VapU0ZYtW2zannnmGWVmZmrRokUF+l7L+fPnFRcXJ0lq1arVzRVvp+eee069evVSbm6ufv31V8XFxalTp05at26dHnjggRKpoThNmTJFrVu3Vk5OjpKSkjRx4kR988032rVrl6pVq1Zs2zXj6wG4HsINTGf69OkKDQ3VmjVr5Or6f3/i//jHPzR9+nQnVnbjfv75Z0lSdHS0KleuXGgfDw8P3XfffTZtPj4+ysnJKdB+K6pevbq1zubNm6tWrVpq2bKl5s+fb4pwU6tWLev+PfDAA6pQoYL69++v+Ph4jRkz5qbWff78eXl7exf6mBlfD8D1cFoKpnPq1ClVqlTJ5o08X5kytn/yFotFEyZMKNDvylMS8fHxslgsWrdunQYOHCh/f3/5+PgoKipK586dU0ZGhnr06KEKFSqoSpUqGjVqlC5dunTdWvPy8jR9+nTVrVtXHh4eqly5sqKionTkyBGbWl5++WVJUkBAwFVrtldaWpqefPJJVa5cWR4eHgoLC9OMGTOUl5cn6c/TRHfccYckKS4uzno6Jf/5OHDggPr27atatWrJ29tb1apVU5cuXbRr164brqkw4eHhkqTjx4/btGdkZGjQoEG688475e7urtDQUMXFxeny5cs2/eLi4tS0aVP5+fnJx8dHf//73zV//nxd+V3Bly5d0gsvvKDAwEB5e3vr/vvv1w8//FCgnvPnz2vUqFEKDQ2Vp6en/Pz8FB4ersWLF9/Q/uUHnUOHDlnbEhISFBERobJly6pcuXLq2LGjdu7cabNcnz59VK5cOe3atUsdOnRQ+fLl1bZt26tux5HXQ0JCgjp06KAqVarIy8tLYWFhGj16tM6dO1doDb/++qs6duyosmXLqkqVKnr11VclSd9//73uv/9+lS1bVrVr19YHH3xgs3z+6ykxMVF9+/aVn5+fypYtqy5duujgwYPXfe4Mw9DcuXN1zz33yMvLSxUrVtTjjz9u17K4PRBuYDoRERHaunWroqOjtXXrVrtChr0GDBggX19fffzxx3r55Zf10UcfaeDAgercubMaNmyoTz/9VL1799aMGTP01ltvXXd9Q4YM0Ysvvqj27dtrxYoVeuWVV7R69Wo1a9ZMJ0+elCQtW7ZM/fv3lyStXr1aW7Zs0YABA26o/t9//13NmjXT2rVr9corr2jFihVq166dRo0apWeffVbSn6esVq9eLUnq37+/tmzZoi1btmjs2LGSpGPHjsnf31+vvvqqVq9erbfffluurq5q2rSp9u7de0N1FSYlJUWSVLt2bWtbRkaGmjRpojVr1mjcuHH66quv1L9/f02dOlUDBw60WT41NVWDBg3SkiVLtHTpUj322GN67rnn9Morr9j0GzhwoF5//XVFRUXp888/V7du3fTYY4/pf//7n02/mJgYzZs3T9HR0Vq9erU+/PBDde/eXadOnbqh/Ttw4IAkWYPklClT9M9//lP16tXTkiVL9OGHH+rMmTNq0aKFdu/ebbNsTk6OHn74YbVp00aff/659RRiYRx5Pezfv1+dOnXS/PnztXr1ag0fPlxLlixRly5dCvS9dOmSHnvsMXXu3Fmff/65IiMjFRsbq5deekm9e/dWv379tGzZMtWpU0d9+vTR9u3bC6yjf//+KlOmjD766CPNmjVLP/zwg1q1aqU//vjjms/doEGDNHz4cLVr107Lly/X3Llz9csvv6hZs2YFwjBuUwZgMidPnjTuv/9+Q5IhyXBzczOaNWtmTJ061Thz5oxNX0nG+PHjC6wjODjY6N27t/X++++/b0gynnvuOZt+Xbt2NSQZM2fOtGm/5557jL///e/XrHPPnj2GJOOZZ56xad+6dashyXjppZesbePHjzckGb///vs113mlli1bGn/729+s90ePHm1IMrZu3WrTb8iQIYbFYjH27t1rGIZh/P7771d9bq50+fJlIycnx6hVq5YxYsQIa3tKSoohyXj//fevuXx+v2nTphmXLl0yLl68aCQnJxsRERFGlSpVjJSUFGvfQYMGGeXKlTMOHTpks47XX3/dkGT88ssvhW4jNzfXuHTpkjFx4kTD39/fyMvLMwzj/34Hf63bMAxj0aJFhiSbv4H69esbXbt2ve7zcaX//ve/hiQjISHBuHTpknH+/Hlj48aNRs2aNQ0XFxfjxx9/NNLS0gxXV9cCf19nzpwxAgMDjR49eljbevfubUgyFixYYNf2HXk9/FVeXp5x6dIlY8OGDYYk48cffyxQw2effWZtu3TpknHHHXcYkowdO3ZY20+dOmW4uLgYMTEx1rb819Ojjz5qs83NmzcbkoxJkybZbCs4ONh6f8uWLYYkY8aMGTbLHj582PDy8jJeeOEFu54XmNttfeRm48aN6tKli6pWrSqLxaLly5c7vI41a9bovvvuU/ny5XXHHXeoW7du1v844Rz+/v7atGmTtm3bpldffVWPPPKI9u3bp9jYWN19993WIyI34qGHHrK5HxYWJknq3Llzgfa/nm4ozH//+19JKjAjp0mTJgoLC9M333xzw3Vezbp161SvXj01adLEpr1Pnz4yDEPr1q277jouX76sKVOmqF69enJ3d5erq6vc3d21f//+m5r98+KLL8rNzU2enp6655579PPPP2vlypU2M2W++OILtW7dWlWrVtXly5ett8jISEnShg0bbPa1Xbt28vX1lYuLi9zc3DRu3DidOnVKJ06ckPR/v4MnnnjCppYePXoUOI3TpEkTffXVVxo9erTWr1+vCxcuOLR/PXv2lJubm7y9vfXAAw8oNzdXn376qRo0aKA1a9bo8uXLioqKstkvT09PtWzZstBZa926dbNru468Hg4ePKhevXopMDDQ+py1bNlSkgr8bi0Wizp16mS97+rqqpo1a6pKlSpq1KiRtd3Pz0+VK1cu9PVw5fPerFkzBQcHW38vhfniiy9ksVj05JNP2jxXgYGBatiw4U3N8IN53Nbh5ty5c2rYsKHmzJlzQ8sfPHhQjzzyiNq0aaPk5GStWbNGJ0+e1GOPPVbEleJGhIeH68UXX9Qnn3yiY8eOacSIEUpNTb2pQZR+fn42993d3a/afvHixWuuK/90RmEzl6pWrXrDpzuut82rbe+vNV1LTEyMxo4dq65du2rlypXaunWrtm3bpoYNGzr8gf9Xw4YN07Zt2/Ttt9/q9ddf16VLl/TII4/Y1HT8+HGtXLlSbm5uNre//e1vkmT9oP7hhx/UoUMHSdK///1vbd68Wdu2bbMO3M2vM3/dgYGBNrW4urrK39/fpu3NN9/Uiy++qOXLl6t169by8/NT165dtX//frv2b9q0adq2bZt27NihtLQ0HTx4UF27drXulyTde++9BfYtISGhQCD39vaWj4+PXdvNd73Xw9mzZ9WiRQtt3bpVkyZN0vr167Vt2zYtXbpUkgr8br29veXp6WnT5u7uXuC1kN9e2Ovhyuc9v+1af4fHjx+XYRgKCAgo8Fx9//33N/XPC8zjtp4tFRkZaf2PrzA5OTl6+eWXtWjRIv3xxx+qX7++pk2bZp0au2PHDuXm5mrSpEnWgXmjRo3SI488okuXLsnNza0kdgN2cHNz0/jx4/XGG29YZx5Jf84wys7OLtC/OILFlfI/PNPT03XnnXfaPHbs2DFVqlSpWLaZnp5eoP3YsWOSZNc2//Of/ygqKkpTpkyxaT958mSh19+x15133mkdRNy8eXMFBgbqySef1Pjx463/gFSqVEkNGjTQ5MmTC11Hfkj7+OOP5ebmpi+++MLmA/jKo7P5v4OMjAyb6diXL18u8DdQtmxZxcXFKS4uTsePH7cexenSpYt+/fXX6+5fjRo1rPt3pfzn/dNPP1VwcPB112WxWK7b51oKez2sW7dOx44d0/r1661HayRdd/zLzcjIyCi0rWbNmlddplKlSrJYLNq0aVOhl0QorA23n9v6yM319O3bV5s3b9bHH3+sn376Sd27d9eDDz5o/U8tPDxcLi4uev/995Wbm6vMzEx9+OGH6tChA8HGiQr78Jb+77B6/geg9OdMpJ9++smm37p163T27NniK/D/a9OmjaQ/w8Jfbdu2TXv27LnmDJgb1bZtW+3evVs7duywaV+4cKEsFotat24t6f8+IAo7EmOxWAp8gHz55Zc6evRokdb6xBNPqFWrVvr3v/9tPaXx0EMP6eeff9Zdd92l8PDwArf8363FYpGrq6tcXFys67tw4YI+/PBDm23k/6Ny5XWAlixZUmD21V8FBASoT58++uc//6m9e/fq/PnzN7WvHTt2lKurq3777bdC9+tqocge9r4e8gPTlb/bd99994a3fT1XPu/fffedDh06dM1rKz300EMyDENHjx4t9Hm6++67i61elB639ZGba/ntt9+0ePFiHTlyxPriHzVqlFavXq33339fU6ZMUUhIiNauXavu3btr0KBBys3NVUREhFatWuXk6m9vHTt21J133qkuXbqobt26ysvLU3JysmbMmKFy5cpp2LBh1r5PPfWUxo4dq3Hjxqlly5bavXu35syZI19f32Kvs06dOnr66af11ltvqUyZMoqMjFRqaqrGjh2roKAgjRgxosi3OWLECC1cuFCdO3fWxIkTFRwcrC+//FJz587VkCFDrDOTypcvr+DgYH3++edq27at/Pz8VKlSJYWEhOihhx5SfHy86tatqwYNGmj79u167bXXChx9KgrTpk1T06ZN9corr+i9997TxIkTlZiYqGbNmik6Olp16tTRxYsXlZqaqlWrVumdd97RnXfeqc6dO2vmzJnq1auXnn76aZ06dUqvv/56gQ/usLAwPfnkk5o1a5bc3NzUrl07/fzzz3r99dcLnPZp2rSpHnroITVo0EAVK1bUnj179OGHHyoiIuKq15ixV0hIiCZOnKgxY8bo4MGDevDBB1WxYkUdP35cP/zwg/Wo0Y2w9/XQrFkzVaxYUYMHD9b48ePl5uamRYsW6ccff7ypfbuWpKQkDRgwQN27d9fhw4c1ZswYVatWTc8888xVl2nevLmefvpp9e3bV0lJSXrggQdUtmxZpaen69tvv9Xdd9+tIUOGFFvNKCWcPKD5liHJWLZsmfX+kiVLDElG2bJlbW6urq7WmQvp6elGrVq1jOeff97YsWOHsWHDBqNly5ZG27ZtrbMxUPISEhKMXr16GbVq1TLKlStnuLm5GdWrVzeeeuopY/fu3TZ9s7OzjRdeeMEICgoyvLy8jJYtWxrJyclXnS21bds2m+WvNoupd+/eRtmyZa9ba25urjFt2jSjdu3ahpubm1GpUiXjySefNA4fPmzXdq7nytlShmEYhw4dMnr16mX4+/sbbm5uRp06dYzXXnvNyM3Nten39ddfG40aNTI8PDxsZg7973//M/r3729UrlzZ8Pb2Nu6//35j06ZNRsuWLY2WLVtal3d0ttRrr71W6OPdu3c3XF1djQMHDhiG8edMrujoaCM0NNRwc3Mz/Pz8jMaNGxtjxowxzp49a11uwYIFRp06dQwPDw+jRo0axtSpU4358+cbkmxmYGVnZxsjR440KleubHh6ehr33XefsWXLlgJ/A6NHjzbCw8ONihUrWtc5YsQI4+TJk9fcv/zZUp988sk1+xmGYSxfvtxo3bq14ePjY3h4eBjBwcHG448/bnz99dfWPvb+beVz5PXw3XffGREREYa3t7dxxx13GAMGDDB27NhR4Pd4tRoK+3szjD9nH3bu3Nl6P//1tHbtWuOpp54yKlSoYHh5eRmdOnUy9u/fb7PslbOl8i1YsMBo2rSpUbZsWcPLy8u46667jKioKCMpKcnu5wbmZTGMK65odZuyWCxatmyZdYBfQkKCnnjiCf3yyy82h7YlqVy5cgoMDNTYsWP11VdfKSkpyfrYkSNHFBQUpC1btpSKq8ICQEmLj49X3759tW3btps65QZcDaelrqJRo0bKzc3ViRMn1KJFi0L7nD9/vkDwyb+ff7VXAABQsm7rAcVnz55VcnKykpOTJf15RdTk5GSlpaWpdu3aeuKJJxQVFaWlS5cqJSVF27Zt07Rp06xjajp37qxt27Zp4sSJ2r9/v3bs2KG+ffsqODjY5joPAACg5NzWp6XWr19vnR3yV71791Z8fLwuXbqkSZMmaeHChTp69Kj8/f0VERGhuLg464j8jz/+WNOnT9e+ffvk7e2tiIgITZs2TXXr1i3p3QEAALrNww0AADCf2/q0FAAAMB/CDQAAMJXbbrZUXl6ejh07pvLly9/0JcwBAEDJMAxDZ86cUdWqVa1feXQ1t124OXbsmIKCgpxdBgAAuAGHDx++7hXRb7twU758eUl/PjmOfqsuAABwjqysLAUFBVk/x6/ltgs3+aeifHx8CDcAAJQy9gwpYUAxAAAwFcINAAAwFcINAAAwldtuzI09DMPQ5cuXlZub6+xSYAcXFxe5uroytR8AIIlwU0BOTo7S09N1/vx5Z5cCB3h7e6tKlSpyd3d3dikAACcj3PxFXl6eUlJS5OLioqpVq8rd3Z2jAbc4wzCUk5Oj33//XSkpKapVq9Z1L+4EADA3ws1f5OTkKC8vT0FBQfL29nZ2ObCTl5eX3NzcdOjQIeXk5MjT09PZJQEAnIh/cQvBf/6lD78zAEA+PhEAAICpEG4AAICpMObGTiGjvyzR7aW+2rlkt5eaqtDQUO3cuVP33HNPiW4bAICixJEbkzl8+LD69+9vne0VHBysYcOG6dSpU9dcLigoSOnp6apfv34JVQoAQPEg3JjIwYMHFR4ern379mnx4sU6cOCA3nnnHX3zzTeKiIjQ6dOnC10uJydHLi4uCgwMlKsrB/MAAKUb4cZEhg4dKnd3d61du1YtW7ZU9erVFRkZqa+//lpHjx7VmDFjJEkhISGaNGmS+vTpI19fXw0cOFCpqamyWCxKTk62rm/FihWqVauWvLy81Lp1a33wwQeyWCz6448/JEmHDh1Sly5dVLFiRZUtW1Z/+9vftGrVKifsOQAA/4d/003i9OnTWrNmjSZPniwvLy+bxwIDA/XEE08oISFBc+fOlSS99tprGjt2rF5++eVC15eamqrHH39cw4YN04ABA7Rz506NGjXKps/QoUOVk5OjjRs3qmzZstq9e7fKlStXPDsImERJj99DyY9hhPMRbkxi//79MgxDYWFhhT4eFham//3vf/r9998lSW3atLEJK6mpqTb933nnHdWpU0evvfaaJKlOnTr6+eefNXnyZGuftLQ0devWTXfffbckqUaNGkW5SwAA3BBOS90mDMOQJOvXSYSHh1+z/969e3XvvffatDVp0sTmfnR0tCZNmqTmzZtr/Pjx+umnn4qwYgAAbgzhxiRq1qwpi8Wi3bt3F/r4r7/+qooVK6pSpUqSpLJly15zfYZhFPherfyAlG/AgAE6ePCgnnrqKe3atUvh4eF66623bmIvAAC4eYQbk/D391f79u01d+5cXbhwweaxjIwMLVq0SD179rT7i0Dr1q2rbdu22bQlJSUV6BcUFKTBgwdr6dKlGjlypP7973/f+E4AAFAECDcmMmfOHGVnZ6tjx47auHGjDh8+rNWrV6t9+/aqVq2azXiZ6xk0aJB+/fVXvfjii9q3b5+WLFmi+Ph4Sf93amv48OFas2aNUlJStGPHDq1bt+6qY34AACgpDCi2U2kYbV+rVi0lJSVpwoQJ6tmzp06dOqXAwEB17dpV48ePl5+fn93rCg0N1aeffqqRI0dq9uzZioiI0JgxYzRkyBB5eHhIknJzczV06FAdOXJEPj4+evDBB/XGG28U1+4BAGAXwo3JBAcH6/33379mnytnRkl/XvvmyjE1Dz/8sB5++GHr/cmTJ+vOO++Up6enJDG+BgBwSyLc4Krmzp2re++9V/7+/tq8ebNee+01Pfvss84uCwCAayLc4Kr279+vSZMm6fTp06pevbpGjhyp2NhYZ5cFAMA1OXVA8caNG9WlSxdVrVpVFotFy5cvv+4y2dnZGjNmjIKDg+Xh4aG77rpLCxYsKP5ib0NvvPGGjh07posXL2rfvn0aO3Ys3z0FALjlOfWT6ty5c2rYsKH69u2rbt262bVMjx49dPz4cc2fP181a9bUiRMndPny5WKuFAAAlBZODTeRkZGKjIy0u//q1au1YcMGHTx40DrzJyQkpJiqAwAApVGpus7NihUrFB4erunTp6tatWqqXbu2Ro0aVeCidX+VnZ2trKwsmxsAADCvUjWA4uDBg/r222/l6empZcuW6eTJk3rmmWd0+vTpq467mTp1quLi4kq4UgAA4Cyl6shNXl6eLBaLFi1apCZNmqhTp06aOXOm4uPjr3r0JjY2VpmZmdbb4cOHS7hqAABQkkrVkZsqVaqoWrVq8vX1tbaFhYXJMAwdOXJEtWrVKrCMh4eH9Yq6AADA/EpVuGnevLk++eQTnT17VuXKlZMk7du3T2XKlNGdd95ZvBuf4Hv9PkW6vcxiXX18fLyGDx+uP/74o1i3U5QsFouWLVumrl27OrsUAMAtzKmnpc6ePavk5GQlJydLklJSUpScnKy0tDRJf55SioqKsvbv1auX/P391bdvX+3evVsbN27U888/r379+snLy8sZu3BLsFgs17z16dOnwDI9e/bUvn37Sr5YAACKmVOP3CQlJal169bW+zExMZKk3r17Kz4+Xunp6dagI0nlypVTYmKinnvuOYWHh8vf3189evTQpEmTSrz2W0l6err154SEBI0bN0579+61tl0Z/C5duiQvL6/bOhACAMzLqUduWrVqJcMwCtzi4+Ml/XnqZP369TbL1K1bV4mJiTp//rwOHz6sGTNm3PYf0oGBgdabr6+vLBaL9f7FixdVoUIFLVmyRK1atZKnp6f+85//KD4+XhUqVLCuY8KECbrnnnu0YMECVa9eXeXKldOQIUOUm5ur6dOnKzAwUJUrV9bkyZNttm2xWPTuu+/qoYcekre3t8LCwrRlyxYdOHBArVq1UtmyZRUREaHffvvNZrmVK1eqcePG8vT0VI0aNRQXF2dzMcb9+/frgQcekKenp+rVq6fExMRifQ4BAOZRqmZL4ca9+OKLio6O1p49e9SxY8dC+/z222/66quvtHr1ai1evFgLFixQ586ddeTIEW3YsEHTpk3Tyy+/rO+//95muVdeeUVRUVFKTk5W3bp11atXLw0aNEixsbFKSkqSJJsv3FyzZo2efPJJRUdHa/fu3Xr33XcVHx9vDU55eXl67LHH5OLiou+//17vvPOOXnzxxWJ6ZgAAZlOqBhTjxg0fPlyPPfbYNfvk5eVpwYIFKl++vOrVq6fWrVtr7969WrVqlcqUKaM6depo2rRpWr9+ve677z7rcn379lWPHj0k/RmiIiIiNHbsWGuIGjZsmPr27WvtP3nyZI0ePVq9e/eWJNWoUUOvvPKKXnjhBY0fP15ff/219uzZo9TUVOtA8SlTpjh0NWsAwO2LcHObCA8Pv26fkJAQlS9f3no/ICBALi4uKlOmjE3biRMnbJZr0KCBzeOSdPfdd9u0Xbx4UVlZWfLx8dH27du1bds2m1Ncubm5unjxos6fP689e/aoevXqNjPgIiIiHNhbAMDtjHBzmyhbtux1+7i5udnct1gshbbl5eVddTmLxXLVtvzl8vLyFBcXV+iRJE9PTxmGUaA9fx0AAFwP4QYl7u9//7v27t2rmjVrFvp4vXr1lJaWpmPHjqlq1aqSpC1btpRkiQCAUoxwgxI3btw4PfTQQwoKClL37t1VpkwZ/fTTT9q1a5cmTZqkdu3aqU6dOoqKitKMGTOUlZWlMWPGOLtsAEApQbixVzFfMfh20rFjR33xxReaOHGipk+fLjc3N9WtW1cDBgyQJJUpU0bLli1T//791aRJE4WEhOjNN9/Ugw8+6OTKAQClgcUobICDiWVlZcnX11eZmZny8fGxeezixYtKSUlRaGioPD09nVQhbgS/O5QWIaO/dHYJt53UVzs7uwQUgWt9fl+J69wAAABTIdwAAABTIdwAAABTIdwAAABTIdwU4jYbY20K/M4AAPkIN3+Rf1Xd8+fPO7kSOCr/d3blFZUBALcfrnPzFy4uLqpQoYL1u5O8vb257P8tzjAMnT9/XidOnFCFChXk4uLi7JIAAE5GuLlCYGCgJBX4ckjc2ipUqGD93QEAbm+EmytYLBZVqVJFlStX1qVLl5xdDuzg5ubGERsAgBXh5ipcXFz4wAQAoBRiQDEAADAVwg0AADAVwg0AADAVwg0AADAVwg0AADAVwg0AADAVwg0AADAVwg0AADAVwg0AADAVwg0AADAVwg0AADAVwg0AADAVwg0AADAVwg0AADAVp4abjRs3qkuXLqpataosFouWL19u97KbN2+Wq6ur7rnnnmKrDwAAlD5ODTfnzp1Tw4YNNWfOHIeWy8zMVFRUlNq2bVtMlQEAgNLK1Zkbj4yMVGRkpMPLDRo0SL169ZKLi8t1j/ZkZ2crOzvbej8rK8vh7QEAgNKj1I25ef/99/Xbb79p/PjxdvWfOnWqfH19rbegoKBirhAAADhTqQo3+/fv1+jRo7Vo0SK5utp30Ck2NlaZmZnW2+HDh4u5SgAA4ExOPS3liNzcXPXq1UtxcXGqXbu23ct5eHjIw8OjGCsDAAC3klITbs6cOaOkpCTt3LlTzz77rCQpLy9PhmHI1dVVa9euVZs2bZxcJQAAcLZSE258fHy0a9cum7a5c+dq3bp1+vTTTxUaGuqkygAAwK3EqeHm7NmzOnDggPV+SkqKkpOT5efnp+rVqys2NlZHjx7VwoULVaZMGdWvX99m+cqVK8vT07NAOwAAuH05NdwkJSWpdevW1vsxMTGSpN69eys+Pl7p6elKS0tzVnkAAKAUshiGYTi7iJKUlZUlX19fZWZmysfHx9nlALjNhIz+0tkl3HZSX+3s7BJQBBz5/C5VU8EBAACuh3ADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMhXADAABMxanhZuPGjerSpYuqVq0qi8Wi5cuXX7P/0qVL1b59e91xxx3y8fFRRESE1qxZUzLFAgCAUsGp4ebcuXNq2LCh5syZY1f/jRs3qn379lq1apW2b9+u1q1bq0uXLtq5c2cxVwoAAEoLV2duPDIyUpGRkXb3nzVrls39KVOm6PPPP9fKlSvVqFGjIq4OAACURk4NNzcrLy9PZ86ckZ+f31X7ZGdnKzs723o/KyurJEoDAABOUqoHFM+YMUPnzp1Tjx49rtpn6tSp8vX1td6CgoJKsEIAAFDSSm24Wbx4sSZMmKCEhARVrlz5qv1iY2OVmZlpvR0+fLgEqwQAACWtVJ6WSkhIUP/+/fXJJ5+oXbt21+zr4eEhDw+PEqoMAAA4W6k7crN48WL16dNHH330kTp37uzscgAAwC3GqUduzp49qwMHDljvp6SkKDk5WX5+fqpevbpiY2N19OhRLVy4UNKfwSYqKkqzZ8/Wfffdp4yMDEmSl5eXfH19nbIPAADg1uLwkZsPPvhAX375pfX+Cy+8oAoVKqhZs2Y6dOiQQ+tKSkpSo0aNrNO4Y2Ji1KhRI40bN06SlJ6errS0NGv/d999V5cvX9bQoUNVpUoV623YsGGO7gYAADApi2EYhiML1KlTR/PmzVObNm20ZcsWtW3bVrNmzdIXX3whV1dXLV26tLhqLRJZWVny9fVVZmamfHx8nF0OgNtMyOgvr98JRSr1VYYwmIEjn98On5Y6fPiwatasKUlavny5Hn/8cT399NNq3ry5WrVqdUMFAwAAFBWHT0uVK1dOp06dkiStXbvWOlvJ09NTFy5cKNrqAAAAHOTwkZv27dtrwIABatSokfbt22edsfTLL78oJCSkqOsDAABwiMNHbt5++21FRETo999/12effSZ/f39J0vbt2/XPf/6zyAsEAABwhMNHbipUqFDot3jHxcUVSUEAAAA344auc/PHH3/ohx9+0IkTJ5SXl2dtt1gseuqpp4qsuNKImRAlj5kQAIC/cjjcrFy5Uk888YTOnTun8uXLy2KxWB8j3AAAAGdzeMzNyJEj1a9fP505c0Z//PGH/ve//1lvp0+fLo4aAQAA7OZwuDl69Kiio6Pl7e1dHPUAAADcFIfDTceOHZWUlFQctQAAANw0u8bcrFixwvpz586d9fzzz2v37t26++675ebmZtP34YcfLtoKAQAAHGBXuOnatWuBtokTJxZos1gsys3NvemiAAAAbpRd4eav070BAABuZQ6PuVm4cKGys7MLtOfk5GjhwoVFUhQAAMCNcjjc9O3bV5mZmQXaz5w5o759+xZJUQAAADfK4XBjGIbNhfvyHTlyRL6+vkVSFAAAwI2y+wrFjRo1ksVikcViUdu2beXq+n+L5ubmKiUlRQ8++GCxFAkAAGAvu8NN/oyp5ORkdezYUeXKlbM+5u7urpCQEHXr1q3ICwQAAHCE3eFm/PjxkqSQkBD17NlTnp6exVYUAADAjXL4izN79+5dHHUAAAAUCYfDTcWKFQsdUGyxWOTp6amaNWuqT58+zJwCAABO4XC4GTdunCZPnqzIyEg1adJEhmFo27ZtWr16tYYOHaqUlBQNGTJEly9f1sCBA4ujZgAAgKtyONx8++23mjRpkgYPHmzT/u6772rt2rX67LPP1KBBA7355puEGwAAUOIcvs7NmjVr1K5duwLtbdu21Zo1ayRJnTp10sGDB2++OgAAAAc5HG78/Py0cuXKAu0rV66Un5+fJOncuXMqX778zVcHAADgIIdPS40dO1ZDhgzRf//7XzVp0kQWi0U//PCDVq1apXfeeUeSlJiYqJYtWxZ5sQAAANfjcLgZOHCg6tWrpzlz5mjp0qUyDEN169bVhg0b1KxZM0nSyJEji7xQAAAAezgcbiSpefPmat68eVHXAgAAcNNuKNzk5eXpwIEDOnHihPLy8mwee+CBB4qkMAAAgBvhcLj5/vvv1atXLx06dEiGYdg8ZrFYlJubW2TFAQAAOMrhcDN48GCFh4fryy+/VJUqVQq9WjEAAICzODwVfP/+/ZoyZYrCwsJUoUIF+fr62twcsXHjRnXp0kVVq1aVxWLR8uXLr7vMhg0b1LhxY3l6eqpGjRrWGVoAAADSDYSbpk2b6sCBA0Wy8XPnzqlhw4aaM2eOXf1TUlLUqVMntWjRQjt37tRLL72k6OhoffbZZ0VSDwAAKP0cPi313HPPaeTIkcrIyNDdd98tNzc3m8cbNGhg97oiIyMVGRlpd/933nlH1atX16xZsyRJYWFhSkpK0uuvv65u3brZvR4AAGBeDoeb/BDRr18/a5vFYpFhGMU+oHjLli3q0KGDTVvHjh01f/58Xbp0qUDQkqTs7GxlZ2db72dlZRVbfQAAwPkcDjcpKSnFUYddMjIyFBAQYNMWEBCgy5cv6+TJk6pSpUqBZaZOnaq4uLiSKhEAADiZw+EmODi4OOqw25Wzs/Kno19t1lZsbKxiYmKs97OyshQUFFR8BQIAAKdyeECxJH344Ydq3ry5qlatqkOHDkmSZs2apc8//7xIi7tSYGCgMjIybNpOnDghV1dX+fv7F7qMh4eHfHx8bG4AAMC8HA438+bNU0xMjDp16qQ//vjDOsamQoUK1oG+xSUiIkKJiYk2bWvXrlV4eHih420AAMDtx+Fw89Zbb+nf//63xowZIxcXF2t7eHi4du3a5dC6zp49q+TkZCUnJ0v6czxPcnKy0tLSJP15SikqKsraf/DgwTp06JBiYmK0Z88eLViwQPPnz9eoUaMc3Q0AAGBSNzSguFGjRgXaPTw8dO7cOYfWlZSUpNatW1vv54+N6d27t+Lj45Wenm4NOpIUGhqqVatWacSIEXr77bdVtWpVvfnmm0wDBwAAVg6Hm9DQUCUnJxcYWPzVV1+pXr16Dq2rVatWBb6f6q/i4+MLtLVs2VI7duxwaDsAAOD24XC4ef755zV06FBdvHhRhmHohx9+0OLFizV16lS99957xVEjAACA3RwON3379tXly5f1wgsv6Pz58+rVq5eqVaum2bNn6x//+Edx1AgAAGC3G5oKPnDgQB06dEgnTpxQRkaGDh8+rH/84x/auHFjUdcHAADgEIeP3PxVpUqVrD8fOHBArVu3LtavXwAAALieGzpyAwAAcKsi3AAAAFMh3AAAAFOxe8zNihUrrvm4M78tHAAAIJ/d4aZr167X7XO1b+YGAAAoKXaHm7y8vOKsAwAAoEgw5gYAAJgK4QYAAJgK4QYAAJgK4QYAAJgK4QYAAJgK4QYAAJiKXVPBK1asaPc1bE6fPn1TBQEAANwMu8LNrFmzrD+fOnVKkyZNUseOHRURESFJ2rJli9asWaOxY8cWS5EAAAD2sivc9O7d2/pzt27dNHHiRD377LPWtujoaM2ZM0dff/21RowYUfRVAgAA2MnhMTdr1qzRgw8+WKC9Y8eO+vrrr4ukKAAAgBvlcLjx9/fXsmXLCrQvX75c/v7+RVIUAADAjbL7u6XyxcXFqX///lq/fr11zM3333+v1atX67333ivyAgEAABzhcLjp06ePwsLC9Oabb2rp0qUyDEP16tXT5s2b1bRp0+KoEQAAwG4OhxtJatq0qRYtWlTUtQAAANy0Gwo3+S5cuKBLly7ZtPn4+NxUQQAAADfD4QHF58+f17PPPqvKlSurXLlyqlixos0NAADAmRwON88//7zWrVunuXPnysPDQ++9957i4uJUtWpVLVy4sDhqBAAAsJvDp6VWrlyphQsXqlWrVurXr59atGihmjVrKjg4WIsWLdITTzxRHHUCAADYxeEjN6dPn1ZoaKikP8fX5H+X1P3336+NGzcWbXUAAAAOcjjc1KhRQ6mpqZKkevXqacmSJZL+PKJToUKFoqwNAADAYQ6Hm759++rHH3+UJMXGxlrH3owYMULPP/98kRcIAADgCIfDzYgRIxQdHS1Jat26tX799VctXrxYO3bs0LBhwxwuYO7cuQoNDZWnp6caN26sTZs2XbP/okWL1LBhQ3l7e6tKlSrq27evTp065fB2AQCAOTkcbv7q4sWLql69uh577DE1bNjQ4eUTEhI0fPhwjRkzRjt37lSLFi0UGRmptLS0Qvt/++23ioqKUv/+/fXLL7/ok08+0bZt2zRgwICb2Q0AAGAiDoeb3NxcvfLKK6pWrZrKlSungwcPSpLGjh2r+fPnO7SumTNnqn///howYIDCwsI0a9YsBQUFad68eYX2//777xUSEqLo6GiFhobq/vvv16BBg5SUlOTobgAAAJNyONxMnjxZ8fHxmj59utzd3a3td999t0NfnJmTk6Pt27erQ4cONu0dOnTQd999V+gyzZo105EjR7Rq1SoZhqHjx4/r008/VefOna+6nezsbGVlZdncAACAeTkcbhYuXKh//etfeuKJJ+Ti4mJtb9CggX799Ve713Py5Enl5uYqICDApj0gIEAZGRmFLtOsWTMtWrRIPXv2lLu7uwIDA1WhQgW99dZbV93O1KlT5evra70FBQXZXSMAACh9HA43R48eVc2aNQu05+XlFfieKXtYLBab+4ZhFGjLt3v3bkVHR2vcuHHavn27Vq9erZSUFA0ePPiq64+NjVVmZqb1dvjwYYdrBAAApYfDVyj+29/+pk2bNik4ONim/ZNPPlGjRo3sXk+lSpXk4uJS4CjNiRMnChzNyTd16lQ1b97cOuW8QYMGKlu2rFq0aKFJkyapSpUqBZbx8PCQh4eH3XUBAIDSzeFwM378eD311FM6evSo8vLytHTpUu3du1cLFy7UF198Yfd63N3d1bhxYyUmJurRRx+1ticmJuqRRx4pdJnz58/L1dW25PxTY4ZhOLorAADAhBw+LdWlSxclJCRo1apVslgsGjdunPbs2aOVK1eqffv2Dq0rJiZG7733nhYsWKA9e/ZoxIgRSktLs55mio2NVVRUlM22ly5dqnnz5ungwYPavHmzoqOj1aRJE1WtWtXRXQEAACbk8JEbSerYsaM6duxYoH3btm2699577V5Pz549derUKU2cOFHp6emqX7++Vq1aZT3llZ6ebnPNmz59+ujMmTOaM2eORo4cqQoVKqhNmzaaNm3ajewGAAAwIYvh4Pmcs2fPysXFRV5eXta25ORkjR07VqtWrVJubm6RF1mUsrKy5Ovrq8zMTPn4+BT5+kNGf1nk68S1pb569UsBALca3iNKHu8R5uDI57fdp6WOHDmi5s2bW6dUx8TE6Pz584qKitK9994rDw8PffvttzddPAAAwM2w+7TU6NGjdfbsWc2ePVufffaZZs+erQ0bNqhhw4bat2+fQkNDi7NOAAAAu9gdbv773/9qyZIlat68uR5//HFVrVpV3bt31+jRo4uzPgAAAIfYfVoqIyNDd911lyQpMDBQXl5eV52yDQAA4CwOTQX/69ctlClTRp6enkVeEAAAwM2w+7SUYRhq27at9SJ6Fy5cUJcuXWy+PFOSduzYUbQVAgAAOMDucDN+/Hib+5ySAgAAt6IbDjcAAAC3Ioe/fgEAAOBWRrgBAACmckPfLQUAQKkxwdfZFdx+JmQ6dfMcuQEAAKZCuAEAAKZi12mpN9980+4VRkdH33AxAAAAN8uucPPGG2/YtTKLxUK4AQAATmVXuElJSSnuOgAAAIoEY24AAICp3NBU8CNHjmjFihVKS0tTTk6OzWMzZ84sksIAAABuhMPh5ptvvtHDDz+s0NBQ7d27V/Xr11dqaqoMw9Df//734qgRAADAbg6floqNjdXIkSP1888/y9PTU5999pkOHz6sli1bqnv37sVRIwAAgN0cDjd79uxR7969JUmurq66cOGCypUrp4kTJ2ratGlFXiAAAIAjHA43ZcuWVXZ2tiSpatWq+u2336yPnTx5sugqAwAAuAEOj7m57777tHnzZtWrV0+dO3fWyJEjtWvXLi1dulT33XdfcdQIAABgN4fDzcyZM3X27FlJ0oQJE3T27FklJCSoZs2adl/sDwAAoLg4HG5q1Khh/dnb21tz584t0oIAAABuhsNjbmrUqKFTp04VaP/jjz9sgg8AAIAzOBxuUlNTlZubW6A9OztbR48eLZKiAAAAbpTdp6VWrFhh/XnNmjXy9fW13s/NzdU333yjkJCQIi0OAADAUXaHm65du0r685u/869zk8/NzU0hISGaMWNGkRYHAADgKLvDTV5eniQpNDRU27ZtU6VKlYqtKAAAgBvl8GyplJSU4qgDAACgSDg8oFiSNmzYoC5duqhmzZqqVauWHn74YW3atKmoawMAAHCYw+HmP//5j9q1aydvb29FR0fr2WeflZeXl9q2bauPPvrI4QLmzp2r0NBQeXp6qnHjxtcNSdnZ2RozZoyCg4Pl4eGhu+66SwsWLHB4uwAAwJwcPi01efJkTZ8+XSNGjLC2DRs2TDNnztQrr7yiXr162b2uhIQEDR8+XHPnzlXz5s317rvvKjIyUrt371b16tULXaZHjx46fvy45s+fr5o1a+rEiRO6fPmyo7sBAABMyuFwc/DgQXXp0qVA+8MPP6yXXnrJoXXNnDlT/fv314ABAyRJs2bN0po1azRv3jxNnTq1QP/Vq1drw4YNOnjwoPz8/CTputPPs7OzrV/0KUlZWVkO1QgAAEoXh09LBQUF6ZtvvinQ/s033ygoKMju9eTk5Gj79u3q0KGDTXuHDh303XffFbrMihUrFB4erunTp6tatWqqXbu2Ro0apQsXLlx1O1OnTpWvr6/15kiNAACg9LH7yE2/fv00e/ZsjRw5UtHR0UpOTlazZs1ksVj07bffKj4+XrNnz7Z7wydPnlRubq4CAgJs2gMCApSRkVHoMgcPHtS3334rT09PLVu2TCdPntQzzzyj06dPX3XcTWxsrGJiYqz3s7KyCDgAAJiY3eHmgw8+0KuvvqohQ4YoMDBQM2bM0JIlSyRJYWFhSkhI0COPPOJwARaLxea+YRgF2vLl5eXJYrFo0aJF1iskz5w5U48//rjefvtteXl5FVjGw8NDHh4eDtcFAABKJ7vDjWEY1p8fffRRPfrooze14UqVKsnFxaXAUZoTJ04UOJqTr0qVKqpWrZrNVz+EhYXJMAwdOXJEtWrVuqmaAABA6efQmJurHVG5Ee7u7mrcuLESExNt2hMTE9WsWbNCl2nevLmOHTums2fPWtv27dunMmXK6M477yyy2gAAQOnlULipXbu2/Pz8rnlzRExMjN577z0tWLBAe/bs0YgRI5SWlqbBgwdL+nO8TFRUlLV/r1695O/vr759+2r37t3auHGjnn/+efXr16/QU1IAAOD249BU8Li4OJtTQjerZ8+eOnXqlCZOnKj09HTVr19fq1atUnBwsCQpPT1daWlp1v7lypVTYmKinnvuOYWHh8vf3189evTQpEmTiqwmAABQulmMvw6muYYyZcooIyNDlStXLu6ailVWVpZ8fX2VmZkpHx+fIl9/yOgvi3yduLbUVzs7uwTAbrxHlLxUT/svLosiMiGzyFfpyOe33aelinK8DQAAQHGxO9zYeYAHAADAqewec5OXl1ecdQAAABQJh79+AQAA4FZGuAEAAKbi8LeCA7ecCUV3eQLYqRhmQgBAUeHIDQAAMBXCDQAAMBXCDQAAMBXCDQAAMBXCDQAAMBXCDQAAMBXCDQAAMBXCDQAAMBXCDQAAMBXCDQAAMBXCDQAAMBXCDQAAMBXCDQAAMBXCDQAAMBXCDQAAMBXCDQAAMBXCDQAAMBXCDQAAMBXCDQAAMBXCDQAAMBXCDQAAMBXCDQAAMBXCDQAAMBXCDQAAMBXCDQAAMBWnh5u5c+cqNDRUnp6eaty4sTZt2mTXcps3b5arq6vuueee4i0QAACUKk4NNwkJCRo+fLjGjBmjnTt3qkWLFoqMjFRaWto1l8vMzFRUVJTatm1bQpUCAIDSwqnhZubMmerfv78GDBigsLAwzZo1S0FBQZo3b941lxs0aJB69eqliIiIEqoUAACUFk4LNzk5Odq+fbs6dOhg096hQwd99913V13u/fff12+//abx48fbtZ3s7GxlZWXZ3AAAgHk5LdycPHlSubm5CggIsGkPCAhQRkZGocvs379fo0eP1qJFi+Tq6mrXdqZOnSpfX1/rLSgo6KZrBwAAty6nDyi2WCw29w3DKNAmSbm5uerVq5fi4uJUu3Ztu9cfGxurzMxM6+3w4cM3XTMAALh12Xf4oxhUqlRJLi4uBY7SnDhxosDRHEk6c+aMkpKStHPnTj377LOSpLy8PBmGIVdXV61du1Zt2rQpsJyHh4c8PDyKZycAAMAtx2lHbtzd3dW4cWMlJibatCcmJqpZs2YF+vv4+GjXrl1KTk623gYPHqw6deooOTlZTZs2LanSAQDALcxpR24kKSYmRk899ZTCw8MVERGhf/3rX0pLS9PgwYMl/XlK6ejRo1q4cKHKlCmj+vXr2yxfuXJleXp6FmgHAAC3L6eGm549e+rUqVOaOHGi0tPTVb9+fa1atUrBwcGSpPT09Ote8wYAAOCvLIZhGM4uoiRlZWXJ19dXmZmZ8vHxKfL1h4z+ssjXiWtL9ezl7BJuPxMynV1BqcV7RMnjPcIJiuE9wpHPb6fPlgIAAChKhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqhBsAAGAqTg83c+fOVWhoqDw9PdW4cWNt2rTpqn2XLl2q9u3b64477pCPj48iIiK0Zs2aEqwWAADc6pwabhISEjR8+HCNGTNGO3fuVIsWLRQZGam0tLRC+2/cuFHt27fXqlWrtH37drVu3VpdunTRzp07S7hyAABwq3JquJk5c6b69++vAQMGKCwsTLNmzVJQUJDmzZtXaP9Zs2bphRde0L333qtatWppypQpqlWrllauXFnClQMAgFuV08JNTk6Otm/frg4dOti0d+jQQd99951d68jLy9OZM2fk5+d31T7Z2dnKysqyuQEAAPNyWrg5efKkcnNzFRAQYNMeEBCgjIwMu9YxY8YMnTt3Tj169Lhqn6lTp8rX19d6CwoKuqm6AQDArc3pA4otFovNfcMwCrQVZvHixZowYYISEhJUuXLlq/aLjY1VZmam9Xb48OGbrhkAANy6XJ214UqVKsnFxaXAUZoTJ04UOJpzpYSEBPXv31+ffPKJ2rVrd82+Hh4e8vDwuOl6AQBA6eC0Izfu7u5q3LixEhMTbdoTExPVrFmzqy63ePFi9enTRx999JE6d+5c3GUCAIBSxmlHbiQpJiZGTz31lMLDwxUREaF//etfSktL0+DBgyX9eUrp6NGjWrhwoaQ/g01UVJRmz56t++67z3rUx8vLS76+vk7bDwAAcOtwarjp2bOnTp06pYkTJyo9PV3169fXqlWrFBwcLElKT0+3uebNu+++q8uXL2vo0KEaOnSotb13796Kj48v6fIBAMAtyKnhRpKeeeYZPfPMM4U+dmVgWb9+ffEXBAAASjWnz5YCAAAoSoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQbAABgKk4PN3PnzlVoaKg8PT3VuHFjbdq06Zr9N2zYoMaNG8vT01M1atTQO++8U0KVAgCA0sCp4SYhIUHDhw/XmDFjtHPnTrVo0UKRkZFKS0srtH9KSoo6deqkFi1aaOfOnXrppZcUHR2tzz77rIQrBwAAtyqnhpuZM2eqf//+GjBggMLCwjRr1iwFBQVp3rx5hfZ/5513VL16dc2aNUthYWEaMGCA+vXrp9dff72EKwcAALcqV2dtOCcnR9u3b9fo0aNt2jt06KDvvvuu0GW2bNmiDh062LR17NhR8+fP16VLl+Tm5lZgmezsbGVnZ1vvZ2ZmSpKysrJudhcKlZd9vljWi6vLshjOLuH2U0yvn9sB7xElj/cIJyiG94j8z23DuP7v02nh5uTJk8rNzVVAQIBNe0BAgDIyMgpdJiMjo9D+ly9f1smTJ1WlSpUCy0ydOlVxcXEF2oOCgm6ietxKfJ1dwO3oVZ51lB78tTpBMb5HnDlzRr6+116/08JNPovFYnPfMIwCbdfrX1h7vtjYWMXExFjv5+Xl6fTp0/L397/mdlA6ZGVlKSgoSIcPH5aPj4+zywFwi+E9wjwMw9CZM2dUtWrV6/Z1WripVKmSXFxcChylOXHiRIGjM/kCAwML7e/q6ip/f/9Cl/Hw8JCHh4dNW4UKFW68cNySfHx8eOMCcFW8R5jD9Y7Y5HPagGJ3d3c1btxYiYmJNu2JiYlq1qxZoctEREQU6L927VqFh4cXOt4GAADcfpw6WyomJkbvvfeeFixYoD179mjEiBFKS0vT4MGDJf15SikqKsraf/DgwTp06JBiYmK0Z88eLViwQPPnz9eoUaOctQsAAOAW49QxNz179tSpU6c0ceJEpaenq379+lq1apWCg4MlSenp6TbXvAkNDdWqVas0YsQIvf3226patarefPNNdevWzVm7ACfz8PDQ+PHjC5x6BACJ94jblcWwZ04VAABAKeH0r18AAAAoSoQbAABgKoQbAABgKoQbAABgKoQbAABgKoQblBqtWrXS8OHDbdqGDRumxo0by8PDQ/fcc49T6gJwa7jyPeLHH3/UP//5TwUFBcnLy0thYWGaPXu28wpEiXH6d0sBN8MwDPXr109bt27VTz/95OxyANxCtm/frjvuuEP/+c9/FBQUpO+++05PP/20XFxc9Oyzzzq7PBQjrnODUqFPnz764IMPbNpSUlIUEhIiSZowYYKWL1+u5OTkki8OgNNd7z0i39ChQ7Vnzx6tW7euBKtDSePIDUqF2bNna9++fapfv74mTpwoSbrjjjucXBWAW4W97xGZmZny8/Mr6fJQwgg3KBV8fX3l7u4ub29vBQYGOrscALcYe94jtmzZoiVLlujLL78s4epQ0hhQDAAwvV9++UWPPPKIxo0bp/bt2zu7HBQzwg0AwNR2796tNm3aaODAgXr55ZedXQ5KAOEGpYa7u7tyc3OdXQaAW1Rh7xG//PKLWrdurd69e2vy5MlOqgwljTE3KDVCQkK0detWpaamqly5cvLz89PBgwd19uxZZWRk6MKFC9bZUvXq1ZO7u7tzCwZQoq58jzh+/Lhat26tDh06KCYmRhkZGZIkFxcXJiSYHFPBUWrs27dPvXv31o8//qgLFy4oJSVFffr00YYNGwr0LWwKKABzu/I9onfv3gWmh0tScHCwUlNTS75AlBjCDQAAMBXG3AAAAFMh3AAAAFMh3AAAAFMh3AAAAFMh3AAAAFMh3AAAAFMh3AAAAFMh3AAAAFMh3AAAAFMh3AAAAFMh3AAAAFP5f9nQPBsTGvCWAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "make_table(ins, outs, \"Origs\", \"Trimmed\", \"Total Read Lengths\", samples, \"Sum of Total Reads Per Sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "f13cdc61-db2e-4955-9fe6-78993f9014d6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjMAAAHFCAYAAAAHcXhbAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8g+/7EAAAACXBIWXMAAA9hAAAPYQGoP6dpAAA2gUlEQVR4nO3dd3xUVf7/8feQHggBwpIQ6QLSQYmyFCkiSJUiCoISKStIEQxIkRUIKKFIxBUBQRdwleIqKspX10hTmnRUdEEpCWAwSkmoIeX8/vCXWYYkJAyTzFx4PR+PeTy855577ueeScybW2ZsxhgjAAAAiyri7gIAAABuBmEGAABYGmEGAABYGmEGAABYGmEGAABYGmEGAABYGmEGAABYGmEGAABYGmEGAABYGmEGt4Vt27bp0UcfVdmyZeXr66uwsDD16NFDW7duvaFxJk+eLJvN5lQNGzZskM1m04YNG5zaPr9atmypli1b5qufzWazv/z9/VWrVi299NJLunLlSoHWmJubmd+cnD17VqVLl9aKFSuy7aNMmTI6d+5ctm0qVaqkTp063fS+s97v/LxyY7PZNHny5JuuxZXy+/N1M86cOaMSJUro448/LtD94Nbh7e4CgIL2+uuva+TIkbrvvvs0c+ZMVaxYUQkJCXrjjTfUrFkzvfbaaxo2bFi+xho4cKDatWvnVB333HOPtm7dqlq1ajm1fUGoUqWK3nvvPUnS77//rrfeeksvvviiEhIStHDhQjdXd/Oio6MVHh6unj17Zlv3+++/a+bMmZo6dWqB7Dvr/b5at27ddOedd+qVV17J1xhbt25VuXLlCqI8p82bN6/A91GyZEk999xzev7559WhQwf5+voW+D5hcQa4hW3atMkUKVLEdOrUyaSlpTmsS0tLM506dTJFihQxmzZtuu44Fy5cKMgyXapFixamRYsW+epXu3Zth7a0tDRTrVo14+vray5dulRAFeZu0qRJxlX/Wzp16pQJCAgwCxYsyHEf7dq1M0WLFjWJiYkO6ytWrGg6duzokhqulZ+xMzMzzcWLFwtk/1Zy8uRJ4+3tbd577z13lwIL4DITbmkxMTGy2WyaP3++vL0dT0R6e3tr3rx5stlsmj59ur096zLE7t271aNHD5UsWVJ33nmnw7qrpaamatSoUQoLC1NgYKCaN2+uXbt2qVKlSnrqqafs/XK6zPTUU0+pWLFi+uWXX9ShQwcVK1ZM5cuX16hRo5Samuqwn+joaDVq1EilSpVS8eLFdc899+jtt9+WceF3xXp7e6tBgwa6cuWKzp49a283xmjevHlq0KCBAgICVLJkSfXo0UOHDx922D4uLk5dunRRuXLl5O/vr6pVq2rQoEH6448/su1rzZo1atCggfz8/FS5cuVcz1b8+9//VqNGjRQcHKzAwEBVqVJF/fv3z/NYlixZovT09BzPykjSSy+9pPT09Hxdxjl9+rSGDBmiO+64Q76+vqpSpYomTJiQ7T1yhs1m07Bhw7RgwQLVrFlTfn5+Wrp0qX3d1fUtWbJENptN69at09/+9jeFhISoePHi6tu3ry5cuKCTJ0/qscceU4kSJVS2bFmNHj1aaWlp9u2PHj0qm82mWbNmacaMGapUqZICAgLUsmVLHTx4UGlpaRo3bpzCw8MVHBysbt26KSkpyaHeay8zZY35yiuvKDY2VpUrV1axYsXUuHFjbdu2LdvxLlq0SNWrV5efn59q1aqlZcuW6amnnlKlSpUc+oWGhqpNmzZasGDBTc8xbn1cZsItKyMjQ+vXr1dERESup+rLly+vhg0bat26dcrIyJCXl5d9Xffu3dWrVy8NHjxYFy5cyHU//fr108qVKzVmzBg98MAD+vHHH9WtWzelpKTkq860tDQ9/PDDGjBggEaNGqWvv/5aU6dOVXBwsCZOnGjvd/ToUQ0aNEgVKlSQ9Od9QMOHD9eJEycc+t2sI0eOqESJEvrLX/5ibxs0aJCWLFmiZ599VjNmzNDp06c1ZcoUNWnSRPv27VNoaKgk6dChQ2rcuLEGDhyo4OBgHT16VLGxsWrWrJm+//57+fj4SJLWrl2rLl26qHHjxlqxYoUyMjI0c+ZM/fbbbw61bN26VT179lTPnj01efJk+fv7Kz4+XuvWrcvzONasWaO7775bJUqUyHF9xYoVNWTIEL3++uuKiopS9erVc+x3+fJltWrVSocOHVJ0dLTq1aunb775RjExMdq7d6/WrFmTn2m9ro8//ljffPONJk6cqLCwMJUpU+a6/QcOHKju3btrxYoV2rNnj1544QWlp6frwIED6t69u55++ml99dVXmjFjhsLDwxUVFeWw/RtvvKF69erpjTfe0NmzZzVq1Ch17txZjRo1ko+Pj/75z38qPj5eo0eP1sCBA7V69eo8j+GNN95QjRo1NGfOHEnSiy++qA4dOujIkSMKDg6WJC1cuFCDBg3SI488oldffVXJycmKjo7ONRS2bNlS48eP19mzZ3N9HwFJXGbCrevkyZNGkunVq9d1+/Xs2dNIMr/99psx5n+XISZOnJit77WXQfbv328kmbFjxzr0W758uZFkIiMj7W3r1683ksz69evtbZGRkUaSef/99x2279Chg7nrrrtyrTkjI8OkpaWZKVOmmJCQEJOZmWlfd6OXmdLS0kxaWppJTEw0EydONJIcLs1s3brVSDKzZ8922P7YsWMmICDAjBkzJsfxMzMzTVpamomPjzeSzCeffGJf16hRIxMeHu5wKSslJcWUKlXKYX5feeUVI8mcPXs2z+O5VmBgoBk8eHC29qz38Pfffzd//PGHCQ4ONo888oh9/bWXghYsWJDjezRjxgwjyXz55Zf5rimny0ySTHBwsDl9+nS2/pLMpEmT7MuLFy82kszw4cMd+nXt2tVIMrGxsQ7tDRo0MPfcc499+ciRI0aSqV+/vsnIyLC3z5kzx0gyDz/8sMP2I0eONJJMcnKyve3an6+sMevWrWvS09Pt7du3bzeSzPLly40xf/7MhoWFmUaNGjnsIz4+3vj4+JiKFStmO/64uDgjyXz++efZ1gFX4zITbnvm/1+mufby0SOPPJLnths3bpQkPfbYYw7tPXr0yHZZKzc2m02dO3d2aKtXr57i4+Md2tatW6cHH3xQwcHB8vLyko+PjyZOnKhTp05luxSQX/v375ePj498fHxUtmxZTZkyRePHj9egQYPsfT777DPZbDY98cQTSk9Pt7/CwsJUv359h8tmSUlJGjx4sMqXLy9vb2/5+PioYsWKkqSffvpJknThwgXt2LFD3bt3l7+/v33boKCgbPNw7733Svpzft9//32dOHEiX8d19uxZXbx4Mc8zHCEhIRo7dqw+/PBDffvttzn2WbdunYoWLaoePXo4tGddQly7dm2+arqeBx54QCVLlsx3/2uftqpZs6YkqWPHjtnar/05kqQOHTqoSJEiDv1y216SEhIS8qypY8eODmc269WrJ0n2/R84cMB+GexqFSpUUNOmTXMcM+v9y+/7jtsXYQa3rNKlSyswMFBHjhy5br+jR48qMDBQpUqVcmgvW7Zsnvs4deqUJNkvs2Tx9vZWSEhIvuoMDAx0+KMuSX5+frp8+bJ9efv27Wrbtq2kP+852Lx5s3bs2KEJEyZIki5dupSvfV3rzjvv1I4dO7R9+3b9+9//Vv369RUTE+PwKPNvv/0mY4xCQ0PtwSfrtW3bNvv9MJmZmWrbtq1WrVqlMWPGaO3atdq+fbv9vomsGs+cOaPMzEyFhYVlq+fatubNm+vjjz9Wenq6+vbtq3LlyqlOnTpavnz5dY8ra1/XzmtORo4cqfDwcI0ZMybH9adOnVJYWFi2sFumTBl5e3vbfwZuRn5+1q527c9q1tM+ObVf/XPkzPaSchzjWtf+vPv5+Un633uR2+9Kbm3S/94/Z3++cfvgnhncsry8vNSqVSt98cUXOn78eI73zRw/fly7du1S+/btHf5VKWU/U5OTrP+B//bbb7rjjjvs7enp6S75I5dlxYoV8vHx0WeffebwB/pmP4fD399fERERkv48C9KqVSvVrl1bI0eOVKdOnVSsWDGVLl1aNptN33zzjf0P1NWy2n744Qft27dPS5YsUWRkpH39L7/84tC/ZMmSstlsOnnyZLaxcmrr0qWLunTpotTUVG3btk0xMTHq3bu3KlWqpMaNG+d4XFnvy+nTp/Ocg4CAAE2ePFlPP/10jve/hISE6Ntvv5UxxuFnIikpSenp6SpdunSe+8iLKz9bx1Nd/btyrZzed+l/758r5hi3Ns7M4JY2fvx4GWM0ZMgQZWRkOKzLyMjQM888I2OMxo8f79T4zZs3lyStXLnSof2DDz5Qenq6c0XnwGazydvb2yFwXbp0Sf/6179ctg/pzz8406dP12+//abXX39d0p+XNIwxOnHihCIiIrK96tata69RUrbA8+abbzosFy1aVPfdd59WrVrl8C/+c+fO6dNPP821Nj8/P7Vo0UIzZsyQJO3ZsyfXvllPHB06dChfx92/f3/VrFlT48aNU2ZmpsO61q1b6/z589mC4zvvvGNfj7zdddddCgsL0/vvv+/QnpCQoC1btuS4TdbTcp702UzwTJyZwS2tadOmmjNnjkaOHKlmzZpp2LBhqlChgv1D87799lvNmTNHTZo0cWr82rVr6/HHH9fs2bPl5eWlBx54QPv379fs2bMVHBzscF/CzejYsaNiY2PVu3dvPf300zp16pReeeWVHM+U3Ky+ffsqNjZWr7zyioYOHaqmTZvq6aefVr9+/bRz5041b95cRYsWVWJiojZt2qS6devqmWeeUY0aNXTnnXdq3LhxMsaoVKlS+vTTTxUXF5dtH1OnTlW7du3Upk0bjRo1ShkZGZoxY4aKFi3qcDZl4sSJOn78uFq3bq1y5crp7Nmzeu211+Tj46MWLVpc9zhatmypzz//PF/H7OXlpWnTpqlbt26S/ne/R9Z8vPHGG4qMjNTRo0dVt25dbdq0SdOmTVOHDh304IMP5msft7siRYooOjpagwYNUo8ePdS/f3+dPXtW0dHRKlu2bI6/K9u2bVNISIg9MAO54cwMbnnDhw/X5s2bVa5cOY0aNUoPPPCAoqKiVLZsWW3atEnDhw+/qfEXL16sESNG6O2331bnzp21YsUK+78+XfU46QMPPKB//vOf+v7779W5c2dNmDBBPXr00Lhx41wy/tWKFCmi6dOn6/Tp0/bHbN98803NnTtXX3/9tXr16qWOHTtq4sSJunDhgu677z5Jko+Pjz799FNVr15dgwYN0uOPP66kpCR99dVX2fbRpk0bffzxx0pJSVHPnj0VFRWlRx55JNvnxzRq1EgnT57U2LFj1bZtWz399NMKCAjQunXrVLt27eseR58+fZSYmKgdO3bk67i7du2aY6j19/fX+vXr1adPH82aNUvt27fXkiVLNHr0aK1atSpfY+NPTz/9tBYuXKh9+/apW7duio6O1rhx43J8hN4Yo9WrV6t37963xWU43BybMS78xC0AkqQtW7aoadOmeu+999S7d293l3Pbqlevnpo2bar58+e7uxTk4uzZs6pevbq6du3q8BUaa9euVdu2bbV//37VqFHDjRXCCggzwE2Ki4vT1q1b1bBhQwUEBGjfvn2aPn26goOD9d133+XriRoUjC+++ELdunXTzz//7HHfcXQ7OnnypF5++WW1atVKISEhio+P16uvvqr//ve/2rlzp8PZtlatWqlq1apatGiRGyuGVXDPDHCTihcvri+//FJz5szRuXPnVLp0abVv314xMTEEGTdr166dZs2apSNHjhBmPICfn5+OHj2qIUOG6PTp0woMDNRf//pXLViwwCHInDlzRi1atNCQIUPcWC2shDMzAADA0rgBGAAAWBphBgAAWBphBgAAWNotfwNwZmamfv31VwUFBfFZBQAAWIQxRufOnVN4eHieH0B6y4eZX3/9VeXLl3d3GQAAwAnHjh3L82nEWz7MBAUFSfpzMooXL+7magAAQH6kpKSofPny9r/j13PLh5msS0vFixcnzAAAYDH5uUWEG4ABAIClEWYAAIClEWYAAIClEWYAAIClEWYAAIClEWYAAIClEWYAAIClEWYAAIClEWYAAIClEWYAAICluTXMfP311+rcubPCw8Nls9n08ccfO6w3xmjy5MkKDw9XQECAWrZsqf3797unWAAA4JHcGmYuXLig+vXra+7cuTmunzlzpmJjYzV37lzt2LFDYWFhatOmjc6dO1fIlQIAAE/l1i+abN++vdq3b5/jOmOM5syZowkTJqh79+6SpKVLlyo0NFTLli3ToEGDCrNUAADgoTz2npkjR47o5MmTatu2rb3Nz89PLVq00JYtW9xYGQAA8CRuPTNzPSdPnpQkhYaGOrSHhoYqPj4+1+1SU1OVmppqX05JSSmYAgEAgEfw2DCTxWazOSwbY7K1XS0mJkbR0dEFXRYAD1Fp3Bp3l3DDjk7v6O4SgFuKx15mCgsLk/S/MzRZkpKSsp2tudr48eOVnJxsfx07dqxA6wQAAO7lsWGmcuXKCgsLU1xcnL3typUr2rhxo5o0aZLrdn5+fipevLjDCwAA3Lrcepnp/Pnz+uWXX+zLR44c0d69e1WqVClVqFBBI0eO1LRp01StWjVVq1ZN06ZNU2BgoHr37u3GqgEAgCdxa5jZuXOnWrVqZV+OioqSJEVGRmrJkiUaM2aMLl26pCFDhujMmTNq1KiRvvzySwUFBbmrZAAA4GFsxhjj7iIKUkpKioKDg5WcnMwlJ+AWxA3AwK3pRv5+e+w9MwAAAPlBmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJZGmAEAAJbm0WEmPT1df//731W5cmUFBASoSpUqmjJlijIzM91dGgAA8BDe7i7gembMmKEFCxZo6dKlql27tnbu3Kl+/fopODhYI0aMcHd5AADAA3h0mNm6dau6dOmijh07SpIqVaqk5cuXa+fOnW6uDAAAeAqPvszUrFkzrV27VgcPHpQk7du3T5s2bVKHDh1y3SY1NVUpKSkOLwAAcOvy6DMzY8eOVXJysmrUqCEvLy9lZGTo5Zdf1uOPP57rNjExMYqOji7EKgEAgDt59JmZlStX6t1339WyZcu0e/duLV26VK+88oqWLl2a6zbjx49XcnKy/XXs2LFCrBgAABQ2jz4z8/zzz2vcuHHq1auXJKlu3bqKj49XTEyMIiMjc9zGz89Pfn5+hVkmAABwI48+M3Px4kUVKeJYopeXF49mAwAAO48+M9O5c2e9/PLLqlChgmrXrq09e/YoNjZW/fv3d3dpAADAQ3h0mHn99df14osvasiQIUpKSlJ4eLgGDRqkiRMnurs0AADgITw6zAQFBWnOnDmaM2eOu0sBAAAeyqPvmQEAAMgLYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFgaYQYAAFiaU2HmyJEjrq4DAADAKU6FmapVq6pVq1Z69913dfnyZVfXBAAAkG9OhZl9+/bp7rvv1qhRoxQWFqZBgwZp+/btrq5NknTixAk98cQTCgkJUWBgoBo0aKBdu3YVyL4AAID1OBVm6tSpo9jYWJ04cUKLFy/WyZMn1axZM9WuXVuxsbH6/fffXVLcmTNn1LRpU/n4+Ojzzz/Xjz/+qNmzZ6tEiRIuGR8AAFifzRhjbnaQ1NRUzZs3T+PHj9eVK1fk4+Ojnj17asaMGSpbtqzT444bN06bN2/WN9984/QYKSkpCg4OVnJysooXL+70OAA8U6Vxa9xdwg07Or2ju0sAPN6N/P2+qaeZdu7cqSFDhqhs2bKKjY3V6NGjdejQIa1bt04nTpxQly5dbmZ4rV69WhEREXr00UdVpkwZ3X333Vq0aNFNjQkAAG4t3s5sFBsbq8WLF+vAgQPq0KGD3nnnHXXo0EFFivyZjSpXrqw333xTNWrUuKniDh8+rPnz5ysqKkovvPCCtm/frmeffVZ+fn7q27dvjtukpqYqNTXVvpySknJTNQAAAM/mVJiZP3+++vfvr379+iksLCzHPhUqVNDbb799U8VlZmYqIiJC06ZNkyTdfffd2r9/v+bPn59rmImJiVF0dPRN7RcAAFiHU2Hm559/zrOPr6+vIiMjnRnermzZsqpVq5ZDW82aNfXhhx/mus348eMVFRVlX05JSVH58uVvqg4AAOC5nAozixcvVrFixfToo486tP/73//WxYsXbzrEZGnatKkOHDjg0Hbw4EFVrFgx1238/Pzk5+fnkv0DAADP59QNwNOnT1fp0qWztZcpU8Z+ScgVnnvuOW3btk3Tpk3TL7/8omXLlmnhwoUaOnSoy/YBAACszakwEx8fr8qVK2drr1ixohISEm66qCz33nuvPvroIy1fvlx16tTR1KlTNWfOHPXp08dl+wAAANbm1GWmMmXK6LvvvlOlSpUc2vft26eQkBBX1GXXqVMnderUyaVjAgCAW4dTZ2Z69eqlZ599VuvXr1dGRoYyMjK0bt06jRgxQr169XJ1jQAAALly6szMSy+9pPj4eLVu3Vre3n8OkZmZqb59+7r0nhkAAIC8OBVmfH19tXLlSk2dOlX79u1TQECA6tate92njAAAAAqCU2EmS/Xq1VW9enVX1QIAAHDDnAozGRkZWrJkidauXaukpCRlZmY6rF+3bp1LigMAAMiLU2FmxIgRWrJkiTp27Kg6derIZrO5ui4AAIB8cSrMrFixQu+//746dOjg6noAAABuiFOPZvv6+qpq1aqurgUAAOCGORVmRo0apddee03GGFfXAwAAcEOcusy0adMmrV+/Xp9//rlq164tHx8fh/WrVq1ySXEAAAB5cSrMlChRQt26dXN1LQAAADfMqTCzePFiV9cBAADgFKfumZGk9PR0ffXVV3rzzTd17tw5SdKvv/6q8+fPu6w4AACAvDh1ZiY+Pl7t2rVTQkKCUlNT1aZNGwUFBWnmzJm6fPmyFixY4Oo6AQAAcuTUmZkRI0YoIiJCZ86cUUBAgL29W7duWrt2rcuKAwAAyIvTTzNt3rxZvr6+Du0VK1bUiRMnXFIYAABAfjh1ZiYzM1MZGRnZ2o8fP66goKCbLgoAACC/nAozbdq00Zw5c+zLNptN58+f16RJk/iKAwAAUKicusz06quvqlWrVqpVq5YuX76s3r176+eff1bp0qW1fPlyV9cIAACQK6fCTHh4uPbu3avly5dr9+7dyszM1IABA9SnTx+HG4IBAAAKmlNhRpICAgLUv39/9e/f35X1AAAA3BCnwsw777xz3fV9+/Z1qhgAAIAb5VSYGTFihMNyWlqaLl68KF9fXwUGBhJmAABAoXHqaaYzZ844vM6fP68DBw6oWbNm3AAMAAAKldPfzXStatWqafr06dnO2gAAABQkl4UZSfLy8tKvv/7qyiEBAACuy6l7ZlavXu2wbIxRYmKi5s6dq6ZNm7qkMAAAgPxwKsx07drVYdlms+kvf/mLHnjgAc2ePdsVdQEAAOSLU2EmMzPT1XUAAAA4xaX3zAAAABQ2p87MREVF5btvbGysM7sAAADIF6fCzJ49e7R7926lp6frrrvukiQdPHhQXl5euueee+z9bDaba6oEAADIhVNhpnPnzgoKCtLSpUtVsmRJSX9+kF6/fv10//33a9SoUS4tEgAAIDdO3TMze/ZsxcTE2IOMJJUsWVIvvfQSTzMBAIBC5VSYSUlJ0W+//ZatPSkpSefOnbvpogAAAPLLqTDTrVs39evXTx988IGOHz+u48eP64MPPtCAAQPUvXt3V9cIAACQK6fumVmwYIFGjx6tJ554QmlpaX8O5O2tAQMGaNasWS4tEAAA4HqcCjOBgYGaN2+eZs2apUOHDskYo6pVq6po0aKurg8AAOC6bupD8xITE5WYmKjq1auraNGiMsa4qi4AAIB8cSrMnDp1Sq1bt1b16tXVoUMHJSYmSpIGDhzIY9kAAKBQORVmnnvuOfn4+CghIUGBgYH29p49e+qLL75wWXEAAAB5ceqemS+//FL/+c9/VK5cOYf2atWqKT4+3iWFAQAA5IdTZ2YuXLjgcEYmyx9//CE/P7+bLgoAACC/nAozzZs31zvvvGNfttlsyszM1KxZs9SqVSuXFQcAAJAXpy4zzZo1Sy1bttTOnTt15coVjRkzRvv379fp06e1efNmV9cIAACQK6fOzNSqVUvfffed7rvvPrVp00YXLlxQ9+7dtWfPHt15552urhEAACBXN3xmJi0tTW3bttWbb76p6OjogqgJAAAg3274zIyPj49++OEH2Wy2gqgHAADghjh1malv3756++23XV0LAADADXPqBuArV67orbfeUlxcnCIiIrJ9J1NsbKxLigMAAMjLDYWZw4cPq1KlSvrhhx90zz33SJIOHjzo0IfLTwAAoDDdUJipVq2aEhMTtX79ekl/fn3BP/7xD4WGhhZIcQAAAHm5oXtmrv1W7M8//1wXLlxwaUEAAAA3wqkbgLNcG24AAAAK2w2FGZvNlu2eGO6RAQAA7nRD98wYY/TUU0/Zv0zy8uXLGjx4cLanmVatWuW6CgEAAK7jhsJMZGSkw/ITTzzh0mIAAABu1A2FmcWLFxdUHQAAAE65qRuAAQAA3I0wAwAALM1SYSYmJkY2m00jR450dykAAMBDWCbM7NixQwsXLlS9evXcXQoAAPAglggz58+fV58+fbRo0SKVLFnS3eUAAAAPYokwM3ToUHXs2FEPPvhgnn1TU1OVkpLi8AIAALeuG3o02x1WrFih3bt3a8eOHfnqHxMTo+jo6AKuCgAAeAqPPjNz7NgxjRgxQu+++678/f3ztc348eOVnJxsfx07dqyAqwQAAO7k0Wdmdu3apaSkJDVs2NDelpGRoa+//lpz585VamqqvLy8HLbx8/Ozf90CAAC49Xl0mGndurW+//57h7Z+/fqpRo0aGjt2bLYgAwAAbj8eHWaCgoJUp04dh7aiRYsqJCQkWzsAALg9efQ9MwAAAHnx6DMzOdmwYYO7SwAAAB6EMzMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSPDrMxMTE6N5771VQUJDKlCmjrl276sCBA+4uCwAAeBCPDjMbN27U0KFDtW3bNsXFxSk9PV1t27bVhQsX3F0aAADwEN7uLuB6vvjiC4flxYsXq0yZMtq1a5eaN2/upqoAAIAn8egwc63k5GRJUqlSpXLtk5qaqtTUVPtySkpKgdcFAADcx6MvM13NGKOoqCg1a9ZMderUybVfTEyMgoOD7a/y5csXYpUAAKCwWSbMDBs2TN99952WL19+3X7jx49XcnKy/XXs2LFCqhAAALiDJS4zDR8+XKtXr9bXX3+tcuXKXbevn5+f/Pz8CqkyAADgbh4dZowxGj58uD766CNt2LBBlStXdndJAADAw3h0mBk6dKiWLVumTz75REFBQTp58qQkKTg4WAEBAW6uDgAAeAKPvmdm/vz5Sk5OVsuWLVW2bFn7a+XKle4uDQAAeAiPPjNjjHF3CQAAwMN59JkZAACAvBBmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApRFmAACApVkizMybN0+VK1eWv7+/GjZsqG+++cbdJQEAAA/h8WFm5cqVGjlypCZMmKA9e/bo/vvvV/v27ZWQkODu0gAAgAfw+DATGxurAQMGaODAgapZs6bmzJmj8uXLa/78+e4uDQAAeACPDjNXrlzRrl271LZtW4f2tm3basuWLW6qCgAAeBJvdxdwPX/88YcyMjIUGhrq0B4aGqqTJ0/muE1qaqpSU1Pty8nJyZKklJSUgisUgNtkpl50dwk3jP8fAXnL+j0xxuTZ16PDTBabzeawbIzJ1pYlJiZG0dHR2drLly9fILUBwI0KnuPuCgDrOHfunIKDg6/bx6PDTOnSpeXl5ZXtLExSUlK2szVZxo8fr6ioKPtyZmamTp8+rZCQkFwD0O0kJSVF5cuX17Fjx1S8eHF3l3PLYp4LB/NcOJjnwsE8OzLG6Ny5cwoPD8+zr0eHGV9fXzVs2FBxcXHq1q2bvT0uLk5dunTJcRs/Pz/5+fk5tJUoUaIgy7Sk4sWL88tSCJjnwsE8Fw7muXAwz/+T1xmZLB4dZiQpKipKTz75pCIiItS4cWMtXLhQCQkJGjx4sLtLAwAAHsDjw0zPnj116tQpTZkyRYmJiapTp47+7//+TxUrVnR3aQAAwAN4fJiRpCFDhmjIkCHuLuOW4Ofnp0mTJmW7FAfXYp4LB/NcOJjnwsE8O89m8vPMEwAAgIfy6A/NAwAAyAthBgAAWBphBgAAWBphBgAAWBphxiK+/vprde7cWeHh4bLZbPr444/t69LS0jR27FjVrVtXRYsWVXh4uPr27atff/3VYYzU1FQNHz5cpUuXVtGiRfXwww/r+PHjee77xIkTeuKJJxQSEqLAwEA1aNBAu3btcvUhegR3zXN6err+/ve/q3LlygoICFCVKlU0ZcoUZWZmFsRhegRXzPXChQvVsmVLFS9eXDabTWfPns3XvufNm6fKlSvL399fDRs21DfffOPCI/Ms7prnmJgY3XvvvQoKClKZMmXUtWtXHThwwMVH5znc+fOcJSYmRjabTSNHjrz5A7IYwoxFXLhwQfXr19fcuXOzrbt48aJ2796tF198Ubt379aqVat08OBBPfzwww79Ro4cqY8++kgrVqzQpk2bdP78eXXq1EkZGRm57vfMmTNq2rSpfHx89Pnnn+vHH3/U7Nmzb9lPVXbXPM+YMUMLFizQ3Llz9dNPP2nmzJmaNWuWXn/9dZcfo6dwxVxfvHhR7dq10wsvvJDv/a5cuVIjR47UhAkTtGfPHt1///1q3769EhISbvqYPJG75nnjxo0aOnSotm3bpri4OKWnp6tt27a6cOHCTR+TJ3LXPGfZsWOHFi5cqHr16jl9DJZmYDmSzEcffXTdPtu3bzeSTHx8vDHGmLNnzxofHx+zYsUKe58TJ06YIkWKmC+++CLXccaOHWuaNWvmkrqtpjDnuWPHjqZ///4Obd27dzdPPPGE8wdgIc7M9dXWr19vJJkzZ87kua/77rvPDB482KGtRo0aZty4cTdSsiUV5jxfKykpyUgyGzduvOFtraaw5/ncuXOmWrVqJi4uzrRo0cKMGDHixou2OM7M3KKSk5Nls9nsZ1B27dqltLQ0tW3b1t4nPDxcderU0ZYtW3IdZ/Xq1YqIiNCjjz6qMmXK6O6779aiRYsKunzLcNU8N2vWTGvXrtXBgwclSfv27dOmTZvUoUOHAq3fSq6da2dcuXJFu3btcnh/JKlt27bXfX9uJ66Y59zGlaRSpUq5dFyrcuU8Dx06VB07dtSDDz5484VZlCU+ARg35vLlyxo3bpx69+5t/7KykydPytfXVyVLlnToGxoamu1bya92+PBhzZ8/X1FRUXrhhRe0fft2Pfvss/Lz81Pfvn0L9Dg8nSvneezYsUpOTlaNGjXk5eWljIwMvfzyy3r88ccL9BisIqe5dsYff/yhjIwMhYaGOrTn9f7cLlw1z9cyxigqKkrNmjVTnTp1XDauVblynlesWKHdu3drx44dLqrOmggzt5i0tDT16tVLmZmZmjdvXp79jTGy2Wy5rs/MzFRERISmTZsmSbr77ru1f/9+zZ8//7YOM66e55UrV+rdd9/VsmXLVLt2be3du1cjR45UeHi4IiMjXVm65dzoXOfHte9FXu/P7aAg5jnLsGHD9N1332nTpk0uHdeKXDnPx44d04gRI/Tll1/K39/fRRVaE5eZbiFpaWl67LHHdOTIEcXFxTkk/rCwMF25ckVnzpxx2CYpKSnbv1KvVrZsWdWqVcuhrWbNmrfszZL5URDz/Pzzz2vcuHHq1auX6tatqyeffFLPPfecYmJiCuw4rOB6c+2M0qVLy8vLK9tZmLzen1udq+f5asOHD9fq1au1fv16lStXzmXjWpGr53nXrl1KSkpSw4YN5e3tLW9vb23cuFH/+Mc/5O3tfd2HDm41hJlbRNYvyc8//6yvvvpKISEhDusbNmwoHx8fxcXF2dsSExP1ww8/qEmTJrmO27Rp02yPUx48ePC2/dbygprnixcvqkgRx19HLy+vW/rR7LzkNdfO8PX1VcOGDR3eH0mKi4u77vtzKyuIeZb+PNs1bNgwrVq1SuvWrVPlypVdMq5VFcQ8t27dWt9//7327t1rf0VERKhPnz7au3evvLy8XFC5NXCZySLOnz+vX375xb585MgR7d27V6VKlVJ4eLh69Oih3bt367PPPlNGRob9X56lSpWSr6+vgoODNWDAAI0aNUohISEqVaqURo8erbp16zrcNNa6dWt169ZNw4YNkyQ999xzatKkiaZNm6bHHntM27dv18KFC7Vw4cLCnYBC4q557ty5s15++WVVqFBBtWvX1p49exQbG6v+/fsX7gQUopuda+nPe5ROnjxpH+f7779XUFCQKlSoYL/R9Nq5joqK0pNPPqmIiAg1btxYCxcuVEJCggYPHlyYh19o3DXPQ4cO1bJly/TJJ58oKCjIPm5wcLACAgIK7fgLizvmOSgoKNs9SEWLFlVISMjtd2+SW5+lQr5lPap37SsyMtIcOXIkx3WSzPr16+1jXLp0yQwbNsyUKlXKBAQEmE6dOpmEhASH/VSsWNFMmjTJoe3TTz81derUMX5+fqZGjRpm4cKFhXDE7uGueU5JSTEjRowwFSpUMP7+/qZKlSpmwoQJJjU1tZCOvPC5Yq4nTZqUY5/Fixfb++T0M/3GG2+YihUrGl9fX3PPPffc0o8Lu2uecxv36m1uJe78eb7a7fpots0YY5yLQQAAAO7HPTMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAAMDSCDMAPMrkyZPVoEEDd5fhlKNHj8pms2nv3r3uLgW4rRBmAGTz1FNPyWazyWazydvbWxUqVNAzzzyT7Qs03cFTAsNTTz2lrl27urUGAH8izADIUbt27ZSYmKijR4/qrbfe0qeffqohQ4a4uywAyIYwAyBHfn5+CgsLU7ly5dS2bVv17NlTX375pUOfxYsXq2bNmvL391eNGjU0b948h/Vjx45V9erVFRgYqCpVqujFF19UWlqaQ5/p06crNDRUQUFBGjBggC5fvnxTdRtjNHPmTFWpUkUBAQGqX7++PvjgA/v6DRs2yGazae3atYqIiFBgYKCaNGmS7dvhX3rpJZUpU0ZBQUEaOHCgxo0bZ7/8NXnyZC1dulSffPKJ/QzWhg0b7NsePnxYrVq1UmBgoOrXr6+tW7fe1DEByIObvxsKgAeKjIw0Xbp0sS8fOnTI1KpVy4SGhtrbFi5caMqWLWs+/PBDc/jwYfPhhx+aUqVKmSVLltj7TJ061WzevNkcOXLErF692oSGhpoZM2bY169cudL4+vqaRYsWmf/+979mwoQJJigoyNSvXz/X2rK+tG/Pnj05rn/hhRdMjRo1zBdffGEOHTpkFi9ebPz8/MyGDRuMMf/7QsBGjRqZDRs2mP3795v777/fNGnSxD7Gu+++a/z9/c0///lPc+DAARMdHW2KFy9ur+vcuXPmscceM+3atTOJiYkmMTHRpKam2murUaOG+eyzz8yBAwdMjx49TMWKFU1aWtoNvAMAbgRhBkA2kZGRxsvLyxQtWtT4+/vbv703NjbW3qd8+fJm2bJlDttNnTrVNG7cONdxZ86caRo2bGhfbty4sRk8eLBDn0aNGjkdZs6fP2/8/f3Nli1bHNoHDBhgHn/8cWPM/8LMV199ZV+/Zs0aI8lcunTJXsPQoUMdxmjatKlDXdcGvqtre+utt+xt+/fvN5LMTz/9lOsxAbg5XGYCkKNWrVpp7969+vbbbzV8+HA99NBDGj58uCTp999/17FjxzRgwAAVK1bM/nrppZd06NAh+xgffPCBmjVrprCwMBUrVkwvvviiEhIS7Ot/+uknNW7c2GG/1y7fiB9//FGXL19WmzZtHOp65513HOqSpHr16tn/u2zZspKkpKQkSdKBAwd03333OfS/dvl6rjc2ANfzdncBADxT0aJFVbVqVUnSP/7xD7Vq1UrR0dGaOnWqMjMzJUmLFi1So0aNHLbz8vKSJG3btk29evVSdHS0HnroIQUHB2vFihWaPXt2gdWcVdeaNWt0xx13OKzz8/NzWPbx8bH/t81mc9j+6rYsxph815HX2ABcizADIF8mTZqk9u3b65lnnlF4eLjuuOMOHT58WH369Mmx/+bNm1WxYkVNmDDB3hYfH+/Qp2bNmtq2bZv69u1rb9u2bZvTNdaqVUt+fn5KSEhQixYtnB7nrrvu0vbt2/Xkk0/a23bu3OnQx9fXVxkZGU7vA4DrEGYA5EvLli1Vu3ZtTZs2TXPnztXkyZP17LPPqnjx4mrfvr1SU1O1c+dOnTlzRlFRUapataoSEhK0YsUK3XvvvVqzZo0++ugjhzFHjBihyMhIRUREqFmzZnrvvfe0f/9+ValSJc96rn36SPozzIwePVrPPfecMjMz1axZM6WkpGjLli0qVqyYIiMj83Wsw4cP19/+9jdFRESoSZMmWrlypb777juHuipVqqT//Oc/OnDggEJCQhQcHJyvsQEUAHfftAPA8+R0c6sxxrz33nvG19fXJCQk2JcbNGhgfH19TcmSJU3z5s3NqlWr7P2ff/55ExISYooVK2Z69uxpXn31VRMcHOww5ssvv2xKly5tihUrZiIjI82YMWPydQNwTq8jR46YzMxM89prr5m77rrL+Pj4mL/85S/moYceMhs3bjTG/O8G4DNnztjH3LNnj337LFOmTLHX1b9/f/Pss8+av/71r/b1SUlJpk2bNqZYsWJGklm/fn2ONyefOXPGvh5AwbAZcwMXggHgNtWmTRuFhYXpX//6l7tLAXANLjMBwDUuXryoBQsW6KGHHpKXl5eWL1+ur776SnFxce4uDUAOODMDANe4dOmSOnfurN27dys1NVV33XWX/v73v6t79+7uLg1ADggzAADA0vjQPAAAYGmEGQAAYGmEGQAAYGmEGQAAYGmEGQAAYGmEGQAAYGmEGQAAYGmEGQAAYGmEGQAAYGn/D0ysjj2jIQ4qAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "reads = !cat {in_fn} | seqtk seq -A |grep -v '>'\n",
    "ds = [len(read) for read in reads]\n",
    "make_histogram(ds, \"Read Length\", \"Original Reads (No Trimming)\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d14099b9-2baf-4547-9998-f3d9656fda5b",
   "metadata": {},
   "source": [
    "Sample log file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "25a99080",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is cutadapt 4.6 with Python 3.10.8\n",
      "Command line parameters: -j 0 --nextseq-trim=15 --action=trim -a ATCACGAGATCGGAAGAGCACACGTCT;e=0.15;o=6;anywhere; -n 2 -u 5 -u -5 --max-n=0 -q 15 -m 20 --rename={id}_{cut_prefix}{cut_suffix} {comment} --too-short-output=trimmed/t1.fastq_tooshort -o trimmed/t1.fq.gz samples/t1.fq.gz\n",
      "Processing single-end reads on 32 cores ...\n",
      "Finished in 1.341 s (1.433 µs/read; 41.87 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:                 935,632\n",
      "Reads with adapters:                   909,387 (97.2%)\n",
      "\n",
      "== Read fate breakdown ==\n",
      "Reads that were too short:               5,278 (0.6%)\n",
      "Reads with too many N:                     233 (0.0%)\n",
      "Reads written (passing filters):       930,121 (99.4%)\n",
      "\n",
      "Total basepairs processed:   113,211,472 bp\n",
      "Quality-trimmed:                 840,606 bp (0.7%)\n",
      "Total written (filtered):     53,041,029 bp (46.9%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: ATCACGAGATCGGAAGAGCACACGTCT; Type: regular 3'; Length: 27; Trimmed: 911187 times\n",
      "\n",
      "Minimum overlap: 6\n",
      "No. of allowed errors:\n",
      "1-5 bp: 0; 6-12 bp: 1; 13-19 bp: 2; 20-25 bp: 3; 26-27 bp: 4\n",
      "\n",
      "Bases preceding removed adapters:\n",
      "  A: 31.4%\n",
      "  C: 22.3%\n",
      "  G: 11.6%\n",
      "  T: 34.6%\n",
      "  none/other: 0.0%\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "6\t654\t228.4\t0\t28 626\n",
      "7\t1728\t57.1\t1\t513 1215\n",
      "8\t51\t14.3\t1\t0 51\n",
      "9\t351\t3.6\t1\t316 35\n",
      "10\t433\t0.9\t1\t402 31\n",
      "11\t1793\t0.2\t1\t1721 72\n",
      "12\t4\t0.1\t1\t0 4\n",
      "13\t7\t0.0\t1\t0 4 3\n",
      "14\t893\t0.0\t2\t831 52 10\n",
      "15\t2139\t0.0\t2\t2014 95 30\n",
      "16\t27\t0.0\t2\t0 25 2\n",
      "17\t2494\t0.0\t2\t2352 115 27\n",
      "18\t24\t0.0\t2\t0 22 1 1\n",
      "19\t1491\t0.0\t2\t1390 84 17\n",
      "20\t1726\t0.0\t3\t1569 112 34 11\n",
      "21\t1963\t0.0\t3\t1815 107 33 8\n",
      "22\t2211\t0.0\t3\t2014 146 25 26\n",
      "23\t5064\t0.0\t3\t4693 265 54 52\n",
      "24\t40\t0.0\t3\t0 33 3 4\n",
      "25\t2876\t0.0\t3\t2688 150 24 14\n",
      "26\t3126\t0.0\t3\t2823 236 44 22 1\n",
      "27\t7188\t0.0\t4\t6761 319 60 28 20\n",
      "28\t126\t0.0\t4\t56 61 5 1 3\n",
      "29\t4255\t0.0\t4\t3923 250 47 16 19\n",
      "30\t4704\t0.0\t4\t4288 292 52 42 30\n",
      "31\t5034\t0.0\t4\t4634 314 53 24 9\n",
      "32\t5565\t0.0\t4\t5156 337 39 20 13\n",
      "33\t6194\t0.0\t4\t5719 382 55 21 17\n",
      "34\t6739\t0.0\t4\t6177 470 55 20 17\n",
      "35\t14931\t0.0\t4\t13985 724 116 62 44\n",
      "36\t215\t0.0\t4\t70 132 8 3 2\n",
      "37\t8513\t0.0\t4\t7953 472 44 29 15\n",
      "38\t9375\t0.0\t4\t8743 504 70 35 23\n",
      "39\t10059\t0.0\t4\t9433 509 65 27 25\n",
      "40\t10972\t0.0\t4\t10216 605 83 40 28\n",
      "41\t12024\t0.0\t4\t11194 689 70 40 31\n",
      "42\t12852\t0.0\t4\t12024 686 65 50 27\n",
      "43\t14070\t0.0\t4\t13097 810 94 41 28\n",
      "44\t14811\t0.0\t4\t13898 756 81 36 40\n",
      "45\t16173\t0.0\t4\t15152 859 90 40 32\n",
      "46\t17215\t0.0\t4\t16171 874 86 49 35\n",
      "47\t18458\t0.0\t4\t17066 989 167 117 119\n",
      "48\t19158\t0.0\t4\t17961 1024 83 43 47\n",
      "49\t25604\t0.0\t4\t24100 1254 155 55 40\n",
      "50\t17835\t0.0\t4\t16419 1276 87 26 27\n",
      "51\t46449\t0.0\t4\t44362 1778 166 93 50\n",
      "52\t888\t0.0\t4\t372 459 30 17 10\n",
      "53\t26632\t0.0\t4\t25157 1260 117 51 47\n",
      "54\t27057\t0.0\t4\t25059 1716 158 67 57\n",
      "55\t55029\t0.0\t4\t52693 2003 161 92 80\n",
      "56\t1537\t0.0\t4\t798 599 84 26 30\n",
      "57\t29501\t0.0\t4\t27572 1712 116 55 46\n",
      "58\t62664\t0.0\t4\t59598 2520 273 156 117\n",
      "59\t1170\t0.0\t4\t485 590 53 23 19\n",
      "60\t31193\t0.0\t4\t29443 1534 106 62 48\n",
      "61\t37396\t0.0\t4\t35462 1721 120 49 44\n",
      "62\t25465\t0.0\t4\t23719 1474 142 62 68\n",
      "63\t37775\t0.0\t4\t35513 1904 193 76 89\n",
      "64\t32036\t0.0\t4\t30138 1699 101 44 54\n",
      "65\t39768\t0.0\t4\t37941 1601 118 54 54\n",
      "66\t780\t0.0\t4\t313 399 38 15 15\n",
      "67\t24649\t0.0\t4\t23443 1047 78 41 40\n",
      "68\t15095\t0.0\t4\t14008 965 61 28 33\n",
      "69\t29423\t0.0\t4\t28146 1103 78 51 45\n",
      "70\t550\t0.0\t4\t248 239 23 18 22\n",
      "71\t10542\t0.0\t4\t9935 483 38 46 40\n",
      "72\t10385\t0.0\t4\t9568 533 106 76 102\n",
      "73\t11939\t0.0\t4\t10788 849 91 82 129\n",
      "74\t41326\t0.0\t4\t39098 1687 175 167 199\n",
      "75\t16166\t0.0\t4\t15109 774 98 94 91\n",
      "76\t4311\t0.0\t4\t4001 220 36 24 30\n",
      "77\t176\t0.0\t4\t133 30 2 7 4\n",
      "78\t53\t0.0\t4\t30 14 2 3 4\n",
      "79\t24\t0.0\t4\t13 5 3 2 1\n",
      "80\t20\t0.0\t4\t17 0 2 1\n",
      "81\t7\t0.0\t4\t4 1 1 0 1\n",
      "82\t7\t0.0\t4\t3 1 0 1 2\n",
      "83\t2\t0.0\t4\t0 1 1\n",
      "84\t1\t0.0\t4\t1\n",
      "87\t1\t0.0\t4\t0 0 0 1\n",
      "88\t1\t0.0\t4\t0 1\n",
      "91\t1\t0.0\t4\t0 1\n",
      "92\t1\t0.0\t4\t0 1\n",
      "93\t1\t0.0\t4\t1\n",
      "102\t1\t0.0\t4\t0 1\n"
     ]
    }
   ],
   "source": [
    "!cat {fname(out_path, samples[0], \"log\")}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f01aabd",
   "metadata": {},
   "source": [
    "## Step: extract_umi"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f3b1fed-c1f7-49e3-b7ba-6ba2e04cab6d",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Trim 8-nt umi from 5' end and 4-nt from random primer on 3' end"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "dd5644cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_path  = mkpath(\"adapter_trimmed\")\n",
    "out_path = mkpath(\"extract_umi\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc67f4cc-a8a4-4db8-a006-9c0a517948d0",
   "metadata": {},
   "source": [
    "#### cutadapt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39bff7bb-f62d-4788-b753-799b74b1800b",
   "metadata": {},
   "source": [
    "https://cutadapt.readthedocs.io/en/stable/guide.html"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d69157c7-7084-443b-b967-c3ef25d49528",
   "metadata": {},
   "source": [
    "##### paramaters\n",
    "- -j, number of cores\n",
    "- -q 20, trim low quality seq from 3' end\n",
    "- -m 20, fiter out reads shorter than 20\n",
    "- --action=trim, trim up and downstream sequences\n",
    "- -u 8,  trim 8 reads from the beginning of read\n",
    "- -u -4, trim 4 reads from the end of the read\n",
    "- --rename='{id}_{cut_prefix} {comment}', rename reads using TEMPLATE with variables such as {id}, {adapter_name}. (This is important as it is used by **umi** to remove pcr duplicates later in pipeline)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "d2e469da",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done           00:00:03    13,144,772 reads @   0.2 µs/read; 241.84 M reads/minute\n",
      "Done           00:00:01     3,907,314 reads @   0.3 µs/read; 185.23 M reads/minute\n",
      "Done           00:00:02     8,779,388 reads @   0.3 µs/read; 198.77 M reads/minute\n",
      "Done           00:00:01     5,636,274 reads @   0.3 µs/read; 201.67 M reads/minute\n",
      "Done           00:00:03    12,845,154 reads @   0.2 µs/read; 243.90 M reads/minute\n",
      "Done           00:00:02    14,555,266 reads @   0.2 µs/read; 293.99 M reads/minute\n",
      "Done           00:00:01     5,530,889 reads @   0.3 µs/read; 184.19 M reads/minute\n",
      "Done           00:00:02     7,068,411 reads @   0.3 µs/read; 189.63 M reads/minute\n"
     ]
    }
   ],
   "source": [
    "for sample in samples:\n",
    "    !cutadapt -j 0 -q 20 -m 20 --action=trim -u 8 -u -4 --rename='{{id}}_{{cut_prefix}} {{comment}}' \\\n",
    "              -o {fname(out_path,sample,\"fq\")}  \\\n",
    "              {fname(in_path,sample,\"fq\")} > {fname(out_path,sample,\"log\")}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9a469bf-82f1-4582-a4ae-8056b578067a",
   "metadata": {},
   "source": [
    "#### Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "c2a76d43-e024-4433-8fb2-4f68b06f6b33",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('adapter_trimmed/a4_2.fq')"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_fn  = fname(in_path,treated[1], 'fq')\n",
    "out_fn = fname(out_path,treated[1],'fq')\n",
    "in_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "83fb6298-bf01-47e5-aa2b-c75e236663d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def show_ends(reads):\n",
    "    for read in reads.split('\\n'):\n",
    "        display(HTML(f'<span style=\"color:blue\">{read[:8]}</span>{read[8:-4]}<span style=\"color:blue\">{read[-4:]}</span>'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "930edce2-0a4b-4170-a17c-1a1d1c1c7b0f",
   "metadata": {},
   "source": [
    "Show the 5' and 3' ends to be trimmed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "b33bb5e0-d7e3-4161-87d1-df2737385816",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<span style=\"color:blue\">ANTTGAAG</span>TGTGCTCTTC<span style=\"color:blue\">TAAT</span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"color:blue\">TNAAATCT</span>GCCCGG<span style=\"color:blue\">AGAG</span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"color:blue\">GNTCTACG</span>GAGACGGAGCAGAGCGAAAGCATTTGCCAAGAATGTTTTCATTAATCAAGGAC<span style=\"color:blue\">GAAA</span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"color:blue\">GNCTACCT</span>GGGCAGCTGCGGGGAAACGAAAGTCTTTGGGTTCCGGGGGGAGTATGGTTGCAA<span style=\"color:blue\">AGCT</span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"color:blue\">GNCCTACC</span>GCAGCTTCCGGGAAACCAAAGTCTTTGGGTTCCGGGGGGAGTATGGTTGCAAAG<span style=\"color:blue\">CTGA</span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"color:blue\">ANTTTGAG</span>ATGGTTGCAAAGCTGAAACTCAAAGG<span style=\"color:blue\">AAGT</span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"color:blue\">TNCCCTGA</span>GCGCAATAGATATAGGACCGC<span style=\"color:blue\">GGGA</span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"color:blue\">GNAACATA</span>GGGGAGTGGGGGTGCAAAGCTGAAACTTAAAGGAA<span style=\"color:blue\">AATC</span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<span style=\"color:blue\">GNAATGGA</span>GGGTAGCGGCGGTAGTAGGCCTAAAAGCAGCCACCAATTAAGAAAGGGTTCAAG<span style=\"color:blue\">CTCA</span>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "reads = !head -36 {in_fn} | seqtk seq -A |grep -v '>'\n",
    "reads = ('\\n').join(reads)\n",
    "show_ends(reads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "b509e2f0-fc0c-46f6-861e-e0f199f60f85",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "GAGACGGAGCAGAGCGAAAGCATTTGCCAAGAATGTTTTCATTAATCAAGGAC"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "GGGCAGCTGCGGGGAAACGAAAGTCTTTGGGTTCCGGGGGGAGTATGGTTGCAA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "GCAGCTTCCGGGAAACCAAAGTCTTTGGGTTCCGGGGGGAGTATGGTTGCAAAG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "ATGGTTGCAAAGCTGAAACTCAAAGG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "GCGCAATAGATATAGGACCGC"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "GGGGAGTGGGGGTGCAAAGCTGAAACTTAAAGGAA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "GGGTAGCGGCGGTAGTAGGCCTAAAAGCAGCCACCAATTAAGAAAGGGTTCAAG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "GAGCGAAAGCATTTGCCAAGAATGTTTTCATTAATCAAGT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTTTCCATTTAAACAAGCTCCTTAGCTGTTTAAAGCCACA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "reads = !head -36 {out_fn} | seqtk seq -A |grep -v '>'\n",
    "reads = ('\\n').join(reads)\n",
    "show_adapter(reads) # hack"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e51b20c-8064-46ac-b2fd-a375f765f3d8",
   "metadata": {},
   "source": [
    "- Both ends of reads have been trimmed\n",
    "- In several cases, the 5' end has been trimmed further due to poor reads, shown as extra N's\n",
    "- Short reads have been deleted"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e07bf322-c426-43eb-82e6-a6aced4d2bc8",
   "metadata": {},
   "source": [
    "Determine how many short reads have been deleted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "4b4e476b-a1e7-4b91-9594-90c240871fc1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[13144772, 3907314, 8779388, 5636274, 12845154, 14555266, 5530889, 7068411]"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Adapter Trimmed\n",
    "files = [fname(in_path,sample, \"fq\") for sample in samples]\n",
    "ins = [nseqs(f) for f in files]\n",
    "ins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "66e3f504-7207-49a8-8b5c-8114578ddc64",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[8463966, 2710933, 7345700, 4864834, 7500090, 5671357, 4618416, 6082913]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# UMI and Adapter Trimmed\n",
    "files = [fname(out_path,sample, \"fq\") for sample in samples]\n",
    "outs = [nseqs(f) for f in files]\n",
    "outs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "8918e05d-ac0b-47f1-92ec-1d275353ec7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjcAAAGsCAYAAAAhYYazAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABEBElEQVR4nO3deViU1f//8deAsiiKO0qiYG4QbkkLmqmVu2ZZaVnhgpYfd8lUMhW0sk1DK7dMUT9mWmmrqZS7ZiZBmfLJJAxTlLACNQWE+f3h1/k1gcbgDAM3z8d1zXVxnzn3ud8HJF6dexmT2Ww2CwAAwCBcnF0AAACAPRFuAACAoRBuAACAoRBuAACAoRBuAACAoRBuAACAoRBuAACAoRBuAACAoRBuAACAoRBuAACAoZTrcLNz50716dNHvr6+MplM+vDDD23aPyoqSiaTqcCrcuXKjikYAAD8q3Idbs6fP69WrVrpjTfeKNb+EydOVFpamtUrKChIDz30kJ0rBQAARVWuw02PHj303HPPqV+/foW+n5OTo0mTJumGG25Q5cqVddttt2n79u2W9728vFS3bl3L6/Tp0zp8+LDCw8NLaAYAAOCfKji7gNJsyJAhOnbsmN599135+vpqw4YN6t69uw4ePKgmTZoU6L906VI1bdpUHTp0cEK1AABAKucrN9eSnJysNWvW6L333lOHDh104403auLEibrjjju0fPnyAv2zs7O1evVqVm0AAHAyVm6u4ttvv5XZbFbTpk2t2rOzs1WzZs0C/devX6+zZ88qLCyspEoEAACFINxcRX5+vlxdXRUfHy9XV1er97y8vAr0X7p0qXr37q26deuWVIkAAKAQhJuraNOmjfLy8pSenv6v19CkpKRo27Zt+vjjj0uoOgAAcDXlOtycO3dOR48etWynpKQoMTFRNWrUUNOmTfXoo48qLCxMc+bMUZs2bZSRkaGtW7eqRYsW6tmzp2W/ZcuWqV69eurRo4czpgEAAP7GZDabzc4uwlm2b9+uzp07F2gfNGiQYmNjlZubq+eee04rV67UiRMnVLNmTYWGhio6OlotWrSQdPn0VcOGDRUWFqbnn3++pKcAAAD+oVyHGwAAYDzcCg4AAAyFcAMAAAyl3F1QnJ+fr5MnT6pKlSoymUzOLgcAABSB2WzW2bNn5evrKxeXa6/NlLtwc/LkSfn5+Tm7DAAAUAzHjx9X/fr1r9mn3IWbKlWqSLr8zalataqTqwEAAEWRlZUlPz8/y9/xayl34ebKqaiqVasSbgAAKGOKckkJFxQDAABDIdwAAABDIdwAAABDKXfX3ADAP+Xl5Sk3N9fZZQDlXsWKFeXq6nrd4xBuAJRr586d06+//io+iQZwPpPJpPr168vLy+u6xiHcACi38vLy9Ouvv6pSpUqqXbs2D/YEnMhsNuu3337Tr7/+qiZNmlzXCg7hBkC5lZubK7PZrNq1a8vT09PZ5QDlXu3atXXs2DHl5uZeV7jhgmIA5R4rNkDpYK/fRcINAAAwFMINAAAwFK65AYB/8J/yWYke79iLvRx/jGPHFBAQoISEBLVu3drhx3OG2NhYjR8/Xn/++aezSykyk8mkDRs26L777nN2KYbCyg0AoNg6deqk8ePHO2z8wYMHF/kP/4ABA3TkyBGH1YKyg5UbAIDT5eTkyM3Nrdj75+bmytPTk7veIImVGwAoczZt2qQ77rhD1apVU82aNdW7d28lJydb9dm/f7/atGkjDw8PhYSEKCEhwer9vLw8hYeHKyAgQJ6enmrWrJnmzZtn1efKqkl0dLTq1KmjqlWr6sknn1ROTo7l/R07dmjevHkymUwymUw6duyYJOnw4cPq2bOnvLy85OPjo8cff1wZGRmWsTt16qTRo0crIiJCtWrVUpcuXQrMMyoqSitWrNBHH31kGX/79u06duyYTCaT1q1bp06dOsnDw0P//e9/FRsbq2rVqlnt37p1ay1btkwNGjSQl5eX/vOf/ygvL08vv/yy6tatqzp16uj555+3Oq7JZNLixYvVu3dvVapUSYGBgfrqq6909OhRderUSZUrV1ZoaGiB7/knn3yitm3bysPDQ40aNVJ0dLQuXbpkef+nn37SnXfeKQ8PDwUFBSkuLu5fftIoLlZuAKCMOX/+vCIiItSiRQudP39e06dP1/3336/ExES5uLjo/Pnz6t27t+666y5Ne3WBThz/RWPHR0iSjpw+K5df/1Rubq4qVq2l519/W9Vq1NR38V9ryuQJynGrqm597pck/XE+R1988aUu5rto0ZqPdPLXVM18eoxq1aql559/XvPmzdORI0cUHBysmTNnSrr8nJK0tDR17NhRw4cP19y5c3XhwgVNnjxZ/fv319atWy3zWLFihf7zn/9oz549hT4heuLEiUpKSlJWVpaWL18uSapRo4ZOnjwpSZo8ebLmzJmj5cuXy93dXVu2bCkwRnJysj7//HNt2rRJycnJevDBB5WSkqKmTZtqx44d2rt3r4YOHaq7775bt99+u2W/WbNmae7cuZo7d64mT56sgQMHqlGjRoqMjFSDBg00dOhQjR49Wp9//rkkafPmzXrsscc0f/58dejQQcnJyXriiSckSTNmzFB+fr769eunWrVqad++fcrKynLo6bzyjnADAGXMAw88YLX99ttvq06dOjp8+LCCg4O1evVq5eXladmyZTr6e44aNwvU6bSTev6Zpyz7VKxYUSOfirRs12/QUN8d2K8tn35oCTdX+kXPeUOenpXUuFmgZs6cqaefflqzZs2St7e33NzcVKlSJdWtW9eyz8KFC3XzzTfrhRdesLQtW7ZMfn5+OnLkiJo2bSpJaty4sV5++eWrztPLy0uenp7Kzs62Gv+K8ePHq1+/ftf8XuXn52vZsmWqUqWKgoKC1LlzZ/3444/auHGjXFxc1KxZM7300kvavn27VbgZMmSI+vfvL+lyiAoNDdW0adPUrVs3SdK4ceM0ZMgQS//nn39eU6ZM0aBBgyRJjRo10qxZszRp0iTNmDFDX3zxhZKSknTs2DHVr19fkvTCCy+oR48e16wfxUO4AYAyJjk5WdOmTdO+ffuUkZGh/Px8SVJqaqqCg4OVlJSkVq1aqVKlStLvl08htWp7S4Fx1q1apg1rVintxHFdvHhRubk5ahbUwqpP06BgeXpWsmyHhobq3LlzOn78uBo2bFhoffHx8dq2bVuhnw+UnJxsCTchISHF+wb8n6Ls7+/vrypVqli2fXx85OrqKhcXF6u29PR0q/1atmxp9b4ktWjRwqrt4sWLysrKUtWqVRUfH69vvvnG6hRXXl6eLl68qL/++ktJSUlq0KCBJdhIl7+XcAzCDQCUMX369JGfn5/eeust+fr6Kj8/X8HBwZZrYYryIaCbP9mgV6On6qlps9Sy7a2qXNlLsYvn64eE+CLVcK0nyebn56tPnz566aWXCrxXr149y9eVK1cu0rGupij7V6xY0WrbZDIV2nYlIBa235W5FtZ2Zb/8/HxFR0cXupLk4eFR6M+EJ2M7DuEGAMqQM2fOKCkpSYsXL1aHDh0kSbt377bqExQUpFWrVunChQuWtu+/PWDV59v9X6lVyK0aMGiYpe3XX44VON6Rwz/o4oUL8vi/u5D27dsnLy8vywqEm5ub8vLyrPa5+eab9cEHH8jf318VKlzfn5nCxi+Nbr75Zv34449q3Lhxoe8HBQUpNTVVJ0+elK+vryTpq6++KskSyxXulgKAMqR69eqqWbOmlixZoqNHj2rr1q2KiIiw6jNw4EC5uLgoPDxcyUf+p11bt2jl4jes+jTwb6TD3ydoz/Yvdezno3rjled16LtvCxwvNzdXUU+PUfKR/2n3tjjNmDFDo0ePtpzW8ff319dff61jx45ZTpGNGjVKv//+ux555BHt379fP//8s7Zs2aKhQ4faHFT8/f31/fff68cff1RGRoZyc3Nt/I6VjOnTp2vlypWKiorSoUOHlJSUpLVr1+rZZ5+VJN1zzz1q1qyZwsLC9N1332nXrl2aOnWqk6s2LlZuAOAfSuKJwcXl4uKid999V2PHjlVwcLCaNWum+fPnq1OnTpY+Xl5e+uSTTzRixAgN6NFRjZo007hnovTUE2GWPg89NkT/O3RQk0cNlUwm9bj3AfUPC9eebV9YHe+29neqQcCNGvpgL+Xk5GjgIw8rKirK8v7EiRM1aNAgBQUF6cKFC0pJSZG/v7/27NmjyZMnq1u3bsrOzlbDhg3VvXt3q2tdimL48OHavn27QkJCdO7cOW3btk3+/v7F+dY5VLdu3fTpp59q5syZevnll1WxYkU1b95cw4ZdXhlzcXHRhg0bFB4erltvvVX+/v6aP3++unfv7uTKjclkLsrJWQPJysqSt7e3MjMzVbVqVWeXA8CJLl68qJSUFAUEBMjDw8PZ5TjE97/+Wex9p00YqbNZmYp5e7WlrWX9atdfFHAV1/qdtOXvN6elAACAoRBuAACAoTg13OzcuVN9+vSRr6+vTCaTPvzwwyLvu2fPHlWoUMGwn24LAM4267UFVqekgLLCqeHm/PnzatWqld54441/7/w3mZmZCgsL09133+2gygAAQFnl1LulevToUaxHTz/55JMaOHCgXF1dbVrtAQAAxlfmrrlZvny5kpOTNWPGjCL1z87OVlZWltULAAAYV5kKNz/99JOmTJmi1atXF/mpl7Nnz5a3t7fl5efn5+AqAQCAM5WZcJOXl6eBAwcqOjra8qFrRREZGanMzEzL6/jx4w6sEgAAOFuZeULx2bNndeDAASUkJGj06NGSLn9QmdlsVoUKFbRlyxbdddddBfZzd3eXu7t7SZcLAACcpMyEm6pVq+rgwYNWbQsWLNDWrVv1/vvvKyAgwEmVATCcKO8SPl5myR7vKhbOfVHbNn+mdZt3lfix/f39NX78eI0fP96u45pMJm3YsEH33XefXcctLbZv367OnTvrjz/+ULVq1ZxdTpE46mf9d04NN+fOndPRo0ct2ykpKUpMTFSNGjXUoEEDRUZG6sSJE1q5cqVcXFwUHBxstX+dOnXk4eFRoB0AYLtBT47WI0OecHYZpd7gwYP1559/Ouxu3aioKH344YdKTEz8177t2rVTWlqavL1LOJCXck4NNwcOHFDnzp0t21c+2XbQoEGKjY1VWlqaUlNTnVUeAJQrlSp7qVJlZ1dRfuTk5MjNza3Y++fm5srNzU1169a1Y1XG4NQLijt16iSz2VzgFRsbK0mKjY3V9u3br7p/VFRUkZItABiJv7+/YmJirNpat25t9WndJpNJixcv1ujBA3RbE1/d1/k2fRe/X6kpPyv8od66rekNerxvVx0/lmLZZ+HcF9W/W4erHjcvL0/h4eEKCAiQp6enmjVrpnnz5ln1GTx4sO677z69+uqrqlevnmrWrKlRo0YpNzfX0ic9PV19+vSRp6enAgICtHr1vz8F+ZtvvlGXLl1Uq1YteXt7q2PHjvr222+t+vz000+688475eHhoaCgIMXFxRUYZ/LkyWratKkqVaqkRo0aadq0aVa1RUVFqXXr1lq8eLH8/PxUqVIlPfTQQ/rzzz8t769YsUIfffSRTCaTTCaT5e/UiRMnNGDAAFWvXl01a9ZU3759dezYsQLfm9mzZ8vX17fQm2NiY2MVHR2t7777zjL+lb+JJpNJixYtUt++fVW5cmU999xz2r59u0wmk6W+2NhYVatWTZ9++qmaNWumSpUq6cEHH9T58+e1YsUK+fv7q3r16hozZozy8vIsx/X399dzzz2nsLAweXl5qWHDhvroo4/022+/qW/fvvLy8lKLFi104MABq3r37t2rO++8U56envLz89PYsWN1/vz56/pZ20OZuVsKAGCbWbNmqc8DD2vt5p3yv7GJpowZrlmREzR01ASt+WyrJGn2tKeLPF5+fr7q16+vdevW6fDhw5o+fbqeeeYZrVu3zqrftm3blJycrG3btmnFihWKjY21/IGWLv+RP3bsmOWayQULFig9Pf2axz579qwGDRqkXbt2ad++fWrSpIl69uyps2fPWmrr16+fXF1dtW/fPi1atEiTJ08uME6VKlUUGxurw4cPa968eXrrrbf02muvWfU5evSo1q1bp08++USbNm1SYmKiRo0aJUmaOHGi+vfvr+7duystLU1paWlq166d/vrrL3Xu3FleXl7auXOndu/eLS8vL3Xv3l05OTmWsb/88kslJSUpLi5On376aYH6BgwYoKeeeko33XSTZfwBAwZY3p8xY4b69u2rgwcPaujQoYV+r/766y/Nnz9f7777rjZt2qTt27erX79+2rhxozZu3KhVq1ZpyZIlev/99632e+2119S+fXslJCSoV69eevzxxxUWFqbHHntM3377rRo3bqywsDCZzWZJ0sGDB9WtWzf169dP33//vdauXavdu3dbbvop7s/aHsrMBcUAANsMGTJE3frcL0kaOnKcHu/bVU+MfVrtO13+6JqBQ5/UjKdGX2sIKxUrVlR0dLRlOyAgQHv37tW6devUv39/S3v16tX1xhtvyNXVVc2bN1evXr305Zdfavjw4Tpy5Ig+//xz7du3T7fddpsk6e2331ZgYOA1j/3Pu2EXL16s6tWra8eOHerdu7e++OILJSUl6dixY6pfv74k6YUXXijwFPxnn33W8rW/v7+eeuoprV27VpMmTbK0X7x4UStWrLCM8/rrr6tXr16aM2eO6tatK09PT2VnZ1udDvrvf/8rFxcXLV26VCaTSdLlh85Wq1ZN27dvV9euXSVJlStX1tKlS696OsrT01NeXl6qUKFCoaebBg4caBVqUlJSCvTJzc3VwoULdeONN0qSHnzwQa1atUqnT5+Wl5eXgoKC1LlzZ23bts0qOPXs2VNPPvmkJGn69OlauHChbrnlFj300EOSLq96hYaG6vTp06pbt65eeeUVDRw40HJhcJMmTTR//nx17NhRCxcuVGpqarF+1vZAuAEAg2rZsqXl6xq16kiSmjQPsrTVrF1H2dkXde5slryqVC3SmIsWLdLSpUv1yy+/6MKFC8rJySnwAcY33XSTXF1dLdv16tWz3O2alJSkChUqKCQkxPJ+8+bN//VOn/T0dE2fPl1bt27V6dOnlZeXp7/++styXWZSUpIaNGhgCSSSFBoaWmCc999/XzExMTp69KjOnTunS5cuqWpV67kXNk5+fr5+/PHHq17fEh8fr6NHj6pKlSpW7RcvXlRycrJlu0WLFtd1nc3fv29XU6lSJUuwkSQfHx/5+/vLy8vLqu2fKyh///fi4+Njqfefbenp6apbt65lzn8/1WQ2m5Wfn6+UlBQdOXKkWD9reyDcAEAZ4+LiYjk1cMXfrxu5omLFipavr6wmVCikLT/feqyr2fzJBk2PmKA5c+YoNDRUVapU0SuvvKKvv/76qse9cpz8/HxJstR95dhFNXjwYP3222+KiYlRw4YN5e7urtDQUMspn39+Pwo7xr59+/Twww8rOjpa3bp1k7e3t959913NmTPnmse+Ms61as7Pz1fbtm0Lvaakdu3alq8rV76+K7aLsn9h3/9r/UwK2+/KXAtru7Jffn6+nnzySY0dO7ZADQ0aNNCPP/5otV9JItwAQBlTu3ZtpaWlWbazsrIKPT1hb9/u/0rt2rXTyJEjLW1/X5UoisDAQF26dEkHDhzQrbfeKkn68ccfLRfEXs2uXbu0YMEC9ezZU5J0/PhxZWRkWN4PCgpSamqqTp48KV9fX0nSV199ZTXGnj171LBhQ02dOtXS9ssvvxQ4VmHjuLi4WC4AdnNzs7oYV5JuvvlmrV27VnXq1CmwEmSrwsYvjW6++WYdOnRIjRs3LvT94v6s7YELigGgjLnrrru0atUq7dq1Sz/88IMGDRpkdRrIURr4N9KBAwe0efNmHTlyRNOmTdM333xj0xjNmjVT9+7dNXz4cH399deKj4/XsGHD5Onpec39GjdurFWrVikpKUlff/21Hn30Uat97rnnHjVr1kxhYWH67rvvtGvXLqsQc2WM1NRUvfvuu0pOTtb8+fO1YcOGAsfy8PDQoEGDLOOMHTtW/fv3t5yS8vf31/fff68ff/xRGRkZys3N1aOPPqpatWqpb9++2rVrl1JSUrRjxw6NGzdOv/76q03fI39/f8tz3zIyMpSdnW3T/iVl8uTJ+uqrrzRq1CglJibqp59+0scff6wxY8ZIKv7P2h5YuQGAfyolTwy+msjISP3888/q3bu3vL29NWvWrBJZuXnosSFKP/ajBgwYIJPJpEceeUQjR47U559/btM4y5cv17Bhw9SxY0f5+Pjoueee07Rp0665z7Jly/TEE0+oTZs2atCggV544QVNnDjR8r6Li4s2bNig8PBw3XrrrfL399f8+fPVvXt3S5++fftqwoQJGj16tLKzs9WrVy9NmzbN6hZ66XII6tevn3r27Knff/9dPXv21IIFCyzvDx8+XNu3b1dISIjOnTunbdu2qVOnTtq5c6cmT56sfv366ezZs7rhhht0991327yS88ADD2j9+vXq3Lmz/vzzTy1fvlyDBw+2aYyS0LJlS+3YsUNTp05Vhw4dZDabdeONN1pdpFycn7U9mMyFnag0sKysLHl7eyszM/O6lw4BlG0XL15USkqKAgIC5OHh4exyHOL7X/+063gt61ez63iljS1PB4b9Xet30pa/35yWAgAAhkK4AQAAhkK4AQDg//CxPsZAuAEAAIZCuAFQ7pWz+yqAUstev4vcCm5n/lM+s+t4x17sZdfxAPx/V54Nk5OTUyLP3gBwbVeeOH29z20i3AAotypUqKBKlSrpt99+U8WKFeXiYrzFbPOlnH/vZIOLFy/adTzgivz8fP3222+qVKmSKlS4vnhCuAFQbplMJtWrV08pKSmFPobfCNL/uGDX8dwusMIFx3FxcVGDBg2u+/OoCDcAyjU3Nzc1adLEshxuNMPWb7freF8+1cmu4wF/5+bmZpcVVMINgHLPxcXFsE8oPnHWvh/AaNTvE4zFeCeYAQBAuUa4AQAAhkK4AQAAhsI1NwDKHXs/j0rimVRAacLKDQAAMBTCDQAAMBTCDQAAMBTCDQAAMBTCDQAAMBTCDQAAMBTCDQAAMBTCDQAAMBTCDQAAMBTCDQAAMBTCDQAAMBTCDQAAMBTCDQAAMBTCDQAAMBTCDQAAMBSnhpudO3eqT58+8vX1lclk0ocffnjN/uvXr1eXLl1Uu3ZtVa1aVaGhodq8eXPJFAsAAMoEp4ab8+fPq1WrVnrjjTeK1H/nzp3q0qWLNm7cqPj4eHXu3Fl9+vRRQkKCgysFAABlRQVnHrxHjx7q0aNHkfvHxMRYbb/wwgv66KOP9Mknn6hNmzZ2rg4AAJRFTg031ys/P19nz55VjRo1rtonOztb2dnZlu2srKySKA0AADhJmb6geM6cOTp//rz69+9/1T6zZ8+Wt7e35eXn51eCFQIAgJJWZsPNmjVrFBUVpbVr16pOnTpX7RcZGanMzEzL6/jx4yVYJQAAKGll8rTU2rVrFR4ervfee0/33HPPNfu6u7vL3d29hCoDAADOVuZWbtasWaPBgwfrnXfeUa9evZxdDgAAKGWcunJz7tw5HT161LKdkpKixMRE1ahRQw0aNFBkZKROnDihlStXSrocbMLCwjRv3jzdfvvtOnXqlCTJ09NT3t7eTpkDAAAoXZy6cnPgwAG1adPGcht3RESE2rRpo+nTp0uS0tLSlJqaaum/ePFiXbp0SaNGjVK9evUsr3HjxjmlfgAAUPo4deWmU6dOMpvNV30/NjbWanv79u2OLQgAAJR5Ze6aGwAAgGsh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAENxarjZuXOn+vTpI19fX5lMJn344Yf/us+OHTvUtm1beXh4qFGjRlq0aJHjCwUAAGWGU8PN+fPn1apVK73xxhtF6p+SkqKePXuqQ4cOSkhI0DPPPKOxY8fqgw8+cHClAACgrKjgzIP36NFDPXr0KHL/RYsWqUGDBoqJiZEkBQYG6sCBA3r11Vf1wAMPOKhKAABQlti8cnP8+HH9+uuvlu39+/dr/PjxWrJkiV0LK8xXX32lrl27WrV169ZNBw4cUG5ubqH7ZGdnKysry+oFAACMy+ZwM3DgQG3btk2SdOrUKXXp0kX79+/XM888o5kzZ9q9wL87deqUfHx8rNp8fHx06dIlZWRkFLrP7Nmz5e3tbXn5+fk5tEYAAOBcNoebH374Qbfeeqskad26dQoODtbevXv1zjvvKDY21t71FWAymay2zWZzoe1XREZGKjMz0/I6fvy4w2sEAADOY/M1N7m5uXJ3d5ckffHFF7r33nslSc2bN1daWpp9q/uHunXr6tSpU1Zt6enpqlChgmrWrFnoPu7u7pZ6AQCA8dm8cnPTTTdp0aJF2rVrl+Li4tS9e3dJ0smTJ68aMOwlNDRUcXFxVm1btmxRSEiIKlas6NBjAwCAssHmcPPSSy9p8eLF6tSpkx555BG1atVKkvTxxx9bTlcV1blz55SYmKjExERJl2/1TkxMVGpqqqTLp5TCwsIs/UeMGKFffvlFERERSkpK0rJly/T2229r4sSJtk4DAAAYlM2npTp16qSMjAxlZWWpevXqlvYnnnhClSpVsmmsAwcOqHPnzpbtiIgISdKgQYMUGxurtLQ0S9CRpICAAG3cuFETJkzQm2++KV9fX82fP5/bwAEAgEWxnnPj6upqFWwkyd/f3+ZxOnXqZLkguDCFXaDcsWNHffvttzYfCwAAlA9FCjdt2rS56t1I/0TwAAAAzlSkcHPfffdZvr548aIWLFigoKAghYaGSpL27dunQ4cOaeTIkQ4pEgAAoKiKFG5mzJhh+XrYsGEaO3asZs2aVaAPz5ABAADOZvPdUu+9957VHUxXPPbYY3yAJQAAcDqbw42np6d2795doH337t3y8PCwS1EAAADFZfPdUuPHj9d//vMfxcfH6/bbb5d0+ZqbZcuWafr06XYvEAAAwBY2h5spU6aoUaNGmjdvnt555x1JUmBgoGJjY9W/f3+7FwgAAGCLYj3npn///gQZAABQKtl8zQ0AAEBpZvPKTV5enl577TWtW7dOqampysnJsXr/999/t1txAAAAtrJ55SY6Olpz585V//79lZmZqYiICPXr108uLi6KiopyQIkAAABFZ3O4Wb16td566y1NnDhRFSpU0COPPKKlS5dq+vTp2rdvnyNqBAAAKDKbw82pU6fUokULSZKXl5cyMzMlSb1799Znn31m3+oAAABsZHO4qV+/vtLS0iRJjRs31pYtWyRJ33zzjdzd3e1bHQAAgI1sDjf333+/vvzyS0nSuHHjNG3aNDVp0kRhYWEaOnSo3QsEAACwhc13S7344ouWrx988EHVr19fe/fuVePGjXXvvffatTgAAABbFeshfn93++23Wz6GAQAAwNmK9RC/VatWqX379vL19dUvv/wiSYqJidFHH31k1+IAAABsZXO4WbhwoSIiItSzZ0/9+eefysvLkyRVq1ZNMTEx9q4PAADAJjaHm9dff11vvfWWpk6dKldXV0t7SEiIDh48aNfiAAAAbGVzuElJSVGbNm0KtLu7u+v8+fN2KQoAAKC4bA43AQEBSkxMLND++eefKygoyB41AQAAFJvNd0s9/fTTGjVqlC5evCiz2az9+/drzZo1mj17tpYuXeqIGgEAAIrM5nAzZMgQXbp0SZMmTdJff/2lgQMH6oYbbtC8efP08MMPO6JGwKn8p9j/Y0WOvdjL7mMCAC4r1nNuhg8fruHDhysjI0P5+fmqU6eOJOnEiRO64YYb7FogAACALYr1nJsratWqpTp16ujUqVMaM2aMGjdubK+6AAAAiqXI4ebPP//Uo48+qtq1a8vX11fz589Xfn6+pk+frkaNGmnfvn1atmyZI2sFAAD4V0U+LfXMM89o586dGjRokDZt2qQJEyZo06ZNunjxoj7//HN17NjRkXUCAAAUSZHDzWeffably5frnnvu0ciRI9W4cWM1bdqUpxIDAIBSpcinpU6ePGl5jk2jRo3k4eGhYcOGOawwAACA4ihyuMnPz1fFihUt266urqpcubJDigIAACiuIp+WMpvNGjx4sNzd3SVJFy9e1IgRIwoEnPXr19u3wvIuytsBY2baf0wAAEqJIoebQYMGWW0/9thjdi8GAADgehU53CxfvtyRdQAAANjFdT3EDwAAoLQh3AAAAEMh3AAAAEMh3AAAAEMpUri5+eab9ccff0iSZs6cqb/++stuBSxYsEABAQHy8PBQ27ZttWvXrmv2X716tVq1aqVKlSqpXr16GjJkiM6cOWO3egAAQNlWpHCTlJSk8+fPS5Kio6N17tw5uxx87dq1Gj9+vKZOnaqEhAR16NBBPXr0UGpqaqH9d+/erbCwMIWHh+vQoUN677339M033/CkZAAAYFGkW8Fbt26tIUOG6I477pDZbNarr74qLy+vQvtOnz69yAefO3euwsPDLeEkJiZGmzdv1sKFCzV79uwC/fft2yd/f3+NHTtWkhQQEKAnn3xSL7/8cpGPCQAAjK1I4SY2NlYzZszQp59+KpPJpM8//1wVKhTc1WQyFTnc5OTkKD4+XlOmTLFq79q1q/bu3VvoPu3atdPUqVO1ceNG9ejRQ+np6Xr//ffVq1evqx4nOztb2dnZlu2srKwi1QcAAMqmIoWbZs2a6d1335Ukubi46Msvv1SdOnWu68AZGRnKy8uTj4+PVbuPj49OnTpV6D7t2rXT6tWrNWDAAF28eFGXLl3Svffeq9dff/2qx5k9e7aio6Ovq1YAAFB22Hy3VH5+/nUHm78zmUxW22azuUDbFYcPH9bYsWM1ffp0xcfHa9OmTUpJSdGIESOuOn5kZKQyMzMtr+PHj9utdgAAUPoU+eMX/i45OVkxMTFKSkqSyWRSYGCgxo0bpxtvvLHIY9SqVUuurq4FVmnS09MLrOZcMXv2bLVv315PP/20JKlly5aqXLmyOnTooOeee0716tUrsI+7u7vlwz4BAIDx2bxys3nzZgUFBWn//v1q2bKlgoOD9fXXX+umm25SXFxckcdxc3NT27ZtC+wTFxendu3aFbrPX3/9JRcX65JdXV0lXV7xAQAAsHnlZsqUKZowYYJefPHFAu2TJ09Wly5dijxWRESEHn/8cYWEhCg0NFRLlixRamqq5TRTZGSkTpw4oZUrV0qS+vTpo+HDh2vhwoXq1q2b0tLSNH78eN16663y9fW1dSoAAMCAbA43SUlJWrduXYH2oUOHKiYmxqaxBgwYoDNnzmjmzJlKS0tTcHCwNm7cqIYNG0qS0tLSrJ55M3jwYJ09e1ZvvPGGnnrqKVWrVk133XWXXnrpJVunAQAADMrmcFO7dm0lJiaqSZMmVu2JiYnFutB45MiRGjlyZKHvxcbGFmgbM2aMxowZY/NxAABA+WBzuBk+fLieeOIJ/fzzz2rXrp1MJpN2796tl156SU899ZQjagQAACgym8PNtGnTVKVKFc2ZM0eRkZGSJF9fX0VFRVmeHAwAAOAsNocbk8mkCRMmaMKECTp79qwkqUqVKnYvDAAAoDiK9ZybKwg1AACgtLH5OTcAAACl2XWt3AAA/k+Ut53Hy7TveEA5wsoNAAAwFJvCTW5urjp37qwjR444qh4AAIDrYlO4qVixon744Yerfmo3AACAs9l8zU1YWJjefvvtAp8tBcAGXJ8BAA5jc7jJycnR0qVLFRcXp5CQEFWuXNnq/blz59qtOAAAAFvZHG5++OEH3XzzzZJU4NobTlcBAABnszncbNu2zRF1AAAA2EWxbwU/evSoNm/erAsXLkiSzGaz3YoCAAAoLpvDzZkzZ3T33XeradOm6tmzp9LS0iRJw4YN41PBAQCA09l8WmrChAmqWLGiUlNTFRgYaGkfMGCAJkyYoDlz5ti1QAAA8P/5T/nMruMde7GXXccrDWwON1u2bNHmzZtVv359q/YmTZrol19+sVthAAAAxWHzaanz58+rUqVKBdozMjLk7u5ul6IAAACKy+Zwc+edd2rlypWWbZPJpPz8fL3yyivq3LmzXYsDAACwlc2npV555RV16tRJBw4cUE5OjiZNmqRDhw7p999/1549exxRIwAAQJHZvHITFBSk77//Xrfeequ6dOmi8+fPq1+/fkpISNCNN97oiBoBAACKzOaVG0mqW7euoqOj7V0LAADAdStWuPnjjz/09ttvKykpSSaTSYGBgRoyZIhq1Khh7/oAAABsYvNpqR07diggIEDz58/XH3/8od9//13z589XQECAduzY4YgaAQAAiszmlZtRo0apf//+WrhwoVxdXSVJeXl5GjlypEaNGqUffvjB7kUCAAAUlc0rN8nJyXrqqacswUaSXF1dFRERoeTkZLsWBwAAYCubw83NN9+spKSkAu1JSUlq3bq1PWoCAAAotiKdlvr+++8tX48dO1bjxo3T0aNHdfvtt0uS9u3bpzfffFMvvviiY6oEAAAooiKFm9atW8tkMslsNlvaJk2aVKDfwIEDNWDAAPtVBwAAYKMihZuUlBRH1wEAAGAXRQo3DRs2dHQdAAAAdlGsh/idOHFCe/bsUXp6uvLz863eGzt2rF0KAwAAKA6bw83y5cs1YsQIubm5qWbNmjKZTJb3TCYT4QYAADiVzeFm+vTpmj59uiIjI+XiYvOd5AAAAA5lczr566+/9PDDDxNsAABAqWRzQgkPD9d7773niFoAAACum82npWbPnq3evXtr06ZNatGihSpWrGj1/ty5c+1WHAAAgK1sXrl54YUXtHnzZp0+fVoHDx5UQkKC5ZWYmGhzAQsWLFBAQIA8PDzUtm1b7dq165r9s7OzNXXqVDVs2FDu7u668cYbtWzZMpuPCwAAjMnmlZu5c+dq2bJlGjx48HUffO3atRo/frwWLFig9u3ba/HixerRo4cOHz6sBg0aFLpP//79dfr0ab399ttq3Lix0tPTdenSpeuuBQAAGIPN4cbd3V3t27e3y8Hnzp2r8PBwDRs2TJIUExOjzZs3a+HChZo9e3aB/ps2bdKOHTv0888/q0aNGpIkf39/u9QCAEC5FOXtgDEz7T+mDWw+LTVu3Di9/vrr133gnJwcxcfHq2vXrlbtXbt21d69ewvd5+OPP1ZISIhefvll3XDDDWratKkmTpyoCxcuXPU42dnZysrKsnoBAADjsnnlZv/+/dq6das+/fRT3XTTTQUuKF6/fn2RxsnIyFBeXp58fHys2n18fHTq1KlC9/n555+1e/dueXh4aMOGDcrIyNDIkSP1+++/X/W6m9mzZys6OrpINQEAgLLP5nBTrVo19evXz24F/P0Jx5JkNpsLtF2Rn58vk8mk1atXy9v78jLa3Llz9eCDD+rNN9+Up6dngX0iIyMVERFh2c7KypKfn5/d6gcAAKVLsT5+wR5q1aolV1fXAqs06enpBVZzrqhXr55uuOEGS7CRpMDAQJnNZv36669q0qRJgX3c3d3l7u5ul5oBAEDp57THDLu5ualt27aKi4uzao+Li1O7du0K3ad9+/Y6efKkzp07Z2k7cuSIXFxcVL9+fYfWCwAAygabV24CAgKuetpIunxdTFFFRETo8ccfV0hIiEJDQ7VkyRKlpqZqxIgRki6fUjpx4oRWrlwpSRo4cKBmzZqlIUOGKDo6WhkZGXr66ac1dOjQQk9JAQCA8sfmcDN+/Hir7dzcXCUkJGjTpk16+umnbRprwIABOnPmjGbOnKm0tDQFBwdr48aNatiwoSQpLS1Nqamplv5eXl6Ki4vTmDFjFBISopo1a6p///567rnnbJ0GAAAwKJvDzbhx4wptf/PNN3XgwAGbCxg5cqRGjhxZ6HuxsbEF2po3b17gVBYAAMAVdrvmpkePHvrggw/sNRwAAECx2LxyczXvv/++5anBAACDMuDTbGE8NoebNm3aWF1QbDabderUKf32229asGCBXYsDULL8p3xm1/GOvdjLruMBQFHYHG7uu+8+q20XFxfVrl1bnTp1UvPmze1VFwAAQLHYHG5mzJjhiDoAAADswm7X3AAA4Cz2PqUqcVq1LCtyuHFxcbnmw/uky58TdenSpesuCgAAoLiKHG42bNhw1ff27t2r119/XWaz2S5FAQAAFFeRw03fvn0LtP3vf/9TZGSkPvnkEz366KOaNWuWXYsDAACwVbEe4nfy5EkNHz5cLVu21KVLl5SYmKgVK1aoQYMG9q4PAADAJjaFm8zMTE2ePFmNGzfWoUOH9OWXX+qTTz5RcHCwo+oDAACwSZFPS7388st66aWXVLduXa1Zs6bQ01QAAADOVuRwM2XKFHl6eqpx48ZasWKFVqxYUWi/9evX2604AAAAWxU53ISFhf3rreAAAADOVuRwExsb68AyAAAA7KNYd0sBAACUVoQbAABgKIQbAABgKIQbAABgKIQbAABgKIQbAABgKEW+FRwAbBbl7YAxM+0/JgBDYeUGAAAYCuEGAAAYCuEGAAAYCuEGAAAYCuEGAAAYCndLAQBQGHvf7cedfiWGlRsAAGAohBsAAGAohBsAAGAohBsAAGAohBsAAGAohBsAAGAohBsAAGAohBsAAGAohBsAAGAohBsAAGAoTg83CxYsUEBAgDw8PNS2bVvt2rWrSPvt2bNHFSpUUOvWrR1bIAAAKFOcGm7Wrl2r8ePHa+rUqUpISFCHDh3Uo0cPpaamXnO/zMxMhYWF6e677y6hSgEAQFnh1HAzd+5chYeHa9iwYQoMDFRMTIz8/Py0cOHCa+735JNPauDAgQoNDS2hSgEAQFnhtHCTk5Oj+Ph4de3a1aq9a9eu2rt371X3W758uZKTkzVjxowiHSc7O1tZWVlWLwAAYFxOCzcZGRnKy8uTj4+PVbuPj49OnTpV6D4//fSTpkyZotWrV6tChQpFOs7s2bPl7e1tefn5+V137QAAoPRy+gXFJpPJattsNhdok6S8vDwNHDhQ0dHRatq0aZHHj4yMVGZmpuV1/Pjx664ZAACUXkVb/nCAWrVqydXVtcAqTXp6eoHVHEk6e/asDhw4oISEBI0ePVqSlJ+fL7PZrAoVKmjLli266667Cuzn7u4ud3d3x0yinPOf8pldxzv2Yi+7jgcAKJ+ctnLj5uamtm3bKi4uzqo9Li5O7dq1K9C/atWqOnjwoBITEy2vESNGqFmzZkpMTNRtt91WUqUDAIBSzGkrN5IUERGhxx9/XCEhIQoNDdWSJUuUmpqqESNGSLp8SunEiRNauXKlXFxcFBwcbLV/nTp15OHhUaAdAACUX04NNwMGDNCZM2c0c+ZMpaWlKTg4WBs3blTDhg0lSWlpaf/6zBsAAIC/c2q4kaSRI0dq5MiRhb4XGxt7zX2joqIUFRVl/6IAAECZ5fS7pQAAAOyJcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAyFcAMAAAylgrMLACyivO08XqZ9xwMAlAms3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAEMh3AAAAENxerhZsGCBAgIC5OHhobZt22rXrl1X7bt+/Xp16dJFtWvXVtWqVRUaGqrNmzeXYLUAAKC0c2q4Wbt2rcaPH6+pU6cqISFBHTp0UI8ePZSamlpo/507d6pLly7auHGj4uPj1blzZ/Xp00cJCQklXDkAACitnBpu5s6dq/DwcA0bNkyBgYGKiYmRn5+fFi5cWGj/mJgYTZo0SbfccouaNGmiF154QU2aNNEnn3xSwpUDAIDSymnhJicnR/Hx8eratatVe9euXbV3794ijZGfn6+zZ8+qRo0aV+2TnZ2trKwsqxcAADAup4WbjIwM5eXlycfHx6rdx8dHp06dKtIYc+bM0fnz59W/f/+r9pk9e7a8vb0tLz8/v+uqGwAAlG5Ov6DYZDJZbZvN5gJthVmzZo2ioqK0du1a1alT56r9IiMjlZmZaXkdP378umsGAAClVwVnHbhWrVpydXUtsEqTnp5eYDXnn9auXavw8HC99957uueee67Z193dXe7u7tddLwAAKBuctnLj5uamtm3bKi4uzqo9Li5O7dq1u+p+a9as0eDBg/XOO++oV69eji4TAACUMU5buZGkiIgIPf744woJCVFoaKiWLFmi1NRUjRgxQtLlU0onTpzQypUrJV0ONmFhYZo3b55uv/12y6qPp6envL29nTYPAABQejg13AwYMEBnzpzRzJkzlZaWpuDgYG3cuFENGzaUJKWlpVk982bx4sW6dOmSRo0apVGjRlnaBw0apNjY2JIuHwAAlEJODTeSNHLkSI0cObLQ9/4ZWLZv3+74ggAAQJnm9LulAAAA7IlwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADIVwAwAADMXp4WbBggUKCAiQh4eH2rZtq127dl2z/44dO9S2bVt5eHioUaNGWrRoUQlVCgAAygKnhpu1a9dq/Pjxmjp1qhISEtShQwf16NFDqamphfZPSUlRz5491aFDByUkJOiZZ57R2LFj9cEHH5Rw5QAAoLRyariZO3euwsPDNWzYMAUGBiomJkZ+fn5auHBhof0XLVqkBg0aKCYmRoGBgRo2bJiGDh2qV199tYQrBwAApVUFZx04JydH8fHxmjJlilV7165dtXfv3kL3+eqrr9S1a1ertm7duuntt99Wbm6uKlasWGCf7OxsZWdnW7YzMzMlSVlZWdc7hULlZ/9l1/GyTGa7jnd5UPvMvdTPtZTOUyo/cy2t/375mRYfP9PrGZCf6fUNeXlMs7kI9Zqd5MSJE2ZJ5j179li1P//88+amTZsWuk+TJk3Mzz//vFXbnj17zJLMJ0+eLHSfGTNmmCXx4sWLFy9evAzwOn78+L9mDKet3FxhMpmsts1mc4G2f+tfWPsVkZGRioiIsGzn5+fr999/V82aNa95nNIgKytLfn5+On78uKpWrerschyKuRpPeZmnVH7mWl7mKZWfuZaleZrNZp09e1a+vr7/2tdp4aZWrVpydXXVqVOnrNrT09Pl4+NT6D5169YttH+FChVUs2bNQvdxd3eXu7u7VVu1atWKX7gTVK1atdT/o7MX5mo85WWeUvmZa3mZp1R+5lpW5unt7V2kfk67oNjNzU1t27ZVXFycVXtcXJzatWtX6D6hoaEF+m/ZskUhISGFXm8DAADKH6feLRUREaGlS5dq2bJlSkpK0oQJE5SamqoRI0ZIunxKKSwszNJ/xIgR+uWXXxQREaGkpCQtW7ZMb7/9tiZOnOisKQAAgFLGqdfcDBgwQGfOnNHMmTOVlpam4OBgbdy4UQ0bNpQkpaWlWT3zJiAgQBs3btSECRP05ptvytfXV/Pnz9cDDzzgrCk4lLu7u2bMmFHgtJoRMVfjKS/zlMrPXMvLPKXyM1ejztNkNhflnioAAICywekfvwAAAGBPhBsAAGAohBsAAGAohBsAAGAohJtSJjs7W61bt5bJZFJiYmKR9xs3bpzatm0rd3d3tW7d2mH12Utx5vndd9/pkUcekZ+fnzw9PRUYGKh58+Y5tlA7KM5cz5w5o+7du8vX11fu7u7y8/PT6NGjHfaZaPZQ3H+7V5w5c0b169eXyWTSn3/+aff6nCktLU0DBw5Us2bN5OLiovHjxzu7JIdZv369unTpotq1a6tq1aoKDQ3V5s2bnV2W3e3evVvt27dXzZo15enpqebNm+u1115zdlkOt2fPHlWoUKHU/50h3JQykyZNKtKjpf/JbDZr6NChGjBggAOqsr/izDM+Pl61a9fWf//7Xx06dEhTp05VZGSk3njjDQdVaR/FmauLi4v69u2rjz/+WEeOHFFsbKy++OILyzOgSqPi/tu9Ijw8XC1btrRjRaVHdna2ateuralTp6pVq1bOLsehdu7cqS5dumjjxo2Kj49X586d1adPHyUkJDi7NLuqXLmyRo8erZ07dyopKUnPPvusnn32WS1ZssTZpTlMZmamwsLCdPfddzu7lH/3r58+Bbv4/PPPze3btzd7e3uba9SoYe7Vq5f56NGjVn02btxobt68ufnQoUNmSeaEhASbjzNjxgxzq1at7FN0MZTUPK8YOXKkuXPnztdZdfGU9FznzZtnrl+//nVWbbuSmOeCBQvMHTt2NH/55ZdmSeY//vjDfhMooo4dO5rHjBljfvrpp83Vq1c3+/j4mGfMmGF5/5dffjHfe++95sqVK5urVKlifuihh8ynTp0q1nHGjRtnv8KLoaTmekVQUJA5OjraDpXbpqTnef/995sfe+wxO1Ruu5KY64ABA8zPPvus0//OFAUrNyXk/PnzioiI0DfffKMvv/xSLi4uuv/++5Wfny9JOn36tIYPH65Vq1apUqVKTq62+Ep6npmZmapRo8Z1j1McJTnXkydPav369erYsaM9SreJo+d5+PBhzZw5UytXrpSLi3P/k7RixQpVrlxZX3/9tV5++WXNnDlTcXFxMpvNuu+++/T7779rx44diouLU3JycplZKS1MSc01Pz9fZ8+eddrvaUnNMyEhQXv37nXK7+gVjpzr8uXLlZycrBkzZjhwBnbk3GxVfqWnp5slmQ8ePGjOz883d+/e3Txr1iyz2Ww2p6SklNmVm39y1DzNZrN579695ooVK5q3bNlix4qLzxFzffjhh82enp5mSeY+ffqYL1y44IDKbWPPeV68eNHcsmVL86pVq8xms9m8bds2p67c3HHHHVZtt9xyi3ny5MnmLVu2mF1dXc2pqamW966sUu3fv9/m45SGlZuSmKvZbDa//PLL5ho1aphPnz593XXbqiTmecMNN5jd3NzMLi4u5pkzZ9qtdls5cq5Hjhwx16lTx/zjjz+azebS93emMKzclJDk5GQNHDhQjRo1UtWqVRUQECBJSk1N1euvv66srCxFRkY6ucrrV1LzPHTokPr27avp06erS5cu1z1ecZTEXF977TV9++23+vDDD5WcnKyIiAh7lG4TR84zMjJSgYGBeuyxx+xZcrH985qfevXqKT09XUlJSfLz85Ofn5/lvaCgIFWrVk1JSUklXaZdlMRc16xZo6ioKK1du1Z16tSxS922cvQ8d+3apQMHDmjRokWKiYnRmjVr7Fa7rRwx17y8PA0cOFDR0dFq2rSpQ+p2BKd+tlR50qdPH/n5+emtt96Sr6+v8vPzFRwcrJycHG3dulX79u0r8NkeISEhevTRR7VixQonVW27kpjn4cOHddddd2n48OF69tlnHTGNIimJudatW1d169ZV8+bNVbNmTXXo0EHTpk1TvXr1HDGlQjlynlu3btXBgwf1/vvvS7p8Ybwk1apVS1OnTlV0dLRjJnUVFStWtNo2mUzKz8+X2WyWyWQq0P9q7WWBo+e6du1ahYeH67333tM999xz3fUWl6PneSXst2jRQqdPn1ZUVJQeeeSR6yu6mBwx17Nnz+rAgQNKSEjQ6NGjJckyZoUKFbRlyxbddddd9puEnRBuSsCZM2eUlJSkxYsXq0OHDpIu30Z4xfz58/Xcc89Ztk+ePKlu3bpp7dq1uu2220q83uIqiXkeOnRId911lwYNGqTnn3/evhOwgTN+plf+8GdnZ19H5bZx9Dw/+OADXbhwwbL9zTffaOjQodq1a5duvPFGO87k+gQFBSk1NVXHjx+3/N/v4cOHlZmZqcDAQCdXZ1/2mOuaNWs0dOhQrVmzRr169XJkucXmiJ+p2Wwu0d/PorqeuVatWlUHDx60aluwYIG2bt2q999/3xLuShvCTQmoXr26atasqSVLlqhevXpKTU3VlClTLO83aNDAqr+Xl5ck6cYbb1T9+vWLdIyjR4/q3LlzOnXqlC5cuGB5zkhQUJDc3NzsM5F/4eh5Hjp0SJ07d1bXrl0VERGhU6dOSZJcXV1Vu3ZtO87k3zl6rhs3btTp06d1yy23yMvLS4cPH9akSZPUvn17+fv723Uu1+Loef4zwGRkZEiSAgMDVa1ateus3n7uuecetWzZUo8++qhiYmJ06dIljRw5Uh07dlRISEiRxrjyO3nu3Dn99ttvSkxMlJubm4KCghxYue2ud65r1qxRWFiY5s2bp9tvv93ye+rp6Slvb29Hl19k1zvPN998Uw0aNFDz5s0lXQ79r776qsaMGePo0m12PXN1cXFRcHCwVVudOnXk4eFRoL004ZqbEuDi4qJ3331X8fHxCg4O1oQJE/TKK6/Y9RjDhg1TmzZttHjxYh05ckRt2rRRmzZtdPLkSbse51ocPc/33ntPv/32m1avXq169epZXrfccovdjlFUjp6rp6en3nrrLd1xxx0KDAzU+PHj1bt3b3366ad2O0ZRlMS/3bLAZDLpww8/VPXq1XXnnXfqnnvuUaNGjbR27doij3HldzI+Pl7vvPOO2rRpo549ezqw6uK53rkuXrxYly5d0qhRo6x+T8eNG+fgym1zvfPMz89XZGSkWrdurZCQEL3++ut68cUXNXPmTAdXbjt7/Psta0zmK2vdAAAABsDKDQAAMBTCTRkwYsQIeXl5FfoqzY/jt1V5madUfuZaXuZ50003XXWeq1evdnZ5dlVe5lpe5ikZc66clioD0tPTr/qBiVWrVnXa8yPsrbzMUyo/cy0v8/zll1+Um5tb6Hs+Pj6qUqVKCVfkOOVlruVlnpIx50q4AQAAhsJpKQAAYCiEGwAAYCiEGwAAYCiEGwAAYCiEGwAAYCiEGwAAYCiEGwAAYCiEGwAAYCj/D1fgcZ86vjfxAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "make_table(ins, outs, \"adapter trimmed\", \"umi and adapter trimmed\", \"Number of Reads\", samples)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "699fb601-82e1-4e55-896b-98e1b11c92f4",
   "metadata": {},
   "source": [
    "Frequency of Read Length for Adapter Only Trimmed vs. Adapter and UMI Trimmed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "2f0111ab-c1ff-4357-b64c-4a033dd60c32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAkAAAAHFCAYAAAAaD0bAAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABBOElEQVR4nO3deVhV5f7//9eOUUVIIRlKESsTI4egDJXKTBz7qlkHzym1tE40oZDnY2qdckg0y8wcqETN6qil1tGkAlPJEisMrYxsEMUMMqxALRnv3x/+3KftRhMCN7qej+va1+W+13ut/V63KK9rTdtmjDECAACwkPNc3QAAAMCZRgACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACGpg5c+bIZrMpIiKiRustWbJENptNe/bsqZ/G/n/Tpk3Tm2++Wa+fUZ0jR45o+vTp6ty5s3x8fNSkSRN16tRJ06ZN05EjR2q93bqatz179shms53W62Sf1bp1a91xxx1/qY+G5vHHH5fNZnN1G4ATd1c3AMDRokWLJEk7d+7URx99pC5duri4I0fTpk3TLbfcokGDBp2xz/zxxx9144036rvvvlNCQoKefPJJSdKGDRs0depULVu2TOvXr1dgYOAZ6+lEwcHBysrKchi77777VFxcrFdffdWptjpvvPGGfH19661HAP9DAAIakOzsbO3YsUP9+/fXunXrlJqa2uACUH2orKxURUWFvLy8ql0+fPhwffXVV9q4caO6d+9uH+/Vq5f69++vHj16aMSIEXrnnXfOVMtOvLy8dM011ziM+fr6qqyszGn8RL///rsaNWqkzp0712eLAP6AU2BAA5KamipJmj59urp27arly5frt99+c6rbunWrunXrJm9vb4WEhGj8+PEqLy93qluxYoViY2MVHBysRo0aKTw8XA8//LDTKaM77rhDPj4+2rlzp3r27KkmTZroggsu0AMPPODw+TabTUeOHNFLL71kP51z/fXX25cXFhbqnnvu0UUXXSRPT0+FhYVp0qRJqqiosNccP1X05JNPaurUqQoLC5OXl5c2btxY7ZxkZ2crPT1do0aNcgg/x3Xv3l0jR47Uu+++q23btjn0+sADD+jll19WeHi4GjdurI4dO+qtt946yewfM2XKFLm7u2vfvn1Oy0aOHCl/f38dPXr0lNs4ldatW2vAgAFavXq1OnfuLG9vb02aNMm+7I+nwDZt2iSbzab//Oc/GjdunIKDg+Xj46ObbrpJP/74ow4dOqR//vOfCggIUEBAgO68804dPnzY4fOOz8PixYt12WWXqVGjRoqKitLWrVtljNHMmTMVFhYmHx8f3XDDDfr222+del6/fr169uwpX19fNW7cWN26ddN7773nVLdu3Tp16tRJXl5eCgsL01NPPVXreQLqnQHQIPz222/Gz8/PXHXVVcYYYxYuXGgkmSVLljjU7dy50zRu3Ni0b9/eLFu2zPz3v/81vXv3Nq1atTKSTF5enr12ypQp5plnnjHr1q0zmzZtMikpKSYsLMz06NHDYZsjRowwnp6eplWrVuaJJ54w6enp5vHHHzfu7u5mwIAB9rqsrCzTqFEj069fP5OVlWWysrLMzp07jTHGFBQUmJYtW5rQ0FDz/PPPm/Xr15spU6YYLy8vc8cdd9i3kZeXZySZCy+80PTo0cOsXLnSpKenO/T9R9OmTTOSzNtvv33SuUtLSzOSTHJysn1MkmndurW5+uqrzWuvvWbS0tLM9ddfb9zd3c13331nr1u8eLHDvP3444/Gy8vLTJw40eEzDh48aBo1amT+9a9/nbSPE1133XXm8ssvdxgLDQ01wcHBpk2bNmbRokVm48aN5uOPP7YvGzFihL1248aNRpIJDQ01d9xxh3nnnXdMSkqK8fHxMT169DC9evUyY8eONenp6WbGjBnGzc3NPPjggw6fd3z9rl27mtWrV5s33njDtG3b1jRv3twkJiaagQMHmrfeesu8+uqrJjAw0HTo0MFUVVXZ13/55ZeNzWYzgwYNMqtXrzZr1641AwYMMG5ubmb9+vX2uvXr1xs3NzfTvXt3s3r1avP666+bq666yv5zCTQ0/FQCDcTSpUuNJJOSkmKMMebQoUPGx8fHxMTEONTFxcWZRo0amcLCQvtYRUWFadeunVMA+qOqqipTXl5uMjMzjSSzY8cO+7IRI0YYSebZZ591WOeJJ54wkswHH3xgH2vSpInDL+nj7rnnHuPj42P27t3rMP7UU08ZSfagdDwAXXzxxaasrOxP5yU+Pt5IMl999dVJa3Jzc40kc++999rHJJnAwEBTUlJiHyssLDTnnXeeQ1A6MQAZc2w+WrRoYUpLS+1jM2bMMOedd95J57c6JwtAbm5uZteuXU71JwtAN910k0PdmDFjjCSTkJDgMD5o0CDTvHlzhzFJJigoyBw+fNg+9uabbxpJplOnTg5hZ/bs2UaS+eyzz4wxxhw5csQ0b97c6fMrKytNx44dzdVXX20f69KliwkJCTG///67faykpMQ0b96cAIQGiVNgQAORmpqqRo0aaejQoZIkHx8f3Xrrrdq8ebO++eYbe93GjRvVs2dPhwt+3dzcFBcX57TN3bt36x//+IeCgoLk5uYmDw8PXXfddZKk3Nxcp/rbbrvN4f0//vEP+2f+mbfeeks9evRQSEiIKioq7K++fftKkjIzMx3q/9//+3/y8PD40+2eDmOMJDndbdSjRw81bdrU/j4wMFAtWrTQ3r17T7m90aNH68CBA3r99dclSVVVVVqwYIH69++v1q1b/+V+O3TooLZt2552/YABAxzeh4eHS5L69+/vNP7zzz87nQbr0aOHmjRp4rR+3759Hebs+Pjx+dmyZYt+/vlnjRgxwuHvtKqqSn369NEnn3yiI0eO6MiRI/rkk0908803y9vb2769pk2b6qabbjrt/QTOJAIQ0AB8++23ev/999W/f38ZY/Trr7/q119/1S233CLpf3eGSdLBgwcVFBTktI0Txw4fPqyYmBh99NFHmjp1qjZt2qRPPvlEq1evlnTswts/cnd3l7+/f7XbPHjw4J/uw48//qi1a9fKw8PD4XX55ZdLkoqKihzqT3Yn1IlatWolScrLyztpzfHbylu2bOkwfuL+SMcuVj5x30/UuXNnxcTEaN68eZKOhbs9e/bogQceOK2e/8zp7vtxzZs3d3jv6el5yvETr1Gq7fo//vijJOmWW25x+nudMWOGjDH6+eef9csvv6iqquq0fi6BhoK7wIAGYNGiRTLGaOXKlVq5cqXT8pdeeklTp06Vm5ub/P39VVhY6FRz4tiGDRv0ww8/aNOmTfajPpL066+/VttDRUWFDh486BAajm+zuiBxooCAAHXo0EFPPPFEtctDQkIc3p/us2F69eqlCRMm6M0331SfPn2qrTn+XKJevXqd1jZPR0JCgm699VZ9+umnmjt3rtq2bVtn2z9bnosTEBAgSXruuedOeidbYGCgysvLZbPZTuvnEmgoCECAi1VWVuqll17SxRdfrIULFzotf+utt/T000/r7bff1oABA9SjRw+tWbNGP/74o/00WGVlpVasWOGw3vFfsifeWv7888+ftJdXX31VCQkJ9vf/+c9/JMnhTq+THUEZMGCA0tLSdPHFF6tZs2Z/stenLyoqSrGxsUpNTdWwYcPUrVs3h+UffPCBFi1apD59+igyMrLOPnfw4MFq1aqVHnroIWVmZuqZZ545a4JLXenWrZvOP/98ffnll6c8+uXp6amrr75aq1ev1syZM+2nwQ4dOqS1a9eeqXaBGiEAAS729ttv64cfftCMGTMcgsZxERERmjt3rlJTUzVgwAA98sgjWrNmjW644Qb9+9//VuPGjTVv3jynW9u7du2qZs2aKT4+Xo899pg8PDz06quvaseOHdX24enpqaefflqHDx/WVVddpS1btmjq1Knq27evw+3nV1xxhTZt2qS1a9cqODhYTZs21WWXXabJkycrIyNDXbt2VUJCgi677DIdPXpUe/bsUVpamlJSUnTRRRfVao6WLl2qG2+8UbGxsUpISFDPnj0lHTvK9eyzz6pdu3ZasmRJrbZ9Mm5ubrr//vs1btw4NWnS5Jx7QvPp8PHx0XPPPacRI0bo559/1i233KIWLVrop59+0o4dO/TTTz9pwYIFko49PqBPnz7q1auXHnroIVVWVmrGjBlq0qSJfv75ZxfvCeCMa4AAF0tNTZWnp6fuvPPOapcHBARo8ODBeuutt/Tjjz8qIiJC69evl6+vr0aMGKF//vOf6tChgx599FGH9fz9/bVu3To1btxYt99+u0aOHCkfHx+nI0XHeXh46K233lJGRoYGDhyoOXPm6O6777ZfCHzcs88+q0svvVRDhw7VVVddpXvuuUfSsetasrOzFRsbq5kzZ6pPnz4aNmyYFi1apE6dOv2lo0KBgYHaunWrHnnkEb3zzjsaOHCgBg4cqHXr1mnChAnaunVrvTwF+viF5cOGDZOfn1+db/9scPvtt2vjxo06fPiw7rnnHt14440aPXq0Pv30U3sQlY6dfnzzzTdVUlKiuLg4JSUlaciQIRo5cqQLuwdOzmaO3z4BwLLuuOMOrVy50unuIat77rnnlJCQoC+++MJ+MTeAcwOnwADgBDk5OcrLy9PkyZM1cOBAwg9wDiIAAcAJBg8erMLCQsXExCglJcXV7QCoB5wCAwAAlsNF0AAAwHIIQAAAwHIIQAAAwHK4CLoaVVVV+uGHH9S0aVPLPfkVAICzlTFGhw4dUkhIiM4779THeAhA1fjhhx+cvlQRAACcHfbt2/enT54nAFWjadOmko5NoK+vr4u7AQAAp6OkpEQtW7a0/x4/FQJQNY6f9vL19SUAAQBwljmdy1e4CBoAAFgOAQgAAFgOAQgAAFgOAQgAAFgOAQgAAFgOAQgAAFgOAQgAAFiOywPQ/PnzFRYWJm9vb0VGRmrz5s2nrM/MzFRkZKS8vb3Vpk0bpaSkOCy//vrrZbPZnF79+/evz90AAABnEZcGoBUrVmjMmDGaOHGicnJyFBMTo759+yo/P7/a+ry8PPXr108xMTHKycnRhAkTlJCQoFWrVtlrVq9erYKCAvvriy++kJubm2699dYztVsAAKCBsxljjKs+vEuXLrryyiu1YMEC+1h4eLgGDRqk5ORkp/px48ZpzZo1ys3NtY/Fx8drx44dysrKqvYzZs+erX//+98qKChQkyZNTquvkpIS+fn5qbi4mCdBAwBwlqjJ72+XHQEqKyvTtm3bFBsb6zAeGxurLVu2VLtOVlaWU33v3r2VnZ2t8vLyatdJTU3V0KFDTxl+SktLVVJS4vACAADnLpcFoKKiIlVWViowMNBhPDAwUIWFhdWuU1hYWG19RUWFioqKnOo//vhjffHFF7rrrrtO2UtycrL8/PzsL74JHgCAc5vLL4I+8QvLjDGn/BKz6uqrG5eOHf2JiIjQ1Vdffcoexo8fr+LiYvtr3759p9s+AAA4C7ns2+ADAgLk5ubmdLTnwIEDTkd5jgsKCqq23t3dXf7+/g7jv/32m5YvX67Jkyf/aS9eXl7y8vKq4R4AAICzlcuOAHl6eioyMlIZGRkO4xkZGeratWu160RHRzvVp6enKyoqSh4eHg7jr732mkpLS3X77bfXbeMAAOCs57IjQJKUlJSkYcOGKSoqStHR0XrhhReUn5+v+Ph4ScdOTe3fv19Lly6VdOyOr7lz5yopKUl33323srKylJqaqmXLljltOzU1VYMGDXI6MgQAwLmm9cPrXN1Cje2Z7trn87k0AMXFxengwYOaPHmyCgoKFBERobS0NIWGhkqSCgoKHJ4JFBYWprS0NCUmJmrevHkKCQnRnDlzNGTIEIftfv311/rggw+Unp5+RvcHAACcHVz6HKCGiucAAQDOJhwBOuaseA4QAACAqxCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5bg8AM2fP19hYWHy9vZWZGSkNm/efMr6zMxMRUZGytvbW23atFFKSopTza+//qr7779fwcHB8vb2Vnh4uNLS0uprFwAAwFnGpQFoxYoVGjNmjCZOnKicnBzFxMSob9++ys/Pr7Y+Ly9P/fr1U0xMjHJycjRhwgQlJCRo1apV9pqysjL16tVLe/bs0cqVK7Vr1y69+OKLuvDCC8/UbgEAgAbOZowxrvrwLl266Morr9SCBQvsY+Hh4Ro0aJCSk5Od6seNG6c1a9YoNzfXPhYfH68dO3YoKytLkpSSkqKZM2fqq6++koeHR636KikpkZ+fn4qLi+Xr61urbQAAcKa0fnidq1uosT3T+9f5Nmvy+9tlR4DKysq0bds2xcbGOozHxsZqy5Yt1a6TlZXlVN+7d29lZ2ervLxckrRmzRpFR0fr/vvvV2BgoCIiIjRt2jRVVlbWz44AAICzjrurPrioqEiVlZUKDAx0GA8MDFRhYWG16xQWFlZbX1FRoaKiIgUHB2v37t3asGGDbrvtNqWlpembb77R/fffr4qKCv373/+udrulpaUqLS21vy8pKfmLewcAABoyl18EbbPZHN4bY5zG/qz+j+NVVVVq0aKFXnjhBUVGRmro0KGaOHGiw2m2EyUnJ8vPz8/+atmyZW13BwAAnAVcFoACAgLk5ubmdLTnwIEDTkd5jgsKCqq23t3dXf7+/pKk4OBgtW3bVm5ubvaa8PBwFRYWqqysrNrtjh8/XsXFxfbXvn37/squAQCABs5lAcjT01ORkZHKyMhwGM/IyFDXrl2rXSc6OtqpPj09XVFRUfYLnrt166Zvv/1WVVVV9pqvv/5awcHB8vT0rHa7Xl5e8vX1dXgBAIBzl0tPgSUlJWnhwoVatGiRcnNzlZiYqPz8fMXHx0s6dmRm+PDh9vr4+Hjt3btXSUlJys3N1aJFi5SamqqxY8faa+69914dPHhQo0eP1tdff61169Zp2rRpuv/++8/4/gEAgIbJZRdBS1JcXJwOHjyoyZMnq6CgQBEREUpLS1NoaKgkqaCgwOGZQGFhYUpLS1NiYqLmzZunkJAQzZkzR0OGDLHXtGzZUunp6UpMTFSHDh104YUXavTo0Ro3btwZ3z8AANAwufQ5QA0VzwECAJxNeA7QMWfFc4AAAABchQAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAsx+UBaP78+QoLC5O3t7ciIyO1efPmU9ZnZmYqMjJS3t7eatOmjVJSUhyWL1myRDabzel19OjR+twNAABwFnFpAFqxYoXGjBmjiRMnKicnRzExMerbt6/y8/Orrc/Ly1O/fv0UExOjnJwcTZgwQQkJCVq1apVDna+vrwoKChxe3t7eZ2KXAADAWcDdlR8+a9YsjRo1SnfddZckafbs2Xr33Xe1YMECJScnO9WnpKSoVatWmj17tiQpPDxc2dnZeuqppzRkyBB7nc1mU1BQ0BnZBwAAcPZx2RGgsrIybdu2TbGxsQ7jsbGx2rJlS7XrZGVlOdX37t1b2dnZKi8vt48dPnxYoaGhuuiiizRgwADl5OTU/Q4AAICzlssCUFFRkSorKxUYGOgwHhgYqMLCwmrXKSwsrLa+oqJCRUVFkqR27dppyZIlWrNmjZYtWyZvb29169ZN33zzzUl7KS0tVUlJicMLAACcu1x+EbTNZnN4b4xxGvuz+j+OX3PNNbr99tvVsWNHxcTE6LXXXlPbtm313HPPnXSbycnJ8vPzs79atmxZ290BAABnAZcFoICAALm5uTkd7Tlw4IDTUZ7jgoKCqq13d3eXv79/teucd955uuqqq055BGj8+PEqLi62v/bt21fDvQEAAGcTlwUgT09PRUZGKiMjw2E8IyNDXbt2rXad6Ohop/r09HRFRUXJw8Oj2nWMMdq+fbuCg4NP2ouXl5d8fX0dXgAA4Nzl0lNgSUlJWrhwoRYtWqTc3FwlJiYqPz9f8fHxko4dmRk+fLi9Pj4+Xnv37lVSUpJyc3O1aNEipaamauzYsfaaSZMm6d1339Xu3bu1fft2jRo1Stu3b7dvEwAAwKW3wcfFxengwYOaPHmyCgoKFBERobS0NIWGhkqSCgoKHJ4JFBYWprS0NCUmJmrevHkKCQnRnDlzHG6B//XXX/XPf/5ThYWF8vPzU+fOnfX+++/r6quvPuP7BwAAGiabOX4VMexKSkrk5+en4uJiTocBABq81g+vc3ULNbZnev8632ZNfn+7/C4wAACAM40ABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALKdWASgvL6+u+wAAADhjahWALrnkEvXo0UOvvPKKjh49Wtc9AQAA1KtaBaAdO3aoc+fOeuihhxQUFKR77rlHH3/8cV33BgAAUC9qFYAiIiI0a9Ys7d+/X4sXL1ZhYaG6d++uyy+/XLNmzdJPP/1U130CAADUmb90EbS7u7sGDx6s1157TTNmzNB3332nsWPH6qKLLtLw4cNVUFBQV30CAADUmb8UgLKzs3XfffcpODhYs2bN0tixY/Xdd99pw4YN2r9/vwYOHFhXfQIAANQZ99qsNGvWLC1evFi7du1Sv379tHTpUvXr10/nnXcsT4WFhen5559Xu3bt6rRZAACAulCrALRgwQKNHDlSd955p4KCgqqtadWqlVJTU/9ScwAAAPWhVgHom2+++dMaT09PjRgxojabBwAAqFe1ugZo8eLFev31153GX3/9db300kt/uSkAAID6VKsANH36dAUEBDiNt2jRQtOmTavRtubPn6+wsDB5e3srMjJSmzdvPmV9ZmamIiMj5e3trTZt2iglJeWktcuXL5fNZtOgQYNq1BMAADi31SoA7d27V2FhYU7joaGhys/PP+3trFixQmPGjNHEiROVk5OjmJgY9e3b96TbyMvLU79+/RQTE6OcnBxNmDBBCQkJWrVqVbU9jh07VjExMae/YwAAwBJqFYBatGihzz77zGl8x44d8vf3P+3tzJo1S6NGjdJdd92l8PBwzZ49Wy1bttSCBQuqrU9JSVGrVq00e/ZshYeH66677tLIkSP11FNPOdRVVlbqtttu06RJk9SmTZua7RwAADjn1SoADR06VAkJCdq4caMqKytVWVmpDRs2aPTo0Ro6dOhpbaOsrEzbtm1TbGysw3hsbKy2bNlS7TpZWVlO9b1791Z2drbKy8vtY5MnT9YFF1ygUaNGnVYvpaWlKikpcXgBAIBzV63uAps6dar27t2rnj17yt392Caqqqo0fPjw074GqKioSJWVlQoMDHQYDwwMVGFhYbXrFBYWVltfUVGhoqIiBQcH68MPP1Rqaqq2b99+2vuTnJysSZMmnXY9AAA4u9UqAHl6emrFihWaMmWKduzYoUaNGumKK65QaGhojbdls9kc3htjnMb+rP74+KFDh3T77bfrxRdfrPYi7ZMZP368kpKS7O9LSkrUsmXL014fAACcXWoVgI5r27at2rZtW6t1AwIC5Obm5nS058CBA05HeY4LCgqqtt7d3V3+/v7auXOn9uzZo5tuusm+vKqqStKx7y3btWuXLr74Yqftenl5ycvLq1b7AQAAzj61CkCVlZVasmSJ3nvvPR04cMAeMo7bsGHDn27D09NTkZGRysjI0ODBg+3jGRkZJ/0OsejoaK1du9ZhLD09XVFRUfLw8FC7du30+eefOyx/5JFHdOjQIT377LMc1QEAAJJqGYBGjx6tJUuWqH///oqIiDjlKatTSUpK0rBhwxQVFaXo6Gi98MILys/PV3x8vKRjp6b279+vpUuXSpLi4+M1d+5cJSUl6e6771ZWVpZSU1O1bNkySZK3t7ciIiIcPuP888+XJKdxAABgXbUKQMuXL9drr72mfv36/aUPj4uL08GDBzV58mQVFBQoIiJCaWlp9muJCgoKHJ4JFBYWprS0NCUmJmrevHkKCQnRnDlzNGTIkL/UBwAAsBabOX4VcQ2EhIRo06ZNtb7+p6ErKSmRn5+fiouL5evr6+p2AAA4pdYPr3N1CzW2Z3r/Ot9mTX5/1+o5QA899JCeffZZ1SI7AQAAuFytToF98MEH2rhxo95++21dfvnl8vDwcFi+evXqOmkOAACgPtQqAJ1//vkOd24BAACcTWoVgBYvXlzXfQAAAJwxtboGSJIqKiq0fv16Pf/88zp06JAk6YcfftDhw4frrDkAAID6UKsjQHv37lWfPn2Un5+v0tJS9erVS02bNtWTTz6po0ePKiUlpa77BAAAqDO1OgI0evRoRUVF6ZdfflGjRo3s44MHD9Z7771XZ80BAADUh1rfBfbhhx/K09PTYTw0NFT79++vk8YAAADqS62OAFVVVamystJp/Pvvv1fTpk3/clMAAAD1qVYBqFevXpo9e7b9vc1m0+HDh/XYY4/95a/HAAAAqG+1OgX2zDPPqEePHmrfvr2OHj2qf/zjH/rmm28UEBBg/2JSAACAhqpWASgkJETbt2/XsmXL9Omnn6qqqkqjRo3Sbbfd5nBRNAAAQENUqwAkSY0aNdLIkSM1cuTIuuwHAACg3tUqAC1duvSUy4cPH16rZgAAAM6EWgWg0aNHO7wvLy/Xb7/9Jk9PTzVu3JgABAAAGrRa3QX2yy+/OLwOHz6sXbt2qXv37lwEDQAAGrxafxfYiS699FJNnz7d6egQAABAQ1NnAUiS3Nzc9MMPP9TlJgEAAOpcra4BWrNmjcN7Y4wKCgo0d+5cdevWrU4aAwAAqC+1CkCDBg1yeG+z2XTBBRfohhtu0NNPP10XfQEAANSbWgWgqqqquu4DAADgjKnTa4AAAADOBrU6ApSUlHTatbNmzarNRwAAANSbWgWgnJwcffrpp6qoqNBll10mSfr666/l5uamK6+80l5ns9nqpksAAIA6VKsAdNNNN6lp06Z66aWX1KxZM0nHHo545513KiYmRg899FCdNgkAAFCXanUN0NNPP63k5GR7+JGkZs2aaerUqdwFBgAAGrxaBaCSkhL9+OOPTuMHDhzQoUOH/nJTAAAA9alWAWjw4MG68847tXLlSn3//ff6/vvvtXLlSo0aNUo333xzXfcIAABQp2p1DVBKSorGjh2r22+/XeXl5cc25O6uUaNGaebMmXXaIAAAQF2rVQBq3Lix5s+fr5kzZ+q7776TMUaXXHKJmjRpUtf9AQAA1Lm/9CDEgoICFRQUqG3btmrSpImMMXXVFwAAQL2pVQA6ePCgevbsqbZt26pfv34qKCiQJN11113cAg8AABq8WgWgxMREeXh4KD8/X40bN7aPx8XF6Z133qmz5gAAAOpDra4BSk9P17vvvquLLrrIYfzSSy/V3r1766QxAACA+lKrI0BHjhxxOPJzXFFRkby8vP5yUwAAAPWpVgHo2muv1dKlS+3vbTabqqqqNHPmTPXo0aPOmgMAAKgPtToFNnPmTF1//fXKzs5WWVmZ/u///k87d+7Uzz//rA8//LCuewQAAKhTtToC1L59e3322We6+uqr1atXLx05ckQ333yzcnJydPHFF9d1jwAAAHWqxkeAysvLFRsbq+eff16TJk2qj54AAADqVY2PAHl4eOiLL76QzWarj34AAADqXa1OgQ0fPlypqal10sD8+fMVFhYmb29vRUZGavPmzaesz8zMVGRkpLy9vdWmTRulpKQ4LF+9erWioqJ0/vnnq0mTJurUqZNefvnlOukVAACcG2p1EXRZWZkWLlyojIwMRUVFOX0H2KxZs05rOytWrNCYMWM0f/58devWTc8//7z69u2rL7/8Uq1atXKqz8vLU79+/XT33XfrlVde0Ycffqj77rtPF1xwgYYMGSJJat68uSZOnKh27drJ09NTb731lu688061aNFCvXv3rs3uAgCAc4zN1OALvHbv3q3WrVurZ8+eJ9+gzaYNGzac1va6dOmiK6+8UgsWLLCPhYeHa9CgQUpOTnaqHzdunNasWaPc3Fz7WHx8vHbs2KGsrKyTfs6VV16p/v37a8qUKafVV0lJifz8/FRcXCxfX9/TWgcAAFdp/fA6V7dQY3um96/zbdbk93eNjgBdeumlKigo0MaNGyUd++qLOXPmKDAwsMZNlpWVadu2bXr44YcdxmNjY7Vly5Zq18nKylJsbKzDWO/evZWamqry8nJ5eHg4LDPGaMOGDdq1a5dmzJhx0l5KS0tVWlpqf19SUlLT3QEAAGeRGl0DdOLBorfffltHjhyp1QcXFRWpsrLSKTwFBgaqsLCw2nUKCwurra+oqFBRUZF9rLi4WD4+PvL09FT//v313HPPqVevXiftJTk5WX5+fvZXy5Yta7VPAADg7FCri6CPq8HZs5M68W4yY8wp7zCrrv7E8aZNm2r79u365JNP9MQTTygpKUmbNm066TbHjx+v4uJi+2vfvn212BMAAHC2qNEpMJvN5hRAans7fEBAgNzc3JyO9hw4cOCkp9SCgoKqrXd3d5e/v7997LzzztMll1wiSerUqZNyc3OVnJys66+/vtrtenl58R1mAABYSI0CkDFGd9xxhz0sHD16VPHx8U53ga1evfpPt+Xp6anIyEhlZGRo8ODB9vGMjAwNHDiw2nWio6O1du1ah7H09HRFRUU5Xf9zYt9/vMYHAABYW40C0IgRIxze33777X/pw5OSkjRs2DBFRUUpOjpaL7zwgvLz8xUfHy/p2Kmp/fv32794NT4+XnPnzlVSUpLuvvtuZWVlKTU1VcuWLbNvMzk5WVFRUbr44otVVlamtLQ0LV261OFOMwAAYG01CkCLFy+u0w+Pi4vTwYMHNXnyZBUUFCgiIkJpaWkKDQ2VJBUUFCg/P99eHxYWprS0NCUmJmrevHkKCQnRnDlz7M8AkqQjR47ovvvu0/fff69GjRqpXbt2euWVVxQXF1envQMAgLNXjZ4DZBU8BwgAcDbhOUDH1OT391+6CwwAAOBsRAACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACW4+7qBgD8T+uH17m6hRrbM72/q1sAgBrjCBAAALAcAhAAALAcAhAAALAcAhAAALAclweg+fPnKywsTN7e3oqMjNTmzZtPWZ+ZmanIyEh5e3urTZs2SklJcVj+4osvKiYmRs2aNVOzZs1044036uOPP67PXQAAAGcZlwagFStWaMyYMZo4caJycnIUExOjvn37Kj8/v9r6vLw89evXTzExMcrJydGECROUkJCgVatW2Ws2bdqkv//979q4caOysrLUqlUrxcbGav/+/WdqtwAAQANnM8YYV314ly5ddOWVV2rBggX2sfDwcA0aNEjJyclO9ePGjdOaNWuUm5trH4uPj9eOHTuUlZVV7WdUVlaqWbNmmjt3roYPH35afZWUlMjPz0/FxcXy9fWt4V4Btcdt8ABqg/87jqnJ72+XHQEqKyvTtm3bFBsb6zAeGxurLVu2VLtOVlaWU33v3r2VnZ2t8vLyatf57bffVF5erubNm5+0l9LSUpWUlDi8AADAuctlAaioqEiVlZUKDAx0GA8MDFRhYWG16xQWFlZbX1FRoaKiomrXefjhh3XhhRfqxhtvPGkvycnJ8vPzs79atmxZw70BAABnE5dfBG2z2RzeG2Ocxv6svrpxSXryySe1bNkyrV69Wt7e3ifd5vjx41VcXGx/7du3rya7AAAAzjIu+yqMgIAAubm5OR3tOXDggNNRnuOCgoKqrXd3d5e/v7/D+FNPPaVp06Zp/fr16tChwyl78fLykpeXVy32AgAAnI1cdgTI09NTkZGRysjIcBjPyMhQ165dq10nOjraqT49PV1RUVHy8PCwj82cOVNTpkzRO++8o6ioqLpvHgAAnNVcegosKSlJCxcu1KJFi5Sbm6vExETl5+crPj5e0rFTU3+8cys+Pl579+5VUlKScnNztWjRIqWmpmrs2LH2mieffFKPPPKIFi1apNatW6uwsFCFhYU6fPjwGd8/AADQMLn02+Dj4uJ08OBBTZ48WQUFBYqIiFBaWppCQ0MlSQUFBQ7PBAoLC1NaWpoSExM1b948hYSEaM6cORoyZIi9Zv78+SorK9Mtt9zi8FmPPfaYHn/88TOyXwAAoGFz6XOAGiqeAwRX4VkegOudjf8Oz0aWfQ4QAACAqxCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5RCAAACA5bi7ugGgvrR+eJ2rWwAANFAcAQIAAJZDAAIAAJZDAAIAAJZDAAIAAJZDAAIAAJZDAAIAAJZDAAIAAJZDAAIAAJZDAAIAAJZDAAIAAJZDAAIAAJZDAAIAAJZDAAIAAJZDAAIAAJZDAAIAAJZDAAIAAJZDAAIAAJZDAAIAAJZDAAIAAJZDAAIAAJZDAAIAAJbj8gA0f/58hYWFydvbW5GRkdq8efMp6zMzMxUZGSlvb2+1adNGKSkpDst37typIUOGqHXr1rLZbJo9e3Y9dg8AAM5GLg1AK1as0JgxYzRx4kTl5OQoJiZGffv2VX5+frX1eXl56tevn2JiYpSTk6MJEyYoISFBq1atstf89ttvatOmjaZPn66goKAztSsAAOAs4tIANGvWLI0aNUp33XWXwsPDNXv2bLVs2VILFiyotj4lJUWtWrXS7NmzFR4errvuuksjR47UU089Za+56qqrNHPmTA0dOlReXl5nalcAAMBZxGUBqKysTNu2bVNsbKzDeGxsrLZs2VLtOllZWU71vXv3VnZ2tsrLy2vdS2lpqUpKShxeAADg3OWyAFRUVKTKykoFBgY6jAcGBqqwsLDadQoLC6utr6ioUFFRUa17SU5Olp+fn/3VsmXLWm8LAAA0fC6/CNpmszm8N8Y4jf1ZfXXjNTF+/HgVFxfbX/v27av1tgAAQMPn7qoPDggIkJubm9PRngMHDjgd5TkuKCio2np3d3f5+/vXuhcvLy+uFwIAwEJcdgTI09NTkZGRysjIcBjPyMhQ165dq10nOjraqT49PV1RUVHy8PCot14BAMC5xaWnwJKSkrRw4UItWrRIubm5SkxMVH5+vuLj4yUdOzU1fPhwe318fLz27t2rpKQk5ebmatGiRUpNTdXYsWPtNWVlZdq+fbu2b9+usrIy7d+/X9u3b9e33357xvcPAAA0TC47BSZJcXFxOnjwoCZPnqyCggJFREQoLS1NoaGhkqSCggKHZwKFhYUpLS1NiYmJmjdvnkJCQjRnzhwNGTLEXvPDDz+oc+fO9vdPPfWUnnrqKV133XXatGnTGds3AADQcNnM8auIYVdSUiI/Pz8VFxfL19fX1e2gllo/vM7VLaCB2jO9v6tbQAPG/x1nRn38O6zJ72+X3wUGAABwphGAAACA5RCAAACA5RCAAACA5bj0LjAAwLmLi4nRkHEECAAAWA4BCAAAWA4BCAAAWA4BCAAAWA4BCAAAWA4BCAAAWA4BCAAAWA4BCAAAWA4BCAAAWA4BCAAAWA4BCAAAWA4BCAAAWA4BCAAAWA4BCAAAWA4BCAAAWI67qxsAAPy51g+vc3ULwDmFI0AAAMByCEAAAMByOAUGwHI4nQSAI0AAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMBy+DJUnBa+PBIAcC7hCBAAALAcAhAAALAclweg+fPnKywsTN7e3oqMjNTmzZtPWZ+ZmanIyEh5e3urTZs2SklJcapZtWqV2rdvLy8vL7Vv315vvPFGfbUPAADOQi4NQCtWrNCYMWM0ceJE5eTkKCYmRn379lV+fn619Xl5eerXr59iYmKUk5OjCRMmKCEhQatWrbLXZGVlKS4uTsOGDdOOHTs0bNgw/e1vf9NHH310pnYLAAA0cDZjjHHVh3fp0kVXXnmlFixYYB8LDw/XoEGDlJyc7FQ/btw4rVmzRrm5ufax+Ph47dixQ1lZWZKkuLg4lZSU6O2337bX9OnTR82aNdOyZctOq6+SkhL5+fmpuLhYvr6+td29cwoXQQMA6tKe6f3rfJs1+f3tsiNAZWVl2rZtm2JjYx3GY2NjtWXLlmrXycrKcqrv3bu3srOzVV5efsqak20TAABYj8tugy8qKlJlZaUCAwMdxgMDA1VYWFjtOoWFhdXWV1RUqKioSMHBwSetOdk2Jam0tFSlpaX298XFxZKOJcn6EPHYu/WyXQAAzhb18Tv2+DZP5+SWy58DZLPZHN4bY5zG/qz+xPGabjM5OVmTJk1yGm/ZsuXJGwcAALXmN7v+tn3o0CH5+fmdssZlASggIEBubm5OR2YOHDjgdATnuKCgoGrr3d3d5e/vf8qak21TksaPH6+kpCT7+6qqKv3888/y9/c/ZXCqjZKSErVs2VL79u3j+iIxHydiPpwxJ46YD0fMhzMrz4kxRocOHVJISMif1rosAHl6eioyMlIZGRkaPHiwfTwjI0MDBw6sdp3o6GitXbvWYSw9PV1RUVHy8PCw12RkZCgxMdGhpmvXriftxcvLS15eXg5j559/fk13qUZ8fX0t94N5KsyHI+bDGXPiiPlwxHw4s+qc/NmRn+NcegosKSlJw4YNU1RUlKKjo/XCCy8oPz9f8fHxko4dmdm/f7+WLl0q6dgdX3PnzlVSUpLuvvtuZWVlKTU11eHurtGjR+vaa6/VjBkzNHDgQP33v//V+vXr9cEHH7hkHwEAQMPj0gAUFxengwcPavLkySooKFBERITS0tIUGhoqSSooKHB4JlBYWJjS0tKUmJioefPmKSQkRHPmzNGQIUPsNV27dtXy5cv1yCOP6NFHH9XFF1+sFStWqEuXLmd8/wAAQMPk8oug77vvPt13333VLluyZInT2HXXXadPP/30lNu85ZZbdMstt9RFe3XOy8tLjz32mNMpN6tiPhwxH86YE0fMhyPmwxlzcnpc+iBEAAAAV3D5d4EBAACcaQQgAABgOQQgAABgOQQgAABgOQSgM2j+/PkKCwuTt7e3IiMjtXnzZle3dMa8//77uummmxQSEiKbzaY333zTYbkxRo8//rhCQkLUqFEjXX/99dq5c6drmj0DkpOTddVVV6lp06Zq0aKFBg0apF27djnUWGlOFixYoA4dOtgf3BYdHa23337bvtxKc1Gd5ORk2Ww2jRkzxj5mtTl5/PHHZbPZHF5BQUH25VabD0nav3+/br/9dvn7+6tx48bq1KmTtm3bZl9uxTmpCQLQGbJixQqNGTNGEydOVE5OjmJiYtS3b1+H5xydy44cOaKOHTtq7ty51S5/8sknNWvWLM2dO1effPKJgoKC1KtXLx06dOgMd3pmZGZm6v7779fWrVuVkZGhiooKxcbG6siRI/YaK83JRRddpOnTpys7O1vZ2dm64YYbNHDgQPt/1laaixN98skneuGFF9ShQweHcSvOyeWXX66CggL76/PPP7cvs9p8/PLLL+rWrZs8PDz09ttv68svv9TTTz/t8C0GVpuTGjM4I66++moTHx/vMNauXTvz8MMPu6gj15Fk3njjDfv7qqoqExQUZKZPn24fO3r0qPHz8zMpKSku6PDMO3DggJFkMjMzjTHMiTHGNGvWzCxcuNDSc3Ho0CFz6aWXmoyMDHPdddeZ0aNHG2Os+fPx2GOPmY4dO1a7zIrzMW7cONO9e/eTLrfinNQUR4DOgLKyMm3btk2xsbEO47GxsdqyZYuLumo48vLyVFhY6DA/Xl5euu666ywzP8XFxZKk5s2bS7L2nFRWVmr58uU6cuSIoqOjLT0X999/v/r3768bb7zRYdyqc/LNN98oJCREYWFhGjp0qHbv3i3JmvOxZs0aRUVF6dZbb1WLFi3UuXNnvfjii/blVpyTmiIAnQFFRUWqrKx0+kb6wMBAp2+ut6Ljc2DV+THGKCkpSd27d1dERIQka87J559/Lh8fH3l5eSk+Pl5vvPGG2rdvb8m5kKTly5fr008/VXJystMyK85Jly5dtHTpUr377rt68cUXVVhYqK5du+rgwYOWnI/du3drwYIFuvTSS/Xuu+8qPj5eCQkJ9u/OtOKc1JTLvwrDSmw2m8N7Y4zTmJVZdX4eeOABffbZZ9V+Ya+V5uSyyy7T9u3b9euvv2rVqlUaMWKEMjMz7cutNBf79u3T6NGjlZ6eLm9v75PWWWlO+vbta//zFVdcoejoaF188cV66aWXdM0110iy1nxUVVUpKipK06ZNkyR17txZO3fu1IIFCzR8+HB7nZXmpKY4AnQGBAQEyM3NzSl1HzhwwCmdW9HxOzmsOD8PPvig1qxZo40bN+qiiy6yj1txTjw9PXXJJZcoKipKycnJ6tixo5599llLzsW2bdt04MABRUZGyt3dXe7u7srMzNScOXPk7u5u328rzcmJmjRpoiuuuELffPONJX9GgoOD1b59e4ex8PBw+401VpyTmiIAnQGenp6KjIxURkaGw3hGRoa6du3qoq4ajrCwMAUFBTnMT1lZmTIzM8/Z+THG6IEHHtDq1au1YcMGhYWFOSy34pycyBij0tJSS85Fz5499fnnn2v79u32V1RUlG677TZt375dbdq0sdycnKi0tFS5ubkKDg625M9It27dnB6d8fXXXys0NFQS/4ecFlddfW01y5cvNx4eHiY1NdV8+eWXZsyYMaZJkyZmz549rm7tjDh06JDJyckxOTk5RpKZNWuWycnJMXv37jXGGDN9+nTj5+dnVq9ebT7//HPz97//3QQHB5uSkhIXd14/7r33XuPn52c2bdpkCgoK7K/ffvvNXmOlORk/frx5//33TV5envnss8/MhAkTzHnnnWfS09ONMdaai5P5411gxlhvTh566CGzadMms3v3brN161YzYMAA07RpU/v/oVabj48//ti4u7ubJ554wnzzzTfm1VdfNY0bNzavvPKKvcZqc1JTBKAzaN68eSY0NNR4enqaK6+80n7LsxVs3LjRSHJ6jRgxwhhz7JbNxx57zAQFBRkvLy9z7bXXms8//9y1Tdej6uZCklm8eLG9xkpzMnLkSPu/jQsuuMD07NnTHn6MsdZcnMyJAchqcxIXF2eCg4ONh4eHCQkJMTfffLPZuXOnfbnV5sMYY9auXWsiIiKMl5eXadeunXnhhRcclltxTmrCZowxrjn2BAAA4BpcAwQAACyHAAQAACyHAAQAACyHAAQAACyHAAQAACyHAAQAACyHAAQAACyHAATgnPD444+rU6dOrm6jVvbs2SObzabt27e7uhXAMghAAOrMHXfcIZvNJpvNJnd3d7Vq1Ur33nuvfvnlF1e31mBCxh133KFBgwa5tAcABCAAdaxPnz4qKCjQnj17tHDhQq1du1b33Xefq9sCAAcEIAB1ysvLS0FBQbrooosUGxuruLg4paenO9QsXrxY4eHh8vb2Vrt27TR//nyH5ePGjVPbtm3VuHFjtWnTRo8++qjKy8sdaqZPn67AwEA1bdpUo0aN0tGjR/9S38YYPfnkk2rTpo0aNWqkjh07auXKlfblmzZtks1m03vvvaeoqCg1btxYXbt2dfpG7qlTp6pFixZq2rSp7rrrLj388MP2U3OPP/64XnrpJf33v/+1HynbtGmTfd3du3erR48eaty4sTp27KisrKy/tE8ATsHF30UG4BwyYsQIM3DgQPv77777zrRv394EBgbax1544QUTHBxsVq1aZXbv3m1WrVplmjdvbpYsWWKvmTJlivnwww9NXl6eWbNmjQkMDDQzZsywL1+xYoXx9PQ0L774ovnqq6/MxIkTTdOmTU3Hjh1P2lteXp6RZHJycqpdPmHCBNOuXTvzzjvvmO+++84sXrzYeHl5mU2bNhlj/veFvl26dDGbNm0yO3fuNDExMaZr1672bbzyyivG29vbLFq0yOzatctMmjTJ+Pr62vs6dOiQ+dvf/mb69OljCgoKTEFBgSktLbX31q5dO/PWW2+ZXbt2mVtuucWEhoaa8vLyGvwNADhdBCAAdWbEiBHGzc3NNGnSxHh7e9u/5X7WrFn2mpYtW5r//Oc/DutNmTLFREdHn3S7Tz75pImMjLS/j46ONvHx8Q41Xbp0qXUAOnz4sPH29jZbtmxxGB81apT5+9//boz5XwBav369ffm6deuMJPP777/be7j//vsdttGtWzeHvk4MiX/sbeHChfaxnTt3GkkmNzf3pPsEoPY4BQagTvXo0UPbt2/XRx99pAcffFC9e/fWgw8+KEn66aeftG/fPo0aNUo+Pj7219SpU/Xdd9/Zt7Fy5Up1795dQUFB8vHx0aOPPqr8/Hz78tzcXEVHRzt87onva+LLL7/U0aNH1atXL4e+li5d6tCXJHXo0MH+5+DgYEnSgQMHJEm7du3S1Vdf7VB/4vtTOdW2AdQtd1c3AODc0qRJE11yySWSpDlz5qhHjx6aNGmSpkyZoqqqKknSiy++qC5dujis5+bmJknaunWrhg4dqkmTJql3797y8/PT8uXL9fTTT9dbz8f7WrdunS688EKHZV5eXg7vPTw87H+22WwO6/9x7DhjzGn38WfbBlB3CEAA6tVjjz2mvn376t5771VISIguvPBC7d69W7fddlu19R9++KFCQ0M1ceJE+9jevXsdasLDw7V161YNHz7cPrZ169Za99i+fXt5eXkpPz9f1113Xa23c9lll+njjz/WsGHD7GPZ2dkONZ6enqqsrKz1ZwCoGwQgAPXq+uuv1+WXX65p06Zp7ty5evzxx5WQkCBfX1/17dtXpaWlys7O1i+//KKkpCRdcsklys/P1/Lly3XVVVdp3bp1euONNxy2OXr0aI0YMUJRUVHq3r27Xn31Ve3cuVNt2rT5035OvGtLOhaAxo4dq8TERFVVVal79+4qKSnRli1b5OPjoxEjRpzWvj744IO6++67FRUVpa5du2rFihX67LPPHPpq3bq13n33Xe3atUv+/v7y8/M7rW0DqGOuvggJwLmjugt8jTHm1VdfNZ6eniY/P9/+vlOnTsbT09M0a9bMXHvttWb16tX2+n/961/G39/f+Pj4mLi4OPPMM88YPz8/h20+8cQTJiAgwPj4+JgRI0aY//u//zuti6Cre+Xl5Zmqqirz7LPPmssuu8x4eHiYCy64wPTu3dtkZmYaY/53EfQvv/xi32ZOTo59/eMmT55s72vkyJEmISHBXHPNNfblBw4cML169TI+Pj5Gktm4cWO1F2j/8ssv9uUA6p7NmBqcoAYA1EivXr0UFBSkl19+2dWtAPgDToEBQB357bfflJKSot69e8vNzU3Lli3T+vXrlZGR4erWAJyAI0AAUEd+//133XTTTfr0009VWlqqyy67TI888ohuvvlmV7cG4AQEIAAAYDk8CBEAAFgOAQgAAFgOAQgAAFgOAQgAAFgOAQgAAFgOAQgAAFgOAQgAAFgOAQgAAFgOAQgAAFjO/we+aWtAm7AOAQAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "reads = !cat {in_fn} | seqtk seq -A |grep -v '>'\n",
    "ds = [len(read) for read in reads]\n",
    "make_histogram(ds, \"Read Length\", \"Adapter Only Trimmed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "27157fb5-606b-4411-a4df-08f1a21f6291",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAkAAAAHFCAYAAAAaD0bAAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABEXUlEQVR4nO3de1xVVeL///eRqzfIRLkYIVoaRlpC40CSWYm3LmZ+wjQvpU2U5YWcSUPzmngpR30omhe8fKaUb2rzsaQZqdRxkjINrDHTJkUcg0jHwMsICuv3hz/OdAQUTuAR9+v5eOzHw7322muvvdjBu7X32cdmjDECAACwkHqu7gAAAMDVRgACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACAACWQwACasGCBQtks9kUHh5erf1WrVolm82m7Ozs2unY/2/GjBn685//XKvHcJVt27bJZrNp27Ztl603efJk2Ww2HT9+vMLt4eHhuu++++zr2dnZstlsstlsmjx5coX7PPPMM/Y6v3Tfffdd9loo63NVlspcrl911dChQ9WyZUtXdwPXKXdXdwC4HqWkpEiS9u3bp88//1ydOnVycY8czZgxQ/369VOfPn1c3ZU6p3Hjxlq1apVee+011av33/+HPH36tN599135+PiosLCwWm127NhRGRkZDmWPPfaYWrdurTfeeKNKbWRkZOimm26q1nEBK2MGCKhhu3fv1t69e9W7d29J0ooVK1zco6ujpKRERUVFru5GrYuLi9ORI0f08ccfO5SnpqaqpKREjzzySLXb9PHx0W9/+1uHxcvLSzfccEO58l8yxug///mPJOm3v/0tAQioBgIQUMPKAs/MmTMVHR2tdevW6ezZs+XqffbZZ7rnnnvk7e2toKAgjR8/XufPny9XLzU1VbGxsQoMDFT9+vUVFhamcePG6cyZMw71hg4dqkaNGmnfvn164IEH1LBhQzVr1kwvvviiw/FtNpvOnDmj1atX22+r/PJWT15enp577jnddNNN8vT0VGhoqKZMmaILFy7Y65TdDpo9e7amT5+u0NBQeXl5aevWrZWOy6JFi3TvvfeqefPmatiwoe644w7Nnj273DmX3S764osvFBMTowYNGqhVq1aaOXOmSktLHep+++236tGjhxo0aCA/Pz/Fx8fr1KlTlfahJrRt21bR0dH2Wb4yKSkp6tu3r3x9fWvt2DabTS+++KKWLFmisLAweXl5afXq1fZtv7wFVnY79ZNPPtGzzz6rpk2bysfHR4MHD9aZM2eUl5enJ554QjfccIMCAwM1duxYh59F2c94zpw5mjVrllq2bKn69evrvvvu08GDB3X+/HmNGzdOQUFB8vX11WOPPab8/PxyfU5NTVVUVJQaNmyoRo0aqXv37srMzCxXb9WqVWrbtq28vLwUFhamNWvW1PwAAr/ALTCgBv3nP//R2rVrdffddys8PFzPPPOMhg8frnfffVdDhgyx1/vmm2/0wAMPqGXLllq1apUaNGig5ORkvfPOO+Xa/O6779SrVy+NHj1aDRs21LfffqtZs2Zp165d+uSTTxzqnj9/Xr169dJzzz2ncePGaefOnZo+fbqOHDmi999/X9LFWyX333+/unbtqokTJ0q6OAMhXQw/v/nNb1SvXj299tprat26tTIyMjR9+nRlZ2dr5cqVDsdbsGCB2rRpozfeeEM+Pj669dZbKx2b77//XgMGDFBoaKg8PT21d+9evf766/r222/LhYm8vDwNHDhQL7/8siZNmqT33ntP48ePV1BQkAYPHixJ+vHHH9WlSxd5eHgoOTlZ/v7+evvtt/Xiiy9W9cfltGHDhmnEiBE6efKkmjRpogMHDtjHesOGDbV67D//+c/asWOHXnvtNQUEBKh58+aXrT98+HD17dtX69atU2Zmpl599VVduHBBBw4cUN++ffW73/1OH330kWbNmqWgoCAlJCQ47L9o0SK1b99eixYt0s8//6yXX35ZDz/8sDp16iQPDw+lpKToyJEjGjt2rIYPH65NmzbZ950xY4YmTJigp59+WhMmTFBxcbHmzJmjmJgY7dq1S+3atZN0Mfw8/fTTevTRR/Xmm2+qoKBAkydPVlFRkcNtRqBGGQA1Zs2aNUaSWbJkiTHGmFOnTplGjRqZmJgYh3pxcXGmfv36Ji8vz1524cIFc9tttxlJ5vDhwxW2X1paas6fP2+2b99uJJm9e/fatw0ZMsRIMvPnz3fY5/XXXzeSzN///nd7WcOGDc2QIUPKtf/cc8+ZRo0amSNHjjiUv/HGG0aS2bdvnzHGmMOHDxtJpnXr1qa4uPjKA3OJkpISc/78ebNmzRrj5uZm/v3vf9u3denSxUgyn3/+ucM+7dq1M927d7evv/LKK8Zms5msrCyHet26dTOSzNatWy/bh0mTJhlJ5qeffqpw++233266dOliXy875zlz5th/rgsXLjTGGPP73//ehIaGmtLSUjNixAhz6a/WLl26mNtvv/2y/blUSEiI6d27t0OZJOPr6+swXr/cNmnSJPv6ypUrjSTz0ksvOdTr06ePkWTmzp3rUH7nnXeajh07ljvfDh06mJKSEnv5vHnzjCTzyCOPOOw/evRoI8kUFBQYY4zJyckx7u7u5Y5/6tQpExAQYJ544gljzMVrISgoyHTs2NGUlpba62VnZxsPDw8TEhJS2RABvwrRGqhBK1asUP369dW/f39JUqNGjfQ///M/2rFjh7777jt7va1bt+qBBx6Qv7+/vczNzU1xcXHl2jx06JAGDBiggIAAubm5ycPDQ126dJEk7d+/v1z9gQMHOqwPGDDAfswr+eCDD9S1a1cFBQXpwoUL9qVnz56SpO3btzvUf+SRR+Th4XHFdiUpMzNTjzzyiJo2bWo/j8GDB6ukpEQHDx50qBsQEKDf/OY3DmXt27fXkSNH7Otbt27V7bffrg4dOlR4vrWp7OeakpKiCxcuaM2aNXr66acv+ymtmnL//ferSZMmVa7/0EMPOayHhYVJkv0ZtV+W/3J8y/Tq1cthFuZy+0tSTk6OJOmvf/2rLly4oMGDBztcS97e3urSpYv9U3oHDhzQDz/8oAEDBjiMX0hIiKKjo6t8nkB1cQsMqCH//Oc/9be//U2PP/64jDH6+eefJUn9+vXTypUrlZKSoqSkJEnSiRMnFBAQUK6NS8tOnz6tmJgYeXt7a/r06WrTpo0aNGigo0ePqm/fvvYHYMu4u7uradOmFbZ54sSJK57Djz/+qPfff7/SUHPpR8YDAwOv2KZ08Y9iTEyM2rZtq/nz56tly5by9vbWrl27NGLEiHLncek5SJKXl5dDvRMnTig0NLRcvYrGtSLu7hd//ZWUlFS4/cKFC5cNd8OGDVPnzp31+uuv66efftLQoUOrdNxfq6pjXubGG290WPf09Ky0/Ny5c79qf0n2Nn788UdJ0t13311hv8pCVdl1Wdl/D7X9SghYFwEIqCEpKSkyxmj9+vVav359ue2rV6/W9OnT5ebmpqZNmyovL69cnUvLPvnkE/3www/atm2bfdZHkj1cXerChQs6ceKEQ4Aoa7OiUHEpPz8/tW/fXq+//nqF24OCghzWqzrj8ec//1lnzpzRxo0bFRISYi/Pysqq0v4VqeoYVqZs9u3YsWMOM3HSxU9X5ebmKjIystL977nnHrVt21ZTp05Vt27dFBwcXI3eO+9qzDLVBD8/P0nS+vXrHX7mlyq7Ln/NzxJwBgEIqAElJSVavXq1WrdureXLl5fb/sEHH+jNN9/Uhx9+qIceekhdu3bVpk2b9OOPP9r/+JaUlCg1NdVhv7I/dl5eXg7lb731VqV9efvttzVy5Ej7etmD1b/8pNelsyllHnroIaWlpal169bVus1yJRWdhzFGy5Ytc7rNrl27avbs2dq7d6/DbbCKHiSvyP333y+bzabU1FR17NjRYdtf/vIXFRYW6sEHH7xsGxMmTND69es1YsSI6p/Ada579+5yd3fX999/r8cff7zSem3btlVgYKDWrl2rhIQE+7Vy5MgR7dy5s1zoBmoKAQioAR9++KF++OEHzZo1yyFolAkPD9fChQu1YsUKPfTQQ5owYYI2bdqk+++/X6+99poaNGigRYsWlftoe3R0tJo0aaL4+HhNmjRJHh4eevvtt7V3794K++Hp6ak333xTp0+f1t13323/ZFLPnj3VuXNne7077rhD27Zt0/vvv6/AwEA1btzYPpuRnp6u6OhojRw5Um3bttW5c+eUnZ2ttLQ0LVmyxKl3zXTr1k2enp568skn9Yc//EHnzp3T4sWLdfLkyWq3VWb06NFKSUlR7969NX36dPunwL799tsq7d+6dWu9+OKLmjNnjn7++Wf16tVL9evX1xdffKGZM2cqMjLyis8TPfXUU3rqqaecPofrWcuWLTV16lQlJibq0KFD6tGjh5o0aaIff/xRu3btUsOGDTVlyhTVq1dP06ZN0/Dhw/XYY4/p2Wef1c8//6zJkydX+XYm4AweggZqwIoVK+Tp6amnn366wu1+fn567LHH9MEHH+jHH39UeHi4PvroI/n4+GjIkCH63e9+p/bt29s/ll6madOm2rx5sxo0aKCnnnpKzzzzjBo1alRupqiMh4eHPvjgA6Wnp+vRRx/VggUL9Oyzz+rdd991qDd//nzdeuut6t+/v+6++24999xzki4+X7J7927FxsZqzpw56tGjhwYNGqSUlBTdeeedTs8K3XbbbdqwYYNOnjypvn376qWXXtKdd96pBQsWONWedPH5kO3bt6tdu3Z6/vnn9dRTT8nb21sLFy6schvz589XcnKyvvzySw0YMEAPP/ywVq9erREjRmjr1q3251rgnPHjx2v9+vU6ePCghgwZou7du+sPf/iDjhw5onvvvddeb9iwYVq+fLm++eYb9e3bV1OnTtWrr76q+++/34W9x/XOZowxru4EgF9v6NChWr9+vU6fPu3qrgDANY8ZIAAAYDkEIAAAYDncAgMAAJbDDBAAALAcAhAAALAcAhAAALAcXoRYgdLSUv3www9q3LhxnXntPAAAVmeM0alTpxQUFOTwJb4VIQBV4Icffrhq3+sDAABq1tGjR6/41noCUAUaN24s6eIA+vj4uLg3AACgKgoLCxUcHGz/O345BKAKlN328vHxIQABAFDHVOXxFR6CBgAAlkMAAgAAlkMAAgAAlkMAAgAAlkMAAgAAlkMAAgAAlkMAAgAAlkMAAgAAlkMAAgAAlkMAAgAAlkMAAgAAlkMAAgAAlkMAAgAAlkMAAgAAlkMAAgAAluPu6g4AAIBfp+W4za7uQrVlz+zt0uMzAwQAACyHAAQAACyHAAQAACyHAAQAACyHAAQAACyHAAQAACyHAAQAACyHAAQAACzH5QEoOTlZoaGh8vb2VkREhHbs2FFp3dzcXA0YMEBt27ZVvXr1NHr06Arr/fzzzxoxYoQCAwPl7e2tsLAwpaWl1dIZAACAusalASg1NVWjR49WYmKiMjMzFRMTo549eyonJ6fC+kVFRWrWrJkSExPVoUOHCusUFxerW7duys7O1vr163XgwAEtW7ZMLVq0qM1TAQAAdYhLvwpj7ty5GjZsmIYPHy5Jmjdvnv76179q8eLFSkpKKle/ZcuWmj9/viQpJSWlwjZTUlL073//Wzt37pSHh4ckKSQkpJbOAAAA1EUumwEqLi7Wnj17FBsb61AeGxurnTt3Ot3upk2bFBUVpREjRsjf31/h4eGaMWOGSkpKKt2nqKhIhYWFDgsAALh+uSwAHT9+XCUlJfL393co9/f3V15entPtHjp0SOvXr1dJSYnS0tI0YcIEvfnmm3r99dcr3ScpKUm+vr72JTg42OnjAwCAa5/LH4K22WwO68aYcmXVUVpaqubNm2vp0qWKiIhQ//79lZiYqMWLF1e6z/jx41VQUGBfjh496vTxAQDAtc9lzwD5+fnJzc2t3GxPfn5+uVmh6ggMDJSHh4fc3NzsZWFhYcrLy1NxcbE8PT3L7ePl5SUvLy+njwkAAOoWl80AeXp6KiIiQunp6Q7l6enpio6Odrrde+65R//85z9VWlpqLzt48KACAwMrDD8AAMB6XHoLLCEhQcuXL1dKSor279+vMWPGKCcnR/Hx8ZIu3poaPHiwwz5ZWVnKysrS6dOn9dNPPykrK0vffPONffvzzz+vEydOaNSoUTp48KA2b96sGTNmaMSIEVf13AAAwLXLpR+Dj4uL04kTJzR16lTl5uYqPDxcaWlp9o+t5+bmlnsn0F133WX/9549e/TOO+8oJCRE2dnZkqTg4GBt2bJFY8aMUfv27dWiRQuNGjVKr7zyylU7LwAAcG2zGWOMqztxrSksLJSvr68KCgrk4+Pj6u4AAHBZLcdtdnUXqi17Zu8ab7M6f79d/ikwAACAq40ABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALIcABAAALMflASg5OVmhoaHy9vZWRESEduzYUWnd3NxcDRgwQG3btlW9evU0evToy7a9bt062Ww29enTp2Y7DQAA6jSXBqDU1FSNHj1aiYmJyszMVExMjHr27KmcnJwK6xcVFalZs2ZKTExUhw4dLtv2kSNHNHbsWMXExNRG1wEAQB3m0gA0d+5cDRs2TMOHD1dYWJjmzZun4OBgLV68uML6LVu21Pz58zV48GD5+vpW2m5JSYkGDhyoKVOmqFWrVrXVfQAAUEe5LAAVFxdrz549io2NdSiPjY3Vzp07f1XbU6dOVbNmzTRs2LAq1S8qKlJhYaHDAgAArl8uC0DHjx9XSUmJ/P39Hcr9/f2Vl5fndLuffvqpVqxYoWXLllV5n6SkJPn6+tqX4OBgp48PAACufS5/CNpmszmsG2PKlVXVqVOn9NRTT2nZsmXy8/Or8n7jx49XQUGBfTl69KhTxwcAAHWDu6sO7OfnJzc3t3KzPfn5+eVmharq+++/V3Z2th5++GF7WWlpqSTJ3d1dBw4cUOvWrcvt5+XlJS8vL6eOCQAA6h6XzQB5enoqIiJC6enpDuXp6emKjo52qs3bbrtNX3/9tbKysuzLI488oq5duyorK4tbWwAAQJILZ4AkKSEhQYMGDVJkZKSioqK0dOlS5eTkKD4+XtLFW1PHjh3TmjVr7PtkZWVJkk6fPq2ffvpJWVlZ8vT0VLt27eTt7a3w8HCHY9xwww2SVK4cAABYl0sDUFxcnE6cOKGpU6cqNzdX4eHhSktLU0hIiKSLLz689J1Ad911l/3fe/bs0TvvvKOQkBBlZ2dfza4DAIA6zGaMMa7uxLWmsLBQvr6+KigokI+Pj6u7AwDAZbUct9nVXai27Jm9a7zN6vz9dvmnwAAAAK42AhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAcAhAAALAclweg5ORkhYaGytvbWxEREdqxY0eldXNzczVgwAC1bdtW9erV0+jRo8vVWbZsmWJiYtSkSRM1adJEDz74oHbt2lWLZwAAAOoalwag1NRUjR49WomJicrMzFRMTIx69uypnJycCusXFRWpWbNmSkxMVIcOHSqss23bNj355JPaunWrMjIydPPNNys2NlbHjh2rzVMBAAB1iM0YY1x18E6dOqljx45avHixvSwsLEx9+vRRUlLSZfe97777dOedd2revHmXrVdSUqImTZpo4cKFGjx4cJX6VVhYKF9fXxUUFMjHx6dK+wAA4Cotx212dReqLXtm7xpvszp/v102A1RcXKw9e/YoNjbWoTw2NlY7d+6sseOcPXtW58+f14033lhjbQIAgLrN3VUHPn78uEpKSuTv7+9Q7u/vr7y8vBo7zrhx49SiRQs9+OCDldYpKipSUVGRfb2wsLDGjg8AAK49Ln8I2mazOawbY8qVOWv27Nlau3atNm7cKG9v70rrJSUlydfX174EBwfXyPEBAMC1yWUByM/PT25ubuVme/Lz88vNCjnjjTfe0IwZM7Rlyxa1b9/+snXHjx+vgoIC+3L06NFffXwAAHDtclkA8vT0VEREhNLT0x3K09PTFR0d/avanjNnjqZNm6a//OUvioyMvGJ9Ly8v+fj4OCwAAOD65bJngCQpISFBgwYNUmRkpKKiorR06VLl5OQoPj5e0sWZmWPHjmnNmjX2fbKysiRJp0+f1k8//aSsrCx5enqqXbt2ki7e9po4caLeeecdtWzZ0j7D1KhRIzVq1OjqniAAALgmuTQAxcXF6cSJE5o6dapyc3MVHh6utLQ0hYSESLr44sNL3wl011132f+9Z88evfPOOwoJCVF2drakiy9WLC4uVr9+/Rz2mzRpkiZPnlyr5wMAAOoGl74H6FrFe4AAAHUJ7wG6qE68BwgAAMBVCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMByCEAAAMBynApAhw8frul+AAAAXDVOBaBbbrlFXbt21Z/+9CedO3eupvsEAABQq5wKQHv37tVdd92ll19+WQEBAXruuee0a9cupzqQnJys0NBQeXt7KyIiQjt27Ki0bm5urgYMGKC2bduqXr16Gj16dIX1NmzYoHbt2snLy0vt2rXTe++951TfAADA9cmpABQeHq65c+fq2LFjWrlypfLy8tS5c2fdfvvtmjt3rn766acqtZOamqrRo0crMTFRmZmZiomJUc+ePZWTk1Nh/aKiIjVr1kyJiYnq0KFDhXUyMjIUFxenQYMGae/evRo0aJCeeOIJff75586cKgAAuA7ZjDHm1zZSVFSk5ORkjR8/XsXFxfLw8FBcXJxmzZqlwMDASvfr1KmTOnbsqMWLF9vLwsLC1KdPHyUlJV32mPfdd5/uvPNOzZs3z6E8Li5OhYWF+vDDD+1lPXr0UJMmTbR27doqnU9hYaF8fX1VUFAgHx+fKu0DAICrtBy32dVdqLbsmb1rvM3q/P3+VZ8C2717t1544QUFBgZq7ty5Gjt2rL7//nt98sknOnbsmB599NFK9y0uLtaePXsUGxvrUB4bG6udO3c63aeMjIxybXbv3v2ybRYVFamwsNBhAQAA1y93Z3aaO3euVq5cqQMHDqhXr15as2aNevXqpXr1Luap0NBQvfXWW7rtttsqbeP48eMqKSmRv7+/Q7m/v7/y8vKc6ZYkKS8vr9ptJiUlacqUKU4fEwAA1C1OzQAtXrxYAwYMUE5Ojv785z/roYcesoefMjfffLNWrFhxxbZsNpvDujGmXFl1VbfN8ePHq6CgwL4cPXr0Vx0fAABc25yaAfruu++uWMfT01NDhgypdLufn5/c3NzKzczk5+eXm8GpjoCAgGq36eXlJS8vL6ePCQAA6hanZoBWrlypd999t1z5u+++q9WrV1epDU9PT0VERCg9Pd2hPD09XdHR0c50S5IUFRVVrs0tW7b8qjYBAMD1xakANHPmTPn5+ZUrb968uWbMmFHldhISErR8+XKlpKRo//79GjNmjHJychQfHy/p4q2pwYMHO+yTlZWlrKwsnT59Wj/99JOysrL0zTff2LePGjVKW7Zs0axZs/Ttt99q1qxZ+uijjyp9ZxAAALAep26BHTlyRKGhoeXKQ0JCKn2HT0Xi4uJ04sQJTZ06Vbm5uQoPD1daWppCQkIkXXzx4aXt3XXXXfZ/79mzR++8845CQkKUnZ0tSYqOjta6des0YcIETZw4Ua1bt1Zqaqo6derkxJkCAIDrkVMBqHnz5vrqq6/UsmVLh/K9e/eqadOm1WrrhRde0AsvvFDhtlWrVpUrq8pri/r166d+/fpVqx8AAMA6nLoF1r9/f40cOVJbt25VSUmJSkpK9Mknn2jUqFHq379/TfcRAACgRjk1AzR9+nQdOXJEDzzwgNzdLzZRWlqqwYMHV+sZIAAAAFdwKgB5enoqNTVV06ZN0969e1W/fn3dcccd9md3AAAArmVOBaAybdq0UZs2bWqqLwAAAFeFUwGopKREq1at0scff6z8/HyVlpY6bP/kk09qpHMAAAC1wakANGrUKK1atUq9e/dWeHj4r/7qCgAAgKvJqQC0bt06/b//9//Uq1evmu4PAABArXPqY/Cenp665ZZbarovAAAAV4VTAejll1/W/Pnzq/RSQgAAgGuNU7fA/v73v2vr1q368MMPdfvtt8vDw8Nh+8aNG2ukcwAAALXBqQB0ww036LHHHqvpvgAAAFwVTgWglStX1nQ/AAAArhqnngGSpAsXLuijjz7SW2+9pVOnTkmSfvjhB50+fbrGOgcAAFAbnJoBOnLkiHr06KGcnBwVFRWpW7duaty4sWbPnq1z585pyZIlNd1PAACAGuPUDNCoUaMUGRmpkydPqn79+vbyxx57TB9//HGNdQ4AAKA2OP0psE8//VSenp4O5SEhITp27FiNdAwAAKC2ODUDVFpaqpKSknLl//rXv9S4ceNf3SkAAIDa5FQA6tatm+bNm2dft9lsOn36tCZNmsTXYwAAgGueU7fA/vjHP6pr165q166dzp07pwEDBui7776Tn5+f1q5dW9N9BAAAqFFOBaCgoCBlZWVp7dq1+vLLL1VaWqphw4Zp4MCBDg9FAwAAXIucCkCSVL9+fT3zzDN65plnarI/AAAAtc6pALRmzZrLbh88eLBTnQEAALganApAo0aNclg/f/68zp49K09PTzVo0IAABAAArmlOfQrs5MmTDsvp06d14MABde7cmYegAQDANc/p7wK71K233qqZM2eWmx0CAAC41tRYAJIkNzc3/fDDDzXZJAAAQI1z6hmgTZs2OawbY5Sbm6uFCxfqnnvuqZGOAQAA1BanAlCfPn0c1m02m5o1a6b7779fb775Zk30CwAAoNY4FYBKS0truh8AAABXTY0+AwQAAFAXODUDlJCQUOW6c+fOdeYQAAAAtcapGaDMzEytWLFCb731lrZt26Zt27Zp6dKlWrFihTIzM+1LVlbWFdtKTk5WaGiovL29FRERoR07dly2/vbt2xURESFvb2+1atVKS5YsKVdn3rx5atu2rerXr6/g4GCNGTNG586dc+ZUAQDAdcipGaCHH35YjRs31urVq9WkSRNJF1+O+PTTTysmJkYvv/xyldpJTU3V6NGjlZycrHvuuUdvvfWWevbsqW+++UY333xzufqHDx9Wr1699Oyzz+pPf/qTPv30U73wwgtq1qyZHn/8cUnS22+/rXHjxiklJUXR0dE6ePCghg4dKunit9gDAADYjDGmuju1aNFCW7Zs0e233+5Q/o9//EOxsbFVfhdQp06d1LFjRy1evNheFhYWpj59+igpKalc/VdeeUWbNm3S/v377WXx8fHau3evMjIyJEkvvvii9u/fr48//the5+WXX9auXbuuOLtUprCwUL6+viooKJCPj0+V9gEAwFVajtvs6i5UW/bM3jXeZnX+fjt1C6ywsFA//vhjufL8/HydOnWqSm0UFxdrz549io2NdSiPjY3Vzp07K9wnIyOjXP3u3btr9+7dOn/+vCSpc+fO2rNnj3bt2iVJOnTokNLS0tS7d+UDXVRUpMLCQocFAABcv5wKQI899piefvpprV+/Xv/617/0r3/9S+vXr9ewYcPUt2/fKrVx/PhxlZSUyN/f36Hc399feXl5Fe6Tl5dXYf0LFy7o+PHjkqT+/ftr2rRp6ty5szw8PNS6dWt17dpV48aNq7QvSUlJ8vX1tS/BwcFVOgcAAFA3ORWAlixZot69e+upp55SSEiIQkJCNHDgQPXs2VPJycnVastmszmsG2PKlV2p/i/Lt23bptdff13Jycn68ssvtXHjRn3wwQeaNm1apW2OHz9eBQUF9uXo0aPVOgcAAFC3OPUQdIMGDZScnKw5c+bo+++/lzFGt9xyixo2bFjlNvz8/OTm5lZutic/P7/cLE+ZgICACuu7u7uradOmkqSJEydq0KBBGj58uCTpjjvu0JkzZ/S73/1OiYmJqlevfObz8vKSl5dXlfsOAADqtl/1IsTc3Fzl5uaqTZs2atiwoarzPLWnp6ciIiKUnp7uUJ6enq7o6OgK94mKiipXf8uWLYqMjJSHh4ck6ezZs+VCjpubm4wx1eofAAC4fjkVgE6cOKEHHnhAbdq0Ua9evZSbmytJGj58eJU/Ai9dfKHi8uXLlZKSov3792vMmDHKyclRfHy8pIu3pgYPHmyvHx8fryNHjighIUH79+9XSkqKVqxYobFjx9rrPPzww1q8eLHWrVunw4cPKz09XRMnTtQjjzwiNzc3Z04XAABcZ5y6BTZmzBh5eHgoJydHYWFh9vK4uDiNGTOmyl+IGhcXpxMnTmjq1KnKzc1VeHi40tLSFBISIuniDFNOTo69fmhoqNLS0jRmzBgtWrRIQUFBWrBggf0dQJI0YcIE2Ww2TZgwQceOHVOzZs308MMP6/XXX3fmVAEAwHXIqfcABQQE6K9//as6dOigxo0ba+/evWrVqpUOHz6sO+64Q6dPn66Nvl41vAcIAFCX8B6gi2r9PUBnzpxRgwYNypUfP36ch4kBAMA1z6kAdO+992rNmjX2dZvNptLSUs2ZM0ddu3atsc4BAADUBqeeAZozZ47uu+8+7d69W8XFxfrDH/6gffv26d///rc+/fTTmu4jAABAjXJqBqhdu3b66quv9Jvf/EbdunXTmTNn1LdvX2VmZqp169Y13UcAAIAaVe0ZoPPnzys2NlZvvfWWpkyZUht9AgAAqFXVngHy8PDQP/7xj8t+XQUAAMC1zKlbYIMHD9aKFStqui8AAABXhVMPQRcXF2v58uVKT09XZGRkue8Amzt3bo10DgAAoDZUKwAdOnRILVu21D/+8Q917NhRknTw4EGHOtwaAwAA17pqBaBbb71Vubm52rp1q6SLX2WxYMGCSr+9HQAA4FpUrWeALv3WjA8//FBnzpyp0Q4BAADUNqcegi7jxNeIAQAAuFy1ApDNZiv3jA/P/AAAgLqmWs8AGWM0dOhQ+xeenjt3TvHx8eU+BbZx48aa6yEAAEANq1YAGjJkiMP6U089VaOdAQAAuBqqFYBWrlxZW/0AAAC4an7VQ9AAAAB1EQEIAABYDgEIAABYDgEIAABYDgEIAABYDgEIAABYDgEIAABYDgEIAABYDgEIAABYDgEIAABYDgEIAABYDgEIAABYDgEIAABYDgEIAABYDgEIAABYjssDUHJyskJDQ+Xt7a2IiAjt2LHjsvW3b9+uiIgIeXt7q1WrVlqyZEm5Oj///LNGjBihwMBAeXt7KywsTGlpabV1CgAAoI5xaQBKTU3V6NGjlZiYqMzMTMXExKhnz57KycmpsP7hw4fVq1cvxcTEKDMzU6+++qpGjhypDRs22OsUFxerW7duys7O1vr163XgwAEtW7ZMLVq0uFqnBQAArnE2Y4xx1cE7deqkjh07avHixfaysLAw9enTR0lJSeXqv/LKK9q0aZP2799vL4uPj9fevXuVkZEhSVqyZInmzJmjb7/9Vh4eHk71q7CwUL6+viooKJCPj49TbQAAcLW0HLfZ1V2otuyZvWu8zer8/XbZDFBxcbH27Nmj2NhYh/LY2Fjt3Lmzwn0yMjLK1e/evbt2796t8+fPS5I2bdqkqKgojRgxQv7+/goPD9eMGTNUUlJSOycCAADqHHdXHfj48eMqKSmRv7+/Q7m/v7/y8vIq3CcvL6/C+hcuXNDx48cVGBioQ4cO6ZNPPtHAgQOVlpam7777TiNGjNCFCxf02muvVdhuUVGRioqK7OuFhYW/8uwAAMC1zOUPQdtsNod1Y0y5sivV/2V5aWmpmjdvrqVLlyoiIkL9+/dXYmKiw222SyUlJcnX19e+BAcHO3s6AACgDnBZAPLz85Obm1u52Z78/PxyszxlAgICKqzv7u6upk2bSpICAwPVpk0bubm52euEhYUpLy9PxcXFFbY7fvx4FRQU2JejR4/+mlMDAADXOJcFIE9PT0VERCg9Pd2hPD09XdHR0RXuExUVVa7+li1bFBkZaX/g+Z577tE///lPlZaW2uscPHhQgYGB8vT0rLBdLy8v+fj4OCwAAOD65dJbYAkJCVq+fLlSUlK0f/9+jRkzRjk5OYqPj5d0cWZm8ODB9vrx8fE6cuSIEhIStH//fqWkpGjFihUaO3asvc7zzz+vEydOaNSoUTp48KA2b96sGTNmaMSIEVf9/AAAwLXJZQ9BS1JcXJxOnDihqVOnKjc3V+Hh4UpLS1NISIgkKTc31+GdQKGhoUpLS9OYMWO0aNEiBQUFacGCBXr88cftdYKDg7VlyxaNGTNG7du3V4sWLTRq1Ci98sorV/38AADAtcml7wG6VvEeIABAXcJ7gC6qE+8BAgAAcBUCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsBwCEAAAsByXB6Dk5GSFhobK29tbERER2rFjx2Xrb9++XREREfL29larVq20ZMmSSuuuW7dONptNffr0qeFeAwCAusylASg1NVWjR49WYmKiMjMzFRMTo549eyonJ6fC+ocPH1avXr0UExOjzMxMvfrqqxo5cqQ2bNhQru6RI0c0duxYxcTE1PZpAACAOsalAWju3LkaNmyYhg8frrCwMM2bN0/BwcFavHhxhfWXLFmim2++WfPmzVNYWJiGDx+uZ555Rm+88YZDvZKSEg0cOFBTpkxRq1atrsapAACAOsRlAai4uFh79uxRbGysQ3lsbKx27txZ4T4ZGRnl6nfv3l27d+/W+fPn7WVTp05Vs2bNNGzYsCr1paioSIWFhQ4LAAC4frksAB0/flwlJSXy9/d3KPf391deXl6F++Tl5VVY/8KFCzp+/Lgk6dNPP9WKFSu0bNmyKvclKSlJvr6+9iU4OLiaZwMAAOoSlz8EbbPZHNaNMeXKrlS/rPzUqVN66qmntGzZMvn5+VW5D+PHj1dBQYF9OXr0aDXOAAAA1DXurjqwn5+f3Nzcys325Ofnl5vlKRMQEFBhfXd3dzVt2lT79u1Tdna2Hn74Yfv20tJSSZK7u7sOHDig1q1bl2vXy8tLXl5ev/aUAABAHeGyGSBPT09FREQoPT3doTw9PV3R0dEV7hMVFVWu/pYtWxQZGSkPDw/ddttt+vrrr5WVlWVfHnnkEXXt2lVZWVnc2gIAAJJcOAMkSQkJCRo0aJAiIyMVFRWlpUuXKicnR/Hx8ZIu3po6duyY1qxZI0mKj4/XwoULlZCQoGeffVYZGRlasWKF1q5dK0ny9vZWeHi4wzFuuOEGSSpXDgAArMulASguLk4nTpzQ1KlTlZubq/DwcKWlpSkkJESSlJub6/BOoNDQUKWlpWnMmDFatGiRgoKCtGDBAj3++OOuOgUAAFAH2UzZU8SwKywslK+vrwoKCuTj4+Pq7gAAcFktx212dReqLXtm7xpvszp/v13+KTAAAICrjQAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAshwAEAAAsx6XfBQYAuH7Vxa9nkGrnKxpw7WEGCAAAWA4BCAAAWA4BCAAAWA4BCAAAWA4BCAAAWA4BCAAAWA4BCAAAWA4BCAAAWA4BCAAAWA5vggYA4Bfq6husUT3MAAEAAMshAAEAAMshAAEAAMshAAEAAMshAAEAAMshAAEAAMshAAEAAMshAAEAAMshAAEAAMshAAEAAMtxeQBKTk5WaGiovL29FRERoR07dly2/vbt2xURESFvb2+1atVKS5Yscdi+bNkyxcTEqEmTJmrSpIkefPBB7dq1qzZPAQAA1DEuDUCpqakaPXq0EhMTlZmZqZiYGPXs2VM5OTkV1j98+LB69eqlmJgYZWZm6tVXX9XIkSO1YcMGe51t27bpySef1NatW5WRkaGbb75ZsbGxOnbs2NU6LQAAcI2zGWOMqw7eqVMndezYUYsXL7aXhYWFqU+fPkpKSipX/5VXXtGmTZu0f/9+e1l8fLz27t2rjIyMCo9RUlKiJk2aaOHChRo8eHCV+lVYWChfX18VFBTIx8enmmcFAJD4UlFcXvbM3jXeZnX+frvs2+CLi4u1Z88ejRs3zqE8NjZWO3furHCfjIwMxcbGOpR1795dK1as0Pnz5+Xh4VFun7Nnz+r8+fO68cYbK+1LUVGRioqK7OuFhYXVORUAqHWECaBmuewW2PHjx1VSUiJ/f3+Hcn9/f+Xl5VW4T15eXoX1L1y4oOPHj1e4z7hx49SiRQs9+OCDlfYlKSlJvr6+9iU4OLiaZwMAAOoSlz8EbbPZHNaNMeXKrlS/onJJmj17ttauXauNGzfK29u70jbHjx+vgoIC+3L06NHqnAIAAKhjXHYLzM/PT25ubuVme/Lz88vN8pQJCAiosL67u7uaNm3qUP7GG29oxowZ+uijj9S+ffvL9sXLy0teXl5OnAUAAKiLXDYD5OnpqYiICKWnpzuUp6enKzo6usJ9oqKiytXfsmWLIiMjHZ7/mTNnjqZNm6a//OUvioyMrPnOAwCAOs2lt8ASEhK0fPlypaSkaP/+/RozZoxycnIUHx8v6eKtqV9+cis+Pl5HjhxRQkKC9u/fr5SUFK1YsUJjx46115k9e7YmTJiglJQUtWzZUnl5ecrLy9Pp06ev+vkBAIBrk8tugUlSXFycTpw4oalTpyo3N1fh4eFKS0tTSEiIJCk3N9fhnUChoaFKS0vTmDFjtGjRIgUFBWnBggV6/PHH7XWSk5NVXFysfv36ORxr0qRJmjx58lU5LwAAcG1z6XuArlW8BwjAtYaPweN64+r3ALn8U2AAAABXm0tvgQGAKzCbAoAZIAAAYDkEIAAAYDkEIAAAYDkEIAAAYDkEIAAAYDkEIAAAYDkEIAAAYDkEIAAAYDkEIAAAYDkEIAAAYDkEIAAAYDkEIAAAYDkEIAAAYDkEIAAAYDkEIAAAYDkEIAAAYDkEIAAAYDnuru4AgLqt5bjNru4CAFQbAcgF6uIfjOyZvV3dBQAAagy3wAAAgOUQgAAAgOVwCwzXrbp4qxEAcHUwAwQAACyHAAQAACyHAAQAACyHZ4BQJTxPAwC4njADBAAALIcABAAALIcABAAALMflASg5OVmhoaHy9vZWRESEduzYcdn627dvV0REhLy9vdWqVSstWbKkXJ0NGzaoXbt28vLyUrt27fTee+/VVvcBAEAd5NIAlJqaqtGjRysxMVGZmZmKiYlRz549lZOTU2H9w4cPq1evXoqJiVFmZqZeffVVjRw5Uhs2bLDXycjIUFxcnAYNGqS9e/dq0KBBeuKJJ/T5559frdMCAADXOJsxxrjq4J06dVLHjh21ePFie1lYWJj69OmjpKSkcvVfeeUVbdq0Sfv377eXxcfHa+/evcrIyJAkxcXFqbCwUB9++KG9To8ePdSkSROtXbu2Sv0qLCyUr6+vCgoK5OPj4+zpVYpPVAEArK42vmS7On+/XTYDVFxcrD179ig2NtahPDY2Vjt37qxwn4yMjHL1u3fvrt27d+v8+fOXrVNZmwAAwHpc9h6g48ePq6SkRP7+/g7l/v7+ysvLq3CfvLy8CutfuHBBx48fV2BgYKV1KmtTkoqKilRUVGRfLygokHQxSdaG0qKztdIuAAB1RW38jS1rsyo3t1z+IkSbzeawbowpV3al+peWV7fNpKQkTZkypVx5cHBw5R0HAABO851Xe22fOnVKvr6+l63jsgDk5+cnNze3cjMz+fn55WZwygQEBFRY393dXU2bNr1sncralKTx48crISHBvl5aWqp///vfatq06WWDkzMKCwsVHByso0eP1srzRXUF43AR4/BfjMVFjMNFjMN/MRYXVWUcjDE6deqUgoKCrtieywKQp6enIiIilJ6erscee8xenp6erkcffbTCfaKiovT+++87lG3ZskWRkZHy8PCw10lPT9eYMWMc6kRHR1faFy8vL3l5eTmU3XDDDdU9pWrx8fGx9IVchnG4iHH4L8biIsbhIsbhvxiLi640Dlea+Snj0ltgCQkJGjRokCIjIxUVFaWlS5cqJydH8fHxki7OzBw7dkxr1qyRdPETXwsXLlRCQoKeffZZZWRkaMWKFQ6f7ho1apTuvfdezZo1S48++qj+7//+Tx999JH+/ve/u+QcAQDAtcelASguLk4nTpzQ1KlTlZubq/DwcKWlpSkkJESSlJub6/BOoNDQUKWlpWnMmDFatGiRgoKCtGDBAj3++OP2OtHR0Vq3bp0mTJigiRMnqnXr1kpNTVWnTp2u+vkBAIBrk8sfgn7hhRf0wgsvVLht1apV5cq6dOmiL7/88rJt9uvXT/369auJ7tU4Ly8vTZo0qdwtN6thHC5iHP6LsbiIcbiIcfgvxuKimh4Hl74IEQAAwBVc/l1gAAAAVxsBCAAAWA4BCAAAWA4BCAAAWA4BqBYkJSXp7rvvVuPGjdW8eXP16dNHBw4ccKhjjNHkyZMVFBSk+vXr67777tO+fftc1OPaUZVxGDp0qGw2m8Py29/+1kU9rj2LFy9W+/bt7S/wioqK0ocffmjfboXrQbryOFjlerhUUlKSbDabRo8ebS+zyjXxSxWNg1WuicmTJ5c7z4CAAPt2q1wPVxqHmrweCEC1YPv27RoxYoQ+++wzpaen68KFC4qNjdWZM2fsdWbPnq25c+dq4cKF+uKLLxQQEKBu3brp1KlTLux5zarKOEhSjx49lJuba1/S0tJc1OPac9NNN2nmzJnavXu3du/erfvvv1+PPvqo/ReYFa4H6crjIFnjevilL774QkuXLlX79u0dyq1yTZSpbBwk61wTt99+u8N5fv311/ZtVroeLjcOUg1eDwa1Lj8/30gy27dvN8YYU1paagICAszMmTPtdc6dO2d8fX3NkiVLXNXNWnfpOBhjzJAhQ8yjjz7quk65UJMmTczy5cstez2UKRsHY6x3PZw6dcrceuutJj093XTp0sWMGjXKGGO93xGVjYMx1rkmJk2aZDp06FDhNitdD5cbB2Nq9npgBugqKCgokCTdeOONkqTDhw8rLy9PsbGx9jpeXl7q0qWLdu7c6ZI+Xg2XjkOZbdu2qXnz5mrTpo2effZZ5efnu6J7V01JSYnWrVunM2fOKCoqyrLXw6XjUMZK18OIESPUu3dvPfjggw7lVrsmKhuHMla5Jr777jsFBQUpNDRU/fv316FDhyRZ73qobBzK1NT14PI3QV/vjDFKSEhQ586dFR4eLkn2b6u/9Bvq/f39deTIkavex6uhonGQpJ49e+p//ud/FBISosOHD2vixIm6//77tWfPnuvuradff/21oqKidO7cOTVq1Ejvvfee2rVrZ/8FZpXrobJxkKx1Paxbt05ffvmlvvjii3LbrPQ74nLjIFnnmujUqZPWrFmjNm3a6Mcff9T06dMVHR2tffv2Wep6uNw4NG3atEavBwJQLXvxxRf11VdfVfhlrDabzWHdGFOu7HpR2TjExcXZ/x0eHq7IyEiFhIRo8+bN6tu379XuZq1q27atsrKy9PPPP2vDhg0aMmSItm/fbt9uleuhsnFo166dZa6Ho0ePatSoUdqyZYu8vb0rrXe9XxNVGQerXBM9e/a0//uOO+5QVFSUWrdurdWrV9sf8r3erwfp8uOQkJBQo9cDt8Bq0UsvvaRNmzZp69atuummm+zlZU+0l6X6Mvn5+eUS/vWgsnGoSGBgoEJCQvTdd99dpd5dPZ6enrrlllsUGRmppKQkdejQQfPnz7fc9VDZOFTker0e9uzZo/z8fEVERMjd3V3u7u7avn27FixYIHd3d/vP/Xq/Jq40DiUlJeX2uV6viUs1bNhQd9xxh7777jvL/Y74pV+OQ0V+zfVAAKoFxhi9+OKL2rhxoz755BOFhoY6bA8NDVVAQIDS09PtZcXFxdq+fbuio6OvdndrzZXGoSInTpzQ0aNHFRgYeBV66FrGGBUVFVnmeqhM2ThU5Hq9Hh544AF9/fXXysrKsi+RkZEaOHCgsrKy1KpVK0tcE1caBzc3t3L7XK/XxKWKioq0f/9+BQYGWvp3xC/HoSK/6nqokUep4eD55583vr6+Ztu2bSY3N9e+nD171l5n5syZxtfX12zcuNF8/fXX5sknnzSBgYGmsLDQhT2vWVcah1OnTpmXX37Z7Ny50xw+fNhs3brVREVFmRYtWlxX42CMMePHjzd/+9vfzOHDh81XX31lXn31VVOvXj2zZcsWY4w1rgdjLj8OVroeKnLpp5+sck1c6pfjYKVr4uWXXzbbtm0zhw4dMp999pl56KGHTOPGjU12drYxxjrXw+XGoaavBwJQLZBU4bJy5Up7ndLSUjNp0iQTEBBgvLy8zL333mu+/vpr13W6FlxpHM6ePWtiY2NNs2bNjIeHh7n55pvNkCFDTE5Ojms7XgueeeYZExISYjw9PU2zZs3MAw88YA8/xljjejDm8uNgpeuhIpcGIKtcE5f65ThY6ZqIi4szgYGBxsPDwwQFBZm+ffuaffv22bdb5Xq43DjU9PVgM8YYJ2alAAAA6iyeAQIAAJZDAAIAAJZDAAIAAJZDAAIAAJZDAAIAAJZDAAIAAJZDAAIAAJZDAAJwXZg8ebLuvPNOV3fDKdnZ2bLZbMrKynJ1VwDLIAABqDFDhw6VzWaTzWaTu7u7br75Zj3//PM6efKkq7t2zYSMoUOHqk+fPi7tAwACEIAa1qNHD+Xm5io7O1vLly/X+++/rxdeeMHV3QIABwQgADXKy8tLAQEBuummmxQbG6u4uDht2bLFoc7KlSsVFhYmb29v3XbbbUpOTnbY/sorr6hNmzZq0KCBWrVqpYkTJ+r8+fMOdWbOnCl/f381btxYw4YN07lz535Vv40xmj17tlq1aqX69eurQ4cOWr9+vX37tm3bZLPZ9PHHHysyMlINGjRQdHS0Dhw44NDO9OnT1bx5czVu3FjDhw/XuHHj7LfmJk+erNWrV+v//u//7DNl27Zts+976NAhde3aVQ0aNFCHDh2UkZHxq84JwGXU2DeYAbC8IUOGmEcffdS+/v3335t27doZf39/e9nSpUtNYGCg2bBhgzl06JDZsGGDufHGG82qVavsdaZNm2Y+/fRTc/jwYbNp0ybj7+9vZs2aZd+emppqPD09zbJly8y3335rEhMTTePGjU2HDh0q7dvhw4eNJJOZmVnh9ldffdXcdttt5i9/+Yv5/vvvzcqVK42Xl5fZtm2bMcaYrVu3GkmmU6dOZtu2bWbfvn0mJibGREdH29v405/+ZLy9vU1KSoo5cOCAmTJlivHx8bH369SpU+aJJ54wPXr0MLm5uSY3N9cUFRXZ+3bbbbeZDz74wBw4cMD069fPhISEmPPnz1fjJwCgqghAAGrMkCFDjJubm2nYsKHx9vY2kowkM3fuXHud4OBg88477zjsN23aNBMVFVVpu7NnzzYRERH29aioKBMfH+9Qp1OnTk4HoNOnTxtvb2+zc+dOh/Jhw4aZJ5980hjz3wD00Ucf2bdv3rzZSDL/+c9/7H0YMWKEQxv33HOPQ78uDYm/7Nvy5cvtZfv27TOSzP79+ys9JwDO4xYYgBrVtWtXZWVl6fPPP9dLL72k7t2766WXXpIk/fTTTzp69KiGDRumRo0a2Zfp06fr+++/t7exfv16de7cWQEBAWrUqJEmTpyonJwc+/b9+/crKirK4biXrlfHN998o3Pnzqlbt24O/VqzZo1DvySpffv29n8HBgZKkvLz8yVJBw4c0G9+8xuH+peuX87l2gZQs9xd3QEA15eGDRvqlltukSQtWLBAXbt21ZQpUzRt2jSVlpZKkpYtW6ZOnTo57Ofm5iZJ+uyzz9S/f39NmTJF3bt3l6+vr9atW6c333yz1vpc1q/NmzerRYsWDtu8vLwc1j08POz/ttlsDvv/sqyMMabK/bhS2wBqDgEIQK2aNGmSevbsqeeff15BQUFq0aKFDh06pIEDB1ZY/9NPP1VISIgSExPtZUeOHHGoExYWps8++0yDBw+2l3322WdO97Fdu3by8vJSTk6OunTp4nQ7bdu21a5duzRo0CB72e7dux3qeHp6qqSkxOljAKgZBCAAteq+++7T7bffrhkzZmjhwoWaPHmyRo4cKR8fH/Xs2VNFRUXavXu3Tp48qYSEBN1yyy3KycnRunXrdPfdd2vz5s167733HNocNWqUhgwZosjISHXu3Flvv/229u3bp1atWl2xP5d+aku6GIDGjh2rMWPGqLS0VJ07d1ZhYaF27typRo0aaciQIVU615deeknPPvusIiMjFR0drdTUVH311VcO/WrZsqX++te/6sCBA2ratKl8fX2r1DaAGubqh5AAXD8qesDXGGPefvtt4+npaXJycuzrd955p/H09DRNmjQx9957r9m4caO9/u9//3vTtGlT06hRIxMXF2f++Mc/Gl9fX4c2X3/9dePn52caNWpkhgwZYv7whz9U6SHoipbDhw+b0tJSM3/+fNO2bVvj4eFhmjVrZrp37262b99ujPnvQ9AnT560t5mZmWnfv8zUqVPt/XrmmWfMyJEjzW9/+1v79vz8fNOtWzfTqFEjI8ls3bq1wge0T548ad8OoObZjKnGDWoAQLV069ZNAQEB+t///V9XdwXAL3ALDABqyNmzZ7VkyRJ1795dbm5uWrt2rT766COlp6e7umsALsEMEADUkP/85z96+OGH9eWXX6qoqEht27bVhAkT1LdvX1d3DcAlCEAAAMByeBEiAACwHAIQAACwHAIQAACwHAIQAACwHAIQAACwHAIQAACwHAIQAACwHAIQAACwHAIQAACwnP8PMTleUL6HvTAAAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "reads = !cat {out_fn} | seqtk seq -A |grep -v '>'\n",
    "ds = [len(read) for read in reads]\n",
    "make_histogram(ds, \"Read Length\", \"Adapter and UMI Trimmed\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e211e5a3",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Step: extract rRNA "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd6d76ec-6da9-4d26-b7da-76daaa33f260",
   "metadata": {},
   "source": [
    "#### Remove reads containing rRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "5a5471c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_path  = mkpath(\"extract_umi\")\n",
    "out_path = mkpath(\"extract_rRNA\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b864fcd0-b80f-45f1-9b05-cda5485dc4dc",
   "metadata": {},
   "source": [
    "#### bowtie2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8b42dcd-22a1-4152-b206-cf947913a3f8",
   "metadata": {},
   "source": [
    "https://bowtie-bio.sourceforge.net/bowtie2/index.shtml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "cf5767fd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bowtie 2 version 2.5.1 by Ben Langmead (langmea@cs.jhu.edu, www.cs.jhu.edu/~langmea)\n",
      "Usage: \n",
      "  bowtie2 [options]* -x <bt2-idx> {-1 <m1> -2 <m2> | -U <r> | --interleaved <i> | -b <bam>} [-S <sam>]\n",
      "\n",
      "  <bt2-idx>  Index filename prefix (minus trailing .X.bt2).\n",
      "             NOTE: Bowtie 1 and Bowtie 2 indexes are not compatible.\n",
      "  <m1>       Files with #1 mates, paired with files in <m2>.\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\n",
      "  <m2>       Files with #2 mates, paired with files in <m1>.\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\n",
      "  <r>        Files with unpaired reads.\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\n",
      "  <i>        Files with interleaved paired-end FASTQ/FASTA reads\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\n",
      "  <bam>      Files are unaligned BAM sorted by read name.\n",
      "  <sam>      File for SAM output (default: stdout)\n",
      "\n",
      "  <m1>, <m2>, <r> can be comma-separated lists (no whitespace) and can be\n",
      "  specified many times.  E.g. '-U file1.fq,file2.fq -U file3.fq'.\n",
      "\n",
      "Options (defaults in parentheses):\n",
      "\n",
      " Input:\n",
      "  -q                 query input files are FASTQ .fq/.fastq (default)\n",
      "  --tab5             query input files are TAB5 .tab5\n",
      "  --tab6             query input files are TAB6 .tab6\n",
      "  --qseq             query input files are in Illumina's qseq format\n",
      "  -f                 query input files are (multi-)FASTA .fa/.mfa\n",
      "  -r                 query input files are raw one-sequence-per-line\n",
      "  -F k:<int>,i:<int> query input files are continuous FASTA where reads\n",
      "                     are substrings (k-mers) extracted from a FASTA file <s>\n",
      "                     and aligned at offsets 1, 1+i, 1+2i ... end of reference\n",
      "  -c                 <m1>, <m2>, <r> are sequences themselves, not files\n",
      "  -s/--skip <int>    skip the first <int> reads/pairs in the input (none)\n",
      "  -u/--upto <int>    stop after first <int> reads/pairs (no limit)\n",
      "  -5/--trim5 <int>   trim <int> bases from 5'/left end of reads (0)\n",
      "  -3/--trim3 <int>   trim <int> bases from 3'/right end of reads (0)\n",
      "  --trim-to [3:|5:]<int> trim reads exceeding <int> bases from either 3' or 5' end\n",
      "                     If the read end is not specified then it defaults to 3 (0)\n",
      "  --phred33          qualities are Phred+33 (default)\n",
      "  --phred64          qualities are Phred+64\n",
      "  --int-quals        qualities encoded as space-delimited integers\n",
      "\n",
      " Presets:                 Same as:\n",
      "  For --end-to-end:\n",
      "   --very-fast            -D 5 -R 1 -N 0 -L 22 -i S,0,2.50\n",
      "   --fast                 -D 10 -R 2 -N 0 -L 22 -i S,0,2.50\n",
      "   --sensitive            -D 15 -R 2 -N 0 -L 22 -i S,1,1.15 (default)\n",
      "   --very-sensitive       -D 20 -R 3 -N 0 -L 20 -i S,1,0.50\n",
      "\n",
      "  For --local:\n",
      "   --very-fast-local      -D 5 -R 1 -N 0 -L 25 -i S,1,2.00\n",
      "   --fast-local           -D 10 -R 2 -N 0 -L 22 -i S,1,1.75\n",
      "   --sensitive-local      -D 15 -R 2 -N 0 -L 20 -i S,1,0.75 (default)\n",
      "   --very-sensitive-local -D 20 -R 3 -N 0 -L 20 -i S,1,0.50\n",
      "\n",
      " Alignment:\n",
      "  -N <int>           max # mismatches in seed alignment; can be 0 or 1 (0)\n",
      "  -L <int>           length of seed substrings; must be >3, <32 (22)\n",
      "  -i <func>          interval between seed substrings w/r/t read len (S,1,1.15)\n",
      "  --n-ceil <func>    func for max # non-A/C/G/Ts permitted in aln (L,0,0.15)\n",
      "  --dpad <int>       include <int> extra ref chars on sides of DP table (15)\n",
      "  --gbar <int>       disallow gaps within <int> nucs of read extremes (4)\n",
      "  --ignore-quals     treat all quality values as 30 on Phred scale (off)\n",
      "  --nofw             do not align forward (original) version of read (off)\n",
      "  --norc             do not align reverse-complement version of read (off)\n",
      "  --no-1mm-upfront   do not allow 1 mismatch alignments before attempting to\n",
      "                     scan for the optimal seeded alignments\n",
      "  --end-to-end       entire read must align; no clipping (on)\n",
      "   OR\n",
      "  --local            local alignment; ends might be soft clipped (off)\n",
      "\n",
      " Scoring:\n",
      "  --ma <int>         match bonus (0 for --end-to-end, 2 for --local) \n",
      "  --mp <int>         max penalty for mismatch; lower qual = lower penalty (6)\n",
      "  --np <int>         penalty for non-A/C/G/Ts in read/ref (1)\n",
      "  --rdg <int>,<int>  read gap open, extend penalties (5,3)\n",
      "  --rfg <int>,<int>  reference gap open, extend penalties (5,3)\n",
      "  --score-min <func> min acceptable alignment score w/r/t read length\n",
      "                     (G,20,8 for local, L,-0.6,-0.6 for end-to-end)\n",
      "\n",
      " Reporting:\n",
      "  (default)          look for multiple alignments, report best, with MAPQ\n",
      "   OR\n",
      "  -k <int>           report up to <int> alns per read; MAPQ not meaningful\n",
      "   OR\n",
      "  -a/--all           report all alignments; very slow, MAPQ not meaningful\n",
      "\n",
      " Effort:\n",
      "  -D <int>           give up extending after <int> failed extends in a row (15)\n",
      "  -R <int>           for reads w/ repetitive seeds, try <int> sets of seeds (2)\n",
      "\n",
      " Paired-end:\n",
      "  -I/--minins <int>  minimum fragment length (0)\n",
      "  -X/--maxins <int>  maximum fragment length (500)\n",
      "  --fr/--rf/--ff     -1, -2 mates align fw/rev, rev/fw, fw/fw (--fr)\n",
      "  --no-mixed         suppress unpaired alignments for paired reads\n",
      "  --no-discordant    suppress discordant alignments for paired reads\n",
      "  --dovetail         concordant when mates extend past each other\n",
      "  --no-contain       not concordant when one mate alignment contains other\n",
      "  --no-overlap       not concordant when mates overlap at all\n",
      "\n",
      " BAM:\n",
      "  --align-paired-reads\n",
      "                     Bowtie2 will, by default, attempt to align unpaired BAM reads.\n",
      "                     Use this option to align paired-end reads instead.\n",
      "  --preserve-tags    Preserve tags from the original BAM record by\n",
      "                     appending them to the end of the corresponding SAM output.\n",
      "\n",
      " Output:\n",
      "  -t/--time          print wall-clock time taken by search phases\n",
      "  --un <path>        write unpaired reads that didn't align to <path>\n",
      "  --al <path>        write unpaired reads that aligned at least once to <path>\n",
      "  --un-conc <path>   write pairs that didn't align concordantly to <path>\n",
      "  --al-conc <path>   write pairs that aligned concordantly at least once to <path>\n",
      "    (Note: for --un, --al, --un-conc, or --al-conc, add '-gz' to the option name, e.g.\n",
      "    --un-gz <path>, to gzip compress output, or add '-bz2' to bzip2 compress output.)\n",
      "  --quiet            print nothing to stderr except serious errors\n",
      "  --met-file <path>  send metrics to file at <path> (off)\n",
      "  --met-stderr       send metrics to stderr (off)\n",
      "  --met <int>        report internal counters & metrics every <int> secs (1)\n",
      "  --no-unal          suppress SAM records for unaligned reads\n",
      "  --no-head          suppress header lines, i.e. lines starting with @\n",
      "  --no-sq            suppress @SQ header lines\n",
      "  --rg-id <text>     set read group id, reflected in @RG line and RG:Z: opt field\n",
      "  --rg <text>        add <text> (\"lab:value\") to @RG line of SAM header.\n",
      "                     Note: @RG line only printed when --rg-id is set.\n",
      "  --omit-sec-seq     put '*' in SEQ and QUAL fields for secondary alignments.\n",
      "  --sam-no-qname-trunc\n",
      "                     Suppress standard behavior of truncating readname at first whitespace \n",
      "                     at the expense of generating non-standard SAM.\n",
      "  --xeq              Use '='/'X', instead of 'M,' to specify matches/mismatches in SAM record.\n",
      "  --soft-clipped-unmapped-tlen\n",
      "                     Exclude soft-clipped bases when reporting TLEN\n",
      "  --sam-append-comment\n",
      "                     Append FASTA/FASTQ comment to SAM record\n",
      "\n",
      " Performance:\n",
      "  -p/--threads <int> number of alignment threads to launch (1)\n",
      "  --reorder          force SAM output order to match order of input reads\n",
      "  --mm               use memory-mapped I/O for index; many 'bowtie's can share\n",
      "\n",
      " Other:\n",
      "  --qc-filter        filter out reads that are bad according to QSEQ filter\n",
      "  --seed <int>       seed for random number generator (0)\n",
      "  --non-deterministic\n",
      "                     seed rand. gen. arbitrarily instead of using read attributes\n",
      "  --version          print version information and quit\n",
      "  -h/--help          print this usage message\n"
     ]
    }
   ],
   "source": [
    "!bowtie2 --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ceaf71bf-4db4-4a00-9ecf-ab0d3f05c14b",
   "metadata": {},
   "source": [
    "##### paramaters\n",
    "- --threads, number of alignment threads to launch\n",
    "- -L 15, length of seed substrings  3< -L < 32\n",
    "- -x {rRNA_index}, index dir/basename\n",
    "- -U {in_fn}, input files with unpaired reads\n",
    "- --un {out_fn}, output file of unaligned reads"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "327dff28",
   "metadata": {},
   "source": [
    "`--un {out_fn}` is the output file for reads that DO NOT ALIGN with `{rRNA_index}`, i.e., reads without rRNA\n",
    "\n",
    "`alignment rate` means (%) aligned with rRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "834bef7d-4f25-4b99-a4e9-9c2e166bce21",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "extract_umi/a4_1.fq\n",
      "8463966 reads; of these:\n",
      "  8463966 (100.00%) were unpaired; of these:\n",
      "    3605893 (42.60%) aligned 0 times\n",
      "    4857586 (57.39%) aligned exactly 1 time\n",
      "    487 (0.01%) aligned >1 times\n",
      "57.40% overall alignment rate\n",
      "++++++++++++++++++++++++++++++++++++\n",
      "extract_umi/a4_2.fq\n",
      "2710933 reads; of these:\n",
      "  2710933 (100.00%) were unpaired; of these:\n",
      "    1094741 (40.38%) aligned 0 times\n",
      "    1616175 (59.62%) aligned exactly 1 time\n",
      "    17 (0.00%) aligned >1 times\n",
      "59.62% overall alignment rate\n",
      "++++++++++++++++++++++++++++++++++++\n",
      "extract_umi/a4_3.fq\n",
      "7345700 reads; of these:\n",
      "  7345700 (100.00%) were unpaired; of these:\n",
      "    2478074 (33.74%) aligned 0 times\n",
      "    4867539 (66.26%) aligned exactly 1 time\n",
      "    87 (0.00%) aligned >1 times\n",
      "66.26% overall alignment rate\n",
      "++++++++++++++++++++++++++++++++++++\n",
      "extract_umi/a4_4.fq\n",
      "4864834 reads; of these:\n",
      "  4864834 (100.00%) were unpaired; of these:\n",
      "    1741926 (35.81%) aligned 0 times\n",
      "    3122862 (64.19%) aligned exactly 1 time\n",
      "    46 (0.00%) aligned >1 times\n",
      "64.19% overall alignment rate\n",
      "++++++++++++++++++++++++++++++++++++\n",
      "extract_umi/no_1.fq\n",
      "7500090 reads; of these:\n",
      "  7500090 (100.00%) were unpaired; of these:\n",
      "    2432106 (32.43%) aligned 0 times\n",
      "    5067929 (67.57%) aligned exactly 1 time\n",
      "    55 (0.00%) aligned >1 times\n",
      "67.57% overall alignment rate\n",
      "++++++++++++++++++++++++++++++++++++\n",
      "extract_umi/no_2.fq\n",
      "5671357 reads; of these:\n",
      "  5671357 (100.00%) were unpaired; of these:\n",
      "    2027040 (35.74%) aligned 0 times\n",
      "    3643828 (64.25%) aligned exactly 1 time\n",
      "    489 (0.01%) aligned >1 times\n",
      "64.26% overall alignment rate\n",
      "++++++++++++++++++++++++++++++++++++\n",
      "extract_umi/no_3.fq\n",
      "4618416 reads; of these:\n",
      "  4618416 (100.00%) were unpaired; of these:\n",
      "    1433973 (31.05%) aligned 0 times\n",
      "    3184399 (68.95%) aligned exactly 1 time\n",
      "    44 (0.00%) aligned >1 times\n",
      "68.95% overall alignment rate\n",
      "++++++++++++++++++++++++++++++++++++\n",
      "extract_umi/no_4.fq\n",
      "6082913 reads; of these:\n",
      "  6082913 (100.00%) were unpaired; of these:\n",
      "    1892186 (31.11%) aligned 0 times\n",
      "    4190718 (68.89%) aligned exactly 1 time\n",
      "    9 (0.00%) aligned >1 times\n",
      "68.89% overall alignment rate\n",
      "++++++++++++++++++++++++++++++++++++\n"
     ]
    }
   ],
   "source": [
    "for sample in samples:\n",
    "    print(fname(in_path,sample,\"fq\"))\n",
    "    !bowtie2 --threads {nc} -L 15 -x {rRNA_index} \\\n",
    "               -U {fname(in_path,sample,\"fq\")} --un {fname(out_path,sample,\"fq\")} > /dev/null\n",
    "    print(\"++++++++++++++++++++++++++++++++++++\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4565988-83c6-4ab5-9b75-288ace093f82",
   "metadata": {},
   "source": [
    "#### Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "5d1acbdc-b184-43ea-824e-d70f9e1f97eb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[8463966, 2710933, 7345700, 4864834, 7500090, 5671357, 4618416, 6082913]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Pre rRNA Extraction\n",
    "ins = [nseqs(fname(in_path,sample,'fq')) for sample in samples]\n",
    "ins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "ac263016-357d-4f14-a5bb-80b74f292549",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[3605893, 1094741, 2478074, 1741926, 2432106, 2027040, 1433973, 1892186]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Post rRNA Extraction\n",
    "outs = [nseqs(fname(out_path,sample,'fq')) for sample in samples]\n",
    "outs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "7aaf4ef6-974b-4469-ac86-adda3025be9a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAioAAAGxCAYAAABMeZ2uAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABIbUlEQVR4nO3deVxN+f8H8Ndtu9KqlIqoEGWNxpalxjLWYRYGDbINsve1ZStrGIw9xpIwYmYsM2MN2ZdRaGzZSYNkLULb/fz+8Oj+5ip0697uSa/n43Eej+7nnPs570/3ml7zOZtMCCFAREREJEF6ui6AiIiI6H0YVIiIiEiyGFSIiIhIshhUiIiISLIYVIiIiEiyGFSIiIhIshhUiIiISLIYVIiIiEiyGFSIiIhIshhUiN5j7dq1kMlkysXAwAD29vbo2rUrrl+/ruvyCp23tze8vb012ufEiRNRvnx5GBgYwNLSMl99+Pn5qXxORkZGqFixIkaNGoWUlJQc22dvN2vWrBzrsj/zmJiYXPcVEBAAmUyG9u3bq1Wjk5OTSo0mJiaoU6cOlixZAl3dHNzPzw9OTk462TeROhhUiD4iLCwMJ0+exP79+zFkyBD8+eefaNy4MZ49e6br0oq0P/74AzNmzEDPnj1x+PBh7N+/P999GRsb4+TJkzh58iT+/PNP+Pj4YN68efj222/f+55Zs2bh6dOned5HRkYGNmzYAADYs2cP7t27p1aNXl5eyhrXr1+PkiVLYujQoQgJCVGrH6LihkGF6COqV6+OBg0awNvbGxMmTMC4ceOQlJSE7du367q0Iu3ixYsAgGHDhsHLywuenp5qvf/Vq1fKn/X09NCgQQM0aNAArVu3xqpVq+Dj44N9+/bh9u3bOd7bokULpKamYsaMGXne3x9//IFHjx6hXbt2yMrKQnh4uFr1WlpaKmv8+uuvsWvXLlhYWGDFihVq9UNU3DCoEKkp+w/qw4cPVdpjYmLw5ZdfwsrKCiVKlICHhwd+/fVXlW2yDy1ERUWhf//+sLa2hrm5OXr27InU1FQkJiaiS5cusLS0hL29PUaNGoWMjAyVPp4+fQp/f3+ULVsWRkZGcHFxwYQJE5CWlqbcxsPDA02aNMlRe1ZWFsqWLYuvv/5a2TZlyhTUr18fVlZWMDc3R506dbB69ep8H5JQKBSYM2cOqlatCrlcDltbW/Ts2RP//vuvchsnJydMnDgRAFCmTBnIZDIEBwe/t08/Pz+YmpriwoULaNWqFczMzNC8efMP1vG+zwkAqlSpgr59+2Lp0qWIj4/P07hWr14NIyMjhIWFwdHREWFhYQU6bGNubg5XV9cc9aWnp2P69OnK35+NjQ169+6NR48eqWy3efNmtGrVCvb29jA2NoabmxvGjRuH1NTUHPtau3YtqlSpArlcDjc3N6xbty7XmkJDQ1GrVi2YmprCzMwMVatWxfjx4/M9RiJNMNB1AURFTfb/obu6uirbDh48iNatW6N+/fpYvnw5LCwssGnTJnz33Xd49eoV/Pz8VPro168fvv76a2zatAnnzp3D+PHjkZmZiatXr+Lrr7/GDz/8gP3792P27NlwcHBAQEAAAODNmzfw8fHBzZs3MWXKFNSsWRNHjx5FSEgIYmNjsXPnTgBA7969MXz4cFy/fh2VK1dW7jcyMhL3799H7969lW137tzBgAEDUL58eQDAqVOnMHToUNy7dw+TJ09W+/czaNAg/PzzzxgyZAjat2+PO3fuYNKkSTh06BDOnj2L0qVLY9u2bVi6dClWr16NPXv2wMLCAuXKlftgv+np6fjyyy8xYMAAjBs3DpmZmR/c/vbt2zAwMICLi0uu64ODg7F+/XpMmjTpvX+4s/3777+IjIzEN998AxsbG/Tq1QvTp0/HkSNH0KxZsw//Qt4jMzMTCQkJKt8jhUKBjh074ujRoxgzZgwaNWqE+Ph4BAUFwdvbGzExMTA2NgYAXL9+HW3btsWIESNgYmKCK1euYPbs2Th9+jSioqKUfa5duxa9e/dGx44dMW/ePCQnJyM4OBhpaWnQ0/v//1fdtGkT/P39MXToUMydOxd6enq4ceMGLl++nK/xEWmMIKJchYWFCQDi1KlTIiMjQ7x48ULs2bNH2NnZiaZNm4qMjAzltlWrVhUeHh4qbUII0b59e2Fvby+ysrJU+hw6dKjKdp06dRIAxPz581Xaa9euLerUqaN8vXz5cgFA/PrrryrbzZ49WwAQkZGRQgghHj9+LIyMjMT48eNVtuvSpYsoU6ZMjjqzZWVliYyMDDF16lRhbW0tFAqFcl2zZs1Es2bNPvQrE3FxcQKA8Pf3V2n/+++/BQCVeoKCggQA8ejRow/2KYQQvXr1EgDEmjVrcl1nYmIiMjIyREZGhnj8+LEIDQ0Venp6OcYvhBAAxODBg4UQQkyYMEHo6emJf/75Rwjx/59PdHS0ynumTp0qAIg9e/YIIYS4deuWkMlkokePHh+tXQghKlSoINq2bausMT4+XvTv318YGhqKHTt2KLeLiIgQAMSWLVtU3h8dHS0AiGXLluXav0KhEBkZGeLw4cMCgHI8WVlZwsHBQdSpU0fls7xz544wNDQUFSpUULYNGTJEWFpa5mk8RIXpkwkqhw8fVv5RACC2bdumdh8KhUL8+OOPonLlysLIyEiUK1dOzJgxQ/PFUpGQ/Ufr3cXNzU08e/ZMud3169cFADF37lzlH6LsZdmyZQKAuHz5skqfe/fuVdlXYGCgACCuXr2q0t6tWzdhbW2tfN2lSxdhYmKi8kdHCCEePnwoAIixY8cq27755htRtmxZZUh6+vSpkMvlYvTo0SrvPXDggGjevLkwNzfPMdbExETldnkJKtnjPX36dI51bm5uon79+srX+QkqycnJ71337tKtW7dc+/pvUElOThalS5cWrVu3FkLkHlQUCoVwdnYWjo6Oyt+lEEL4+PiIkiVL5lrTuypUqJBrjcuXL1fZztfXV1haWor09PQc3yU7OzvRpUsX5bY3b94U3bp1E2XKlBEymUyl302bNgkhhLh8+bLyu/muZs2aqQSVdevWCQCia9euYvv27Xn6XIgKwydzjkpqaipq1aqFJUuW5LuP4cOHY9WqVZg7dy6uXLmCv/76C/Xq1dNglVQUrVu3DtHR0YiKisKAAQMQFxeHbt26Kddnn2MwatQoGBoaqiz+/v4AgMePH6v0aWVlpfLayMjove1v3rxRvn7y5Ans7Owgk8lUtrO1tYWBgQGePHmibOvTpw/u3buHffv2AQAiIiKQlpamchjq9OnTaNWqFQBg5cqVOH78OKKjozFhwgQAwOvXr/P4W/r/+gDA3t4+xzoHBweV+tRVsmRJmJub57rO2NgY0dHRiI6Oxl9//QVvb29ERETkegnyf5mbm2PixInYs2cPDh48mOs2UVFRuH37Njp37oyUlBQ8f/4cz58/R5cuXfDq1StERETkqf7GjRsjOjoap06dwvr16+Hk5IQhQ4bg2LFjym0ePnyI58+fw8jIKMd3KTExUfk9evnyJZo0aYK///4b06dPx6FDhxAdHY2tW7cC+P/PLfv3bWdnl6Oed9t69OiBNWvWID4+Ht988w1sbW1Rv3595feHSFc+mXNU2rRpgzZt2rx3fXp6OiZOnIhffvkFz58/R/Xq1TF79mzlfSHi4uIQGhqKixcvokqVKoVUNRUFbm5uyhMzfXx8kJWVhVWrVuH333/Ht99+i9KlSwMAAgMDVU5S/S9Nfaesra3x999/QwihElaSkpKQmZmprAUAvvjiCzg4OCAsLAxffPEFwsLCUL9+fbi7uyu32bRpEwwNDbFjxw6UKFFC2Z7fK5qsra0BAA8ePMhxzsn9+/dV6lPXu+Hsv/T09FSuGmrZsiXq1q2LKVOmwNfXF46Oju9976BBg7Bw4UKMHTsWgwYNyrF+9erVAID58+dj/vz5ua4fMGDAR+u3sLBQ1li/fn3Ur18ftWrVgr+/P2JjY6Gnp4fSpUvD2toae/bsybUPMzMzAG/D0/3793Ho0CGVc2SeP3+usn3255GYmJijr9zaevfujd69eyM1NRVHjhxBUFAQ2rdvj2vXrqFChQofHSORNnwyMyof07t3bxw/fhybNm3C+fPn0blzZ7Ru3Vp5466//voLLi4u2LFjB5ydneHk5IR+/fqpdZ8FKh7mzJmDUqVKYfLkyVAoFKhSpQoqV66Mf/75B56enrku2X9gCqp58+Z4+fJljiCRfTLof6+E0dfXR48ePbB9+3YcPXoUMTEx6NOnj8r7sm9kp6+vr2x7/fo11q9fn6/6Pv/8cwBQ3m8kW3R0NOLi4j56pY6myOVyLF26FG/evMH06dM/uK2RkRGmT5+O6Oho/Pbbbyrrnj17hm3btsHLywsHDx7Msfj6+iI6Olp5qbU6KleujDFjxuDChQvYvHkzAKB9+/Z48uQJsrKycv0eZQfe7NAml8tV+nz3UucqVarA3t4eERERKlcoxcfH48SJE++tzcTEBG3atMGECROQnp6OS5cuqT0+Io3R9bEnbcA756jcuHFDyGQyce/ePZXtmjdvLgIDA4UQQgwYMEDI5XJRv359ceTIEXHw4EFRu3Zt4ePjU5ilk4S878RKIYSYM2eOACDWr18vhBAiKipKyOVy0apVK7Fx40Zx+PBhsW3bNjFz5kzx7bfffrTP952vkX2iaLbXr1+LmjVrCjMzMzF//nyxb98+ERQUJAwNDUXbtm1z1Hn16lUBQJQrV04YGxuL58+fq6w/cOCAACC+/fZbERkZKSIiIkTdunVF5cqVBQBx+/Zt5bZ5OUdFCCF++OEHIZPJxIgRI8TevXvFihUrhK2trXB0dBSPHz/+6Jhz8+7vIa/r2rZtKwwNDcWtW7eUbfjPOSrZFAqF8PDwUJ7jkf35LF68WAAQmzdvzrX/8+fPCwBixIgRH6y/QoUKol27djnaX7x4IcqUKSOqVKkiMjMzRWZmpmjTpo2wsrISU6ZMEbt37xb79+8Xa9euFb169RJbt24VQrw9WbpUqVKiVq1aYuvWreKvv/4SXbt2VX5uYWFhyn2sWrVKABAdO3YUO3bsEBs2bBCVKlUSjo6OKueo9OvXTwwdOlRs2rRJHD58WGzevFnUrl1bWFhYiKSkpA+Oj0ibikVQ+fXXXwUAYWJiorIYGBgoT07r379/jpMZz5w5IwCIK1euFPYQSAI+FFRev34typcvLypXriwyMzOFEEL8888/okuXLsLW1lYYGhoKOzs78fnnn6ucMFnQoCKEEE+ePBEDBw4U9vb2wsDAQFSoUEEEBgaKN2/e5DqORo0aCQDC19c31/Vr1qwRVapUEXK5XLi4uIiQkBCxevXqfAeVrKwsMXv2bOHq6ioMDQ1F6dKlxffffy8SEhLyNObc5DeoXLhwQejp6YnevXsr23ILKkIIERkZmSOo1K5dW9ja2oq0tLT31tagQQNRunTpD27zvqAihBBLly4VAER4eLgQQoiMjAwxd+5cUatWLVGiRAlhamoqqlatKgYMGCCuX7+ufN+JEydEw4YNRcmSJYWNjY3o16+fOHv2bI6gIsTbsJJ9kYCrq6tYs2aN6NWrl0pQCQ8PFz4+PqJMmTLCyMhIODg4iC5duojz58+/d1xEhUEmhI4eNKFFMpkM27ZtQ6dOnQC8vTGSr68vLl26pDLFDQCmpqaws7NDUFAQZs6cqXJzrdevX6NkyZKIjIxEy5YtC3MIREREhE/oZNoP8fDwQFZWFpKSknK9Wyfw9jkcmZmZuHnzJipWrAgAuHbtGgDwJDIiIiId+WRmVF6+fIkbN24AeBtM5s+fDx8fH1hZWaF8+fL4/vvvcfz4ccybNw8eHh54/PgxoqKiUKNGDbRt2xYKhQKfffYZTE1NsWDBAigUCgwePBjm5uaIjIzU8eiIiIiKp08mqBw6dAg+Pj452nv16oW1a9ciIyMD06dPx7p163Dv3j1YW1ujYcOGmDJlCmrUqAHg7eWTQ4cORWRkpPKs93nz5uW4twUREREVjk8mqBAREdGnp9jcR4WIiIiKHgYVIiIikqwifdWPQqHA/fv3YWZm9sHbaxMREZF0CCHw4sULODg4QE/vw3MmRTqo3L9//4PP8CAiIiLpSkhIyPFcsHcV6aCS/fyUhISE9z5VlYiIiKQlJSUFjo6OeXoOWpEOKtmHe8zNzRlUiIiIipi8nLbBk2mJiIhIshhUiIiISLIYVIiIiEiyivQ5KkRElH9ZWVkqT4wn0hR9fX0YGBho5NYhDCpERMXQy5cv8e+//4JPUSFtKVmyJOzt7WFkZFSgfhhUiIiKmaysLPz7778oWbIkbGxseMNM0ighBNLT0/Ho0SPcvn0blStX/uhN3T6EQYWIqJjJyMiAEAI2NjYwNjbWdTn0CTI2NoahoSHi4+ORnp6OEiVK5LsvnkxLRFRMcSaFtKkgsygq/WikFyIiIiItYFAhIiIiyeI5KkREBABwGrezUPd3Z1a7Qt1fcXPnzh04Ozvj3LlzqF27tq7LyTfOqBARUZHg5+cHmUwGmUwGQ0NDuLi4YNSoUUhNTdX4vrL38+6yadOmPL3/zp07kMlkiI2N1XhtufHz80OnTp1U2hwdHfHgwQNUr169UGrQFs6oEBFRkdG6dWuEhYUhIyMDR48eRb9+/ZCamorQ0NAc22ZkZMDQ0DDf+woLC0Pr1q1V2iwtLfPdX27S09MLfJ+R99HX14ednZ1W+i5MnFEhIqIiQy6Xw87ODo6OjujevTt8fX2xfft2AEBwcDBq166NNWvWwMXFBXK5HEIIJCcn44cffoCtrS3Mzc3x+eef459//vnoviwtLWFnZ6eyZF9m26dPH9SsWRNpaWkA3oaiunXrwtfXFwDg7OwMAPDw8IBMJoO3tzeA/5/5CAkJgYODA1xdXQEAGzZsgKenJ8zMzGBnZ4fu3bsjKSlJpZ5Lly6hXbt2MDc3h5mZGZo0aYKbN28iODgY4eHh+OOPP5QzP4cOHcp1Vufw4cOoV68e5HI57O3tMW7cOGRmZirXe3t7Y9iwYRgzZgysrKxgZ2eH4OBgtT8nTeKMygdo+ngtj8cSEWmWsbGxymMAbty4gV9//RVbtmyBvr4+AKBdu3awsrLCrl27YGFhgRUrVqB58+a4du0arKys8rXfRYsWoVatWhg3bhx++uknTJo0CY8fP8ayZcsAAKdPn0a9evWwf/9+VKtWTWXW5MCBAzA3N8e+ffuUdwZOT0/HtGnTUKVKFSQlJWHkyJHw8/PDrl27AAD37t1D06ZN4e3tjaioKJibm+P48ePIzMzEqFGjEBcXh5SUFISFhQEArKyscP/+fZWa7927h7Zt28LPzw/r1q3DlStX0L9/f5QoUUIljISHhyMgIAB///03Tp48CT8/P3h5eaFly5b5+l0VFIMKEREVSadPn8bGjRvRvHlzZVt6ejrWr18PGxsbAEBUVBQuXLiApKQkyOVyAMDcuXOxfft2/P777/jhhx/e23+3bt2UYSfb+fPn4eLiAlNTU2zYsAHNmjWDmZkZ5s2bhwMHDsDCwgIAlPu3trbOcfjFxMQEq1atUgkvffr0Uf7s4uKCRYsWoV69enj58iVMTU2xdOlSWFhYYNOmTcrDWdmzMcDbwJaWlvbBQz3Lli2Do6MjlixZAplMhqpVq+L+/fsYO3YsJk+erLzvSc2aNREUFAQAqFy5MpYsWYIDBw4wqBAREX3Mjh07YGpqiszMTGRkZKBjx45YvHixcn2FChWUIQEAzpw5g5cvX8La2lqln9evX+PmzZsf3NdPP/2EFi1aqLQ5Ojoqf27YsCFGjRqFadOmYezYsWjatGmexlCjRo0c56WcO3cOwcHBiI2NxdOnT6FQKAAAd+/ehbu7O2JjY9GkSZMCnXMTFxeHhg0bqtzoz8vLS/ncp/LlywN4G1T+y97ePsdhqMLEoEJEREWGj48PQkNDYWhoCAcHhxx/uE1MTFReKxQK2Nvb49ChQzn6+tiJsXZ2dqhUqdJ71ysUChw/fhz6+vq4fv16nsfwbo2pqalo1aoVWrVqhQ0bNsDGxgZ3797FF198gfT0dADQyKMOhBA57kacfejpv+3v/k5lMpkyOOkCgwoRERUZJiYmHwwP76pTpw4SExNhYGAAJycnjdby448/Ii4uDocPH8YXX3yBsLAw9O7dGwCUMyZZWVkf7efKlSt4/PgxZs2apZyxiYmJUdmmZs2aCA8Pf++VTEZGRh/dl7u7O7Zs2aISWE6cOAEzMzOULVv24wPWEV71Q0REn6wWLVqgYcOG6NSpE/bu3Ys7d+7gxIkTmDhxYo4w8K7nz58jMTFRZcm+Z0tsbCwmT56M1atXw8vLCwsXLsTw4cNx69YtAICtrS2MjY2xZ88ePHz4EMnJye/dT/ny5WFkZITFixfj1q1b+PPPPzFt2jSVbYYMGYKUlBR07doVMTExuH79OtavX4+rV68CAJycnHD+/HlcvXoVjx8/VjnBOJu/vz8SEhIwdOhQXLlyBX/88QeCgoIQEBCgsefyaANnVIiICMCneWWiTCbDrl27MGHCBPTp0wePHj2CnZ0dmjZtijJlynzwvdmzI/8VEhKCESNGwNfXF35+fujQoQMAoG/fvti5cyd69OiBI0eOwMDAAIsWLcLUqVMxefJkNGnSJNfDT8DbE2/Xrl2L8ePHY9GiRahTpw7mzp2LL7/8UrmNtbU1oqKiMHr0aDRr1gz6+vqoXbs2vLy8AAD9+/fHoUOH4OnpiZcvX+LgwYM5ZpDKli2LXbt2YfTo0ahVqxasrKzQt29fTJw4UY3faOGTiewDVEVQSkoKLCwskJycDHNzc433z8uTiehT9ObNG9y+fRvOzs7K+4IQadqHvmfq/P2W7lwPERERFXsMKkRERCRZDCpEREQkWQwqREREJFkMKkRERCRZDCpEREQkWQwqREREJFkMKkRERCRZOg0qmZmZmDhxIpydnWFsbAwXFxdMnTpVpw8/IiIiIunQ6S30Z8+ejeXLlyM8PBzVqlVDTEwMevfuDQsLCwwfPlyXpRERFT/BFoW8v/c//4a069ChQ/Dx8cGzZ88++hRpXdPpjMrJkyfRsWNHtGvXDk5OTvj222/RqlWrjz4oioiIih8/Pz/IZDLIZDIYGhrCxcUFo0aNUj4osCDu3LkDmUyG2NjYPG2X23Lq1Kk87evQoUOQyWR4/vx5gevOC29vb4wYMUKlrVGjRnjw4AEsLAo5nOaDTmdUGjdujOXLl+PatWtwdXXFP//8g2PHjmHBggW5bp+Wloa0tDTl65SUlEKqlIiIpKB169YICwtDRkYGjh49in79+iE1NRWhoaGFWsf+/ftRrVo1lTZra2uN7iM9PR1GRkYa7TObkZER7OzstNK3pul0RmXs2LHo1q0bqlatCkNDQ3h4eGDEiBHo1q1brtuHhITAwsJCuTg6OhZyxUREpEtyuRx2dnZwdHRE9+7d4evri+3btwN4+z+zw4YNg62tLUqUKIHGjRsjOjpa+d5nz57B19cXNjY2MDY2RuXKlREWFgYAcHZ2BgB4eHhAJpPB29v7g3VYW1vDzs5OZTE0NIQQAi1atEDr1q2R/czf58+fo3z58pgwYQLu3LkDHx8fAECpUqUgk8ng5+cH4O3Mx5AhQxAQEIDSpUujZcuWAID58+ejRo0aMDExgaOjI/z9/fHy5UuVeo4fP45mzZqhZMmSKFWqFL744gs8e/YMfn5+OHz4MBYuXKic+blz506uszpbtmxBtWrVIJfL4eTkhHnz5qnsw8nJCTNnzkSfPn1gZmaG8uXL4+eff877h5dPOg0qmzdvxoYNG7Bx40acPXsW4eHhmDt3LsLDw3PdPjAwEMnJycolISGhkCsmIiIpMTY2RkZGBgBgzJgx2LJlC8LDw3H27FlUqlQJX3zxBZ4+fQoAmDRpEi5fvozdu3cjLi4OoaGhKF26NADg9OnTAN7OlDx48ABbt27NVz0ymQzh4eE4ffo0Fi1aBAAYOHAgypQpg+DgYDg6OmLLli0AgKtXr+LBgwdYuHCh8v3h4eEwMDDA8ePHsWLFCgCAnp4eFi1ahIsXLyI8PBxRUVEYM2aM8j2xsbFo3rw5qlWrhpMnT+LYsWPo0KEDsrKysHDhQjRs2BD9+/fHgwcP8ODBg1z/J//MmTPo0qULunbtigsXLiA4OBiTJk3C2rVrVbabN28ePD09ce7cOfj7+2PQoEG4cuVKvn5XeaXTQz+jR4/GuHHj0LVrVwBAjRo1EB8fj5CQEPTq1SvH9nK5HHK5vLDLJCIiCTp9+jQ2btyI5s2bKw//rF27Fm3atAEArFy5Evv27cPq1asxevRo3L17Fx4eHvD09ATwdoYgm42NDYD/nyn5mEaNGkFPT/X/9ZOTk6Gvr4+yZctixYoV6NGjBx4+fIi//voL586dg6GhIQDAysoKAGBra5vjRNZKlSphzpw5Km3/Pb/E2dkZ06ZNw6BBg7Bs2TIAwJw5c+Dp6al8DUDlsJSRkRFKliz5wXHNnz8fzZs3x6RJkwAArq6uuHz5Mn788UfljA8AtG3bFv7+/gDeHhX56aefcOjQIVStWvVDv64C0WlQefXqVY4PWl9fn5cnExFRrnbs2AFTU1NkZmYiIyMDHTt2xOLFi3Hz5k1kZGTAy8tLua2hoSHq1auHuLg4AMCgQYPwzTff4OzZs2jVqhU6deqERo0a5auOzZs3w83NTaVNX19f+XPnzp2xbds2hISEIDQ0FK6urnnqNztE/dfBgwcxc+ZMXL58GSkpKcjMzMSbN2+QmpoKExMTxMbGonPnzvkaR7a4uDh07NhRpc3LywsLFixAVlaWcmw1a9ZUrpfJZLCzs0NSUlKB9v0xOj3006FDB8yYMQM7d+7EnTt3sG3bNsyfPx9fffWVLssiIiKJ8vHxQWxsLK5evYo3b95g69atsLW1VZ4PIpPJVLYXQijb2rRpg/j4eIwYMQL3799H8+bNMWrUqHzV4ejoiEqVKqks//Xq1SucOXMG+vr6uH79ep77NTExUXkdHx+Ptm3bonr16tiyZQvOnDmDpUuXAoDykJexsXG+xvBf//09/bftXdmzQtlkMpnWJxd0GlQWL16Mb7/9Fv7+/nBzc8OoUaMwYMAATJs2TZdlERGRRJmYmKBSpUqoUKGCyh/NSpUqwcjICMeOHVO2ZWRkICYmRmXmw8bGBn5+ftiwYQMWLFigPBk0++qarKwsjdT5v//9D3p6eti9ezcWLVqEqKgo5Tp19hUTE4PMzEzMmzcPDRo0gKurK+7fv6+yTc2aNXHgwIH39mFkZPTRfbm7u6v87gDgxIkTcHV1VZkp0gWdHvoxMzPDggUL3ns5MhERUV6YmJhg0KBBGD16NKysrFC+fHnMmTMHr169Qt++fQEAkydPRt26dVGtWjWkpaVhx44dyhBja2sLY2Nj7NmzB+XKlUOJEiU+eI+RJ0+eIDExUaXN0tISJUqUwM6dO7FmzRqcPHkSderUwbhx49CrVy+cP38epUqVQoUKFSCTybBjxw60bdsWxsbGMDU1zXU/FStWRGZmJhYvXowOHTrg+PHjWL58uco2gYGBqFGjBvz9/TFw4EAYGRnh4MGD6Ny5M0qXLg0nJyf8/fffuHPnDkxNTZXnyPzX//73P3z22WeYNm0avvvuO5w8eRJLlixROe9FV3QaVIiISEKK+J1iZ82aBYVCgR49euDFixfw9PTE3r17UapUKQBvZxYCAwNx584dGBsbo0mTJti0aRMAwMDAAIsWLcLUqVMxefJkNGnSBIcOHXrvvlq0aJGjLSIiAs2bN0ffvn0RHByMOnXqAACCgoIQGRmJgQMHYvPmzShbtiymTJmCcePGoXfv3ujZs2eOq2uy1a5dG/Pnz8fs2bMRGBiIpk2bIiQkBD179lRu4+rqisjISIwfPx716tWDsbEx6tevr7zVx6hRo9CrVy+4u7vj9evXuH37do791KlTB7/++ismT56MadOmwd7eHlOnTlU5kVZXZCK3g1BFREpKCiwsLJCcnAxzc3ON9+80bqdG+7szq51G+yMiyo83b97g9u3bcHZ2RokSJXRdDn2iPvQ9U+fvN5+eTERERJLFoEJERESSxaBCREREksWgQkRERJLFoEJEVEwV4WspqAjQ1PeLQYWIqJjJvoFXenq6jiuhT9mrV68A5Lybrbp4HxUiomLGwMAAJUuWxKNHj2BoaJjjmWtEBSGEwKtXr5CUlARLS8sC39mWQYWIqJiRyWSwt7fH7du3ER8fr+ty6BNlaWmZpydRfwyDChFRMWRkZITKlSvz8A9phaGhocaeEcSgQkRUTOnp6fHOtCR5PDBJREREksUZFSIq0vhMLqJPG2dUiIiISLIYVIiIiEiyGFSIiIhIshhUiIiISLIYVIiIiEiyGFSIiIhIsnh5MhUrvJSViKho4YwKERERSRaDChEREUkWgwoRERFJFoMKERERSRaDChEREUkWgwoRERFJFoMKERERSRaDChEREUkWgwoRERFJlk6DipOTE2QyWY5l8ODBuiyLiIiIJEKnt9CPjo5GVlaW8vXFixfRsmVLdO7cWYdVERERkVToNKjY2NiovJ41axYqVqyIZs2a6agiIiIikhLJPJQwPT0dGzZsQEBAAGQyWa7bpKWlIS0tTfk6JSWlsMojIiIiHZDMybTbt2/H8+fP4efn995tQkJCYGFhoVwcHR0Lr0AiIiIqdJIJKqtXr0abNm3g4ODw3m0CAwORnJysXBISEgqxQiIiIipskjj0Ex8fj/3792Pr1q0f3E4ul0MulxdSVURERKRrkphRCQsLg62tLdq1a6frUoiIiEhCdB5UFAoFwsLC0KtXLxgYSGKCh4iIiCRC50Fl//79uHv3Lvr06aPrUoiIiEhidD6F0apVKwghdF0GERERSZDOZ1SIiIiI3odBhYiIiCSLQYWIiIgki0GFiIiIJItBhYiIiCSLQYWIiIgki0GFiIiIJItBhYiIiCSLQYWIiIgki0GFiIiIJItBhYiIiCSLQYWIiIgki0GFiIiIJItBhYiIiCSLQYWIiIgki0GFiIiIJItBhYiIiCSLQYWIiIgki0GFiIiIJItBhYiIiCSLQYWIiIgki0GFiIiIJItBhYiIiCSLQYWIiIgki0GFiIiIJItBhYiIiCSLQYWIiIgki0GFiIiIJItBhYiIiCSLQYWIiIgki0GFiIiIJMtA1wXcu3cPY8eOxe7du/H69Wu4urpi9erVqFu3rq5LIyIi0iqncTs12t+dWe002p8U6DSoPHv2DF5eXvDx8cHu3btha2uLmzdvwtLSUpdlERERkUToNKjMnj0bjo6OCAsLU7Y5OTnpriAiIiKSFJ2eo/Lnn3/C09MTnTt3hq2tLTw8PLBy5cr3bp+WloaUlBSVhYiIiD5dOg0qt27dQmhoKCpXroy9e/di4MCBGDZsGNatW5fr9iEhIbCwsFAujo6OhVwxERERFSadBhWFQoE6depg5syZ8PDwwIABA9C/f3+Ehobmun1gYCCSk5OVS0JCQiFXTERERIVJp0HF3t4e7u7uKm1ubm64e/durtvL5XKYm5urLERERPTp0mlQ8fLywtWrV1Xarl27hgoVKuioIiIiIpISnQaVkSNH4tSpU5g5cyZu3LiBjRs34ueff8bgwYN1WRYRERFJhE6DymeffYZt27YhIiIC1atXx7Rp07BgwQL4+vrqsiwiIiKSCLXvo5KQkACZTIZy5coBAE6fPo2NGzfC3d0dP/zwg9oFtG/fHu3bt1f7fURERPTpU3tGpXv37jh48CAAIDExES1btsTp06cxfvx4TJ06VeMFEhERUfGldlC5ePEi6tWrBwD49ddfUb16dZw4cQIbN27E2rVrNV0fERERFWNqB5WMjAzI5XIAwP79+/Hll18CAKpWrYoHDx5otjoiIiIq1tQOKtWqVcPy5ctx9OhR7Nu3D61btwYA3L9/H9bW1hovkIiIiIovtYPK7NmzsWLFCnh7e6Nbt26oVasWgLfP7ck+JERERESkCWpf9ePt7Y3Hjx8jJSUFpUqVUrb/8MMPKFmypEaLIyIiouJN7aACAPr6+iohBQCcnJw0UQ8REeXCadxOjfd5Z1Y7jfdJpGl5CioeHh6QyWR56vDs2bMFKoiIiIgoW56CSqdOnZQ/v3nzBsuWLYO7uzsaNmwIADh16hQuXboEf39/rRRJRERExVOegkpQUJDy5379+mHYsGGYNm1ajm0SEhI0Wx0REREVa2pf9fPbb7+hZ8+eOdq///57bNmyRSNFEREREQH5CCrGxsY4duxYjvZjx46hRIkSGimKiIiICMjHVT8jRozAoEGDcObMGTRo0ADA23NU1qxZg8mTJ2u8QCIiIiq+1A4q48aNg4uLCxYuXIiNGzcCANzc3LB27Vp06dJF4wUSERFR8ZWv+6h06dKFoYSIiIi0Tu1zVIiIiIgKi9ozKllZWfjpp5/w66+/4u7du0hPT1dZ//TpU40VR0RERMWb2jMqU6ZMwfz589GlSxckJycjICAAX3/9NfT09BAcHKyFEomIiKi4Ujuo/PLLL1i5ciVGjRoFAwMDdOvWDatWrcLkyZNx6tQpbdRIRERExZTaQSUxMRE1atQAAJiamiI5ORkA0L59e+zcqfmHZhEREVHxpXZQKVeuHB48eAAAqFSpEiIjIwEA0dHRkMvlmq2OiIiIijW1g8pXX32FAwcOAACGDx+OSZMmoXLlyujZsyf69Omj8QKJiIio+FL7qp9Zs2Ypf/72229Rrlw5nDhxApUqVcKXX36p0eKIiIioeMvXDd/+q0GDBspb6RMRERFpUr5u+LZ+/Xp4eXnBwcEB8fHxAIAFCxbgjz/+0GhxREREVLypHVRCQ0MREBCAtm3b4vnz58jKygIAWFpaYsGCBZquj4iIiIoxtYPK4sWLsXLlSkyYMAH6+vrKdk9PT1y4cEGjxREREVHxpnZQuX37Njw8PHK0y+VypKamaqQoIiIiIiAfQcXZ2RmxsbE52nfv3g13d3dN1EREREQEIB9X/YwePRqDBw/GmzdvIITA6dOnERERgZCQEKxatUobNRIREVExpXZQ6d27NzIzMzFmzBi8evUK3bt3R9myZbFw4UJ07dpVrb6Cg4MxZcoUlbYyZcogMTFR3bKIiIjoE5Sv+6j0798f/fv3x+PHj6FQKGBrawsAuHfvHsqWLatWX9WqVcP+/fuVr/97gi4REREVbwW64Vvp0qUBvH1Q4YwZM7Bq1Sq8fv1avQIMDGBnZ1eQMoiIiOgTleeTaZ8/fw5fX1/Y2NjAwcEBixYtgkKhwOTJk+Hi4oJTp05hzZo1ahdw/fp1ODg4wNnZGV27dsWtW7feu21aWhpSUlJUFiIiIvp05TmojB8/HkeOHEGvXr1gZWWFkSNHon379jh27Bh2796N6OhodOvWTa2d169fH+vWrcPevXuxcuVKJCYmolGjRnjy5Emu24eEhMDCwkK5ODo6qrU/IiIiKlryHFR27tyJsLAwzJ07F3/++SeEEHB1dUVUVBSaNWuWr523adMG33zzDWrUqIEWLVpg586dAIDw8PBctw8MDERycrJySUhIyNd+iYiIqGjI8zkq9+/fV94nxcXFBSVKlEC/fv00WoyJiQlq1KiB69ev57peLpdDLpdrdJ9EREQkXXmeUVEoFDA0NFS+1tfXh4mJiUaLSUtLQ1xcHOzt7TXaLxERERVNeZ5REULAz89POaPx5s0bDBw4MEdY2bp1a553PmrUKHTo0AHly5dHUlISpk+fjpSUFPTq1SvPfRAREdGnK89B5d3w8P333xd45//++y+6deuGx48fw8bGBg0aNMCpU6dQoUKFAvdNRERERV+eg0pYWJjGd75p0yaN90lEgNO4nRrv886sdhrvk4joY9R+KCERERFRYWFQISIiIsliUCEiIiLJYlAhIiIiycpTUKlTpw6ePXsGAJg6dSpevXql1aKIiIiIgDwGlbi4OKSmpgIApkyZgpcvX2q1KCIiIiIgj5cn165dG71790bjxo0hhMDcuXNhamqa67aTJ0/WaIFERERUfOUpqKxduxZBQUHYsWMHZDIZdu/eDQODnG+VyWQMKkRERKQxeQoqVapUUd6cTU9PDwcOHICtra1WCyMiIiLK851psykUCm3UQUREBIB3ViZVagcVALh58yYWLFiAuLg4yGQyuLm5Yfjw4ahYsaKm6yMiIqJiTO37qOzduxfu7u44ffo0atasierVq+Pvv/9GtWrVsG/fPm3USERERMWU2jMq48aNw8iRIzFr1qwc7WPHjkXLli01VhwREREVb2rPqMTFxaFv37452vv06YPLly9rpCgiIiIiIB9BxcbGBrGxsTnaY2NjeSUQERERaZTah3769++PH374Abdu3UKjRo0gk8lw7NgxzJ49G//73/+0USMREREVU2oHlUmTJsHMzAzz5s1DYGAgAMDBwQHBwcEYNmyYxgskIiKi4kvtoCKTyTBy5EiMHDkSL168AACYmZlpvDAiIiKifN1HJRsDChEREWmT2ifTEhERERUWBhUiIiKSLAYVIiIikiy1gkpGRgZ8fHxw7do1bdVDREREpKRWUDE0NMTFixchk8m0VQ8RERGRktqHfnr27InVq1droxYiIiIiFWpfnpyeno5Vq1Zh37598PT0hImJicr6+fPna6w4IiIiKt7UDioXL15EnTp1ACDHuSo8JERERESapHZQOXjwoDbqICIiIsoh35cn37hxA3v37sXr168BAEIIjRVFREREBOQjqDx58gTNmzeHq6sr2rZtiwcPHgAA+vXrx6cnExERkUapHVRGjhwJQ0ND3L17FyVLllS2f/fdd9izZ49GiyMiIqLiTe2gEhkZidmzZ6NcuXIq7ZUrV0Z8fHy+CwkJCYFMJsOIESPy3QcRERF9WtQOKqmpqSozKdkeP34MuVyeryKio6Px888/o2bNmvl6PxEREX2a1A4qTZs2xbp165SvZTIZFAoFfvzxR/j4+KhdwMuXL+Hr64uVK1eiVKlSar+fiIiIPl1qX578448/wtvbGzExMUhPT8eYMWNw6dIlPH36FMePH1e7gMGDB6Ndu3Zo0aIFpk+f/sFt09LSkJaWpnydkpKi9v6IiIio6FB7RsXd3R3nz59HvXr10LJlS6SmpuLrr7/GuXPnULFiRbX62rRpE86ePYuQkJA8bR8SEgILCwvl4ujoqG75REREVISoPaMCAHZ2dpgyZUqBdpyQkIDhw4cjMjISJUqUyNN7AgMDERAQoHydkpLCsEJERPQJy1dQefbsGVavXo24uDjIZDK4ubmhd+/esLKyynMfZ86cQVJSEurWratsy8rKwpEjR7BkyRKkpaVBX19f5T1yuTzfJ+wSERFR0aP2oZ/Dhw/D2dkZixYtwrNnz/D06VMsWrQIzs7OOHz4cJ77ad68OS5cuIDY2Fjl4unpCV9fX8TGxuYIKURERFT8qD2jMnjwYHTp0gWhoaHKMJGVlQV/f38MHjwYFy9ezFM/ZmZmqF69ukqbiYkJrK2tc7QTERFR8aT2jMrNmzfxv//9T2XGQ19fHwEBAbh586ZGiyMiIqLiTe0ZlTp16iAuLg5VqlRRaY+Li0Pt2rULVMyhQ4cK9H4iIiL6tOQpqJw/f17587BhwzB8+HDcuHEDDRo0AACcOnUKS5cuxaxZs7RTJRERERVLeQoqtWvXhkwmgxBC2TZmzJgc23Xv3h3fffed5qojIiKiYi1PQeX27dvaroOIiIgohzwFlQoVKmi7DiIiIqIc8nXDt3v37uH48eNISkqCQqFQWTds2DCNFEZERESkdlAJCwvDwIEDYWRkBGtra8hkMuU6mUzGoEJEREQao3ZQmTx5MiZPnozAwEDo6al9GxYiIiKiPFM7abx69Qpdu3ZlSCEiIiKtUztt9O3bF7/99ps2aiEiIiJSofahn5CQELRv3x579uxBjRo1YGhoqLJ+/vz5GivukxNsoYU+kzXfJxERkUSoHVRmzpyJvXv3Km+h/+7JtERERESaonZQmT9/PtasWQM/Pz8tlENERET0/9Q+R0Uul8PLy0sbtRARERGpUDuoDB8+HIsXL9ZGLUREREQq1D70c/r0aURFRWHHjh2oVq1ajpNpt27dqrHiiIiIqHhTO6hYWlri66+/1kYtRERERCrydQt9IiIiosLA28sSERGRZKk9o+Ls7PzB+6XcunWrQAURERERZVM7qIwYMULldUZGBs6dO4c9e/Zg9OjRmqqLiIiISP2gMnz48Fzbly5dipiYmAIXRERERJRNY+eotGnTBlu2bNFUd0RERESaCyq///47rKysNNUdERERkfqHfjw8PFROphVCIDExEY8ePcKyZcs0WhwREREVb2oHlU6dOqm81tPTg42NDby9vVG1alVN1UVERESkflAJCgrSRh1EREREOfCGb0RERCRZeZ5R0dPT++CN3gBAJpMhMzOzwEURERERAWoElW3btr133YkTJ7B48WIIITRSFBERERGgRlDp2LFjjrYrV64gMDAQf/31F3x9fTFt2jSNFkdERETFW77OUbl//z769++PmjVrIjMzE7GxsQgPD0f58uU1XR8REREVY2oFleTkZIwdOxaVKlXCpUuXcODAAfz111+oXr16vnYeGhqKmjVrwtzcHObm5mjYsCF2796dr76IiIjo05PnoDJnzhy4uLhgx44diIiIwIkTJ9CkSZMC7bxcuXKYNWsWYmJiEBMTg88//xwdO3bEpUuXCtQvERERfRryfI7KuHHjYGxsjEqVKiE8PBzh4eG5brd169Y877xDhw4qr2fMmIHQ0FCcOnUK1apVy3M/RERE9GnKc1Dp2bPnRy9PLoisrCz89ttvSE1NRcOGDXPdJi0tDWlpacrXKSkpWquHiIiIdC/PQWXt2rVaKeDChQto2LAh3rx5A1NTU2zbtg3u7u65bhsSEoIpU6ZopQ4iIiKSHp3fmbZKlSqIjY3FqVOnMGjQIPTq1QuXL1/OddvAwEAkJycrl4SEhEKuloiIiAqT2s/60TQjIyNUqlQJAODp6Yno6GgsXLgQK1asyLGtXC6HXC4v7BKJiIhIR3Q+o/IuIYTKeShERERUfOl0RmX8+PFo06YNHB0d8eLFC2zatAmHDh3Cnj17dFkWERERSYROg8rDhw/Ro0cPPHjwABYWFqhZsyb27NmDli1b6rIsIiIikgidBpXVq1frcvdEREQkcZI7R4WIiIgoG4MKERERSRaDChEREUkWgwoRERFJFoMKERERSZbO70xLuuc0bqfG+7wzq53G+yQiouKHMypEREQkWQwqREREJFkMKkRERCRZDCpEREQkWQwqREREJFkMKkRERCRZDCpEREQkWQwqREREJFm84RtRQQRbaKHPZM33SURURHFGhYiIiCSLMypERP/FWTIiSeGMChEREUkWgwoRERFJFoMKERERSRaDChEREUkWgwoRERFJFoMKERERSRaDChEREUkWgwoRERFJFoMKERERSRaDChEREUkWgwoRERFJFoMKERERSRaDChEREUmWToNKSEgIPvvsM5iZmcHW1hadOnXC1atXdVkSERERSYhOg8rhw4cxePBgnDp1Cvv27UNmZiZatWqF1NRUXZZFREREEmGgy53v2bNH5XVYWBhsbW1x5swZNG3aVEdVERERkVToNKi8Kzk5GQBgZWWV6/q0tDSkpaUpX6ekpBRKXUREn6RgCw33l6zZ/oggoZNphRAICAhA48aNUb169Vy3CQkJgYWFhXJxdHQs5CqJiIioMEkmqAwZMgTnz59HRETEe7cJDAxEcnKycklISCjEComIiKiwSeLQz9ChQ/Hnn3/iyJEjKFeu3Hu3k8vlkMvlhVgZERER6ZJOg4oQAkOHDsW2bdtw6NAhODs767IcIiIikhidBpXBgwdj48aN+OOPP2BmZobExEQAgIWFBYyNjXVZGhEREUmAToNKaGgoAMDb21ulPSwsDH5+foVfEBERUVGm6Su5AJ1fzaXzQz9ERERE7yOZq36IiIiI3sWgQkRERJLFoEJERESSxaBCREREkiWJG74RURHA58IQkQ5wRoWIiIgki0GFiIiIJItBhYiIiCSLQYWIiIgki0GFiIiIJItBhYiIiCSLQYWIiIgki0GFiIiIJItBhYiIiCSLQYWIiIgki0GFiIiIJItBhYiIiCSLDyUkIqJPHx+qWWRxRoWIiIgki0GFiIiIJItBhYiIiCSLQYWIiIgki0GFiIiIJItBhYiIiCSLQYWIiIgki0GFiIiIJItBhYiIiCSLQYWIiIgki0GFiIiIJItBhYiIiCSLQYWIiIgkS6dPTz5y5Ah+/PFHnDlzBg8ePMC2bdvQqVMnXZZEmsInlRIRkQbodEYlNTUVtWrVwpIlS3RZBhEREUmUTmdU2rRpgzZt2uiyBCIiIpIwnQYVdaWlpSEtLU35OiUlRYfVEBERkbYVqZNpQ0JCYGFhoVwcHR11XRIRERFpUZEKKoGBgUhOTlYuCQkJui6JiIiItKhIHfqRy+WQy+W6LoOIiIgKSZGaUSEiIqLiRaczKi9fvsSNGzeUr2/fvo3Y2FhYWVmhfPnyOqyMiIiIpECnQSUmJgY+Pj7K1wEBAQCAXr16Ye3atTqqioiIiKRCp0HF29sbQghdlkBEREQSxnNUiIiISLIYVIiIiEiyGFSIiIhIshhUiIiISLIYVIiIiEiyGFSIiIhIshhUiIiISLIYVIiIiEiyGFSIiIhIshhUiIiISLIYVIiIiEiyGFSIiIhIshhUiIiISLIYVIiIiEiyGFSIiIhIshhUiIiISLIYVIiIiEiyGFSIiIhIshhUiIiISLIYVIiIiEiyGFSIiIhIshhUiIiISLIYVIiIiEiyGFSIiIhIshhUiIiISLIYVIiIiEiyGFSIiIhIshhUiIiISLIYVIiIiEiyGFSIiIhIsnQeVJYtWwZnZ2eUKFECdevWxdGjR3VdEhEREUmEToPK5s2bMWLECEyYMAHnzp1DkyZN0KZNG9y9e1eXZREREZFE6DSozJ8/H3379kW/fv3g5uaGBQsWwNHREaGhobosi4iIiCTCQFc7Tk9Px5kzZzBu3DiV9latWuHEiRO5victLQ1paWnK18nJyQCAlJQUrdSoSHul0f5SZEKj/b3ttOBj1/Q4AS2MVUOfMT/T/ONnWpBO+Zmqg2PNP6l+f3N2+bZPIfJQr9CRe/fuCQDi+PHjKu0zZswQrq6uub4nKChIAODChQsXLly4fAJLQkLCR/OCzmZUsslkMpXXQogcbdkCAwMREBCgfK1QKPD06VNYW1u/9z1SkZKSAkdHRyQkJMDc3FzX5WhVcRlrcRknUHzGWlzGCXCsn6KiNE4hBF68eAEHB4ePbquzoFK6dGno6+sjMTFRpT0pKQllypTJ9T1yuRxyuVylzdLSUlslaoW5ubnkv0CaUlzGWlzGCRSfsRaXcQIc66eoqIzTwsIiT9vp7GRaIyMj1K1bF/v27VNp37dvHxo1aqSjqoiIiEhKdHroJyAgAD169ICnpycaNmyIn3/+GXfv3sXAgQN1WRYRERFJhE6DynfffYcnT55g6tSpePDgAapXr45du3ahQoUKuixLK+RyOYKCgnIcuvoUFZexFpdxAsVnrMVlnADH+in6VMcpEyIv1wYRERERFT6d30KfiIiI6H0YVIiIiEiyGFSIiIhIshhUiIiISLIYVIiIiEiyGFS0JC0tDbVr14ZMJkNsbGye3zd8+HDUrVsXcrkctWvX1lp9mpSfsf7zzz/o1q0bHB0dYWxsDDc3NyxcuFC7hWpAfsb65MkTtG7dGg4ODpDL5XB0dMSQIUO09jBNTcjv9zfbkydPUK5cOchkMjx//lzj9enSgwcP0L17d1SpUgV6enoYMWKErkvSmq1bt6Jly5awsbGBubk5GjZsiL179+q6LI07duwYvLy8YG1tDWNjY1StWhU//fSTrsvSuuPHj8PAwEDyf2sYVLRkzJgxeXqGwbuEEOjTpw++++47LVSlHfkZ65kzZ2BjY4MNGzbg0qVLmDBhAgIDA7FkyRItVakZ+Rmrnp4eOnbsiD///BPXrl3D2rVrsX//fknf2DC/399sffv2Rc2aNTVYkXSkpaXBxsYGEyZMQK1atXRdjlYdOXIELVu2xK5du3DmzBn4+PigQ4cOOHfunK5L0ygTExMMGTIER44cQVxcHCZOnIiJEyfi559/1nVpWpOcnIyePXuiefPmui7l4wr6FOTiZvfu3cLLy0tYWFgIKysr0a5dO3Hjxg2VbXbt2iWqVq0qLl26JACIc+fOqb2foKAgUatWLc0UnU+FNdZs/v7+wsfHp4BV509hj3XhwoWiXLlyBaxafYUxzmXLlolmzZqJAwcOCADi2bNnmhtAHjVr1kwMHTpUjB49WpQqVUqUKVNGBAUFKdfHx8eLL7/8UpiYmAgzMzPRuXNnkZiYmK/9DB8+XHOF50NhjTWbu7u7mDJligYqV09hj/Orr74S33//vQYqV19hjPW7774TEydOlMTfmo/hjIqaUlNTERAQgOjoaBw4cAB6enr46quvoFAoAAAPHz5E//79sX79epQsWVLH1RZMYY81OTkZVlZWBe4nPwpzrPfv38fWrVvRrFkzTZSuFm2P8/Lly5g6dSrWrVsHPT3d/uclPDwcJiYm+PvvvzFnzhxMnToV+/btgxACnTp1wtOnT3H48GHs27cPN2/eLFKzmO8qrLEqFAq8ePFCZ/9OC2uc586dw4kTJ3TybzSbNscaFhaGmzdvIigoSIsj0CDd5qSiLykpSQAQFy5cEAqFQrRu3VpMmzZNCCHE7du3i/SMyru0NVYhhDhx4oQwNDQUkZGRGqw4/7Qx1q5duwpjY2MBQHTo0EG8fv1aC5WrR5PjfPPmjahZs6ZYv369EEKIgwcP6nRGpXHjxiptn332mRg7dqyIjIwU+vr64u7du8p12bNHp0+fVns/UphRKYyxCiHEnDlzhJWVlXj48GGB61ZXYYyzbNmywsjISOjp6YmpU6dqrHZ1aXOs165dE7a2tuLq1atCCGn+rXkXZ1TUdPPmTXTv3h0uLi4wNzeHs7MzAODu3btYvHgxUlJSEBgYqOMqNaOwxnrp0iV07NgRkydPRsuWLQvcX34Uxlh/+uknnD17Ftu3b8fNmzcREBCgidLVos1xBgYGws3NDd9//70mS863d8+Rsbe3R1JSEuLi4uDo6AhHR0flOnd3d1haWiIuLq6wy9SIwhhrREQEgoODsXnzZtja2mqkbnVpe5xHjx5FTEwMli9fjgULFiAiIkJjtatLG2PNyspC9+7dMWXKFLi6umqlbm3Q6UMJi6IOHTrA0dERK1euhIODAxQKBapXr4709HRERUXh1KlTOR4I5enpCV9fX4SHh+uo6vwpjLFevnwZn3/+Ofr374+JEydqYxh5UhhjtbOzg52dHapWrQpra2s0adIEkyZNgr29vTaGlCttjjMqKgoXLlzA77//DuDtieEAULp0aUyYMAFTpkzRzqDew9DQUOW1TCaDQqGAEAIymSzH9u9rLwq0PdbNmzejb9+++O2339CiRYsC15tf2h5ndnCvUaMGHj58iODgYHTr1q1gReeTNsb64sULxMTE4Ny5cxgyZAgAKPs0MDBAZGQkPv/8c80NQkMYVNTw5MkTxMXFYcWKFWjSpAmAt5e1ZVu0aBGmT5+ufH3//n188cUX2Lx5M+rXr1/o9RZEYYz10qVL+Pzzz9GrVy/MmDFDswNQgy4+1+w/4mlpaQWoXD3aHueWLVvw+vVr5evo6Gj06dMHR48eRcWKFTU4koJxd3fH3bt3kZCQoPy/0suXLyM5ORlubm46rk6zNDHWiIgI9OnTBxEREWjXrp02y803bXymQohC/feZVwUZq7m5OS5cuKDStmzZMkRFReH3339XBjWpYVBRQ6lSpWBtbY2ff/4Z9vb2uHv3LsaNG6dcX758eZXtTU1NAQAVK1ZEuXLl8rSPGzdu4OXLl0hMTMTr16+V97Bwd3eHkZGRZgaSB9oe66VLl+Dj44NWrVohICAAiYmJAAB9fX3Y2NhocCQfp+2x7tq1Cw8fPsRnn30GU1NTXL58GWPGjIGXlxecnJw0OpYP0fY43w0jjx8/BgC4ubnB0tKygNVrTosWLVCzZk34+vpiwYIFyMzMhL+/P5o1awZPT8889ZH97/Lly5d49OgRYmNjYWRkBHd3dy1Wrr6CjjUiIgI9e/bEwoUL0aBBA+W/U2NjY1hYWGi7/Dwr6DiXLl2K8uXLo2rVqgDeBvi5c+di6NCh2i5dbQUZq56eHqpXr67SZmtrixIlSuRolxKeo6IGPT09bNq0CWfOnEH16tUxcuRI/PjjjxrdR79+/eDh4YEVK1bg2rVr8PDwgIeHB+7fv6/R/XyMtsf622+/4dGjR/jll19gb2+vXD777DON7SOvtD1WY2NjrFy5Eo0bN4abmxtGjBiB9u3bY8eOHRrbR14Uxve3KJDJZNi+fTtKlSqFpk2bokWLFnBxccHmzZvz3Ef2v8szZ85g48aN8PDwQNu2bbVYdf4UdKwrVqxAZmYmBg8erPLvdPjw4VquXD0FHadCoUBgYCBq164NT09PLF68GLNmzcLUqVO1XLn6NPH9LWpkInsOmoiIiEhiOKNCREREksWgUogGDhwIU1PTXBcp3049PzjWT2+sxWWc1apVe+84f/nlF12Xp1HFZazFZZzApzlWHvopRElJSe99EJ25ubnO7k2gDRzrW5/SWIvLOOPj45GRkZHrujJlysDMzKyQK9Ke4jLW4jJO4NMcK4MKERERSRYP/RAREZFkMagQERGRZDGoEBERkWQxqBAREZFkMagQERGRZDGoEBERkWQxqBAREZFk/R9GKcjrUs3OSAAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "make_table(ins, outs, \"Pre Extraction\", \"Post Extraction\", \"Number of Reads\", samples, \"Removal of rRNA Reads\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25a152bb",
   "metadata": {},
   "source": [
    "## Step: align"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffaf8d74-6d7e-4961-9ac0-1ffa13ea2fc5",
   "metadata": {},
   "source": [
    "#### Align reads to genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "52d0b8b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_path = mkpath(\"extract_rRNA\")\n",
    "out_path= mkpath(\"align\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2d03a5e",
   "metadata": {},
   "source": [
    "#### STAR"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1295ee30-6dbe-4ae0-9a40-e790c2c70574",
   "metadata": {},
   "source": [
    "https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "630aa25d-2a33-4283-ad22-9b663b6d9c30",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: STAR  [options]... --genomeDir /path/to/genome/index/   --readFilesIn R1.fq R2.fq\n",
      "Spliced Transcripts Alignment to a Reference (c) Alexander Dobin, 2009-2022\n",
      "\n",
      "STAR version=2.7.11a\n",
      "STAR compilation time,server,dir=2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "For more details see:\n",
      "<https://github.com/alexdobin/STAR>\n",
      "<https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf>\n",
      "### versions\n",
      "versionGenome           2.7.4a\n",
      "    string: earliest genome index version compatible with this STAR release. Please do not change this value!\n",
      "\n",
      "### Parameter Files\n",
      "parametersFiles          -\n",
      "    string: name of a user-defined parameters file, \"-\": none. Can only be defined on the command line.\n",
      "\n",
      "### System\n",
      "sysShell            -\n",
      "    string: path to the shell binary, preferably bash, e.g. /bin/bash.\n",
      "                    - ... the default shell is executed, typically /bin/sh. This was reported to fail on some Ubuntu systems - then you need to specify path to bash.\n",
      "\n",
      "### Run Parameters\n",
      "runMode                         alignReads\n",
      "    string: type of the run.\n",
      "                                alignReads             ... map reads\n",
      "                                genomeGenerate         ... generate genome files\n",
      "                                inputAlignmentsFromBAM ... input alignments from BAM. Presently only works with --outWigType and --bamRemoveDuplicates options.\n",
      "                                liftOver               ... lift-over of GTF files (--sjdbGTFfile) between genome assemblies using chain file(s) from --genomeChainFiles.\n",
      "                                soloCellFiltering  </path/to/raw/count/dir/>   </path/to/output/prefix>    ... STARsolo cell filtering (\"calling\") without remapping, followed by the path to raw count directory and output (filtered) prefix\n",
      "\n",
      "runThreadN                      1\n",
      "    int: number of threads to run STAR\n",
      "\n",
      "runDirPerm                      User_RWX\n",
      "    string: permissions for the directories created at the run-time.\n",
      "                                User_RWX ... user-read/write/execute\n",
      "                                All_RWX  ... all-read/write/execute (same as chmod 777)\n",
      "\n",
      "runRNGseed                      777\n",
      "    int: random number generator seed.\n",
      "\n",
      "\n",
      "### Genome Parameters\n",
      "genomeDir                   ./GenomeDir/\n",
      "    string: path to the directory where genome files are stored (for --runMode alignReads) or will be generated (for --runMode generateGenome)\n",
      "\n",
      "genomeLoad                NoSharedMemory\n",
      "    string: mode of shared memory usage for the genome files. Only used with --runMode alignReads.\n",
      "                          LoadAndKeep     ... load genome into shared and keep it in memory after run\n",
      "                          LoadAndRemove   ... load genome into shared but remove it after run\n",
      "                          LoadAndExit     ... load genome into shared memory and exit, keeping the genome in memory for future runs\n",
      "                          Remove          ... do not map anything, just remove loaded genome from memory\n",
      "                          NoSharedMemory  ... do not use shared memory, each job will have its own private copy of the genome\n",
      "\n",
      "genomeFastaFiles            -\n",
      "    string(s): path(s) to the fasta files with the genome sequences, separated by spaces. These files should be plain text FASTA files, they *cannot* be zipped.\n",
      "                            Required for the genome generation (--runMode genomeGenerate). Can also be used in the mapping (--runMode alignReads) to add extra (new) sequences to the genome (e.g. spike-ins).\n",
      "\n",
      "genomeChainFiles            -\n",
      "    string: chain files for genomic liftover. Only used with --runMode liftOver .\n",
      "\n",
      "genomeFileSizes             0\n",
      "    uint(s)>0: genome files exact sizes in bytes. Typically, this should not be defined by the user.\n",
      "    \n",
      "genomeTransformOutput       None\n",
      "    string(s)               which output to transform back to original genome\n",
      "                            SAM     ... SAM/BAM alignments\n",
      "                            SJ      ... splice junctions (SJ.out.tab)\n",
      "                            Quant   ... quantifications (from --quantMode option)\n",
      "                            None    ... no transformation of the output        \n",
      "\n",
      "genomeChrSetMitochondrial   chrM M MT\n",
      "    string(s)               names of the mitochondrial chromosomes. Presently only used for STARsolo statisics output/\n",
      "\n",
      "### Genome Indexing Parameters - only used with --runMode genomeGenerate\n",
      "genomeChrBinNbits           18\n",
      "    int: =log2(chrBin), where chrBin is the size of the bins for genome storage: each chromosome will occupy an integer number of bins. For a genome with large number of contigs, it is recommended to scale this parameter as min(18, log2[max(GenomeLength/NumberOfReferences,ReadLength)]).\n",
      "\n",
      "genomeSAindexNbases         14\n",
      "    int: length (bases) of the SA pre-indexing string. Typically between 10 and 15. Longer strings will use much more memory, but allow faster searches. For small genomes, the parameter --genomeSAindexNbases must be scaled down to min(14, log2(GenomeLength)/2 - 1).\n",
      "\n",
      "genomeSAsparseD             1\n",
      "    int>0: suffux array sparsity, i.e. distance between indices: use bigger numbers to decrease needed RAM at the cost of mapping speed reduction\n",
      "\n",
      "genomeSuffixLengthMax       -1\n",
      "    int: maximum length of the suffixes, has to be longer than read length. -1 = infinite.\n",
      "    \n",
      "genomeTransformType         None\n",
      "    string: type of genome transformation\n",
      "                            None       ... no transformation\n",
      "                            Haploid    ... replace reference alleles with alternative alleles from VCF file (e.g. consensus allele)\n",
      "                            Diploid    ... create two haplotypes for each chromosome listed in VCF file, for genotypes 1|2, assumes perfect phasing (e.g. personal genome)\n",
      "\n",
      "genomeTransformVCF          -\n",
      "    string: path to VCF file for genome transformation\n",
      "\n",
      "\n",
      "    \n",
      "#####UnderDevelopment_begin : not supported - do not use\n",
      "genomeType                  Full\n",
      "    string: type of genome to generate\n",
      "                            Full                ... full (normal) genome\n",
      "                            Transcriptome       ... genome consists of transcript sequences\n",
      "                            SuperTransriptome   ... genome consists of superTranscript sequences\n",
      "#####UnderDevelopment_end\n",
      "\n",
      "# DEPRECATED: please use --genomeTransformVCF and --genomeTransformType options instead.\n",
      "#genomeConsensusFile         -\n",
      "#    string: VCF file with consensus SNPs (i.e. alternative allele is the major (AF>0.5) allele)\n",
      "# DEPRECATED \n",
      "\n",
      "\n",
      "\n",
      "### Splice Junctions Database\n",
      "sjdbFileChrStartEnd                     -\n",
      "    string(s): path to the files with genomic coordinates (chr <tab> start <tab> end <tab> strand) for the splice junction introns. Multiple files can be supplied wand will be concatenated.\n",
      "\n",
      "sjdbGTFfile                             -\n",
      "    string: path to the GTF file with annotations\n",
      "\n",
      "sjdbGTFchrPrefix                        -\n",
      "    string: prefix for chromosome names in a GTF file (e.g. 'chr' for using ENSMEBL annotations with UCSC genomes)\n",
      "\n",
      "sjdbGTFfeatureExon                      exon\n",
      "    string: feature type in GTF file to be used as exons for building transcripts\n",
      "\n",
      "sjdbGTFtagExonParentTranscript          transcript_id\n",
      "    string: GTF attribute name for parent transcript ID (default \"transcript_id\" works for GTF files)\n",
      "\n",
      "sjdbGTFtagExonParentGene                gene_id\n",
      "    string: GTF attribute name for parent gene ID (default \"gene_id\" works for GTF files)\n",
      "\n",
      "sjdbGTFtagExonParentGeneName            gene_name\n",
      "    string(s): GTF attrbute name for parent gene name\n",
      "\n",
      "sjdbGTFtagExonParentGeneType            gene_type gene_biotype\n",
      "    string(s): GTF attrbute name for parent gene type\n",
      "\n",
      "sjdbOverhang                            100\n",
      "    int>0: length of the donor/acceptor sequence on each side of the junctions, ideally = (mate_length - 1)\n",
      "\n",
      "sjdbScore                               2\n",
      "    int: extra alignment score for alignments that cross database junctions\n",
      "\n",
      "sjdbInsertSave                          Basic\n",
      "    string: which files to save when sjdb junctions are inserted on the fly at the mapping step\n",
      "                    Basic ... only small junction / transcript files\n",
      "                    All   ... all files including big Genome, SA and SAindex - this will create a complete genome directory\n",
      "\n",
      "### Variation parameters\n",
      "varVCFfile                              -\n",
      "    string: path to the VCF file that contains variation data. The 10th column should contain the genotype information, e.g. 0/1\n",
      "\n",
      "### Input Files\n",
      "inputBAMfile                -\n",
      "    string: path to BAM input file, to be used with --runMode inputAlignmentsFromBAM\n",
      "\n",
      "### Read Parameters\n",
      "readFilesType               Fastx\n",
      "    string: format of input read files\n",
      "                            Fastx       ... FASTA or FASTQ\n",
      "                            SAM SE      ... SAM or BAM single-end reads; for BAM use --readFilesCommand samtools view\n",
      "                            SAM PE      ... SAM or BAM paired-end reads; for BAM use --readFilesCommand samtools view\n",
      "                            \n",
      "readFilesSAMattrKeep        All\n",
      "    string(s): for --readFilesType SAM SE/PE, which SAM tags to keep in the output BAM, e.g.: --readFilesSAMtagsKeep RG PL\n",
      "                            All     ... keep all tags\n",
      "                            None    ... do not keep any tags\n",
      "\n",
      "readFilesIn                 Read1 Read2\n",
      "    string(s): paths to files that contain input read1 (and, if needed,  read2)\n",
      "\n",
      "readFilesManifest           -\n",
      "    string: path to the \"manifest\" file with the names of read files. The manifest file should contain 3 tab-separated columns:\n",
      "            paired-end reads: read1_file_name $tab$ read2_file_name $tab$ read_group_line.\n",
      "            single-end reads: read1_file_name $tab$ -               $tab$ read_group_line.\n",
      "            Spaces, but not tabs are allowed in file names.\n",
      "            If read_group_line does not start with ID:, it can only contain one ID field, and ID: will be added to it.\n",
      "            If read_group_line starts with ID:, it can contain several fields separated by $tab$, and all fields will be be copied verbatim into SAM @RG header line.\n",
      "\n",
      "readFilesPrefix             -\n",
      "    string: prefix for the read files names, i.e. it will be added in front of the strings in --readFilesIn\n",
      "\n",
      "readFilesCommand             -\n",
      "    string(s): command line to execute for each of the input file. This command should generate FASTA or FASTQ text and send it to stdout\n",
      "               For example: zcat - to uncompress .gz files, bzcat - to uncompress .bz2 files, etc.\n",
      "\n",
      "readMapNumber               -1\n",
      "    int: number of reads to map from the beginning of the file\n",
      "                            -1: map all reads\n",
      "\n",
      "readMatesLengthsIn          NotEqual\n",
      "    string: Equal/NotEqual - lengths of names,sequences,qualities for both mates are the same  / not the same. NotEqual is safe in all situations.\n",
      "\n",
      "readNameSeparator           /\n",
      "    string(s): character(s) separating the part of the read names that will be trimmed in output (read name after space is always trimmed)\n",
      "\n",
      "readQualityScoreBase        33\n",
      "    int>=0: number to be subtracted from the ASCII code to get Phred quality score\n",
      "\n",
      "### Read Clipping\n",
      "\n",
      "clipAdapterType             Hamming\n",
      "    string:                 adapter clipping type\n",
      "                            Hamming ... adapter clipping based on Hamming distance, with the number of mismatches controlled by --clip5pAdapterMMp\n",
      "                            CellRanger4 ... 5p and 3p adapter clipping similar to CellRanger4. Utilizes Opal package by Martin Šošić: https://github.com/Martinsos/opal\n",
      "                            None ... no adapter clipping, all other clip* parameters are disregarded\n",
      "                            \n",
      "clip3pNbases                 0\n",
      "    int(s): number(s) of bases to clip from 3p of each mate. If one value is given, it will be assumed the same for both mates.\n",
      "\n",
      "clip3pAdapterSeq            -\n",
      "    string(s): adapter sequences to clip from 3p of each mate.  If one value is given, it will be assumed the same for both mates.\n",
      "                            polyA ... polyA sequence with the length equal to read length\n",
      "\n",
      "clip3pAdapterMMp            0.1\n",
      "    double(s): max proportion of mismatches for 3p adapter clipping for each mate.  If one value is given, it will be assumed the same for both mates.\n",
      "\n",
      "clip3pAfterAdapterNbases    0\n",
      "    int(s): number of bases to clip from 3p of each mate after the adapter clipping. If one value is given, it will be assumed the same for both mates.\n",
      "\n",
      "clip5pNbases                 0\n",
      "    int(s): number(s) of bases to clip from 5p of each mate. If one value is given, it will be assumed the same for both mates.\n",
      "\n",
      "#####UnderDevelopment_begin : not supported - do not use   \n",
      "clip5pAdapterSeq            -\n",
      "    string(s): adapter sequences to clip from 5p of each mate, separated by space.\n",
      "\n",
      "clip5pAdapterMMp            0.1\n",
      "    double(s): max proportion of mismatches for 5p adapter clipping for each mate, separated by space\n",
      "\n",
      "clip5pAfterAdapterNbases    0\n",
      "    int(s): number of bases to clip from 5p of each mate after the adapter clipping, separated by space.\n",
      "#####UnderDevelopment_end\n",
      "\n",
      "### Limits\n",
      "limitGenomeGenerateRAM               31000000000\n",
      "    int>0: maximum available RAM (bytes) for genome generation\n",
      "\n",
      "limitIObufferSize                    30000000 50000000\n",
      "    int>0: max available buffers size (bytes) for input/output, per thread\n",
      "\n",
      "limitOutSAMoneReadBytes              100000\n",
      "    int>0: max size of the SAM record (bytes) for one read. Recommended value: >(2*(LengthMate1+LengthMate2+100)*outFilterMultimapNmax\n",
      "\n",
      "limitOutSJoneRead                    1000\n",
      "    int>0: max number of junctions for one read (including all multi-mappers)\n",
      "\n",
      "limitOutSJcollapsed                  1000000\n",
      "    int>0: max number of collapsed junctions\n",
      "\n",
      "limitBAMsortRAM                         0\n",
      "    int>=0: maximum available RAM (bytes) for sorting BAM. If =0, it will be set to the genome index size. 0 value can only be used with --genomeLoad NoSharedMemory option.\n",
      "\n",
      "limitSjdbInsertNsj                     1000000\n",
      "    int>=0: maximum number of junction to be inserted to the genome on the fly at the mapping stage, including those from annotations and those detected in the 1st step of the 2-pass run\n",
      "\n",
      "limitNreadsSoft                        -1\n",
      "    int: soft limit on the number of reads\n",
      "\n",
      "### Output: general\n",
      "outFileNamePrefix               ./\n",
      "    string: output files name prefix (including full or relative path). Can only be defined on the command line.\n",
      "\n",
      "outTmpDir                       -\n",
      "    string: path to a directory that will be used as temporary by STAR. All contents of this directory will be removed!\n",
      "            - the temp directory will default to outFileNamePrefix_STARtmp\n",
      "\n",
      "outTmpKeep                      None\n",
      "    string: whether to keep the tempporary files after STAR runs is finished\n",
      "                                None ... remove all temporary files\n",
      "                                All .. keep all files\n",
      "\n",
      "outStd                          Log\n",
      "    string: which output will be directed to stdout (standard out)\n",
      "                                Log                    ... log messages\n",
      "                                SAM                    ... alignments in SAM format (which normally are output to Aligned.out.sam file), normal standard output will go into Log.std.out\n",
      "                                BAM_Unsorted           ... alignments in BAM format, unsorted. Requires --outSAMtype BAM Unsorted\n",
      "                                BAM_SortedByCoordinate ... alignments in BAM format, sorted by coordinate. Requires --outSAMtype BAM SortedByCoordinate\n",
      "                                BAM_Quant              ... alignments to transcriptome in BAM format, unsorted. Requires --quantMode TranscriptomeSAM\n",
      "\n",
      "outReadsUnmapped                None\n",
      "   string: output of unmapped and partially mapped (i.e. mapped only one mate of a paired end read) reads in separate file(s).\n",
      "                                None    ... no output\n",
      "                                Fastx   ... output in separate fasta/fastq files, Unmapped.out.mate1/2\n",
      "\n",
      "outQSconversionAdd              0\n",
      "   int: add this number to the quality score (e.g. to convert from Illumina to Sanger, use -31)\n",
      "\n",
      "outMultimapperOrder             Old_2.4\n",
      "    string: order of multimapping alignments in the output files\n",
      "                                Old_2.4             ... quasi-random order used before 2.5.0\n",
      "                                Random              ... random order of alignments for each multi-mapper. Read mates (pairs) are always adjacent, all alignment for each read stay together. This option will become default in the future releases.\n",
      "\n",
      "### Output: SAM and BAM\n",
      "outSAMtype                      SAM\n",
      "    strings: type of SAM/BAM output\n",
      "                                1st word:\n",
      "                                BAM  ... output BAM without sorting\n",
      "                                SAM  ... output SAM without sorting\n",
      "                                None ... no SAM/BAM output\n",
      "                                2nd, 3rd:\n",
      "                                Unsorted           ... standard unsorted\n",
      "                                SortedByCoordinate ... sorted by coordinate. This option will allocate extra memory for sorting which can be specified by --limitBAMsortRAM.\n",
      "\n",
      "outSAMmode                      Full\n",
      "    string: mode of SAM output\n",
      "                                None ... no SAM output\n",
      "                                Full ... full SAM output\n",
      "                                NoQS ... full SAM but without quality scores\n",
      "\n",
      "outSAMstrandField               None\n",
      "    string: Cufflinks-like strand field flag\n",
      "                                None        ... not used\n",
      "                                intronMotif ... strand derived from the intron motif. This option changes the output alignments: reads with inconsistent and/or non-canonical introns are filtered out.\n",
      "\n",
      "outSAMattributes                Standard\n",
      "    string: a string of desired SAM attributes, in the order desired for the output SAM. Tags can be listed in any combination/order.\n",
      "                                ***Presets:\n",
      "                                None        ... no attributes\n",
      "                                Standard    ... NH HI AS nM\n",
      "                                All         ... NH HI AS nM NM MD jM jI MC ch                                                                    \n",
      "                                ***Alignment:\n",
      "                                NH          ... number of loci the reads maps to: =1 for unique mappers, >1 for multimappers. Standard SAM tag.\n",
      "                                HI          ... multiple alignment index, starts with --outSAMattrIHstart (=1 by default). Standard SAM tag.\n",
      "                                AS          ... local alignment score, +1/-1 for matches/mismateches, score* penalties for indels and gaps. For PE reads, total score for two mates. Stadnard SAM tag.\n",
      "                                nM          ... number of mismatches. For PE reads, sum over two mates.\n",
      "                                NM          ... edit distance to the reference (number of mismatched + inserted + deleted bases) for each mate. Standard SAM tag.\n",
      "                                MD          ... string encoding mismatched and deleted reference bases (see standard SAM specifications). Standard SAM tag.\n",
      "                                jM          ... intron motifs for all junctions (i.e. N in CIGAR): 0: non-canonical; 1: GT/AG, 2: CT/AC, 3: GC/AG, 4: CT/GC, 5: AT/AC, 6: GT/AT. If splice junctions database is used, and a junction is annotated, 20 is added to its motif value.\n",
      "                                jI          ... start and end of introns for all junctions (1-based).\n",
      "                                XS          ... alignment strand according to --outSAMstrandField.\n",
      "                                MC          ... mate's CIGAR string. Standard SAM tag.\n",
      "                                ch          ... marks all segment of all chimeric alingments for --chimOutType WithinBAM output.\n",
      "                                cN          ... number of bases clipped from the read ends: 5' and 3'\n",
      "                                ***Variation:\n",
      "                                vA          ... variant allele\n",
      "                                vG          ... genomic coordinate of the variant overlapped by the read.\n",
      "                                vW          ... 1 - alignment passes WASP filtering; 2,3,4,5,6,7 - alignment does not pass WASP filtering. Requires --waspOutputMode SAMtag.\n",
      "                                ha          ... haplotype (1/2) when mapping to the diploid genome. Requires genome generated with --genomeTransformType Diploid .                               \n",
      "                                ***STARsolo:\n",
      "                                CR CY UR UY ... sequences and quality scores of cell barcodes and UMIs for the solo* demultiplexing.\n",
      "                                GX GN       ... gene ID and gene name for unique-gene reads.\n",
      "                                gx gn       ... gene IDs and gene names for unique- and multi-gene reads.\n",
      "                                CB UB       ... error-corrected cell barcodes and UMIs for solo* demultiplexing. Requires --outSAMtype BAM SortedByCoordinate.\n",
      "                                sM          ... assessment of CB and UMI.\n",
      "                                sS          ... sequence of the entire barcode (CB,UMI,adapter).\n",
      "                                sQ          ... quality of the entire barcode.\n",
      "                                sF          ... type of feature overlap and number of features for each alignment\n",
      "                                ***Unsupported/undocumented:\n",
      "                                rB          ... alignment block read/genomic coordinates.\n",
      "                                vR          ... read coordinate of the variant.\n",
      "\n",
      "outSAMattrIHstart               1\n",
      "    int>=0:                     start value for the IH attribute. 0 may be required by some downstream software, such as Cufflinks or StringTie.\n",
      "\n",
      "outSAMunmapped                  None\n",
      "    string(s): output of unmapped reads in the SAM format\n",
      "                                1st word:\n",
      "                                None   ... no output\n",
      "                                Within ... output unmapped reads within the main SAM file (i.e. Aligned.out.sam)\n",
      "                                2nd word:\n",
      "                                KeepPairs ... record unmapped mate for each alignment, and, in case of unsorted output, keep it adjacent to its mapped mate. Only affects multi-mapping reads.\n",
      "\n",
      "outSAMorder                     Paired\n",
      "    string: type of sorting for the SAM output\n",
      "                                Paired: one mate after the other for all paired alignments\n",
      "                                PairedKeepInputOrder: one mate after the other for all paired alignments, the order is kept the same as in the input FASTQ files\n",
      "\n",
      "outSAMprimaryFlag        OneBestScore\n",
      "    string: which alignments are considered primary - all others will be marked with 0x100 bit in the FLAG\n",
      "                                OneBestScore ... only one alignment with the best score is primary\n",
      "                                AllBestScore ... all alignments with the best score are primary\n",
      "\n",
      "outSAMreadID            Standard\n",
      "    string: read ID record type\n",
      "                                Standard ... first word (until space) from the FASTx read ID line, removing /1,/2 from the end\n",
      "                                Number   ... read number (index) in the FASTx file\n",
      "\n",
      "outSAMmapqUnique        255\n",
      "    int: 0 to 255: the MAPQ value for unique mappers\n",
      "\n",
      "outSAMflagOR           0\n",
      "    int: 0 to 65535: sam FLAG will be bitwise OR'd with this value, i.e. FLAG=FLAG | outSAMflagOR. This is applied after all flags have been set by STAR, and after outSAMflagAND. Can be used to set specific bits that are not set otherwise.\n",
      "\n",
      "outSAMflagAND           65535\n",
      "    int: 0 to 65535: sam FLAG will be bitwise AND'd with this value, i.e. FLAG=FLAG & outSAMflagOR. This is applied after all flags have been set by STAR, but before outSAMflagOR. Can be used to unset specific bits that are not set otherwise.\n",
      "\n",
      "outSAMattrRGline        -\n",
      "    string(s): SAM/BAM read group line. The first word contains the read group identifier and must start with \"ID:\", e.g. --outSAMattrRGline ID:xxx CN:yy \"DS:z z z\".\n",
      "            xxx will be added as RG tag to each output alignment. Any spaces in the tag values have to be double quoted.\n",
      "            Comma separated RG lines correspons to different (comma separated) input files in --readFilesIn. Commas have to be surrounded by spaces, e.g.\n",
      "            --outSAMattrRGline ID:xxx , ID:zzz \"DS:z z\" , ID:yyy DS:yyyy\n",
      "\n",
      "outSAMheaderHD          -\n",
      "    strings: @HD (header) line of the SAM header\n",
      "\n",
      "outSAMheaderPG          -\n",
      "    strings: extra @PG (software) line of the SAM header (in addition to STAR)\n",
      "\n",
      "outSAMheaderCommentFile -\n",
      "    string: path to the file with @CO (comment) lines of the SAM header\n",
      "\n",
      "outSAMfilter            None\n",
      "    string(s): filter the output into main SAM/BAM files\n",
      "                        KeepOnlyAddedReferences ... only keep the reads for which all alignments are to the extra reference sequences added with --genomeFastaFiles at the mapping stage.\n",
      "                        KeepAllAddedReferences ...  keep all alignments to the extra reference sequences added with --genomeFastaFiles at the mapping stage.\n",
      "\n",
      "\n",
      "outSAMmultNmax          -1\n",
      "    int: max number of multiple alignments for a read that will be output to the SAM/BAM files. Note that if this value is not equal to -1, the top scoring alignment will be output first\n",
      "                        -1 ... all alignments (up to --outFilterMultimapNmax) will be output\n",
      "\n",
      "outSAMtlen              1\n",
      "    int: calculation method for the TLEN field in the SAM/BAM files\n",
      "                        1 ... leftmost base of the (+)strand mate to rightmost base of the (-)mate. (+)sign for the (+)strand mate\n",
      "                        2 ... leftmost base of any mate to rightmost base of any mate. (+)sign for the mate with the leftmost base. This is different from 1 for overlapping mates with protruding ends\n",
      "\n",
      "outBAMcompression       1\n",
      "    int: -1 to 10  BAM compression level, -1=default compression (6?), 0=no compression, 10=maximum compression\n",
      "\n",
      "outBAMsortingThreadN    0\n",
      "    int: >=0: number of threads for BAM sorting. 0 will default to min(6,--runThreadN).\n",
      "\n",
      "outBAMsortingBinsN      50\n",
      "    int: >0:  number of genome bins fo coordinate-sorting\n",
      "\n",
      "### BAM processing\n",
      "bamRemoveDuplicatesType  -\n",
      "    string: mark duplicates in the BAM file, for now only works with (i) sorted BAM fed with inputBAMfile, and (ii) for paired-end alignments only\n",
      "                        -                       ... no duplicate removal/marking\n",
      "                        UniqueIdentical         ... mark all multimappers, and duplicate unique mappers. The coordinates, FLAG, CIGAR must be identical\n",
      "                        UniqueIdenticalNotMulti  ... mark duplicate unique mappers but not multimappers.\n",
      "\n",
      "bamRemoveDuplicatesMate2basesN   0\n",
      "    int>0: number of bases from the 5' of mate 2 to use in collapsing (e.g. for RAMPAGE)\n",
      "\n",
      "### Output Wiggle\n",
      "outWigType          None\n",
      "    string(s): type of signal output, e.g. \"bedGraph\" OR \"bedGraph read1_5p\". Requires sorted BAM: --outSAMtype BAM SortedByCoordinate .\n",
      "                    1st word:\n",
      "                    None       ... no signal output\n",
      "                    bedGraph   ... bedGraph format\n",
      "                    wiggle     ... wiggle format\n",
      "                    2nd word:\n",
      "                    read1_5p   ... signal from only 5' of the 1st read, useful for CAGE/RAMPAGE etc\n",
      "                    read2      ... signal from only 2nd read\n",
      "\n",
      "outWigStrand        Stranded\n",
      "    string: strandedness of wiggle/bedGraph output\n",
      "                    Stranded   ...  separate strands, str1 and str2\n",
      "                    Unstranded ...  collapsed strands\n",
      "\n",
      "outWigReferencesPrefix    -\n",
      "    string: prefix matching reference names to include in the output wiggle file, e.g. \"chr\", default \"-\" - include all references\n",
      "\n",
      "outWigNorm              RPM\n",
      "    string: type of normalization for the signal\n",
      "                        RPM    ... reads per million of mapped reads\n",
      "                        None   ... no normalization, \"raw\" counts\n",
      "\n",
      "### Output Filtering\n",
      "outFilterType                   Normal\n",
      "    string: type of filtering\n",
      "                                Normal  ... standard filtering using only current alignment\n",
      "                                BySJout ... keep only those reads that contain junctions that passed filtering into SJ.out.tab\n",
      "\n",
      "outFilterMultimapScoreRange     1\n",
      "    int: the score range below the maximum score for multimapping alignments\n",
      "\n",
      "outFilterMultimapNmax           10\n",
      "    int: maximum number of loci the read is allowed to map to. Alignments (all of them) will be output only if the read maps to no more loci than this value.\n",
      "         Otherwise no alignments will be output, and the read will be counted as \"mapped to too many loci\" in the Log.final.out .\n",
      "\n",
      "outFilterMismatchNmax           10\n",
      "    int: alignment will be output only if it has no more mismatches than this value.\n",
      "\n",
      "outFilterMismatchNoverLmax      0.3\n",
      "    real: alignment will be output only if its ratio of mismatches to *mapped* length is less than or equal to this value.\n",
      "\n",
      "outFilterMismatchNoverReadLmax  1.0\n",
      "    real: alignment will be output only if its ratio of mismatches to *read* length is less than or equal to this value.\n",
      "\n",
      "\n",
      "outFilterScoreMin               0\n",
      "    int: alignment will be output only if its score is higher than or equal to this value.\n",
      "\n",
      "outFilterScoreMinOverLread      0.66\n",
      "    real: same as outFilterScoreMin, but  normalized to read length (sum of mates' lengths for paired-end reads)\n",
      "\n",
      "outFilterMatchNmin              0\n",
      "    int: alignment will be output only if the number of matched bases is higher than or equal to this value.\n",
      "\n",
      "outFilterMatchNminOverLread     0.66\n",
      "    real: sam as outFilterMatchNmin, but normalized to the read length (sum of mates' lengths for paired-end reads).\n",
      "\n",
      "outFilterIntronMotifs           None\n",
      "    string: filter alignment using their motifs\n",
      "                None                           ... no filtering\n",
      "                RemoveNoncanonical             ... filter out alignments that contain non-canonical junctions\n",
      "                RemoveNoncanonicalUnannotated  ... filter out alignments that contain non-canonical unannotated junctions when using annotated splice junctions database. The annotated non-canonical junctions will be kept.\n",
      "\n",
      "outFilterIntronStrands          RemoveInconsistentStrands\n",
      "    string: filter alignments\n",
      "                RemoveInconsistentStrands      ... remove alignments that have junctions with inconsistent strands\n",
      "                None                           ... no filtering\n",
      "\n",
      "### Output splice junctions (SJ.out.tab)\n",
      "outSJtype                       Standard\n",
      "    string: type of splice junction output\n",
      "                                Standard    ... standard SJ.out.tab output\n",
      "                                None        ... no splice junction output\n",
      "\n",
      "### Output Filtering: Splice Junctions\n",
      "outSJfilterReads                All\n",
      "    string: which reads to consider for collapsed splice junctions output\n",
      "                                All     ... all reads, unique- and multi-mappers\n",
      "                                Unique  ... uniquely mapping reads only\n",
      "\n",
      "outSJfilterOverhangMin          30  12  12  12\n",
      "    4 integers:    minimum overhang length for splice junctions on both sides for: (1) non-canonical motifs, (2) GT/AG and CT/AC motif, (3) GC/AG and CT/GC motif, (4) AT/AC and GT/AT motif. -1 means no output for that motif\n",
      "                                does not apply to annotated junctions\n",
      "\n",
      "outSJfilterCountUniqueMin       3   1   1   1\n",
      "    4 integers: minimum uniquely mapping read count per junction for: (1) non-canonical motifs, (2) GT/AG and CT/AC motif, (3) GC/AG and CT/GC motif, (4) AT/AC and GT/AT motif. -1 means no output for that motif\n",
      "                                Junctions are output if one of outSJfilterCountUniqueMin OR outSJfilterCountTotalMin conditions are satisfied\n",
      "                                does not apply to annotated junctions\n",
      "\n",
      "outSJfilterCountTotalMin     3   1   1   1\n",
      "    4 integers: minimum total (multi-mapping+unique) read count per junction for: (1) non-canonical motifs, (2) GT/AG and CT/AC motif, (3) GC/AG and CT/GC motif, (4) AT/AC and GT/AT motif. -1 means no output for that motif\n",
      "                                Junctions are output if one of outSJfilterCountUniqueMin OR outSJfilterCountTotalMin conditions are satisfied\n",
      "                                does not apply to annotated junctions\n",
      "\n",
      "outSJfilterDistToOtherSJmin     10  0   5   10\n",
      "    4 integers>=0: minimum allowed distance to other junctions' donor/acceptor\n",
      "                                does not apply to annotated junctions\n",
      "\n",
      "outSJfilterIntronMaxVsReadN        50000 100000 200000\n",
      "    N integers>=0: maximum gap allowed for junctions supported by 1,2,3,,,N reads\n",
      "                                i.e. by default junctions supported by 1 read can have gaps <=50000b, by 2 reads: <=100000b, by 3 reads: <=200000. by >=4 reads any gap <=alignIntronMax\n",
      "                                does not apply to annotated junctions\n",
      "\n",
      "### Scoring\n",
      "scoreGap                     0\n",
      "    int: splice junction penalty (independent on intron motif)\n",
      "\n",
      "scoreGapNoncan               -8\n",
      "    int: non-canonical junction penalty (in addition to scoreGap)\n",
      "\n",
      "scoreGapGCAG                 -4\n",
      "    GC/AG and CT/GC junction penalty (in addition to scoreGap)\n",
      "\n",
      "scoreGapATAC                 -8\n",
      "    AT/AC  and GT/AT junction penalty  (in addition to scoreGap)\n",
      "\n",
      "scoreGenomicLengthLog2scale   -0.25\n",
      "    extra score logarithmically scaled with genomic length of the alignment: scoreGenomicLengthLog2scale*log2(genomicLength)\n",
      "\n",
      "scoreDelOpen                 -2\n",
      "    deletion open penalty\n",
      "\n",
      "scoreDelBase                 -2\n",
      "    deletion extension penalty per base (in addition to scoreDelOpen)\n",
      "\n",
      "scoreInsOpen                 -2\n",
      "    insertion open penalty\n",
      "\n",
      "scoreInsBase                 -2\n",
      "    insertion extension penalty per base (in addition to scoreInsOpen)\n",
      "\n",
      "scoreStitchSJshift           1\n",
      "    maximum score reduction while searching for SJ boundaries in the stitching step\n",
      "\n",
      "\n",
      "### Alignments and Seeding\n",
      "\n",
      "seedSearchStartLmax             50\n",
      "    int>0: defines the search start point through the read - the read is split into pieces no longer than this value\n",
      "\n",
      "seedSearchStartLmaxOverLread    1.0\n",
      "    real: seedSearchStartLmax normalized to read length (sum of mates' lengths for paired-end reads)\n",
      "\n",
      "seedSearchLmax       0\n",
      "    int>=0: defines the maximum length of the seeds, if =0 seed length is not limited\n",
      "\n",
      "seedMultimapNmax      10000\n",
      "    int>0: only pieces that map fewer than this value are utilized in the stitching procedure\n",
      "\n",
      "seedPerReadNmax       1000\n",
      "    int>0: max number of seeds per read\n",
      "\n",
      "seedPerWindowNmax     50\n",
      "    int>0: max number of seeds per window\n",
      "\n",
      "seedNoneLociPerWindow    10\n",
      "    int>0: max number of one seed loci per window\n",
      "\n",
      "seedSplitMin                12\n",
      "    int>0: min length of the seed sequences split by Ns or mate gap\n",
      "\n",
      "seedMapMin              5\n",
      "    int>0: min length of seeds to be mapped\n",
      "\n",
      "alignIntronMin              21\n",
      "    minimum intron size: genomic gap is considered intron if its length>=alignIntronMin, otherwise it is considered Deletion\n",
      "\n",
      "alignIntronMax              0\n",
      "    maximum intron size, if 0, max intron size will be determined by (2^winBinNbits)*winAnchorDistNbins\n",
      "\n",
      "alignMatesGapMax            0\n",
      "    maximum gap between two mates, if 0, max intron gap will be determined by (2^winBinNbits)*winAnchorDistNbins\n",
      "\n",
      "alignSJoverhangMin          5\n",
      "    int>0: minimum overhang (i.e. block size) for spliced alignments\n",
      "\n",
      "alignSJstitchMismatchNmax   0 -1 0 0\n",
      "    4*int>=0: maximum number of mismatches for stitching of the splice junctions (-1: no limit).\n",
      "                            (1) non-canonical motifs, (2) GT/AG and CT/AC motif, (3) GC/AG and CT/GC motif, (4) AT/AC and GT/AT motif.\n",
      "\n",
      "alignSJDBoverhangMin        3\n",
      "    int>0: minimum overhang (i.e. block size) for annotated (sjdb) spliced alignments\n",
      "\n",
      "alignSplicedMateMapLmin     0\n",
      "    int>0: minimum mapped length for a read mate that is spliced\n",
      "\n",
      "alignSplicedMateMapLminOverLmate 0.66\n",
      "    real>0: alignSplicedMateMapLmin normalized to mate length\n",
      "\n",
      "alignWindowsPerReadNmax     10000\n",
      "    int>0: max number of windows per read\n",
      "\n",
      "alignTranscriptsPerWindowNmax       100\n",
      "    int>0: max number of transcripts per window\n",
      "\n",
      "alignTranscriptsPerReadNmax               10000\n",
      "    int>0: max number of different alignments per read to consider\n",
      "\n",
      "alignEndsType           Local\n",
      "    string: type of read ends alignment\n",
      "                        Local             ... standard local alignment with soft-clipping allowed\n",
      "                        EndToEnd          ... force end-to-end read alignment, do not soft-clip\n",
      "                        Extend5pOfRead1   ... fully extend only the 5p of the read1, all other ends: local alignment\n",
      "                        Extend5pOfReads12 ... fully extend only the 5p of the both read1 and read2, all other ends: local alignment\n",
      "\n",
      "alignEndsProtrude       0    ConcordantPair\n",
      "    int, string:        allow protrusion of alignment ends, i.e. start (end) of the +strand mate downstream of the start (end) of the -strand mate\n",
      "                        1st word: int: maximum number of protrusion bases allowed\n",
      "                        2nd word: string:\n",
      "                                            ConcordantPair ... report alignments with non-zero protrusion as concordant pairs\n",
      "                                            DiscordantPair ... report alignments with non-zero protrusion as discordant pairs\n",
      "\n",
      "alignSoftClipAtReferenceEnds    Yes\n",
      "    string: allow the soft-clipping of the alignments past the end of the chromosomes\n",
      "                                Yes ... allow\n",
      "                                No  ... prohibit, useful for compatibility with Cufflinks\n",
      "\n",
      "alignInsertionFlush     None\n",
      "    string: how to flush ambiguous insertion positions\n",
      "                        None    ... insertions are not flushed\n",
      "                        Right   ... insertions are flushed to the right\n",
      "\n",
      "### Paired-End reads\n",
      "peOverlapNbasesMin          0\n",
      "    int>=0:             minimum number of overlap bases to trigger mates merging and realignment. Specify >0 value to switch on the \"merginf of overlapping mates\" algorithm.\n",
      "\n",
      "peOverlapMMp                0.01\n",
      "    real, >=0 & <1:     maximum proportion of mismatched bases in the overlap area\n",
      "\n",
      "### Windows, Anchors, Binning\n",
      "\n",
      "winAnchorMultimapNmax           50\n",
      "    int>0: max number of loci anchors are allowed to map to\n",
      "\n",
      "winBinNbits                     16\n",
      "    int>0: =log2(winBin), where winBin is the size of the bin for the windows/clustering, each window will occupy an integer number of bins.\n",
      "\n",
      "winAnchorDistNbins              9\n",
      "    int>0: max number of bins between two anchors that allows aggregation of anchors into one window\n",
      "\n",
      "winFlankNbins                   4\n",
      "    int>0: log2(winFlank), where win Flank is the size of the left and right flanking regions for each window\n",
      "\n",
      "winReadCoverageRelativeMin      0.5\n",
      "    real>=0: minimum relative coverage of the read sequence by the seeds in a window, for STARlong algorithm only.\n",
      "\n",
      "winReadCoverageBasesMin      0\n",
      "    int>0: minimum number of bases covered by the seeds in a window , for STARlong algorithm only.\n",
      "\n",
      "### Chimeric Alignments\n",
      "chimOutType                 Junctions\n",
      "    string(s): type of chimeric output\n",
      "                            Junctions       ... Chimeric.out.junction\n",
      "                            SeparateSAMold  ... output old SAM into separate Chimeric.out.sam file\n",
      "                            WithinBAM       ... output into main aligned BAM files (Aligned.*.bam)\n",
      "                            WithinBAM HardClip  ... (default) hard-clipping in the CIGAR for supplemental chimeric alignments (default if no 2nd word is present)\n",
      "                            WithinBAM SoftClip  ... soft-clipping in the CIGAR for supplemental chimeric alignments\n",
      "\n",
      "chimSegmentMin              0\n",
      "    int>=0: minimum length of chimeric segment length, if ==0, no chimeric output\n",
      "\n",
      "chimScoreMin                0\n",
      "    int>=0: minimum total (summed) score of the chimeric segments\n",
      "\n",
      "chimScoreDropMax            20\n",
      "    int>=0: max drop (difference) of chimeric score (the sum of scores of all chimeric segments) from the read length\n",
      "\n",
      "chimScoreSeparation         10\n",
      "    int>=0: minimum difference (separation) between the best chimeric score and the next one\n",
      "\n",
      "chimScoreJunctionNonGTAG    -1\n",
      "    int: penalty for a non-GT/AG chimeric junction\n",
      "\n",
      "chimJunctionOverhangMin     20\n",
      "    int>=0: minimum overhang for a chimeric junction\n",
      "\n",
      "chimSegmentReadGapMax       0\n",
      "    int>=0: maximum gap in the read sequence between chimeric segments\n",
      "\n",
      "chimFilter                  banGenomicN\n",
      "    string(s): different filters for chimeric alignments\n",
      "                            None ... no filtering\n",
      "                            banGenomicN ... Ns are not allowed in the genome sequence around the chimeric junction\n",
      "\n",
      "chimMainSegmentMultNmax        10\n",
      "    int>=1: maximum number of multi-alignments for the main chimeric segment. =1 will prohibit multimapping main segments.\n",
      "\n",
      "chimMultimapNmax                    0\n",
      "    int>=0: maximum number of chimeric multi-alignments\n",
      "                                0 ... use the old scheme for chimeric detection which only considered unique alignments\n",
      "\n",
      "chimMultimapScoreRange          1\n",
      "    int>=0: the score range for multi-mapping chimeras below the best chimeric score. Only works with --chimMultimapNmax > 1\n",
      "\n",
      "chimNonchimScoreDropMin         20\n",
      "    int>=0: to trigger chimeric detection, the drop in the best non-chimeric alignment score with respect to the read length has to be greater than this value\n",
      "\n",
      "chimOutJunctionFormat           0\n",
      "    int: formatting type for the Chimeric.out.junction file\n",
      "                                0 ... no comment lines/headers\n",
      "                                1 ... comment lines at the end of the file: command line and Nreads: total, unique/multi-mapping\n",
      "\n",
      "### Quantification of Annotations\n",
      "quantMode                   -\n",
      "    string(s): types of quantification requested\n",
      "                            -                ... none\n",
      "                            TranscriptomeSAM ... output SAM/BAM alignments to transcriptome into a separate file\n",
      "                            GeneCounts       ... count reads per gene\n",
      "\n",
      "quantTranscriptomeBAMcompression    1       1\n",
      "    int: -2 to 10  transcriptome BAM compression level\n",
      "                            -2  ... no BAM output\n",
      "                            -1  ... default compression (6?)\n",
      "                             0  ... no compression\n",
      "                             10 ... maximum compression\n",
      "\n",
      "quantTranscriptomeBan       IndelSoftclipSingleend\n",
      "    string: prohibit various alignment type\n",
      "                            IndelSoftclipSingleend  ... prohibit indels, soft clipping and single-end alignments - compatible with RSEM\n",
      "                            Singleend               ... prohibit single-end alignments\n",
      "\n",
      "### 2-pass Mapping\n",
      "twopassMode                 None\n",
      "    string: 2-pass mapping mode.\n",
      "                            None        ... 1-pass mapping\n",
      "                            Basic       ... basic 2-pass mapping, with all 1st pass junctions inserted into the genome indices on the fly\n",
      "\n",
      "twopass1readsN              -1\n",
      "    int: number of reads to process for the 1st step. Use very large number (or default -1) to map all reads in the first step.\n",
      "\n",
      "\n",
      "### WASP parameters\n",
      "waspOutputMode              None\n",
      "    string: WASP allele-specific output type. This is re-implementation of the original WASP mappability filtering by Bryce van de Geijn, Graham McVicker, Yoav Gilad & Jonathan K Pritchard. Please cite the original WASP paper: Nature Methods 12, 1061–1063 (2015), https://www.nature.com/articles/nmeth.3582 .\n",
      "                            SAMtag      ... add WASP tags to the alignments that pass WASP filtering\n",
      "\n",
      "### STARsolo (single cell RNA-seq) parameters\n",
      "soloType                    None\n",
      "    string(s): type of single-cell RNA-seq\n",
      "                            CB_UMI_Simple   ... (a.k.a. Droplet) one UMI and one Cell Barcode of fixed length in read2, e.g. Drop-seq and 10X Chromium.\n",
      "                            CB_UMI_Complex  ... multiple Cell Barcodes of varying length, one UMI of fixed length and one adapter sequence of fixed length are allowed in read2 only (e.g. inDrop, ddSeq).\n",
      "                            CB_samTagOut    ... output Cell Barcode as CR and/or CB SAm tag. No UMI counting. --readFilesIn cDNA_read1 [cDNA_read2 if paired-end] CellBarcode_read . Requires --outSAMtype BAM Unsorted [and/or SortedByCoordinate]\n",
      "                            SmartSeq        ... Smart-seq: each cell in a separate FASTQ (paired- or single-end), barcodes are corresponding read-groups, no UMI sequences, alignments deduplicated according to alignment start and end (after extending soft-clipped bases)\n",
      "\n",
      "soloCBtype                  Sequence\n",
      "    string: cell barcode type\n",
      "                            Sequence: cell barcode is a sequence (standard option)\n",
      "                            String: cell barcode is an arbitrary string\n",
      "\n",
      "soloCBwhitelist             -\n",
      "    string(s): file(s) with whitelist(s) of cell barcodes. Only --soloType CB_UMI_Complex allows more than one whitelist file.\n",
      "                            None            ... no whitelist: all cell barcodes are allowed\n",
      "\n",
      "soloCBstart                 1\n",
      "    int>0: cell barcode start base\n",
      "\n",
      "soloCBlen                   16\n",
      "    int>0: cell barcode length\n",
      "\n",
      "soloUMIstart                17\n",
      "    int>0: UMI start base\n",
      "\n",
      "soloUMIlen                  10\n",
      "    int>0: UMI length\n",
      "\n",
      "soloBarcodeReadLength       1\n",
      "    int: length of the barcode read\n",
      "                            1   ... equal to sum of soloCBlen+soloUMIlen\n",
      "                            0   ... not defined, do not check\n",
      "\n",
      "soloBarcodeMate             0                            \n",
      "    int: identifies which read mate contains the barcode (CB+UMI) sequence\n",
      "                            0   ... barcode sequence is on separate read, which should always be the last file in the --readFilesIn listed\n",
      "                            1   ... barcode sequence is a part of mate 1\n",
      "                            2   ... barcode sequence is a part of mate 2\n",
      "\n",
      "soloCBposition              -\n",
      "    strings(s)              position of Cell Barcode(s) on the barcode read.\n",
      "                            Presently only works with --soloType CB_UMI_Complex, and barcodes are assumed to be on Read2.\n",
      "                            Format for each barcode: startAnchor_startPosition_endAnchor_endPosition\n",
      "                            start(end)Anchor defines the Anchor Base for the CB: 0: read start; 1: read end; 2: adapter start; 3: adapter end\n",
      "                            start(end)Position is the 0-based position with of the CB start(end) with respect to the Anchor Base\n",
      "                            String for different barcodes are separated by space.\n",
      "                            Example: inDrop (Zilionis et al, Nat. Protocols, 2017):\n",
      "                            --soloCBposition  0_0_2_-1  3_1_3_8\n",
      "\n",
      "soloUMIposition             -\n",
      "    string                  position of the UMI on the barcode read, same as soloCBposition\n",
      "                            Example: inDrop (Zilionis et al, Nat. Protocols, 2017):\n",
      "                            --soloCBposition  3_9_3_14\n",
      "\n",
      "soloAdapterSequence         -\n",
      "    string:                 adapter sequence to anchor barcodes. Only one adapter sequence is allowed.\n",
      "\n",
      "soloAdapterMismatchesNmax   1\n",
      "    int>0:                  maximum number of mismatches allowed in adapter sequence.\n",
      "\n",
      "soloCBmatchWLtype           1MM_multi\n",
      "    string:                 matching the Cell Barcodes to the WhiteList\n",
      "                            Exact                           ... only exact matches allowed\n",
      "                            1MM                             ... only one match in whitelist with 1 mismatched base allowed. Allowed CBs have to have at least one read with exact match.\n",
      "                            1MM_multi                       ... multiple matches in whitelist with 1 mismatched base allowed, posterior probability calculation is used choose one of the matches.\n",
      "                                                                Allowed CBs have to have at least one read with exact match. This option matches best with CellRanger 2.2.0\n",
      "                            1MM_multi_pseudocounts          ... same as 1MM_Multi, but pseudocounts of 1 are added to all whitelist barcodes.\n",
      "                            1MM_multi_Nbase_pseudocounts    ... same as 1MM_multi_pseudocounts, multimatching to WL is allowed for CBs with N-bases. This option matches best with CellRanger >= 3.0.0\n",
      "                            EditDist_2                    ... allow up to edit distance of 3 fpr each of the barcodes. May include one deletion + one insertion. Only works with --soloType CB_UMI_Complex. Matches to multiple passlist barcdoes are not allowed. Similar to ParseBio Split-seq pipeline. \n",
      "\n",
      "soloInputSAMattrBarcodeSeq  -\n",
      "    string(s):              when inputting reads from a SAM file (--readsFileType SAM SE/PE), these SAM attributes mark the barcode sequence (in proper order).\n",
      "                            For instance, for 10X CellRanger or STARsolo BAMs, use --soloInputSAMattrBarcodeSeq CR UR .\n",
      "                            This parameter is required when running STARsolo with input from SAM. \n",
      "    \n",
      "soloInputSAMattrBarcodeQual  -\n",
      "    string(s):              when inputting reads from a SAM file (--readsFileType SAM SE/PE), these SAM attributes mark the barcode qualities (in proper order).\n",
      "                            For instance, for 10X CellRanger or STARsolo BAMs, use --soloInputSAMattrBarcodeQual CY UY .\n",
      "                            If this parameter is '-' (default), the quality 'H' will be assigned to all bases.\n",
      "\n",
      "soloStrand                  Forward\n",
      "    string: strandedness of the solo libraries:\n",
      "                            Unstranded  ... no strand information\n",
      "                            Forward     ... read strand same as the original RNA molecule\n",
      "                            Reverse     ... read strand opposite to the original RNA molecule\n",
      "\n",
      "soloFeatures                Gene\n",
      "    string(s): genomic features for which the UMI counts per Cell Barcode are collected\n",
      "                            Gene            ... genes: reads match the gene transcript\n",
      "                            SJ              ... splice junctions: reported in SJ.out.tab\n",
      "                            GeneFull        ... full gene (pre-mRNA): count all reads overlapping genes' exons and introns\n",
      "                            GeneFull_ExonOverIntron ... full gene (pre-mRNA): count all reads overlapping genes' exons and introns: prioritize 100% overlap with exons\n",
      "                            GeneFull_Ex50pAS        ... full gene (pre-RNA): count all reads overlapping genes' exons and introns: prioritize >50% overlap with exons. Do not count reads with 100% exonic overlap in the antisense direction.\n",
      "\n",
      "#####UnderDevelopment_begin : not supported - do not use\n",
      "                            Transcript3p    ... quantification of transcript for 3' protocols\n",
      "#####UnderDevelopment_end\n",
      "\n",
      "soloMultiMappers            Unique\n",
      "    string(s): counting method for reads mapping to multiple genes           \n",
      "                            Unique     ... count only reads that map to unique genes\n",
      "                            Uniform    ... uniformly distribute multi-genic UMIs to all genes\n",
      "                            Rescue     ... distribute UMIs proportionally to unique+uniform counts (~ first iteration of EM)\n",
      "                            PropUnique ... distribute UMIs proportionally to unique mappers, if present, and uniformly if not.\n",
      "                            EM         ... multi-gene UMIs are distributed using Expectation Maximization algorithm\n",
      "\n",
      "soloUMIdedup                1MM_All\n",
      "    string(s):              type of UMI deduplication (collapsing) algorithm\n",
      "                            1MM_All                     ... all UMIs with 1 mismatch distance to each other are collapsed (i.e. counted once).\n",
      "                            1MM_Directional_UMItools    ... follows the \"directional\" method from the UMI-tools by Smith, Heger and Sudbery (Genome Research 2017).\n",
      "                            1MM_Directional             ... same as 1MM_Directional_UMItools, but with more stringent criteria for duplicate UMIs\n",
      "                            Exact                       ... only exactly matching UMIs are collapsed.\n",
      "                            NoDedup                     ... no deduplication of UMIs, count all reads.\n",
      "                            1MM_CR                      ... CellRanger2-4 algorithm for 1MM UMI collapsing.\n",
      "\n",
      "soloUMIfiltering            -\n",
      "    string(s)               type of UMI filtering (for reads uniquely mapping to genes)\n",
      "                            -                  ... basic filtering: remove UMIs with N and homopolymers (similar to CellRanger 2.2.0).\n",
      "                            MultiGeneUMI       ... basic + remove lower-count UMIs that map to more than one gene.\n",
      "                            MultiGeneUMI_All   ... basic + remove all UMIs that map to more than one gene.\n",
      "                            MultiGeneUMI_CR    ... basic + remove lower-count UMIs that map to more than one gene, matching CellRanger > 3.0.0 .\n",
      "                                                   Only works with --soloUMIdedup 1MM_CR\n",
      "                                                \n",
      "soloOutFileNames            Solo.out/          features.tsv barcodes.tsv        matrix.mtx\n",
      "    string(s)               file names for STARsolo output:\n",
      "                            file_name_prefix   gene_names   barcode_sequences   cell_feature_count_matrix\n",
      "\n",
      "soloCellFilter              CellRanger2.2 3000 0.99 10\n",
      "    string(s):              cell filtering type and parameters\n",
      "                            None            ... do not output filtered cells\n",
      "                            TopCells        ... only report top cells by UMI count, followed by the exact number of cells\n",
      "                            CellRanger2.2   ... simple filtering of CellRanger 2.2. \n",
      "                                                Can be followed by numbers: number of expected cells, robust maximum percentile for UMI count, maximum to minimum ratio for UMI count\n",
      "                                                The harcoded values are from CellRanger: nExpectedCells=3000;  maxPercentile=0.99;  maxMinRatio=10\n",
      "                            EmptyDrops_CR   ... EmptyDrops filtering in CellRanger flavor. Please cite the original EmptyDrops paper: A.T.L Lun et al, Genome Biology, 20, 63 (2019): https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1662-y\n",
      "                                                Can be followed by 10 numeric parameters:  nExpectedCells   maxPercentile   maxMinRatio   indMin   indMax   umiMin   umiMinFracMedian   candMaxN   FDR   simN \n",
      "                                                The harcoded values are from CellRanger:             3000            0.99            10    45000    90000      500               0.01      20000  0.01  10000\n",
      "\n",
      "soloOutFormatFeaturesGeneField3    \"Gene Expression\"\n",
      "    string(s):                field 3 in the Gene features.tsv file. If \"-\", then no 3rd field is output.\n",
      "\n",
      "soloCellReadStats           None\n",
      "    string:                 Output reads statistics for each CB\n",
      "                            Standard    ... standard output\n",
      "\n",
      "#####UnderDevelopment_begin : not supported - do not use\n",
      "soloClusterCBfile           -\n",
      "    string:                 file containing the cluster information for cell barcodes, two columns: CB cluster_index. Only used with --soloFeatures Transcript3p\n",
      "#####UnderDevelopment_end\n"
     ]
    }
   ],
   "source": [
    "!STAR -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8117db0-0cda-418b-90bb-db68a82f3827",
   "metadata": {},
   "source": [
    "##### paramaters\n",
    "- --runThreadN {nc},\n",
    "- --runMode alignReads,\n",
    "- --readFilesCommand cat,\n",
    "- --genomeDir {genome_index},\n",
    "- --alignEndsType EndToEnd,\n",
    "---quantMode TranscriptomeSAM,\n",
    "- --readFilesIn {fname(in_path,sample,\"fq\")},\n",
    "- --outFileNamePrefix {out_path}/{sample}_,\n",
    "- --outFilterMultimapNmax 1,\n",
    "- --outFilterMultimapScoreRange 1,\n",
    "- --outSAMattributes All,\n",
    "- --outSAMtype BAM SortedByCoordinate,\n",
    "- --outFilterType BySJout,\n",
    "- --outReadsUnmapped Fastx,\n",
    "- --outFilterScoreMin 10,\n",
    "- --outFilterMatchNmin 24,"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "b93de45a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "extract_rRNA/a4_1.fq\n",
      "\t/home/cdaniels/mambaforge/envs/biop/bin/STAR-avx2 --runThreadN 32 --runMode alignReads --readFilesCommand cat --genomeDir /home/cdaniels/uofc_data/ref/homo_sapian_39/hg_39 --alignEndsType EndToEnd --quantMode TranscriptomeSAM --readFilesIn extract_rRNA/a4_1.fq --outFileNamePrefix align/a4_1_ --outFilterMultimapNmax 1 --outFilterMultimapScoreRange 1 --outSAMattributes All --outSAMtype BAM SortedByCoordinate --outFilterType BySJout --outReadsUnmapped Fastx --outFilterScoreMin 10 --outFilterMatchNmin 24\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Oct 13 23:58:38 ..... started STAR run\n",
      "Oct 13 23:58:38 ..... loading genome\n",
      "Oct 13 23:58:50 ..... started mapping\n",
      "Oct 13 23:58:58 ..... finished mapping\n",
      "Oct 13 23:58:59 ..... started sorting BAM\n",
      "Oct 13 23:58:59 ..... finished successfully\n",
      "++++++++++++++++++++++++++++++++++++\n",
      "extract_rRNA/a4_2.fq\n",
      "\t/home/cdaniels/mambaforge/envs/biop/bin/STAR-avx2 --runThreadN 32 --runMode alignReads --readFilesCommand cat --genomeDir /home/cdaniels/uofc_data/ref/homo_sapian_39/hg_39 --alignEndsType EndToEnd --quantMode TranscriptomeSAM --readFilesIn extract_rRNA/a4_2.fq --outFileNamePrefix align/a4_2_ --outFilterMultimapNmax 1 --outFilterMultimapScoreRange 1 --outSAMattributes All --outSAMtype BAM SortedByCoordinate --outFilterType BySJout --outReadsUnmapped Fastx --outFilterScoreMin 10 --outFilterMatchNmin 24\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Oct 13 23:59:00 ..... started STAR run\n",
      "Oct 13 23:59:00 ..... loading genome\n",
      "Oct 13 23:59:16 ..... started mapping\n",
      "Oct 13 23:59:22 ..... finished mapping\n",
      "Oct 13 23:59:22 ..... started sorting BAM\n",
      "Oct 13 23:59:23 ..... finished successfully\n",
      "++++++++++++++++++++++++++++++++++++\n",
      "extract_rRNA/a4_3.fq\n",
      "\t/home/cdaniels/mambaforge/envs/biop/bin/STAR-avx2 --runThreadN 32 --runMode alignReads --readFilesCommand cat --genomeDir /home/cdaniels/uofc_data/ref/homo_sapian_39/hg_39 --alignEndsType EndToEnd --quantMode TranscriptomeSAM --readFilesIn extract_rRNA/a4_3.fq --outFileNamePrefix align/a4_3_ --outFilterMultimapNmax 1 --outFilterMultimapScoreRange 1 --outSAMattributes All --outSAMtype BAM SortedByCoordinate --outFilterType BySJout --outReadsUnmapped Fastx --outFilterScoreMin 10 --outFilterMatchNmin 24\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Oct 13 23:59:23 ..... started STAR run\n",
      "Oct 13 23:59:23 ..... loading genome\n",
      "Oct 13 23:59:35 ..... started mapping\n",
      "Oct 13 23:59:42 ..... finished mapping\n",
      "Oct 13 23:59:43 ..... started sorting BAM\n",
      "Oct 13 23:59:43 ..... finished successfully\n",
      "++++++++++++++++++++++++++++++++++++\n",
      "extract_rRNA/a4_4.fq\n",
      "\t/home/cdaniels/mambaforge/envs/biop/bin/STAR-avx2 --runThreadN 32 --runMode alignReads --readFilesCommand cat --genomeDir /home/cdaniels/uofc_data/ref/homo_sapian_39/hg_39 --alignEndsType EndToEnd --quantMode TranscriptomeSAM --readFilesIn extract_rRNA/a4_4.fq --outFileNamePrefix align/a4_4_ --outFilterMultimapNmax 1 --outFilterMultimapScoreRange 1 --outSAMattributes All --outSAMtype BAM SortedByCoordinate --outFilterType BySJout --outReadsUnmapped Fastx --outFilterScoreMin 10 --outFilterMatchNmin 24\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Oct 13 23:59:44 ..... started STAR run\n",
      "Oct 13 23:59:44 ..... loading genome\n",
      "Oct 13 23:59:55 ..... started mapping\n",
      "Oct 14 00:00:02 ..... finished mapping\n",
      "Oct 14 00:00:02 ..... started sorting BAM\n",
      "Oct 14 00:00:03 ..... finished successfully\n",
      "++++++++++++++++++++++++++++++++++++\n",
      "extract_rRNA/no_1.fq\n",
      "\t/home/cdaniels/mambaforge/envs/biop/bin/STAR-avx2 --runThreadN 32 --runMode alignReads --readFilesCommand cat --genomeDir /home/cdaniels/uofc_data/ref/homo_sapian_39/hg_39 --alignEndsType EndToEnd --quantMode TranscriptomeSAM --readFilesIn extract_rRNA/no_1.fq --outFileNamePrefix align/no_1_ --outFilterMultimapNmax 1 --outFilterMultimapScoreRange 1 --outSAMattributes All --outSAMtype BAM SortedByCoordinate --outFilterType BySJout --outReadsUnmapped Fastx --outFilterScoreMin 10 --outFilterMatchNmin 24\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Oct 14 00:00:03 ..... started STAR run\n",
      "Oct 14 00:00:03 ..... loading genome\n",
      "Oct 14 00:00:15 ..... started mapping\n",
      "Oct 14 00:00:22 ..... finished mapping\n",
      "Oct 14 00:00:23 ..... started sorting BAM\n",
      "Oct 14 00:00:23 ..... finished successfully\n",
      "++++++++++++++++++++++++++++++++++++\n",
      "extract_rRNA/no_2.fq\n",
      "\t/home/cdaniels/mambaforge/envs/biop/bin/STAR-avx2 --runThreadN 32 --runMode alignReads --readFilesCommand cat --genomeDir /home/cdaniels/uofc_data/ref/homo_sapian_39/hg_39 --alignEndsType EndToEnd --quantMode TranscriptomeSAM --readFilesIn extract_rRNA/no_2.fq --outFileNamePrefix align/no_2_ --outFilterMultimapNmax 1 --outFilterMultimapScoreRange 1 --outSAMattributes All --outSAMtype BAM SortedByCoordinate --outFilterType BySJout --outReadsUnmapped Fastx --outFilterScoreMin 10 --outFilterMatchNmin 24\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Oct 14 00:00:23 ..... started STAR run\n",
      "Oct 14 00:00:23 ..... loading genome\n",
      "Oct 14 00:00:35 ..... started mapping\n",
      "Oct 14 00:00:42 ..... finished mapping\n",
      "Oct 14 00:00:43 ..... started sorting BAM\n",
      "Oct 14 00:00:43 ..... finished successfully\n",
      "++++++++++++++++++++++++++++++++++++\n",
      "extract_rRNA/no_3.fq\n",
      "\t/home/cdaniels/mambaforge/envs/biop/bin/STAR-avx2 --runThreadN 32 --runMode alignReads --readFilesCommand cat --genomeDir /home/cdaniels/uofc_data/ref/homo_sapian_39/hg_39 --alignEndsType EndToEnd --quantMode TranscriptomeSAM --readFilesIn extract_rRNA/no_3.fq --outFileNamePrefix align/no_3_ --outFilterMultimapNmax 1 --outFilterMultimapScoreRange 1 --outSAMattributes All --outSAMtype BAM SortedByCoordinate --outFilterType BySJout --outReadsUnmapped Fastx --outFilterScoreMin 10 --outFilterMatchNmin 24\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Oct 14 00:00:44 ..... started STAR run\n",
      "Oct 14 00:00:44 ..... loading genome\n",
      "Oct 14 00:00:56 ..... started mapping\n",
      "Oct 14 00:01:02 ..... finished mapping\n",
      "Oct 14 00:01:03 ..... started sorting BAM\n",
      "Oct 14 00:01:03 ..... finished successfully\n",
      "++++++++++++++++++++++++++++++++++++\n",
      "extract_rRNA/no_4.fq\n",
      "\t/home/cdaniels/mambaforge/envs/biop/bin/STAR-avx2 --runThreadN 32 --runMode alignReads --readFilesCommand cat --genomeDir /home/cdaniels/uofc_data/ref/homo_sapian_39/hg_39 --alignEndsType EndToEnd --quantMode TranscriptomeSAM --readFilesIn extract_rRNA/no_4.fq --outFileNamePrefix align/no_4_ --outFilterMultimapNmax 1 --outFilterMultimapScoreRange 1 --outSAMattributes All --outSAMtype BAM SortedByCoordinate --outFilterType BySJout --outReadsUnmapped Fastx --outFilterScoreMin 10 --outFilterMatchNmin 24\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Oct 14 00:01:03 ..... started STAR run\n",
      "Oct 14 00:01:03 ..... loading genome\n",
      "Oct 14 00:01:15 ..... started mapping\n",
      "Oct 14 00:01:21 ..... finished mapping\n",
      "Oct 14 00:01:22 ..... started sorting BAM\n",
      "Oct 14 00:01:22 ..... finished successfully\n",
      "++++++++++++++++++++++++++++++++++++\n"
     ]
    }
   ],
   "source": [
    "for sample in samples:\n",
    "    print(fname(in_path,sample,\"fq\"))\n",
    "    !STAR --runThreadN {nc} \\\n",
    "              --runMode alignReads \\\n",
    "              --readFilesCommand cat \\\n",
    "              --genomeDir {genome_index} \\\n",
    "              --alignEndsType EndToEnd \\\n",
    "              --quantMode TranscriptomeSAM \\\n",
    "              --readFilesIn {fname(in_path,sample,\"fq\")} \\\n",
    "              --outFileNamePrefix {out_path}/{sample}_ \\\n",
    "              --outFilterMultimapNmax 1 \\\n",
    "              --outFilterMultimapScoreRange 1 \\\n",
    "              --outSAMattributes All \\\n",
    "              --outSAMtype BAM SortedByCoordinate \\\n",
    "              --outFilterType BySJout \\\n",
    "              --outReadsUnmapped Fastx \\\n",
    "              --outFilterScoreMin 10 \\\n",
    "              --outFilterMatchNmin 24\n",
    "    print(\"++++++++++++++++++++++++++++++++++++\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e1ec9d6",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "#### Samtools"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50c2008b-836f-4925-a41f-b67b6a206502",
   "metadata": {},
   "source": [
    "https://www.htslib.org"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "3ce26f12-98a0-4278-9dd0-cbc4369ab251",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: samtools index -M [-bc] [-m INT] <in1.bam> <in2.bam>...\n",
      "   or: samtools index [-bc] [-m INT] <in.bam> [out.index]\n",
      "Options:\n",
      "  -b       Generate BAI-format index for BAM files [default]\n",
      "  -c       Generate CSI-format index for BAM files\n",
      "  -m INT   Set minimum interval size for CSI indices to 2^INT [14]\n",
      "  -M       Interpret all filename arguments as files to be indexed\n",
      "  -o FILE  Write index to FILE [alternative to <out.index> as an argument]\n",
      "  -@ INT   Sets the number of threads [none]\n"
     ]
    }
   ],
   "source": [
    "!samtools index"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0fcb989-05b4-41e5-adf3-f75d475065e0",
   "metadata": {},
   "source": [
    "##### paramaters\n",
    "- index, generate an index file from a bam file: *.bam -> *.bai\n",
    "- @ {nc}, number of threads\n",
    "- {in_bam}, input *.bam file; out *.bai file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "b6331ed8",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for sample in samples:\n",
    "    aligned_bam = fname(out_path,f\"{sample}_Aligned.sortedByCoord.out\",\"bam\")\n",
    "    !cp {aligned_bam} {fname(out_path,sample,\"bam\")}\n",
    "    !samtools index -@ {nc} {fname(out_path,sample,\"bam\")}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c42d8e6b-34c6-43fa-aaee-a0cc6ab0c25e",
   "metadata": {},
   "source": [
    "#### Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "230097c3-8e3e-438d-8634-a4dd3ba9cfd0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[3605893, 1094741, 2478074, 1741926, 2432106, 2027040, 1433973, 1892186]"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Pre genomic alignment\n",
    "ins = [nseqs(fname(in_path,sample,'fq')) for sample in samples]\n",
    "ins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "322bf0dc-6464-4a76-9329-1453dcee5357",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2595303, 819585, 1630323, 1176367, 1766129, 1290541, 967229, 1417404]"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Post genomic alignment\n",
    "outs = [nseqs(fname(out_path,sample,'bam')) for sample in samples]\n",
    "outs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "96cab0d0-efbe-4f26-8278-7b02d58ed1fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjcAAAGxCAYAAACeKZf2AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABJ9ElEQVR4nO3deVyN+f8//sdpO+1RWkdUllREZEb2xlp4Y/gwgyn7GLveTaQZxJhsY8KQLWKsQ/EeZMkSYxsyZUxiBilDyTKK0Hp9//Dr/OZMxTl1TqeuHvfb7bq9Xa/zul7X83Xk3WOuVSIIggAiIiIikdDSdAFEREREqsRwQ0RERKLCcENERESiwnBDREREosJwQ0RERKLCcENERESiwnBDREREosJwQ0RERKLCcENERESiwnBDpAErV66ERCJB8+bNy+0jkUgwb9482Xp8fDwkEgni4+PVX2A1ExsbK/ddvMuOHTsQHh6utnoOHjyI/v37w87ODnp6ejAxMYGHhwfmzp2L9PR0te2XiBTDcEOkAZs2bQIAJCcn45dfflFom9atW+PChQto3bq1OkurlmJjYxEaGqpwf3WFm+LiYvj7+6Nfv34oKChAWFgY4uLisGfPHnz00Uf44Ycf0KFDB5Xvl4iUo6PpAohqm4SEBFy9ehV9+vTBoUOHEBkZiQ8++OCd25mamqJdu3ZVUCGVZ/Hixdi6dSvCwsIwa9Ysuc969+6N4OBgrFu3TkPVEVEJHrkhqmKRkZEAgEWLFqF9+/bYtWsXXr58+c7tyjsttWHDBjRt2hRSqRSurq7YsWMHRo4cCQcHB1mfu3fvQiKRYNmyZVi+fDkcHR1hbGwMLy8vXLx4UW68kSNHwtjYGDdu3ECvXr1gZGQEW1tbLFq0CABw8eJFdOzYEUZGRmjatCm2bNlSqtbMzEx89tlnqF+/PvT09ODo6IjQ0FAUFhYqXdPIkSOxevVqAG9O1ZUsd+/eLfN76tq1Kw4dOoS0tDS5/iWePn2KiRMn4r333oOenh6cnJwQEhKCvLy8t37/+fn5WLJkCZo3b14q2JTQ0dHBpEmTSrXv3r0bXl5eMDIygrGxMXr16oXExES5PiXf+61bt+Dr6wtjY2PY29vjv//9b6naFJ2DRCLB5MmTsXnzZjg7O8PAwACenp64ePEiBEHA0qVLZd/7hx9+iFu3bpWq/fjx4+jWrRtMTU1haGiIDh064MSJE2/9rog0TiCiKvPy5UvBzMxMaNu2rSAIgrBx40YBgBAVFVWqLwBh7ty5svVTp04JAIRTp07J2tatWycAEAYNGiQcPHhQ2L59u9C0aVOhYcOGQsOGDWX9UlNTBQCCg4OD0Lt3b2H//v3C/v37hRYtWgh169YVnj17Juvr7+8v6OnpCS4uLsKKFSuEuLg4YdSoUQIAITg4WGjatKkQGRkpHD16VOjbt68AQEhISJBtn5GRIdjb2wsNGzYU1q1bJxw/flxYsGCBIJVKhZEjRypd061bt4TBgwcLAIQLFy7IltevX5f5HScnJwsdOnQQbGxs5PoLgiC8evVKcHd3F4yMjIRly5YJx44dE7766itBR0dH8PX1fevf3blz52TfgTIWLlwoSCQSYfTo0cLBgweFmJgYwcvLSzAyMhKSk5PL/N6XLVsmHD9+XJgzZ44gkUiE0NBQWT9l5gBAaNiwodC+fXshJiZG2Ldvn9C0aVPB3NxcmDFjhtC/f3/Zz421tbXg7u4uFBcXy7b/4YcfBIlEIgwYMECIiYkRDhw4IPTt21fQ1tYWjh8/rtT3QFSVGG6IqtDWrVsFAMLatWsFQRCE58+fC8bGxkKnTp1K9X1XuCkqKhJsbGyEDz74QG67tLQ0QVdXt8xw06JFC6GwsFDWfunSJQGAsHPnTlmbv7+/AECIjo6WtRUUFAiWlpYCAOHXX3+VtT958kTQ1tYWAgICZG2fffaZYGxsLKSlpcnVtWzZMgGA7Be6MjVNmjRJUOa/xfr06SM3/xJr164VAAg//vijXPvixYsFAMKxY8fKHXPXrl1yf3f/VFBQILeUSE9PF3R0dIQpU6bI9X/+/LlgY2MjDBkyRNZW8r3/uzZfX1/B2dm5QnMAINjY2AgvXryQte3fv18AILRq1UouyISHhwsAhN9++00QBEHIzc0VzM3NhX79+sntp6ioSGjZsqXw/vvvl/tdEWlarT4tdebMGfTr1w92dnaQSCTYv3+/0mMIgoBly5bJTgvY29vjm2++UX2xJAqRkZEwMDDAxx9/DAAwNjbG//3f/+Hnn3/Gn3/+qdRYN2/eRGZmJoYMGSLX3qBBg3Ivau3Tpw+0tbVl6+7u7gCAtLQ0uX4SiQS+vr6ydR0dHTRu3Bi2trbw8PCQtZubm8PKykpu+4MHD8Lb2xt2dnYoLCyULT4+PgCA06dPV6gmVTh58iSMjIwwePBgufaRI0cCQIVOtzx79gy6urpyS0JCAgDg6NGjKCwshJ+fn9x3oa+vjy5dupQ6xSiRSNCvXz+5Nnd3d7nvQtk5eHt7w8jISLbu4uICAPDx8ZE7XVfSXrKv8+fP4+nTp/D395ervbi4GL1798bly5eRm5ur7NdFVCVq9QXFubm5aNmyJUaNGoVBgwZVaIxp06bh2LFjWLZsGVq0aIHs7Gw8fvxYxZWSGNy6dQtnzpzBoEGDIAgCnj17BgAYPHgwNm/ejE2bNiEsLEzh8Z48eQIAsLa2LvWZtbU1UlNTS7VbWFjIrUulUgDAq1ev5NoNDQ2hr68v16anpwdzc/NSY+rp6eH169ey9YcPH+LAgQPQ1dUts+5///tQtCZVePLkCWxsbOR+qQOAlZUVdHR0ZN9pWRo0aACgdOgyMTHB5cuXAbwJdv+8q+vhw4cAgLZt25Y5ppaW/H9flvW9S6VSue9X2Tn8++9MT0/vre0l+yqp/d8h6p+ePn0qF5yIqotaHW58fHxk/zVZlvz8fHz55ZfYvn07nj17hubNm2Px4sXo2rUrACAlJQURERH4/fff4ezsXEVVU021adMmCIKAvXv3Yu/evaU+37JlC77++mu5oxhvUxIKSn4J/VNmZmbliq2EevXqwd3dHQsXLizzczs7uyqu6P9nYWGBX375BYIgyIWDrKwsFBYWol69euVu26ZNG9StWxcHDhyQOzqrra0NT09PAMDvv/8ut03JeHv37kXDhg01PgdllIyzatWqcu/SKytYE1UHtTrcvMuoUaNw9+5d7Nq1C3Z2dti3bx969+6Na9euoUmTJjhw4ACcnJxw8OBB9O7dG4IgoHv37liyZEmZ/4VLtVdRURG2bNmCRo0aYePGjaU+P3jwIL799lscPnwYffv2VWhMZ2dn2NjY4Mcff0RAQICsPT09HefPn9dYiOjbty9iY2PRqFEj1K1bVyVj/vNojoGBgUL9yzry061bN/z444/Yv38/Bg4cKGvfunWr7PPy6Onp4YsvvsDs2bOxePFizJw585119OrVCzo6Orh9+3aFjw7/W2XmoIwOHTqgTp06uH79OiZPnqySMYmqCsNNOW7fvo2dO3fir7/+kv2SCAwMxJEjR7B582Z88803uHPnDtLS0rBnzx5s3boVRUVFmDFjBgYPHoyTJ09qeAZUnRw+fBgPHjyQO/L3T82bN8f333+PyMhIhcONlpYWQkND8dlnn2Hw4MEYPXo0nj17htDQUNja2pY65VFV5s+fj7i4OLRv3x5Tp06Fs7MzXr9+jbt37yI2NhZr165F/fr1lRqzRYsWAN48Z8bHxwfa2tpwd3eXnUopq39MTAwiIiLQpk0baGlpwdPTE35+fli9ejX8/f1x9+5dtGjRAmfPnsU333wDX19fdO/e/a11zJw5Ezdu3MCsWbNw5swZDB06FA4ODsjLy8OdO3ewceNGaGtrw9DQEADg4OCA+fPnIyQkBHfu3EHv3r1Rt25dPHz4EJcuXYKRkZFSDycEUOk5KMrY2BirVq2Cv78/nj59isGDB8PKygqPHj3C1atX8ejRI0RERKhkX0SqxnBTjl9//RWCIKBp06Zy7Xl5ebLTAcXFxcjLy8PWrVtl/SIjI9GmTRvcvHmTp6pIJjIyEnp6ehg1alSZn9erVw8DBw7E3r178fDhQ4UP948fPx4SiQRLlizBwIED4eDggFmzZuF///ufxl4DYGtri4SEBCxYsABLly7FX3/9BRMTEzg6Osp+uStr2LBhOHfuHNasWYP58+dDEASkpqbKPcvnn6ZNm4bk5GTMnj0b2dnZEN7cGQp9fX2cOnUKISEhWLp0KR49eoT33nsPgYGBmDt37jvr0NLSwpYtWzB48GBs2LABQUFBePLkCQwMDNCoUSN069YN27Ztk/u3HxwcDFdXV6xYsQI7d+5EXl4ebGxs0LZtW0yYMEHp76Kyc1DGiBEj0KBBAyxZsgSfffYZnj9/DisrK7Rq1Up2ATNRdSQRBEHQdBHVgUQiwb59+zBgwAAAbx66NXz4cCQnJ5e6BsLY2Bg2NjaYO3cuvvnmGxQUFMg+e/XqFQwNDXHs2DH06NGjKqdABODN3TtNmzbFgAEDsH79ek2XQ0RU5XjkphweHh4oKipCVlYWOnXqVGafDh06oLCwELdv30ajRo0AAH/88QcAqOziQaK3yczMxMKFC+Ht7Q0LCwukpaXhu+++w/PnzzFt2jRNl0dEpBG1+sjNixcvZI8b9/DwwPLly+Ht7Q1zc3M0aNAAI0aMwLlz5/Dtt9/Cw8MDjx8/xsmTJ9GiRQv4+vqiuLgYbdu2hbGxMcLDw1FcXIxJkybB1NQUx44d0/DsqDb4+++/4efnh8uXL+Pp06cwNDREu3btEBoaqtD7qoiIxKhWh5v4+Hh4e3uXavf390dUVBQKCgrw9ddfY+vWrbh//z4sLCzg5eWF0NBQ2QWODx48wJQpU3Ds2DEYGRnBx8cH3377Le+WIiIi0pBaHW6IiIhIfGr16xeIiIhIfBhuiIiISFRq3d1SxcXFePDgAUxMTEq9m4WIiIiqJ0EQ8Pz5c9jZ2b3zIaW1Ltw8ePAA9vb2mi6DiIiIKuDevXvvfMp5rQs3JiYmAN58OaamphquhoiIiBSRk5MDe3t72e/xt6l14abkVJSpqSnDDRERUQ2jyCUlvKCYiIiIRIXhhoiIiESF4YaIiIhEpdZdc0NEROpRVFSEgoICTZdBNZiuri60tbUrPQ7DDRERVdqLFy/w119/gW/0ocqQSCSoX78+jI2NKzUOww0REVVKUVER/vrrLxgaGsLS0pIPSKUKEQQBjx49wl9//YUmTZpU6ggOww0REVVKQUEBBEGApaUlDAwMNF0O1WCWlpa4e/cuCgoKKhVueEExERGpBI/YUGWp6meI4YaIiIhEheGGiIiIRIXX3BARkVo4zDpUpfu7u6hPle5P1UaOHIlnz55h//79AICuXbuiVatWCA8P12hdNRGP3BARUa00cuRISCQSSCQS6OrqwsnJCYGBgcjNzVXbPsePHw9tbW3s2rXrnX1jYmKwYMECtdVS1ebNm4dWrVpVyb4YboiIqNbq3bs3MjIycOfOHXz99ddYs2YNAgMDy+xb2QcUvnz5Ert378YXX3yByMjId/Y3NzdX6A3YVBrDDRER1VpSqRQ2Njawt7fHsGHDMHz4cNlpoZIjDZs2bYKTkxOkUikEQUB2djbGjx8PKysrmJqa4sMPP8TVq1ffua89e/bA1dUVwcHBOHfuHO7evfvW/l27dsX06dNl6xkZGejTpw8MDAzg6OiIHTt2wMHBQe60lUQiwcaNGzFw4EAYGhqiSZMm+Omnn2Sfx8fHQyKR4OjRo/Dw8ICBgQE+/PBDZGVl4fDhw3BxcYGpqSk++eQTvHz5UradIAhYsmQJnJycYGBggJYtW2Lv3r2lxj1x4gQ8PT1haGiI9u3b4+bNmwCAqKgohIaG4urVq7KjZVFRUe/8ziqK19yomKrPMdf0c8hERDWJgYGB3BGaW7du4ccff0R0dLTsuSt9+vSBubk5YmNjYWZmhnXr1qFbt274448/YG5uXu7YkZGRGDFiBMzMzODr64vNmzcjNDRU4dr8/Pzw+PFjxMfHQ1dXFwEBAcjKyirVLzQ0FEuWLMHSpUuxatUqDB8+HGlpaXK1zZs3D99//z0MDQ0xZMgQDBkyBFKpFDt27MCLFy8wcOBArFq1CjNnzgQAfPnll4iJiUFERASaNGmCM2fOYMSIEbC0tESXLl1k44aEhODbb7+FpaUlJkyYgNGjR+PcuXMYOnQofv/9dxw5cgTHjx8HAJiZmSk8d2Ux3BAREQG4dOkSduzYgW7dusna8vPz8cMPP8DS0hIAcPLkSVy7dg1ZWVmQSqUAgGXLlmH//v3Yu3cvxo8fX+bYf/75Jy5evIiYmBgAwIgRIzB16lTMnTsXWlrvPoly48YNHD9+HJcvX4anpycAYOPGjWjSpEmpviNHjsQnn3wCAPjmm2+watUqXLp0Cb1795b1+frrr9GhQwcAwJgxYxAcHIzbt2/DyckJADB48GCcOnUKM2fORG5uLpYvX46TJ0/Cy8sLAODk5ISzZ89i3bp1cuFm4cKFsvVZs2ahT58+eP36NQwMDGBsbAwdHR3Y2Ni8c76VxdNSRERUax08eBDGxsbQ19eHl5cXOnfujFWrVsk+b9iwoSzYAMCVK1fw4sULWFhYwNjYWLakpqbi9u3b5e4nMjISvXr1Qr169QAAvr6+yM3NlR3FeJebN29CR0cHrVu3lrU1btwYdevWLdXX3d1d9mcjIyOYmJiUOsLzzz7W1tYwNDSUBZuStpJtrl+/jtevX6NHjx5yc966dWupOf9zXFtbWwAo8+iSuvHIDRER1Vre3t6IiIiArq4u7OzsoKurK/e5kZGR3HpxcTFsbW0RHx9faqw6deqUuY+ioiJs3boVmZmZ0NHRkWuPjIxEz54931lneS8kLav933OQSCQoLi4ut0/J3WLlbVPyv4cOHcJ7770n16/k6FV54/5z+6rEcENERLWWkZERGjdurHD/1q1by0KKg4ODQtvExsbi+fPnSExMlHtf0o0bNzB8+HA8efIEFhYWbx2jWbNmKCwsRGJiItq0aQPgzfVAz549U7j2inJ1dYVUKkV6errcKShl6enpoaioSIWVlY/hhoiISEHdu3eHl5cXBgwYgMWLF8PZ2RkPHjxAbGwsBgwYILse5p8iIyPRp08ftGzZUq7dzc0N06dPx7Zt2zBt2rS37rdZs2bo3r07xo8fLzvS9N///hcGBgZqf6eXiYkJAgMDMWPGDBQXF6Njx47IycnB+fPnYWxsDH9/f4XGcXBwQGpqKpKSklC/fn2YmJiUOvKjKgw3RESkFmK821MikSA2NhYhISEYPXo0Hj16BBsbG3Tu3BnW1tal+j98+BCHDh3Cjh07yhzro48+QmRk5DvDDQBs3boVY8aMQefOnWFjY4OwsDAkJydDX19fJXN7mwULFsDKygphYWG4c+cO6tSpg9atW2P27NkKjzFo0CDExMTA29sbz549w+bNmzFy5Ei11CsRyjuRVwUiIiIQEREhu9ffzc0Nc+bMgY+PT5n94+Pj4e3tXao9JSUFzZo1U2ifOTk5MDMzQ3Z2NkxNTStce3l4KzgR1TavX79GamoqHB0dq+QXLb3x119/wd7eHsePH5e7w6sme9vPkjK/vzV65KZ+/fpYtGiR7Hznli1b0L9/fyQmJsLNza3c7W7evCk3sX9eyU5ERCRGJ0+exIsXL9CiRQtkZGQgKCgIDg4O6Ny5s6ZLq3Y0Gm769esnt75w4UJERETg4sWLbw03VlZW5V6VTkREJEYFBQWYPXs27ty5AxMTE7Rv3x7bt28vdacTVaNrboqKirBnzx7k5ubKHhJUHg8PD7x+/Rqurq748ssvyzxVVSIvLw95eXmy9ZycHJXVTEREVFV69eqFXr16abqMGkHjD/G7du0ajI2NIZVKMWHCBOzbtw+urq5l9rW1tcX69esRHR2NmJgYODs7o1u3bjhz5ky544eFhcHMzEy22Nvbq2sqREREVA1o9IJi4M2jrdPT0/Hs2TNER0dj48aNOH36dLkB59/69esHiUQi92KwfyrryI29vT0vKCYiUhFeUEyqIooLioE3D/UpuaDY09MTly9fxooVK7Bu3TqFtm/Xrh22bdtW7udSqVRt99ETERFR9aPx01L/JgiC3JGWd0lMTJS9v4KIiIhIo0duZs+eDR8fH9jb2+P58+fYtWsX4uPjceTIEQBAcHAw7t+/j61btwIAwsPD4eDgADc3N+Tn52Pbtm2Ijo5GdHS0JqdBRERE1YhGw83Dhw/x6aefIiMjA2ZmZnB3d8eRI0fQo0cPAEBGRgbS09Nl/fPz8xEYGIj79+/DwMAAbm5uOHToEHx9fTU1BSIiIqpmNBpuIiMj3/p5VFSU3HpQUBCCgoLUWBEREanMPLMq3l921e6vCnXt2hWtWrVCeHg4gDfvaZo+fTqmT5+u0bqqWy0lqt01N0RERFVh5MiRkEgkkEgk0NXVhZOTEwIDA5Gbm1vpse/evQuJRIKkpCSFt+nZsye0tbVx8eLFd/a9fPkyxo8fX4kKxY3hhoiIaq3evXsjIyMDd+7cwddff401a9YgMDCwyutIT0/HhQsXMHny5Hee1QDevHbI0NCwCiqrmRhuiIio1pJKpbCxsYG9vT2GDRuG4cOHY//+/QDePCdt6tSpsLKygr6+Pjp27IjLly/Ltv37778xfPhwWFpawsDAAE2aNMHmzZsBAI6OjgDePFFfIpGga9eub61j8+bN6Nu3Lz7//HPs3r37nUePHBwcZKeoAODGjRvo2LEj9PX14erqiuPHj0MikcjmUnIkqeSt3IaGhmjZsiUuXLggN+758+fRuXNnGBgYwN7eHlOnTpWrJSsrC/369YOBgQEcHR2xffv2t9apKQw3RERE/x8DAwMUFBQAeHOdZ3R0NLZs2YJff/0VjRs3Rq9evfD06VMAwFdffYXr16/j8OHDSElJQUREBOrVqwcAuHTpEgDg+PHjyMjIQExMTLn7FAQBmzdvxogRI9CsWTM0bdoUP/74o8I1FxcXY8CAATA0NMQvv/yC9evXIyQkpMy+ISEhCAwMRFJSEpo2bYpPPvkEhYWFAN68MaBXr1746KOP8Ntvv2H37t04e/YsJk+eLNt+5MiRuHv3Lk6ePIm9e/dizZo1yMrKUrjWqqLxh/gRERFVB5cuXcKOHTvQrVs35ObmIiIiAlFRUfDx8QEAbNiwAXFxcYiMjMQXX3yB9PR0eHh4wNPTE8CboyklLC0tAQAWFhawsbF5636PHz+Oly9fyt4bNWLECERGRmLUqFEK1X3s2DHcvn0b8fHxsn0tXLhQdufxPwUGBqJPnzdPvg8NDYWbmxtu3bqFZs2aYenSpRg2bJjswuAmTZpg5cqV6NKlCyIiIpCeno7Dhw/j4sWL+OCDDwC8uTHIxcVFoTqrEo/cEBFRrXXw4EEYGxtDX18fXl5e6Ny5M1atWoXbt2+joKAAHTp0kPXV1dXF+++/j5SUFADA559/jl27dqFVq1YICgrC+fPnK1RDZGQkhg4dCh2dN8cbPvnkE/zyyy+4efOmQtvfvHkT9vb2ciHq/fffL7Ovu7u77M8lD8AtOfJy5coVREVFwdjYWLb06tULxcXFSE1NRUpKCnR0dGRhDgCaNWuGOnXqKDXfqsBwQ0REtZa3tzeSkpJw8+ZNvH79GjExMbCyskLJaxclEolcf0EQZG0+Pj5IS0vD9OnT8eDBA3Tr1k3pi5GfPn2K/fv3Y82aNdDR0YGOjg7ee+89FBYWYtOmTQqN8c+a3kVXV1f255JtiouLZf/72WefISkpSbZcvXoVf/75Jxo1alTud1IdMdwQEVGtZWRkhMaNG6Nhw4Zyv/gbN24MPT09nD17VtZWUFCAhIQEudMwlpaWGDlyJLZt24bw8HCsX78ewJv3JgJAUVHRW/e/fft21K9fH1evXpULFeHh4diyZYvsepi3adasGdLT0/Hw4UNZ2z8vfFZU69atkZycjMaNG5da9PT04OLigsLCQiQkJMi2uXnzJp49e6b0vtSN19wQERH9i5GRET7//HN88cUXMDc3R4MGDbBkyRK8fPkSY8aMAQDMmTMHbdq0gZubG/Ly8nDw4EFZ8LGysoKBgQGOHDmC+vXrQ19fH2ZmpR9qGBkZicGDB6N58+Zy7Q0bNsTMmTNx6NAh9O/f/6219ujRA40aNYK/vz+WLFmC58+fyy4oVuYoy8yZM9GuXTtMmjQJ48aNg5GREVJSUhAXF4dVq1bB2dkZvXv3xrhx47B+/Xro6Ohg+vTpMDAwUHgfVYXhhoiI1KOGPzF40aJFKC4uxqeffornz5/D09MTR48eRd26dQG8OToTHByMu3fvwsDAAJ06dcKuXbsAADo6Oli5ciXmz5+POXPmoFOnToiPj5cb/8qVK7h69So2bNhQat8mJibo2bMnIiMj3xlutLW1sX//fowdOxZt27aFk5MTli5din79+kFfX1/h+bq7u+P06dMICQlBp06dIAgCGjVqhKFDh8r6bN68GWPHjkWXLl1gbW2Nr7/+Gl999ZXC+6gqEqHkJFotkZOTAzMzM2RnZ8PU1FTl4zvMOqTS8e4u6qPS8YiIVO3169dITU2Fo6OjUr9MSX3OnTuHjh074tatW2jUqJGmy1HY236WlPn9zSM3RERENdy+fftgbGyMJk2a4NatW5g2bRo6dOhQo4KNKjHcEBER1XDPnz9HUFAQ7t27h3r16qF79+749ttvNV2WxjDcEBER1XB+fn7w8/PTdBnVBm8FJyIiIlFhuCEiIpWoZfenkBqo6meI4YaIiCpFW1sbAJCfn6/hSqimK/kZKvmZqihec0NERJWio6MDQ0NDPHr0CLq6utDS4n83k/KKi4vx6NEjGBoayt6zVVEMN0REVCkSiQS2trZITU1FWlqapsuhGkxLSwsNGjSo9PurGG6IiKjS9PT00KRJE56aokrR09NTyZE/hhsiIlIJLS0tPqGYqgWeGCUiIiJRYbghIiIiUWG4ISIiIlFhuCEiIiJRYbghIiIiUWG4ISIiIlFhuCEiIiJRYbghIiIiUWG4ISIiIlFhuCEiIiJRYbghIiIiUWG4ISIiIlFhuCEiIiJRYbghIiIiUWG4ISIiIlFhuCEiIiJRYbghIiIiUdFouImIiIC7uztMTU1hamoKLy8vHD58+K3bnD59Gm3atIG+vj6cnJywdu3aKqqWiIiIagKNhpv69etj0aJFSEhIQEJCAj788EP0798fycnJZfZPTU2Fr68vOnXqhMTERMyePRtTp05FdHR0FVdORERE1ZWOJnfer18/ufWFCxciIiICFy9ehJubW6n+a9euRYMGDRAeHg4AcHFxQUJCApYtW4ZBgwaVuY+8vDzk5eXJ1nNyclQ3ASIiIqp2qs01N0VFRdi1axdyc3Ph5eVVZp8LFy6gZ8+ecm29evVCQkICCgoKytwmLCwMZmZmssXe3l7ltRMREVH1ofFwc+3aNRgbG0MqlWLChAnYt28fXF1dy+ybmZkJa2truTZra2sUFhbi8ePHZW4THByM7Oxs2XLv3j2Vz4GIiIiqD42elgIAZ2dnJCUl4dmzZ4iOjoa/vz9Onz5dbsCRSCRy64IglNleQiqVQiqVqrZoIiIiqrY0Hm709PTQuHFjAICnpycuX76MFStWYN26daX62tjYIDMzU64tKysLOjo6sLCwqJJ6iYiIqHrT+GmpfxMEQe4C4H/y8vJCXFycXNuxY8fg6ekJXV3dqiiPiIiIqjmNhpvZs2fj559/xt27d3Ht2jWEhIQgPj4ew4cPB/Dmehk/Pz9Z/wkTJiAtLQ0BAQFISUnBpk2bEBkZicDAQE1NgYiIiKoZjZ6WevjwIT799FNkZGTAzMwM7u7uOHLkCHr06AEAyMjIQHp6uqy/o6MjYmNjMWPGDKxevRp2dnZYuXJlubeBExERUe0jEUquyK0lcnJyYGZmhuzsbJiamqp8fIdZh1Q63t1FfVQ6HhERUU2kzO/vanfNDREREVFlMNwQERGRqDDcEBERkagw3BAREZGoMNwQERGRqDDcEBERkagw3BAREZGoMNwQERGRqDDcEBERkagw3BAREZGoMNwQERGRqDDcEBERkagw3BAREZGoMNwQERGRqDDcEBERkagw3BAREZGoMNwQERGRqDDcEBERkagw3BAREZGoMNwQERGRqDDcEBERkagw3BAREZGoMNwQERGRqDDcEBERkagw3BAREZGoMNwQERGRqDDcEBERkagw3BAREZGoMNwQERGRqDDcEBERkagw3BAREZGoMNwQERGRqOhougB6h3lmahgzW/VjEhERVRM8ckNERESiwnBDREREosJwQ0RERKLCcENERESiotFwExYWhrZt28LExARWVlYYMGAAbt68+dZt4uPjIZFISi03btyooqqJiIioOtNouDl9+jQmTZqEixcvIi4uDoWFhejZsydyc3Pfue3NmzeRkZEhW5o0aVIFFRMREVF1p9FbwY8cOSK3vnnzZlhZWeHKlSvo3LnzW7e1srJCnTp11FgdERER1UTV6pqb7Ow3z18xNzd/Z18PDw/Y2tqiW7duOHXqVLn98vLykJOTI7cQERGReFWbcCMIAgICAtCxY0c0b9683H62trZYv349oqOjERMTA2dnZ3Tr1g1nzpwps39YWBjMzMxki729vbqmQERERNWARBAEQZkN7t27B4lEgvr16wMALl26hB07dsDV1RXjx4+vcCGTJk3CoUOHcPbsWdnYiurXrx8kEgl++umnUp/l5eUhLy9Ptp6TkwN7e3tkZ2fD1NS0wvWWx2HWIZWOd1d/mErHA8AnFBMRUY2Tk5MDMzMzhX5/K33NzbBhwzB+/Hh8+umnyMzMRI8ePeDm5oZt27YhMzMTc+bMUbrgKVOm4KeffsKZM2eUDjYA0K5dO2zbtq3Mz6RSKaRSqdJjEpVQdWAFgLuL+qh8TCIiekPp01K///473n//fQDAjz/+iObNm+P8+fPYsWMHoqKilBpLEARMnjwZMTExOHnyJBwdHZUtBwCQmJgIW1vbCm1LRERE4qL0kZuCggLZkZDjx4/jP//5DwCgWbNmyMjIUGqsSZMmYceOHfjf//4HExMTZGZmAgDMzMxgYGAAAAgODsb9+/exdetWAEB4eDgcHBzg5uaG/Px8bNu2DdHR0YiOjlZ2KkRERCRCSocbNzc3rF27Fn369EFcXBwWLFgAAHjw4AEsLCyUGisiIgIA0LVrV7n2zZs3Y+TIkQCAjIwMpKenyz7Lz89HYGAg7t+/DwMDA7i5ueHQoUPw9fVVdipEREQkQkqHm8WLF2PgwIFYunQp/P390bJlSwDATz/9JDtdpShFrmX+96muoKAgBAUFKbUfIiIiqj2UDjddu3bF48ePkZOTg7p168rax48fD0NDQ5UWR0SkDrxInEjcKvSEYm1tbblgAwAODg6qqIeIiIioUhQKNx4eHpBIJAoN+Ouvv1aqICIiIqLKUCjcDBgwQPbn169fY82aNXB1dYWXlxcA4OLFi0hOTsbEiRPVUiQRERGRohQKN3PnzpX9eezYsZg6darsLql/9rl3755qqyMiIiJSktIP8duzZw/8/PxKtY8YMYLPmiEiIiKNUzrcGBgY4OzZs6Xaz549C319fZUURURERFRRSt8tNX36dHz++ee4cuUK2rVrB+DNNTebNm2q0HuliIiIiFRJ6XAza9YsODk5YcWKFdixYwcAwMXFBVFRURgyZIjKCyQiIiJSRoWeczNkyBAGGSIiIqqWlL7mhoiIiKg6U/rITVFREb777jv8+OOPSE9PR35+vtznT58+VVlxRERERMpS+shNaGgoli9fjiFDhiA7OxsBAQH46KOPoKWlhXnz5qmhRCIiIiLFKR1utm/fjg0bNiAwMBA6Ojr45JNPsHHjRsyZMwcXL15UR41EREREClM63GRmZqJFixYAAGNjY2RnZwMA+vbti0OHVP+mXSIiIiJlKB1u6tevj4yMDABA48aNcezYMQDA5cuXIZVKVVsdERERkZKUDjcDBw7EiRMnAADTpk3DV199hSZNmsDPzw+jR49WeYFEREREylD6bqlFixbJ/jx48GDUr18f58+fR+PGjfGf//xHpcURERERKatCD/H7p3bt2slew0BERESkaRV6iN8PP/yADh06wM7ODmlpaQCA8PBw/O9//1NpcURERETKUjrcREREICAgAL6+vnj27BmKiooAAHXq1EF4eLiq6yMiIiJSitLhZtWqVdiwYQNCQkKgra0ta/f09MS1a9dUWhwRERGRspQON6mpqfDw8CjVLpVKkZubq5KiiIiIiCpK6XDj6OiIpKSkUu2HDx+Gq6urKmoiIiIiqjCl75b64osvMGnSJLx+/RqCIODSpUvYuXMnwsLCsHHjRnXUSERERKQwpcPNqFGjUFhYiKCgILx8+RLDhg3De++9hxUrVuDjjz9WR41ERERECqvQc27GjRuHcePG4fHjxyguLoaVlRUA4P79+3jvvfdUWiARERGRMir0nJsS9erVg5WVFTIzMzFlyhQ0btxYVXURERERVYjC4ebZs2cYPnw4LC0tYWdnh5UrV6K4uBhz5syBk5MTLl68iE2bNqmzViIiIqJ3Uvi01OzZs3HmzBn4+/vjyJEjmDFjBo4cOYLXr1/j8OHD6NKlizrrJCIiIlKIwuHm0KFD2Lx5M7p3746JEyeicePGaNq0KZ9KTERERNWKwqelHjx4IHuOjZOTE/T19TF27Fi1FUZERERUEQqHm+LiYujq6srWtbW1YWRkpJaiiIiIiCpK4dNSgiBg5MiRkEqlAIDXr19jwoQJpQJOTEyMaiskIiIiUoLC4cbf319ufcSIESovhoiIiKiyFA43mzdvVmcdRERERCpRqYf4EREREVU3DDdEREQkKhoNN2FhYWjbti1MTExgZWWFAQMG4ObNm+/c7vTp02jTpg309fXh5OSEtWvXVkG1REREVBNU6MWZqnL69GlMmjQJbdu2RWFhIUJCQtCzZ09cv3693NvMU1NT4evri3HjxmHbtm04d+4cJk6cCEtLSwwaNKiKZ0BEVL05zDqk0vHuLuqj0vGI1EGhcNO6dWucOHECdevWxfz58xEYGAhDQ8NK7/zIkSNy65s3b4aVlRWuXLmCzp07l7nN2rVr0aBBA9mTkV1cXJCQkIBly5Yx3BAREZFip6VSUlKQm5sLAAgNDcWLFy/UUkx2djYAwNzcvNw+Fy5cQM+ePeXaevXqhYSEBBQUFJTqn5eXh5ycHLmFiIiIxEuhIzetWrXCqFGj0LFjRwiCgGXLlsHY2LjMvnPmzKlQIYIgICAgAB07dkTz5s3L7ZeZmQlra2u5NmtraxQWFuLx48ewtbWV+ywsLAyhoaEVqomIiIhqHoXCTVRUFObOnYuDBw9CIpHg8OHD0NEpvalEIqlwuJk8eTJ+++03nD179p19JRKJ3LogCGW2A0BwcDACAgJk6zk5ObC3t69QjURERFT9KRRunJ2dsWvXLgCAlpYWTpw4ASsrK5UVMWXKFPz00084c+YM6tev/9a+NjY2yMzMlGvLysqCjo4OLCwsSvWXSqWyV0YQERGR+Cl9t1RxcbHKdi4IAqZMmYJ9+/YhPj4ejo6O79zGy8sLBw4ckGs7duwYPD095V7sSURERLVThZ5zc/v2bUyZMgXdu3dHjx49MHXqVNy+fVvpcSZNmoRt27Zhx44dMDExQWZmJjIzM/Hq1StZn+DgYPj5+cnWJ0yYgLS0NAQEBCAlJQWbNm1CZGQkAgMDKzIVIiIiEhmlw83Ro0fh6uqKS5cuwd3dHc2bN8cvv/wCNzc3xMXFKTVWREQEsrOz0bVrV9ja2sqW3bt3y/pkZGQgPT1dtu7o6IjY2FjEx8ejVatWWLBgAVauXMnbwImIiAhABU5LzZo1CzNmzMCiRYtKtc+cORM9evRQeKySC4HfJioqqlRbly5d8Ouvvyq8HyIiIqo9lD5yk5KSgjFjxpRqHz16NK5fv66SooiIiIgqSulwY2lpiaSkpFLtSUlJKr2DioiIiKgilD4tNW7cOIwfPx537txB+/btIZFIcPbsWSxevBj//e9/1VEjERERkcKUDjdfffUVTExM8O233yI4OBgAYGdnh3nz5mHq1KkqL5CIiIhIGUqHG4lEghkzZmDGjBl4/vw5AMDExETlhREREVFpfNP7uykdbv6JoYaIiIiqmwo9xI+IiIioumK4ISIiIlFhuCEiIiJRUSrcFBQUwNvbG3/88Ye66iEiIiKqFKXCja6uLn7//XdIJBJ11UNERERUKUqflvLz80NkZKQ6aiEiIiKqNKVvBc/Pz8fGjRsRFxcHT09PGBkZyX2+fPlylRVHREREpCylw83vv/+O1q1bA0Cpa294uoqIiIg0Telwc+rUKXXUQURERKQSFb4V/NatWzh69ChevXoFABAEQWVFEREREVWU0kdunjx5giFDhuDUqVOQSCT4888/4eTkhLFjx6JOnTr49ttv1VEnEVH1Ns9MxeNlq3Y8olpE6SM3M2bMgK6uLtLT02FoaChrHzp0KI4cOaLS4oiIiIiUpfSRm2PHjuHo0aOoX7++XHuTJk2QlpamssKIqOrxbcNEJAZKH7nJzc2VO2JT4vHjx5BKpSopioiIiKiilA43nTt3xtatW2XrEokExcXFWLp0Kby9vVVaHBEREZGylD4ttXTpUnTt2hUJCQnIz89HUFAQkpOT8fTpU5w7d04dNRIREREpTOkjN66urvjtt9/w/vvvo0ePHsjNzcVHH32ExMRENGrUSB01EhERESlM6SM3AGBjY4PQ0FBV10JERERUaRUKN3///TciIyORkpICiUQCFxcXjBo1Cubm5qquj4iIiEgpSp+WOn36NBwdHbFy5Ur8/fffePr0KVauXAlHR0ecPn1aHTUSERERKUzpIzeTJk3CkCFDEBERAW1tbQBAUVERJk6ciEmTJuH3339XeZFEREREilL6yM3t27fx3//+VxZsAEBbWxsBAQG4ffu2SosjIiIiUpbS4aZ169ZISUkp1Z6SkoJWrVqpoiYiIiKiClPotNRvv/0m+/PUqVMxbdo03Lp1C+3atQMAXLx4EatXr8aiRYvUUyWR2PAli0REaqNQuGnVqhUkEgkEQZC1BQUFleo3bNgwDB06VHXVERERESlJoXCTmpqq7jqIiIiIVEKhcNOwYUN110FERESkEhV6iN/9+/dx7tw5ZGVlobi4WO6zqVOnqqQwIiIioopQOtxs3rwZEyZMgJ6eHiwsLCCRSGSfSSQShhsiIiLSKKXDzZw5czBnzhwEBwdDS0vpO8mJiIiI1ErpdPLy5Ut8/PHHDDZERERULSmdUMaMGYM9e/aooxYiIiKiSlP6tFRYWBj69u2LI0eOoEWLFtDV1ZX7fPny5QqPdebMGSxduhRXrlxBRkYG9u3bhwEDBpTbPz4+Ht7e3qXaU1JS0KxZM4X3S0REROKldLj55ptvcPToUTg7OwNAqQuKlZGbm4uWLVti1KhRGDRokMLb3bx5E6amprJ1S0tLpfZLRERE4qV0uFm+fDk2bdqEkSNHVnrnPj4+8PHxUXo7Kysr1KlTp9L7JyIiIvFR+pobqVSKDh06qKMWhXl4eMDW1hbdunXDqVOn3to3Ly8POTk5cgsRERGJl9LhZtq0aVi1apU6anknW1tbrF+/HtHR0YiJiYGzszO6deuGM2fOlLtNWFgYzMzMZIu9vX0VVkxERERVTenTUpcuXcLJkydx8OBBuLm5lbqgOCYmRmXF/Zuzs7PsWh8A8PLywr1797Bs2TJ07ty5zG2Cg4MREBAgW8/JyWHAISIiEjGlw02dOnXw0UcfqaOWCmnXrh22bdtW7udSqRRSqbQKKyIiIiJNqtDrF6qTxMRE2NraaroMIiIiqiYq9OJMVXnx4gVu3bolW09NTUVSUhLMzc3RoEEDBAcH4/79+9i6dSsAIDw8HA4ODnBzc0N+fj62bduG6OhoREdHa2oKREREVM0oHW4cHR3f+jybO3fuKDxWQkKC3EP5Sq6N8ff3R1RUFDIyMpCeni77PD8/H4GBgbh//z4MDAzg5uaGQ4cOwdfXV9lpEBERkUgpHW6mT58ut15QUIDExEQcOXIEX3zxhVJjde3aFYIglPt5VFSU3HpQUBCCgoKU2gcREYmfw6xDKh/z7qI+Kh+TqobS4WbatGlltq9evRoJCQmVLoiIiIioMlR2zY2Pjw+Cg4Or3QXHRERE9BbzzNQwZrbqx1SC0g/xK8/evXthbm6uquGIiIiIKkTpIzceHh5yFxQLgoDMzEw8evQIa9asUWlxRERERMpSOtwMGDBAbl1LSwuWlpbo2rUrmjVrpqq6iIiIiCpE6XAzd+5cddRBREREpBIqu+aGiIiIqDpQ+MiNlpbWWx/eBwASiQSFhYWVLoqIiIioohQON/v27Sv3s/Pnz2PVqlVvfSAfERERUVVQONz079+/VNuNGzcQHByMAwcOYPjw4ViwYIFKiyMiIiJSVoWuuXnw4AHGjRsHd3d3FBYWIikpCVu2bEGDBg1UXR8RERGRUpQKN9nZ2Zg5cyYaN26M5ORknDhxAgcOHEDz5s3VVR8RERGRUhQ+LbVkyRIsXrwYNjY22LlzZ5mnqYiIiIg0TeFwM2vWLBgYGKBx48bYsmULtmzZUma/mJgYlRVHRETVjAjfQ0Tio3C48fPze+et4ERERESapnC4iYqKUmMZRERERKrBJxQTERGRqDDcEBERkagw3BAREZGoMNwQERGRqDDcEBERkagw3BAREZGoMNwQERGRqDDcEBERkago/BA/IiKl8VH9RKQBPHJDREREosJwQ0RERKLCcENERESiwnBDREREosJwQ0RERKLCcENERESiwnBDREREosLn3FCFOcw6pNLx7i7qo9LxiIioduKRGyIiIhIVhhsiIiISFYYbIiIiEhWGGyIiIhIVhhsiIiISFY2GmzNnzqBfv36ws7ODRCLB/v3737nN6dOn0aZNG+jr68PJyQlr165Vf6FERERUY2g03OTm5qJly5b4/vvvFeqfmpoKX19fdOrUCYmJiZg9ezamTp2K6OhoNVdKRERENYVGn3Pj4+MDHx8fhfuvXbsWDRo0QHh4OADAxcUFCQkJWLZsGQYNGqSmKomIiKgmqVHX3Fy4cAE9e/aUa+vVqxcSEhJQUFBQ5jZ5eXnIycmRW4iIiEi8alS4yczMhLW1tVybtbU1CgsL8fjx4zK3CQsLg5mZmWyxt7evilKJiIhIQ2pUuAEAiUQity4IQpntJYKDg5GdnS1b7t27p/YaiYiISHNq1LulbGxskJmZKdeWlZUFHR0dWFhYlLmNVCqFVCqtivKIiIioGqhR4cbLywsHDhyQazt27Bg8PT2hq6uroaqIiEiU5pmpeLxs1Y5H5dLoaakXL14gKSkJSUlJAN7c6p2UlIT09HQAb04p+fn5yfpPmDABaWlpCAgIQEpKCjZt2oTIyEgEBgZqonwiIiKqhjR65CYhIQHe3t6y9YCAAACAv78/oqKikJGRIQs6AODo6IjY2FjMmDEDq1evhp2dHVauXMnbwImIiEhGo+Gma9eusguCyxIVFVWqrUuXLvj111/VWBURERHVZDXubikiIiKit2G4ISIiIlFhuCEiIiJRYbghIiIiUWG4ISIiIlFhuCEiIiJRYbghIiIiUWG4ISIiIlFhuCEiIiJRYbghIiIiUalRbwUnkeMbeImISAV45IaIiIhEheGGiIiIRIXhhoiIiESF4YaIiIhEheGGiIiIRIXhhoiIiESF4YaIiIhEheGGiIiIRIXhhoiIiESF4YaIiIhEheGGiIiIRIXhhoiIiESF4YaIiIhEheGGiIiIRIXhhoiIiESF4YaIiIhEheGGiIiIRIXhhoiIiESF4YaIiIhEheGGiIiIRIXhhoiIiESF4YaIiIhEheGGiIiIRIXhhoiIiESF4YaIiIhEheGGiIiIRIXhhoiIiERF4+FmzZo1cHR0hL6+Ptq0aYOff/653L7x8fGQSCSllhs3blRhxURERFSdaTTc7N69G9OnT0dISAgSExPRqVMn+Pj4ID09/a3b3bx5ExkZGbKlSZMmVVQxERERVXcaDTfLly/HmDFjMHbsWLi4uCA8PBz29vaIiIh463ZWVlawsbGRLdra2uX2zcvLQ05OjtxCRERE4qWxcJOfn48rV66gZ8+ecu09e/bE+fPn37qth4cHbG1t0a1bN5w6deqtfcPCwmBmZiZb7O3tK107ERERVV8aCzePHz9GUVERrK2t5dqtra2RmZlZ5ja2trZYv349oqOjERMTA2dnZ3Tr1g1nzpwpdz/BwcHIzs6WLffu3VPpPIiIiKh60dF0ARKJRG5dEIRSbSWcnZ3h7OwsW/fy8sK9e/ewbNkydO7cucxtpFIppFKp6gomIiKiak1jR27q1asHbW3tUkdpsrKySh3NeZt27drhzz//VHV5REREVENpLNzo6emhTZs2iIuLk2uPi4tD+/btFR4nMTERtra2qi6PiIiIaiiNnpYKCAjAp59+Ck9PT3h5eWH9+vVIT0/HhAkTALy5Xub+/fvYunUrACA8PBwODg5wc3NDfn4+tm3bhujoaERHR2tyGkRERFSNaDTcDB06FE+ePMH8+fORkZGB5s2bIzY2Fg0bNgQAZGRkyD3zJj8/H4GBgbh//z4MDAzg5uaGQ4cOwdfXV1NTICIiompG4xcUT5w4ERMnTizzs6ioKLn1oKAgBAUFVUFVREREVFNp/PULRERERKrEcENERESiwnBDREREosJwQ0RERKLCcENERESiwnBDREREosJwQ0RERKLCcENERESiwnBDREREosJwQ0RERKLCcENERESiwnBDREREosJwQ0RERKLCcENERESiwnBDREREosJwQ0RERKLCcENERESiwnBDREREosJwQ0RERKLCcENERESiwnBDREREosJwQ0RERKLCcENERESiwnBDREREosJwQ0RERKLCcENERESiwnBDREREosJwQ0RERKLCcENERESiwnBDREREosJwQ0RERKLCcENERESiwnBDREREosJwQ0RERKLCcENERESiwnBDREREosJwQ0RERKKi8XCzZs0aODo6Ql9fH23atMHPP//81v6nT59GmzZtoK+vDycnJ6xdu7aKKiUiIqKaQKPhZvfu3Zg+fTpCQkKQmJiITp06wcfHB+np6WX2T01Nha+vLzp16oTExETMnj0bU6dORXR0dBVXTkRERNWVRsPN8uXLMWbMGIwdOxYuLi4IDw+Hvb09IiIiyuy/du1aNGjQAOHh4XBxccHYsWMxevRoLFu2rIorJyIioupKR1M7zs/Px5UrVzBr1iy59p49e+L8+fNlbnPhwgX07NlTrq1Xr16IjIxEQUEBdHV1S22Tl5eHvLw82Xp2djYAICcnp7JTKFNx3kuVjpcjEVQ63ptBVTP3aj/XajpPoPbMtbr+/PLvtOL4d1qZAfl3Wrkh34wpCArUK2jI/fv3BQDCuXPn5NoXLlwoNG3atMxtmjRpIixcuFCu7dy5cwIA4cGDB2VuM3fuXAEAFy5cuHDhwkUEy717996ZMTR25KaERCKRWxcEoVTbu/qX1V4iODgYAQEBsvXi4mI8ffoUFhYWb91PdZCTkwN7e3vcu3cPpqammi5HrThX8akt8wRqz1xryzyB2jPXmjRPQRDw/Plz2NnZvbOvxsJNvXr1oK2tjczMTLn2rKwsWFtbl7mNjY1Nmf11dHRgYWFR5jZSqRRSqVSurU6dOhUvXANMTU2r/Q+dqnCu4lNb5gnUnrnWlnkCtWeuNWWeZmZmCvXT2AXFenp6aNOmDeLi4uTa4+Li0L59+zK38fLyKtX/2LFj8PT0LPN6GyIiIqp9NHq3VEBAADZu3IhNmzYhJSUFM2bMQHp6OiZMmADgzSklPz8/Wf8JEyYgLS0NAQEBSElJwaZNmxAZGYnAwEBNTYGIiIiqGY1eczN06FA8efIE8+fPR0ZGBpo3b47Y2Fg0bNgQAJCRkSH3zBtHR0fExsZixowZWL16Nezs7LBy5UoMGjRIU1NQK6lUirlz55Y6rSZGnKv41JZ5ArVnrrVlnkDtmatY5ykRBEXuqSIiIiKqGTT++gUiIiIiVWK4ISIiIlFhuCEiIiJRYbghIiIiUWG4ISIiIlFhuKlm8vLy0KpVK0gkEiQlJSm83bRp09CmTRtIpVK0atVKbfWpSkXmefXqVXzyySewt7eHgYEBXFxcsGLFCvUWqgIVmeuTJ0/Qu3dv2NnZQSqVwt7eHpMnT1bbC19VoaI/uyWePHmC+vXrQyKR4NmzZyqvT5MyMjIwbNgwODs7Q0tLC9OnT9d0SWoTExODHj16wNLSEqampvDy8sLRo0c1XZbKnT17Fh06dICFhQUMDAzQrFkzfPfdd5ouS+3OnTsHHR2dav97huGmmgkKClLovRn/JggCRo8ejaFDh6qhKtWryDyvXLkCS0tLbNu2DcnJyQgJCUFwcDC+//57NVWpGhWZq5aWFvr374+ffvoJf/zxB6KionD8+HHZAy6ro4r+7JYYM2YM3N3dVVhR9ZGXlwdLS0uEhISgZcuWmi5Hrc6cOYMePXogNjYWV65cgbe3N/r164fExERNl6ZSRkZGmDx5Ms6cOYOUlBR8+eWX+PLLL7F+/XpNl6Y22dnZ8PPzQ7du3TRdyru989WapBKHDx8WOnToIJiZmQnm5uZCnz59hFu3bsn1iY2NFZo1ayYkJycLAITExESl9zN37lyhZcuWqim6AqpqniUmTpwoeHt7V7Lqiqnqua5YsUKoX79+JatWXlXMc82aNUKXLl2EEydOCACEv//+W3UTUFCXLl2EKVOmCF988YVQt25dwdraWpg7d67s87S0NOE///mPYGRkJJiYmAj/93//J2RmZlZoP9OmTVNd4RVQVXMt4erqKoSGhqqgcuVU9TwHDhwojBgxQgWVK68q5jp06FDhyy+/1PjvGUXwyE0Vyc3NRUBAAC5fvowTJ05AS0sLAwcORHFxMQDg4cOHGDduHH744QcYGhpquNqKq+p5Zmdnw9zcvNLjVERVzvXBgweIiYlBly5dVFG6UtQ9z+vXr2P+/PnYunUrtLQ0+39JW7ZsgZGREX755RcsWbIE8+fPR1xcHARBwIABA/D06VOcPn0acXFxuH37do05UlqWqpprcXExnj9/rrF/p1U1z8TERJw/f14j/0ZLqHOumzdvxu3btzF37lw1zkCFNJutaq+srCwBgHDt2jWhuLhY6N27t7BgwQJBEAQhNTW1xh65+Td1zVMQBOH8+fOCrq6ucOzYMRVWXHHqmOvHH38sGBgYCACEfv36Ca9evVJD5cpR5Txfv34tuLu7Cz/88IMgCIJw6tQpjR656dixo1xb27ZthZkzZwrHjh0TtLW1hfT0dNlnJUepLl26pPR+qsORm6qYqyAIwpIlSwRzc3Ph4cOHla5bWVUxz/fee0/Q09MTtLS0hPnz56usdmWpc65//PGHYGVlJdy8eVMQhOr3e6YsPHJTRW7fvo1hw4bByckJpqamcHR0BACkp6dj1apVyMnJQXBwsIarrLyqmmdycjL69++POXPmoEePHpUeryKqYq7fffcdfv31V+zfvx+3b99GQECAKkpXijrnGRwcDBcXF4wYMUKVJVfYv6/5sbW1RVZWFlJSUmBvbw97e3vZZ66urqhTpw5SUlKqukyVqIq57ty5E/PmzcPu3bthZWWlkrqVpe55/vzzz0hISMDatWsRHh6OnTt3qqx2ZaljrkVFRRg2bBhCQ0PRtGlTtdStDhp9cWZt0q9fP9jb22PDhg2ws7NDcXExmjdvjvz8fJw8eRIXL14s9eIyT09PDB8+HFu2bNFQ1cqrinlev34dH374IcaNG4cvv/xSHdNQSFXM1cbGBjY2NmjWrBksLCzQqVMnfPXVV7C1tVXHlMqkznmePHkS165dw969ewG8uTAeAOrVq4eQkBCEhoaqZ1Ll0NXVlVuXSCQoLi6GIAiQSCSl+pfXXhOoe667d+/GmDFjsGfPHnTv3r3S9VaUuudZEvZbtGiBhw8fYt68efjkk08qV3QFqWOuz58/R0JCAhITEzF58mQAkI2po6ODY8eO4cMPP1TdJFSE4aYKPHnyBCkpKVi3bh06deoE4M1thCVWrlyJr7/+Wrb+4MED9OrVC7t378YHH3xQ5fVWVFXMMzk5GR9++CH8/f2xcOFC1U5ACZr4Oy35xZ+Xl1eJypWj7nlGR0fj1atXsvXLly9j9OjR+Pnnn9GoUSMVzqRyXF1dkZ6ejnv37sn+6/f69evIzs6Gi4uLhqtTLVXMdefOnRg9ejR27tyJPn36qLPcClPH36kgCFX671NRlZmrqakprl27Jte2Zs0anDx5Env37pWFu+qG4aYK1K1bFxYWFli/fj1sbW2Rnp6OWbNmyT5v0KCBXH9jY2MAQKNGjVC/fn2F9nHr1i28ePECmZmZePXqlew5I66urtDT01PNRN5B3fNMTk6Gt7c3evbsiYCAAGRmZgIAtLW1YWlpqcKZvJu65xobG4uHDx+ibdu2MDY2xvXr1xEUFIQOHTrAwcFBpXN5G3XP898B5vHjxwAAFxcX1KlTp5LVq0737t3h7u6O4cOHIzw8HIWFhZg4cSK6dOkCT09PhcYo+Tf54sULPHr0CElJSdDT04Orq6saK1deZee6c+dO+Pn5YcWKFWjXrp3s36mBgQHMzMzUXb7CKjvP1atXo0GDBmjWrBmAN6F/2bJlmDJlirpLV1pl5qqlpYXmzZvLtVlZWUFfX79Ue3XCa26qgJaWFnbt2oUrV66gefPmmDFjBpYuXarSfYwdOxYeHh5Yt24d/vjjD3h4eMDDwwMPHjxQ6X7eRt3z3LNnDx49eoTt27fD1tZWtrRt21Zl+1CUuudqYGCADRs2oGPHjnBxccH06dPRt29fHDx4UGX7UERV/OzWBBKJBPv370fdunXRuXNndO/eHU5OTti9e7fCY5T8m7xy5Qp27NgBDw8P+Pr6qrHqiqnsXNetW4fCwkJMmjRJ7t/ptGnT1Fy5cio7z+LiYgQHB6NVq1bw9PTEqlWrsGjRIsyfP1/NlStPFT+/NY1EKDnWTURERCQCPHJDREREosJwUwNMmDABxsbGZS7V+XH8yqot8wRqz1xryzzd3NzKnef27ds1XZ5K1Za51pZ5AuKcK09L1QBZWVnlvjDR1NRUY8+PULXaMk+g9sy1tswzLS0NBQUFZX5mbW0NExOTKq5IfWrLXGvLPAFxzpXhhoiIiESFp6WIiIhIVBhuiIiISFQYboiIiEhUGG6IiIhIVBhuiIiISFQYboiIiEhUGG6IiIhIVP4fv753ssFdou0AAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "make_table(ins, outs, \"Pre Alignment\", \"Post Aligned\", \"Number of Reads\", samples, \"Alignment to Genome\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2af7b272",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Step: dedup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d57fa5ba-4e41-40ca-97c9-9c2553ff0d1d",
   "metadata": {},
   "source": [
    "#### Remove duplicate pcr reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "d6391aaa",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_path= mkpath(\"align\")\n",
    "out_path = mkpath(\"dedup\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68cf4a9c-5da0-448d-b0b4-82479f32a16f",
   "metadata": {},
   "source": [
    "#### umi_tools"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "058f7378-c4ca-4cc9-a778-7e07df98ceec",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "https://umi-tools.readthedocs.io/en/latest/index.html"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c399b23d-89e9-4187-a006-3ead12c827b1",
   "metadata": {},
   "source": [
    "pip install umi_tools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "bdeeb064-8c8b-4da5-8ee3-b28345777337",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "For full UMI-tools documentation, see: https://umi-tools.readthedocs.io/en/latest/\n",
      "\n",
      "\n",
      "umi_tools.py - Tools for UMI analyses\n",
      "=====================================\n",
      "\n",
      ":Author: Tom Smith & Ian Sudbury, CGAT\n",
      ":Tags: Genomics UMI\n",
      "\n",
      "There are 6 tools:\n",
      "\n",
      "  - whitelist\n",
      "  - extract\n",
      "  - group\n",
      "  - dedup\n",
      "  - count\n",
      "  - count_tab\n",
      "\n",
      "To get help on a specific tool, type:\n",
      "\n",
      "    umi_tools <tool> --help\n",
      "\n",
      "To use a specific tool, type::\n",
      "\n",
      "    umi_tools <tool> [tool options] [tool arguments]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "! umi_tools -h"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "82326ae9-28e6-4aa7-b28f-c56d061ade25",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "dedup - Deduplicate reads using UMI and mapping coordinates\n",
      "\n",
      "Usage: umi_tools dedup [OPTIONS] [--stdin=IN_BAM] [--stdout=OUT_BAM]\n",
      "\n",
      "       note: If --stdout is ommited, standard out is output. To\n",
      "             generate a valid BAM file on standard out, please\n",
      "             redirect log with --log=LOGFILE or --log2stderr \n",
      "\n",
      "For full UMI-tools documentation, see https://umi-tools.readthedocs.io/en/latest/\n",
      "\n",
      "Required options missing, see --help for more details\n"
     ]
    }
   ],
   "source": [
    "! umi_tools dedup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1cdeb9c-4bb3-4f94-b342-9ea841fbb743",
   "metadata": {},
   "source": [
    "##### paramaters\n",
    "- dedup, deduplicate reads using UMI and mapping coordinates\n",
    "- --method directional, specifies the algorithm used for selecting a representative read from a group of reads that share the same alignment position and UMI (after accounting for potential sequencing errors in the UMI).\n",
    "- -I {in_fn}, indexed BAM file for deduplication\n",
    "- --output-stats {stats_fn}, filename where the deduplication statistics will be saved\n",
    "- -L {log_fn}, filename where log output should be written\n",
    "- --temp-dir {temp_dir}, path to temporary directory\n",
    "- S {out_fn}, stdout for BAM file with deduplicated reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "d0c238ea",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for sample in samples:\n",
    "    !umi_tools dedup --method directional              \\\n",
    "        -I {fname(in_path,sample,\"bam\")}               \\\n",
    "        --output-stats={out_path}/dedup.stats          \\\n",
    "        -L {out_path}/umitools.log                     \\\n",
    "        --temp-dir={out_path}/                         \\\n",
    "        -S {fname(out_path,sample,\"bam\")} 2> /dev/null"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "b1a38da9-6c53-4c14-82bc-252bdca96bfe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 324M\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  61M Oct 14 00:04 a4_1.bam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  17M Oct 14 00:04 a4_2.bam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  51M Oct 14 00:05 a4_3.bam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  38M Oct 14 00:05 a4_4.bam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  300 Oct 14 00:07 dedup.stats_edit_distance.tsv\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 1.4K Oct 14 00:07 dedup.stats_per_umi_per_position.tsv\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 1.5M Oct 14 00:07 dedup.stats_per_umi.tsv\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  42M Oct 14 00:06 no_1.bam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  43M Oct 14 00:06 no_2.bam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  31M Oct 14 00:07 no_3.bam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  44M Oct 14 00:07 no_4.bam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  36K Oct 14 00:07 umitools.log\n"
     ]
    }
   ],
   "source": [
    "! ls -lh {out_path}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94a15938-2c5e-4f6d-81e0-cb27eabecc79",
   "metadata": {},
   "source": [
    "#### Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "839a5e4e-1d77-4b00-80f8-24ec426cc9dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2595303, 819585, 1630323, 1176367, 1766129, 1290541, 967229, 1417404]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Pre dedup\n",
    "ins = [nseqs(fname(in_path,sample,'bam')) for sample in samples]\n",
    "ins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "4e3a8104-5eb3-4781-8d41-20567f55d7ff",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1479247, 379872, 1263524, 919344, 978142, 997874, 767642, 1090610]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Post dedup\n",
    "outs = [nseqs(fname(out_path,sample,'bam')) for sample in samples]\n",
    "outs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "0a869095-5214-4a7f-905a-a7af72eced51",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjcAAAGxCAYAAACeKZf2AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy81sbWrAAAACXBIWXMAAA9hAAAPYQGoP6dpAABJAElEQVR4nO3deVgVZf8/8PdhOyCyCIJAokAqoiaiuG8QiqKmPlZaWu6WucujJNojYppLaKi5b7jkUqGW4obibqYQpCmuqZhA7qCobOf+/eGP8+0E6Dkwh4PD+3Vdc13OnJl7PvcciHcz98wohBACRERERDJhZOgCiIiIiKTEcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQxVGVFQUFAqFejIxMYGzszM++OADXLlyxdDllQvTpk3TOEaVKlVC9erV0alTJyxatAiPHz8u0zr+yc/PD35+fnrb56ZNmxAZGSl5u/8+pmZmZnB3d8fYsWPx6NGjQuufPXsWgwYNgru7O8zNzVG5cmU0btwYc+fOxYMHD9Tr+fn5abRrbm6OevXqYcaMGcjJydGqNjc3N402LC0t0bhxY3z77bcw1MPrBw4cCDc3N4Psm+TDxNAFEJW1tWvXom7dunj+/DlOnDiBmTNn4tChQ7h48SKqVKli6PLKhb1798LGxgY5OTlITU3FwYMHERISgq+//ho7d+6Et7d3mde0ZMkSvba/adMm/PHHHxg3bpxe2i84po8fP8bu3buxYMECnD59GidPnlQHuZUrV2LEiBHw9PTExIkTUa9ePeTm5iI+Ph7Lli3DL7/8gu3bt6vb9PDwwHfffQcAuHv3LlatWoX//e9/SElJwYoVK7Sqq3Xr1oiIiAAApKamYv78+Rg9ejQyMzMxefJkiY8CURkRRBXE2rVrBQBx5swZjeXh4eECgFizZo2BKis/wsLCBABx9+7dQp8lJSUJGxsbUaNGDfH8+fMyqaMsde3aVdSsWVPydos7ph9//LEAII4fPy6EEOLkyZPC2NhYdO7cucjjm52dLX766Sf1fPv27UX9+vU11snNzRW1a9cWZmZm4tmzZ6+srWbNmqJr164ayzIyMtTfsyEMGDBAL98DVSy8LEUVnq+vLwDg77//1lgeHx+P7t27w87ODubm5vDx8cH333+vsU7Bpa64uDgMGzYM9vb2sLa2Rv/+/ZGVlYX09HT07t0btra2cHZ2xoQJE5Cbm6vRxoMHDzBixAi88cYbMDMzg4eHB6ZMmYLs7Gz1Oj4+Pmjbtm2h2vPz8/HGG2+gV69e6mU5OTmYMWMG6tatC6VSCQcHBwwaNAh3794t1XHy9vbGlClTkJKSgq1bt6qXu7m5YeDAgYXW//dlpMOHD0OhUGDjxo0IDg6Gk5MTLCws0L59eyQmJr5y/0VdlsrOzsb06dPh5eUFc3Nz2Nvbw9/fHydPnlSvs3jxYrRr1w6Ojo6wtLTEW2+9hblz52p8D35+foiJicHNmzc1LtMU0McxbdGiBQDg5s2bAICvvvoKCoUCK1asgFKpLLS+mZkZunfv/tI2TUxM0KhRI+Tk5BR5yUsb1tbWqFOnTqHfB22PwdatWxEYGAhnZ2dYWFjAy8sLkyZNQlZWVqF9RUVFwdPTE0qlEl5eXli/fn2RNS1duhTe3t6oXLkyrKysULduXZ5VopfiZSmq8K5fvw4AqFOnjnrZoUOH0LlzZzRv3hzLli2DjY0NtmzZgj59+uDp06eF/pgPHToUvXr1wpYtW5CYmIjJkycjLy8Ply5dQq9evfDJJ5/gwIEDmDNnDlxcXBAcHAwAeP78Ofz9/XHt2jWEh4ejYcOGOHbsGGbNmoWkpCTExMQAAAYNGoSxY8fiypUrqF27tnq/+/fvR2pqKgYNGgQAUKlU6NGjB44dO4aQkBC0atUKN2/eRFhYGPz8/BAfHw8LC4sSH6vu3bsjJCQER48eRf/+/UvUxuTJk9G4cWOsWrUKGRkZmDZtGvz8/JCYmAgPDw+t28nLy0NQUBCOHTuGcePG4e2330ZeXh5OnTqFlJQUtGrVCgBw7do19O3bF+7u7jAzM8Pvv/+OmTNn4uLFi1izZg2AF5e8PvnkE1y7dk3jsg+gv2N69epVAICDgwPy8/MRFxeHJk2awNXVVee2/un69euwtbWFg4NDibbPy8vDrVu3NH4fdDkGV65cQZcuXTBu3DhYWlri4sWLmDNnDk6fPo24uDh1m1FRURg0aBB69OiBefPmqX8WsrOzYWT0f//fvWXLFowYMQKjR49GREQEjIyMcPXqVVy4cKGER4gqBEOfOiIqKwWXpU6dOiVyc3PF48ePxd69e4WTk5No166dyM3NVa9bt25d4ePjo7FMCCG6desmnJ2dRX5+vkabo0eP1livZ8+eAoCYP3++xvJGjRqJxo0bq+eXLVsmAIjvv/9eY705c+YIAGL//v1CCCHu3bsnzMzMxOTJkzXW6927t6hWrZq6zs2bNwsAIjo6WmO9M2fOCABiyZIlLz1GL7ssJYQQz549EwBEUFCQelnNmjXFgAEDCq3bvn170b59e/X8oUOHBADRuHFjoVKp1Mtv3LghTE1NxdChQwvV8bL21q9fLwCIlStXvrRP/5Sfny9yc3PF+vXrhbGxsXjw4IH6s+IuS0l1TNPT00Vubq54+PCh2Lhxo7CwsBCurq7i2bNnIj09XQAQH3zwgdZ9KbgslZubK3Jzc0VaWpqYOnWqACCWLVumVRs1a9YUXbp0Ubdx8+ZNMWzYMGFqaip27dpV6mOgUqlEbm6uOHLkiAAgfv/9dyHEi+/BxcWl2J+Ff34Po0aNEra2ttoeFiIhRAW/LHX06FG88847cHFxgUKhwI4dO3RuQwiBiIgI1KlTB0qlEq6urvjqq6+kL5Yk06JFC5iamsLKygqdO3dGlSpV8NNPP8HE5MWJzKtXr+LixYvo168fgBf/J1swdenSBWlpabh06ZJGm926ddOY9/LyAgB07dq10PKCyxAAEBcXB0tLS7z33nsa6xWcGTp48CAAwN7eHu+88w7WrVsHlUoFAHj48CF++ukn9O/fX137rl27YGtri3feeUej7kaNGsHJyQmHDx8u6WEDAEnuoOnbt6/GJZ+aNWuiVatWOHTokE7t7NmzB+bm5hg8ePBL10tMTET37t1hb28PY2NjmJqaon///sjPz8fly5dfuR+pjqmTkxNMTU1RpUoVfPTRR2jcuDH27t0Lc3NzrbYvyvnz52FqagpTU1M4Oztj+vTpCA0Nxaeffqp1G7t371a3UbNmTaxcuRKLFi3S+NnV5Rj8+eef6Nu3L5ycnNTHu3379gCA5ORkAMClS5eQmppa7M/CPzVr1gyPHj3Chx9+iJ9++gn37t0ryaGiCqZCh5usrCx4e3vj22+/LXEbY8eOxapVqxAREYGLFy9i586daNasmYRVktTWr1+PM2fOIC4uDp9++imSk5Px4Ycfqj8vGGswYcIE9X/0C6YRI0YAQKH/wNrZ2WnMm5mZFbv8+fPn6vn79+/Dycmp0G3Pjo6OMDExwf3799XLBg8ejNu3byM2NhYAsHnzZmRnZ2tcIvv777/x6NEjmJmZFao9PT291H8YCoKZi4tLidtwcnIqctk/+6qNu3fvwsXFReMSxr+lpKSgbdu2uH37NhYsWIBjx47hzJkzWLx4MQDg2bNnr9yPVMf0wIEDOHPmDJKSknDv3j0cP34c9erVAwBUrVoVlSpVUl8i1dabb76JM2fO4PTp0/jhhx/g7e2NWbNmYcuWLVq30aZNG5w5cwanTp3Chg0b4ObmhlGjRuH48eM6H4MnT56gbdu2+PXXXzFjxgwcPnwYZ86cwbZt2wD83/Eu+K6L+1n4p48//hhr1qzBzZs38e6778LR0RHNmzdX/x4QFaVCj7kJCgpCUFBQsZ/n5OTgiy++wHfffYdHjx6hQYMGmDNnjnpQY3JyMpYuXYo//vgDnp6eZVQ1lZaXl5d6ELG/vz/y8/OxatUq/Pjjj3jvvfdQtWpVAEBoaKjGQN1/kur7tre3x6+//gohhEbAuXPnDvLy8tS1AECnTp3g4uKCtWvXolOnTli7di2aN2+u/gMJvPgjaW9vj7179xa5Pysrq1LV+/PPPwOAxsBec3NzjcHPBe7du6dRf4H09PQil9nb2+tUi4ODA44fPw6VSlVswNmxYweysrKwbds21KxZU708KSlJ6/1IdUy9vb2LPB4AYGxsjICAAOzZswd//fUXqlevrlWb5ubm6p/lpk2bwt/fH/Xr18e4cePQrVs3VK5c+ZVt2NjYqNto3rw5mjdvDm9vb4wYMQJJSUkwMjLS+hjExcUhNTUVhw8fVp+tAVBocHPBd13cz8K/DRo0CIMGDUJWVhaOHj2KsLAwdOvWDZcvX9b4XokKVOgzN68yaNAgnDhxAlu2bMHZs2fx/vvvo3PnzuoHvu3cuRMeHh7YtWsX3N3d4ebmhqFDh2o8aIvKv7lz56JKlSqYOnUqVCoVPD09Ubt2bfz+++/w9fUtciptSCgQEBCAJ0+eFLokWnDXSEBAgHqZsbExPv74Y+zYsQPHjh1DfHx8oUsy3bp1w/3795Gfn19k3aUJZb///ju++uoruLm5oXfv3urlbm5uOHv2rMa6ly9fLnTprsDmzZs1Lm/dvHkTJ0+e1PkBfUFBQXj+/DmioqKKXacgMP7z7iMhBFauXFloXaVSWeSZHH0e038KDQ2FEALDhg0r8iF8ubm52Llz50vbsLe3x+zZs/H3339j0aJFJaqjdu3aCAkJwblz59R3xWl7DIo63gCwfPlyjXlPT084OzsX+7NQHEtLSwQFBWHKlCnIycnB+fPnS9RHqgAMON6nXAEgtm/frp6/evWqUCgU4vbt2xrrBQQEiNDQUCGEEJ9++qlQKpWiefPm4ujRo+LQoUOiUaNGwt/fvyxLJy0V95wbIYSYO3euACA2bNgghBAiLi5OKJVKERgYKDZt2iSOHDkitm/fLr766ivx3nvvvbLN4gbmDhgwQFhaWqrnnz17Jho2bCisrKzE/PnzRWxsrAgLCxOmpqaiS5cuheq8dOmSACCqV68uLCwsxKNHjzQ+z8vLE0FBQcLOzk6Eh4eLPXv2iAMHDoioqCgxYMAAsW3btpceo4K69+7dK3755Rdx9OhRsWXLFjFs2DBhbm4uatSoIc6dO6exzcaNGwUA8dlnn4kDBw6I1atXC09PT+Hs7FzkgGJXV1fRo0cPsWvXLvHdd9+JWrVqCSsrK3H16tVCdfzTvwcU5+bmCn9/f2FqaipCQkLEnj17RExMjJg6darYvHmzEEKI5ORkYWZmJvz8/MTu3bvFtm3bRMeOHUXt2rUFAHHo0KFC+1yyZIn49ddf1d+pVMe0uEHa/7RixQphYmIiGjRoIBYvXiwOHz4sYmNjxdy5c0WtWrVEz549NY7Hv59zI8SLwbpvvfWWsLOzExkZGS/dX1HPuRFCiMePH4tq1aoJT09PkZeXp/UxuHfvnqhSpYrw9vYW27ZtEzt37hQffPCB+nivXbtWvY9Vq1YJAOqfhY0bN4patWoJV1dXjQHFQ4cOFaNHjxZbtmwRR44cEVu3bhWNGjUSNjY24s6dO688plQxMdz8f/8ON99//70AICwtLTUmExMT0bt3byGEEMOGDRMAxKVLl9TbJSQkCADi4sWLZd0FeoWXhZtnz56JGjVqiNq1a4u8vDwhhBC///676N27t3B0dBSmpqbCyclJvP322xp3opQ23AghxP3798Xw4cOFs7OzMDExETVr1hShoaHFPiivVatWAoDo169fkZ/n5uaKiIgI4e3tLczNzUXlypVF3bp1xaeffiquXLny0mNUUHfBpFQqhbOzswgMDBQLFiwQmZmZhbZRqVRi7ty5wsPDQ5ibmwtfX18RFxdX7N1SGzZsEGPGjBEODg5CqVSKtm3bivj4+CLr+Kd/tyfEi+9t6tSp6gfX2dvbi7ffflucPHlSvc7OnTvVx+KNN94QEydOFHv27CkUbh48eCDee+89YWtrKxQKhcb+pTim2oQbIV48LHHAgAGiRo0awszMTFhaWgofHx8xdepUjT/mxYUbIYSIiYkRAER4ePhL91VcuBFCiMWLFwsAYt26dUII7Y/ByZMnRcuWLUWlSpWEg4ODGDp0qPjtt98KhRshXgScgu+uTp06Ys2aNYUe4rdu3Trh7+8vqlWrJszMzISLi4vo3bu3OHv27Ev7RhWbQggDvUCknFEoFNi+fTt69uwJ4MWDqPr164fz58/D2NhYY93KlSvDyckJYWFh+OqrrzQeBvbs2TNUqlQJ+/fvR8eOHcuyC0Tl2uHDh+Hv748ffvih0N1hRERSqtADil/Gx8cH+fn5uHPnTpFPhgVevJMlLy8P165dw5tvvgkA6ltLOciNiIjIMCp0uHny5In6KaHAiyd7JiUlwc7ODnXq1EG/fv3Qv39/zJs3Dz4+Prh37x7i4uLw1ltvoUuXLujQoQMaN26MwYMHIzIyEiqVCiNHjkTHjh01nu5JREREZadCX5YqOE3+bwMGDEBUVBRyc3MxY8YMrF+/Hrdv34a9vT1atmyJ8PBwvPXWWwBevEV39OjR2L9/v3ok/7x58wo934SIiIjKRoUON0RERCQ/fM4NERERyQrDDREREclKhRtQrFKpkJqaCisrq0Lv8yEiIqLySQiBx48fv/KdckAFDDepqalwdXU1dBlERERUArdu3Xrl+9cqXLgpeCfQrVu3YG1tbeBqiIiISBuZmZlwdXXV6t1+FS7cFFyKsra2ZrghIiJ6zWgzpIQDiomIiEhWGG6IiIhIVhhuiIiISFYq3JgbIiIq34QQyMvLQ35+vqFLoTJmamoKY2PjUrfDcENEROVGTk4O0tLS8PTpU0OXQgagUChQvXp1VK5cuVTtMNwQEVG5oFKpcP36dRgbG8PFxQVmZmZ82GoFIoTA3bt38ddff6F27dqlOoPDcENEROVCTk4OVCoVXF1dUalSJUOXQwbg4OCAGzduIDc3t1ThhgOKiYioXHnVo/VJvqQ6U8efICIiIpIVhhsiIiKSFY65ISKics9tUkyZ7u/G7K5luj+SFs/cEBERldLAgQOhUCigUChgamoKDw8PTJgwAVlZWZLvq2A/CoUClpaWqF27NgYOHIiEhATJ9/W6YrghIiKSQOfOnZGWloY///wTM2bMwJIlSzBhwoQi183NzS3VvtauXYu0tDScP38eixcvxpMnT9C8eXOsX7++VO3KBcMNERGRBJRKJZycnODq6oq+ffuiX79+2LFjBwBg2rRpaNSoEdasWQMPDw8olUoIIZCRkYFPPvkEjo6OsLa2xttvv43ff//9lfuytbWFk5MT3NzcEBgYiB9//BH9+vXDqFGj8PDhQ/V6J0+eRLt27WBhYQFXV1eMGTNG42ySm5sbvvzyS/Tt2xeVK1eGi4sLFi1apP78xo0bUCgUSEpKUi979OgRFAoFDh8+DAA4fPgwFAoFYmJi4O3tDXNzczRv3hznzp0r3QEtBY65kZjU14V53ZeI6PVkYWGhcYbm6tWr+P777xEdHa1+hkvXrl1hZ2eH3bt3w8bGBsuXL0dAQAAuX74MOzs7nfY3fvx4rF+/HrGxsejduzfOnTuHTp064csvv8Tq1atx9+5djBo1CqNGjcLatWvV23399deYPHkypk2bhn379mH8+PGoW7cuOnbsqNP+J06ciAULFsDJyQmTJ09G9+7dcfnyZZiamurUjhQYboiIiCR2+vRpbNq0CQEBAeplOTk52LBhAxwcHAAAcXFxOHfuHO7cuQOlUgkAiIiIwI4dO/Djjz/ik08+0WmfdevWBfDibAvwIrT07dsX48aNAwDUrl0bCxcuRPv27bF06VKYm5sDAFq3bo1JkyYBAOrUqYMTJ07gm2++0TnchIWFqbdZt24dqlevju3bt6N37946tSMFhhsiIiIJ7Nq1C5UrV0ZeXh5yc3PRo0cPjUs8NWvWVAcbAEhISMCTJ09gb2+v0c6zZ89w7do1nfcvhADwfw/CS0hIwNWrV/Hdd99prFPwmgsvLy8AQMuWLTXaadmyJSIjI3Xe/z/bsbOzg6enJ5KTk3VuRwoMN0RERBLw9/fH0qVLYWpqChcXl0KXYywtLTXmVSoVnJ2d1WNX/snW1lbn/RcECXd3d3X7n376KcaMGVNo3Ro1ary0rYKAVPC06ILgBOg2GNpQ7wZjuCEiIpKApaUlatWqpfX6jRs3Rnp6OkxMTODm5lbq/UdGRsLa2hodOnRQt3/+/PlX1nTq1KlC8wWXuArONKWlpcHHxwcANAYX/3u7gtD08OFDXL58Wd1OWWO4ISIiMoAOHTqgZcuW6NmzJ+bMmQNPT0+kpqZi9+7d6NmzJ3x9fYvd9tGjR0hPT0d2djYuX76M5cuXY8eOHVi/fr36rM/nn3+OFi1aYOTIkRg2bBgsLS2RnJyM2NhYjctlJ06cwNy5c9GzZ0/Exsbihx9+QEzMi5tjLCws0KJFC8yePRtubm64d+8evvjiiyJrmj59Ouzt7VGtWjVMmTIFVatWRc+ePSU7XrpguCEionJPjneOKhQK7N69G1OmTMHgwYNx9+5dODk5oV27dqhWrdpLtx00aBAAwNzcHG+88QbatGmD06dPo3Hjxup1GjZsiCNHjmDKlClo27YthBB488030adPH422/vvf/yIhIQHh4eGwsrLCvHnz0KlTJ/Xna9asweDBg+Hr6wtPT0/MnTsXgYGBhWqaPXs2xo4diytXrsDb2xs///wzzMzMSnOISkwh/nkhrQLIzMyEjY0NMjIyYG1tLXn7vBWciKhknj9/juvXr8Pd3V19Jw/pl5ubG8aNG6e+o6okDh8+DH9/fzx8+LBEY4X+6WU/A7r8/eZD/IiIiEhWDBpuZs2ahaZNm8LKygqOjo7o2bMnLl269NJtCp6E+O/p4sWLZVQ1ERERlWcGHXNz5MgRjBw5Ek2bNkVeXh6mTJmCwMBAXLhwodAtc/926dIljdNS/3x2ABEREb1awQP/SsPPzw/lbYSLQcPN3r17NebXrl0LR0dHJCQkoF27di/d1tHRsdTX9oiIiEh+ytWYm4yMDADQ6n0aPj4+cHZ2RkBAAA4dOlTsetnZ2cjMzNSYiIiISL7KTbgRQiA4OBht2rRBgwYNil3P2dkZK1asQHR0NLZt2wZPT08EBATg6NGjRa4/a9Ys2NjYqCdXV1d9dYGIiIjKgXLznJtRo0bh7NmzOH78+EvX8/T0hKenp3q+ZcuWuHXrFiIiIoq8lBUaGorg4GD1fGZmJgMOERGRjJWLMzejR4/Gzz//jEOHDqF69eo6b9+iRQtcuXKlyM+USiWsra01JiIiIpIvg565EUJg9OjR2L59Ow4fPqx+2ZeuEhMT4ezsLHF1RERE9DoyaLgZOXIkNm3ahJ9++glWVlZIT08HANjY2MDCwgLAi8tKt2/fxvr16wG8eDGYm5sb6tevj5ycHGzcuBHR0dGIjo42WD+IiEjPptmU8f4yynZ/rwmFQoHt27cb7J1R2jLoZamlS5ciIyMDfn5+cHZ2Vk9bt25Vr5OWloaUlBT1fE5ODiZMmICGDRuibdu2OH78OGJiYtCrVy9DdIGIiAgDBw5UP1TW1NQUHh4emDBhArKyskrd9o0bN6BQKIp9G/e/1yuYrKysUL9+fYwcObLYoRtyZfDLUq8SFRWlMR8SEoKQkBA9VURERFQynTt3xtq1a5Gbm4tjx45h6NChyMrKwtKlS8u0jgMHDqB+/fp4+vQpzp07hwULFsDb2xs7d+5EQEBAmdZiKOViQDEREdHrTqlUwsnJCa6urujbty/69euHHTt2AHjxzLUxY8bA0dER5ubmaNOmDc6cOaPe9uHDh+jXrx8cHBxgYWGB2rVrY+3atQCgHo/q4+MDhUIBPz+/l9Zhb28PJycneHh4oEePHjhw4ACaN2+OIUOGID8/X73ezp070aRJE5ibm8PDwwPh4eHIy8tTf37lyhW0a9cO5ubmqFevHmJjYzX2U/A6pEePHqmXJSUlQaFQqJ98HBUVBVtbW+zYsQN16tSBubk5OnbsiFu3bul6eHXCcENERKQHFhYWyM3NBfDiqkN0dDTWrVuH3377DbVq1UKnTp3w4MEDAMD//vc/XLhwAXv27EFycjKWLl2KqlWrAgBOnz4N4MUZmbS0NGzbtk2nOoyMjDB27FjcvHkTCQkJAIB9+/bho48+wpgxY3DhwgUsX74cUVFRmDlzJgBApVKhV69eMDY2xqlTp7Bs2TJ8/vnnJToOT58+xcyZM7Fu3TqcOHECmZmZ+OCDD0rUlrbKzXNuiIiI5OL06dPYtGkTAgIC1JemoqKiEBQUBABYuXIlYmNjsXr1akycOBEpKSnw8fGBr68vAMDNzU3dVsG7EwvOyJRE3bp1AbwYl9OsWTPMnDkTkyZNwoABAwAAHh4e+PLLLxESEoKwsDAcOHAAycnJuHHjhvoRLV999ZW6fl3k5ubi22+/RfPmzQEA69atg5eXF06fPo1mzZqVqD+vwnBDREQkgV27dqFy5crIy8tDbm4uevTogUWLFuHatWvIzc1F69at1euampqiWbNmSE5OBgB89tlnePfdd/Hbb78hMDAQPXv2RKtWrSSrrWCMq0KhAAAkJCTgzJkz6jM1AJCfn4/nz5/j6dOnSE5ORo0aNTSePdeyZcsS7dvExEQd2oAXQcvW1hbJyckMN0REROWZv78/li5dClNTU7i4uMDU1BTAi7t+gf8LFgWEEOplQUFBuHnzJmJiYnDgwAEEBARg5MiRiIiIkKS2ghBVMH5HpVIhPDy8yDuNzc3Ni7zh59/1GxkZqftRoOAy3Ku2LW6ZVDjmhoiISAKWlpaoVasWatasqQ42AFCrVi2YmZlpvF4oNzcX8fHx8PLyUi9zcHDAwIEDsXHjRkRGRmLFihUAADMzMwDQGAysC5VKhYULF8Ld3R0+Pj4AgMaNG+PSpUuoVatWocnIyAj16tVDSkoKUlNT1e388ssvGu0WXC4rCG8AirxdPS8vD/Hx8er5S5cu4dGjR+pLZfrAMzdERER6ZGlpic8++wwTJ06EnZ0datSogblz5+Lp06cYMmQIAGDq1Klo0qQJ6tevj+zsbOzatUsdfBwdHWFhYYG9e/eievXqMDc3h41N8Q81vH//PtLT0/H06VP88ccfiIyMxOnTpxETEwNjY2P1/rp16wZXV1e8//77MDIywtmzZ3Hu3DnMmDEDHTp0gKenJ/r374958+YhMzMTU6ZM0dhPrVq14OrqimnTpmHGjBm4cuUK5s2bV6geU1NTjB49GgsXLoSpqSlGjRqFFi1a6O2SFMBwQ0REr4PX/InBs2fPhkqlwscff4zHjx/D19cX+/btQ5UqVQC8ODsTGhqKGzduwMLCAm3btsWWLVsAvBizsnDhQkyfPh1Tp05F27Ztcfjw4WL31aFDBwBApUqVULNmTfj7+2PFihWoVauWep1OnTph165dmD59OubOnQtTU1PUrVsXQ4cOBfDiktP27dsxZMgQNGvWDG5ubli4cCE6d+6sbsPU1BSbN2/GZ599Bm9vbzRt2hQzZszA+++/r1FPpUqV8Pnnn6Nv377466+/0KZNG6xZs0aS41ochdDmSXoykpmZCRsbG2RkZOjlJZpuk2Ikbe/G7K6StkdEVF49f/4c169fh7u7O8zNzQ1dDkkgKioK48aN03gWzsu87GdAl7/fHHNDREREssJwQ0RERLLCcENERER6MXDgQK0vSUmJ4YaIiIhkheGGiIjKlQp2nwv9g1TfPcMNERGVCwUPvnv69KmBKyFDycnJAQD183hKis+5ISKicsHY2Bi2tra4c+cOgBfPR9HnI/qpfFGpVLh79y4qVaoEE5PSxROGGyIiKjcK3npdEHCoYjEyMkKNGjVKHWoZboiIqNxQKBRwdnaGo6NjsS9hJPkyMzNTv5CzNBhuiIio3DE2Ni71uAuquDigmIiIiGSF4YaIiIhkheGGiIiIZIXhhoiIiGSF4YaIiIhkheGGiIiIZIXhhoiIiGSF4YaIiIhkheGGiIiIZIXhhoiIiGSF4YaIiIhkheGGiIiIZIXhhoiIiGSF4YaIiIhkheGGiIiIZIXhhoiIiGSF4YaIiIhkheGGiIiIZIXhhoiIiGSF4YaIiIhkheGGiIiIZIXhhoiIiGSF4YaIiIhkheGGiIiIZIXhhoiIiGSF4YaIiIhkheGGiIiIZIXhhoiIiGSF4YaIiIhkheGGiIiIZIXhhoiIiGSF4YaIiIhkheGGiIiIZIXhhoiIiGSF4YaIiIhkheGGiIiIZIXhhoiIiGSF4YaIiIhkxaDhZtasWWjatCmsrKzg6OiInj174tKlS6/c7siRI2jSpAnMzc3h4eGBZcuWlUG1RERE9DowaLg5cuQIRo4ciVOnTiE2NhZ5eXkIDAxEVlZWsdtcv34dXbp0Qdu2bZGYmIjJkydjzJgxiI6OLsPKiYiIqLwyMeTO9+7dqzG/du1aODo6IiEhAe3atStym2XLlqFGjRqIjIwEAHh5eSE+Ph4RERF499139V0yERERlXPlasxNRkYGAMDOzq7YdX755RcEBgZqLOvUqRPi4+ORm5tbaP3s7GxkZmZqTERERCRf5SbcCCEQHByMNm3aoEGDBsWul56ejmrVqmksq1atGvLy8nDv3r1C68+aNQs2NjbqydXVVfLaiYiIqPwoN+Fm1KhROHv2LDZv3vzKdRUKhca8EKLI5QAQGhqKjIwM9XTr1i1pCiYiIqJyyaBjbgqMHj0aP//8M44ePYrq1au/dF0nJyekp6drLLtz5w5MTExgb29faH2lUgmlUilpvURERFR+GfTMjRACo0aNwrZt2xAXFwd3d/dXbtOyZUvExsZqLNu/fz98fX1hamqqr1KJiIjoNWHQcDNy5Ehs3LgRmzZtgpWVFdLT05Geno5nz56p1wkNDUX//v3V88OHD8fNmzcRHByM5ORkrFmzBqtXr8aECRMM0QUiIiIqZwwabpYuXYqMjAz4+fnB2dlZPW3dulW9TlpaGlJSUtTz7u7u2L17Nw4fPoxGjRrhyy+/xMKFC3kbOBEREQEw8JibgoHALxMVFVVoWfv27fHbb7/poSIiIiJ63ZWbu6WIiIiIpMBwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESywnBDREREssJwQ0RERLLCcENERESyonO4uXXrFv766y/1/OnTpzFu3DisWLFC0sKIiIiISkLncNO3b18cOnQIAJCeno6OHTvi9OnTmDx5MqZPny55gURERES60Dnc/PHHH2jWrBkA4Pvvv0eDBg1w8uRJbNq0CVFRUVLXR0RERKQTncNNbm4ulEolAODAgQPo3r07AKBu3bpIS0uTtjoiIiIiHZnoukH9+vWxbNkydO3aFbGxsfjyyy8BAKmpqbC3t5e8QCIiqblNipG8zRuzu0reJhGVjM5nbubMmYPly5fDz88PH374Iby9vQEAP//8s/pyFREREZGh6Hzmxs/PD/fu3UNmZiaqVKmiXv7JJ5+gUqVKkhZHREREpCudww0AGBsbawQbAHBzc5OiHiIiIqJS0Src+Pj4QKFQaNXgb7/9VqqCiIiIiEpDq3DTs2dP9b+fP3+OJUuWoF69emjZsiUA4NSpUzh//jxGjBihlyKJiIiItKVVuAkLC1P/e+jQoRgzZoz6Lql/rnPr1i1pqyMiIiLSkc53S/3www/o379/oeUfffQRoqOjJSmKiIiIqKR0DjcWFhY4fvx4oeXHjx+Hubm5JEURERERlZTOd0uNGzcOn332GRISEtCiRQsAL8bcrFmzBlOnTpW8QCIiIiJd6BxuJk2aBA8PDyxYsACbNm0CAHh5eSEqKgq9e/eWvEAiIiIiXZToOTe9e/dmkCEiIqJySecxN0RERETlmc5nbvLz8/HNN9/g+++/R0pKCnJycjQ+f/DggWTFEREREelK5zM34eHhmD9/Pnr37o2MjAwEBwejV69eMDIywrRp0/RQIhEREZH2dA433333HVauXIkJEybAxMQEH374IVatWoWpU6fi1KlT+qiRiIiISGs6h5v09HS89dZbAIDKlSsjIyMDANCtWzfExMRIWx0RERGRjnQON9WrV0daWhoAoFatWti/fz8A4MyZM1AqldJWR0RERKQjnQcU/+c//8HBgwfRvHlzjB07Fh9++CFWr16NlJQUjB8/Xh81EhmU2yTpz0jemN1V8jaJiOgFncPN7Nmz1f9+7733UL16dZw8eRK1atVC9+7dJS2OiIiISFcleojfP7Vo0UL9GgYiIiIiQyvRQ/w2bNiA1q1bw8XFBTdv3gQAREZG4qeffpK0OCIiIiJd6Rxuli5diuDgYHTp0gWPHj1Cfn4+AMDW1haRkZFS10dERESkE53DzaJFi7By5UpMmTIFxsbG6uW+vr44d+6cpMURERER6UrncHP9+nX4+PgUWq5UKpGVlSVJUUREREQlpXO4cXd3R1JSUqHle/bsQb169aSoiYiIiKjEdL5bauLEiRg5ciSeP38OIQROnz6NzZs3Y9asWVi1apU+aiQiIiLSms7hZtCgQcjLy0NISAiePn2Kvn374o033sCCBQvwwQcf6KNGIiIiIq2V6Dk3w4YNw7Bhw3Dv3j2oVCo4OjoCAG7fvo033nhD0gKJiIiIdFGi59wUqFq1KhwdHZGeno7Ro0ejVq1aUtVFREREVCJah5tHjx6hX79+cHBwgIuLCxYuXAiVSoWpU6fCw8MDp06dwpo1a/RZKxEREdEraX1ZavLkyTh69CgGDBiAvXv3Yvz48di7dy+eP3+OPXv2oH379vqsk4iIiEgrWp+5iYmJwdq1axEREYGff/4ZQgjUqVMHcXFxJQ42R48exTvvvAMXFxcoFArs2LHjpesfPnwYCoWi0HTx4sUS7Z+IiIjkR+szN6mpqern2Hh4eMDc3BxDhw4t1c6zsrLg7e2NQYMG4d1339V6u0uXLsHa2lo97+DgUKo6iIiISD60DjcqlQqmpqbqeWNjY1haWpZq50FBQQgKCtJ5O0dHR9ja2pZq30RERCRPWocbIQQGDhwIpVIJAHj+/DmGDx9eKOBs27ZN2gqL4OPjg+fPn6NevXr44osv4O/vX+y62dnZyM7OVs9nZmbqvT4iIiIyHK3DzYABAzTmP/roI8mLeRVnZ2esWLECTZo0QXZ2NjZs2ICAgAAcPnwY7dq1K3KbWbNmITw8vIwrJSIiIkPROtysXbtWn3VoxdPTE56enur5li1b4tatW4iIiCg23ISGhiI4OFg9n5mZCVdXV73XSkRERIZRqof4lQctWrTAlStXiv1cqVTC2tpaYyIiIiL5KtHrF8qTxMREODs7G7oM/Zlmo4c2M6Rvk4iIqJwwaLh58uQJrl69qp6/fv06kpKSYGdnhxo1aiA0NBS3b9/G+vXrAQCRkZFwc3ND/fr1kZOTg40bNyI6OhrR0dGG6gIRERGVMwYNN/Hx8Rp3OhWMjRkwYACioqKQlpaGlJQU9ec5OTmYMGECbt++DQsLC9SvXx8xMTHo0qVLmddORERE5ZNW4aZx48Y4ePAgqlSpgunTp2PChAmoVKlSqXfu5+cHIUSxn0dFRWnMh4SEICQkpNT7JSIiIvnSakBxcnIysrKyAADh4eF48uSJXosiIiIiKimtztw0atQIgwYNQps2bSCEQEREBCpXrlzkulOnTpW0QCIiIiJdaBVuoqKiEBYWhl27dkGhUGDPnj0wMSm8qUKhYLghIiIig9Iq3Hh6emLLli0AACMjIxw8eBCOjo56LYyIiIioJHS+W0qlUumjDiIiItKC26QYSdu7MburpO2VByW6FfzatWuIjIxEcnIyFAoFvLy8MHbsWLz55ptS10dERESkE51fv7Bv3z7Uq1cPp0+fRsOGDdGgQQP8+uuvqF+/PmJjY/VRIxEREZHWdD5zM2nSJIwfPx6zZ88utPzzzz9Hx44dJSuOiIiISFc6n7lJTk7GkCFDCi0fPHgwLly4IElRRERERCWlc7hxcHBAUlJSoeVJSUm8g4qIiIgMTufLUsOGDcMnn3yCP//8E61atYJCocDx48cxZ84c/Pe//9VHjURERERa0znc/O9//4OVlRXmzZuH0NBQAICLiwumTZuGMWPGSF4gERERkS50DjcKhQLjx4/H+PHj8fjxYwCAlZWV5IURERERlUSJnnNTgKGGiIiIyhudBxQTERERlWcMN0RERCQrDDdEREQkKzqFm9zcXPj7++Py5cv6qoeIiIioVHQKN6ampvjjjz+gUCj0VQ8RERFRqeh8Wap///5YvXq1PmohIiIiKjWdbwXPycnBqlWrEBsbC19fX1haWmp8Pn/+fMmKIyIiItKVzuHmjz/+QOPGjQGg0NgbXq4iIipf3CbFSNrejdldJW2PSB90DjeHDh3SRx1EREREkijxreBXr17Fvn378OzZMwCAEEKyooiIiIhKSudwc//+fQQEBKBOnTro0qUL0tLSAABDhw7lW8GJiIjI4HS+LDV+/HiYmpoiJSUFXl5e6uV9+vTB+PHjMW/ePEkLJJKlaTYSt5chbXtERK8xncPN/v37sW/fPlSvXl1jee3atXHz5k3JCiMiIiIqCZ0vS2VlZaFSpUqFlt+7dw9KpVKSooiIiIhKSudw065dO6xfv149r1AooFKp8PXXX8Pf31/S4oiIiIh0pfNlqa+//hp+fn6Ij49HTk4OQkJCcP78eTx48AAnTpzQR41EREREWtP5zE29evVw9uxZNGvWDB07dkRWVhZ69eqFxMREvPnmm/qokYiIiEhrOp+5AQAnJyeEh4dLXQsRERFRqZUo3Dx8+BCrV69GcnIyFAoFvLy8MGjQINjZ2UldHxEREZFOdL4sdeTIEbi7u2PhwoV4+PAhHjx4gIULF8Ld3R1HjhzRR41EREREWtP5zM3IkSPRu3dvLF26FMbGxgCA/Px8jBgxAiNHjsQff/wheZFERERE2tL5zM21a9fw3//+Vx1sAMDY2BjBwcG4du2apMURERER6UrncNO4cWMkJycXWp6cnIxGjRpJURMRERFRiWl1Wers2bPqf48ZMwZjx47F1atX0aJFCwDAqVOnsHjxYsyePVs/VRJRmXCbFCNpezdmd5W0PSIibWgVbho1agSFQgEhhHpZSEhIofX69u2LPn36SFcdERERkY60CjfXr1/Xdx1EREREktAq3NSsWVPfdRARERFJokQP8bt9+zZOnDiBO3fuQKVSaXw2ZswYSQojIiIiKgmdw83atWsxfPhwmJmZwd7eHgqFQv2ZQqFguCEiIiKD0jncTJ06FVOnTkVoaCiMjHS+k5yIiIhIr3ROJ0+fPsUHH3zAYENERETlks4JZciQIfjhhx/0UQsRERFRqel8WWrWrFno1q0b9u7di7feegumpqYan8+fP1+y4oiIiIh0pXO4+eqrr7Bv3z54enoCQKEBxURERESGpHO4mT9/PtasWYOBAwfqoRwiIiIqU9Ns9NBmhvRt6kDnMTdKpRKtW7fWRy1EREREpaZzuBk7diwWLVqkj1qIiIiISk3ny1KnT59GXFwcdu3ahfr16xcaULxt2zbJiiMiIiLSlc7hxtbWFr169dJHLURERESlVqLXLxARERGVV3zMMBEREcmKzmdu3N3dX/o8mz///LNUBRERERGVhs7hZty4cRrzubm5SExMxN69ezFx4kSp6iIiIiIqEZ3DzdixY4tcvnjxYsTHx+vU1tGjR/H1118jISEBaWlp2L59O3r27PnSbY4cOYLg4GCcP38eLi4uCAkJwfDhw3XaLxEREcmXZGNugoKCEB0drdM2WVlZ8Pb2xrfffqvV+tevX0eXLl3Qtm1bJCYmYvLkyRgzZozO+yUiIiL50vnMTXF+/PFH2NnZ6bRNUFAQgoKCtF5/2bJlqFGjBiIjIwEAXl5eiI+PR0REBN59912d9k1ERETypHO48fHx0RhQLIRAeno67t69iyVLlkha3L/98ssvCAwM1FjWqVMnrF69Grm5uYUeKAgA2dnZyM7OVs9nZmbqtUYiIiIyLJ3Dzb/HxBgZGcHBwQF+fn6oW7euVHUVKT09HdWqVdNYVq1aNeTl5eHevXtwdnYutM2sWbMQHh6u17qIiCoMGb5kkeRH53ATFhamjzq09u/b0IUQRS4vEBoaiuDgYPV8ZmYmXF1d9VcgERERGZRkY27KgpOTE9LT0zWW3blzByYmJrC3ty9yG6VSCaVSWRblERERUTmgdbgxMjJ66cP7gBdnT/Ly8kpdVHFatmyJnTt3aizbv38/fH19ixxvQ0RUZqS+XMNLNUQlpnW42b59e7GfnTx5EosWLVJfItLWkydPcPXqVfX89evXkZSUBDs7O9SoUQOhoaG4ffs21q9fDwAYPnw4vv32WwQHB2PYsGH45ZdfsHr1amzevFmn/RIRkby4TYqRvM0bs7tK3iaVDa3DTY8ePQotu3jxIkJDQ7Fz507069cPX375pU47j4+Ph7+/v3q+YGzMgAEDEBUVhbS0NKSkpKg/d3d3x+7duzF+/HgsXrwYLi4uWLhwIW8DJyIiIrUSjblJTU1FWFgY1q1bh06dOiEpKQkNGjTQuR0/P7+Xnu2JiooqtKx9+/b47bffdN4XERERVQw6PaE4IyMDn3/+OWrVqoXz58/j4MGD2LlzZ4mCDREREZE+aH3mZu7cuZgzZw6cnJywefPmIi9TERERERma1uFm0qRJsLCwQK1atbBu3TqsW7euyPW2bdsmWXFEREREutI63PTv3/+Vt4ITERERGZrW4aaowb1ERC/FR/UTkQHoNKCYiIiIqLxjuCEiIiJZYbghIiIiWWG4ISIiIllhuCEiIiJZYbghIiIiWWG4ISIiIllhuCEiIiJZYbghIiIiWWG4ISIiIllhuCEiIiJZYbghIiIiWdH6xZlE/+Y2KUbS9m7M7ippe0REVDHxzA0RERHJCsMNERERyQrDDREREckKww0RERHJCsMNERERyQrDDREREckKbwUnIiIqyjQbidvLkLY9KhbP3BAREZGsMNwQERGRrDDcEBERkaww3BAREZGsMNwQERGRrDDcEBERkaww3BAREZGsMNwQERGRrDDcEBERkaww3BAREZGsMNwQERGRrDDcEBERkaww3BAREZGsMNwQERGRrDDcEBERkaww3BAREZGsMNwQERGRrDDcEBERkaww3BAREZGsMNwQERGRrDDcEBERkaww3BAREZGsMNwQERGRrDDcEBERkaww3BAREZGsMNwQERGRrDDcEBERkaww3BAREZGsMNwQERGRrDDcEBERkaww3BAREZGsMNwQERGRrBg83CxZsgTu7u4wNzdHkyZNcOzYsWLXPXz4MBQKRaHp4sWLZVgxERERlWcGDTdbt27FuHHjMGXKFCQmJqJt27YICgpCSkrKS7e7dOkS0tLS1FPt2rXLqGIiIiIq7wwabubPn48hQ4Zg6NCh8PLyQmRkJFxdXbF06dKXbufo6AgnJyf1ZGxsXOy62dnZyMzM1JiIiIhIvgwWbnJycpCQkIDAwECN5YGBgTh58uRLt/Xx8YGzszMCAgJw6NChl647a9Ys2NjYqCdXV9dS105ERETll8HCzb1795Cfn49q1appLK9WrRrS09OL3MbZ2RkrVqxAdHQ0tm3bBk9PTwQEBODo0aPF7ic0NBQZGRnq6datW5L2g4iIiMoXE0MXoFAoNOaFEIWWFfD09ISnp6d6vmXLlrh16xYiIiLQrl27IrdRKpVQKpXSFUxERETlmsHO3FStWhXGxsaFztLcuXOn0Nmcl2nRogWuXLkidXlERET0mjJYuDEzM0OTJk0QGxursTw2NhatWrXSup3ExEQ4OztLXR4RERG9pgx6WSo4OBgff/wxfH190bJlS6xYsQIpKSkYPnw4gBfjZW7fvo3169cDACIjI+Hm5ob69esjJycHGzduRHR0NKKjow3ZDSIiIipHDBpu+vTpg/v372P69OlIS0tDgwYNsHv3btSsWRMAkJaWpvHMm5ycHEyYMAG3b9+GhYUF6tevj5iYGHTp0sVQXSAiIqJyxuADikeMGIERI0YU+VlUVJTGfEhICEJCQsqgKiIiInpdGfz1C0RERERSYrghIiIiWWG4ISIiIllhuCEiIiJZYbghIiIiWWG4ISIiIllhuCEiIiJZYbghIiIiWWG4ISIiIlkx+BOKidSm2UjcXoa07RER0WuBZ26IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVhhuiIiISFYYboiIiEhWGG6IiIhIVgwebpYsWQJ3d3eYm5ujSZMmOHbs2EvXP3LkCJo0aQJzc3N4eHhg2bJlZVQpERERvQ4MGm62bt2KcePGYcqUKUhMTETbtm0RFBSElJSUIte/fv06unTpgrZt2yIxMRGTJ0/GmDFjEB0dXcaVExERUXll0HAzf/58DBkyBEOHDoWXlxciIyPh6uqKpUuXFrn+smXLUKNGDURGRsLLywtDhw7F4MGDERERUcaVExERUXllYqgd5+TkICEhAZMmTdJYHhgYiJMnTxa5zS+//ILAwECNZZ06dcLq1auRm5sLU1PTQttkZ2cjOztbPZ+RkQEAyMzMLG0XiqTKfippe5kKIWl7LxqVpu/lvq/ltJ9Axelref355XdacvxOS9Mgv9PSNfmiTSG0qFcYyO3btwUAceLECY3lM2fOFHXq1Clym9q1a4uZM2dqLDtx4oQAIFJTU4vcJiwsTADgxIkTJ06cOMlgunXr1iszhsHO3BRQKBQa80KIQstetX5RywuEhoYiODhYPa9SqfDgwQPY29u/dD/lQWZmJlxdXXHr1i1YW1sbuhy9Yl/lp6L0E6g4fa0o/QQqTl9fp34KIfD48WO4uLi8cl2DhZuqVavC2NgY6enpGsvv3LmDatWqFbmNk5NTkeubmJjA3t6+yG2USiWUSqXGMltb25IXbgDW1tbl/odOKuyr/FSUfgIVp68VpZ9Axenr69JPGxsbrdYz2IBiMzMzNGnSBLGxsRrLY2Nj0apVqyK3admyZaH19+/fD19f3yLH2xAREVHFY9C7pYKDg7Fq1SqsWbMGycnJGD9+PFJSUjB8+HAALy4p9e/fX73+8OHDcfPmTQQHByM5ORlr1qzB6tWrMWHCBEN1gYiIiMoZg4656dOnD+7fv4/p06cjLS0NDRo0wO7du1GzZk0AQFpamsYzb9zd3bF7926MHz8eixcvhouLCxYuXIh3333XUF3QK6VSibCwsEKX1eSIfZWfitJPoOL0taL0E6g4fZVrPxVCaHNPFREREdHrweCvXyAiIiKSEsMNERERyQrDDREREckKww0RERHJCsMNERERyQrDTTmTnZ2NRo0aQaFQICkpSevtxo4diyZNmkCpVKJRo0Z6q08qJenn77//jg8//BCurq6wsLCAl5cXFixYoN9CJVCSvt6/fx+dO3eGi4sLlEolXF1dMWrUKL298FUKJf3ZLXD//n1Ur14dCoUCjx49krw+Q0pLS0Pfvn3h6ekJIyMjjBs3ztAl6c22bdvQsWNHODg4wNraGi1btsS+ffsMXZbkjh8/jtatW8Pe3h4WFhaoW7cuvvnmG0OXpXcnTpyAiYlJuf87w3BTzoSEhGj13ox/E0Jg8ODB6NOnjx6qkl5J+pmQkAAHBwds3LgR58+fx5QpUxAaGopvv/1WT1VKoyR9NTIyQo8ePfDzzz/j8uXLiIqKwoEDB9QPuCyPSvqzW2DIkCFo2LChhBWVH9nZ2XBwcMCUKVPg7e1t6HL06ujRo+jYsSN2796NhIQE+Pv745133kFiYqKhS5OUpaUlRo0ahaNHjyI5ORlffPEFvvjiC6xYscLQpelNRkYG+vfvj4CAAEOX8mqvfLUmSWLPnj2idevWwsbGRtjZ2YmuXbuKq1evaqyze/duUbduXXH+/HkBQCQmJuq8n7CwMOHt7S1N0SVQVv0sMGLECOHv71/KqkumrPu6YMECUb169VJWrbuy6OeSJUtE+/btxcGDBwUA8fDhQ+k6oKX27duL0aNHi4kTJ4oqVaqIatWqibCwMPXnN2/eFN27dxeWlpbCyspKvP/++yI9Pb1E+xk7dqx0hZdAWfW1QL169UR4eLgEleumrPv5n//8R3z00UcSVK67suhrnz59xBdffGHwvzPa4JmbMpKVlYXg4GCcOXMGBw8ehJGREf7zn/9ApVIBAP7++28MGzYMGzZsQKVKlQxcbcmVdT8zMjJgZ2dX6nZKoiz7mpqaim3btqF9+/ZSlK4TfffzwoULmD59OtavXw8jI8P+J2ndunWwtLTEr7/+irlz52L69OmIjY2FEAI9e/bEgwcPcOTIEcTGxuLatWuvzZnSopRVX1UqFR4/fmyw39Oy6mdiYiJOnjxpkN/RAvrs69q1a3Ht2jWEhYXpsQcSMmy2qrju3LkjAIhz584JlUolOnfuLL788kshhBDXr19/bc/c/Ju++imEECdPnhSmpqZi//79ElZccvro6wcffCAsLCwEAPHOO++IZ8+e6aFy3UjZz+fPn4uGDRuKDRs2CCGEOHTokEHP3LRp00ZjWdOmTcXnn38u9u/fL4yNjUVKSor6s4KzVKdPn9Z5P+XhzE1Z9FUIIebOnSvs7OzE33//Xeq6dVUW/XzjjTeEmZmZMDIyEtOnT5esdl3ps6+XL18Wjo6O4tKlS0KI8vd3pig8c1NGrl27hr59+8LDwwPW1tZwd3cHAKSkpGDRokXIzMxEaGiogassvbLq5/nz59GjRw9MnToVHTt2LHV7JVEWff3mm2/w22+/YceOHbh27RqCg4OlKF0n+uxnaGgovLy88NFHH0lZcon9e8yPs7Mz7ty5g+TkZLi6usLV1VX9Wb169WBra4vk5OSyLlMSZdHXzZs3Y9q0adi6dSscHR0lqVtX+u7nsWPHEB8fj2XLliEyMhKbN2+WrHZd6aOv+fn56Nu3L8LDw1GnTh291K0PBn1xZkXyzjvvwNXVFStXroSLiwtUKhUaNGiAnJwcxMXF4dSpU4VeXObr64t+/fph3bp1Bqpad2XRzwsXLuDtt9/GsGHD8MUXX+ijG1opi746OTnByckJdevWhb29Pdq2bYv//e9/cHZ21keXiqTPfsbFxeHcuXP48ccfAbwYGA8AVatWxZQpUxAeHq6fThXD1NRUY16hUEClUkEIAYVCUWj94pa/DvTd161bt2LIkCH44Ycf0KFDh1LXW1L67mdB2H/rrbfw999/Y9q0afjwww9LV3QJ6aOvjx8/Rnx8PBITEzFq1CgAULdpYmKC/fv34+2335auExJhuCkD9+/fR3JyMpYvX462bdsCeHEbYYGFCxdixowZ6vnU1FR06tQJW7duRfPmzcu83pIqi36eP38eb7/9NgYMGICZM2dK2wEdGOI7LfjDn52dXYrKdaPvfkZHR+PZs2fq+TNnzmDw4ME4duwY3nzzTQl7Ujr16tVDSkoKbt26pf6/3wsXLiAjIwNeXl4Grk5aUvR18+bNGDx4MDZv3oyuXbvqs9wS08d3KoQo099PbZWmr9bW1jh37pzGsiVLliAuLg4//vijOtyVNww3ZaBKlSqwt7fHihUr4OzsjJSUFEyaNEn9eY0aNTTWr1y5MgDgzTffRPXq1bXax9WrV/HkyROkp6fj2bNn6ueM1KtXD2ZmZtJ05BX03c/z58/D398fgYGBCA4ORnp6OgDA2NgYDg4OEvbk1fTd1927d+Pvv/9G06ZNUblyZVy4cAEhISFo3bo13NzcJO3Ly+i7n/8OMPfu3QMAeHl5wdbWtpTVS6dDhw5o2LAh+vXrh8jISOTl5WHEiBFo3749fH19tWqj4HfyyZMnuHv3LpKSkmBmZoZ69erpsXLdlbavmzdvRv/+/bFgwQK0aNFC/XtqYWEBGxsbfZevtdL2c/HixahRowbq1q0L4EXoj4iIwOjRo/Vdus5K01cjIyM0aNBAY5mjoyPMzc0LLS9POOamDBgZGWHLli1ISEhAgwYNMH78eHz99deS7mPo0KHw8fHB8uXLcfnyZfj4+MDHxwepqamS7udl9N3PH374AXfv3sV3330HZ2dn9dS0aVPJ9qEtfffVwsICK1euRJs2beDl5YVx48ahW7du2LVrl2T70EZZ/Oy+DhQKBXbs2IEqVaqgXbt26NChAzw8PLB161at2yj4nUxISMCmTZvg4+ODLl266LHqkiltX5cvX468vDyMHDlS4/d07Nixeq5cN6Xtp0qlQmhoKBo1agRfX18sWrQIs2fPxvTp0/Vcue6k+Pl93ShEwbluIiIiIhngmRsiIiKSFYab18Dw4cNRuXLlIqfy/Dh+XVWUfgIVp68VpZ/169cvtp/fffedocuTVEXpa0XpJyDPvvKy1Gvgzp07xb4w0dra2mDPj5BaReknUHH6WlH6efPmTeTm5hb5WbVq1WBlZVXGFelPRelrReknIM++MtwQERGRrPCyFBEREckKww0RERHJCsMNERERyQrDDREREckKww0RERHJCsMNERERyQrDDREREcnK/wM/AbRxF5wTMQAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "make_table(ins, outs, \"Pre Depup\", \"Post Dedup\", \"Number of Reads\", samples, \"Remove Duplicate PCR Reads\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cb41ed9",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Step: split_fwd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e3bb70a-e471-4e48-ac98-fb800b30c9b8",
   "metadata": {},
   "source": [
    "#### Split forward reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "8e7f2f40",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_path  = mkpath(\"dedup\")\n",
    "out_path = mkpath(\"split_fwd\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87fb1071-811c-4441-93e1-5fbf2025812e",
   "metadata": {},
   "source": [
    "#### samtools view"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01814e9d-279b-46c9-85d6-a6c2933c07c7",
   "metadata": {},
   "source": [
    "https://www.htslib.org"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "f7f26c58",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Usage: samtools view [options] <in.bam>|<in.sam>|<in.cram> [region ...]\n",
      "\n",
      "Output options:\n",
      "  -b, --bam                  Output BAM\n",
      "  -C, --cram                 Output CRAM (requires -T)\n",
      "  -1, --fast                 Use fast BAM compression (and default to --bam)\n",
      "  -u, --uncompressed         Uncompressed BAM output (and default to --bam)\n",
      "  -h, --with-header          Include header in SAM output\n",
      "  -H, --header-only          Print SAM header only (no alignments)\n",
      "      --no-header            Print SAM alignment records only [default]\n",
      "  -c, --count                Print only the count of matching records\n",
      "  -o, --output FILE          Write output to FILE [standard output]\n",
      "  -U, --unoutput FILE, --output-unselected FILE\n",
      "                             Output reads not selected by filters to FILE\n",
      "  -p, --unmap                Set flag to UNMAP on reads not selected\n",
      "                             then write to output file.\n",
      "  -P, --fetch-pairs          Retrieve complete pairs even when outside of region\n",
      "Input options:\n",
      "  -t, --fai-reference FILE   FILE listing reference names and lengths\n",
      "  -M, --use-index            Use index and multi-region iterator for regions\n",
      "      --region[s]-file FILE  Use index to include only reads overlapping FILE\n",
      "  -X, --customized-index     Expect extra index file argument after <in.bam>\n",
      "\n",
      "Filtering options (Only include in output reads that...):\n",
      "  -L, --target[s]-file FILE  ...overlap (BED) regions in FILE\n",
      "  -r, --read-group STR       ...are in read group STR\n",
      "  -R, --read-group-file FILE ...are in a read group listed in FILE\n",
      "  -N, --qname-file FILE      ...whose read name is listed in FILE\n",
      "  -d, --tag STR1[:STR2]      ...have a tag STR1 (with associated value STR2)\n",
      "  -D, --tag-file STR:FILE    ...have a tag STR whose value is listed in FILE\n",
      "  -q, --min-MQ INT           ...have mapping quality >= INT\n",
      "  -l, --library STR          ...are in library STR\n",
      "  -m, --min-qlen INT         ...cover >= INT query bases (as measured via CIGAR)\n",
      "  -e, --expr STR             ...match the filter expression STR\n",
      "  -f, --require-flags FLAG   ...have all of the FLAGs present\n",
      "  -F, --excl[ude]-flags FLAG ...have none of the FLAGs present\n",
      "      --rf, --incl-flags, --include-flags FLAG\n",
      "                             ...have some of the FLAGs present\n",
      "  -G FLAG                    EXCLUDE reads with all of the FLAGs present\n",
      "      --subsample FLOAT      Keep only FLOAT fraction of templates/read pairs\n",
      "      --subsample-seed INT   Influence WHICH reads are kept in subsampling [0]\n",
      "  -s INT.FRAC                Same as --subsample 0.FRAC --subsample-seed INT\n",
      "\n",
      "Processing options:\n",
      "      --add-flags FLAG       Add FLAGs to reads\n",
      "      --remove-flags FLAG    Remove FLAGs from reads\n",
      "  -x, --remove-tag STR\n",
      "               Comma-separated read tags to strip (repeatable) [null]\n",
      "      --keep-tag STR\n",
      "               Comma-separated read tags to preserve (repeatable) [null].\n",
      "               Equivalent to \"-x ^STR\"\n",
      "  -B, --remove-B             Collapse the backward CIGAR operation\n",
      "  -z, --sanitize FLAGS       Perform sanitity checking and fixing on records.\n",
      "                             FLAGS is comma separated (see manual). [off]\n",
      "\n",
      "General options:\n",
      "  -?, --help   Print long help, including note about region specification\n",
      "  -S           Ignored (input format is auto-detected)\n",
      "      --no-PG  Do not add a PG line\n",
      "      --input-fmt-option OPT[=VAL]\n",
      "               Specify a single input file format option in the form\n",
      "               of OPTION or OPTION=VALUE\n",
      "  -O, --output-fmt FORMAT[,OPT[=VAL]]...\n",
      "               Specify output format (SAM, BAM, CRAM)\n",
      "      --output-fmt-option OPT[=VAL]\n",
      "               Specify a single output file format option in the form\n",
      "               of OPTION or OPTION=VALUE\n",
      "  -T, --reference FILE\n",
      "               Reference sequence FASTA FILE [null]\n",
      "  -@, --threads INT\n",
      "               Number of additional threads to use [0]\n",
      "      --write-index\n",
      "               Automatically index the output files [off]\n",
      "      --verbosity INT\n",
      "               Set level of verbosity\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!samtools view "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a19e8c26-9b8e-4dc8-b04a-287837fc9ffd",
   "metadata": {},
   "source": [
    "##### paramaters\n",
    "- view,\n",
    "- -@ {nc}, number of threads\n",
    "- -F 16, excludes `flag 16`, meaning select forward reads (-f 16, selects reverse reads)\n",
    "- -b,\n",
    "- {in_fn}, input bam file \n",
    "- -o, ouput bam file with forward reads"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b10ad99-8063-4824-803a-395ca908affb",
   "metadata": {},
   "source": [
    "#### samtools index"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6eae2694-4b7c-4829-8ecc-20922373e867",
   "metadata": {},
   "source": [
    "https://www.htslib.org"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "50b827f2",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: samtools index -M [-bc] [-m INT] <in1.bam> <in2.bam>...\n",
      "   or: samtools index [-bc] [-m INT] <in.bam> [out.index]\n",
      "Options:\n",
      "  -b       Generate BAI-format index for BAM files [default]\n",
      "  -c       Generate CSI-format index for BAM files\n",
      "  -m INT   Set minimum interval size for CSI indices to 2^INT [14]\n",
      "  -M       Interpret all filename arguments as files to be indexed\n",
      "  -o FILE  Write index to FILE [alternative to <out.index> as an argument]\n",
      "  -@ INT   Sets the number of threads [none]\n"
     ]
    }
   ],
   "source": [
    "!samtools index"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e5782f3-d12e-484b-8ce9-d071d4fa65ff",
   "metadata": {},
   "source": [
    "##### paramters\n",
    "- index, creates a bam index file for a bam file\n",
    "- -@ {nc}, number of threads\n",
    "- {in_fn}, input sorted bam file and ouputs an index; *.bam -> *.bam.bai"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "39fb914c",
   "metadata": {},
   "outputs": [],
   "source": [
    "for sample in samples:\n",
    "    !samtools view  -@ {nc} -F 16 {fname(in_path,sample,\"bam\")} -b -o {fname(out_path,sample,\"bam\")} \n",
    "    !samtools index -@ {nc} {fname(out_path,sample,\"bam\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "3bd84be6-eae8-4cfc-bc57-f254d3474716",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 324M\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  61M Oct 14 00:04 a4_1.bam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  17M Oct 14 00:04 a4_2.bam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  51M Oct 14 00:05 a4_3.bam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  38M Oct 14 00:05 a4_4.bam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  300 Oct 14 00:07 dedup.stats_edit_distance.tsv\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 1.4K Oct 14 00:07 dedup.stats_per_umi_per_position.tsv\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 1.5M Oct 14 00:07 dedup.stats_per_umi.tsv\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  42M Oct 14 00:06 no_1.bam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  43M Oct 14 00:06 no_2.bam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  31M Oct 14 00:07 no_3.bam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  44M Oct 14 00:07 no_4.bam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  36K Oct 14 00:07 umitools.log\n"
     ]
    }
   ],
   "source": [
    "!ls -lh {in_path}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "360074b8-d88d-4971-a8b3-0c053fc626bf",
   "metadata": {},
   "source": [
    "#### Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "e54d1f81-c5d4-471e-b9c4-a68c4811352d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('dedup/a4_2.bam')"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_fn  = fname(in_path,treated[1], 'bam')\n",
    "out_fn = fname(out_path,treated[1],'bam')\n",
    "in_fn"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8a90a2a-c5a8-4be5-8a21-8b9463db29ed",
   "metadata": {},
   "source": [
    "Look at the format of a bam file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "07000b50",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VH01331:96:AACYNKFM5:1:2306:27131:33336_TTTTTTTA\t16\tchr1\t405127\t255\t31M2D23M\t*\t0\t0\tGGGTAAAGAGGCAAAGAGGCACCTTTTATTTTTTTTTTTTTTTTTTTTTTGAGT\tCCC--CC--CC;CCC---C--;--CC;-;CCCCCCCCCCCCCCCCCC;C;CCC;\tNH:i:1\tHI:i:1\tAS:i:43\tnM:i:2\tNM:i:4\tMD:Z:20T6C3^TC23\tjM:B:c,-1\tjI:B:i,-1\n",
      "VH01331:96:AACYNKFM5:1:2606:44911:40284_TGCACAAC\t16\tchr1\t607282\t255\t54M\t*\t0\t0\tGACAGAAAGGCAGGAGGAAGAGGAGGAGGAGAATCTTTGCACAGGGGGTGTCAG\tC-CCCC---;CC-CC;;CCC-C;CCCCCC;CC;CCCCCC;CCCCCCCCCC-CCC\tNH:i:1\tHI:i:1\tAS:i:53\tnM:i:0\tNM:i:0\tMD:Z:54\tjM:B:c,-1\tjI:B:i,-1\n",
      "samtools view: writing to standard output failed: Broken pipe\n",
      "samtools view: error closing standard output: -1\n"
     ]
    }
   ],
   "source": [
    "!samtools view {in_fn} |head -2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cb34f22-da56-4555-9755-a309920d3a86",
   "metadata": {},
   "source": [
    "Number of forward and reverse reads should equal total reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "5058f7a0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "233566\n",
      "146306\n",
      "379872\n"
     ]
    }
   ],
   "source": [
    "!samtools view -c {in_fn} -F 16\n",
    "!samtools view -c {in_fn} -f 16\n",
    "!samtools view -c {in_fn}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "151516de",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Step: split_rev"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7299fd66-f718-4c46-abd1-4f2504338ab8",
   "metadata": {},
   "source": [
    "#### Split reverse reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "a7bda4bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_path  = mkpath(\"dedup\")\n",
    "out_path = mkpath(\"split_rev\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d07bbab-303e-4978-949a-f1cc979b12d0",
   "metadata": {},
   "source": [
    "#### samtools view"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6340272-8e6c-4aed-9a2f-fb502b0769e4",
   "metadata": {},
   "source": [
    "##### paramaters\n",
    "- view,\n",
    "- -@ {nc}, number of threads\n",
    "- -F 16, excludes `flag 16`, meaning select forward reads (-f 16, selects reverse reads)\n",
    "- -b,\n",
    "- {in_fn}, input bam file \n",
    "- -o, ouput bam file with forward reads"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "915ae262-106a-4b62-962c-cb9f40204138",
   "metadata": {},
   "source": [
    "#### samtools index"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2591337-70f1-403c-b9b7-a84bf3a64730",
   "metadata": {},
   "source": [
    "##### paramters\n",
    "- index, creates a bam index file for a bam file\n",
    "- -@ {nc}, number of threads\n",
    "- {in_fn}, input sorted bam file and ouputs an index; *.bam -> *.bam.bai"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "35f8c90f",
   "metadata": {},
   "outputs": [],
   "source": [
    "for sample in samples:\n",
    "    !samtools view  -@ {nc} -f 16 {fname(in_path,sample,\"bam\")} -b -o {fname(out_path,sample,\"bam\")} \n",
    "    !samtools index -@ {nc} {fname(out_path,sample,\"bam\")}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3147326e",
   "metadata": {},
   "source": [
    "## Step: peak_fwd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "55001850-9733-48ff-8645-43dea137e873",
   "metadata": {},
   "source": [
    "#### Call peaks from forward reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "569fd44e",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_path   = mkpath(\"split_fwd\")\n",
    "out_path  = mkpath(\"peak_fwd\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fc1e3aa-37cc-4344-865c-de9b0d1d97df",
   "metadata": {},
   "source": [
    "#### macs3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbdd8d47-19df-4697-b707-c825f15a5a0b",
   "metadata": {},
   "source": [
    "https://macs3-project.github.io/MACS/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "49d758aa",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: macs3 callpeak [-h] -t TFILE [TFILE ...] [-c [CFILE ...]]\n",
      "                      [-f {AUTO,BAM,SAM,BED,ELAND,ELANDMULTI,ELANDEXPORT,BOWTIE,BAMPE,BEDPE}]\n",
      "                      [-g GSIZE] [-s TSIZE] [--keep-dup KEEPDUPLICATES]\n",
      "                      [--outdir OUTDIR] [-n NAME] [-B] [--verbose VERBOSE]\n",
      "                      [--trackline] [--SPMR] [--nomodel] [--shift SHIFT]\n",
      "                      [--extsize EXTSIZE] [--bw BW] [--d-min D_MIN]\n",
      "                      [-m MFOLD MFOLD] [--fix-bimodal] [-q QVALUE | -p PVALUE]\n",
      "                      [--scale-to {large,small}] [--down-sample] [--seed SEED]\n",
      "                      [--tempdir TEMPDIR] [--nolambda] [--slocal SMALLLOCAL]\n",
      "                      [--llocal LARGELOCAL] [--max-gap MAXGAP]\n",
      "                      [--min-length MINLEN] [--broad]\n",
      "                      [--broad-cutoff BROADCUTOFF] [--cutoff-analysis]\n",
      "                      [--call-summits] [--fe-cutoff FECUTOFF] [--to-large]\n",
      "                      [--ratio RATIO] [--buffer-size BUFFER_SIZE]\n",
      "\n",
      "options:\n",
      "  -h, --help            show this help message and exit\n",
      "\n",
      "Input files arguments:\n",
      "  -t TFILE [TFILE ...], --treatment TFILE [TFILE ...]\n",
      "                        ChIP-seq treatment file. If multiple files are given\n",
      "                        as '-t A B C', then they will all be read and pooled\n",
      "                        together. REQUIRED.\n",
      "  -c [CFILE ...], --control [CFILE ...]\n",
      "                        Control file. If multiple files are given as '-c A B\n",
      "                        C', they will be pooled to estimate ChIP-seq\n",
      "                        background noise.\n",
      "  -f {AUTO,BAM,SAM,BED,ELAND,ELANDMULTI,ELANDEXPORT,BOWTIE,BAMPE,BEDPE}, --format {AUTO,BAM,SAM,BED,ELAND,ELANDMULTI,ELANDEXPORT,BOWTIE,BAMPE,BEDPE}\n",
      "                        Format of tag file, \"AUTO\", \"BED\" or \"ELAND\" or\n",
      "                        \"ELANDMULTI\" or \"ELANDEXPORT\" or \"SAM\" or \"BAM\" or\n",
      "                        \"BOWTIE\" or \"BAMPE\" or \"BEDPE\". The default AUTO\n",
      "                        option will let MACS decide which format (except for\n",
      "                        BAMPE and BEDPE which should be implicitly set) the\n",
      "                        file is. Please check the definition in README. Please\n",
      "                        note that if the format is set as BAMPE or BEDPE,\n",
      "                        MACS3 will call its special Paired-end mode to call\n",
      "                        peaks by piling up the actual ChIPed fragments defined\n",
      "                        by both aligned ends, instead of predicting the\n",
      "                        fragment size first and extending reads. Also please\n",
      "                        note that the BEDPE only contains three columns, and\n",
      "                        is NOT the same BEDPE format used by BEDTOOLS.\n",
      "                        DEFAULT: \"AUTO\"\n",
      "  -g GSIZE, --gsize GSIZE\n",
      "                        Effective genome size. It can be 1.0e+9 or 1000000000,\n",
      "                        or shortcuts:'hs' for human (2.9e9), 'mm' for mouse\n",
      "                        (2.6e9), 'ce' for C. elegans (1e8) and 'dm' for\n",
      "                        fruitfly (1.4e8), Default:hs. The effective genome\n",
      "                        size numbers are collected from Deeptools https://deep\n",
      "                        tools.readthedocs.io/en/develop/content/feature/effect\n",
      "                        iveGenomeSize.html Please refer to deeptools to define\n",
      "                        the best genome size you plan to use.\n",
      "  -s TSIZE, --tsize TSIZE\n",
      "                        Tag size/read length. This will override the auto\n",
      "                        detected tag size. DEFAULT: Not set\n",
      "  --keep-dup KEEPDUPLICATES\n",
      "                        It controls the behavior towards duplicate tags at the\n",
      "                        exact same location -- the same coordination and the\n",
      "                        same strand. The 'auto' option makes MACS calculate\n",
      "                        the maximum tags at the exact same location based on\n",
      "                        binomal distribution using 1e-5 as pvalue cutoff; and\n",
      "                        the 'all' option keeps every tags. If an integer is\n",
      "                        given, at most this number of tags will be kept at the\n",
      "                        same location. Note, if you've used samtools or picard\n",
      "                        to flag reads as 'PCR/Optical duplicate' in bit 1024,\n",
      "                        MACS3 will still read them although the reads may be\n",
      "                        decided by MACS3 as duplicate later. If you plan to\n",
      "                        rely on samtools/picard/any other tool to filter\n",
      "                        duplicates, please remove those duplicate reads and\n",
      "                        save a new alignment file then ask MACS3 to keep all\n",
      "                        by '--keep-dup all'. The default is to keep one tag at\n",
      "                        the same location. Default: 1\n",
      "\n",
      "Output arguments:\n",
      "  --outdir OUTDIR       If specified all output files will be written to that\n",
      "                        directory. Default: the current working directory\n",
      "  -n NAME, --name NAME  Experiment name, which will be used to generate output\n",
      "                        file names. DEFAULT: \"NA\"\n",
      "  -B, --bdg             Whether or not to save extended fragment pileup, and\n",
      "                        local lambda tracks (two files) at every bp into a\n",
      "                        bedGraph file. DEFAULT: False\n",
      "  --verbose VERBOSE     Set verbose level of runtime message. 0: only show\n",
      "                        critical message, 1: show additional warning message,\n",
      "                        2: show process information, 3: show debug messages.\n",
      "                        DEFAULT:2\n",
      "  --trackline           Tells MACS to include trackline with bedGraph files.\n",
      "                        To include this trackline while displaying bedGraph at\n",
      "                        UCSC genome browser, can show name and description of\n",
      "                        the file as well. However my suggestion is to convert\n",
      "                        bedGraph to bigWig, then show the smaller and faster\n",
      "                        binary bigWig file at UCSC genome browser, as well as\n",
      "                        downstream analysis. Require -B to be set. Default:\n",
      "                        Not include trackline.\n",
      "  --SPMR                If True, MACS will SAVE signal per million reads for\n",
      "                        fragment pileup profiles. It won't interfere with\n",
      "                        computing pvalue/qvalue during peak calling, since\n",
      "                        internally MACS3 keeps using the raw pileup and\n",
      "                        scaling factors between larger and smaller dataset to\n",
      "                        calculate statistics measurements. If you plan to use\n",
      "                        the signal output in bedGraph to call peaks using\n",
      "                        bdgcmp and bdgpeakcall, you shouldn't use this option\n",
      "                        because you will end up with different results.\n",
      "                        However, this option is recommended for displaying\n",
      "                        normalized pileup tracks across many datasets. Require\n",
      "                        -B to be set. Default: False\n",
      "\n",
      "Shifting model arguments:\n",
      "  --nomodel             Whether or not to build the shifting model. If True,\n",
      "                        MACS will not build model. by default it means\n",
      "                        shifting size = 100, try to set extsize to change it.\n",
      "                        It's highly recommended that while you have many\n",
      "                        datasets to process and you plan to compare different\n",
      "                        conditions, aka differential calling, use both\n",
      "                        'nomodel' and 'extsize' to make signal files from\n",
      "                        different datasets comparable. DEFAULT: False\n",
      "  --shift SHIFT         (NOT the legacy --shiftsize option!) The arbitrary\n",
      "                        shift in bp. Use discretion while setting it other\n",
      "                        than default value. When NOMODEL is set, MACS will use\n",
      "                        this value to move cutting ends (5') towards 5'->3'\n",
      "                        direction then apply EXTSIZE to extend them to\n",
      "                        fragments. When this value is negative, ends will be\n",
      "                        moved toward 3'->5' direction. Recommended to keep it\n",
      "                        as default 0 for ChIP-Seq datasets, or -1 * half of\n",
      "                        EXTSIZE together with EXTSIZE option for detecting\n",
      "                        enriched cutting loci such as certain DNAseI-Seq\n",
      "                        datasets. Note, you can't set values other than 0 if\n",
      "                        format is BAMPE or BEDPE for paired-end data. DEFAULT:\n",
      "                        0.\n",
      "  --extsize EXTSIZE     The arbitrary extension size in bp. When nomodel is\n",
      "                        true, MACS will use this value as fragment size to\n",
      "                        extend each read towards 3' end, then pile them up.\n",
      "                        It's exactly twice the number of obsolete SHIFTSIZE.\n",
      "                        In previous language, each read is moved 5'->3'\n",
      "                        direction to middle of fragment by 1/2 d, then\n",
      "                        extended to both direction with 1/2 d. This is\n",
      "                        equivalent to say each read is extended towards 5'->3'\n",
      "                        into a d size fragment. DEFAULT: 200. EXTSIZE and\n",
      "                        SHIFT can be combined when necessary. Check SHIFT\n",
      "                        option.\n",
      "  --bw BW               Band width for picking regions to compute fragment\n",
      "                        size. This value is only used while building the\n",
      "                        shifting model. Tweaking this is not recommended.\n",
      "                        DEFAULT: 300\n",
      "  --d-min D_MIN         Minimum fragment size in basepair. Any predicted\n",
      "                        fragment size less than this will be excluded.\n",
      "                        DEFAULT: 20\n",
      "  -m MFOLD MFOLD, --mfold MFOLD MFOLD\n",
      "                        Select the regions within MFOLD range of high-\n",
      "                        confidence enrichment ratio against background to\n",
      "                        build model. Fold-enrichment in regions must be lower\n",
      "                        than upper limit, and higher than the lower limit. Use\n",
      "                        as \"-m 10 30\". This setting is only used while\n",
      "                        building the shifting model. Tweaking it is not\n",
      "                        recommended. DEFAULT:5 50\n",
      "  --fix-bimodal         Whether turn on the auto pair model process. If set,\n",
      "                        when MACS failed to build paired model, it will use\n",
      "                        the nomodel settings, the --exsize parameter to extend\n",
      "                        each tags towards 3' direction. Not to use this\n",
      "                        automate fixation is a default behavior now. DEFAULT:\n",
      "                        False\n",
      "\n",
      "Peak calling arguments:\n",
      "  -q QVALUE, --qvalue QVALUE\n",
      "                        Minimum FDR (q-value) cutoff for peak detection.\n",
      "                        DEFAULT: 0.05. -q, and -p are mutually exclusive.\n",
      "  -p PVALUE, --pvalue PVALUE\n",
      "                        Pvalue cutoff for peak detection. DEFAULT: not set.\n",
      "                        -q, and -p are mutually exclusive. If pvalue cutoff is\n",
      "                        set, qvalue will not be calculated and reported as -1\n",
      "                        in the final .xls file.\n",
      "  --scale-to {large,small}\n",
      "                        When set to 'small', scale the larger sample up to the\n",
      "                        smaller sample. When set to 'larger', scale the\n",
      "                        smaller sample up to the bigger sample. By default,\n",
      "                        scale to 'small'. This option replaces the obsolete '\n",
      "                        --to-large' option. The default behavior is\n",
      "                        recommended since it will lead to less significant\n",
      "                        p/q-values in general but more specific results. Keep\n",
      "                        in mind that scaling down will influence control/input\n",
      "                        sample more. DEFAULT: 'small', the choice is either\n",
      "                        'small' or 'large'.\n",
      "  --down-sample         When set, random sampling method will scale down the\n",
      "                        bigger sample. By default, MACS uses linear scaling.\n",
      "                        Warning: This option will make your result unstable\n",
      "                        and irreproducible since each time, random reads would\n",
      "                        be selected. Consider to use 'randsample' script\n",
      "                        instead. <not implmented>If used together with --SPMR,\n",
      "                        1 million unique reads will be randomly picked.</not\n",
      "                        implemented> Caution: due to the implementation, the\n",
      "                        final number of selected reads may not be as you\n",
      "                        expected! DEFAULT: False\n",
      "  --seed SEED           Set the random seed while down sampling data. Must be\n",
      "                        a non-negative integer in order to be effective.\n",
      "                        DEFAULT: not set\n",
      "  --tempdir TEMPDIR     Optional directory to store temp files. DEFAULT: /tmp\n",
      "  --nolambda            If True, MACS will use fixed background lambda as\n",
      "                        local lambda for every peak region. Normally, MACS\n",
      "                        calculates a dynamic local lambda to reflect the local\n",
      "                        bias due to the potential chromatin accessibility.\n",
      "  --slocal SMALLLOCAL   The small nearby region in basepairs to calculate\n",
      "                        dynamic lambda. This is used to capture the bias near\n",
      "                        the peak summit region. Invalid if there is no control\n",
      "                        data. If you set this to 0, MACS will skip slocal\n",
      "                        lambda calculation. *Note* that MACS will always\n",
      "                        perform a d-size local lambda calculation while the\n",
      "                        control data is available. The final local bias would\n",
      "                        be the maximum of the lambda value from d, slocal, and\n",
      "                        llocal size windows. While control is not available, d\n",
      "                        and slocal lambda won't be considered. DEFAULT: 1000\n",
      "  --llocal LARGELOCAL   The large nearby region in basepairs to calculate\n",
      "                        dynamic lambda. This is used to capture the surround\n",
      "                        bias. If you set this to 0, MACS will skip llocal\n",
      "                        lambda calculation. *Note* that MACS will always\n",
      "                        perform a d-size local lambda calculation while the\n",
      "                        control data is available. The final local bias would\n",
      "                        be the maximum of the lambda value from d, slocal, and\n",
      "                        llocal size windows. While control is not available, d\n",
      "                        and slocal lambda won't be considered. DEFAULT: 10000.\n",
      "  --max-gap MAXGAP      Maximum gap between significant sites to cluster them\n",
      "                        together. The DEFAULT value is the detected read\n",
      "                        length/tag size.\n",
      "  --min-length MINLEN   Minimum length of a peak. The DEFAULT value is the\n",
      "                        predicted fragment size d. Note, if you set a value\n",
      "                        smaller than the fragment size, it may have NO effect\n",
      "                        on the result. For BROAD peak calling, try to set a\n",
      "                        large value such as 500bps. You can also use '--\n",
      "                        cutoff-analysis' option with default setting, and\n",
      "                        check the column 'avelpeak' under different cutoff\n",
      "                        values to decide a reasonable minlen value.\n",
      "  --broad               If set, MACS will try to call broad peaks using the\n",
      "                        --broad-cutoff setting. Please tweak '--broad-cutoff'\n",
      "                        setting to control the peak calling behavior. At the\n",
      "                        meantime, either -q or -p cutoff will be used to\n",
      "                        define regions with 'stronger enrichment' inside of\n",
      "                        broad peaks. The maximum gap is expanded to 4 * MAXGAP\n",
      "                        (--max-gap parameter). As a result, MACS will output a\n",
      "                        'gappedPeak' and a 'broadPeak' file instead of\n",
      "                        'narrowPeak' file. Note, a broad peak will be reported\n",
      "                        even if there is no 'stronger enrichment' inside.\n",
      "                        DEFAULT: False\n",
      "  --broad-cutoff BROADCUTOFF\n",
      "                        Cutoff for broad region. This option is not available\n",
      "                        unless --broad is set. If -p is set, this is a pvalue\n",
      "                        cutoff, otherwise, it's a qvalue cutoff. Please note\n",
      "                        that in broad peakcalling mode, MACS3 uses this\n",
      "                        setting to control the overall peak calling behavior,\n",
      "                        then uses -q or -p setting to define regions inside\n",
      "                        broad region as 'stronger' enrichment. DEFAULT: 0.1\n",
      "  --cutoff-analysis     While set, MACS3 will analyze number or total length\n",
      "                        of peaks that can be called by different p-value\n",
      "                        cutoff then output a summary table to help user decide\n",
      "                        a better cutoff. The table will be saved in\n",
      "                        NAME_cutoff_analysis.txt file. Note, minlen and maxgap\n",
      "                        may affect the results. WARNING: May take ~30 folds\n",
      "                        longer time to finish. The result can be useful for\n",
      "                        users to decide a reasonable cutoff value. DEFAULT:\n",
      "                        False\n",
      "\n",
      "Post-processing options:\n",
      "  --call-summits        If set, MACS will use a more sophisticated signal\n",
      "                        processing approach to find subpeak summits in each\n",
      "                        enriched peak region. DEFAULT: False\n",
      "  --fe-cutoff FECUTOFF  When set, the value will be used as the minimum\n",
      "                        requirement to filter out peaks with low fold-\n",
      "                        enrichment. Note, MACS3 adds one as pseudocount while\n",
      "                        calculating fold-enrichment. By default, it is set as\n",
      "                        1 so there is no filtering. DEFAULT: 1.0\n",
      "\n",
      "Obsolete options:\n",
      "  --to-large            Obsolete option. Please use '--scale-to large'\n",
      "                        instead.\n",
      "  --ratio RATIO         Obsolete option. Originally designed to normalize\n",
      "                        treatment and control with customized ratio, now it\n",
      "                        won't have any effect.\n",
      "\n",
      "Other options:\n",
      "  --buffer-size BUFFER_SIZE\n",
      "                        Buffer size for incrementally increasing internal\n",
      "                        array size to store reads alignment information. In\n",
      "                        most cases, you don't have to change this parameter.\n",
      "                        However, if there are large number of\n",
      "                        chromosomes/contigs/scaffolds in your alignment, it's\n",
      "                        recommended to specify a smaller buffer size in order\n",
      "                        to decrease memory usage (but it will take longer time\n",
      "                        to read alignment files). Minimum memory requested for\n",
      "                        reading an alignment file is about # of CHROMOSOME *\n",
      "                        BUFFER_SIZE * 8 Bytes. DEFAULT: 100000\n",
      "\n",
      "Examples:\n",
      "1. Peak calling for regular TF ChIP-seq:\n",
      "    $ macs3 callpeak -t ChIP.bam -c Control.bam -f BAM -g hs -n test -B -q 0.01\n",
      "2. Broad peak calling on Histone Mark ChIP-seq:\n",
      "    $ macs3 callpeak -t ChIP.bam -c Control.bam --broad -g hs --broad-cutoff 0.1\n",
      "3. Peak calling on ATAC-seq (paired-end mode):\n",
      "    $ macs3 callpeak -f BAMPE -t ATAC.bam -g hs -n test -B -q 0.01\n",
      "4. Peak calling on ATAC-seq ( focusing on insertion sites, and using single-end mode):\n",
      "    $ macs3 callpeak -f BAM -t ATAC.bam -g hs -n test -B -q 0.01 --shift -50 --extension 100\n"
     ]
    }
   ],
   "source": [
    "! macs3 callpeak -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34ef50d5-10cd-4224-9b76-c8f1ec4b77ab",
   "metadata": {},
   "source": [
    "##### paramaters\n",
    "- --treatment, space delimited list of treated sample bam files\n",
    "- --control, space delimited list of control sample bam files\n",
    "- -f BAM, file type is BAM\n",
    "- -n fwd, prefix name for files\n",
    "- -g hs, genome size, where hs means Homo sapiens (human) ~ 2.7e9\n",
    "- --keep-dup all, how to handle duplicate tags aligned to the exact same position in the genome. all means don't discard any\n",
    "- --outdir {out_path}, path of output\n",
    "- --nomodel, do not build the shifting model from the data. The default is to model the shift size by considering distances reads positive and negative strands of reads\n",
    "- --extsize 30, extend reads from 5' end by 30"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a182d25-23e6-4820-9b4e-42d8719801d1",
   "metadata": {},
   "source": [
    "Need a space delimted list of samples for **mac3**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "af9ef914-16d6-48b1-b607-890d36c1c4e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "split_fwd/a4_1.bam split_fwd/a4_2.bam split_fwd/a4_3.bam split_fwd/a4_4.bam\n",
      "split_fwd/no_1.bam split_fwd/no_2.bam split_fwd/no_3.bam split_fwd/no_4.bam\n"
     ]
    }
   ],
   "source": [
    "print(samples_string(treated,in_path))\n",
    "print(samples_string(control,in_path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "ccdf9906-6d4d-4403-b6c8-c02e01f1002f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  @ 14 Oct 2023 00:13:43: [38 MB] \n",
      "# Command line: callpeak --treatment split_fwd/a4_1.bam split_fwd/a4_2.bam split_fwd/a4_3.bam split_fwd/a4_4.bam --control split_fwd/no_1.bam split_fwd/no_2.bam split_fwd/no_3.bam split_fwd/no_4.bam -f BAM -n fwd -g hs --keep-dup all --outdir peak_fwd --nomodel --extsize 30\n",
      "# ARGUMENTS LIST:\n",
      "# name = fwd\n",
      "# format = BAM\n",
      "# ChIP-seq file = ['split_fwd/a4_1.bam', 'split_fwd/a4_2.bam', 'split_fwd/a4_3.bam', 'split_fwd/a4_4.bam']\n",
      "# control file = ['split_fwd/no_1.bam', 'split_fwd/no_2.bam', 'split_fwd/no_3.bam', 'split_fwd/no_4.bam']\n",
      "# effective genome size = 2.91e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 5.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 1000 bps and 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is off\n",
      " \n",
      "INFO  @ 14 Oct 2023 00:13:43: [38 MB] #1 read tag files... \n",
      "INFO  @ 14 Oct 2023 00:13:43: [38 MB] #1 read treatment tags... \n",
      "INFO  @ 14 Oct 2023 00:13:44: [85 MB] 869510 reads have been read. \n",
      "INFO  @ 14 Oct 2023 00:13:44: [96 MB] 233566 reads have been read. \n",
      "INFO  @ 14 Oct 2023 00:13:44: [96 MB] 703905 reads have been read. \n",
      "INFO  @ 14 Oct 2023 00:13:45: [99 MB] 520846 reads have been read. \n",
      "INFO  @ 14 Oct 2023 00:13:45: [99 MB] #1.2 read input tags... \n",
      "INFO  @ 14 Oct 2023 00:13:45: [115 MB] 557795 reads have been read. \n",
      "INFO  @ 14 Oct 2023 00:13:46: [126 MB] 555966 reads have been read. \n",
      "INFO  @ 14 Oct 2023 00:13:46: [132 MB] 431273 reads have been read. \n",
      "INFO  @ 14 Oct 2023 00:13:47: [139 MB] 615633 reads have been read. \n",
      "INFO  @ 14 Oct 2023 00:13:47: [139 MB] #1 tag size is determined as 50 bps \n",
      "INFO  @ 14 Oct 2023 00:13:47: [139 MB] #1 tag size = 50.0 \n",
      "INFO  @ 14 Oct 2023 00:13:47: [139 MB] #1  total tags in treatment: 2327827 \n",
      "INFO  @ 14 Oct 2023 00:13:47: [139 MB] #1  total tags in control: 2160667 \n",
      "INFO  @ 14 Oct 2023 00:13:47: [139 MB] #1 finished! \n",
      "INFO  @ 14 Oct 2023 00:13:47: [139 MB] #2 Build Peak Model... \n",
      "INFO  @ 14 Oct 2023 00:13:47: [139 MB] #2 Skipped... \n",
      "INFO  @ 14 Oct 2023 00:13:47: [139 MB] #2 Use 30 as fragment length \n",
      "INFO  @ 14 Oct 2023 00:13:47: [139 MB] #3 Call peaks... \n",
      "INFO  @ 14 Oct 2023 00:13:47: [139 MB] #3 Pre-compute pvalue-qvalue table... \n",
      "INFO  @ 14 Oct 2023 00:13:52: [180 MB] #3 Call peaks for each chromosome... \n",
      "INFO  @ 14 Oct 2023 00:13:53: [180 MB] #4 Write output xls file... peak_fwd/fwd_peaks.xls \n",
      "INFO  @ 14 Oct 2023 00:13:53: [180 MB] #4 Write peak in narrowPeak format file... peak_fwd/fwd_peaks.narrowPeak \n",
      "INFO  @ 14 Oct 2023 00:13:53: [180 MB] #4 Write summits bed file... peak_fwd/fwd_summits.bed \n",
      "INFO  @ 14 Oct 2023 00:13:53: [180 MB] Done! \n"
     ]
    }
   ],
   "source": [
    "! macs3 callpeak                                  \\\n",
    "    --treatment {samples_string(treated,in_path)} \\\n",
    "    --control   {samples_string(control,in_path)} \\\n",
    "    -f BAM                                        \\\n",
    "    -n fwd                                        \\\n",
    "    -g hs                                         \\\n",
    "    --keep-dup all                                \\\n",
    "    --outdir {out_path}                           \\\n",
    "    --nomodel --extsize 30"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65c223a7",
   "metadata": {},
   "source": [
    "#### Mark 6th column with a **+** to indicate a `fwd` peak (sed is 0 based)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "833ba700",
   "metadata": {},
   "outputs": [],
   "source": [
    "!sed -i 's/\\t[^\\t]*/\\t+/5' {out_path}/*.narrowPeak"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "769377b5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t629931\t629961\tfwd_peak_1\t95\t+\t2.29353\t14.6044\t9.56511\t4\n",
      "chr1\t630717\t630752\tfwd_peak_2\t1066\t+\t2.4527\t112.872\t106.653\t10\n",
      "chr1\t634027\t634058\tfwd_peak_3\t142\t+\t1.36338\t19.4755\t14.2459\t1\n",
      "chr1\t2059577\t2059607\tfwd_peak_4\t93\t+\t7.27896\t14.374\t9.34693\t9\n",
      "chr1\t7780948\t7780978\tfwd_peak_5\t17\t+\t3.96276\t6.17368\t1.75758\t0\n",
      "chr1\t11022187\t11022217\tfwd_peak_6\t114\t+\t2.5373\t16.5636\t11.4402\t5\n",
      "chr1\t11908168\t11908218\tfwd_peak_7\t71\t+\t4.80762\t12.0029\t7.10054\t8\n",
      "chr1\t20651168\t20651198\tfwd_peak_8\t30\t+\t3.45782\t7.60524\t3.02988\t2\n",
      "chr1\t23693852\t23693908\tfwd_peak_9\t139\t+\t2.01178\t19.1214\t13.9047\t5\n",
      "chr1\t24669246\t24669316\tfwd_peak_10\t135\t+\t9.87106\t18.7609\t13.5553\t62\n"
     ]
    }
   ],
   "source": [
    "!head -10 {out_path}/*.narrowPeak"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f7c6bb7",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Step: peak_rev"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b778cfa-72a5-4830-acc8-744486e5752a",
   "metadata": {},
   "source": [
    "#### Call peaks from reverse reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "91b2d3d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_path   = mkpath(\"split_rev\")\n",
    "out_path  = mkpath(\"peak_rev\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "926abaf5-3aa8-4b42-a572-81765b61b470",
   "metadata": {},
   "source": [
    "#### macs3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7749eec4-9912-473c-b68a-3223a86ed090",
   "metadata": {},
   "source": [
    "##### paramaters\n",
    "- --treatment, space delimited list of treated sample bam files\n",
    "- --control, space delimited list of control sample bam files\n",
    "- -f BAM, file type is BAM\n",
    "- -n fwd, prefix name for files\n",
    "- -g hs, genome size, where hs means Homo sapiens (human) ~ 2.7e9\n",
    "- --keep-dup all, how to handle duplicate tags aligned to the exact same position in the genome. all means don't discard any\n",
    "- --outdir {out_path}, path of output\n",
    "- --nomodel, do not build the shifting model from the data. The default is to model the shift size by considering distances reads positive and negative strands of reads\n",
    "- --extsize 30, extend reads from 5' end by 30"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6f7f8b74-ddc4-4fb9-b2fd-be438e38a719",
   "metadata": {},
   "source": [
    "Need a space delimted list of samples for **mac3**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "be3f6268-9e8f-40d6-95f9-3b4fd7b8ef1f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "split_rev/a4_1.bam split_rev/a4_2.bam split_rev/a4_3.bam split_rev/a4_4.bam\n",
      "split_rev/no_1.bam split_rev/no_2.bam split_rev/no_3.bam split_rev/no_4.bam\n"
     ]
    }
   ],
   "source": [
    "print(samples_string(treated,in_path))\n",
    "print(samples_string(control,in_path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "fa6bad3c-7174-4433-8de5-b6f4e0a1877a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  @ 14 Oct 2023 00:14:38: [38 MB] \n",
      "# Command line: callpeak --treatment split_rev/a4_1.bam split_rev/a4_2.bam split_rev/a4_3.bam split_rev/a4_4.bam --control split_rev/no_1.bam split_rev/no_2.bam split_rev/no_3.bam split_rev/no_4.bam -f BAM -n fwd -g hs --keep-dup all --outdir peak_rev --nomodel --extsize 30\n",
      "# ARGUMENTS LIST:\n",
      "# name = fwd\n",
      "# format = BAM\n",
      "# ChIP-seq file = ['split_rev/a4_1.bam', 'split_rev/a4_2.bam', 'split_rev/a4_3.bam', 'split_rev/a4_4.bam']\n",
      "# control file = ['split_rev/no_1.bam', 'split_rev/no_2.bam', 'split_rev/no_3.bam', 'split_rev/no_4.bam']\n",
      "# effective genome size = 2.91e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 5.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 1000 bps and 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is off\n",
      " \n",
      "INFO  @ 14 Oct 2023 00:14:38: [38 MB] #1 read tag files... \n",
      "INFO  @ 14 Oct 2023 00:14:38: [38 MB] #1 read treatment tags... \n",
      "INFO  @ 14 Oct 2023 00:14:39: [76 MB] 609737 reads have been read. \n",
      "INFO  @ 14 Oct 2023 00:14:39: [87 MB] 146306 reads have been read. \n",
      "INFO  @ 14 Oct 2023 00:14:40: [90 MB] 559619 reads have been read. \n",
      "INFO  @ 14 Oct 2023 00:14:40: [92 MB] 398498 reads have been read. \n",
      "INFO  @ 14 Oct 2023 00:14:40: [92 MB] #1.2 read input tags... \n",
      "INFO  @ 14 Oct 2023 00:14:41: [111 MB] 420347 reads have been read. \n",
      "INFO  @ 14 Oct 2023 00:14:41: [121 MB] 441908 reads have been read. \n",
      "INFO  @ 14 Oct 2023 00:14:41: [125 MB] 336369 reads have been read. \n",
      "INFO  @ 14 Oct 2023 00:14:42: [126 MB] 474977 reads have been read. \n",
      "INFO  @ 14 Oct 2023 00:14:42: [126 MB] #1 tag size is determined as 45 bps \n",
      "INFO  @ 14 Oct 2023 00:14:42: [126 MB] #1 tag size = 45.0 \n",
      "INFO  @ 14 Oct 2023 00:14:42: [126 MB] #1  total tags in treatment: 1714160 \n",
      "INFO  @ 14 Oct 2023 00:14:42: [126 MB] #1  total tags in control: 1673601 \n",
      "INFO  @ 14 Oct 2023 00:14:42: [126 MB] #1 finished! \n",
      "INFO  @ 14 Oct 2023 00:14:42: [126 MB] #2 Build Peak Model... \n",
      "INFO  @ 14 Oct 2023 00:14:42: [126 MB] #2 Skipped... \n",
      "INFO  @ 14 Oct 2023 00:14:42: [126 MB] #2 Use 30 as fragment length \n",
      "INFO  @ 14 Oct 2023 00:14:42: [126 MB] #3 Call peaks... \n",
      "INFO  @ 14 Oct 2023 00:14:42: [126 MB] #3 Pre-compute pvalue-qvalue table... \n",
      "INFO  @ 14 Oct 2023 00:14:46: [174 MB] #3 Call peaks for each chromosome... \n",
      "INFO  @ 14 Oct 2023 00:14:46: [174 MB] #4 Write output xls file... peak_rev/fwd_peaks.xls \n",
      "INFO  @ 14 Oct 2023 00:14:46: [174 MB] #4 Write peak in narrowPeak format file... peak_rev/fwd_peaks.narrowPeak \n",
      "INFO  @ 14 Oct 2023 00:14:46: [174 MB] #4 Write summits bed file... peak_rev/fwd_summits.bed \n",
      "INFO  @ 14 Oct 2023 00:14:46: [174 MB] Done! \n"
     ]
    }
   ],
   "source": [
    "! macs3 callpeak                                  \\\n",
    "    --treatment {samples_string(treated,in_path)} \\\n",
    "    --control   {samples_string(control,in_path)} \\\n",
    "    -f BAM                                        \\\n",
    "    -n fwd                                        \\\n",
    "    -g hs                                         \\\n",
    "    --keep-dup all                                \\\n",
    "    --outdir {out_path}                           \\\n",
    "    --nomodel --extsize 30"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e5e0517-82a0-4ac1-9295-6cde3f0ea4f7",
   "metadata": {},
   "source": [
    "#### Mark 6th column with a **-** to indicate a `rev` peak (sed is 0 based)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "78b4e1b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "!sed -i 's/\\t[^\\t]*/\\t-/5' {out_path}/*.narrowPeak"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "1bba2765",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t1255454\t1255484\tfwd_peak_1\t132\t-\t11.2836\t18.3822\t13.26\t8\n",
      "chr1\t1543848\t1543880\tfwd_peak_2\t203\t-\t2.71643\t25.7131\t20.375\t22\n",
      "chr1\t1839165\t1839198\tfwd_peak_3\t562\t-\t2.14733\t62.1373\t56.2356\t26\n",
      "chr1\t8868002\t8868036\tfwd_peak_4\t823\t-\t2.84089\t88.4886\t82.3552\t27\n",
      "chr1\t11088120\t11088151\tfwd_peak_5\t53\t-\t2.23231\t10.1232\t5.3919\t28\n"
     ]
    }
   ],
   "source": [
    "!head -5 {out_path}/*.narrowPeak"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3742bac7",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Step: peak_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d473727b-1bce-4a7b-ab4b-767cd5ca1d40",
   "metadata": {},
   "source": [
    "#### Combine the forward and reverese reads and extend peak sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "65eea53c",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_fwd_path = mkpath(\"peak_fwd\")\n",
    "in_rev_path = mkpath(\"peak_rev\")\n",
    "out_path    = mkpath(\"peak_all\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "152cf773-d7cc-4538-9ec5-84ecf5c9d1ff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "all.narrowPeak\n"
     ]
    }
   ],
   "source": [
    "!ls {out_path}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0fc927c9-89c1-4fea-b760-4a3a2246eb29",
   "metadata": {},
   "source": [
    "##### Combine peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "8589f5eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_peaks    = out_path/\"all.narrowPeak\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "5fd925e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "!cat {in_fwd_path}/*peaks.narrowPeak {in_rev_path}/*peaks.narrowPeak > {all_peaks}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "8cc32762-abdf-43fb-9615-0da8adbf9dd6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1484 peak_all/all.narrowPeak\n"
     ]
    }
   ],
   "source": [
    "!  wc -l {all_peaks} "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2483650",
   "metadata": {},
   "source": [
    "##### Extend peaks by 20 nt in each direction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "id": "a5fb8d9a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629931</td>\n",
       "      <td>629961</td>\n",
       "      <td>fwd_peak_1</td>\n",
       "      <td>95</td>\n",
       "      <td>+</td>\n",
       "      <td>2.29353</td>\n",
       "      <td>14.6044</td>\n",
       "      <td>9.56511</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630717</td>\n",
       "      <td>630752</td>\n",
       "      <td>fwd_peak_2</td>\n",
       "      <td>1066</td>\n",
       "      <td>+</td>\n",
       "      <td>2.45270</td>\n",
       "      <td>112.8720</td>\n",
       "      <td>106.65300</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0       1       2           3     4  5        6         7          8   9\n",
       "0  chr1  629931  629961  fwd_peak_1    95  +  2.29353   14.6044    9.56511   4\n",
       "1  chr1  630717  630752  fwd_peak_2  1066  +  2.45270  112.8720  106.65300  10"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(all_peaks, header = None, sep='\\t')\n",
    "df[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "id": "41c4b3f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629911</td>\n",
       "      <td>629981</td>\n",
       "      <td>fwd_peak_1</td>\n",
       "      <td>95</td>\n",
       "      <td>+</td>\n",
       "      <td>2.29353</td>\n",
       "      <td>14.6044</td>\n",
       "      <td>9.56511</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630697</td>\n",
       "      <td>630772</td>\n",
       "      <td>fwd_peak_2</td>\n",
       "      <td>1066</td>\n",
       "      <td>+</td>\n",
       "      <td>2.45270</td>\n",
       "      <td>112.8720</td>\n",
       "      <td>106.65300</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0       1       2           3     4  5        6         7          8   9\n",
       "0  chr1  629911  629981  fwd_peak_1    95  +  2.29353   14.6044    9.56511   4\n",
       "1  chr1  630697  630772  fwd_peak_2  1066  +  2.45270  112.8720  106.65300  10"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[1] = df[1]-20\n",
    "df[2] = df[2]+20\n",
    "df.to_csv(all_peaks, sep='\\t', header=False,index=False)\n",
    "df = pd.read_csv(all_peaks, header = None, sep='\\t')\n",
    "df[:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "id": "c5baa235-a81f-462c-8275-d5785bf13671",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chr1\t629911\t629981\tfwd_peak_1\t95\t+\t2.29353\t14.6044\t9.56511\t4\n",
      "chr1\t630697\t630772\tfwd_peak_2\t1066\t+\t2.4527\t112.872\t106.653\t10\n",
      "chr1\t634007\t634078\tfwd_peak_3\t142\t+\t1.36338\t19.4755\t14.2459\t1\n",
      "chr1\t2059557\t2059627\tfwd_peak_4\t93\t+\t7.27896\t14.374\t9.34693\t9\n",
      "chr1\t7780928\t7780998\tfwd_peak_5\t17\t+\t3.96276\t6.17368\t1.75758\t0\n",
      "chr1\t11022167\t11022237\tfwd_peak_6\t114\t+\t2.5373\t16.5636\t11.4402\t5\n",
      "chr1\t11908148\t11908238\tfwd_peak_7\t71\t+\t4.80762\t12.0029\t7.10054\t8\n",
      "chr1\t20651148\t20651218\tfwd_peak_8\t30\t+\t3.45782\t7.60524\t3.02988\t2\n",
      "chr1\t23693832\t23693928\tfwd_peak_9\t139\t+\t2.01178\t19.1214\t13.9047\t5\n",
      "chr1\t24669226\t24669336\tfwd_peak_10\t135\t+\t9.87106\t18.7609\t13.5553\t62\n"
     ]
    }
   ],
   "source": [
    "!head -10 {all_peaks}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7add86d0",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "## Step: call motif"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "754e098d-2092-4e4b-bea8-01ed2379fa00",
   "metadata": {},
   "source": [
    "#### Find motifs from peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "id": "e416882f",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_path =  mkpath(\"peak_all\")\n",
    "out_path = mkpath(\"motif\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2b0bb2c-3574-48d1-ad61-4c320dcf9eb8",
   "metadata": {},
   "source": [
    "#### Homer"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54cc177f-854c-4974-a5f4-b046f5322be6",
   "metadata": {},
   "source": [
    "http://homer.ucsd.edu/homer/introduction/install.html"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2ccc580-2c83-4e2d-b457-9c8ea483643b",
   "metadata": {},
   "source": [
    "Install **homer** :\n",
    "- wget http://homer.ucsd.edu/homer/configureHomer.pl\n",
    "- mv configureHomer.pl homer/.\n",
    "- chmod +x configureHomer.pl\n",
    "- ./configureHomer.pl -install\n",
    "- ./configureHomer.pl -install hg19\n",
    "- Update .bashrc to include homer/run in $PATH"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "id": "170c8a94",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\tProgram will find de novo and known motifs in regions in the genome\n",
      "\n",
      "\tUsage: findMotifsGenome.pl <pos file> <genome> <output directory> [additional options]\n",
      "\tExample: findMotifsGenome.pl peaks.txt mm8r peakAnalysis -size 200 -len 8\n",
      "\n",
      "\tPossible Genomes:\n",
      "\t\thg38\thuman\n",
      "\t\t\t-- or --\n",
      "\t\tCustom: provide the path to genome FASTA files (directory or single file)\n",
      "\t\t\tHeads up: will create the directory \"preparsed/\" in same location.\n",
      "\n",
      "\tBasic options:\n",
      "\t\t-mask (mask repeats/lower case sequence, can also add 'r' to genome, i.e. mm9r)\n",
      "\t\t-bg <background position file> (genomic positions to be used as background, default=automatic)\n",
      "\t\t\tremoves background positions overlapping with target positions unless -keepOverlappingBg is used\n",
      "\t\t\t-chopify (chop up large background regions to the avg size of target regions)\n",
      "\t\t-len <#>[,<#>,<#>...] (motif length, default=8,10,12) [NOTE: values greater 12 may cause the program\n",
      "\t\t\tto run out of memory - in these cases decrease the number of sequences analyzed (-N),\n",
      "\t\t\tor try analyzing shorter sequence regions (i.e. -size 100)]\n",
      "\t\t-size <#> (fragment size to use for motif finding, default=200)\n",
      "\t\t\t-size <#,#> (i.e. -size -100,50 will get sequences from -100 to +50 relative from center)\n",
      "\t\t\t-size given (uses the exact regions you give it)\n",
      "\t\t-S <#> (Number of motifs to optimize, default: 25)\n",
      "\t\t-mis <#> (global optimization: searches for strings with # mismatches, default: 2)\n",
      "\t\t-norevopp (don't search reverse strand for motifs)\n",
      "\t\t-nomotif (don't search for de novo motif enrichment)\n",
      "\t\t-rna (output RNA motif logos and compare to RNA motif database, automatically sets -norevopp)\n",
      "\n",
      "\tScanning sequence for motifs\n",
      "\t\t-find <motif file> (This will cause the program to only scan for motifs)\n",
      "\n",
      "\tKnown Motif Options/Visualization\n",
      "\t\t-mset <vertebrates|insects|worms|plants|yeast|all> (check against motif collects, default: auto)\n",
      "\t\t-basic (just visualize de novo motifs, don't check similarity with known motifs)\n",
      "\t\t-bits (scale sequence logos by information content, default: doesn't scale)\n",
      "\t\t-nocheck (don't search for de novo vs. known motif similarity)\n",
      "\t\t-mcheck <motif file> (known motifs to check against de novo motifs,\n",
      "\t\t-float (allow adjustment of the degeneracy threshold for known motifs to improve p-value[dangerous])\n",
      "\t\t-noknown (don't search for known motif enrichment, default: -known)\n",
      "\t\t-mknown <motif file> (known motifs to check for enrichment,\n",
      "\t\t-nofacts (omit humor)\n",
      "\t\t-seqlogo (use weblogo/seqlogo/ghostscript to generate logos, default uses SVG now)\n",
      "\n",
      "\tSequence normalization options:\n",
      "\t\t-gc (use GC% for sequence content normalization, now the default)\n",
      "\t\t-cpg (use CpG% instead of GC% for sequence content normalization)\n",
      "\t\t-noweight (no CG correction)\n",
      "\t\tAlso -nlen <#>, -olen <#>, see homer2 section below.\n",
      "\n",
      "\tAdvanced options:\n",
      "\t\t-h (use hypergeometric for p-values, binomial is default)\n",
      "\t\t-N <#> (Number of sequences to use for motif finding, default=max(50k, 2x input)\n",
      "\t\t-local <#> (use local background, # of equal size regions around peaks to use i.e. 2)\n",
      "\t\t-redundant <#> (Remove redundant sequences matching greater than # percent, i.e. -redundant 0.5)\n",
      "\t\t-maxN <#> (maximum percentage of N's in sequence to consider for motif finding, default: 0.7)\n",
      "\t\t-maskMotif <motif file1> [motif file 2]... (motifs to mask before motif finding)\n",
      "\t\t-opt <motif file1> [motif file 2]... (motifs to optimize or change length of)\n",
      "\t\t-rand (randomize target and background sequences labels)\n",
      "\t\t-ref <peak file> (use file for target and background - first argument is list of peak ids for targets)\n",
      "\t\t-oligo (perform analysis of individual oligo enrichment)\n",
      "\t\t-dumpFasta (Dump fasta files for target and background sequences for use with other programs)\n",
      "\t\t-preparse (force new background files to be created)\n",
      "\t\t-preparsedDir <directory> (location to search for preparsed file and/or place new files)\n",
      "\t\t-keepFiles (keep temporary files)\n",
      "\t\t-fdr <#> (Calculate empirical FDR for de novo discovery #=number of randomizations)\n",
      "\n",
      "\thomer2 specific options:\n",
      "\t\t-homer2 (use homer2 instead of original homer, default)\n",
      "\t\t-nlen <#> (length of lower-order oligos to normalize in background, default: -nlen 3)\n",
      "\t\t\t-nmax <#> (Max normalization iterations, default: 160)\n",
      "\t\t\t-neutral (weight sequences to neutral frequencies, i.e. 25%, 6.25%, etc.)\n",
      "\t\t-olen <#> (lower-order oligo normalization for oligo table, use if -nlen isn't working well)\n",
      "\t\t-p <#> (Number of processors to use, default: 1)\n",
      "\t\t-e <#> (Maximum expected motif instance per bp in random sequence, default: 0.01)\n",
      "\t\t-cache <#> (size in MB for statistics cache, default: 500)\n",
      "\t\t-quickMask (skip full masking after finding motifs, similar to original homer)\n",
      "\t\t-minlp <#> (stop looking for motifs when seed logp score gets above #, default: -10)\n",
      "\n",
      "\tOriginal homer specific options:\n",
      "\t\t-homer1 (to force the use of the original homer)\n",
      "\t\t-depth [low|med|high|allnight] (time spent on local optimization default: med)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!findMotifsGenome.pl -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3fc7490-9498-40eb-9004-6eb26fa93f78",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "##### paramaters\n",
    "- {in_fn}, file containing all.narrowPeak\n",
    "- {out_path}, output dir\n",
    "- hg38, use hg38 human genome\n",
    "- p {nc}, number of processors\n",
    "- rna, output RNA motif logos and compare to RNA motif database, automatically sets -norevopp\n",
    "- S 10, number of motifs to optimize, default: 20\n",
    "- -len 5,6,7,8,9,10, motif length, default=8,10,12\n",
    "- bg {genome_bed12}, background position file; removes background positions overlapping with target positions "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "id": "92a6605b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\tPosition file = peak_all/all.narrowPeak\n",
      "\tGenome = hg38\n",
      "\tOutput Directory = motif\n",
      "\tUsing 32 CPUs\n",
      "\tOperating in RNA mode\n",
      "\tMotif length set at 5,6,7,8,9,\n",
      "\tbackground position file: /home/cdaniels/uofc_data/ref/homo_sapian_39/gencode.v39.annotation.bed12\n",
      "\tFound mset for \"human\", will check against vertebrates motifs\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 1484\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 1484\n",
      "\t\tRedundant Peak IDs: 687\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tRedunant Peaks found: Remove or rename these or some programs may have trouble...\n",
      "\n",
      "\t687 duplicate peak IDs out of 1484 total peaks\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 244939\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\tMax distance to merge: 200 bp\n",
      "\tCalculating co-bound peaks relative to reference: motif/bg.clean.pos\n",
      "\n",
      "\tComparing peaks: (peakfile, overlapping peaks, logRatio(obs/expected), logP)\n",
      "\t**\tmotif/target.clean..pos\t338\t2.06\t-433.38\n",
      "\n",
      "\t**\tPairwise stats are approx with fixed distance (-d 200) and -cobound #\n",
      "\t\tThey get worse as the size increases and peaks from a single file start overlapping\n",
      "\t\tTo get accurate ones, adjust peak sizes first with adjustPeakFile.pl and\n",
      "\t\tthen rerun mergePeaks with the \"-d given\" option (only applies to -cobound #)\n",
      "\n",
      "\tCo-bound by 0 peaks: 244601\n",
      "\tCo-bound by 1 peaks: 338 (max: 338 effective total)\n",
      "\tCustom genome sequence directory: /home/cdaniels/bin/homer/.//data/genomes/hg38//\n",
      "\n",
      "\tExtracting sequences from file: /home/cdaniels/bin/homer/.//data/genomes/hg38///genome.fa\n",
      "\tLooking for peak sequences in a single file (/home/cdaniels/bin/homer/.//data/genomes/hg38///genome.fa)\n",
      "\tExtracting 21671 sequences from chr1\n",
      "\tExtracting 8858 sequences from chr10\n",
      "\tExtracting 14536 sequences from chr11\n",
      "\tExtracting 13567 sequences from chr12\n",
      "\tExtracting 4587 sequences from chr13\n",
      "\tExtracting 8874 sequences from chr14\n",
      "\tExtracting 8939 sequences from chr15\n",
      "\tExtracting 11483 sequences from chr16\n",
      "\tExtracting 14234 sequences from chr17\n",
      "\tExtracting 4748 sequences from chr18\n",
      "\tExtracting 14103 sequences from chr19\n",
      "\tExtracting 18316 sequences from chr2\n",
      "\tExtracting 5868 sequences from chr20\n",
      "\tExtracting 3125 sequences from chr21\n",
      "\tExtracting 5267 sequences from chr22\n",
      "\tExtracting 15501 sequences from chr3\n",
      "\tExtracting 10104 sequences from chr4\n",
      "\tExtracting 11676 sequences from chr5\n",
      "\tExtracting 11471 sequences from chr6\n",
      "\tExtracting 11761 sequences from chr7\n",
      "\tExtracting 10054 sequences from chr8\n",
      "\tExtracting 8486 sequences from chr9\n",
      "\tExtracting 57 sequences from chrM\n",
      "\tExtracting 7788 sequences from chrX\n",
      "\tExtracting 1011 sequences from chrY\n",
      "\t\t45 of 246085 removed because they had >70.00% Ns (i.e. masked repeats)\n",
      "\n",
      "\tNot removing redundant sequences\n",
      "\n",
      "\n",
      "\tSequences processed:\n",
      "\t\tAuto detected maximum sequence length of 201 bp\n",
      "\t\t100000\n",
      "\t\t200000\n",
      "\t\t246040 total\n",
      "\n",
      "\tFrequency Bins: 0.2 0.25 0.3 0.35 0.4 0.45 0.5 0.6 0.7 0.8\n",
      "\tFreq\tBin\tCount\n",
      "\t0.2\t0\t301\n",
      "\t0.25\t1\t2544\n",
      "\t0.3\t2\t13271\n",
      "\t0.35\t3\t29294\n",
      "\t0.4\t4\t38037\n",
      "\t0.45\t5\t40475\n",
      "\t0.5\t6\t40310\n",
      "\t0.6\t7\t58963\n",
      "\t0.7\t8\t20053\n",
      "\t0.8\t9\t2603\n",
      "\t10\t10\t189\n",
      "\tBin\t# Targets\t# Background\tBackground Weight\n",
      "\t1\t6\t2538\t0.389\n",
      "\t2\t59\t13212\t0.735\n",
      "\t3\t205\t29089\t1.160\n",
      "\t4\t336\t37701\t1.467\n",
      "\t5\t281\t40194\t1.151\n",
      "\t6\t230\t40080\t0.944\n",
      "\t7\t276\t58687\t0.774\n",
      "\t8\t76\t19977\t0.626\n",
      "\t9\t14\t2589\t0.890\n",
      "\tNormalizing lower order oligos using homer2\n",
      "\n",
      "\tReading input files...\n",
      "\t245550 total sequences read\n",
      "\tAutonormalization: 1-mers (4 total)\n",
      "\t\tA\t29.26%\t27.21%\t1.075\n",
      "\t\tC\t21.16%\t21.42%\t0.988\n",
      "\t\tG\t22.68%\t22.32%\t1.016\n",
      "\t\tT\t26.90%\t29.05%\t0.926\n",
      "\tAutonormalization: 2-mers (16 total)\n",
      "\t\tAA\t9.57%\t8.41%\t1.139\n",
      "\t\tCA\t7.05%\t7.02%\t1.005\n",
      "\t\tGA\t7.24%\t6.07%\t1.194\n",
      "\t\tTA\t5.40%\t5.75%\t0.939\n",
      "\t\tAC\t5.29%\t4.81%\t1.098\n",
      "\t\tCC\t5.61%\t5.80%\t0.967\n",
      "\t\tGC\t4.65%\t4.83%\t0.964\n",
      "\t\tTC\t5.61%\t5.97%\t0.939\n",
      "\t\tAG\t7.77%\t7.23%\t1.075\n",
      "\t\tCG\t2.02%\t1.31%\t1.542\n",
      "\t\tGG\t5.73%\t6.02%\t0.950\n",
      "\t\tTG\t7.15%\t7.77%\t0.921\n",
      "\t\tAT\t6.63%\t6.79%\t0.976\n",
      "\t\tCT\t6.48%\t7.28%\t0.889\n",
      "\t\tGT\t5.06%\t5.42%\t0.934\n",
      "\t\tTT\t8.73%\t9.52%\t0.917\n",
      "\tAutonormalization: 3-mers (64 total)\n",
      "\tNormalization weights can be found in file: motif/seq.autonorm.tsv\n",
      "\tConverging on autonormalization solution:\n",
      "\t...............................................................................\n",
      "\tFinal normalization:\tAutonormalization: 1-mers (4 total)\n",
      "\t\tA\t29.26%\t29.06%\t1.007\n",
      "\t\tC\t21.16%\t21.19%\t0.999\n",
      "\t\tG\t22.68%\t22.76%\t0.997\n",
      "\t\tT\t26.90%\t26.99%\t0.997\n",
      "\tAutonormalization: 2-mers (16 total)\n",
      "\t\tAA\t9.57%\t9.51%\t1.007\n",
      "\t\tCA\t7.05%\t7.12%\t0.991\n",
      "\t\tGA\t7.24%\t6.94%\t1.044\n",
      "\t\tTA\t5.40%\t5.50%\t0.981\n",
      "\t\tAC\t5.29%\t5.15%\t1.026\n",
      "\t\tCC\t5.61%\t5.65%\t0.993\n",
      "\t\tGC\t4.65%\t4.80%\t0.970\n",
      "\t\tTC\t5.61%\t5.59%\t1.003\n",
      "\t\tAG\t7.77%\t7.74%\t1.005\n",
      "\t\tCG\t2.02%\t1.81%\t1.118\n",
      "\t\tGG\t5.73%\t5.95%\t0.962\n",
      "\t\tTG\t7.15%\t7.25%\t0.986\n",
      "\t\tAT\t6.63%\t6.67%\t0.994\n",
      "\t\tCT\t6.48%\t6.61%\t0.980\n",
      "\t\tGT\t5.06%\t5.07%\t0.999\n",
      "\t\tTT\t8.73%\t8.64%\t1.010\n",
      "\tAutonormalization: 3-mers (64 total)\n",
      "\tFinished preparing sequence/group files\n",
      "\n",
      "\t----------------------------------------------------------\n",
      "\tKnown motif enrichment\n",
      "\t----------------------------------------------------------\n",
      "\tDe novo motif finding (HOMER)\n",
      "\n",
      "\tNumber of Trial motifs (-T) set to 32 (from 10) to work well with 32 CPUs\n",
      "\tScanning input files...\n",
      "\tParsing sequences...\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\tTotal number of Oligos: 1024\n",
      "\tAutoadjustment for sequence coverage in background: 1.00x\n",
      "\n",
      "\tOligos: 1024 of 1034 max\n",
      "\tTree  : 1364 of 5170 max\n",
      "\tOptimizing memory usage...\n",
      "\tCache length = 11180\n",
      "\tUsing binomial scoring\n",
      "\n",
      "\tGlobal Optimization Phase: Looking for enriched oligos with up to 2 mismatches...\n",
      "\n",
      "\tScreening oligos 1024 (allowing 0 mismatches):\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t=====================================================================================\n",
      "\t\t0.00% skipped, 100.00% checked (1024 of 1024), of those checked:\n",
      "\t\t0.00% not in target, 0.00% increased p-value, 0.00% high p-value\n",
      "\n",
      "\tScreening oligos 1024 (allowing 1 mismatches):\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t=====================================================================================\n",
      "\t\t0.00% skipped, 100.00% checked (1024 of 1024), of those checked:\n",
      "\t\t0.00% not in target, 54.39% increased p-value, 54.00% high p-value\n",
      "\n",
      "\tScreening oligos 1024 (allowing 2 mismatches):\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t=====================================================================================\n",
      "\t\t86.62% skipped, 13.38% checked (137 of 1024), of those checked:\n",
      "\t\t0.00% not in target, 2.34% increased p-value, 0.00% high p-value\n",
      "\tReading input files...\n",
      "\t245550 total sequences read\n",
      "\tCache length = 11180\n",
      "\tUsing binomial scoring\n",
      "\n",
      "\tLocal Optimization Phase:\n",
      "\t1 of 25 Initial Sequence: GAAGA... (-134.585)\n",
      "\t\tRound 1: -134.59 GAAGA T:1303.0(58.48%),B:103735.9(37.65%),P:1e-58\n",
      "\t\tRound 2: -134.59 GAAGA T:1303.0(58.48%),B:103735.9(37.65%),P:1e-58\n",
      "\t\t=Final=: -104.92 GAAGA T:762.0(51.38%),B:73260.4(33.37%),P:1e-45\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t2 of 25 Initial Sequence: GGTAA... (-88.080)\n",
      "\t\tRound 1: -88.08 GGTAA T:462.0(26.78%),B:32774.0(13.87%),P:1e-38\n",
      "\t\tRound 2: -88.08 GGTAA T:462.0(26.78%),B:32774.0(13.87%),P:1e-38\n",
      "\t\t=Final=: -101.07 GGTAA T:409.0(27.58%),B:30079.7(13.70%),P:1e-43\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t3 of 25 Initial Sequence: CGGAA... (-30.221)\n",
      "\t\tRound 1: -40.76 CGGAA T:351.0(21.08%),B:30289.9(12.89%),P:1e-17\n",
      "\t\tRound 2: -40.76 CGGAA T:351.0(21.08%),B:30289.9(12.89%),P:1e-17\n",
      "\t\t=Final=: -38.65 CGGAA T:296.0(19.96%),B:26800.7(12.21%),P:1e-16\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t4 of 25 Initial Sequence: GTAAG... (-21.980)\n",
      "\t\tRound 1: -28.45 GTAAG T:565.0(31.69%),B:58689.8(23.46%),P:1e-12\n",
      "\t\tRound 2: -41.43 GTACG T:823.0(42.60%),B:83631.5(31.68%),P:1e-17\n",
      "\t\tRound 3: -54.28 GATCG T:791.0(41.35%),B:75179.5(28.99%),P:1e-23\n",
      "\t\tRound 4: -54.28 GATCG T:791.0(41.35%),B:75179.5(28.99%),P:1e-23\n",
      "\t\t=Final=: -50.48 GATCG T:566.0(38.17%),B:58299.7(26.55%),P:1e-21\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t5 of 25 Initial Sequence: GATGA... (-19.742)\n",
      "\t\tRound 1: -27.45 GATGA T:1109.0(52.67%),B:125584.2(43.56%),P:1e-11\n",
      "\t\tRound 2: -30.27 GACGA T:1209.0(55.76%),B:135600.9(46.08%),P:1e-13\n",
      "\t\tRound 3: -30.27 GACGA T:1209.0(55.76%),B:135600.9(46.08%),P:1e-13\n",
      "\t\t=Final=: -18.07 GACGA T:454.0(30.61%),B:53225.2(24.24%),P:1e-7\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t6 of 25 Initial Sequence: CTGCT... (-12.655)\n",
      "\t\tRound 1: -15.54 CTGCT T:581.0(32.42%),B:67254.1(26.38%),P:1e-6\n",
      "\t\tRound 2: -15.54 CTGCT T:581.0(32.42%),B:67254.1(26.38%),P:1e-6\n",
      "\t\t=Final=: -14.42 CTGCT T:442.0(29.80%),B:53184.7(24.22%),P:1e-6\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t7 of 25 Initial Sequence: TCCGA... (-12.021)\n",
      "\t\tRound 1: -19.38 TCCGA T:224.0(14.02%),B:21411.5(9.29%),P:1e-8\n",
      "\t\tRound 2: -19.38 TCCGA T:224.0(14.02%),B:21411.5(9.29%),P:1e-8\n",
      "\t\t=Final=: -19.93 TCCGA T:205.0(13.82%),B:20013.4(9.12%),P:1e-8\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t8 of 25 Initial Sequence: CTTTT... (-10.079)\n",
      "\t\tRound 1: -10.08 CTTTT T:746.0(39.54%),B:93160.1(34.58%),P:1e-4\n",
      "\t\tRound 2: -10.08 CTTTT T:746.0(39.54%),B:93160.1(34.58%),P:1e-4\n",
      "\t\t=Final=: -15.36 CTTTT T:558.0(37.63%),B:68948.1(31.40%),P:1e-6\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t9 of 25 Initial Sequence: GAGGA... (-9.941)\n",
      "\t\tRound 1: -15.00 GAGSA T:922.0(46.31%),B:111561.1(39.84%),P:1e-6\n",
      "\t\tRound 2: -15.00 GAGSA T:922.0(46.31%),B:111561.1(39.84%),P:1e-6\n",
      "\t\t=Final=: -17.39 GAGSA T:639.0(43.09%),B:79478.9(36.20%),P:1e-7\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t10 of 25 Initial Sequence: GATGC... (-10.016)\n",
      "\t\tRound 1: -10.02 GATGC T:244.0(15.18%),B:27432.3(11.75%),P:1e-4\n",
      "\t\tRound 2: -10.02 GATGC T:244.0(15.18%),B:27432.3(11.75%),P:1e-4\n",
      "\t\t=Final=: -10.82 GATGC T:223.0(15.04%),B:25190.6(11.47%),P:1e-4\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t11 of 25 Initial Sequence: GTATG... (-9.774)\n",
      "\t\tRound 1: -25.46 GTAAG T:493.0(28.29%),B:51362.9(20.86%),P:1e-11\n",
      "\t\tRound 2: -25.46 GTAAG T:493.0(28.29%),B:51362.9(20.86%),P:1e-11\n",
      "\t\t=Final=: -33.05 GTAAG T:423.0(28.52%),B:44048.4(20.06%),P:1e-14\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\tRemaining seeds don't look promising (After initial 5 motifs, logp -9.008 > -9.171)\n",
      "\n",
      "\tFinalizing Enrichment Statistics (new in v3.4)\n",
      "\tReading input files...\n",
      "\t245550 total sequences read\n",
      "\tCache length = 11180\n",
      "\tUsing binomial scoring\n",
      "\tChecking enrichment of 11 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tOutput in file: motif/homerMotifs.motifs5\n",
      "\n",
      "\n",
      "\tNumber of Trial motifs (-T) set to 32 (from 10) to work well with 32 CPUs\n",
      "\tScanning input files...\n",
      "\tParsing sequences...\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\tTotal number of Oligos: 4096\n",
      "\tAutoadjustment for sequence coverage in background: 1.00x\n",
      "\n",
      "\tOligos: 4096 of 4106 max\n",
      "\tTree  : 5460 of 20530 max\n",
      "\tOptimizing memory usage...\n",
      "\tCache length = 11180\n",
      "\tUsing binomial scoring\n",
      "\n",
      "\tGlobal Optimization Phase: Looking for enriched oligos with up to 2 mismatches...\n",
      "\n",
      "\tScreening oligos 4096 (allowing 0 mismatches):\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\t\t0.00% skipped, 100.00% checked (4096 of 4096), of those checked:\n",
      "\t\t0.00% not in target, 0.00% increased p-value, 0.00% high p-value\n",
      "\n",
      "\tScreening oligos 4096 (allowing 1 mismatches):\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\t\t0.00% skipped, 100.00% checked (4096 of 4096), of those checked:\n",
      "\t\t0.00% not in target, 50.32% increased p-value, 50.39% high p-value\n",
      "\n",
      "\tScreening oligos 4096 (allowing 2 mismatches):\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\t\t72.00% skipped, 28.00% checked (1147 of 4096), of those checked:\n",
      "\t\t0.00% not in target, 13.84% increased p-value, 0.00% high p-value\n",
      "\tReading input files...\n",
      "\t245550 total sequences read\n",
      "\tCache length = 11180\n",
      "\tUsing binomial scoring\n",
      "\n",
      "\tLocal Optimization Phase:\n",
      "\t1 of 25 Initial Sequence: CGAAGA... (-96.941)\n",
      "\t\tRound 1: -133.09 NGAAGA T:1294.0(58.22%),B:103232.1(37.51%),P:1e-57\n",
      "\t\tRound 2: -133.09 NGAAGA T:1294.0(58.22%),B:103232.1(37.51%),P:1e-57\n",
      "\t\t=Final=: -104.08 NGAAGA T:759.0(51.18%),B:73004.8(33.25%),P:1e-45\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t2 of 25 Initial Sequence: GTAAGT... (-89.305)\n",
      "\t\tRound 1: -107.29 GTAAGT T:352.0(21.14%),B:20243.3(8.81%),P:1e-46\n",
      "\t\tRound 2: -107.29 GTAAGT T:352.0(21.14%),B:20243.3(8.81%),P:1e-46\n",
      "\t\t=Final=: -137.12 GTAAGT T:337.0(22.72%),B:18986.9(8.65%),P:1e-59\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t3 of 25 Initial Sequence: AGGTAA... (-62.794)\n",
      "\t\tRound 1: -71.28 AGGTAA T:317.0(19.25%),B:21423.7(9.30%),P:1e-30\n",
      "\t\tRound 2: -71.28 AGGTAA T:317.0(19.25%),B:21423.7(9.30%),P:1e-30\n",
      "\t\t=Final=: -81.66 AGGTAA T:296.0(19.96%),B:20315.1(9.25%),P:1e-35\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t4 of 25 Initial Sequence: CGACGA... (-35.106)\n",
      "\t\tRound 1: -43.97 CGACGA T:598.0(33.19%),B:56961.1(22.85%),P:1e-19\n",
      "\t\tRound 2: -43.97 CGACGA T:598.0(33.19%),B:56961.1(22.85%),P:1e-19\n",
      "\t\t=Final=: -42.34 CGACGA T:438.0(29.53%),B:43587.1(19.85%),P:1e-18\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t5 of 25 Initial Sequence: CGGAAA... (-24.457)\n",
      "\t\tRound 1: -44.11 CGGAAA T:187.0(11.85%),B:12698.7(5.62%),P:1e-19\n",
      "\t\tRound 2: -44.11 CGGAAA T:187.0(11.85%),B:12698.7(5.62%),P:1e-19\n",
      "\t\t=Final=: -44.53 CGGAAA T:173.0(11.67%),B:12078.2(5.50%),P:1e-19\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t6 of 25 Initial Sequence: GAGGAG... (-19.747)\n",
      "\t\tRound 1: -20.37 GAGGAG T:222.0(13.91%),B:20997.3(9.12%),P:1e-8\n",
      "\t\tRound 2: -20.37 GAGGAG T:222.0(13.91%),B:20997.3(9.12%),P:1e-8\n",
      "\t\t=Final=: -15.37 GAGGAG T:175.0(11.80%),B:17535.0(7.99%),P:1e-6\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t7 of 25 Initial Sequence: GGTTCG... (-18.591)\n",
      "\t\tRound 1: -30.92 GGTACG T:162.0(10.35%),B:12165.5(5.39%),P:1e-13\n",
      "\t\tRound 2: -30.92 GGTACG T:162.0(10.35%),B:12165.5(5.39%),P:1e-13\n",
      "\t\t=Final=: -33.74 GGTACG T:156.0(10.52%),B:11739.3(5.35%),P:1e-14\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t8 of 25 Initial Sequence: GCACAC... (-15.733)\n",
      "\t\tRound 1: -22.45 GCRCAC T:219.0(13.73%),B:20046.9(8.73%),P:1e-9\n",
      "\t\tRound 2: -22.45 GCRCAC T:219.0(13.73%),B:20046.9(8.73%),P:1e-9\n",
      "\t\t=Final=: -10.17 GCRCAC T:183.0(12.34%),B:20216.9(9.21%),P:1e-4\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t9 of 25 Initial Sequence: CATCGA... (-15.507)\n",
      "\t\tRound 1: -29.71 CATCGA T:799.0(41.66%),B:86270.6(32.49%),P:1e-12\n",
      "\t\tRound 2: -29.71 CATCGA T:799.0(41.66%),B:86270.6(32.49%),P:1e-12\n",
      "\t\t=Final=: -35.47 CATCGA T:564.0(38.03%),B:62133.6(28.30%),P:1e-15\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t10 of 25 Initial Sequence: ATCCGA... (-12.901)\n",
      "\t\tRound 1: -21.15 ATCCGA T:151.0(9.68%),B:12750.9(5.64%),P:1e-9\n",
      "\t\tRound 2: -21.15 ATCCGA T:151.0(9.68%),B:12750.9(5.64%),P:1e-9\n",
      "\t\t=Final=: -19.94 ATCCGA T:123.0(8.29%),B:10346.4(4.71%),P:1e-8\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t11 of 25 Initial Sequence: GCCAAT... (-12.211)\n",
      "\t\tRound 1: -16.36 GCCAAC T:902.0(45.58%),B:107778.0(38.79%),P:1e-7\n",
      "\t\tRound 2: -16.36 GCCAAC T:902.0(45.58%),B:107778.0(38.79%),P:1e-7\n",
      "\t\t=Final=: -16.27 GCCAAC T:660.0(44.50%),B:83046.1(37.83%),P:1e-7\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t12 of 25 Initial Sequence: TACTAC... (-12.991)\n",
      "\t\tRound 1: -24.55 TACTAC T:402.0(23.75%),B:40864.5(16.98%),P:1e-10\n",
      "\t\tRound 2: -24.55 TACTAC T:402.0(23.75%),B:40864.5(16.98%),P:1e-10\n",
      "\t\t=Final=: -22.88 TACTAC T:334.0(22.52%),B:35462.7(16.15%),P:1e-9\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t13 of 25 Initial Sequence: TGCCAA... (-10.514)\n",
      "\t\tRound 1: -14.55 TGCCAA T:66.0(4.35%),B:4863.0(2.19%),P:1e-6\n",
      "\t\tRound 2: -14.62 TGCGAA T:95.0(6.21%),B:7993.4(3.57%),P:1e-6\n",
      "\t\tRound 3: -15.59 CGCCAA T:48.0(3.19%),B:2996.6(1.36%),P:1e-6\n",
      "\t\tRound 4: -15.59 CGCCAA T:48.0(3.19%),B:2996.6(1.36%),P:1e-6\n",
      "\t\t=Final=: -16.50 CGCCAA T:48.0(3.24%),B:2974.8(1.35%),P:1e-7\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t14 of 25 Initial Sequence: TCGCCT... (-9.674)\n",
      "\t\tRound 1: -9.92 TCGCCT T:50.0(3.32%),B:3960.2(1.79%),P:1e-4\n",
      "\t\tRound 2: -12.14 NCGGAM T:158.0(10.11%),B:15789.5(6.94%),P:1e-5\n",
      "\t\tRound 3: -12.14 NCGGAM T:158.0(10.11%),B:15789.5(6.94%),P:1e-5\n",
      "\t\t=Final=: -14.31 NCGGAM T:146.0(9.84%),B:14265.3(6.50%),P:1e-6\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t15 of 25 Initial Sequence: TTCGCG... (-9.683)\n",
      "\t\tRound 1: -18.92 TTCGCG T:202.0(12.74%),B:18993.3(8.29%),P:1e-8\n",
      "\t\tRound 2: -18.92 TTCGCG T:202.0(12.74%),B:18993.3(8.29%),P:1e-8\n",
      "\t\t=Final=: -16.93 TTCGCG T:167.0(11.26%),B:16143.2(7.35%),P:1e-7\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t16 of 25 Initial Sequence: CCCCCT... (-9.564)\n",
      "\t\tRound 1: -18.46 CCCCCT T:383.0(22.77%),B:40904.6(17.00%),P:1e-8\n",
      "\t\tRound 2: -18.46 CCCCCT T:383.0(22.77%),B:40904.6(17.00%),P:1e-8\n",
      "\t\t=Final=: -16.38 CCCCCT T:273.0(18.41%),B:29663.9(13.51%),P:1e-7\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\tRemaining seeds don't look promising (After initial 5 motifs, logp -8.761 > -9.171)\n",
      "\n",
      "\tFinalizing Enrichment Statistics (new in v3.4)\n",
      "\tReading input files...\n",
      "\t245550 total sequences read\n",
      "\tCache length = 11180\n",
      "\tUsing binomial scoring\n",
      "\tChecking enrichment of 16 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tOutput in file: motif/homerMotifs.motifs6\n",
      "\n",
      "\n",
      "\tNumber of Trial motifs (-T) set to 32 (from 10) to work well with 32 CPUs\n",
      "\tScanning input files...\n",
      "\tParsing sequences...\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\tTotal number of Oligos: 16384\n",
      "\tAutoadjustment for sequence coverage in background: 1.00x\n",
      "\n",
      "\tOligos: 16384 of 16394 max\n",
      "\tTree  : 21844 of 81970 max\n",
      "\tOptimizing memory usage...\n",
      "\tCache length = 11180\n",
      "\tUsing binomial scoring\n",
      "\n",
      "\tGlobal Optimization Phase: Looking for enriched oligos with up to 2 mismatches...\n",
      "\n",
      "\tScreening oligos 16384 (allowing 0 mismatches):\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\t\t0.74% skipped, 99.26% checked (16263 of 16384), of those checked:\n",
      "\t\t0.74% not in target, 0.00% increased p-value, 0.00% high p-value\n",
      "\n",
      "\tScreening oligos 16384 (allowing 1 mismatches):\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\t\t0.74% skipped, 99.26% checked (16263 of 16384), of those checked:\n",
      "\t\t0.00% not in target, 42.85% increased p-value, 53.81% high p-value\n",
      "\n",
      "\tScreening oligos 16384 (allowing 2 mismatches):\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\t\t67.54% skipped, 32.46% checked (5319 of 16384), of those checked:\n",
      "\t\t0.00% not in target, 26.97% increased p-value, 0.00% high p-value\n",
      "\tReading input files...\n",
      "\t245550 total sequences read\n",
      "\tCache length = 11180\n",
      "\tUsing binomial scoring\n",
      "\n",
      "\tLocal Optimization Phase:\n",
      "\t1 of 25 Initial Sequence: GGTAAGT... (-115.548)\n",
      "\t\tRound 1: -164.36 GGTAAGT T:461.0(26.73%),B:23567.5(10.18%),P:1e-71\n",
      "\t\tRound 2: -164.36 GGTAAGT T:461.0(26.73%),B:23567.5(10.18%),P:1e-71\n",
      "\t\t=Final=: -222.30 GGTAAGT T:440.0(29.67%),B:22124.2(10.08%),P:1e-96\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t2 of 25 Initial Sequence: GAAGAAG... (-108.635)\n",
      "\t\tRound 1: -193.04 GAAGADG T:924.0(46.38%),B:58004.9(23.22%),P:1e-83\n",
      "\t\tRound 2: -193.04 GAAGADG T:924.0(46.38%),B:58004.9(23.22%),P:1e-83\n",
      "\t\t=Final=: -131.75 GAAGADG T:501.0(33.78%),B:36565.9(16.66%),P:1e-57\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t3 of 25 Initial Sequence: ACGACGA... (-30.635)\n",
      "\t\tRound 1: -45.38 ACGACGA T:526.0(29.87%),B:48333.9(19.76%),P:1e-19\n",
      "\t\tRound 2: -49.95 ACGASGA T:1043.0(50.52%),B:105296.8(38.10%),P:1e-21\n",
      "\t\tRound 3: -49.95 ACGASGA T:1043.0(50.52%),B:105296.8(38.10%),P:1e-21\n",
      "\t\t=Final=: -35.68 ACGASGA T:655.0(44.17%),B:74643.6(34.00%),P:1e-15\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t4 of 25 Initial Sequence: GGTTCGA... (-25.998)\n",
      "\t\tRound 1: -38.77 NVTTCGA T:515.0(29.35%),B:49214.1(20.08%),P:1e-16\n",
      "\t\tRound 2: -38.77 NVTTCGA T:515.0(29.35%),B:49214.1(20.08%),P:1e-16\n",
      "\t\t=Final=: -32.65 NVTTCGA T:411.0(27.71%),B:42598.2(19.40%),P:1e-14\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t5 of 25 Initial Sequence: CAGGTAA... (-25.536)\n",
      "\t\tRound 1: -44.29 MAGGTAM T:145.0(9.32%),B:8809.9(3.93%),P:1e-19\n",
      "\t\tRound 2: -44.29 MAGGTAM T:145.0(9.32%),B:8809.9(3.93%),P:1e-19\n",
      "\t\t=Final=: -48.18 MAGGTAM T:142.0(9.58%),B:8622.6(3.93%),P:1e-20\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t6 of 25 Initial Sequence: AGCACAC... (-22.305)\n",
      "\t\tRound 1: -27.31 AGCACAC T:74.0(4.87%),B:4180.6(1.89%),P:1e-11\n",
      "\t\tRound 2: -27.31 AGCACAC T:74.0(4.87%),B:4180.6(1.89%),P:1e-11\n",
      "\t\t=Final=: -15.87 AGCACAC T:121.0(8.16%),B:10919.1(4.97%),P:1e-6\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t7 of 25 Initial Sequence: CGGAAGA... (-19.382)\n",
      "\t\tRound 1: -24.24 CGGAAGA T:242.0(15.06%),B:22242.4(9.63%),P:1e-10\n",
      "\t\tRound 2: -33.91 CGGAAWR T:142.0(9.13%),B:9740.0(4.34%),P:1e-14\n",
      "\t\tRound 3: -37.14 CGCAAAR T:126.0(8.15%),B:7793.5(3.49%),P:1e-16\n",
      "\t\tRound 4: -37.14 CGCAAAR T:126.0(8.15%),B:7793.5(3.49%),P:1e-16\n",
      "\t\t=Final=: -36.32 CGCAAAR T:118.0(7.96%),B:7550.1(3.44%),P:1e-15\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t8 of 25 Initial Sequence: AGGTGAG... (-15.638)\n",
      "\t\tRound 1: -20.92 AGGTGAG T:123.0(7.96%),B:9838.0(4.38%),P:1e-9\n",
      "\t\tRound 2: -21.06 ASGTGCG T:94.0(6.14%),B:6805.6(3.05%),P:1e-9\n",
      "\t\tRound 3: -21.06 ASGTGCG T:94.0(6.14%),B:6805.6(3.05%),P:1e-9\n",
      "\t\t=Final=: -19.61 ASGTGCG T:88.0(5.93%),B:6592.6(3.00%),P:1e-8\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t9 of 25 Initial Sequence: CTTCTAC... (-15.353)\n",
      "\t\tRound 1: -23.68 CTTCTAC T:807.0(41.98%),B:90705.5(33.84%),P:1e-10\n",
      "\t\tRound 2: -24.80 CWWCYMC T:393.0(23.29%),B:39685.7(16.54%),P:1e-10\n",
      "\t\tRound 3: -31.04 CAACACC T:624.0(34.36%),B:64720.3(25.53%),P:1e-13\n",
      "\t\tRound 4: -31.04 CAACACC T:624.0(34.36%),B:64720.3(25.53%),P:1e-13\n",
      "\t\t=Final=: -22.10 CAACACC T:443.0(29.87%),B:50162.9(22.85%),P:1e-9\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t10 of 25 Initial Sequence: CCAACGA... (-13.063)\n",
      "\t\tRound 1: -24.07 CCAACGA T:490.0(28.15%),B:51605.5(20.95%),P:1e-10\n",
      "\t\tRound 2: -24.07 CCAACGA T:490.0(28.15%),B:51605.5(20.95%),P:1e-10\n",
      "\t\t=Final=: -35.63 CCAACGA T:585.0(39.45%),B:64954.9(29.59%),P:1e-15\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t11 of 25 Initial Sequence: AAGTAAG... (-11.463)\n",
      "\t\tRound 1: -16.52 AAGTAWG T:86.0(5.64%),B:6629.1(2.97%),P:1e-7\n",
      "\t\tRound 2: -16.52 AAGTAWG T:86.0(5.64%),B:6629.1(2.97%),P:1e-7\n",
      "\t\t=Final=: -16.75 AAGTAWG T:83.0(5.60%),B:6491.2(2.96%),P:1e-7\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t12 of 25 Initial Sequence: TTATGTT... (-10.695)\n",
      "\t\tRound 1: -10.69 TTATGTT T:56.0(3.71%),B:4383.5(1.98%),P:1e-4\n",
      "\t\tRound 2: -10.69 TTATGTT T:56.0(3.71%),B:4383.5(1.98%),P:1e-4\n",
      "\t\t=Final=: -12.40 TTATGTT T:56.0(3.78%),B:4286.0(1.95%),P:1e-5\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t13 of 25 Initial Sequence: TGCTTTT... (-9.880)\n",
      "\t\tRound 1: -16.20 TGCTTTT T:622.0(34.27%),B:72193.9(28.02%),P:1e-7\n",
      "\t\tRound 2: -16.20 TGCTTTT T:622.0(34.27%),B:72193.9(28.02%),P:1e-7\n",
      "\t\t=Final=: -20.56 TGCTTTT T:693.0(46.73%),B:85755.6(39.06%),P:1e-8\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t14 of 25 Initial Sequence: GGAGAGC... (-9.920)\n",
      "\t\tRound 1: -18.22 GGAGAGC T:92.0(6.02%),B:7017.2(3.15%),P:1e-7\n",
      "\t\tRound 2: -18.22 GGAGAGC T:92.0(6.02%),B:7017.2(3.15%),P:1e-7\n",
      "\t\t=Final=: -16.70 GGAGAGC T:86.0(5.80%),B:6823.9(3.11%),P:1e-7\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t15 of 25 Initial Sequence: TTCGTGG... (-9.431)\n",
      "\t\tRound 1: -14.97 TTCGTGG T:182.0(11.55%),B:17860.5(7.81%),P:1e-6\n",
      "\t\tRound 2: -14.97 TTCGTGG T:182.0(11.55%),B:17860.5(7.81%),P:1e-6\n",
      "\t\t=Final=: -16.35 TTCGTGG T:172.0(11.60%),B:16900.1(7.70%),P:1e-7\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t16 of 25 Initial Sequence: CGCGGTA... (-9.283)\n",
      "\t\tRound 1: -9.47 CGCGGTA T:88.0(5.76%),B:8299.5(3.71%),P:1e-4\n",
      "\t\tRound 2: -11.09 CGCGAWA T:33.0(2.20%),B:2038.2(0.92%),P:1e-4\n",
      "\t\tRound 3: -11.09 CGCGAWA T:33.0(2.20%),B:2038.2(0.92%),P:1e-4\n",
      "\t\t=Final=: -10.43 CGCGAWA T:31.0(2.09%),B:2004.8(0.91%),P:1e-4\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\tRemaining seeds don't look promising (After initial 5 motifs, logp -9.046 > -9.171)\n",
      "\n",
      "\tFinalizing Enrichment Statistics (new in v3.4)\n",
      "\tReading input files...\n",
      "\t245550 total sequences read\n",
      "\tCache length = 11180\n",
      "\tUsing binomial scoring\n",
      "\tChecking enrichment of 16 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tOutput in file: motif/homerMotifs.motifs7\n",
      "\n",
      "\n",
      "\tNumber of Trial motifs (-T) set to 32 (from 10) to work well with 32 CPUs\n",
      "\tScanning input files...\n",
      "\tParsing sequences...\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\tTotal number of Oligos: 65536\n",
      "\tAutoadjustment for sequence coverage in background: 1.00x\n",
      "\n",
      "\tOligos: 65536 of 65546 max\n",
      "\tTree  : 87380 of 327730 max\n",
      "\tOptimizing memory usage...\n",
      "\tCache length = 11180\n",
      "\tUsing binomial scoring\n",
      "\n",
      "\tGlobal Optimization Phase: Looking for enriched oligos with up to 2 mismatches...\n",
      "\n",
      "\tScreening oligos 65536 (allowing 0 mismatches):\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\t\t13.03% skipped, 86.97% checked (56994 of 65536), of those checked:\n",
      "\t\t13.03% not in target, 0.00% increased p-value, 0.00% high p-value\n",
      "\n",
      "\tScreening oligos 65536 (allowing 1 mismatches):\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\t\t13.03% skipped, 86.97% checked (56994 of 65536), of those checked:\n",
      "\t\t0.00% not in target, 33.89% increased p-value, 52.26% high p-value\n",
      "\n",
      "\tScreening oligos 65536 (allowing 2 mismatches):\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\t\t72.65% skipped, 27.35% checked (17924 of 65536), of those checked:\n",
      "\t\t0.00% not in target, 16.76% increased p-value, 0.00% high p-value\n",
      "\tReading input files...\n",
      "\t245550 total sequences read\n",
      "\tCache length = 11180\n",
      "\tUsing binomial scoring\n",
      "\n",
      "\tLocal Optimization Phase:\n",
      "\t1 of 25 Initial Sequence: GAAGAAGA... (-155.525)\n",
      "\t\tRound 1: -224.47 GAAGAAGA T:852.0(43.71%),B:48029.9(19.65%),P:1e-97\n",
      "\t\tRound 2: -224.47 GAAGAAGA T:852.0(43.71%),B:48029.9(19.65%),P:1e-97\n",
      "\t\t=Final=: -120.24 GAAGAAGA T:669.0(45.11%),B:58379.3(26.59%),P:1e-52\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t2 of 25 Initial Sequence: AGGTAAGT... (-111.138)\n",
      "\t\tRound 1: -161.63 AGGTAAGT T:457.0(26.53%),B:23498.0(10.15%),P:1e-70\n",
      "\t\tRound 2: -161.63 AGGTAAGT T:457.0(26.53%),B:23498.0(10.15%),P:1e-70\n",
      "\t\t=Final=: -234.12 AGGTAAGT T:567.0(38.23%),B:33544.9(15.28%),P:1e-101\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t3 of 25 Initial Sequence: TCGACGAC... (-26.011)\n",
      "\t\tRound 1: -42.65 TCGACGAC T:246.0(15.29%),B:18785.7(8.20%),P:1e-18\n",
      "\t\tRound 2: -43.11 TCGACGAC T:318.0(19.31%),B:26273.3(11.28%),P:1e-18\n",
      "\t\tRound 3: -43.11 TCGACGAC T:318.0(19.31%),B:26273.3(11.28%),P:1e-18\n",
      "\t\t=Final=: -32.17 TCGACGAC T:293.0(19.76%),B:27829.6(12.68%),P:1e-13\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t4 of 25 Initial Sequence: TGAAGAGC... (-24.567)\n",
      "\t\tRound 1: -42.46 TGAAGAGC T:294.0(17.99%),B:23805.9(10.27%),P:1e-18\n",
      "\t\tRound 2: -47.62 TGAAGAGC T:251.0(15.58%),B:18446.0(8.06%),P:1e-20\n",
      "\t\tRound 3: -47.62 TGAAGAGC T:251.0(15.58%),B:18446.0(8.06%),P:1e-20\n",
      "\t\t=Final=: -50.10 TGAAGAGC T:230.0(15.51%),B:17335.1(7.90%),P:1e-21\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t5 of 25 Initial Sequence: AGCACACA... (-23.947)\n",
      "\t\tRound 1: -29.86 AGCACACA T:55.0(3.64%),B:2438.8(1.10%),P:1e-12\n",
      "\t\tRound 2: -29.86 AGCACACA T:55.0(3.64%),B:2438.8(1.10%),P:1e-12\n",
      "\t\t=Final=: -8.86 AGCACACA T:32.0(2.16%),B:2291.8(1.04%),P:1e-3\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t6 of 25 Initial Sequence: GGGTTCGA... (-17.860)\n",
      "\t\tRound 1: -30.42 KGGTTCGA T:53.0(3.51%),B:2248.1(1.02%),P:1e-13\n",
      "\t\tRound 2: -39.17 HBSKACGA T:456.0(26.48%),B:42334.3(17.54%),P:1e-17\n",
      "\t\tRound 3: -41.82 YCCATCGA T:426.0(24.98%),B:38299.3(16.01%),P:1e-18\n",
      "\t\tRound 4: -43.50 HRCATCGA T:244.0(15.18%),B:18521.9(8.09%),P:1e-18\n",
      "\t\tRound 5: -43.56 CGSTTCGA T:274.0(16.87%),B:21569.2(9.36%),P:1e-18\n",
      "\t\tRound 6: -43.56 CGSTTCGA T:274.0(16.87%),B:21569.2(9.36%),P:1e-18\n",
      "\t\t=Final=: -38.88 CGSTTCGA T:230.0(15.51%),B:19066.9(8.68%),P:1e-16\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t7 of 25 Initial Sequence: CGCGGAAA... (-15.123)\n",
      "\t\tRound 1: -27.65 CGCGGAAA T:42.0(2.79%),B:1587.0(0.72%),P:1e-12\n",
      "\t\tRound 2: -31.47 CACGGWAA T:72.0(4.74%),B:3647.6(1.65%),P:1e-13\n",
      "\t\tRound 3: -31.47 CACGGWAA T:72.0(4.74%),B:3647.6(1.65%),P:1e-13\n",
      "\t\t=Final=: -31.28 CACGGWAA T:64.0(4.32%),B:3129.4(1.43%),P:1e-13\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t8 of 25 Initial Sequence: CGCGGTAC... (-14.968)\n",
      "\t\tRound 1: -14.97 CGCGGTAC T:7.0(0.47%),B:39.5(0.02%),P:1e-6\n",
      "\t\tRound 2: -14.97 CGCGGTAC T:7.0(0.47%),B:39.5(0.02%),P:1e-6\n",
      "\t\t=Final=: -11.68 CGCGGTAC T:5.0(0.34%),B:39.5(0.02%),P:1e-5\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t9 of 25 Initial Sequence: TACTACTA... (-14.774)\n",
      "\t\tRound 1: -19.37 TACTACTA T:27.0(1.80%),B:945.2(0.43%),P:1e-8\n",
      "\t\tRound 2: -21.36 TACKATTA T:70.0(4.61%),B:4403.6(1.99%),P:1e-9\n",
      "\t\tRound 3: -25.73 TACTATTA T:103.0(6.71%),B:7014.6(3.14%),P:1e-11\n",
      "\t\tRound 4: -25.73 TACTATTA T:103.0(6.71%),B:7014.6(3.14%),P:1e-11\n",
      "\t\t=Final=: -19.33 TACTATTA T:89.0(6.00%),B:6738.8(3.07%),P:1e-8\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t10 of 25 Initial Sequence: ACAACCCC... (-13.637)\n",
      "\t\tRound 1: -21.01 ACAACCCC T:108.0(7.03%),B:8236.0(3.68%),P:1e-9\n",
      "\t\tRound 2: -21.01 ACAACCCC T:108.0(7.03%),B:8236.0(3.68%),P:1e-9\n",
      "\t\t=Final=: -21.11 ACAACCCC T:102.0(6.88%),B:7853.0(3.58%),P:1e-9\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t11 of 25 Initial Sequence: CCATTGGC... (-12.545)\n",
      "\t\tRound 1: -25.21 CCATTGGC T:207.0(13.03%),B:18087.8(7.91%),P:1e-10\n",
      "\t\tRound 2: -25.21 CCATTGGC T:207.0(13.03%),B:18087.8(7.91%),P:1e-10\n",
      "\t\t=Final=: -27.04 CCATTGGC T:194.0(13.08%),B:17107.3(7.79%),P:1e-11\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t12 of 25 Initial Sequence: GTCAGTAT... (-12.681)\n",
      "\t\tRound 1: -23.03 GTCAGTAT T:88.0(5.76%),B:5911.2(2.66%),P:1e-10\n",
      "\t\tRound 2: -23.03 GTCAGTAT T:88.0(5.76%),B:5911.2(2.66%),P:1e-10\n",
      "\t\t=Final=: -22.34 GTCAGTAT T:84.0(5.66%),B:5816.0(2.65%),P:1e-9\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t13 of 25 Initial Sequence: CAACAGCC... (-12.163)\n",
      "\t\tRound 1: -22.44 CAACAGCC T:120.0(7.78%),B:9235.9(4.12%),P:1e-9\n",
      "\t\tRound 2: -22.44 CAACAGCC T:120.0(7.78%),B:9235.9(4.12%),P:1e-9\n",
      "\t\t=Final=: -22.58 CAACAGCC T:122.0(8.23%),B:9780.4(4.45%),P:1e-9\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t14 of 25 Initial Sequence: CCGTGAAG... (-12.305)\n",
      "\t\tRound 1: -20.38 CCGTGAAG T:166.0(10.59%),B:14637.0(6.45%),P:1e-8\n",
      "\t\tRound 2: -20.38 CCGTGAAG T:166.0(10.59%),B:14637.0(6.45%),P:1e-8\n",
      "\t\t=Final=: -22.00 CCGTGAAG T:160.0(10.79%),B:14171.0(6.45%),P:1e-9\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t15 of 25 Initial Sequence: TGCTGCTG... (-12.114)\n",
      "\t\tRound 1: -15.40 TGCTGCTG T:82.0(5.38%),B:6369.6(2.86%),P:1e-6\n",
      "\t\tRound 2: -18.14 TDCTGCTT T:142.0(9.13%),B:12412.8(5.50%),P:1e-7\n",
      "\t\tRound 3: -18.14 TDCTGCTT T:142.0(9.13%),B:12412.8(5.50%),P:1e-7\n",
      "\t\t=Final=: -13.49 TDCTGCTT T:141.0(9.51%),B:13870.9(6.32%),P:1e-5\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t16 of 25 Initial Sequence: ACGGAAGA... (-10.981)\n",
      "\t\tRound 1: -18.83 ACGGAAGA T:133.0(8.58%),B:11294.2(5.01%),P:1e-8\n",
      "\t\tRound 2: -22.96 ACGGAAGA T:271.0(16.71%),B:25871.7(11.12%),P:1e-9\n",
      "\t\tRound 3: -22.96 ACGGAAGA T:271.0(16.71%),B:25871.7(11.12%),P:1e-9\n",
      "\t\t=Final=: -24.20 ACGGAAGA T:245.0(16.52%),B:23843.4(10.86%),P:1e-10\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t17 of 25 Initial Sequence: AAAGGCGG... (-11.015)\n",
      "\t\tRound 1: -11.02 AAAGGCGG T:10.0(0.67%),B:209.6(0.10%),P:1e-4\n",
      "\t\tRound 2: -11.02 AAAGGCGG T:10.0(0.67%),B:209.6(0.10%),P:1e-4\n",
      "\t\t=Final=: -12.95 AAAGGCGG T:10.0(0.67%),B:209.6(0.10%),P:1e-5\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t18 of 25 Initial Sequence: ACAGGTAA... (-10.986)\n",
      "\t\tRound 1: -21.67 ACAGGTAA T:73.0(4.80%),B:4664.5(2.10%),P:1e-9\n",
      "\t\tRound 2: -25.35 DVCGGTAM T:191.0(12.09%),B:16294.6(7.15%),P:1e-11\n",
      "\t\tRound 3: -26.52 HCCGGTAC T:87.0(5.70%),B:5401.3(2.43%),P:1e-11\n",
      "\t\tRound 4: -27.16 CCAGGTAA T:89.0(5.83%),B:5526.4(2.49%),P:1e-11\n",
      "\t\tRound 5: -27.16 CCAGGTAA T:89.0(5.83%),B:5526.4(2.49%),P:1e-11\n",
      "\t\t=Final=: -28.95 CCAGGTAA T:88.0(5.93%),B:5460.5(2.49%),P:1e-12\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t19 of 25 Initial Sequence: TGTAAGTA... (-10.831)\n",
      "\t\tRound 1: -17.56 TGTAAGTA T:25.0(1.67%),B:893.9(0.41%),P:1e-7\n",
      "\t\tRound 2: -17.86 CGTAAGTA T:27.0(1.80%),B:1020.7(0.46%),P:1e-7\n",
      "\t\tRound 3: -17.86 CGTAAGTA T:27.0(1.80%),B:1020.7(0.46%),P:1e-7\n",
      "\t\t=Final=: -19.25 CGTAAGTA T:27.0(1.82%),B:1017.0(0.46%),P:1e-8\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t20 of 25 Initial Sequence: TAGCTCAG... (-10.704)\n",
      "\t\tRound 1: -10.91 TAGCTCAG T:18.0(1.21%),B:740.3(0.34%),P:1e-4\n",
      "\t\tRound 2: -15.45 TCGCTCAG T:33.0(2.20%),B:1645.4(0.75%),P:1e-6\n",
      "\t\tRound 3: -15.45 TCGCTCAG T:33.0(2.20%),B:1645.4(0.75%),P:1e-6\n",
      "\t\t=Final=: -16.64 TCGCTCAG T:32.0(2.16%),B:1551.9(0.71%),P:1e-7\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t21 of 25 Initial Sequence: CAGGAACT... (-10.549)\n",
      "\t\tRound 1: -13.09 CAGGAACT T:32.0(2.14%),B:1751.2(0.79%),P:1e-5\n",
      "\t\tRound 2: -15.32 CAAGAACT T:38.0(2.53%),B:2091.9(0.95%),P:1e-6\n",
      "\t\tRound 3: -15.32 CAAGAACT T:38.0(2.53%),B:2091.9(0.95%),P:1e-6\n",
      "\t\t=Final=: -16.34 CAAGAACT T:38.0(2.56%),B:2081.1(0.95%),P:1e-7\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t22 of 25 Initial Sequence: TTGCCAAT... (-10.366)\n",
      "\t\tRound 1: -11.62 TTGCCAAT T:26.0(1.74%),B:1347.5(0.61%),P:1e-5\n",
      "\t\tRound 2: -13.17 TCGCCAAT T:30.0(2.00%),B:1568.2(0.71%),P:1e-5\n",
      "\t\tRound 3: -18.34 GCGCCAAT T:32.0(2.14%),B:1366.2(0.62%),P:1e-7\n",
      "\t\tRound 4: -18.34 GCGCCAAT T:32.0(2.14%),B:1366.2(0.62%),P:1e-7\n",
      "\t\t=Final=: -19.58 GCGCCAAT T:32.0(2.16%),B:1363.3(0.62%),P:1e-8\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t23 of 25 Initial Sequence: ATTATGTT... (-10.099)\n",
      "\t\tRound 1: -13.39 ATTATGTT T:55.0(3.64%),B:3970.5(1.79%),P:1e-5\n",
      "\t\tRound 2: -13.39 ATTATGTT T:55.0(3.64%),B:3970.5(1.79%),P:1e-5\n",
      "\t\t=Final=: -14.37 ATTATGTT T:55.0(3.71%),B:3906.9(1.78%),P:1e-6\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t24 of 25 Initial Sequence: CTCTCACG... (-9.938)\n",
      "\t\tRound 1: -14.94 CTCTCACG T:54.0(3.58%),B:3664.8(1.65%),P:1e-6\n",
      "\t\tRound 2: -14.94 CTCTCACG T:54.0(3.58%),B:3664.8(1.65%),P:1e-6\n",
      "\t\t=Final=: -15.30 CTCTCACG T:53.0(3.57%),B:3589.5(1.63%),P:1e-6\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t25 of 25 Initial Sequence: CTCTTTTT... (-9.779)\n",
      "\t\tRound 1: -13.07 CTCTTTTT T:100.0(6.52%),B:8789.5(3.92%),P:1e-5\n",
      "\t\tRound 2: -13.07 CTCTTTTT T:100.0(6.52%),B:8789.5(3.92%),P:1e-5\n",
      "\t\t=Final=: -15.91 CTCTTTTT T:100.0(6.74%),B:8506.4(3.87%),P:1e-6\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\n",
      "\tFinalizing Enrichment Statistics (new in v3.4)\n",
      "\tReading input files...\n",
      "\t245550 total sequences read\n",
      "\tCache length = 11180\n",
      "\tUsing binomial scoring\n",
      "\tChecking enrichment of 25 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tOutput in file: motif/homerMotifs.motifs8\n",
      "\n",
      "\n",
      "\tNumber of Trial motifs (-T) set to 32 (from 10) to work well with 32 CPUs\n",
      "\tScanning input files...\n",
      "\tParsing sequences...\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\tTotal number of Oligos: 261886\n",
      "\tAutoadjustment for sequence coverage in background: 1.00x\n",
      "\n",
      "\tOligos: 261886 of 262154 max\n",
      "\tTree  : 349516 of 1310770 max\n",
      "\tOptimizing memory usage...\n",
      "\tCache length = 11180\n",
      "\tUsing binomial scoring\n",
      "\n",
      "\tGlobal Optimization Phase: Looking for enriched oligos with up to 2 mismatches...\n",
      "\n",
      "\tScreening oligos 261886 (allowing 0 mismatches):\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\t\t46.66% skipped, 53.34% checked (139694 of 261886), of those checked:\n",
      "\t\t46.66% not in target, 0.00% increased p-value, 0.00% high p-value\n",
      "\n",
      "\tScreening oligos 261886 (allowing 1 mismatches):\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\t\t46.66% skipped, 53.34% checked (139694 of 261886), of those checked:\n",
      "\t\t0.00% not in target, 25.39% increased p-value, 31.67% high p-value\n",
      "\n",
      "\tScreening oligos 261886 (allowing 2 mismatches):\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\t\t84.95% skipped, 15.05% checked (39424 of 261886), of those checked:\n",
      "\t\t0.00% not in target, 8.51% increased p-value, 0.00% high p-value\n",
      "\tReading input files...\n",
      "\t245550 total sequences read\n",
      "\tCache length = 11180\n",
      "\tUsing binomial scoring\n",
      "\n",
      "\tLocal Optimization Phase:\n",
      "\t1 of 25 Initial Sequence: CGAAGAAGA... (-137.674)\n",
      "\t\tRound 1: -209.93 AGAAGAAGA T:1251.0(56.99%),B:82628.2(31.36%),P:1e-91\n",
      "\t\tRound 2: -209.93 AGAAGAAGA T:1251.0(56.99%),B:82628.2(31.36%),P:1e-91\n",
      "\t\t=Final=: -98.92 AGAAGAAGA T:593.0(39.99%),B:52163.7(23.76%),P:1e-42\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t2 of 25 Initial Sequence: AAGGTAAGT... (-96.067)\n",
      "\t\tRound 1: -176.96 AAGGTAAGT T:364.0(21.77%),B:15316.2(6.74%),P:1e-76\n",
      "\t\tRound 2: -179.50 VAGGTAAGT T:354.0(21.24%),B:14531.5(6.40%),P:1e-77\n",
      "\t\tRound 3: -179.50 VAGGTAAGT T:354.0(21.24%),B:14531.5(6.40%),P:1e-77\n",
      "\t\t=Final=: -225.22 VAGGTAAGT T:390.0(26.30%),B:17563.1(8.00%),P:1e-97\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t3 of 25 Initial Sequence: CGGAAGAAC... (-34.946)\n",
      "\t\tRound 1: -48.76 CGGAAGAAC T:165.0(10.53%),B:10132.6(4.51%),P:1e-21\n",
      "\t\tRound 2: -57.45 CGGAAGAGC T:538.0(30.43%),B:46517.3(19.09%),P:1e-24\n",
      "\t\tRound 3: -69.01 CDGAAGARC T:362.0(21.67%),B:26106.5(11.21%),P:1e-29\n",
      "\t\tRound 4: -69.01 CDGAAGARC T:362.0(21.67%),B:26106.5(11.21%),P:1e-29\n",
      "\t\t=Final=: -60.58 CDGAAGARC T:306.0(20.63%),B:24163.7(11.01%),P:1e-26\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t4 of 25 Initial Sequence: ATCGACAGC... (-26.937)\n",
      "\t\tRound 1: -39.08 ATCGACAGC T:165.0(10.53%),B:11306.8(5.02%),P:1e-16\n",
      "\t\tRound 2: -41.03 ATCGACAGC T:313.0(19.03%),B:26181.9(11.24%),P:1e-17\n",
      "\t\tRound 3: -44.73 ATCGACAMC T:381.0(22.66%),B:32701.3(13.84%),P:1e-19\n",
      "\t\tRound 4: -44.73 ATCGACAMC T:381.0(22.66%),B:32701.3(13.84%),P:1e-19\n",
      "\t\t=Final=: -45.09 ATCGACAMC T:325.0(21.92%),B:28991.8(13.21%),P:1e-19\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t5 of 25 Initial Sequence: TACTACTAC... (-21.427)\n",
      "\t\tRound 1: -23.97 TACTACTAC T:13.0(0.87%),B:114.1(0.05%),P:1e-10\n",
      "\t\tRound 2: -23.97 TACTACTAC T:13.0(0.87%),B:114.1(0.05%),P:1e-10\n",
      "\t\t=Final=: -7.14 TACTACTAC T:5.0(0.34%),B:104.2(0.05%),P:1e-3\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t6 of 25 Initial Sequence: TCGGTAAGT... (-18.424)\n",
      "\t\tRound 1: -24.02 TCGGTAAGT T:133.0(8.58%),B:10332.1(4.60%),P:1e-10\n",
      "\t\tRound 2: -34.50 TCGGTAAGT T:209.0(13.15%),B:16475.4(7.23%),P:1e-14\n",
      "\t\tRound 3: -34.50 TCGGTAAGT T:209.0(13.15%),B:16475.4(7.23%),P:1e-14\n",
      "\t\t=Final=: -42.29 TCGGTAAGT T:204.0(13.76%),B:15647.6(7.13%),P:1e-18\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t7 of 25 Initial Sequence: CACAAACGC... (-18.106)\n",
      "\t\tRound 1: -21.46 CACAAACGC T:10.0(0.67%),B:60.0(0.03%),P:1e-9\n",
      "\t\tRound 2: -21.80 CACAATCGC T:24.0(1.61%),B:654.6(0.30%),P:1e-9\n",
      "\t\tRound 3: -22.89 CACATACGC T:12.0(0.81%),B:97.5(0.04%),P:1e-9\n",
      "\t\tRound 4: -22.89 CACATACGC T:12.0(0.81%),B:97.5(0.04%),P:1e-9\n",
      "\t\t=Final=: -12.07 CACATACGC T:7.0(0.47%),B:97.5(0.04%),P:1e-5\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t8 of 25 Initial Sequence: GGTAGAGCG... (-17.605)\n",
      "\t\tRound 1: -17.73 GGTAGAGCG T:12.0(0.81%),B:160.7(0.07%),P:1e-7\n",
      "\t\tRound 2: -19.34 GGTMGAGCA T:16.0(1.07%),B:299.1(0.14%),P:1e-8\n",
      "\t\tRound 3: -21.05 GGTCGAGCT T:18.0(1.21%),B:354.4(0.16%),P:1e-9\n",
      "\t\tRound 4: -22.10 GGTCGAGCA T:15.0(1.01%),B:204.8(0.09%),P:1e-9\n",
      "\t\tRound 5: -22.10 GGTCGAGCA T:15.0(1.01%),B:204.8(0.09%),P:1e-9\n",
      "\t\t=Final=: -24.44 GGTCGAGCA T:15.0(1.01%),B:204.8(0.09%),P:1e-10\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t9 of 25 Initial Sequence: TGCTGATGA... (-16.606)\n",
      "\t\tRound 1: -23.64 TGCTGATGA T:263.0(16.26%),B:24849.3(10.70%),P:1e-10\n",
      "\t\tRound 2: -28.76 TGCTGHTGM T:505.0(28.87%),B:51555.9(20.93%),P:1e-12\n",
      "\t\tRound 3: -33.67 TGCTGMTGA T:407.0(24.01%),B:38486.9(16.08%),P:1e-14\n",
      "\t\tRound 4: -35.96 WGCTGMTGA T:609.0(33.69%),B:60971.0(24.25%),P:1e-15\n",
      "\t\tRound 5: -38.79 WGYWGMTGA T:333.0(20.12%),B:28825.3(12.30%),P:1e-16\n",
      "\t\tRound 6: -38.79 WGYWGMTGA T:333.0(20.12%),B:28825.3(12.30%),P:1e-16\n",
      "\t\t=Final=: -34.16 WGYWGMTGA T:271.0(18.27%),B:24734.3(11.27%),P:1e-14\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t10 of 25 Initial Sequence: CGATCCGGA... (-17.312)\n",
      "\t\tRound 1: -17.31 CGATCCGGA T:166.0(10.59%),B:15372.5(6.76%),P:1e-7\n",
      "\t\tRound 2: -17.31 CGATCCGGA T:166.0(10.59%),B:15372.5(6.76%),P:1e-7\n",
      "\t\t=Final=: -18.17 CGATCCGGA T:154.0(10.38%),B:14288.4(6.51%),P:1e-7\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t11 of 25 Initial Sequence: ACCATCAAC... (-14.578)\n",
      "\t\tRound 1: -26.81 ACCATCAAC T:121.0(7.84%),B:8675.8(3.87%),P:1e-11\n",
      "\t\tRound 2: -29.72 ACCATCAAC T:466.0(26.97%),B:46614.8(19.13%),P:1e-12\n",
      "\t\tRound 3: -30.70 ACCATCAAC T:580.0(32.38%),B:59689.1(23.80%),P:1e-13\n",
      "\t\tRound 4: -30.70 ACCATCAAC T:580.0(32.38%),B:59689.1(23.80%),P:1e-13\n",
      "\t\t=Final=: -27.26 ACCATCAAC T:391.0(26.37%),B:41542.7(18.92%),P:1e-11\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t12 of 25 Initial Sequence: TACTCTTTT... (-13.226)\n",
      "\t\tRound 1: -20.35 TACTCTTTT T:69.0(4.55%),B:4419.7(1.99%),P:1e-8\n",
      "\t\tRound 2: -20.35 TACTCTTTT T:69.0(4.55%),B:4419.7(1.99%),P:1e-8\n",
      "\t\t=Final=: -21.57 TACTCTTTT T:68.0(4.59%),B:4335.6(1.97%),P:1e-9\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t13 of 25 Initial Sequence: CGCTAACCC... (-12.148)\n",
      "\t\tRound 1: -18.31 CGCTAACCC T:12.0(0.81%),B:151.6(0.07%),P:1e-7\n",
      "\t\tRound 2: -18.31 CGCTAACCC T:12.0(0.81%),B:151.6(0.07%),P:1e-7\n",
      "\t\t=Final=: -20.73 CGCTAACCC T:12.0(0.81%),B:151.6(0.07%),P:1e-9\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t14 of 25 Initial Sequence: GGCGCTCTC... (-11.682)\n",
      "\t\tRound 1: -11.85 GGCGCTCTC T:10.0(0.67%),B:188.1(0.09%),P:1e-5\n",
      "\t\tRound 2: -16.01 GGCGCTCTC T:74.0(4.87%),B:5505.1(2.48%),P:1e-6\n",
      "\t\tRound 3: -18.79 GGCGBTCTH T:51.0(3.38%),B:2954.6(1.34%),P:1e-8\n",
      "\t\tRound 4: -18.79 GGCGBTCTH T:51.0(3.38%),B:2954.6(1.34%),P:1e-8\n",
      "\t\t=Final=: -16.34 GGCGBTCTH T:45.0(3.03%),B:2711.3(1.23%),P:1e-7\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t15 of 25 Initial Sequence: GACTATGAC... (-11.637)\n",
      "\t\tRound 1: -21.60 GACTATGAC T:195.0(12.33%),B:17520.2(7.67%),P:1e-9\n",
      "\t\tRound 2: -22.34 GACTATGAC T:139.0(8.95%),B:11207.3(4.98%),P:1e-9\n",
      "\t\tRound 3: -22.34 GACTATGAC T:139.0(8.95%),B:11207.3(4.98%),P:1e-9\n",
      "\t\t=Final=: -22.77 GACTATGAC T:131.0(8.83%),B:10739.4(4.89%),P:1e-9\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t16 of 25 Initial Sequence: TAACAGGTA... (-11.747)\n",
      "\t\tRound 1: -22.32 TAACAGGTA T:37.0(2.46%),B:1510.7(0.69%),P:1e-9\n",
      "\t\tRound 2: -22.32 TAACAGGTA T:37.0(2.46%),B:1510.7(0.69%),P:1e-9\n",
      "\t\t=Final=: -23.71 TAACAGGTA T:37.0(2.49%),B:1501.0(0.68%),P:1e-10\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t17 of 25 Initial Sequence: TCATACCCC... (-11.500)\n",
      "\t\tRound 1: -27.06 TCATACCCC T:63.0(4.16%),B:3218.5(1.45%),P:1e-11\n",
      "\t\tRound 2: -27.06 TCATACCCC T:63.0(4.16%),B:3218.5(1.45%),P:1e-11\n",
      "\t\t=Final=: -29.81 TCATACCCC T:63.0(4.25%),B:3153.3(1.44%),P:1e-12\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t18 of 25 Initial Sequence: ATATCGCTC... (-11.307)\n",
      "\t\tRound 1: -17.76 ATATCGCTC T:35.0(2.33%),B:1642.2(0.75%),P:1e-7\n",
      "\t\tRound 2: -17.76 ATATCGCTC T:35.0(2.33%),B:1642.2(0.75%),P:1e-7\n",
      "\t\t=Final=: -17.78 ATATCGCTC T:34.0(2.29%),B:1634.4(0.74%),P:1e-7\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t19 of 25 Initial Sequence: AGTACACGC... (-11.190)\n",
      "\t\tRound 1: -21.88 AGTACACGC T:20.0(1.34%),B:432.6(0.20%),P:1e-9\n",
      "\t\tRound 2: -21.88 AGTACACGC T:20.0(1.34%),B:432.6(0.20%),P:1e-9\n",
      "\t\t=Final=: -23.78 AGTACACGC T:20.0(1.35%),B:432.6(0.20%),P:1e-10\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t20 of 25 Initial Sequence: TTTTAGAGC... (-10.838)\n",
      "\t\tRound 1: -22.15 TTTTAGAGC T:42.0(2.79%),B:1908.1(0.87%),P:1e-9\n",
      "\t\tRound 2: -22.15 TTTTAGAGC T:42.0(2.79%),B:1908.1(0.87%),P:1e-9\n",
      "\t\t=Final=: -23.48 TTTTAGAGC T:42.0(2.83%),B:1892.7(0.86%),P:1e-10\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t21 of 25 Initial Sequence: AGCGGACCA... (-10.794)\n",
      "\t\tRound 1: -20.59 AGCGGACCA T:76.0(5.00%),B:5092.8(2.29%),P:1e-8\n",
      "\t\tRound 2: -20.98 AGCGKACCA T:51.0(3.38%),B:2757.3(1.25%),P:1e-9\n",
      "\t\tRound 3: -22.73 AGCGKAMMW T:53.0(3.51%),B:2786.0(1.26%),P:1e-9\n",
      "\t\tRound 4: -22.73 AGCGKAMMW T:53.0(3.51%),B:2786.0(1.26%),P:1e-9\n",
      "\t\t=Final=: -22.75 AGCGKAMMW T:52.0(3.51%),B:2766.6(1.26%),P:1e-9\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t22 of 25 Initial Sequence: TTTACAGGG... (-10.828)\n",
      "\t\tRound 1: -18.48 TTTACAGGG T:53.0(3.51%),B:3171.1(1.43%),P:1e-8\n",
      "\t\tRound 2: -18.48 TTTACAGGG T:53.0(3.51%),B:3171.1(1.43%),P:1e-8\n",
      "\t\t=Final=: -17.79 TTTACAGGG T:43.0(2.90%),B:2396.0(1.09%),P:1e-7\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t23 of 25 Initial Sequence: CGCGGTACT... (-10.693)\n",
      "\t\tRound 1: -18.33 CGCGGTACT T:6.0(0.40%),B:11.4(0.01%),P:1e-7\n",
      "\t\tRound 2: -18.47 CGCAGTACC T:9.0(0.61%),B:59.0(0.03%),P:1e-8\n",
      "\t\tRound 3: -20.05 CGCGGTAMC T:11.0(0.74%),B:97.6(0.04%),P:1e-8\n",
      "\t\tRound 4: -20.05 CGCGGTAMC T:11.0(0.74%),B:97.6(0.04%),P:1e-8\n",
      "\t\t=Final=: -14.58 CGCGGTAMC T:8.0(0.54%),B:97.6(0.04%),P:1e-6\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t24 of 25 Initial Sequence: CACGCGAAA... (-10.613)\n",
      "\t\tRound 1: -18.82 CACGCGAAA T:18.0(1.21%),B:412.6(0.19%),P:1e-8\n",
      "\t\tRound 2: -20.71 CACGGTAAA T:23.0(1.54%),B:634.0(0.29%),P:1e-8\n",
      "\t\tRound 3: -20.71 CACGGTAAA T:23.0(1.54%),B:634.0(0.29%),P:1e-8\n",
      "\t\t=Final=: -22.60 CACGGTAAA T:23.0(1.55%),B:626.8(0.29%),P:1e-9\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t25 of 25 Initial Sequence: AGATCGAAG... (-10.341)\n",
      "\t\tRound 1: -23.24 AGATCGAAG T:64.0(4.23%),B:3639.3(1.64%),P:1e-10\n",
      "\t\tRound 2: -27.26 AGATCGAAG T:49.0(3.25%),B:2132.1(0.97%),P:1e-11\n",
      "\t\tRound 3: -28.93 AGAYCGWAR T:72.0(4.74%),B:3854.6(1.74%),P:1e-12\n",
      "\t\tRound 4: -28.93 AGAYCGWAR T:72.0(4.74%),B:3854.6(1.74%),P:1e-12\n",
      "\t\t=Final=: -28.97 AGAYCGWAR T:70.0(4.72%),B:3817.4(1.74%),P:1e-12\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\n",
      "\tFinalizing Enrichment Statistics (new in v3.4)\n",
      "\tReading input files...\n",
      "\t245550 total sequences read\n",
      "\tCache length = 11180\n",
      "\tUsing binomial scoring\n",
      "\tChecking enrichment of 25 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tOutput in file: motif/homerMotifs.motifs9\n",
      "\n",
      "\t(Motifs in homer2 format)\n",
      "\tDetermining similar motifs... 93 reduced to 34 motifs\n",
      "\tOutputing HTML and sequence logos for motif comparison...\n",
      "\tChecking de novo motifs against known motifs...\n",
      "\tFormatting HTML page...\n",
      "\t\t1 of 34 (1e-113) similar to hsa-miR-200a* MIMAT0001620 Homo sapiens miR-200a* Targets (miRBase)(0.655)\n",
      "\t\t2 of 34 (1e-58) similar to hsa-miR-4778-3p MIMAT0019937 Homo sapiens miR-4778-3p Targets (miRBase)(0.786)\n",
      "\t\t3 of 34 (1e-27) similar to hsa-miR-29c* MIMAT0004673 Homo sapiens miR-29c* Targets (miRBase)(0.565)\n",
      "\t\t4 of 34 (1e-26) similar to hsa-miR-497 MIMAT0002820 Homo sapiens miR-497 Targets (miRBase)(0.670)\n",
      "\t\t5 of 34 (1e-25) similar to hsa-miR-1306 MIMAT0005950 Homo sapiens miR-1306 Targets (miRBase)(0.743)\n",
      "\t\t6 of 34 (1e-21) similar to hsa-miR-3610 MIMAT0017987 Homo sapiens miR-3610 Targets (miRBase)(0.719)\n",
      "\t\t7 of 34 (1e-19) similar to hsa-miR-1180 MIMAT0005825 Homo sapiens miR-1180 Targets (miRBase)(0.708)\n",
      "\t\t8 of 34 (1e-15) similar to hsa-miR-4440 MIMAT0018958 Homo sapiens miR-4440 Targets (miRBase)(0.637)\n",
      "\t\t9 of 34 (1e-14) similar to hsa-miR-3144-3p MIMAT0015015 Homo sapiens miR-3144-3p Targets (miRBase)(0.699)\n",
      "\t\t10 of 34 (1e-13) similar to hsa-miR-412 MIMAT0002170 Homo sapiens miR-412 Targets (miRBase)(0.608)\n",
      "\t\t11 of 34 (1e-13) similar to hsa-miR-3201 MIMAT0015086 Homo sapiens miR-3201 Targets (miRBase)(0.712)\n",
      "\t\t12 of 34 (1e-13) similar to hsa-miR-4252 MIMAT0016886 Homo sapiens miR-4252 Targets (miRBase)(0.673)\n",
      "\t\t13 of 34 (1e-12) similar to hsa-miR-103b MIMAT0007402 Homo sapiens miR-103b Targets (miRBase)(0.588)\n",
      "\t\t14 of 34 (1e-11) similar to hsa-miR-671-3p MIMAT0004819 Homo sapiens miR-671-3p Targets (miRBase)(0.650)\n",
      "\t\t15 of 34 (1e-10) similar to hsa-miR-18b* MIMAT0004751 Homo sapiens miR-18b* Targets (miRBase)(0.672)\n",
      "\t\t16 of 34 (1e-10) similar to hsa-miR-20b* MIMAT0004752 Homo sapiens miR-20b* Targets (miRBase)(0.712)\n",
      "\t\t17 of 34 (1e-10) similar to hsa-miR-4687-3p MIMAT0019775 Homo sapiens miR-4687-3p Targets (miRBase)(0.844)\n",
      "\t\t18 of 34 (1e-10) similar to hsa-miR-369-5p MIMAT0001621 Homo sapiens miR-369-5p Targets (miRBase)(0.645)\n",
      "\t\t19 of 34 (1e-10) similar to hsa-miR-223* MIMAT0004570 Homo sapiens miR-223* Targets (miRBase)(0.630)\n",
      "\t\t20 of 34 (1e-9) similar to hsa-miR-4422 MIMAT0018935 Homo sapiens miR-4422 Targets (miRBase)(0.796)\n",
      "\t\t21 of 34 (1e-9) similar to hsa-miR-4278 MIMAT0016910 Homo sapiens miR-4278 Targets (miRBase)(0.812)\n",
      "\t\t22 of 34 (1e-9) similar to hsa-miR-583 MIMAT0003248 Homo sapiens miR-583 Targets (miRBase)(0.642)\n",
      "\t\t23 of 34 (1e-8) similar to hsa-miR-125b-1* MIMAT0004592 Homo sapiens miR-125b-1* Targets (miRBase)(0.691)\n",
      "\t\t24 of 34 (1e-8) similar to hsa-miR-1271 MIMAT0005796 Homo sapiens miR-1271 Targets (miRBase)(0.693)\n",
      "\t\t25 of 34 (1e-8) similar to hsa-miR-4681 MIMAT0019766 Homo sapiens miR-4681 Targets (miRBase)(0.619)\n",
      "\t\t26 of 34 (1e-8) similar to hsa-miR-4486 MIMAT0019020 Homo sapiens miR-4486 Targets (miRBase)(0.714)\n",
      "\t\t27 of 34 (1e-6) similar to hsa-miR-339-3p MIMAT0004702 Homo sapiens miR-339-3p Targets (miRBase)(0.690)\n",
      "\t\t28 of 34 (1e-6) similar to hsa-miR-4324 MIMAT0016876 Homo sapiens miR-4324 Targets (miRBase)(0.666)\n",
      "\t\t29 of 34 (1e-5) similar to hsa-miR-4803 MIMAT0019983 Homo sapiens miR-4803 Targets (miRBase)(0.869)\n",
      "\t\t30 of 34 (1e-5) similar to hsa-miR-4642 MIMAT0019702 Homo sapiens miR-4642 Targets (miRBase)(0.729)\n",
      "\t\t31 of 34 (1e-5) similar to hsa-miR-648 MIMAT0003318 Homo sapiens miR-648 Targets (miRBase)(0.724)\n",
      "\t\t32 of 34 (1e-5) similar to hsa-miR-126 MIMAT0000445 Homo sapiens miR-126 Targets (miRBase)(0.722)\n",
      "\t\t33 of 34 (1e-4) similar to hsa-miR-585 MIMAT0003250 Homo sapiens miR-585 Targets (miRBase)(0.642)\n",
      "\t\t34 of 34 (1e-4) similar to hsa-miR-1249 MIMAT0005901 Homo sapiens miR-1249 Targets (miRBase)(0.637)\n",
      "\tJob finished - if results look good, please send beer to ..\n",
      "\n",
      "\tCleaning up tmp files...\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# -S 10 is changed to -S {nc} = 32\n",
    "# -len 5,6,7,8,9\n",
    "# -bg {genome_bed12}\n",
    "!findMotifsGenome.pl {in_path}/all.narrowPeak hg38 {out_path} \\\n",
    "                     -p {nc}                                  \\\n",
    "                     -rna                                     \\\n",
    "                     -len 5,6,7,8,9                           \\\n",
    "                     -bg {genome_bed12}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "545934f0-bbe2-4e91-bd2f-5f3a577d77ac",
   "metadata": {},
   "source": [
    "#### Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "id": "cb64988d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 264K\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  25K Oct 14 00:45 0.0830720188832892.2.tmp.coBoundBy1.txt\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  25K Oct 14 00:27 0.311316585707576.2.tmp.coBoundBy1.txt\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  31K Oct 14 00:51 homerMotifs.all.motifs\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 3.0K Oct 14 00:47 homerMotifs.motifs5\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 4.7K Oct 14 00:47 homerMotifs.motifs6\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 5.1K Oct 14 00:48 homerMotifs.motifs7\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 8.6K Oct 14 00:49 homerMotifs.motifs8\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 9.2K Oct 14 00:51 homerMotifs.motifs9\n",
      "drwxrwxr-x 2 cdaniels cdaniels  12K Oct 14 00:51 homerResults\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 112K Oct 14 00:51 homerResults.html\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  144 Oct 14 00:45 motifFindingParameters.txt\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 1.9K Oct 14 00:45 seq.autonorm.tsv\n"
     ]
    }
   ],
   "source": [
    "!ls -lh {out_path}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "855de52b",
   "metadata": {},
   "source": [
    "**Warning:** Motif Results are not fully rendered in Jupyter. A browser is better."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "id": "4a4c2c44",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<HTML><HEAD><TITLE>motif/ - Homer de novo Motif Results</TITLE></HEAD>\n",
       "<BODY>\n",
       "<H1>Homer <i>de novo</i> Motif Results (motif/)</H1>\n",
       "<A HREF=\"knownResults.html\">Known Motif Enrichment Results</A><BR/>\n",
       "<A HREF=\"geneOntology.html\">Gene Ontology Enrichment Results</A><BR/>\n",
       "If Homer is having trouble matching a motif to a known motif, try copy/pasting the matrix file into \n",
       "<A HREF=\"http://www.benoslab.pitt.edu/stamp/\">STAMP</A><BR/>\n",
       "More information on motif finding results: <A HREF=\"http://biowhat.ucsd.edu/homer/\">HOMER</A>\n",
       " | <A HREF=\"http://biowhat.ucsd.edu/homer/motif/index.html\">Description of Results</A>\n",
       " | <A HREF=\"http://biowhat.ucsd.edu/homer/motif/practicalTips.html\">Tips</A>\n",
       "<BR/>\n",
       "Total target sequences = 1483<BR/>\n",
       "Total background sequences = 219540<BR/>\n",
       "<FONT color=\"red\">* - possible false positive</FONT><BR/>\n",
       "<TABLE border=\"1\" cellpading=\"0\" cellspacing=\"0\">\n",
       "<TR><TD>Rank</TD><TD>Motif</TD><TD>P-value</TD><TD>log P-pvalue</TD><TD>% of Targets</TD><TD>% of Background</TD>\n",
       "<TD>STD(Bg STD)</TD>\n",
       "<TD>Best Match/Details</TD><TD>Motif File</TD></TR>\n",
       "<TR><TD>1\n",
       "</TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.070,0,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.140,0,45.5)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.256,-1.5,38.5)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.534,2,25.7)\">A</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.070,23.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.074,25,45.5)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.163,27,41.8)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.693,25,33.65)\">G</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.032,52,48.95)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.047,50,47.35)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.920,50,45)\">G</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.025,77,48.95)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.046,75,47.7)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.928,73.5,45.4)\">U</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,98.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.066,100,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.197,100,45.65)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.736,102,35.8)\">A</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.034,125,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.060,125,47.3)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.106,123.5,44.3)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.800,127,39)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.028,150,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.039,148.5,47.6)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.093,152,45.65)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.840,150,41)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.127,177,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.142,175,42.65)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.177,175,35.55)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.554,173.5,26.7)\">U</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-113</TD><TD>-2.602e+02</TD><TD>40.32%</TD><TD>15.76%</TD><TD>39.5bp (68.3bp)</TD><TD>hsa-miR-200a* MIMAT0001620 Homo sapiens miR-200a* Targets (miRBase)(0.655)<BR/><A target=\"_blank\" HREF=\"homerResults/motif1.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif1.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif1.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>2\n",
       "</TD><TD><svg width=\"180\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,-1.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.025,2,48.95)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.099,0,47.7)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.875,0,42.75)\">G</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.019,23.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.064,25,48.05)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.149,25,44.85)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.768,27,37.4)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.076,50,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.079,50,45.2)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.184,48.5,41.25)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.661,52,32.05)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,73.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.089,75,48.9)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.909,75,44.45)\">G</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.021,98.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.038,100,47.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.101,100,46.05)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.840,102,41)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.108,125,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.254,125,43.6)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.292,123.5,30.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.346,127,16.3)\">A</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.033,148.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.047,152,47.35)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.316,150,45)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.604,150,29.2)\">G</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-58</TD><TD>-1.355e+02</TD><TD>34.12%</TD><TD>16.71%</TD><TD>45.2bp (80.8bp)</TD><TD>hsa-miR-4778-3p MIMAT0019937 Homo sapiens miR-4778-3p Targets (miRBase)(0.786)<BR/><A target=\"_blank\" HREF=\"homerResults/motif2.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif2.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif2.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>3\n",
       "</TD><TD><svg width=\"130\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,2,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,-1.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.142,0,48.9)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.856,0,41.8)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,25,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,25,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.144,23.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.854,27,41.7)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,50,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,50,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.156,52,48.9)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.842,48.5,41.1)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,73.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,75,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,102,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,100,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,98.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,100,48.85)\">G</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-27</TD><TD>-6.233e+01</TD><TD>40.80%</TD><TD>27.65%</TD><TD>52.7bp (77.2bp)</TD><TD>hsa-miR-29c* MIMAT0004673 Homo sapiens miR-29c* Targets (miRBase)(0.565)<BR/><A target=\"_blank\" HREF=\"homerResults/motif3.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif3.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif3.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>4\n",
       "</TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,0,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.048,0,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.414,2,46.55)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.537,-1.5,25.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.007,25,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.009,23.5,48.65)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.219,27,48.2)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.765,25,37.25)\">G</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,50,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.198,52,48.95)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.352,48.5,39.05)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.449,50,21.45)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,75,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.466,77,48.9)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.532,73.5,25.6)\">U</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,98.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.046,100,48.95)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.073,102,46.65)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.880,100,43)\">G</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.085,125,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.196,123.5,44.75)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.308,127,34.95)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.412,125,19.55)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,150,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.137,150,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.250,152,42.1)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.612,148.5,29.6)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.037,177,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.164,173.5,47.15)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.233,175,38.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.566,175,27.3)\">G</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.013,198.5,49)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.073,200,48.35)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.351,200,44.7)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.563,202,27.15)\">A</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-26</TD><TD>-6.046e+01</TD><TD>24.75%</TD><TD>14.20%</TD><TD>52.3bp (68.8bp)</TD><TD>hsa-miR-497 MIMAT0002820 Homo sapiens miR-497 Targets (miRBase)(0.670)<BR/><A target=\"_blank\" HREF=\"homerResults/motif4.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif4.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif4.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>5\n",
       "</TD><TD><svg width=\"180\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.041,2,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.071,0,46.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.099,-1.5,43.4)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.789,0,38.45)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.040,27,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.132,25,47)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.145,23.5,40.4)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.683,25,33.15)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.039,50,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.081,50,47.05)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.130,48.5,43)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.750,52,36.5)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.036,75,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.047,75,47.2)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.172,73.5,44.85)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.745,77,36.25)\">A</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.034,100,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.103,102,47.3)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.142,98.5,42.15)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.721,100,35.05)\">C</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.087,123.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.131,125,44.65)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.149,127,38.1)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.633,125,30.65)\">G</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.067,148.5,49)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.106,150,45.65)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.141,150,40.35)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.686,152,33.3)\">A</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-25</TD><TD>-5.984e+01</TD><TD>49.09%</TD><TD>35.56%</TD><TD>53.1bp (72.7bp)</TD><TD>hsa-miR-1306 MIMAT0005950 Homo sapiens miR-1306 Targets (miRBase)(0.743)<BR/><A target=\"_blank\" HREF=\"homerResults/motif5.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif5.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif5.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>6\n",
       "</TD><TD><svg width=\"130\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,0,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.296,2,48.9)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.702,-1.5,34.1)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,25,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.177,23.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.821,25,40.05)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,52,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,50,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,50,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,75,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,73.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,75,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,100,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,100,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,98.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.997,102,48.85)\">A</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-21</TD><TD>-4.917e+01</TD><TD>26.23%</TD><TD>16.36%</TD><TD>50.2bp (78.4bp)</TD><TD>hsa-miR-3610 MIMAT0017987 Homo sapiens miR-3610 Targets (miRBase)(0.719)<BR/><A target=\"_blank\" HREF=\"homerResults/motif6.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif6.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif6.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>7\n",
       "</TD><TD><svg width=\"155\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,2,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,-1.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,0,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,25,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,23.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,25,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,52,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.192,50,48.9)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.806,50,39.3)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,75,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,73.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.997,77,48.85)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,100,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.038,98.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.204,100,47.05)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.757,102,36.85)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,125,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,125,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.113,123.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.885,127,43.25)\">A</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-19</TD><TD>-4.400e+01</TD><TD>11.87%</TD><TD>5.67%</TD><TD>48.8bp (79.1bp)</TD><TD>hsa-miR-1180 MIMAT0005825 Homo sapiens miR-1180 Targets (miRBase)(0.708)<BR/><A target=\"_blank\" HREF=\"homerResults/motif7.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif7.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif7.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>8\n",
       "</TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.041,2,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.116,0,46.95)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.120,0,41.15)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.723,-1.5,35.15)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.044,25,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.195,23.5,46.75)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.760,25,37)\">C</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.080,50,48.95)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.229,52,44.95)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.690,50,33.5)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.014,75,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.017,75,48.3)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.111,73.5,47.45)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.858,77,41.9)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.035,102,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.060,98.5,47.25)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.201,100,44.25)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.704,100,34.2)\">C</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.071,123.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.097,125,45.45)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.250,127,40.6)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.582,125,28.1)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,150,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.030,150,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.060,148.5,47.45)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.909,152,44.45)\">A</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.006,173.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.082,177,48.7)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.139,175,44.6)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.773,175,37.65)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-15</TD><TD>-3.620e+01</TD><TD>20.70%</TD><TD>13.05%</TD><TD>49.9bp (75.1bp)</TD><TD>hsa-miR-4440 MIMAT0018958 Homo sapiens miR-4440 Targets (miRBase)(0.637)<BR/><A target=\"_blank\" HREF=\"homerResults/motif8.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif8.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif8.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>9\n",
       "</TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.014,2,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.079,0,48.3)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.250,0,44.35)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.657,-1.5,31.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,25,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.031,23.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.061,25,47.4)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.907,27,44.35)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,50,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.038,50,48.9)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.960,52,47)\">A</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.024,73.5,48.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.123,77,47.75)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.852,75,41.6)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,100,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,100,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.032,98.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.966,102,47.3)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,127,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,125,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,123.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,125,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,152,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,148.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.063,150,48.9)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.935,150,45.75)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,175,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,175,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.092,177,48.9)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.906,173.5,44.3)\">U</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,198.5,49)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.021,200,48.95)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.114,200,47.9)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.864,202,42.2)\">A</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-14</TD><TD>-3.447e+01</TD><TD>3.44%</TD><TD>0.88%</TD><TD>52.0bp (69.4bp)</TD><TD>hsa-miR-3144-3p MIMAT0015015 Homo sapiens miR-3144-3p Targets (miRBase)(0.699)<BR/><A target=\"_blank\" HREF=\"homerResults/motif9.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif9.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif9.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>10\n",
       "</TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.055,-1.5,49)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.060,0,46.25)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.089,2,43.25)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.796,0,38.8)\">C</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.047,23.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.053,27,46.65)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.101,25,44)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.799,25,38.95)\">C</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.037,48.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.050,50,47.15)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.080,52,44.65)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.833,50,40.65)\">G</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.066,75,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.085,75,45.7)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.122,77,41.45)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.727,73.5,35.35)\">U</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.069,98.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.136,102,45.55)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.186,100,38.75)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.609,100,29.45)\">G</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.067,125,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.080,123.5,45.65)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.111,125,41.65)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.742,127,36.1)\">A</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.050,150,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.077,148.5,46.5)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.139,150,42.65)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.734,152,35.7)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.039,175,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.074,173.5,47.05)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.116,177,43.35)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.771,175,37.55)\">G</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-13</TD><TD>-3.163e+01</TD><TD>13.42%</TD><TD>7.66%</TD><TD>54.4bp (75.8bp)</TD><TD>hsa-miR-412 MIMAT0002170 Homo sapiens miR-412 Targets (miRBase)(0.608)<BR/><A target=\"_blank\" HREF=\"homerResults/motif10.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif10.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif10.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>11\n",
       "</TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.031,2,48.95)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.234,0,47.4)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.734,-1.5,35.7)\">U</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,25,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,23.5,48.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.049,27,48.9)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.949,25,46.45)\">C</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.024,48.5,49)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.088,50,47.8)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.146,50,43.4)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.742,52,36.1)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,75,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.997,73.5,48.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,100,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,98.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.050,100,48.9)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.948,102,46.4)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,127,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,125,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,123.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,125,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.027,148.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.032,152,47.65)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.138,150,46.05)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.803,150,39.15)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,177,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,175,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.029,173.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.969,175,47.45)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,202,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,200,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.066,198.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.932,200,45.6)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-13</TD><TD>-3.126e+01</TD><TD>4.52%</TD><TD>1.54%</TD><TD>55.6bp (64.4bp)</TD><TD>hsa-miR-3201 MIMAT0015086 Homo sapiens miR-3201 Targets (miRBase)(0.712)<BR/><A target=\"_blank\" HREF=\"homerResults/motif11.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif11.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif11.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>12\n",
       "</TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.099,0,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.105,2,44.05)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.188,-1.5,38.8)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.608,0,29.4)\">C</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.050,23.5,49)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.058,25,46.5)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.064,27,43.6)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.828,25,40.4)\">C</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.054,50,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.072,50,46.3)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.107,48.5,42.7)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.767,52,37.35)\">A</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.046,75,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.073,77,46.7)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.103,75,43.05)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.778,73.5,37.9)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.023,102,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.030,100,47.85)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.074,100,46.35)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.873,98.5,42.65)\">U</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.050,123.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.073,127,46.5)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.092,125,42.85)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.785,125,38.25)\">G</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.043,148.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.130,152,46.85)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.134,150,40.35)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.693,150,33.65)\">G</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.090,175,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.105,177,44.5)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.125,173.5,39.25)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.680,175,33)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-13</TD><TD>-3.053e+01</TD><TD>13.96%</TD><TD>8.17%</TD><TD>57.9bp (64.6bp)</TD><TD>hsa-miR-4252 MIMAT0016886 Homo sapiens miR-4252 Targets (miRBase)(0.673)<BR/><A target=\"_blank\" HREF=\"homerResults/motif12.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif12.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif12.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>13\n",
       "</TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.003,-1.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.042,0,48.85)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.116,2,46.75)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.839,0,40.95)\">G</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.003,25,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.059,25,48.85)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.061,23.5,45.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.877,27,42.85)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.010,52,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.081,48.5,48.5)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.184,50,44.45)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.725,50,35.25)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.098,75,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.132,75,44.1)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.253,77,37.5)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.517,73.5,24.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.003,100,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.142,98.5,48.85)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.166,100,41.75)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.689,102,33.45)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.003,127,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.142,125,48.85)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.158,125,41.75)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.697,123.5,33.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.044,148.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.060,152,46.8)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.107,150,43.8)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.789,150,38.45)\">G</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.025,175,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.089,173.5,47.75)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.108,175,43.3)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.778,177,37.9)\">A</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.007,200,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.107,198.5,48.65)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.236,202,43.3)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.650,200,31.5)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-12</TD><TD>-2.894e+01</TD><TD>10.52%</TD><TD>5.68%</TD><TD>54.6bp (73.0bp)</TD><TD>hsa-miR-103b MIMAT0007402 Homo sapiens miR-103b Targets (miRBase)(0.588)<BR/><A target=\"_blank\" HREF=\"homerResults/motif13.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif13.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif13.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>14\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,2,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,0,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,-1.5,39)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.700,0,34)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,27,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,25,44)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,23.5,39)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.700,25,34)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,50,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,50,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,48.5,39)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.700,52,34)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,77,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,75,44)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,75,39)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.700,73.5,34)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,102,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,100,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,98.5,39)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.700,100,34)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,127,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,125,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,123.5,39)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.700,125,34)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,152,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,150,44)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,148.5,39)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.700,150,34)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,177,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,175,44)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,173.5,39)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.700,175,34)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,200,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,200,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,198.5,39)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.700,202,34)\">A</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-11</TD><TD>-2.752e+01</TD><TD>12.68%</TD><TD>7.43%</TD><TD>51.6bp (75.3bp)</TD><TD>hsa-miR-671-3p MIMAT0004819 Homo sapiens miR-671-3p Targets (miRBase)(0.650)<BR/><A target=\"_blank\" HREF=\"homerResults/motif14.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif14.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif14.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>15\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,0,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.278,2,48.9)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.720,-1.5,35)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,25,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.023,25,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.026,27,47.8)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.950,23.5,46.5)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,50,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,50,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.106,52,48.9)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.892,48.5,43.6)\">U</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.009,75,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.065,77,48.55)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.081,75,45.3)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.845,73.5,41.25)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,100,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.055,98.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.061,100,46.2)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.883,102,43.15)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,127,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.009,125,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.026,123.5,48.5)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.964,125,47.2)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,150,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,150,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.132,148.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.866,152,42.3)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,175,49)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.016,177,48.95)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.079,173.5,48.15)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.904,175,44.2)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,202,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,200,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,198.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,200,48.85)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-10</TD><TD>-2.510e+01</TD><TD>3.03%</TD><TD>0.92%</TD><TD>54.2bp (65.7bp)</TD><TD>hsa-miR-18b* MIMAT0004751 Homo sapiens miR-18b* Targets (miRBase)(0.672)<BR/><A target=\"_blank\" HREF=\"homerResults/motif15.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif15.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif15.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>16\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"155\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,0,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.124,2,48.9)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.874,-1.5,42.7)\">U</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,25,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,23.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.111,25,48.9)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.887,27,43.35)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,52,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,50,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.179,48.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.819,50,39.95)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,75,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.112,77,48.9)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.886,73.5,43.3)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,100,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,100,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.125,98.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.873,102,42.65)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,127,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,125,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,123.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,125,48.85)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-10</TD><TD>-2.476e+01</TD><TD>24.34%</TD><TD>17.49%</TD><TD>53.8bp (68.9bp)</TD><TD>hsa-miR-20b* MIMAT0004752 Homo sapiens miR-20b* Targets (miRBase)(0.712)<BR/><A target=\"_blank\" HREF=\"homerResults/motif16.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif16.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif16.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>17\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,2,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.065,-1.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.933,0,45.65)\">C</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,25,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.069,23.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.093,25,45.5)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.837,27,40.85)\">A</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.031,50,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.063,50,47.4)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.905,52,44.25)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.006,73.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.031,75,48.65)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.962,75,47.1)\">C</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.019,100,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.044,100,48.05)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.052,98.5,45.85)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.885,102,43.25)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,127,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.069,123.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.154,125,45.5)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.776,125,37.8)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,152,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,148.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.032,150,48.9)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.966,150,47.3)\">C</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,175,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.062,177,48.95)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.178,173.5,45.85)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.759,175,36.95)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-10</TD><TD>-2.432e+01</TD><TD>9.78%</TD><TD>5.48%</TD><TD>50.3bp (67.0bp)</TD><TD>hsa-miR-4687-3p MIMAT0019775 Homo sapiens miR-4687-3p Targets (miRBase)(0.844)<BR/><A target=\"_blank\" HREF=\"homerResults/motif17.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif17.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif17.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>18\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,2,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,0,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,-1.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,0,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,25,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,23.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,25,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,52,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,50,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,50,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.997,48.5,48.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,73.5,48.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.294,77,48.9)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.704,75,34.2)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,102,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,100,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,98.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,100,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,125,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,125,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,123.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.997,127,48.85)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,152,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,150,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,148.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,150,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,177,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,175,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,173.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,175,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,200,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.016,198.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.288,200,48.15)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.695,202,33.75)\">A</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-10</TD><TD>-2.417e+01</TD><TD>1.01%</TD><TD>0.09%</TD><TD>35.1bp (69.0bp)</TD><TD>hsa-miR-369-5p MIMAT0001621 Homo sapiens miR-369-5p Targets (miRBase)(0.645)<BR/><A target=\"_blank\" HREF=\"homerResults/motif18.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif18.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif18.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>19\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,0,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,-1.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.997,2,48.85)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,23.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.077,25,48.9)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.921,25,45.05)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,52,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.051,50,48.95)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.074,50,46.4)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.874,48.5,42.7)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,75,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.072,73.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.926,77,45.3)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,102,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,98.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.093,100,48.9)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.905,100,44.25)\">C</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,125,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.076,125,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.137,123.5,45.15)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.786,127,38.3)\">A</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.019,150,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.042,152,48.05)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.138,148.5,45.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.801,150,39.05)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,175,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.052,173.5,48.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.080,177,46.35)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.867,175,42.35)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,202,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,200,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,198.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,200,48.85)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-10</TD><TD>-2.335e+01</TD><TD>1.42%</TD><TD>0.23%</TD><TD>53.6bp (68.4bp)</TD><TD>hsa-miR-223* MIMAT0004570 Homo sapiens miR-223* Targets (miRBase)(0.630)<BR/><A target=\"_blank\" HREF=\"homerResults/motif19.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif19.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif19.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>20\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"180\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,0,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.051,0,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.065,2,46.4)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.883,-1.5,43.15)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.039,25,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.110,23.5,47)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.850,25,41.5)\">G</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,49)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.057,50,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.133,52,46.1)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.809,50,39.45)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.035,75,48.95)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.052,75,47.2)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.912,73.5,44.6)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,102,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.048,100,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.071,100,46.55)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.880,98.5,43)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,127,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,125,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.080,125,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.918,123.5,44.9)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,152,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.060,150,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.125,150,45.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.814,148.5,39.7)\">U</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-9</TD><TD>-2.180e+01</TD><TD>47.67%</TD><TD>39.73%</TD><TD>63.9bp (63.4bp)</TD><TD>hsa-miR-4422 MIMAT0018935 Homo sapiens miR-4422 Targets (miRBase)(0.796)<BR/><A target=\"_blank\" HREF=\"homerResults/motif20.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif20.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif20.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>21\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"155\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,2,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,-1.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,0,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,25,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,23.5,48.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.176,27,48.9)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.822,25,40.1)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,52,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,50,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,50,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,73.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.117,75,48.9)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.881,75,43.05)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,102,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,100,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.181,98.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.817,100,39.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,127,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,125,48.95)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.173,125,48.9)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.825,123.5,40.25)\">U</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-9</TD><TD>-2.121e+01</TD><TD>20.16%</TD><TD>14.33%</TD><TD>58.3bp (65.1bp)</TD><TD>hsa-miR-4278 MIMAT0016910 Homo sapiens miR-4278 Targets (miRBase)(0.812)<BR/><A target=\"_blank\" HREF=\"homerResults/motif21.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif21.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif21.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>22\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,0,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.052,2,48.9)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.946,-1.5,46.3)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,25,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.099,25,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.148,23.5,44)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.752,27,36.6)\">A</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,49)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.034,50,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.096,52,47.25)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.869,50,42.45)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,75,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.118,75,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.880,73.5,43)\">U</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,98.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.040,102,48.95)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.140,100,46.95)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.819,100,39.95)\">C</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,125,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.007,127,48.95)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.010,125,48.6)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.982,123.5,48.1)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,152,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,150,48.95)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.019,150,48.9)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.979,148.5,47.95)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.023,177,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.052,175,47.85)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.058,175,45.25)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.867,173.5,42.35)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,200,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.011,200,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.012,202,48.4)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.976,198.5,47.8)\">U</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-9</TD><TD>-2.088e+01</TD><TD>4.59%</TD><TD>2.01%</TD><TD>65.4bp (65.1bp)</TD><TD>hsa-miR-583 MIMAT0003248 Homo sapiens miR-583 Targets (miRBase)(0.642)<BR/><A target=\"_blank\" HREF=\"homerResults/motif22.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif22.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif22.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>23\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.032,-1.5,49)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.044,0,47.4)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.187,2,45.2)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.737,0,35.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,23.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.037,25,48.95)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.043,27,47.1)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.919,25,44.95)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,52,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,50,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.173,48.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.825,50,40.25)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,75,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.997,73.5,48.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,100,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.037,100,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.058,98.5,47.1)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.904,102,44.2)\">A</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,125,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,123.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.048,125,48.9)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.950,127,46.5)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,152,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,150,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.084,148.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.914,150,44.7)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,177,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,175,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.077,173.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.921,175,45.05)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,202,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,200,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.073,198.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.925,200,45.25)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-8</TD><TD>-2.002e+01</TD><TD>0.81%</TD><TD>0.07%</TD><TD>57.8bp (71.2bp)</TD><TD>hsa-miR-125b-1* MIMAT0004592 Homo sapiens miR-125b-1* Targets (miRBase)(0.691)<BR/><A target=\"_blank\" HREF=\"homerResults/motif23.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif23.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif23.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>24\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,2,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,0,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.389,-1.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.609,0,29.45)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,25,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.356,23.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.642,25,31.1)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,52,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,50,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,50,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,73.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,75,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,100,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.002,98.5,48.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.003,102,48.85)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.994,100,48.7)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,125,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,125,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,123.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.997,127,48.85)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,150,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,150,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,148.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.997,152,48.85)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,177,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,175,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.027,175,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.971,173.5,47.55)\">U</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-8</TD><TD>-1.899e+01</TD><TD>2.23%</TD><TD>0.67%</TD><TD>56.9bp (73.8bp)</TD><TD>hsa-miR-1271 MIMAT0005796 Homo sapiens miR-1271 Targets (miRBase)(0.693)<BR/><A target=\"_blank\" HREF=\"homerResults/motif24.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif24.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif24.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>25\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"155\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,2,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,0,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.997,-1.5,48.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,25,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,25,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.997,23.5,48.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.048,50,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.067,52,46.6)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.211,48.5,43.25)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.674,50,32.7)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.086,73.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.101,75,44.65)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.812,75,39.6)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,102,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.141,98.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.156,100,41.9)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.702,100,34.1)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,125,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.036,123.5,48.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.113,127,47.15)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.850,125,41.5)\">G</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-8</TD><TD>-1.874e+01</TD><TD>12.74%</TD><TD>8.36%</TD><TD>58.0bp (72.1bp)</TD><TD>hsa-miR-4681 MIMAT0019766 Homo sapiens miR-4681 Targets (miRBase)(0.619)<BR/><A target=\"_blank\" HREF=\"homerResults/motif25.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif25.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif25.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>26\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,0,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.101,2,48.9)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.897,-1.5,43.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,25,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,23.5,48.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.116,27,48.9)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.882,25,43.1)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,52,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,50,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,50,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,73.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,75,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,100,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,100,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.184,102,48.9)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.814,98.5,39.7)\">U</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,125,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.117,127,48.95)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.129,123.5,43.1)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.753,125,36.65)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,150,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,150,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,148.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.997,152,48.85)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,177,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.117,175,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.223,173.5,43.1)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.659,175,31.95)\">G</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-8</TD><TD>-1.853e+01</TD><TD>2.36%</TD><TD>0.76%</TD><TD>38.3bp (71.0bp)</TD><TD>hsa-miR-4486 MIMAT0019020 Homo sapiens miR-4486 Targets (miRBase)(0.714)<BR/><A target=\"_blank\" HREF=\"homerResults/motif26.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif26.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif26.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>27\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.105,0,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.187,-1.5,43.75)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.200,2,34.4)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.508,0,24.4)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.098,27,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.140,25,44.1)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.295,23.5,37.1)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.466,25,22.35)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.005,52,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.027,50,48.75)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.068,48.5,47.4)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.900,50,44)\">C</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.114,73.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.143,77,43.3)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.204,75,36.15)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.539,75,25.95)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.163,102,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.245,100,40.85)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.255,98.5,28.6)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.337,100,15.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,125,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.126,127,44)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.233,125,37.7)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.541,123.5,26.05)\">U</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.013,150,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.173,148.5,48.35)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.181,152,39.7)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.633,150,30.65)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.128,175,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.139,175,42.6)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.202,177,35.65)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.531,173.5,25.55)\">U</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.110,200,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.238,202,43.5)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.273,198.5,31.6)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.379,200,17.95)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-6</TD><TD>-1.599e+01</TD><TD>3.10%</TD><TD>1.29%</TD><TD>57.6bp (65.0bp)</TD><TD>hsa-miR-339-3p MIMAT0004702 Homo sapiens miR-339-3p Targets (miRBase)(0.690)<BR/><A target=\"_blank\" HREF=\"homerResults/motif27.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif27.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif27.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>28\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,2,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,-1.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.184,0,48.9)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.814,0,39.7)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,25,48.95)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.087,25,48.9)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.911,23.5,44.55)\">U</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,49)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.015,50,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.030,52,48.2)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.954,50,46.7)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,75,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.050,75,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.112,77,46.45)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.837,73.5,40.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,100,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,98.5,48.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.042,102,48.9)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.956,100,46.8)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.019,125,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.029,123.5,48.05)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.036,125,46.6)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.916,127,44.8)\">A</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.028,150,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.047,152,47.6)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.078,148.5,45.25)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.847,150,41.35)\">C</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,173.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.109,177,48.95)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.129,175,43.5)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.761,175,37.05)\">G</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-6</TD><TD>-1.592e+01</TD><TD>3.84%</TD><TD>1.78%</TD><TD>51.1bp (66.6bp)</TD><TD>hsa-miR-4324 MIMAT0016876 Homo sapiens miR-4324 Targets (miRBase)(0.666)<BR/><A target=\"_blank\" HREF=\"homerResults/motif28.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif28.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif28.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>29\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,0,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.247,-1.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.751,2,36.55)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,25,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,25,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.997,23.5,48.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,52,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,50,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,50,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.997,48.5,48.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,75,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,73.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.128,75,48.9)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.870,77,42.5)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,102,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,100,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,100,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.997,98.5,48.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,127,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,125,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,123.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,125,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,152,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,150,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,150,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.997,148.5,48.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,177,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,175,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,175,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.997,173.5,48.85)\">U</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-5</TD><TD>-1.371e+01</TD><TD>3.71%</TD><TD>1.82%</TD><TD>62.8bp (63.6bp)</TD><TD>hsa-miR-4803 MIMAT0019983 Homo sapiens miR-4803 Targets (miRBase)(0.869)<BR/><A target=\"_blank\" HREF=\"homerResults/motif29.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif29.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif29.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>30\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"130\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,2,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,0,44)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,-1.5,39)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.700,0,34)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,25,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,25,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,23.5,39)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.700,27,34)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,52,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,50,44)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,50,39)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.700,48.5,34)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,77,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,75,44)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,73.5,39)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.700,75,34)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,102,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,100,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,98.5,39)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.700,100,34)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-5</TD><TD>-1.268e+01</TD><TD>17.19%</TD><TD>13.05%</TD><TD>56.0bp (71.8bp)</TD><TD>hsa-miR-4642 MIMAT0019702 Homo sapiens miR-4642 Targets (miRBase)(0.729)<BR/><A target=\"_blank\" HREF=\"homerResults/motif30.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif30.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif30.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>31\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"155\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,-1.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.188,2,48.95)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.244,0,39.55)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.567,0,27.35)\">G</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.027,25,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.045,23.5,47.65)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.290,27,45.4)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.638,25,30.9)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.094,50,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.136,48.5,44.3)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.372,52,37.5)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.398,50,18.9)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.062,75,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.217,73.5,45.85)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.720,75,35)\">C</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.051,100,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.055,100,46.45)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.073,98.5,43.7)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.821,102,40.05)\">A</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,125,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.026,123.5,48.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.068,127,47.65)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.905,125,44.25)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-5</TD><TD>-1.221e+01</TD><TD>13.08%</TD><TD>9.52%</TD><TD>46.4bp (70.3bp)</TD><TD>hsa-miR-648 MIMAT0003318 Homo sapiens miR-648 Targets (miRBase)(0.724)<BR/><A target=\"_blank\" HREF=\"homerResults/motif31.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif31.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif31.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>32\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,2,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,0,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,-1.5,39)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.700,0,34)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,27,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,25,44)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,23.5,39)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.700,25,34)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,52,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,50,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,48.5,39)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.700,50,34)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,77,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,75,44)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,73.5,39)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.700,75,34)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,102,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,100,44)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,98.5,39)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.700,100,34)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,127,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,125,44)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,125,39)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.700,123.5,34)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,150,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,150,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,148.5,39)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.700,152,34)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,177,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,175,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,173.5,39)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.700,175,34)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-5</TD><TD>-1.168e+01</TD><TD>0.34%</TD><TD>0.02%</TD><TD>40.8bp (70.3bp)</TD><TD>hsa-miR-126 MIMAT0000445 Homo sapiens miR-126 Targets (miRBase)(0.722)<BR/><A target=\"_blank\" HREF=\"homerResults/motif32.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif32.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif32.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>33\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.002,2,48.95)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.035,-1.5,48.85)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.962,0,47.1)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,25,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,25,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,23.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.997,27,48.85)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,52,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,50,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.008,48.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.990,50,48.5)\">C</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,73.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.037,75,48.9)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.961,77,47.05)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,100,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.006,100,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.387,102,48.65)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.606,98.5,29.3)\">U</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,125,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.007,125,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.047,123.5,48.6)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.945,127,46.25)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,152,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,150,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,148.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,150,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,177,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,175,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,173.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,175,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,202,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,200,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,198.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,200,48.85)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-4</TD><TD>-1.138e+01</TD><TD>0.47%</TD><TD>0.05%</TD><TD>38.8bp (61.6bp)</TD><TD>hsa-miR-585 MIMAT0003250 Homo sapiens miR-585 Targets (miRBase)(0.642)<BR/><A target=\"_blank\" HREF=\"homerResults/motif33.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif33.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif33.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>34\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,0,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,0,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,-1.5,39)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.700,2,34)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,25,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,25,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,23.5,39)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.700,27,34)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,50,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,50,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,48.5,39)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.700,52,34)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,77,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,75,44)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,73.5,39)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.700,75,34)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,102,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,100,44)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,98.5,39)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.700,100,34)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,127,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,125,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,123.5,39)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.700,125,34)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,152,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,150,44)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,148.5,39)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.700,150,34)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,177,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,175,44)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,173.5,39)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.700,175,34)\">G</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-4</TD><TD>-1.070e+01</TD><TD>0.67%</TD><TD>0.12%</TD><TD>55.4bp (74.6bp)</TD><TD>hsa-miR-1249 MIMAT0005901 Homo sapiens miR-1249 Targets (miRBase)(0.637)<BR/><A target=\"_blank\" HREF=\"homerResults/motif34.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif34.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif34.motif\">motif file (matrix)</A></TD></TR>\n",
       "</TABLE>\n",
       "<P></P>\n",
       "</BODY></HTML>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import HTML\n",
    "HTML(filename=out_path/\"homerResults.html\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cdf35e2b-9420-4b27-89e1-3c09b1102610",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true
   },
   "source": [
    "# Combination Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 554,
   "id": "68167f2a-13bb-48f7-96d6-a0d01f6e0aca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Original bbbreviated sample names\n",
    "#treated = [\"a4_1\",\"a4_2\",\"a4_3\",\"a4_4\"] # Edit\n",
    "#control = [\"no_1\",\"no_2\",\"no_3\",\"no_4\"] # Edit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 555,
   "id": "10d5f46f-db19-4829-b67e-4e53b12633a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Original samples\n",
    "#samples = treated + control\n",
    "#samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 730,
   "id": "6d38fcbc-0a18-47f8-8f4e-7b91f71570bd",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['a4_1', 'a4_2', 'a4_3', 'a4_4']\n",
      "['no_1', 'no_2', 'no_3', 'no_4']\n",
      "['a4_1', 'a4_2', 'a4_3', 'a4_4', 'no_1', 'no_2', 'no_3', 'no_4']\n",
      "_1234@10-12-14\n"
     ]
    }
   ],
   "source": [
    "# Use these for creating combinations\n",
    "treated = [\"a4_1\",\"a4_2\",\"a4_3\",\"a4_4\"]\n",
    "control = [\"no_1\",\"no_2\",\"no_3\",\"no_4\"]\n",
    "samples = treated + control\n",
    "tag = \"1234@10-12-14\" # Tags should be of form: \"1\", \"234\", \"12_12\", \"1_12; 1234@10-12-14 means samples 1,2,3,4 w/len=10-12-14\"\n",
    "\n",
    "###############\n",
    "tag = \"_\" + tag\n",
    "print(treated)\n",
    "print(control)\n",
    "print(samples)\n",
    "print(tag)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0791bfab-7193-4b90-a802-88140005f8c3",
   "metadata": {},
   "source": [
    "## Step: peak_fwd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc08cddf-198a-438b-bc42-3fe7c73afd9a",
   "metadata": {},
   "source": [
    "#### Call peaks from forward reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 731,
   "id": "857ccc16-071b-42fe-a292-19b743ee915b",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_path   = mkpath(\"split_fwd\")\n",
    "out_path  = mkpath(f\"peak_fwd{tag}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 732,
   "id": "17dba4a6-0d7b-428d-8691-c85e6b8e760f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "peak_fwd_1234@10-12-14\n"
     ]
    }
   ],
   "source": [
    "print(out_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 733,
   "id": "3975418d-8bbe-4e2f-939a-a97f1d51d45c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "split_fwd/a4_1.bam split_fwd/a4_2.bam split_fwd/a4_3.bam split_fwd/a4_4.bam\n",
      "split_fwd/no_1.bam split_fwd/no_2.bam split_fwd/no_3.bam split_fwd/no_4.bam\n"
     ]
    }
   ],
   "source": [
    "print(samples_string(treated,in_path))\n",
    "print(samples_string(control,in_path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 734,
   "id": "69501e11-430c-475e-b5da-ff055be6a563",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  @ 14 Oct 2023 05:24:02: [61 MB] \n",
      "# Command line: callpeak --treatment split_fwd/a4_1.bam split_fwd/a4_2.bam split_fwd/a4_3.bam split_fwd/a4_4.bam --control split_fwd/no_1.bam split_fwd/no_2.bam split_fwd/no_3.bam split_fwd/no_4.bam -f BAM -n fwd -g hs --keep-dup all --outdir peak_fwd_1234@10-12-14 --nomodel --extsize 30\n",
      "# ARGUMENTS LIST:\n",
      "# name = fwd\n",
      "# format = BAM\n",
      "# ChIP-seq file = ['split_fwd/a4_1.bam', 'split_fwd/a4_2.bam', 'split_fwd/a4_3.bam', 'split_fwd/a4_4.bam']\n",
      "# control file = ['split_fwd/no_1.bam', 'split_fwd/no_2.bam', 'split_fwd/no_3.bam', 'split_fwd/no_4.bam']\n",
      "# effective genome size = 2.91e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 5.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 1000 bps and 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is off\n",
      " \n",
      "INFO  @ 14 Oct 2023 05:24:02: [61 MB] #1 read tag files... \n",
      "INFO  @ 14 Oct 2023 05:24:02: [61 MB] #1 read treatment tags... \n",
      "INFO  @ 14 Oct 2023 05:24:02: [85 MB] 869510 reads have been read. \n",
      "INFO  @ 14 Oct 2023 05:24:03: [96 MB] 233566 reads have been read. \n",
      "INFO  @ 14 Oct 2023 05:24:03: [96 MB] 703905 reads have been read. \n",
      "INFO  @ 14 Oct 2023 05:24:04: [99 MB] 520846 reads have been read. \n",
      "INFO  @ 14 Oct 2023 05:24:04: [99 MB] #1.2 read input tags... \n",
      "INFO  @ 14 Oct 2023 05:24:04: [116 MB] 557795 reads have been read. \n",
      "INFO  @ 14 Oct 2023 05:24:05: [127 MB] 555966 reads have been read. \n",
      "INFO  @ 14 Oct 2023 05:24:05: [133 MB] 431273 reads have been read. \n",
      "INFO  @ 14 Oct 2023 05:24:06: [140 MB] 615633 reads have been read. \n",
      "INFO  @ 14 Oct 2023 05:24:06: [140 MB] #1 tag size is determined as 50 bps \n",
      "INFO  @ 14 Oct 2023 05:24:06: [140 MB] #1 tag size = 50.0 \n",
      "INFO  @ 14 Oct 2023 05:24:06: [140 MB] #1  total tags in treatment: 2327827 \n",
      "INFO  @ 14 Oct 2023 05:24:06: [140 MB] #1  total tags in control: 2160667 \n",
      "INFO  @ 14 Oct 2023 05:24:06: [140 MB] #1 finished! \n",
      "INFO  @ 14 Oct 2023 05:24:06: [140 MB] #2 Build Peak Model... \n",
      "INFO  @ 14 Oct 2023 05:24:06: [140 MB] #2 Skipped... \n",
      "INFO  @ 14 Oct 2023 05:24:06: [140 MB] #2 Use 30 as fragment length \n",
      "INFO  @ 14 Oct 2023 05:24:06: [140 MB] #3 Call peaks... \n",
      "INFO  @ 14 Oct 2023 05:24:06: [140 MB] #3 Pre-compute pvalue-qvalue table... \n",
      "INFO  @ 14 Oct 2023 05:24:11: [181 MB] #3 Call peaks for each chromosome... \n",
      "INFO  @ 14 Oct 2023 05:24:11: [181 MB] #4 Write output xls file... peak_fwd_1234@10-12-14/fwd_peaks.xls \n",
      "INFO  @ 14 Oct 2023 05:24:11: [181 MB] #4 Write peak in narrowPeak format file... peak_fwd_1234@10-12-14/fwd_peaks.narrowPeak \n",
      "INFO  @ 14 Oct 2023 05:24:11: [181 MB] #4 Write summits bed file... peak_fwd_1234@10-12-14/fwd_summits.bed \n",
      "INFO  @ 14 Oct 2023 05:24:11: [181 MB] Done! \n"
     ]
    }
   ],
   "source": [
    "! macs3 callpeak                                  \\\n",
    "    --treatment {samples_string(treated,in_path)} \\\n",
    "    --control   {samples_string(control,in_path)} \\\n",
    "    -f BAM                                        \\\n",
    "    -n fwd                                        \\\n",
    "    -g hs                                         \\\n",
    "    --keep-dup all                                \\\n",
    "    --outdir {out_path}                           \\\n",
    "    --nomodel --extsize 30"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c055fc30-8280-4a84-8356-ac8287dd3849",
   "metadata": {},
   "source": [
    "#### Mark 6th column with a **+** to indicate a `fwd` peak (sed is 0 based)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 735,
   "id": "a921df2c-a853-4c72-a581-c1c91cc15853",
   "metadata": {},
   "outputs": [],
   "source": [
    "!sed -i 's/\\t[^\\t]*/\\t+/5' {out_path}/*.narrowPeak"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4dd288b-d5b1-4325-b9e0-7070b4e23ffd",
   "metadata": {},
   "source": [
    "## Step: peak_rev"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "78711214-b7f7-442d-90b8-358b311be6fa",
   "metadata": {},
   "source": [
    "#### Call peaks from reverse reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 736,
   "id": "1f03ae25-b184-456e-a11d-c6e636f6105f",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_path   = mkpath(\"split_rev\")\n",
    "out_path  = mkpath(f\"peak_rev{tag}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 737,
   "id": "bdce4fcb-c6ed-4cac-bc79-d94e894cc337",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "split_rev/a4_1.bam split_rev/a4_2.bam split_rev/a4_3.bam split_rev/a4_4.bam\n",
      "split_rev/no_1.bam split_rev/no_2.bam split_rev/no_3.bam split_rev/no_4.bam\n"
     ]
    }
   ],
   "source": [
    "print(samples_string(treated,in_path))\n",
    "print(samples_string(control,in_path))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 738,
   "id": "24d099fa-fa05-42d1-95f9-379d17bbf346",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "INFO  @ 14 Oct 2023 05:24:12: [61 MB] \n",
      "# Command line: callpeak --treatment split_rev/a4_1.bam split_rev/a4_2.bam split_rev/a4_3.bam split_rev/a4_4.bam --control split_rev/no_1.bam split_rev/no_2.bam split_rev/no_3.bam split_rev/no_4.bam -f BAM -n fwd -g hs --keep-dup all --outdir peak_rev_1234@10-12-14 --nomodel --extsize 30\n",
      "# ARGUMENTS LIST:\n",
      "# name = fwd\n",
      "# format = BAM\n",
      "# ChIP-seq file = ['split_rev/a4_1.bam', 'split_rev/a4_2.bam', 'split_rev/a4_3.bam', 'split_rev/a4_4.bam']\n",
      "# control file = ['split_rev/no_1.bam', 'split_rev/no_2.bam', 'split_rev/no_3.bam', 'split_rev/no_4.bam']\n",
      "# effective genome size = 2.91e+09\n",
      "# band width = 300\n",
      "# model fold = [5, 50]\n",
      "# qvalue cutoff = 5.00e-02\n",
      "# The maximum gap between significant sites is assigned as the read length/tag size.\n",
      "# The minimum length of peaks is assigned as the predicted fragment length \"d\".\n",
      "# Larger dataset will be scaled towards smaller dataset.\n",
      "# Range for calculating regional lambda is: 1000 bps and 10000 bps\n",
      "# Broad region calling is off\n",
      "# Paired-End mode is off\n",
      " \n",
      "INFO  @ 14 Oct 2023 05:24:12: [61 MB] #1 read tag files... \n",
      "INFO  @ 14 Oct 2023 05:24:12: [61 MB] #1 read treatment tags... \n",
      "INFO  @ 14 Oct 2023 05:24:12: [76 MB] 609737 reads have been read. \n",
      "INFO  @ 14 Oct 2023 05:24:12: [87 MB] 146306 reads have been read. \n",
      "INFO  @ 14 Oct 2023 05:24:13: [90 MB] 559619 reads have been read. \n",
      "INFO  @ 14 Oct 2023 05:24:13: [92 MB] 398498 reads have been read. \n",
      "INFO  @ 14 Oct 2023 05:24:13: [92 MB] #1.2 read input tags... \n",
      "INFO  @ 14 Oct 2023 05:24:14: [111 MB] 420347 reads have been read. \n",
      "INFO  @ 14 Oct 2023 05:24:14: [121 MB] 441908 reads have been read. \n",
      "INFO  @ 14 Oct 2023 05:24:15: [124 MB] 336369 reads have been read. \n",
      "INFO  @ 14 Oct 2023 05:24:15: [126 MB] 474977 reads have been read. \n",
      "INFO  @ 14 Oct 2023 05:24:15: [126 MB] #1 tag size is determined as 45 bps \n",
      "INFO  @ 14 Oct 2023 05:24:15: [126 MB] #1 tag size = 45.0 \n",
      "INFO  @ 14 Oct 2023 05:24:15: [126 MB] #1  total tags in treatment: 1714160 \n",
      "INFO  @ 14 Oct 2023 05:24:15: [126 MB] #1  total tags in control: 1673601 \n",
      "INFO  @ 14 Oct 2023 05:24:15: [126 MB] #1 finished! \n",
      "INFO  @ 14 Oct 2023 05:24:15: [126 MB] #2 Build Peak Model... \n",
      "INFO  @ 14 Oct 2023 05:24:15: [126 MB] #2 Skipped... \n",
      "INFO  @ 14 Oct 2023 05:24:15: [126 MB] #2 Use 30 as fragment length \n",
      "INFO  @ 14 Oct 2023 05:24:15: [126 MB] #3 Call peaks... \n",
      "INFO  @ 14 Oct 2023 05:24:15: [126 MB] #3 Pre-compute pvalue-qvalue table... \n",
      "INFO  @ 14 Oct 2023 05:24:19: [174 MB] #3 Call peaks for each chromosome... \n",
      "INFO  @ 14 Oct 2023 05:24:20: [174 MB] #4 Write output xls file... peak_rev_1234@10-12-14/fwd_peaks.xls \n",
      "INFO  @ 14 Oct 2023 05:24:20: [174 MB] #4 Write peak in narrowPeak format file... peak_rev_1234@10-12-14/fwd_peaks.narrowPeak \n",
      "INFO  @ 14 Oct 2023 05:24:20: [174 MB] #4 Write summits bed file... peak_rev_1234@10-12-14/fwd_summits.bed \n",
      "INFO  @ 14 Oct 2023 05:24:20: [174 MB] Done! \n"
     ]
    }
   ],
   "source": [
    "! macs3 callpeak                                  \\\n",
    "    --treatment {samples_string(treated,in_path)} \\\n",
    "    --control   {samples_string(control,in_path)} \\\n",
    "    -f BAM                                        \\\n",
    "    -n fwd                                        \\\n",
    "    -g hs                                         \\\n",
    "    --keep-dup all                                \\\n",
    "    --outdir {out_path}                           \\\n",
    "    --nomodel --extsize 30"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "670aa93b-1bb1-4cbf-bba8-4ded097828e5",
   "metadata": {},
   "source": [
    "#### Mark 6th column with a **-** to indicate a `rev` peak (sed is 0 based)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 739,
   "id": "975f46fc-49c9-4736-a37f-81a4b2a43f41",
   "metadata": {},
   "outputs": [],
   "source": [
    "!sed -i 's/\\t[^\\t]*/\\t-/5' {out_path}/*.narrowPeak"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0abca181-9402-4f65-b444-29a4cb58020d",
   "metadata": {},
   "source": [
    "## Step: peak_all"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "749f3f0a-18d2-4af7-8f6b-055338f79964",
   "metadata": {},
   "source": [
    "#### Combine the forward and reverese reads and extend peak sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 740,
   "id": "9d8810c6-12e0-4921-ba58-b35df1476494",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_fwd_path = mkpath(f\"peak_fwd{tag}\")\n",
    "in_rev_path = mkpath(f\"peak_rev{tag}\")\n",
    "out_path    = mkpath(f\"peak_all{tag}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69b6d8e2-5302-4410-9dd6-3095b5a6cda0",
   "metadata": {},
   "source": [
    "##### Combine peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 741,
   "id": "d4779841-b8fe-43cd-8b34-7214d981e593",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_peaks    = out_path/\"all.narrowPeak\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 742,
   "id": "191101f4-8c41-4336-aa99-b79355288c33",
   "metadata": {},
   "outputs": [],
   "source": [
    "!cat {in_fwd_path}/*peaks.narrowPeak {in_rev_path}/*peaks.narrowPeak > {all_peaks}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 743,
   "id": "78879d4c-c704-4602-a962-a19b6ebd5f3a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1484 peak_all_1234@10-12-14/all.narrowPeak\n"
     ]
    }
   ],
   "source": [
    "!  wc -l {all_peaks} "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70804ec3-31f5-4fd4-8213-2d725f15c95c",
   "metadata": {},
   "source": [
    "##### Extend peaks by 20 nt in each direction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 744,
   "id": "f7de2f61-34a4-40c6-8682-8b6f6ef4adbe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629931</td>\n",
       "      <td>629961</td>\n",
       "      <td>fwd_peak_1</td>\n",
       "      <td>95</td>\n",
       "      <td>+</td>\n",
       "      <td>2.29353</td>\n",
       "      <td>14.6044</td>\n",
       "      <td>9.56511</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>630717</td>\n",
       "      <td>630752</td>\n",
       "      <td>fwd_peak_2</td>\n",
       "      <td>1066</td>\n",
       "      <td>+</td>\n",
       "      <td>2.45270</td>\n",
       "      <td>112.8720</td>\n",
       "      <td>106.65300</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      0       1       2           3     4  5        6         7          8   9\n",
       "0  chr1  629931  629961  fwd_peak_1    95  +  2.29353   14.6044    9.56511   4\n",
       "1  chr1  630717  630752  fwd_peak_2  1066  +  2.45270  112.8720  106.65300  10"
      ]
     },
     "execution_count": 744,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_csv(all_peaks, header = None, sep='\\t')\n",
    "df[:2]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebda2528-3984-483e-ac7a-3547117b4125",
   "metadata": {},
   "source": [
    "## Step: call motif"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41807c6d-6364-4dae-9f23-22e4cbc729ae",
   "metadata": {},
   "source": [
    "#### Find motifs from peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 745,
   "id": "5927b046-2a44-425c-9796-0ee9392f1fe4",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_path =  mkpath(f\"peak_all{tag}\")\n",
    "out_path = mkpath(f\"motif{tag}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c27f04a-1e17-4553-8eb6-5d4266a0c364",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\tPosition file = peak_all_1234@10-12-14/all.narrowPeak\n",
      "\tGenome = hg38\n",
      "\tOutput Directory = motif_1234@10-12-14\n",
      "\tUsing 32 CPUs\n",
      "\tOperating in RNA mode\n",
      "\tMotif length set at 10,12,14,*** might want to hit ctrl+C to quit if it takes too long!\n",
      "*** If you run out of memmory try reducing the number background sequences\n",
      "\n",
      "\tFound mset for \"human\", will check against vertebrates motifs\n",
      "\tPeak/BED file conversion summary:\n",
      "\t\tBED/Header formatted lines: 1484\n",
      "\t\tpeakfile formatted lines: 0\n",
      "\n",
      "\tPeak File Statistics:\n",
      "\t\tTotal Peaks: 1484\n",
      "\t\tRedundant Peak IDs: 687\n",
      "\t\tPeaks lacking information: 0 (need at least 5 columns per peak)\n",
      "\t\tPeaks with misformatted coordinates: 0 (should be integer)\n",
      "\t\tPeaks with misformatted strand: 0 (should be either +/- or 0/1)\n",
      "\n",
      "\tRedunant Peaks found: Remove or rename these or some programs may have trouble...\n",
      "\n",
      "\t687 duplicate peak IDs out of 1484 total peaks\n",
      "\tBackground files for 200 bp fragments found.\n",
      "\tCustom genome sequence directory: /home/cdaniels/bin/homer/.//data/genomes/hg38//\n",
      "\n",
      "\tExtracting sequences from file: /home/cdaniels/bin/homer/.//data/genomes/hg38///genome.fa\n",
      "\tLooking for peak sequences in a single file (/home/cdaniels/bin/homer/.//data/genomes/hg38///genome.fa)\n",
      "\tExtracting 111 sequences from chr1\n",
      "\tExtracting 43 sequences from chr10\n",
      "\tExtracting 77 sequences from chr11\n",
      "\tExtracting 100 sequences from chr12\n",
      "\tExtracting 31 sequences from chr13\n",
      "\tExtracting 52 sequences from chr14\n",
      "\tExtracting 35 sequences from chr15\n",
      "\tExtracting 52 sequences from chr16\n",
      "\tExtracting 105 sequences from chr17\n",
      "\tExtracting 18 sequences from chr18\n",
      "\tExtracting 59 sequences from chr19\n",
      "\tExtracting 218 sequences from chr2\n",
      "\tExtracting 52 sequences from chr20\n",
      "\tExtracting 8 sequences from chr21\n",
      "\tExtracting 26 sequences from chr22\n",
      "\tExtracting 68 sequences from chr3\n",
      "\tExtracting 68 sequences from chr4\n",
      "\tExtracting 70 sequences from chr5\n",
      "\tExtracting 78 sequences from chr6\n",
      "\tExtracting 56 sequences from chr7\n",
      "\tExtracting 31 sequences from chr8\n",
      "\tExtracting 44 sequences from chr9\n",
      "\tExtracting 47 sequences from chrM\n",
      "\tExtracting 34 sequences from chrX\n",
      "\tExtracting 1 sequences from chrY\n",
      "\t\t1 of 1484 removed because they had >70.00% Ns (i.e. masked repeats)\n",
      "\n",
      "\tNot removing redundant sequences\n",
      "\n",
      "\n",
      "\tSequences processed:\n",
      "\t\tAuto detected maximum sequence length of 201 bp\n",
      "\t\t1483 total\n",
      "\n",
      "\tFrequency Bins: 0.2 0.25 0.3 0.35 0.4 0.45 0.5 0.6 0.7 0.8\n",
      "\tFreq\tBin\tCount\n",
      "\t0.25\t1\t6\n",
      "\t0.3\t2\t59\n",
      "\t0.35\t3\t205\n",
      "\t0.4\t4\t336\n",
      "\t0.45\t5\t281\n",
      "\t0.5\t6\t230\n",
      "\t0.6\t7\t276\n",
      "\t0.7\t8\t76\n",
      "\t0.8\t9\t14\n",
      "\n",
      "\tTotal sequences set to 50000\n",
      "\n",
      "\tChoosing background that matches in CpG/GC content...\n",
      "\tBin\t# Targets\t# Background\tBackground Weight\n",
      "\t1\t6\t196\t1.001\n",
      "\t2\t59\t1930\t1.000\n",
      "\t3\t205\t6707\t1.000\n",
      "\t4\t336\t10992\t1.000\n",
      "\t5\t281\t9193\t1.000\n",
      "\t6\t230\t7525\t1.000\n",
      "\t7\t276\t9029\t1.000\n",
      "\t8\t76\t2486\t1.000\n",
      "\t9\t14\t458\t1.000\n",
      "\tAssembling sequence file...\n",
      "\tNormalizing lower order oligos using homer2\n",
      "\n",
      "\tReading input files...\n",
      "\t49999 total sequences read\n",
      "\tAutonormalization: 1-mers (4 total)\n",
      "\t\tA\t29.26%\t27.51%\t1.063\n",
      "\t\tC\t21.16%\t21.62%\t0.979\n",
      "\t\tG\t22.68%\t22.16%\t1.024\n",
      "\t\tT\t26.90%\t28.71%\t0.937\n",
      "\tAutonormalization: 2-mers (16 total)\n",
      "\t\tAA\t9.57%\t8.67%\t1.105\n",
      "\t\tCA\t7.05%\t7.06%\t0.998\n",
      "\t\tGA\t7.24%\t6.06%\t1.195\n",
      "\t\tTA\t5.40%\t5.76%\t0.937\n",
      "\t\tAC\t5.29%\t4.90%\t1.078\n",
      "\t\tCC\t5.61%\t5.85%\t0.959\n",
      "\t\tGC\t4.65%\t4.81%\t0.968\n",
      "\t\tTC\t5.61%\t6.04%\t0.928\n",
      "\t\tAG\t7.77%\t7.17%\t1.085\n",
      "\t\tCG\t2.02%\t1.46%\t1.382\n",
      "\t\tGG\t5.73%\t6.04%\t0.948\n",
      "\t\tTG\t7.15%\t7.51%\t0.953\n",
      "\t\tAT\t6.63%\t6.81%\t0.973\n",
      "\t\tCT\t6.48%\t7.23%\t0.896\n",
      "\t\tGT\t5.06%\t5.26%\t0.962\n",
      "\t\tTT\t8.73%\t9.36%\t0.932\n",
      "\tAutonormalization: 3-mers (64 total)\n",
      "\tNormalization weights can be found in file: motif_1234@10-12-14/seq.autonorm.tsv\n",
      "\tConverging on autonormalization solution:\n",
      "\t...............................................................................\n",
      "\tFinal normalization:\tAutonormalization: 1-mers (4 total)\n",
      "\t\tA\t29.26%\t29.13%\t1.004\n",
      "\t\tC\t21.16%\t21.18%\t0.999\n",
      "\t\tG\t22.68%\t22.71%\t0.999\n",
      "\t\tT\t26.90%\t26.98%\t0.997\n",
      "\tAutonormalization: 2-mers (16 total)\n",
      "\t\tAA\t9.57%\t9.57%\t1.000\n",
      "\t\tCA\t7.05%\t7.10%\t0.993\n",
      "\t\tGA\t7.24%\t6.96%\t1.041\n",
      "\t\tTA\t5.40%\t5.50%\t0.981\n",
      "\t\tAC\t5.29%\t5.18%\t1.020\n",
      "\t\tCC\t5.61%\t5.62%\t0.999\n",
      "\t\tGC\t4.65%\t4.74%\t0.981\n",
      "\t\tTC\t5.61%\t5.63%\t0.995\n",
      "\t\tAG\t7.77%\t7.73%\t1.006\n",
      "\t\tCG\t2.02%\t1.87%\t1.085\n",
      "\t\tGG\t5.73%\t5.94%\t0.963\n",
      "\t\tTG\t7.15%\t7.17%\t0.998\n",
      "\t\tAT\t6.63%\t6.66%\t0.996\n",
      "\t\tCT\t6.48%\t6.59%\t0.983\n",
      "\t\tGT\t5.06%\t5.06%\t1.001\n",
      "\t\tTT\t8.73%\t8.67%\t1.006\n",
      "\tAutonormalization: 3-mers (64 total)\n",
      "\tFinished preparing sequence/group files\n",
      "\n",
      "\t----------------------------------------------------------\n",
      "\tKnown motif enrichment\n",
      "\t----------------------------------------------------------\n",
      "\tDe novo motif finding (HOMER)\n",
      "\n",
      "\tNumber of Trial motifs (-T) set to 32 (from 10) to work well with 32 CPUs\n",
      "\tScanning input files...\n",
      "\tParsing sequences...\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\tTotal number of Oligos: 922025\n",
      "\tAutoadjustment for sequence coverage in background: 1.00x\n",
      "\n",
      "\tOligos: 922025 of 1048586 max\n",
      "\tTree  : 1346736 of 5242930 max\n",
      "\tOptimizing memory usage...\n",
      "\tCache length = 11180\n",
      "\tUsing binomial scoring\n",
      "\n",
      "\tGlobal Optimization Phase: Looking for enriched oligos with up to 2 mismatches...\n",
      "\n",
      "\tScreening oligos 922025 (allowing 0 mismatches):\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\t\t76.50% skipped, 23.50% checked (216715 of 922025), of those checked:\n",
      "\t\t76.50% not in target, 0.00% increased p-value, 0.00% high p-value\n",
      "\n",
      "\tScreening oligos 922025 (allowing 1 mismatches):\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\t\t76.50% skipped, 23.50% checked (216715 of 922025), of those checked:\n",
      "\t\t0.00% not in target, 17.00% increased p-value, 8.64% high p-value\n",
      "\n",
      "\tScreening oligos 922025 (allowing 2 mismatches):\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\t\t95.03% skipped, 4.97% checked (45857 of 922025), of those checked:\n",
      "\t\t0.00% not in target, 2.72% increased p-value, 0.00% high p-value\n",
      "\tReading input files...\n",
      "\t49999 total sequences read\n",
      "\tCache length = 11180\n",
      "\tUsing binomial scoring\n",
      "\n",
      "\tLocal Optimization Phase:\n",
      "\t1 of 25 Initial Sequence: GAAGATGAAG... (-220.128)\n",
      "\t\tRound 1: -310.45 GAAGAWGAAG T:686.0(37.04%),B:5683.1(11.98%),P:1e-134\n",
      "\t\tRound 2: -322.82 GAWGAWGAAG T:766.0(40.35%),B:6562.5(13.70%),P:1e-140\n",
      "\t\tRound 3: -343.89 GAAGAWGAAG T:729.0(38.84%),B:5780.4(12.17%),P:1e-149\n",
      "\t\tRound 4: -343.89 GAAGAWGAAG T:729.0(38.84%),B:5780.4(12.17%),P:1e-149\n",
      "\t\t=Final=: -152.21 GAAGAWGAAG T:400.0(26.97%),B:4822.6(10.83%),P:1e-66\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t2 of 25 Initial Sequence: AAGGTAAGTT... (-76.344)\n",
      "\t\tRound 1: -137.15 AAGGTAAGTT T:192.0(12.15%),B:1252.7(2.77%),P:1e-59\n",
      "\t\tRound 2: -154.81 AAGGTAAGTN T:281.0(17.27%),B:2239.7(4.90%),P:1e-67\n",
      "\t\tRound 3: -161.65 MWGGTRAGTN T:390.0(23.13%),B:3731.7(8.03%),P:1e-70\n",
      "\t\tRound 4: -161.65 MWGGTRAGTN T:390.0(23.13%),B:3731.7(8.03%),P:1e-70\n",
      "\t\t=Final=: -206.51 MWGGTRAGTN T:398.0(26.84%),B:3942.4(8.85%),P:1e-89\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t3 of 25 Initial Sequence: CACAGCACAC... (-43.802)\n",
      "\t\tRound 1: -50.28 CACAGCACAC T:21.0(1.41%),B:22.4(0.05%),P:1e-21\n",
      "\t\tRound 2: -50.28 CACAGCACAC T:21.0(1.41%),B:22.4(0.05%),P:1e-21\n",
      "\t\t=Final=: -5.33 CACAGCACAC T:4.0(0.27%),B:20.4(0.05%),P:1e-2\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t4 of 25 Initial Sequence: ACGACGAGGA... (-39.510)\n",
      "\t\tRound 1: -56.57 ACGACGAGGA T:519.0(29.54%),B:9079.6(18.44%),P:1e-24\n",
      "\t\tRound 2: -56.57 ACGACGAGGA T:519.0(29.54%),B:9079.6(18.44%),P:1e-24\n",
      "\t\t=Final=: -29.88 ACGACGAGGA T:530.0(35.74%),B:12026.2(27.00%),P:1e-12\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t5 of 25 Initial Sequence: CATATACTAA... (-32.386)\n",
      "\t\tRound 1: -37.87 CATATACTAA T:9.0(0.61%),B:2.5(0.00%),P:1e-16\n",
      "\t\tRound 2: -37.87 CATATACTAA T:9.0(0.61%),B:2.5(0.00%),P:1e-16\n",
      "\t\t=Final=: -37.87 CATATACTAA T:8.0(0.54%),B:1.4(0.00%),P:1e-16\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t6 of 25 Initial Sequence: TGGTAGAGCG... (-32.386)\n",
      "\t\tRound 1: -37.27 TGGTAGAGCG T:10.0(0.67%),B:3.5(0.01%),P:1e-16\n",
      "\t\tRound 2: -41.07 TGGAAGAGCG T:85.0(5.57%),B:802.4(1.78%),P:1e-17\n",
      "\t\tRound 3: -54.08 TBGAAGAGCK T:98.0(6.40%),B:826.4(1.84%),P:1e-23\n",
      "\t\tRound 4: -61.65 TNGAAGAGCB T:151.0(9.68%),B:1555.5(3.43%),P:1e-26\n",
      "\t\tRound 5: -61.65 TNGAAGAGCB T:151.0(9.68%),B:1555.5(3.43%),P:1e-26\n",
      "\t\t=Final=: -59.03 TNGAAGAGCB T:137.0(9.24%),B:1464.9(3.29%),P:1e-25\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t7 of 25 Initial Sequence: GGAACGATAC... (-27.033)\n",
      "\t\tRound 1: -43.48 GGAACGATAC T:10.0(0.67%),B:2.2(0.00%),P:1e-18\n",
      "\t\tRound 2: -48.66 GGWWCGATWC T:17.0(1.14%),B:11.5(0.02%),P:1e-21\n",
      "\t\tRound 3: -48.66 GGWWCGATWC T:17.0(1.14%),B:11.5(0.02%),P:1e-21\n",
      "\t\t=Final=: -51.02 GGWWCGATWC T:17.0(1.15%),B:11.5(0.03%),P:1e-22\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t8 of 25 Initial Sequence: CAACGAGATC... (-24.971)\n",
      "\t\tRound 1: -44.00 CAACGAGATC T:385.0(22.87%),B:6755.7(14.07%),P:1e-19\n",
      "\t\tRound 2: -53.62 CAACVAGATC T:358.0(21.45%),B:5762.8(12.13%),P:1e-23\n",
      "\t\tRound 3: -58.20 CAACVAGATC T:473.0(27.32%),B:8008.7(16.45%),P:1e-25\n",
      "\t\tRound 4: -59.99 CAHCVASAWC T:479.0(27.61%),B:8045.7(16.52%),P:1e-26\n",
      "\t\tRound 5: -66.99 CAHCAAGAWC T:697.0(37.51%),B:12385.4(24.27%),P:1e-29\n",
      "\t\tRound 6: -66.99 CAHCAAGAWC T:697.0(37.51%),B:12385.4(24.27%),P:1e-29\n",
      "\t\t=Final=: -50.90 CAHCAAGAWC T:514.0(34.66%),B:10421.0(23.39%),P:1e-22\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t9 of 25 Initial Sequence: ACTACTACTA... (-24.753)\n",
      "\t\tRound 1: -31.09 ACTACTACTA T:10.0(0.67%),B:5.9(0.01%),P:1e-13\n",
      "\t\tRound 2: -33.22 ACTACGACTA T:11.0(0.74%),B:6.7(0.01%),P:1e-14\n",
      "\t\tRound 3: -33.22 ACTACGACTA T:11.0(0.74%),B:6.7(0.01%),P:1e-14\n",
      "\t\t=Final=: -11.63 ACTACGACTA T:5.0(0.34%),B:8.5(0.02%),P:1e-5\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t10 of 25 Initial Sequence: GCTGCTGATG... (-20.383)\n",
      "\t\tRound 1: -27.51 GCTGCTGATG T:69.0(4.55%),B:753.8(1.68%),P:1e-11\n",
      "\t\tRound 2: -31.62 GCTACTGATG T:58.0(3.84%),B:504.9(1.13%),P:1e-13\n",
      "\t\tRound 3: -35.03 GMTRCTGMTG T:57.0(3.77%),B:449.5(1.00%),P:1e-15\n",
      "\t\tRound 4: -35.03 GMTRCTGMTG T:57.0(3.77%),B:449.5(1.00%),P:1e-15\n",
      "\t\t=Final=: -23.99 GMTRCTGMTG T:44.0(2.97%),B:409.4(0.92%),P:1e-10\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t11 of 25 Initial Sequence: AACAAAAAAA... (-19.740)\n",
      "\t\tRound 1: -20.22 AAMAAAAAAA T:562.0(31.55%),B:12613.4(24.66%),P:1e-8\n",
      "\t\tRound 2: -24.28 AAMAAAAAAA T:552.0(31.09%),B:11992.7(23.60%),P:1e-10\n",
      "\t\tRound 3: -32.53 AMMMAAAAAA T:91.0(5.95%),B:1050.8(2.33%),P:1e-14\n",
      "\t\tRound 4: -32.53 AMMMAAAAAA T:91.0(5.95%),B:1050.8(2.33%),P:1e-14\n",
      "\t\t=Final=: -2.94 AMMMAAAAAA T:2.0(0.13%),B:11.6(0.03%),P:1e-1\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t12 of 25 Initial Sequence: TCGAATCCCA... (-18.396)\n",
      "\t\tRound 1: -29.14 TCGAATCCCA T:9.0(0.61%),B:4.2(0.01%),P:1e-12\n",
      "\t\tRound 2: -29.14 TCGAATCCCA T:9.0(0.61%),B:4.2(0.01%),P:1e-12\n",
      "\t\t=Final=: -26.87 TCGAATCCCA T:8.0(0.54%),B:4.2(0.01%),P:1e-11\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t13 of 25 Initial Sequence: ACAAAAAAAA... (-17.084)\n",
      "\t\tRound 1: -28.16 AAAAAAAAAA T:483.0(27.81%),B:9969.4(20.05%),P:1e-12\n",
      "\t\tRound 2: -30.74 AMAAAAAAAA T:537.0(30.39%),B:11057.4(21.98%),P:1e-13\n",
      "\t\tRound 3: -31.34 AMAAAAMAAA T:529.0(30.01%),B:10839.4(21.60%),P:1e-13\n",
      "\t\tRound 4: -31.34 AMAAAAMAAA T:529.0(30.01%),B:10839.4(21.60%),P:1e-13\n",
      "\t\t=Final=: -0.01 AMAAAAMAAA T:692.0(46.66%),B:22259.8(49.97%),P:1e0\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t14 of 25 Initial Sequence: GGTTCGAATC... (-16.818)\n",
      "\t\tRound 1: -13.99 GGTTCGAATC T:18.0(1.21%),B:119.1(0.27%),P:1e-6\n",
      "\t\tRound 2: -18.80 GGTTCAAWTC T:12.0(0.81%),B:30.6(0.07%),P:1e-8\n",
      "\t\tRound 3: -19.72 GGTTCRATYC T:9.0(0.61%),B:11.9(0.02%),P:1e-8\n",
      "\t\tRound 4: -21.16 GGTTCGRTYC T:19.0(1.27%),B:82.2(0.18%),P:1e-9\n",
      "\t\tRound 5: -21.16 GGTTCGRTYC T:19.0(1.27%),B:82.2(0.18%),P:1e-9\n",
      "\t\t=Final=: -23.13 GGTTCGRTYC T:19.0(1.28%),B:81.6(0.18%),P:1e-10\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t15 of 25 Initial Sequence: AGTAAGTAAG... (-16.818)\n",
      "\t\tRound 1: -37.87 AGTAAGTAAG T:9.0(0.61%),B:2.9(0.00%),P:1e-16\n",
      "\t\tRound 2: -37.87 AGTAAGTAAG T:9.0(0.61%),B:2.9(0.00%),P:1e-16\n",
      "\t\t=Final=: -32.36 AGTAAGTAAG T:8.0(0.54%),B:2.9(0.01%),P:1e-14\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t16 of 25 Initial Sequence: GAAAGGTACT... (-16.818)\n",
      "\t\tRound 1: -16.82 GAAAGGTACT T:5.0(0.34%),B:1.2(0.00%),P:1e-7\n",
      "\t\tRound 2: -16.82 GAAAGGTACT T:5.0(0.34%),B:1.2(0.00%),P:1e-7\n",
      "\t\t=Final=: -21.83 GAAAGGTACT T:5.0(0.34%),B:1.2(0.00%),P:1e-9\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t17 of 25 Initial Sequence: TGGTTATCAC... (-16.818)\n",
      "\t\tRound 1: -21.83 TGGTTATCAC T:6.0(0.40%),B:0.9(0.00%),P:1e-9\n",
      "\t\tRound 2: -21.83 TGGTTATCAC T:6.0(0.40%),B:0.9(0.00%),P:1e-9\n",
      "\t\t=Final=: -27.03 TGGTTATCAC T:6.0(0.40%),B:0.9(0.00%),P:1e-11\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t18 of 25 Initial Sequence: CGAGACCCGC... (-16.818)\n",
      "\t\tRound 1: -18.40 CGAGACCCGC T:6.0(0.40%),B:3.7(0.01%),P:1e-7\n",
      "\t\tRound 2: -18.40 CGAGACCCGC T:6.0(0.40%),B:3.7(0.01%),P:1e-7\n",
      "\t\t=Final=: -16.40 CGAGACCCGC T:5.0(0.34%),B:3.7(0.01%),P:1e-7\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t19 of 25 Initial Sequence: GAAGATCATC... (-16.519)\n",
      "\t\tRound 1: -27.95 GAAGATCATC T:91.0(5.95%),B:1144.3(2.54%),P:1e-12\n",
      "\t\tRound 2: -27.95 GAAGATCATC T:91.0(5.95%),B:1144.3(2.54%),P:1e-12\n",
      "\t\t=Final=: -30.76 GAAGATCATC T:88.0(5.93%),B:1071.8(2.41%),P:1e-13\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t20 of 25 Initial Sequence: CTTCGGCAGC... (-16.396)\n",
      "\t\tRound 1: -21.85 CTTCGGCAGC T:15.0(1.01%),B:43.1(0.10%),P:1e-9\n",
      "\t\tRound 2: -29.14 CTTCGGYAGT T:14.0(0.94%),B:20.6(0.04%),P:1e-12\n",
      "\t\tRound 3: -29.14 CTTCGGYAGT T:14.0(0.94%),B:20.6(0.04%),P:1e-12\n",
      "\t\t=Final=: -31.56 CTTCGGYAGT T:14.0(0.94%),B:20.6(0.05%),P:1e-13\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t21 of 25 Initial Sequence: ATGTGAGTAC... (-14.985)\n",
      "\t\tRound 1: -17.69 ATGTGAGTAC T:67.0(4.42%),B:921.4(2.05%),P:1e-7\n",
      "\t\tRound 2: -24.53 ATGTGATTAS T:28.0(1.87%),B:162.1(0.36%),P:1e-10\n",
      "\t\tRound 3: -28.06 ATGTGAGTAG T:45.0(2.99%),B:362.1(0.81%),P:1e-12\n",
      "\t\tRound 4: -28.06 ATGTGAGTAG T:45.0(2.99%),B:362.1(0.81%),P:1e-12\n",
      "\t\t=Final=: -28.33 ATGTGAGTAG T:44.0(2.97%),B:357.9(0.80%),P:1e-12\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t22 of 25 Initial Sequence: TGAAGAAGAC... (-14.985)\n",
      "\t\tRound 1: -18.03 TGAAGAAGAC T:8.0(0.54%),B:9.3(0.02%),P:1e-7\n",
      "\t\tRound 2: -18.45 YGAAGAAGGC T:10.0(0.67%),B:18.5(0.04%),P:1e-8\n",
      "\t\tRound 3: -19.16 CGTAGAAGAC T:12.0(0.81%),B:29.6(0.07%),P:1e-8\n",
      "\t\tRound 4: -19.16 CGTAGAAGAC T:12.0(0.81%),B:29.6(0.07%),P:1e-8\n",
      "\t\t=Final=: -21.83 CGTAGAAGAC T:5.0(0.34%),B:0.9(0.00%),P:1e-9\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t23 of 25 Initial Sequence: TTGCCAATAC... (-14.149)\n",
      "\t\tRound 1: -22.90 TTGCCAATAC T:24.0(1.61%),B:126.6(0.28%),P:1e-9\n",
      "\t\tRound 2: -22.90 TTGCCAATAC T:24.0(1.61%),B:126.6(0.28%),P:1e-9\n",
      "\t\t=Final=: -24.51 TTGCCAATAC T:24.0(1.62%),B:126.6(0.28%),P:1e-10\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t24 of 25 Initial Sequence: TTTTTTTAGG... (-13.436)\n",
      "\t\tRound 1: -24.64 TTTTTTTAGG T:146.0(9.38%),B:2356.3(5.15%),P:1e-10\n",
      "\t\tRound 2: -24.64 TTTTTTTAGG T:146.0(9.38%),B:2356.3(5.15%),P:1e-10\n",
      "\t\t=Final=: -24.51 TTTTTTTAGG T:140.0(9.44%),B:2321.3(5.21%),P:1e-10\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t25 of 25 Initial Sequence: CGGAAGACCA... (-12.929)\n",
      "\t\tRound 1: -22.70 CGGAAGACCA T:63.0(4.16%),B:730.7(1.63%),P:1e-9\n",
      "\t\tRound 2: -22.97 CGGAAGACCA T:50.0(3.32%),B:509.8(1.14%),P:1e-9\n",
      "\t\tRound 3: -23.04 CGGAAGAMCA T:54.0(3.58%),B:578.2(1.29%),P:1e-10\n",
      "\t\tRound 4: -23.04 CGGAAGAMCA T:54.0(3.58%),B:578.2(1.29%),P:1e-10\n",
      "\t\t=Final=: -22.05 CGGAAGAMCA T:52.0(3.51%),B:573.2(1.29%),P:1e-9\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\n",
      "\tFinalizing Enrichment Statistics (new in v3.4)\n",
      "\tReading input files...\n",
      "\t49999 total sequences read\n",
      "\tCache length = 11180\n",
      "\tUsing binomial scoring\n",
      "\tChecking enrichment of 25 motif(s)\n",
      "\t|0%                                    50%                                  100%|\n",
      "\t=================================================================================\n",
      "\tOutput in file: motif_1234@10-12-14/homerMotifs.motifs10\n",
      "\n",
      "\n",
      "\tNumber of Trial motifs (-T) set to 32 (from 10) to work well with 32 CPUs\n",
      "\t-blen automatically set to 2\n",
      "\tScanning input files...\n",
      "\tParsing sequences...\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\tTotal number of Oligos: 5163522\n",
      "\tAutoadjustment for sequence coverage in background: 1.00x\n",
      "\n",
      "\tOligos: 5163522 of 9499604 max\n",
      "\tTree  : 10679144 of 47498020 max\n",
      "\tOptimizing memory usage...\n",
      "\tCache length = 11180\n",
      "\tUsing binomial scoring\n",
      "\n",
      "\tGlobal Optimization Phase: Looking for enriched oligos with up to 2 mismatches...\n",
      "\n",
      "\tScreening oligos 5163522 (allowing 0 mismatches):\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\t\t94.79% skipped, 5.21% checked (269016 of 5163522), of those checked:\n",
      "\t\t94.79% not in target, 0.00% increased p-value, 0.00% high p-value\n",
      "\n",
      "\tScreening oligos 5163522 (allowing 1 mismatches):\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\t\t94.79% skipped, 5.21% checked (269016 of 5163522), of those checked:\n",
      "\t\t0.00% not in target, 4.88% increased p-value, 0.07% high p-value\n",
      "\n",
      "\tScreening oligos 5163522 (allowing 2 mismatches):\n",
      "\t|0%                                   50%                                  100%|\n",
      "\t================================================================================\n",
      "\t\t99.68% skipped, 0.32% checked (16558 of 5163522), of those checked:\n",
      "\t\t0.00% not in target, 0.20% increased p-value, 0.00% high p-value\n",
      "\tReading input files...\n",
      "\t49999 total sequences read\n",
      "\tCache length = 11180\n",
      "\tUsing binomial scoring\n",
      "\n",
      "\tLocal Optimization Phase:\n",
      "\t1 of 25 Initial Sequence: AGAAGAAGAAGA... (-193.860)\n",
      "\t\tRound 1: -279.22 AGAAGAAGAAGA T:716.0(38.31%),B:6585.4(13.74%),P:1e-121\n",
      "\t\tRound 2: -345.76 WGAWGADGAWGA T:622.0(34.27%),B:4423.0(9.45%),P:1e-150\n",
      "\t\tRound 3: -345.76 WGAWGADGAWGA T:622.0(34.27%),B:4423.0(9.45%),P:1e-150\n",
      "\t\t=Final=: -142.88 WGAWGADGAWGA T:308.0(20.77%),B:3222.5(7.23%),P:1e-62\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t2 of 25 Initial Sequence: GGGTTCGATTCC... (-55.020)\n",
      "\t\tRound 1: -79.04 GGGTTCGATTCC T:16.0(1.07%),B:1.3(0.00%),P:1e-34\n",
      "\t\tRound 2: -80.89 GGGTTCGATCCC T:23.0(1.54%),B:8.2(0.02%),P:1e-35\n",
      "\t\tRound 3: -80.89 GGGTTCGATCCC T:23.0(1.54%),B:8.2(0.02%),P:1e-35\n",
      "\t\t=Final=: -82.46 GGGTTCGATCCC T:23.0(1.55%),B:8.2(0.02%),P:1e-35\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t3 of 25 Initial Sequence: CGCACAGCACAC... (-44.443)\n",
      "\t\tRound 1: -64.84 CACACAGCACAC T:20.0(1.34%),B:9.3(0.02%),P:1e-28\n",
      "\t\tRound 2: -68.93 MGCACRGCACAC T:19.0(1.27%),B:6.4(0.01%),P:1e-29\n",
      "\t\tRound 3: -68.93 MGCACRGCACAC T:19.0(1.27%),B:6.4(0.01%),P:1e-29\n",
      "\t\t=Final=: -7.52 MGCACRGCACAC T:2.0(0.13%),B:1.1(0.00%),P:1e-3\n",
      "\t\tPerforming exhaustive masking of motif...\n",
      "\t\tReprioritizing potential motifs...\n",
      "\t4 of 25 Initial Sequence: GAAGGTAAGTTT... (-44.261)\n",
      "\t\tRound 1: -94.17 GAAGGTAAGTTT T:195.0(12.33%),B:1759.1(3.87%),P:1e-40\n",
      "\t\tRound 2: -96.71 VAAGGTAAGTTT T:247.0(15.35%),B:2541.1(5.54%),P:1e-42\n"
     ]
    }
   ],
   "source": [
    "# -S 10 is changed to -S {nc} = 32\n",
    "# -len 5,6,7,8,9\n",
    "# -bg {genome_bed12}\n",
    "!findMotifsGenome.pl {in_path}/all.narrowPeak hg38 {out_path} \\\n",
    "                     -p {nc}                                  \\\n",
    "                     -rna                                     \\\n",
    "                     -len 10,12,14                          "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ea5d205-5c98-4e5d-b148-3c2bc185d048",
   "metadata": {},
   "source": [
    "#### Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ef107b1-8e31-4483-af95-77e563bcc172",
   "metadata": {},
   "outputs": [],
   "source": [
    "!ls -lh {out_path}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f60a3b69-8e84-45a3-854d-9f4cec558b6e",
   "metadata": {},
   "source": [
    "**Warning:** Motif Results are not fully rendered in Jupyter. A browser is better."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 727,
   "id": "8cdeea89-a60c-41d6-9a9c-c1842ce8a851",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<HTML><HEAD><TITLE>motif_23_12/ - Homer de novo Motif Results</TITLE></HEAD>\n",
       "<BODY>\n",
       "<H1>Homer <i>de novo</i> Motif Results (motif_23_12/)</H1>\n",
       "<A HREF=\"knownResults.html\">Known Motif Enrichment Results</A><BR/>\n",
       "<A HREF=\"geneOntology.html\">Gene Ontology Enrichment Results</A><BR/>\n",
       "If Homer is having trouble matching a motif to a known motif, try copy/pasting the matrix file into \n",
       "<A HREF=\"http://www.benoslab.pitt.edu/stamp/\">STAMP</A><BR/>\n",
       "More information on motif finding results: <A HREF=\"http://biowhat.ucsd.edu/homer/\">HOMER</A>\n",
       " | <A HREF=\"http://biowhat.ucsd.edu/homer/motif/index.html\">Description of Results</A>\n",
       " | <A HREF=\"http://biowhat.ucsd.edu/homer/motif/practicalTips.html\">Tips</A>\n",
       "<BR/>\n",
       "Total target sequences = 2597<BR/>\n",
       "Total background sequences = 44805<BR/>\n",
       "<FONT color=\"red\">* - possible false positive</FONT><BR/>\n",
       "<TABLE border=\"1\" cellpading=\"0\" cellspacing=\"0\">\n",
       "<TR><TD>Rank</TD><TD>Motif</TD><TD>P-value</TD><TD>log P-pvalue</TD><TD>% of Targets</TD><TD>% of Background</TD>\n",
       "<TD>STD(Bg STD)</TD>\n",
       "<TD>Best Match/Details</TD><TD>Motif File</TD></TR>\n",
       "<TR><TD>1\n",
       "</TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.053,0,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.128,0,46.35)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.168,-1.5,39.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.651,2,31.55)\">A</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.040,23.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.066,25,47)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.127,27,43.7)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.767,25,37.35)\">G</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.019,48.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.022,52,48.05)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.038,50,46.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.921,50,45.05)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.018,75,49)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.030,77,48.1)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.031,75,46.6)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.921,73.5,45.05)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.020,100,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.056,98.5,48)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.267,100,45.2)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.657,102,31.85)\">A</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.056,125,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.057,125,46.2)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.121,123.5,43.35)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.766,127,37.3)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.062,150,49)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.098,152,45.9)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.103,148.5,41)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.737,150,35.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.120,175,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.141,175,43)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.151,177,35.95)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.588,173.5,28.4)\">U</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-75</TD><TD>-1.743e+02</TD><TD>33.08%</TD><TD>17.95%</TD><TD>44.4bp (64.4bp)</TD><TD>hsa-miR-3124-3p MIMAT0019200 Homo sapiens miR-3124-3p Targets (miRBase)(0.648)<BR/><A target=\"_blank\" HREF=\"homerResults/motif1.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif1.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif1.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>2\n",
       "</TD><TD><svg width=\"180\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,-1.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.061,2,48.95)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.074,0,45.9)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.864,0,42.2)\">G</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,23.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.057,25,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.101,25,46.1)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.841,27,41.05)\">A</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,50,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.007,50,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.045,48.5,48.6)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.947,52,46.35)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,73.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.017,75,48.9)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.981,75,48.05)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,100,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.020,100,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.021,98.5,47.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.958,102,46.9)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.009,125,49)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.015,127,48.55)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.330,125,47.8)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.646,123.5,31.3)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,152,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.035,150,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.112,148.5,47.2)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.852,150,41.6)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-68</TD><TD>-1.581e+02</TD><TD>32.65%</TD><TD>18.24%</TD><TD>46.9bp (69.3bp)</TD><TD>hsa-miR-22* MIMAT0004495 Homo sapiens miR-22* Targets (miRBase)(0.718)<BR/><A target=\"_blank\" HREF=\"homerResults/motif2.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif2.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif2.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>3\n",
       "</TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.083,-1.5,49)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.129,0,44.85)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.195,2,38.4)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.593,0,28.65)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,25,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.072,23.5,48.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.283,27,45.35)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.644,25,31.2)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,50,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.095,50,48.9)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.903,52,44.15)\">A</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.047,73.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.073,75,46.65)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.137,77,43)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.743,75,36.15)\">G</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.088,98.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.170,102,44.6)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.317,100,36.1)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.425,100,20.25)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,125,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,123.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.032,125,48.9)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.966,127,47.3)\">A</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,148.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.047,152,48.95)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.363,150,46.6)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.589,150,28.45)\">G</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,175,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,173.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.042,175,48.9)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.956,177,46.8)\">A</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,200,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.214,202,48.95)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.388,198.5,38.25)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.397,200,18.85)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-32</TD><TD>-7.554e+01</TD><TD>18.17%</TD><TD>10.32%</TD><TD>45.5bp (69.2bp)</TD><TD>hsa-miR-4279 MIMAT0016909 Homo sapiens miR-4279 Targets (miRBase)(0.730)<BR/><A target=\"_blank\" HREF=\"homerResults/motif3.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif3.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif3.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>4\n",
       "</TD><TD><svg width=\"180\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,0,49)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.039,2,48.95)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.270,0,47)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.690,-1.5,33.5)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,23.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.234,25,48.9)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.764,25,37.2)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,50,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,50,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.322,52,48.9)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.676,48.5,32.8)\">U</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.201,73.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.233,75,38.9)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.565,77,27.25)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,102,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,100,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,98.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,100,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,127,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,125,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,123.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,125,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,150,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,148.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.235,150,48.9)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.763,152,37.15)\">A</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-27</TD><TD>-6.303e+01</TD><TD>27.92%</TD><TD>19.05%</TD><TD>51.7bp (69.7bp)</TD><TD>hsa-miR-487b MIMAT0003180 Homo sapiens miR-487b Targets (miRBase)(0.742)<BR/><A target=\"_blank\" HREF=\"homerResults/motif4.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif4.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif4.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>5\n",
       "</TD><TD><svg width=\"155\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,2,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,-1.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,0,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,25,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,23.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,25,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.151,52,48.95)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.313,50,41.4)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.535,50,25.75)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,75,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,73.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.997,77,48.85)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,100,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,98.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.031,100,48.9)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.967,102,47.35)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,125,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.204,123.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.278,125,38.75)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.517,127,24.85)\">A</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-21</TD><TD>-4.921e+01</TD><TD>19.64%</TD><TD>12.91%</TD><TD>51.1bp (70.4bp)</TD><TD>hsa-miR-1180 MIMAT0005825 Homo sapiens miR-1180 Targets (miRBase)(0.703)<BR/><A target=\"_blank\" HREF=\"homerResults/motif5.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif5.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif5.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>6\n",
       "</TD><TD><svg width=\"180\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.012,-1.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.020,0,48.4)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.043,0,47.4)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.925,2,45.25)\">A</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,23.5,49)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.052,25,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.062,27,46.35)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.885,25,43.25)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,52,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.086,50,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.156,48.5,44.65)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.757,50,36.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,75,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,73.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.026,75,48.9)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.972,77,47.6)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.036,102,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.086,98.5,47.2)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.222,100,42.9)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.656,100,31.8)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,125,49)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.023,127,48.95)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.053,123.5,47.8)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.923,125,45.15)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,150,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,150,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.022,148.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.976,152,47.8)\">A</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-20</TD><TD>-4.805e+01</TD><TD>23.22%</TD><TD>16.02%</TD><TD>52.4bp (71.8bp)</TD><TD>hsa-miR-1238 MIMAT0005593 Homo sapiens miR-1238 Targets (miRBase)(0.694)<BR/><A target=\"_blank\" HREF=\"homerResults/motif6.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif6.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif6.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>7\n",
       "</TD><TD><svg width=\"130\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,0,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.045,-1.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.174,0,46.7)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.780,2,38)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.150,25,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.296,25,41.45)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.553,23.5,26.65)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,52,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,50,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.123,48.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.875,50,42.75)\">C</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,73.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.043,77,48.95)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.064,75,46.8)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.892,75,43.6)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,100,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.024,98.5,48.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.047,102,47.75)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.928,100,45.4)\">G</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-19</TD><TD>-4.502e+01</TD><TD>22.64%</TD><TD>15.74%</TD><TD>54.6bp (68.2bp)</TD><TD>hsa-miR-29c* MIMAT0004673 Homo sapiens miR-29c* Targets (miRBase)(0.696)<BR/><A target=\"_blank\" HREF=\"homerResults/motif7.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif7.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif7.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>8\n",
       "</TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.047,2,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.056,0,46.65)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.076,-1.5,43.85)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.821,0,40.05)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.071,25,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.097,25,45.45)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.132,23.5,40.6)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.700,27,34)\">A</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.058,48.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.063,52,46.1)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.079,50,42.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.800,50,39)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.059,77,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.059,75,46.05)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.082,73.5,43.1)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.800,75,39)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.053,100,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.074,100,46.35)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.158,102,42.65)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.715,98.5,34.75)\">U</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.055,123.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.058,125,46.25)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.075,127,43.35)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.812,125,39.6)\">G</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.047,148.5,49)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.070,150,46.65)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.104,152,43.15)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.779,150,37.95)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.045,175,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.080,175,46.75)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.142,173.5,42.75)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.733,177,35.65)\">A</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.048,198.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.051,200,46.6)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.060,202,44.05)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.841,200,41.05)\">G</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-17</TD><TD>-4.116e+01</TD><TD>24.34%</TD><TD>17.52%</TD><TD>53.8bp (64.7bp)</TD><TD>hsa-miR-5047 MIMAT0020541 Homo sapiens miR-5047 Targets (miRBase)(0.772)<BR/><A target=\"_blank\" HREF=\"homerResults/motif8.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif8.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif8.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>9\n",
       "</TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.051,0,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.064,2,46.45)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.066,-1.5,43.25)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.819,0,39.95)\">C</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.050,25,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.078,23.5,46.5)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.082,25,42.6)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.790,27,38.5)\">A</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.069,50,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.083,52,45.55)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.125,50,41.4)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.723,48.5,35.15)\">U</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.039,75,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.058,75,47.05)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.105,77,44.15)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.798,73.5,38.9)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.040,100,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.040,98.5,47)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.061,102,45)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.859,100,41.95)\">G</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.032,125,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.076,123.5,47.4)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.129,125,43.6)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.763,127,37.15)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.051,152,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.071,150,46.45)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.118,150,42.9)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.760,148.5,37)\">U</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.050,173.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.092,177,46.5)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.107,175,41.9)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.751,175,36.55)\">C</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.058,200,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.098,200,46.1)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.101,198.5,41.2)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.743,202,36.15)\">A</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-14</TD><TD>-3.344e+01</TD><TD>14.17%</TD><TD>9.39%</TD><TD>58.2bp (61.9bp)</TD><TD>hsa-miR-3908 MIMAT0018182 Homo sapiens miR-3908 Targets (miRBase)(0.681)<BR/><A target=\"_blank\" HREF=\"homerResults/motif9.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif9.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif9.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>10\n",
       "</TD><TD><svg width=\"180\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,2,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,-1.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,0,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,25,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,25,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,23.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.997,27,48.85)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,50,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.026,50,48.9)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.972,52,47.6)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,73.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,75,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,100,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,98.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.165,100,48.9)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.833,102,40.65)\">A</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,123.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.002,125,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.195,125,48.85)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.802,127,39.1)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,152,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,150,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,148.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,150,48.85)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-14</TD><TD>-3.252e+01</TD><TD>17.71%</TD><TD>12.45%</TD><TD>53.9bp (66.8bp)</TD><TD>hsa-miR-1273g MIMAT0020602 Homo sapiens miR-1273g Targets (miRBase)(0.706)<BR/><A target=\"_blank\" HREF=\"homerResults/motif10.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif10.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif10.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>11\n",
       "</TD><TD><svg width=\"180\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.128,0,48.95)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.213,2,42.55)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.658,-1.5,31.9)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.120,25,49)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.150,27,43)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.295,25,35.5)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.435,23.5,20.75)\">U</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,50,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.030,52,48.95)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,48.5,47.45)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.869,50,42.45)\">C</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.036,73.5,49)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.090,75,47.2)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.208,75,42.7)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.666,77,32.3)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,100,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,98.5,48.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.042,102,48.9)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.956,100,46.8)\">G</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.084,125,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.088,125,44.8)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.127,123.5,40.4)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.701,127,34.05)\">A</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.021,150,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.095,150,47.95)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.147,152,43.2)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.737,148.5,35.85)\">U</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-13</TD><TD>-3.038e+01</TD><TD>45.63%</TD><TD>38.47%</TD><TD>55.1bp (65.0bp)</TD><TD>hsa-miR-452* MIMAT0001636 Homo sapiens miR-452* Targets (miRBase)(0.705)<BR/><A target=\"_blank\" HREF=\"homerResults/motif11.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif11.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif11.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>12\n",
       "</TD><TD><svg width=\"130\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,0,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.068,2,48.9)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.930,-1.5,45.5)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,23.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.010,25,48.9)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.988,25,48.4)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.002,52,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.009,50,48.9)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.059,48.5,48.45)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.930,50,45.5)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,75,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,73.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.997,77,48.85)\">A</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,100,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.032,98.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.436,100,47.35)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.531,102,25.55)\">A</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-12</TD><TD>-2.828e+01</TD><TD>46.13%</TD><TD>39.23%</TD><TD>55.3bp (66.9bp)</TD><TD>hsa-miR-587 MIMAT0003253 Homo sapiens miR-587 Targets (miRBase)(0.693)<BR/><A target=\"_blank\" HREF=\"homerResults/motif12.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif12.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif12.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>13\n",
       "</TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.037,-1.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.094,0,47.15)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.112,0,42.45)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.757,2,36.85)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.035,25,49)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.087,27,47.25)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.108,25,42.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.770,23.5,37.5)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.048,52,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.073,48.5,46.6)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.087,50,42.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.792,50,38.6)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.033,77,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.044,75,47.35)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.086,73.5,45.15)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.837,75,40.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.030,98.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.066,102,47.5)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.128,100,44.2)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.776,100,37.8)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.029,127,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.050,123.5,47.55)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.140,125,45.05)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.781,125,38.05)\">C</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.043,148.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.077,150,46.85)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.109,152,43)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.771,150,37.55)\">G</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.066,173.5,49)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,175,45.7)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.110,177,40.7)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.724,175,35.2)\">C</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.041,200,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.064,202,46.95)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.097,198.5,43.75)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.798,200,38.9)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-12</TD><TD>-2.821e+01</TD><TD>7.97%</TD><TD>4.72%</TD><TD>55.0bp (61.3bp)</TD><TD>hsa-miR-4467 MIMAT0018994 Homo sapiens miR-4467 Targets (miRBase)(0.642)<BR/><A target=\"_blank\" HREF=\"homerResults/motif13.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif13.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif13.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>14\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,-1.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.046,2,48.95)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.061,0,46.65)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.892,0,43.6)\">C</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,25,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.085,23.5,48.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.259,27,44.7)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.655,25,31.75)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,52,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.141,50,48.9)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.857,50,41.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.032,75,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.065,73.5,47.35)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.902,77,44.1)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,102,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.020,98.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.065,100,47.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.914,100,44.7)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,127,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,125,48.95)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.088,125,48.9)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.910,123.5,44.5)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,152,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,148.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.005,150,48.9)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.993,150,48.65)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,177,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,173.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.089,175,48.9)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.909,175,44.45)\">G</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-11</TD><TD>-2.691e+01</TD><TD>6.82%</TD><TD>3.90%</TD><TD>57.7bp (61.4bp)</TD><TD>hsa-miR-4252 MIMAT0016886 Homo sapiens miR-4252 Targets (miRBase)(0.771)<BR/><A target=\"_blank\" HREF=\"homerResults/motif14.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif14.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif14.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>15\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.072,0,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.392,-1.5,45.35)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.535,2,25.75)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,23.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.302,25,48.9)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.696,25,33.8)\">C</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,50,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.230,50,48.9)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.768,52,37.4)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,75,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.436,73.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.562,77,27.1)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,102,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.163,100,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.294,98.5,40.8)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.542,100,26.1)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,127,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,123.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.365,125,48.9)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.633,125,30.65)\">C</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.074,148.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.087,152,45.3)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.251,150,40.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.588,150,28.4)\">G</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,173.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.007,177,48.95)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.380,175,48.6)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.612,175,29.6)\">G</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-11</TD><TD>-2.537e+01</TD><TD>9.16%</TD><TD>5.83%</TD><TD>52.7bp (64.0bp)</TD><TD>hsa-miR-671-3p MIMAT0004819 Homo sapiens miR-671-3p Targets (miRBase)(0.628)<BR/><A target=\"_blank\" HREF=\"homerResults/motif15.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif15.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif15.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>16\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.059,0,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.082,2,46.05)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.133,-1.5,41.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.726,0,35.3)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.078,25,49)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.121,27,45.1)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.122,23.5,39.05)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.679,25,32.95)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.051,50,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.064,50,46.45)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.143,52,43.25)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.742,48.5,36.1)\">U</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.046,75,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.056,77,46.7)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.070,73.5,43.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.828,75,40.4)\">C</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.040,98.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.057,100,47)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.061,100,44.15)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.842,102,41.1)\">A</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.066,125,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.075,125,45.7)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.147,123.5,41.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.712,127,34.6)\">A</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.066,150,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.092,152,45.7)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.098,148.5,41.1)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.744,150,36.2)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.049,175,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.066,173.5,46.55)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.149,177,43.25)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.736,175,35.8)\">G</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-10</TD><TD>-2.464e+01</TD><TD>18.14%</TD><TD>13.50%</TD><TD>51.8bp (67.8bp)</TD><TD>hsa-miR-1306 MIMAT0005950 Homo sapiens miR-1306 Targets (miRBase)(0.655)<BR/><A target=\"_blank\" HREF=\"homerResults/motif16.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif16.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif16.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>17\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.016,2,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.099,0,48.2)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.107,0,43.25)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.778,-1.5,37.9)\">U</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.096,23.5,49)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.103,25,44.2)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.219,25,39.05)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.582,27,28.1)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.047,52,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.048,50,46.65)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.053,50,44.25)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.852,48.5,41.6)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.038,77,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.111,73.5,47.1)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.114,75,41.55)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.737,75,35.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.009,100,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.043,100,48.55)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.129,102,46.4)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.819,98.5,39.95)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.047,127,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.079,125,46.65)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.111,125,42.7)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.763,123.5,37.15)\">U</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.034,148.5,49)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.037,150,47.3)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.118,152,45.45)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.811,150,39.55)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.107,175,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.110,173.5,43.65)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.115,175,38.15)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.668,177,32.4)\">A</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-10</TD><TD>-2.463e+01</TD><TD>38.93%</TD><TD>32.74%</TD><TD>59.1bp (59.0bp)</TD><TD>hsa-miR-3682-3p MIMAT0018110 Homo sapiens miR-3682-3p Targets (miRBase)(0.698)<BR/><A target=\"_blank\" HREF=\"homerResults/motif17.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif17.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif17.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>18\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.079,2,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.083,0,45.05)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.083,-1.5,40.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.755,0,36.75)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.036,27,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.054,25,47.2)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.074,23.5,44.5)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.836,25,40.8)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.061,52,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.063,50,45.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.110,48.5,42.8)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.766,50,37.3)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.041,77,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.049,75,46.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.060,75,44.5)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.850,73.5,41.5)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.077,102,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.080,98.5,45.15)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.096,100,41.15)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.747,100,36.35)\">G</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.077,123.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.078,127,45.15)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.079,125,41.25)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.766,125,37.3)\">C</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.067,150,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.121,150,45.65)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.128,148.5,39.6)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.684,152,33.2)\">A</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.020,173.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.079,177,48)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.127,175,44.05)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.774,175,37.7)\">G</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.043,198.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.049,202,46.85)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.080,200,44.4)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.828,200,40.4)\">G</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-10</TD><TD>-2.378e+01</TD><TD>9.93%</TD><TD>6.56%</TD><TD>55.5bp (62.2bp)</TD><TD>hsa-miR-1205 MIMAT0005869 Homo sapiens miR-1205 Targets (miRBase)(0.789)<BR/><A target=\"_blank\" HREF=\"homerResults/motif18.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif18.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif18.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>19\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"155\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,2,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,0,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.997,-1.5,48.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,23.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.121,25,48.9)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.877,25,42.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,52,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,50,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.228,48.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.770,50,37.5)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,75,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.997,73.5,48.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,102,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,100,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,100,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.997,98.5,48.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,127,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,125,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,125,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.997,123.5,48.85)\">U</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-10</TD><TD>-2.338e+01</TD><TD>40.16%</TD><TD>34.10%</TD><TD>61.9bp (58.5bp)</TD><TD>hsa-miR-4422 MIMAT0018935 Homo sapiens miR-4422 Targets (miRBase)(0.794)<BR/><A target=\"_blank\" HREF=\"homerResults/motif19.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif19.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif19.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>20\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"130\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.234,-1.5,48.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.292,2,37.25)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.473,0,22.65)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,25,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.006,25,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.992,23.5,48.6)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.002,50,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.005,48.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.045,50,48.65)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.948,52,46.4)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.425,73.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.573,75,27.65)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,102,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.045,98.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.433,100,46.7)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.521,100,25.05)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-9</TD><TD>-2.273e+01</TD><TD>45.82%</TD><TD>39.70%</TD><TD>55.5bp (63.0bp)</TD><TD>hsa-miR-4500 MIMAT0019036 Homo sapiens miR-4500 Targets (miRBase)(0.633)<BR/><A target=\"_blank\" HREF=\"homerResults/motif20.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif20.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif20.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>21\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,-1.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.024,2,48.95)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.055,0,47.75)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.920,0,45)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,25,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.084,23.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.110,25,44.75)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.805,27,39.25)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,52,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,50,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,50,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,73.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.021,77,48.95)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.043,75,47.9)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.935,75,45.75)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,102,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.053,100,48.95)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.188,100,46.3)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.758,98.5,36.9)\">U</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,125,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,123.5,48.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.026,127,48.9)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.972,125,47.6)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,150,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,150,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.092,148.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.906,152,44.3)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,175,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,175,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.114,173.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.884,177,43.2)\">A</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,200,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.012,198.5,48.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.063,202,48.35)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.924,200,45.2)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-9</TD><TD>-2.222e+01</TD><TD>2.46%</TD><TD>1.01%</TD><TD>49.9bp (62.1bp)</TD><TD>hsa-miR-1587 MIMAT0019077 Homo sapiens miR-1587 Targets (miRBase)(0.711)<BR/><A target=\"_blank\" HREF=\"homerResults/motif21.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif21.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif21.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>22\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,2,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.258,0,48.95)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.369,0,36.05)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.372,-1.5,17.6)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.127,25,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.214,25,42.65)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.249,23.5,31.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.410,27,19.5)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.018,50,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.120,48.5,48.1)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.192,52,42.1)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.670,50,32.5)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.038,77,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.220,73.5,47.1)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.368,75,36.1)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.374,75,17.7)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.015,100,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.086,100,48.25)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.417,102,43.95)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.482,98.5,23.1)\">U</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.019,123.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.230,127,48.05)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.264,125,36.55)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.487,125,23.35)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.020,152,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.048,148.5,48)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.291,150,45.6)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.641,150,31.05)\">C</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.029,175,49)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.104,175,47.55)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.215,177,42.35)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.652,173.5,31.6)\">U</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-9</TD><TD>-2.178e+01</TD><TD>11.78%</TD><TD>8.25%</TD><TD>55.6bp (62.5bp)</TD><TD>hsa-miR-646 MIMAT0003316 Homo sapiens miR-646 Targets (miRBase)(0.704)<BR/><A target=\"_blank\" HREF=\"homerResults/motif22.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif22.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif22.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>23\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,0,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.013,-1.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.985,2,48.25)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,23.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.168,25,48.9)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.830,25,40.5)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,52,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,50,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,50,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.070,75,48.95)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.129,75,45.45)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.800,73.5,39)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.090,100,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.226,98.5,44.5)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.234,100,33.2)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.450,102,21.5)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,127,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,123.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.163,125,48.9)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.835,125,40.75)\">C</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.002,150,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.011,152,48.9)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.020,150,48.35)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.967,148.5,47.35)\">U</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.094,175,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.188,177,44.3)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.263,173.5,34.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.456,175,21.75)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,202,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.010,200,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.090,198.5,48.45)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.899,200,43.95)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-9</TD><TD>-2.076e+01</TD><TD>0.77%</TD><TD>0.13%</TD><TD>51.6bp (50.4bp)</TD><TD>hsa-miR-16 MIMAT0000069 Homo sapiens miR-16 Targets (miRBase)(0.579)<BR/><A target=\"_blank\" HREF=\"homerResults/motif23.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif23.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif23.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>24\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,2,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,0,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,-1.5,39)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.700,0,34)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,27,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,25,44)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,23.5,39)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.700,25,34)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,52,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,50,44)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,50,39)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.700,48.5,34)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,77,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,75,44)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,75,39)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.700,73.5,34)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,102,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,100,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,98.5,39)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.700,100,34)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,127,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,125,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,123.5,39)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.700,125,34)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,150,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,150,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,148.5,39)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.700,152,34)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,177,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,175,44)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,175,39)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.700,173.5,34)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,200,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,200,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,198.5,39)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.700,202,34)\">A</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-8</TD><TD>-1.908e+01</TD><TD>0.19%</TD><TD>0.00%</TD><TD>47.2bp (0.0bp)</TD><TD>hsa-miR-3688-3p MIMAT0018116 Homo sapiens miR-3688-3p Targets (miRBase)(0.703)<BR/><A target=\"_blank\" HREF=\"homerResults/motif24.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif24.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif24.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>25\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.063,0,49)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.068,2,45.85)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.078,0,42.45)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.791,-1.5,38.55)\">U</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.058,23.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.079,25,46.1)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.113,25,42.15)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.750,27,36.5)\">A</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.056,48.5,49)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.098,50,46.2)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.120,52,41.3)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.726,50,35.3)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.016,75,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.047,75,48.2)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.139,77,45.85)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.798,73.5,38.9)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.057,102,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.074,98.5,46.15)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.125,100,42.45)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.744,100,36.2)\">C</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.145,125,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.146,127,41.75)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.176,125,34.45)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.533,123.5,25.65)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.044,150,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.050,150,46.8)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.076,152,44.3)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.830,148.5,40.5)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.018,177,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.053,175,48.1)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.094,175,45.45)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.835,173.5,40.75)\">U</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-8</TD><TD>-1.851e+01</TD><TD>22.18%</TD><TD>17.82%</TD><TD>60.5bp (58.2bp)</TD><TD>hsa-miR-583 MIMAT0003248 Homo sapiens miR-583 Targets (miRBase)(0.685)<BR/><A target=\"_blank\" HREF=\"homerResults/motif25.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif25.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif25.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>26\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.110,-1.5,49)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.135,2,43.5)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.148,0,36.75)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.607,0,29.35)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.083,27,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.099,25,44.85)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.151,25,39.9)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.667,23.5,32.35)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.089,52,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.091,50,44.55)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.311,50,40)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.509,48.5,24.45)\">U</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.081,73.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.135,75,44.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.142,75,38.2)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.642,77,31.1)\">A</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.055,100,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.129,102,46.25)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.223,100,39.8)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.593,98.5,28.65)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.153,125,49)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.167,127,41.35)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.193,123.5,33)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.487,125,23.35)\">G</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.125,148.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.178,150,42.75)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.198,152,33.85)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.499,150,23.95)\">G</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.118,173.5,49)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.125,175,43.1)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.191,175,36.85)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.566,177,27.3)\">A</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-7</TD><TD>-1.780e+01</TD><TD>10.86%</TD><TD>7.79%</TD><TD>60.7bp (63.1bp)</TD><TD>hsa-miR-587 MIMAT0003253 Homo sapiens miR-587 Targets (miRBase)(0.657)<BR/><A target=\"_blank\" HREF=\"homerResults/motif26.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif26.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif26.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>27\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.056,2,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.067,-1.5,46.2)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.127,0,42.85)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.750,0,36.5)\">C</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.052,25,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.055,27,46.4)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.059,25,43.65)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.834,23.5,40.7)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.036,52,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.076,50,47.2)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.085,48.5,43.4)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.803,50,39.15)\">G</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.029,75,49)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.042,77,47.55)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.131,75,45.45)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.798,73.5,38.9)\">U</text>\n",
       "  \n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.083,98.5,49)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.124,100,44.85)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.135,102,38.65)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.658,100,31.9)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.048,127,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.056,123.5,46.6)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.191,125,43.8)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.705,125,34.25)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.062,150,49)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.103,152,45.9)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.127,150,40.75)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.708,148.5,34.4)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.065,175,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.085,175,45.75)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.095,173.5,41.5)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.755,177,36.75)\">A</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-7</TD><TD>-1.678e+01</TD><TD>15.36%</TD><TD>11.84%</TD><TD>60.8bp (59.5bp)</TD><TD>hsa-miR-140-3p MIMAT0004597 Homo sapiens miR-140-3p Targets (miRBase)(0.732)<BR/><A target=\"_blank\" HREF=\"homerResults/motif27.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif27.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif27.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>28\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,2,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,0,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,-1.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,0,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,25,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.167,25,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.831,23.5,40.55)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,52,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.064,50,48.9)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.934,50,45.7)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,75,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.124,73.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.874,75,42.7)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,100,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,100,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,98.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.997,102,48.85)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,127,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,125,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,123.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,125,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,152,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,150,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,148.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,150,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,177,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,175,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,173.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,175,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,202,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,200,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,198.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,200,48.85)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-7</TD><TD>-1.665e+01</TD><TD>1.35%</TD><TD>0.47%</TD><TD>61.8bp (57.1bp)</TD><TD>hsa-miR-1281 MIMAT0005939 Homo sapiens miR-1281 Targets (miRBase)(0.677)<BR/><A target=\"_blank\" HREF=\"homerResults/motif28.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif28.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif28.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>29\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.014,2,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.041,0,48.3)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.061,0,46.25)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.884,-1.5,43.2)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.041,27,49)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.048,23.5,46.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.096,25,44.55)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.815,25,39.75)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,52,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.052,50,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.062,48.5,46.35)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.885,50,43.25)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,75,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.023,77,48.9)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.975,73.5,47.75)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.073,100,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.087,98.5,45.35)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.181,100,41)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.659,102,31.95)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,125,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.022,125,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.081,123.5,47.85)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.896,127,43.8)\">A</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,150,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,148.5,48.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.019,152,48.9)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.979,150,47.95)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,177,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,175,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.059,173.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.939,175,45.95)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-6</TD><TD>-1.445e+01</TD><TD>5.39%</TD><TD>3.49%</TD><TD>61.4bp (64.8bp)</TD><TD>hsa-miR-409-5p MIMAT0001638 Homo sapiens miR-409-5p Targets (miRBase)(0.773)<BR/><A target=\"_blank\" HREF=\"homerResults/motif29.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif29.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif29.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>30\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"155\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,2,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,-1.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,0,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,25,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,23.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,25,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,50,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,50,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.997,52,48.85)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,48.95)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.354,75,48.9)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.644,73.5,31.2)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,100,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,100,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,98.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.997,102,48.85)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,127,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,125,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,123.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,125,48.85)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-6</TD><TD>-1.431e+01</TD><TD>2.62%</TD><TD>1.36%</TD><TD>51.7bp (72.0bp)</TD><TD>hsa-miR-3713 MIMAT0018164 Homo sapiens miR-3713 Targets (miRBase)(0.599)<BR/><A target=\"_blank\" HREF=\"homerResults/motif30.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif30.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif30.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>31\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"180\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,2,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,0,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.997,-1.5,48.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,25,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,23.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,25,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,50,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,50,48.95)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.332,52,48.9)\">A</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.666,48.5,32.3)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,75,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.997,73.5,48.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,100,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,100,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,98.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.997,102,48.85)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,125,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,125,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,123.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.997,127,48.85)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,152,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,150,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,150,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.997,148.5,48.85)\">U</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-6</TD><TD>-1.397e+01</TD><TD>4.47%</TD><TD>2.78%</TD><TD>62.1bp (62.1bp)</TD><TD>hsa-miR-302c* MIMAT0000716 Homo sapiens miR-302c* Targets (miRBase)(0.698)<BR/><A target=\"_blank\" HREF=\"homerResults/motif31.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif31.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif31.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>32\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"180\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,2,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,0,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.997,-1.5,48.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,25,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,23.5,48.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.102,27,48.9)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.896,25,43.8)\">G</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,50,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,48.95)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.017,52,48.9)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.981,50,48.05)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,75,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,73.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,75,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,102,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,100,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,98.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,100,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,127,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,125,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,123.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,125,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,152,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,150,48.95)\">G</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.081,150,48.9)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.917,148.5,44.85)\">U</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-5</TD><TD>-1.367e+01</TD><TD>9.47%</TD><TD>6.98%</TD><TD>56.5bp (61.5bp)</TD><TD>hsa-miR-4688 MIMAT0019777 Homo sapiens miR-4688 Targets (miRBase)(0.860)<BR/><A target=\"_blank\" HREF=\"homerResults/motif32.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif32.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif32.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>33\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,-1.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.121,0,48.9)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.877,2,42.85)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,25,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,23.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,25,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,50,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.045,50,48.9)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.953,52,46.65)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,75,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,73.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,75,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,102,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,100,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,98.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,100,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,127,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,125,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,123.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,125,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,150,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,148.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.158,150,48.9)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.840,152,41)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,177,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,175,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,173.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,175,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,202,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,200,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,198.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,200,48.85)\">C</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-5</TD><TD>-1.347e+01</TD><TD>0.77%</TD><TD>0.21%</TD><TD>54.6bp (59.1bp)</TD><TD>hsa-miR-760 MIMAT0004957 Homo sapiens miR-760 Targets (miRBase)(0.662)<BR/><A target=\"_blank\" HREF=\"homerResults/motif33.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif33.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif33.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>34\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"230\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,2,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,0,44)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,-1.5,39)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.700,0,34)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,27,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,25,44)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,23.5,39)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.700,25,34)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,52,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,50,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,48.5,39)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.700,50,34)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,77,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,75,44)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,75,39)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.700,73.5,34)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,100,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,100,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,98.5,39)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.700,102,34)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,125,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,125,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,123.5,39)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.700,127,34)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,152,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,150,44)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,148.5,39)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.700,150,34)\">G</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,175,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,175,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,173.5,39)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.700,177,34)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,202,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,200,44)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,200,39)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.700,198.5,34)\">U</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-5</TD><TD>-1.204e+01</TD><TD>0.31%</TD><TD>0.04%</TD><TD>12.3bp (66.1bp)</TD><TD>hsa-miR-103b MIMAT0007402 Homo sapiens miR-103b Targets (miRBase)(0.654)<BR/><A target=\"_blank\" HREF=\"homerResults/motif34.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif34.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif34.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>35\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"180\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,2,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,-1.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,0,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,25,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.079,25,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.919,23.5,44.95)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,52,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,50,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,50,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.997,48.5,48.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,75,49)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,73.5,48.95)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.123,75,48.9)\">G</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.875,77,42.75)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,100,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,100,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.342,98.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.656,102,31.8)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,127,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,125,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,125,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.997,123.5,48.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,152,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,150,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.204,150,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.794,148.5,38.7)\">U</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-4</TD><TD>-1.040e+01</TD><TD>11.32%</TD><TD>8.98%</TD><TD>61.7bp (60.4bp)</TD><TD>hsa-miR-4477a MIMAT0019004 Homo sapiens miR-4477a Targets (miRBase)(0.687)<BR/><A target=\"_blank\" HREF=\"homerResults/motif35.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif35.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif35.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>36\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,2,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,0,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,0,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.997,-1.5,48.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,27,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,25,48.95)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,25,48.9)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.997,23.5,48.85)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,50,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,50,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,48.5,48.9)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.997,52,48.85)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,77,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,75,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,73.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,75,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,102,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,100,48.95)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,98.5,48.9)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.997,100,48.85)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,127,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,125,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,123.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,125,48.85)\">G</text>\n",
       "  \n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.001,150,49)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,148.5,48.95)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.217,150,48.9)\">C</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.781,152,38.05)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.001,177,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.001,175,48.95)\">C</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.001,173.5,48.9)\">U</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.997,175,48.85)\">G</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-4</TD><TD>-9.828e+00</TD><TD>0.23%</TD><TD>0.03%</TD><TD>40.6bp (44.6bp)</TD><TD>hsa-miR-720 MIMAT0005954 Homo sapiens miR-720 Targets (miRBase)(0.650)<BR/><A target=\"_blank\" HREF=\"homerResults/motif36.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif36.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif36.motif\">motif file (matrix)</A></TD></TR>\n",
       "<TR><TD>37\n",
       "<FONT color=\"red\">*</FONT></TD><TD><svg width=\"205\" height=\"50\">\n",
       " <g font-family=\"Arial\" font-weight=\"bold\" font-size=\"66.5\">\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,0,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,0,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,-1.5,39)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.700,2,34)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,25,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,25,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,23.5,39)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.700,27,34)\">A</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,52,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,50,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,48.5,39)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.700,50,34)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,77,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,75,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,73.5,39)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.700,75,34)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,102,49)\">A</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,100,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,98.5,39)\">U</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.700,100,34)\">C</text>\n",
       "  \n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.100,127,49)\">A</text>\n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,125,44)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,125,39)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.700,123.5,34)\">U</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,150,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,150,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,148.5,39)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.700,152,34)\">A</text>\n",
       "  \n",
       "  <text fill=\"#0000EE\" x=\"0\" y=\"0\"  transform=\"matrix(0.59,0,0,0.100,175,49)\">C</text>\n",
       "  <text fill=\"#F9A500\" x=\"0\" y=\"0\"  transform=\"matrix(0.55,0,0,0.100,175,44)\">G</text>\n",
       "  <text fill=\"#DD0000\" x=\"0\" y=\"0\"  transform=\"matrix(0.65,0,0,0.100,173.5,39)\">U</text>\n",
       "  <text fill=\"#00BB00\" x=\"0\" y=\"0\"  transform=\"matrix(0.53,0,0,0.700,177,34)\">A</text>\n",
       "  \n",
       " </g>\n",
       "</svg>\n",
       "</TD><TD>1e-2</TD><TD>-6.523e+00</TD><TD>98.00%</TD><TD>97.04%</TD><TD>55.5bp (63.3bp)</TD><TD>hsa-miR-584 MIMAT0003249 Homo sapiens miR-584 Targets (miRBase)(0.699)<BR/><A target=\"_blank\" HREF=\"homerResults/motif37.info.html\">More Information</A> | <A target=\"blank\" HREF=\"homerResults/motif37.similar.html\">Similar Motifs Found</A></TD><TD><A target=\"_blank\" HREF=\"homerResults/motif37.motif\">motif file (matrix)</A></TD></TR>\n",
       "</TABLE>\n",
       "<P></P>\n",
       "</BODY></HTML>\n"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 727,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import HTML\n",
    "HTML(filename=out_path/\"homerResults.html\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aabbe1e4-9117-4f77-ad88-275e25757124",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec8f3b4f-3c53-46f3-aa75-0df42384b7ee",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
