{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1336fe2c",
   "metadata": {},
   "source": [
    "# UBS-seq Pipeline\n",
    "## Basic Workflow"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4f9191a",
   "metadata": {},
   "source": [
    "**C. Bryan Daniels**\n",
    "\n",
    "**2/1/2024**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1153fdea-42ab-4a16-9376-aae40c035695",
   "metadata": {},
   "source": [
    "## Project: UBS-seq Basic Pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1999833",
   "metadata": {},
   "source": [
    "The purpose of this project is run a minimally viable UBS-seq pipline. For simplicity, it will run several single-end samples, mapping only to the genome. The core steps of the pipeline are:\n",
    "- cut_apapter\n",
    "- quality_control\n",
    "- align2ref\n",
    "- sort2ref\n",
    "- dedupe\n",
    "- filter->all_multi_unique\n",
    "- call_peaks\n",
    "- select_groups\n",
    "- analysis_and_annotation\n",
    "\n",
    "This pipeline is based upon the paper by [Qing Dai, etal](https://doi.org/10.1038/s41587-023-02034-w) and the UBS-seq pipeline developed by [Chang Ye](https://github.com/y9c/m5C-UBSseq)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db5933a4-a627-451d-abe7-7e68204c1d47",
   "metadata": {},
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd477b9e-5c7d-4c3c-8125-89644f367a8d",
   "metadata": {},
   "source": [
    "#### The logic for the Pipeline is defined through a series of Steps using dirs to save intermediate results\n",
    "1. For each **Step** in the pipeline a dir will be created and labeled **Step** and will contain all files created by that **Step**\n",
    "2. Within a **Step**, **in_path** and **out_path** will generically refer to the prior and current **Step**\n",
    "3. Within each **Step**, the appropriate processes will occur. Generally this involves processing files from **in_path** and saving to **out_path**\n",
    "4. **Abbreviated filenames** should not change through the pipeline (suffixes will reflect current file formats). The dir name should reflect the **Step**, not the filename.\n",
    "6. The function **mkpath(step)** returns a path for a dir **Step**. It will create a dir if need be, but not overwrite an existing dir\n",
    "8. The function **fname(path,sample,suffix)** returns a file name without actually creating the file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0852de28-a1f7-4239-a50a-5cba49c7fe72",
   "metadata": {},
   "source": [
    "#### Execution from Command Line"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21bc12e6-1711-424b-b381-66b313c2dbf7",
   "metadata": {},
   "source": [
    "- cd workplace/\n",
    "- juptyer nbconvert ubs_basic.ipynb --to script\n",
    "- ipython ubs_basic.py"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13e28624-07ae-4a54-ba80-dfde893059bc",
   "metadata": {},
   "source": [
    "#### Environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "80595b7b-0802-4dd4-93be-8c15ab0cd809",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ubs_basic.ipnby | ubs_basic.py STARTED: Sat Feb 3 05:58:39 AM CST 2024'"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "start = ! echo 'ubs_basic.ipnby | ubs_basic.py STARTED:' `date`; \n",
    "start = start[0]; start"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "41e148d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, sys, subprocess\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt \n",
    "from pathlib import Path\n",
    "from IPython.display import display, HTML\n",
    "from snakemake import load_configfile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "42cc293d-6fe9-45fc-a0cd-59b7dfffb18b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def fname(path, base, sufix):\n",
    "    'Return a path and suffix complete filename'\n",
    "    return Path(path)/f\"{base}.{sufix}\"\n",
    "\n",
    "def mkpath(path):\n",
    "    'Return dir path name; creates (not over-writting) a new dir within the pwd'\n",
    "    path = Path(path)\n",
    "    if not os.path.exists(path): os.makedirs(path)\n",
    "    return path"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "daf19bdc-0368-4ecf-98cf-785917a77577",
   "metadata": {},
   "source": [
    "#### Path References"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4951cf73-f77f-4bc0-8aab-4c5497aa3049",
   "metadata": {},
   "source": [
    "`Waring:` This notebook should be located and executed from within the directory `home_path/workspace/`\n",
    "\n",
    "`Note:` Github ignores the data/, workspace/ and reference directories, except for `.py` and `.ipynb` files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b02f60de-c362-4078-beba-82c870de2c78",
   "metadata": {},
   "outputs": [],
   "source": [
    "home_path      = Path.cwd()/'..'\n",
    "data_path      = home_path/'data'\n",
    "workspace_path = home_path/'workspace'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e069648-6080-4a32-8ade-8a631634a994",
   "metadata": {},
   "source": [
    "Use `config.yaml` to configure `references/`, but not samples in `data/`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0e9cbdfe-97bf-4c0d-9e86-ef93ad062c5d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "config = load_configfile(\"../config.yaml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "9fd3b929-b9ad-48ed-acf8-e9ec4cfe01dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_fa  = home_path/config['reference']['genome']['fa'].removeprefix('~/')\n",
    "genome_idx = home_path/config['reference']['genome']['hisat3n'].removeprefix('~/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "dbfc6989-8a84-4b33-ae5c-ee507af9442f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#genome_fa = ref_path/'genome/Homo_sapiens.GRCh38.genome.fa'\n",
    "#genome_idx = ref_path/'index/hist3n/Homo_sapiens.GRCh38.genome'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ae9c5015",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add to shell PATH\n",
    "os.environ['PATH'] = f\"{str(home_path)}:\" + os.environ['PATH'] # home_path\n",
    "os.environ['PATH'] = '/home/cdaniels/bin/homer:' + os.environ['PATH'] # homer\n",
    "os.environ['PATH'] = '/home/cdaniels/bin/hisat-3n:' + os.environ['PATH'] # hisat-3n\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e2c24ec3-238a-47d3-99a0-295938a8d597",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set Java Flag\n",
    "os.environ['_JAVA_OPTIONS'] = '-Xmx8g'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "889270cf-bf90-4da3-90fe-397d8d0b57f3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "32"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Number of cores                                                                                                                                                                                               \n",
    "nc = get_ipython().getoutput('nproc')                                                                                                                                                                           \n",
    "nc = int(nc[0])                                                                                                                                                                                                 \n",
    "nc  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f8499228-734a-4c2c-ad0a-deab43ae64c7",
   "metadata": {},
   "source": [
    "#### Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "46c1ab4f-937a-4b6a-a09d-a34462da366b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def nlines(file):\n",
    "    'Returns fast linecout (fast)'\n",
    "    result = subprocess.run(['wc', '-l', file], stdout=subprocess.PIPE)\n",
    "    n = int(result.stdout.split()[0])\n",
    "    return n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ed22df72-94d2-412d-bf77-67e25b974acc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def nseqs(bam_fastq):\n",
    "    'Returns number of sequences in bam, sam, fasta or fastq file'\n",
    "    n = !samtools view -c {bam_fastq}\n",
    "    return int(n[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "db7b2623-6c8a-465f-97af-1cd3e358849b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def samples_string(samples,path,suffix='bam'):\n",
    "    'Returns a space delimited string of sample files'\n",
    "    return \" \".join([str(fname(path,sample,suffix)) for sample in samples])    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f9ffe35a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_table(ds1, ds2, ds1_name, ds2_name, y_label=None, xs_labels=None, table_label=None):\n",
    "    'Makes a bar graph comparing two datasets and their corresponding names'\n",
    "    assert(len(ds1)==len(ds2))\n",
    "    n = np.arange(len(ds1))\n",
    "    width = 0.35\n",
    "    fig, ax = plt.subplots()\n",
    "    rects1 = ax.bar(n - width/2, ds1, width, label=ds1_name)\n",
    "    rects2 = ax.bar(n + width/2, ds2, width,label=ds2_name)\n",
    "    ax.set_ylabel(y_label)\n",
    "    ax.set_title(table_label)\n",
    "    ax.set_xticks(n)\n",
    "    ax.set_xticklabels(range(1,len(ds1)+1)) if xs_labels == None else ax.set_xticklabels(xs_labels)\n",
    "    ax.legend()\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a7b59797-ffd0-4d1c-947e-45f07e301a2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_histogram(ds, ds_name, table_label=None, y_label=\"Frequency\", density=True):\n",
    "    'Makes a histogram for a dataset and its name'\n",
    "    fig, ax = plt.subplots()\n",
    "    ax.hist(ds, density=density)\n",
    "    ax.set_ylabel(y_label)\n",
    "    ax.set_xlabel(ds_name)\n",
    "    if table_label is not None:\n",
    "        ax.set_title(table_label)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a21c1f30-2663-4aa7-9e81-fd90a96b92e5",
   "metadata": {},
   "source": [
    "## Step: select_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf0c6fbb-a4b9-4a0c-97d2-80646efb627e",
   "metadata": {},
   "source": [
    "#### Select Samples and Filenames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "d73f04c9-8803-4e6f-8336-a53a40d78bab",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "out_path = mkpath('samples')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1245b4d6-d147-423a-8d74-038e689e960e",
   "metadata": {},
   "source": [
    "#### Project definitions for treated samples, control samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "67bbf37a-e8c4-4f9f-a267-251bed3f485f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# short sample names\n",
    "treated = [\"t1\",\"t2\",\"t3\"] # Edit\n",
    "control = [] # Edit\n",
    "samples = treated + control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "32c7d8e1-0b1b-495e-af19-7de635060e12",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['t1', 't2', 't3']"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a1738eec-e76c-4104-858a-5319359818a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# sample filenames\n",
    "samples_fn = [data_path/'test0_R1.fq.gz', data_path/'test1_R1.fq.gz', data_path/'test2_R1.fq.gz']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f98ef3fd-e6fc-442d-9147-d2ab1dc04074",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'t1': PosixPath('/home/cdaniels/uofc_data/ubs_seq/UBS-seq_basic/workspace/../data/test0_R1.fq.gz'),\n",
       " 't2': PosixPath('/home/cdaniels/uofc_data/ubs_seq/UBS-seq_basic/workspace/../data/test1_R1.fq.gz'),\n",
       " 't3': PosixPath('/home/cdaniels/uofc_data/ubs_seq/UBS-seq_basic/workspace/../data/test2_R1.fq.gz')}"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# dict\n",
    "s2fn = {name: fname for (name,fname) in zip(samples,samples_fn)};\n",
    "s2fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4c84cd58-777e-4645-bac1-5a3c549ad8dd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ln: failed to create symbolic link 'samples/t1.fq.gz': File exists\n",
      "ln: failed to create symbolic link 'samples/t2.fq.gz': File exists\n",
      "ln: failed to create symbolic link 'samples/t3.fq.gz': File exists\n"
     ]
    }
   ],
   "source": [
    "for (s,fn) in s2fn.items():\n",
    "    ! ln -s {data_path/fn} {fname(out_path,s,\"fq.gz\")}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e11be53",
   "metadata": {},
   "source": [
    "Check that the files look correct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "25371f62",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 647M\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 178M Feb  1 11:41 t1.fq.gz\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 239M Feb  1 11:41 t2.fq.gz\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 230M Feb  1 11:41 t3.fq.gz\n"
     ]
    }
   ],
   "source": [
    "!ls -lLh {out_path}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c7307e9-bcb8-4515-be28-da83ddd2649a",
   "metadata": {},
   "source": [
    "How many sequence reads do we have per Sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "113b16b9-601a-4b74-83e7-12d25a9af329",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samples/t1.fq.gz: 4,678,947\n",
      "samples/t2.fq.gz: 6,830,155\n",
      "samples/t3.fq.gz: 5,919,881\n"
     ]
    }
   ],
   "source": [
    "# Total Reads per Samples\n",
    "files = [fname(out_path,sample, \"fq.gz\") for sample in samples]\n",
    "for f in files:\n",
    "    ns = nseqs(f)\n",
    "    print(f\"{f}: {ns:,}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41fb4092-ae2c-4081-8658-300b9f3ffab8",
   "metadata": {},
   "source": [
    "## Step: fastqc_pre"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e95850c4-5827-459f-bbf9-04cad8347377",
   "metadata": {},
   "source": [
    "#### Pre Trimming Quality Control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "20a87550-8f06-4d37-9c01-025dc8135790",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_path = mkpath(\"samples\")\n",
    "out_path = mkpath(\"fastqc_pre\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16857b29-a33c-424a-a224-22142e6d2e13",
   "metadata": {},
   "source": [
    "#### fastqc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "c0dd3668-5561-44c7-b2ea-b903ebc1737e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "            FastQC - A high throughput sequence QC analysis tool\n",
      "\n",
      "SYNOPSIS\n",
      "\n",
      "\tfastqc seqfile1 seqfile2 .. seqfileN\n",
      "\n",
      "    fastqc [-o output dir] [--(no)extract] [-f fastq|bam|sam] \n",
      "           [-c contaminant file] seqfile1 .. seqfileN\n",
      "\n",
      "DESCRIPTION\n",
      "\n",
      "    FastQC reads a set of sequence files and produces from each one a quality\n",
      "    control report consisting of a number of different modules, each one of \n",
      "    which will help to identify a different potential type of problem in your\n",
      "    data.\n",
      "    \n",
      "    If no files to process are specified on the command line then the program\n",
      "    will start as an interactive graphical application.  If files are provided\n",
      "    on the command line then the program will run with no user interaction\n",
      "    required.  In this mode it is suitable for inclusion into a standardised\n",
      "    analysis pipeline.\n",
      "    \n",
      "    The options for the program as as follows:\n",
      "    \n",
      "    -h --help       Print this help file and exit\n",
      "    \n",
      "    -v --version    Print the version of the program and exit\n",
      "    \n",
      "    -o --outdir     Create all output files in the specified output directory.\n",
      "                    Please note that this directory must exist as the program\n",
      "                    will not create it.  If this option is not set then the \n",
      "                    output file for each sequence file is created in the same\n",
      "                    directory as the sequence file which was processed.\n",
      "                    \n",
      "    --casava        Files come from raw casava output. Files in the same sample\n",
      "                    group (differing only by the group number) will be analysed\n",
      "                    as a set rather than individually. Sequences with the filter\n",
      "                    flag set in the header will be excluded from the analysis.\n",
      "                    Files must have the same names given to them by casava\n",
      "                    (including being gzipped and ending with .gz) otherwise they\n",
      "                    won't be grouped together correctly.\n",
      "                    \n",
      "    --nano          Files come from nanopore sequences and are in fast5 format. In\n",
      "                    this mode you can pass in directories to process and the program\n",
      "                    will take in all fast5 files within those directories and produce\n",
      "                    a single output file from the sequences found in all files.                    \n",
      "                    \n",
      "    --nofilter      If running with --casava then don't remove read flagged by\n",
      "                    casava as poor quality when performing the QC analysis.\n",
      "                   \n",
      "    --extract       If set then the zipped output file will be uncompressed in\n",
      "                    the same directory after it has been created. If --delete is \n",
      "                    also specified then the zip file will be removed after the \n",
      "                    contents are unzipped. \n",
      "                    \n",
      "    -j --java       Provides the full path to the java binary you want to use to\n",
      "                    launch fastqc. If not supplied then java is assumed to be in\n",
      "                    your path.\n",
      "                   \n",
      "    --noextract     Do not uncompress the output file after creating it.  You\n",
      "                    should set this option if you do not wish to uncompress\n",
      "                    the output when running in non-interactive mode.\n",
      "                    \n",
      "    --nogroup       Disable grouping of bases for reads >50bp. All reports will\n",
      "                    show data for every base in the read.  WARNING: Using this\n",
      "                    option will cause fastqc to crash and burn if you use it on\n",
      "                    really long reads, and your plots may end up a ridiculous size.\n",
      "                    You have been warned!\n",
      "                    \n",
      "    --min_length    Sets an artificial lower limit on the length of the sequence\n",
      "                    to be shown in the report.  As long as you set this to a value\n",
      "                    greater or equal to your longest read length then this will be\n",
      "                    the sequence length used to create your read groups.  This can\n",
      "                    be useful for making directly comaparable statistics from \n",
      "                    datasets with somewhat variable read lengths.\n",
      "\n",
      "    --dup_length    Sets a length to which the sequences will be truncated when \n",
      "                    defining them to be duplicates, affecting the duplication and\n",
      "                    overrepresented sequences plot.  This can be useful if you have\n",
      "                    long reads with higher levels of miscalls, or contamination with\n",
      "                    adapter dimers containing UMI sequences.\n",
      "\n",
      "                    \n",
      "    -f --format     Bypasses the normal sequence file format detection and\n",
      "                    forces the program to use the specified format.  Valid\n",
      "                    formats are bam,sam,bam_mapped,sam_mapped and fastq\n",
      "                    \n",
      "\n",
      "    --memory        Sets the base amount of memory, in Megabytes, used to process \n",
      "                    each file.  Defaults to 512MB.  You may need to increase this if\n",
      "                    you have a file with very long sequences in it.\n",
      "                \n",
      "    --svg           Save the graphs in the report in SVG format.\n",
      "\n",
      "    -t --threads    Specifies the number of files which can be processed\n",
      "                    simultaneously.  Each thread will be allocated 250MB of\n",
      "                    memory so you shouldn't run more threads than your\n",
      "                    available memory will cope with, and not more than\n",
      "                    6 threads on a 32 bit machine\n",
      "                  \n",
      "    -c              Specifies a non-default file which contains the list of\n",
      "    --contaminants  contaminants to screen overrepresented sequences against.\n",
      "                    The file must contain sets of named contaminants in the\n",
      "                    form name[tab]sequence.  Lines prefixed with a hash will\n",
      "                    be ignored.\n",
      "\n",
      "    -a              Specifies a non-default file which contains the list of\n",
      "    --adapters      adapter sequences which will be explicity searched against\n",
      "                    the library. The file must contain sets of named adapters\n",
      "                    in the form name[tab]sequence.  Lines prefixed with a hash\n",
      "                    will be ignored.\n",
      "                    \n",
      "    -l              Specifies a non-default file which contains a set of criteria\n",
      "    --limits        which will be used to determine the warn/error limits for the\n",
      "                    various modules.  This file can also be used to selectively \n",
      "                    remove some modules from the output all together.  The format\n",
      "                    needs to mirror the default limits.txt file found in the\n",
      "                    Configuration folder.\n",
      "                    \n",
      "   -k --kmers       Specifies the length of Kmer to look for in the Kmer content\n",
      "                    module. Specified Kmer length must be between 2 and 10. Default\n",
      "                    length is 7 if not specified.\n",
      "                    \n",
      "   -q --quiet       Suppress all progress messages on stdout and only report errors.\n",
      "   \n",
      "   -d --dir         Selects a directory to be used for temporary files written when\n",
      "                    generating report images. Defaults to system temp directory if\n",
      "                    not specified.\n",
      "                    \n",
      "BUGS\n",
      "\n",
      "    Any bugs in fastqc should be reported either to simon.andrews@babraham.ac.uk\n",
      "    or in www.bioinformatics.babraham.ac.uk/bugzilla/\n",
      "                   \n",
      "    \n"
     ]
    }
   ],
   "source": [
    "! fastqc --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb840f81-7880-47b5-aab1-b6017da470e3",
   "metadata": {},
   "source": [
    "##### paramaters:\n",
    "- -o output dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f0f6379b-63d0-4ed9-b3c7-c01accd81bb4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n",
      "Analysis complete for t1.fq.gz\n",
      "application/gzip\n",
      "Analysis complete for t2.fq.gz\n",
      "application/gzip\n",
      "Analysis complete for t3.fq.gz\n"
     ]
    }
   ],
   "source": [
    "for sample in samples:\n",
    "    ! fastqc {fname(in_path,sample,\"fq.gz\")} -o {out_path} 2> /dev/null"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "f009c6d6-c2d9-4b1c-b628-53e2ce7ba2d9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "t1_fastqc.html\tt2_fastqc.html\tt3_fastqc.html\n",
      "t1_fastqc.zip\tt2_fastqc.zip\tt3_fastqc.zip\n"
     ]
    }
   ],
   "source": [
    "! ls {out_path}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51157b74",
   "metadata": {},
   "source": [
    "## Step: trimmed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1731c31-b9b1-4a09-b964-3dece49e1910",
   "metadata": {},
   "source": [
    "#### Trim the adapter and downstream sequence as well as trimmng lower quality downstream sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "8b898be1",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_path = mkpath(\"samples\")\n",
    "out_path = mkpath(\"trim\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "632eaa99",
   "metadata": {},
   "outputs": [],
   "source": [
    "adapter =  \"AGATCGGAAGAGCACACGTCT\"\n",
    "barcode3 = \"ATCACG\"\n",
    "#adapter = \"AGATCGGAAGAGCACACGTCTGAACTCCAG\"\n",
    "#barcode3 = \"TATCACGATCACG\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e73ebe78-4020-4894-8c0c-3239aee02330",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "t1.fq.gz  t2.fq.gz  t3.fq.gz\n"
     ]
    }
   ],
   "source": [
    "!ls {in_path}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca715978-8029-4cfe-a74c-765eb8ed882f",
   "metadata": {},
   "source": [
    "#### cutadapt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "bbb14bd9",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cutadapt version 4.6\n",
      "\n",
      "Copyright (C) 2010 Marcel Martin <marcel.martin@scilifelab.se> and contributors\n",
      "\n",
      "Cutadapt removes adapter sequences from high-throughput sequencing reads.\n",
      "\n",
      "Usage:\n",
      "    cutadapt -a ADAPTER [options] [-o output.fastq] input.fastq\n",
      "\n",
      "For paired-end reads:\n",
      "    cutadapt -a ADAPT1 -A ADAPT2 [options] -o out1.fastq -p out2.fastq in1.fastq in2.fastq\n",
      "\n",
      "Replace \"ADAPTER\" with the actual sequence of your 3' adapter. IUPAC wildcard\n",
      "characters are supported. All reads from input.fastq will be written to\n",
      "output.fastq with the adapter sequence removed. Adapter matching is\n",
      "error-tolerant. Multiple adapter sequences can be given (use further -a\n",
      "options), but only the best-matching adapter will be removed.\n",
      "\n",
      "Input may also be in FASTA format. Compressed input and output is supported and\n",
      "auto-detected from the file name (.gz, .xz, .bz2). Use the file name '-' for\n",
      "standard input/output. Without the -o option, output is sent to standard output.\n",
      "\n",
      "Citation:\n",
      "\n",
      "Marcel Martin. Cutadapt removes adapter sequences from high-throughput\n",
      "sequencing reads. EMBnet.Journal, 17(1):10-12, May 2011.\n",
      "http://dx.doi.org/10.14806/ej.17.1.200\n",
      "\n",
      "Run \"cutadapt --help\" to see all command-line options.\n",
      "See https://cutadapt.readthedocs.io/ for full documentation.\n",
      "\n",
      "Options:\n",
      "  -h, --help            Show this help message and exit\n",
      "  --version             Show version number and exit\n",
      "  --debug               Print debug log. Use twice to also print DP matrices\n",
      "  -j CORES, --cores CORES\n",
      "                        Number of CPU cores to use. Use 0 to auto-detect.\n",
      "                        Default: 1\n",
      "\n",
      "Finding adapters:\n",
      "  Parameters -a, -g, -b specify adapters to be removed from each read (or from\n",
      "  R1 if data is paired-end. If specified multiple times, only the best\n",
      "  matching adapter is trimmed (but see the --times option). Use notation\n",
      "  'file:FILE' to read adapter sequences from a FASTA file.\n",
      "\n",
      "  -a ADAPTER, --adapter ADAPTER\n",
      "                        Sequence of an adapter ligated to the 3' end (paired\n",
      "                        data: of the first read). The adapter and subsequent\n",
      "                        bases are trimmed. If a '$' character is appended\n",
      "                        ('anchoring'), the adapter is only found if it is a\n",
      "                        suffix of the read.\n",
      "  -g ADAPTER, --front ADAPTER\n",
      "                        Sequence of an adapter ligated to the 5' end (paired\n",
      "                        data: of the first read). The adapter and any preceding\n",
      "                        bases are trimmed. Partial matches at the 5' end are\n",
      "                        allowed. If a '^' character is prepended ('anchoring'),\n",
      "                        the adapter is only found if it is a prefix of the read.\n",
      "  -b ADAPTER, --anywhere ADAPTER\n",
      "                        Sequence of an adapter that may be ligated to the 5' or\n",
      "                        3' end (paired data: of the first read). Both types of\n",
      "                        matches as described under -a and -g are allowed. If the\n",
      "                        first base of the read is part of the match, the\n",
      "                        behavior is as with -g, otherwise as with -a. This\n",
      "                        option is mostly for rescuing failed library\n",
      "                        preparations - do not use if you know which end your\n",
      "                        adapter was ligated to!\n",
      "  -e E, --error-rate E, --errors E\n",
      "                        Maximum allowed error rate (if 0 <= E < 1), or absolute\n",
      "                        number of errors for full-length adapter match (if E is\n",
      "                        an integer >= 1). Error rate = no. of errors divided by\n",
      "                        length of matching region. Default: 0.1 (10%)\n",
      "  --no-indels           Allow only mismatches in alignments. Default: allow both\n",
      "                        mismatches and indels\n",
      "  -n COUNT, --times COUNT\n",
      "                        Remove up to COUNT adapters from each read. Default: 1\n",
      "  -O MINLENGTH, --overlap MINLENGTH\n",
      "                        Require MINLENGTH overlap between read and adapter for\n",
      "                        an adapter to be found. Default: 3\n",
      "  --match-read-wildcards\n",
      "                        Interpret IUPAC wildcards in reads. Default: False\n",
      "  -N, --no-match-adapter-wildcards\n",
      "                        Do not interpret IUPAC wildcards in adapters.\n",
      "  --action {trim,retain,mask,lowercase,none}\n",
      "                        What to do if a match was found. trim: trim adapter and\n",
      "                        up- or downstream sequence; retain: trim, but retain\n",
      "                        adapter; mask: replace with 'N' characters; lowercase:\n",
      "                        convert to lowercase; none: leave unchanged. Default:\n",
      "                        trim\n",
      "  --rc, --revcomp       Check both the read and its reverse complement for\n",
      "                        adapter matches. If match is on reverse-complemented\n",
      "                        version, output that one. Default: check only read\n",
      "\n",
      "Additional read modifications:\n",
      "  -u LEN, --cut LEN     Remove LEN bases from each read (or R1 if paired; use -U\n",
      "                        option for R2). If LEN is positive, remove bases from\n",
      "                        the beginning. If LEN is negative, remove bases from the\n",
      "                        end. Can be used twice if LENs have different signs.\n",
      "                        Applied *before* adapter trimming.\n",
      "  --nextseq-trim 3'CUTOFF\n",
      "                        NextSeq-specific quality trimming (each read). Trims\n",
      "                        also dark cycles appearing as high-quality G bases.\n",
      "  -q [5'CUTOFF,]3'CUTOFF, --quality-cutoff [5'CUTOFF,]3'CUTOFF\n",
      "                        Trim low-quality bases from 5' and/or 3' ends of each\n",
      "                        read before adapter removal. Applied to both reads if\n",
      "                        data is paired. If one value is given, only the 3' end\n",
      "                        is trimmed. If two comma-separated cutoffs are given,\n",
      "                        the 5' end is trimmed with the first cutoff, the 3' end\n",
      "                        with the second.\n",
      "  --quality-base N      Assume that quality values in FASTQ are encoded as\n",
      "                        ascii(quality + N). This needs to be set to 64 for some\n",
      "                        old Illumina FASTQ files. Default: 33\n",
      "  --poly-a              Trim poly-A tails\n",
      "  --length LENGTH, -l LENGTH\n",
      "                        Shorten reads to LENGTH. Positive values remove bases at\n",
      "                        the end while negative ones remove bases at the\n",
      "                        beginning. This and the following modifications are\n",
      "                        applied after adapter trimming.\n",
      "  --trim-n              Trim N's on ends of reads.\n",
      "  --length-tag TAG      Search for TAG followed by a decimal number in the\n",
      "                        description field of the read. Replace the decimal\n",
      "                        number with the correct length of the trimmed read. For\n",
      "                        example, use --length-tag 'length=' to correct fields\n",
      "                        like 'length=123'.\n",
      "  --strip-suffix STRIP_SUFFIX\n",
      "                        Remove this suffix from read names if present. Can be\n",
      "                        given multiple times.\n",
      "  -x PREFIX, --prefix PREFIX\n",
      "                        Add this prefix to read names. Use {name} to insert the\n",
      "                        name of the matching adapter.\n",
      "  -y SUFFIX, --suffix SUFFIX\n",
      "                        Add this suffix to read names; can also include {name}\n",
      "  --rename TEMPLATE     Rename reads using TEMPLATE containing variables such as\n",
      "                        {id}, {adapter_name} etc. (see documentation)\n",
      "  --zero-cap, -z        Change negative quality values to zero.\n",
      "\n",
      "Filtering of processed reads:\n",
      "  Filters are applied after above read modifications. Paired-end reads are\n",
      "  always discarded pairwise (see also --pair-filter).\n",
      "\n",
      "  -m LEN[:LEN2], --minimum-length LEN[:LEN2]\n",
      "                        Discard reads shorter than LEN. Default: 0\n",
      "  -M LEN[:LEN2], --maximum-length LEN[:LEN2]\n",
      "                        Discard reads longer than LEN. Default: no limit\n",
      "  --max-n COUNT         Discard reads with more than COUNT 'N' bases. If COUNT\n",
      "                        is a number between 0 and 1, it is interpreted as a\n",
      "                        fraction of the read length.\n",
      "  --max-expected-errors ERRORS, --max-ee ERRORS\n",
      "                        Discard reads whose expected number of errors (computed\n",
      "                        from quality values) exceeds ERRORS.\n",
      "  --max-average-error-rate ERROR_RATE, --max-aer ERROR_RATE\n",
      "                        as --max-expected-errors (see above), but divided by\n",
      "                        length to account for reads of varying length.\n",
      "  --discard-trimmed, --discard\n",
      "                        Discard reads that contain an adapter. Use also -O to\n",
      "                        avoid discarding too many randomly matching reads.\n",
      "  --discard-untrimmed, --trimmed-only\n",
      "                        Discard reads that do not contain an adapter.\n",
      "  --discard-casava      Discard reads that did not pass CASAVA filtering (header\n",
      "                        has :Y:).\n",
      "\n",
      "Output:\n",
      "  --quiet               Print only error messages.\n",
      "  --report {full,minimal}\n",
      "                        Which type of report to print: 'full' or 'minimal'.\n",
      "                        Default: full\n",
      "  --json FILE           Dump report in JSON format to FILE\n",
      "  -o FILE, --output FILE\n",
      "                        Write trimmed reads to FILE. FASTQ or FASTA format is\n",
      "                        chosen depending on input. Summary report is sent to\n",
      "                        standard output. Use '{name}' for demultiplexing (see\n",
      "                        docs). Default: write to standard output\n",
      "  --fasta               Output FASTA to standard output even on FASTQ input.\n",
      "  -Z                    Use compression level 1 for gzipped output files\n",
      "                        (faster, but uses more space)\n",
      "  --info-file FILE      Write information about each read and its adapter\n",
      "                        matches into FILE. See the documentation for the file\n",
      "                        format.\n",
      "  -r FILE, --rest-file FILE\n",
      "                        When the adapter matches in the middle of a read, write\n",
      "                        the rest (after the adapter) to FILE.\n",
      "  --wildcard-file FILE  When the adapter has N wildcard bases, write adapter\n",
      "                        bases matching wildcard positions to FILE. (Inaccurate\n",
      "                        with indels.)\n",
      "  --too-short-output FILE\n",
      "                        Write reads that are too short (according to length\n",
      "                        specified by -m) to FILE. Default: discard reads\n",
      "  --too-long-output FILE\n",
      "                        Write reads that are too long (according to length\n",
      "                        specified by -M) to FILE. Default: discard reads\n",
      "  --untrimmed-output FILE\n",
      "                        Write reads that do not contain any adapter to FILE.\n",
      "                        Default: output to same file as trimmed reads\n",
      "\n",
      "Paired-end options:\n",
      "  The -A/-G/-B/-U/-Q options work like their lowercase counterparts, but are\n",
      "  applied to R2 (second read in pair)\n",
      "\n",
      "  -A ADAPTER            3' adapter to be removed from R2\n",
      "  -G ADAPTER            5' adapter to be removed from R2\n",
      "  -B ADAPTER            5'/3 adapter to be removed from R2\n",
      "  -U LENGTH             Remove LENGTH bases from R2\n",
      "  -Q [5'CUTOFF,]3'CUTOFF\n",
      "                        Quality-trimming cutoff for R2. Default: same as for R1\n",
      "  -p FILE, --paired-output FILE\n",
      "                        Write R2 to FILE.\n",
      "  --pair-adapters       Treat adapters given with -a/-A etc. as pairs. Either\n",
      "                        both or none are removed from each read pair.\n",
      "  --pair-filter {any,both,first}\n",
      "                        Which of the reads in a paired-end read have to match\n",
      "                        the filtering criterion in order for the pair to be\n",
      "                        filtered. Default: any\n",
      "  --interleaved         Read and/or write interleaved paired-end reads.\n",
      "  --untrimmed-paired-output FILE\n",
      "                        Write second read in a pair to this FILE when no adapter\n",
      "                        was found. Use with --untrimmed-output. Default: output\n",
      "                        to same file as trimmed reads\n",
      "  --too-short-paired-output FILE\n",
      "                        Write second read in a pair to this file if pair is too\n",
      "                        short.\n",
      "  --too-long-paired-output FILE\n",
      "                        Write second read in a pair to this file if pair is too\n",
      "                        long.\n"
     ]
    }
   ],
   "source": [
    "! cutadapt --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51c5bf75-0af1-40c7-ba4a-1265587753b9",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "```\n",
    "cutadapt -j {threads} \n",
    "            -n 2 \n",
    "            -a \"{params.barcode3}{params.adapter3};e=0.15;o=6;anywhere;\" \n",
    "            --untrimmed-output={output.fastq_untrimmed} \n",
    "            -o - {input} 2>{output.report1} | \n",
    "cutadapt -j {threads} \n",
    "            -u 5 -u -5 \n",
    "            --rename='{{id}}_{{cut_prefix}}{{cut_suffix}} {{comment}}' \n",
    "            --max-n=0 \n",
    "            -q 15 \n",
    "            --nextseq-trim=15 \n",
    "            -m 20 \n",
    "            --too-short-output={output.fastq_tooshort} \n",
    "            -o {output.fastq_cut} - >{output.report2}\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61d94569-8947-497c-8641-59e0f790ce60",
   "metadata": {},
   "source": [
    "https://cutadapt.readthedocs.io/en/stable/guide.html"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd81c901-05a6-4b85-9dc2-2c2201af649d",
   "metadata": {},
   "source": [
    "##### paramaters:\n",
    "- -j 0, Number of CPU. Use 0 to autodetect\n",
    "- --nextseq-trim=15, is used to trim these low-quality tails of 'G's by introducing a variable quality threshold\n",
    "- --action=trim, trim adapter and downstream sequence\n",
    "- -a '{barcode3}{adapter};e=0.15;o=6;anywhere;', \n",
    "- -n 2, Remove up to COUNT adapters from each read. Default: 1\n",
    "- -u 5 -u -5, Remove LEN bases from each read. If positve, from beginning. If negative, from end\n",
    "- --max-n=0, Discard reads with more than COUNT 'N' bases (here any)\n",
    "- -q 15, Trim low-quality bases from 5' end of eachread before adapter removal.\n",
    "- -m 20, Discard reads shorter than LEN\n",
    "- --rename='{{id}}_{{cut_prefix}}{{cut_suffix}} {{comment}}', rename comment of fastq record\n",
    "- --too-short-output={fname(out_path,sample,\"fastq_tooshort\")},\n",
    "- -o {fname(out_path,sample,\"fq.gz\")}, output file\n",
    "{fname(in_path,sample,\"fq.gz\")}, input file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dbf6f13b-0a83-4436-b929-f2000deb241d",
   "metadata": {},
   "source": [
    "**NOTE:** Added `--length 30` to cutoff reads where C begin to become more common. Not in original code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "c556375c-0e45-402f-b5fe-6418606482ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Done           00:00:03     4,678,947 reads @   0.9 µs/read;  70.32 M reads/minute\n",
      "Done           00:00:05     6,830,155 reads @   0.8 µs/read;  71.94 M reads/minute\n",
      "Done           00:00:05     5,919,881 reads @   0.9 µs/read;  70.33 M reads/minute\n"
     ]
    }
   ],
   "source": [
    "for sample in samples:\n",
    "    !cutadapt -j 0 --nextseq-trim=15 --action=trim -a '{barcode3}{adapter};e=0.15;o=6;anywhere;'\\\n",
    "            -n 2 -u 5 -u -5 --max-n=0 -q 15 -m 20 -l 80\\\n",
    "            --length 30 \\\n",
    "            --rename='{{id}}_{{cut_prefix}}{{cut_suffix}} {{comment}}'\\\n",
    "            --too-short-output={fname(out_path,sample,\"fastq_tooshort\")} \\\n",
    "            -o {fname(out_path,sample,\"fq.gz\")}  \\\n",
    "            {fname(in_path,sample,\"fq.gz\")} > {fname(out_path,sample,\"log\")}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0fd27b4-6fb6-43ec-abda-908fe49df9f8",
   "metadata": {},
   "source": [
    "#### Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0deddf8f-d88b-4b1f-832b-328714f045f3",
   "metadata": {},
   "source": [
    "Take a look at a FASTQ file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "e07c75da-35fd-4c4e-b07a-3696595806f4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('samples/t1.fq.gz')"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "in_fn  = fname(in_path,treated[0],'fq.gz')\n",
    "out_fn = fname(out_path,treated[0],'fq.gz')\n",
    "in_fn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "c450ef4a-6b0d-4ded-a5f1-d387be292b48",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@SRR23538290.50 50 length=121\n",
      "GGAAATTGTGGTAATTTTAGAGTTAATATATGTTGATGGGTGTTGATTTTTTTTGATGGTATCACGAGATCGGAAGAGCACACGTCTGAACTCCAGTCACCACTCAATCGCGTATTCCGTC\n",
      "+\n",
      "FFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,FFFFF,FFF:F\n",
      "@SRR23538290.52 52 length=121\n",
      "CGTTGTTTATTGTTTGGGGGTTTTAGTTATTTTAGTGGAAGATGAATGATAAATCACGAGATCGGAAGAGCACACGTCTGAACTCCATACACCACTCAATCTCGTATGCATTCTTCTGCTT\n",
      "+\n",
      "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,FF,FFFFFFFFF::FF:F,FFFF,FF:FFFFFFFF,FFFFF::FFF,:FFFFFF:FFF:FFFFFF:,F,,F,FF,FFF:F\n",
      "@SRR23538290.65 65 length=121\n",
      "CTCAATATTGGAAGTATTGTTAAAATTTTATAATATGAAAAGAGGTTTATTAATGTAGTGCCGTATCACGAGATCGGAAGAGCACACGTCTGAACTCCAGTCACCACTCAATCTCGTATGC\n",
      "+\n",
      "FFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFF,FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,F\n",
      "@SRR23538290.120 120 length=121\n",
      "CCGCTTAAAATTAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAATTTTATACAGAAAACGGAAGACCAAACGGCAGAACACCAAACACCACACAATAGCGTAAGCCGTAAAGGGGTTGA\n",
      "+\n",
      "FFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFF,,,,,,,,,,,,:,,,:,F,,:,,,,,:,,,,,F,,,:,:,,F,F,,,:FF:,,,,,,:,,,,,::,,,,,,,:,F\n",
      "\n",
      "gzip: stdout: Broken pipe\n"
     ]
    }
   ],
   "source": [
    "! zcat {in_fn}|head -16  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "bbff0c14-bd75-4988-a235-07400fa4a89b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def show_adapter(reads):\n",
    "    for read in reads.split('\\n'):\n",
    "        read = read.replace(adapter, f'<span style=\"color: blue;\">{barcode3}{adapter}</span>')\n",
    "        display(HTML(read))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b131dc87-f706-4aa6-b82d-8fec51b0b7c2",
   "metadata": {},
   "source": [
    "Look for adapter in untrimmed reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "c6560944-596a-443c-9f53-d07613ceedbd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "GGAAATTGTGGTAATTTTAGAGTTAATATATGTTGATGGGTGTTGATTTTTTTTGATGGTATCACG<span style=\"color: blue;\">ATCACGAGATCGGAAGAGCACACGTCT</span>GAACTCCAGTCACCACTCAATCGCGTATTCCGTC"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "CGTTGTTTATTGTTTGGGGGTTTTAGTTATTTTAGTGGAAGATGAATGATAAATCACG<span style=\"color: blue;\">ATCACGAGATCGGAAGAGCACACGTCT</span>GAACTCCATACACCACTCAATCTCGTATGCATTCTTCTGCTT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "CTCAATATTGGAAGTATTGTTAAAATTTTATAATATGAAAAGAGGTTTATTAATGTAGTGCCGTATCACG<span style=\"color: blue;\">ATCACGAGATCGGAAGAGCACACGTCT</span>GAACTCCAGTCACCACTCAATCTCGTATGC"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "CCGCTTAAAATTAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAATTTTATACAGAAAACGGAAGACCAAACGGCAGAACACCAAACACCACACAATAGCGTAAGCCGTAAAGGGGTTGA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTTTTTTGGATATTAGGAAAAAATTTTGTAGAGAGAGTAAAAAATTTACTAGTATCACG<span style=\"color: blue;\">ATCACGAGATCGGAAGAGCACACGTCT</span>GAACTCCAGTCACCACTCAATCTCGTATGCCGTCTTCTGCT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "ATTAATAATGAAGTATAAGGAGTATGTGTAGAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "ATGGATTATATGTGGGAGTGGAGAATTAAAATTTAGGTATAGATAGATTGCATCTATCACG<span style=\"color: blue;\">ATCACGAGATCGGAAGAGCACACGTCT</span>GAACTCCAGTCACCACTCAATCTCGTATTCCGTCTTCTG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "GGCTATGTTGAAAATGAGATTAAGATAGTGTAATCACG<span style=\"color: blue;\">ATCACGAGATCGGAAGAGCACACGTCT</span>GAACTCCAGTCACCACTCAATCTCGTATGCCGTCTTCTGCTTTAAAAGCTGGGGGGGGGGGG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "GGATTTTGTAGGTAGAATATTTGAAGGAGAAGTTTATTAGTTAGGTTTAGGAAACCTAATCACG<span style=\"color: blue;\">ATCACGAGATCGGAAGAGCACACGTCT</span>GAACTCCAGTCACCACTCAATCTCGTATGCCGTCTT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "reads = ! zcat {in_fn} | head -36  | seqtk seq -A |grep -v '>'\n",
    "reads = ('\\n').join(reads)\n",
    "show_adapter(reads)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fad1e27a-a3a6-4bfe-9920-203e65654575",
   "metadata": {},
   "source": [
    "Verify that adapters and all downstream elements of reads have been trimmed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "65a7f174-35a8-4956-bdf3-eb86888a902a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "TTGTGGTAATTTTAGAGTTAATATATGTTG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTTATTGTTTGGGGGTTTTAGTTATTTTAG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TATTGGAAGTATTGTTAAAATTTTATAATA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TAAAATTAAAAAAAAAAAAAAAAAAAAAAA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTGGATATTAGGAAAAAATTTTGTAGAGAG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TAATGAAGTATAAGGAGTATGTGTAGAAAA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTATATGTGGGAGTGGAGAATTAAAATTTA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TGTTGAAAATGAGATTAAGATAGTGTA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTGTAGGTAGAATATTTGAAGGAGAAGTTT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "ATTTTTATTTAGGTTTGATTTTTGGGAAAG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "AAATTAAGTAAATAAGTTTAAATGTTTTTT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "AATGATTTTTGTTGGTATTAAGAAGAAGGA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "AATTATAGTAATGTTTAGTTTATTGAGTTA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTTGTGGGTATTGATTGTTTGGTTTTGGAT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TATTTTTGTTGTTATGGTAAAGTTTGTTGT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "AAATGAGTATAGTTAGTGTTGTGGTAATGT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTAAAAAGATATTTGATATTGATGAAGTTG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "AATGTTTTGGAGTAGAAGTAGAAGGTTTTT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTTAGTTTTTTGGTTTGATTAGTGTTATGA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "AGAGTTTTAATGTTTTTAATGTAGGTGATG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTAAATTTTGGAAATAGTTTATAAATATAG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TAAAATTAATTTGAAATTTAATCATAAAGA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTAGGGAATTATGTTATTGTTATTTTTTAT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTATTTTGGTTATTGATATGGGTATTATGT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "ATTTAAGATTTTGTTGAGGAGTTTTTGTTG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTTATAGAGGGTGAAATAGGAGTTTTTGGA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "GGTGTTTTATGTTGGGGAGGATGATGAGGA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "GTAATGAAGGGTAAGATGAAGTGAAAGGTA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTATGAGATGATTTTTAATAATGAGATTTT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TGTTGGTAGAGGTCTATTAAGTGAGGTTTT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "GAGATTGTTAAAAAGAGAATTAAGAAGTTT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "GAAATTTTTTTGGTTAGTGAGAGAGATAAA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TGGTATTTGGGAATAAAGTTAGTTGTTAGT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTGGTAAAGAATTTAGTATAATATTAAATG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTGGGTTTAGTGTTATGGTTTTGGTTTGTA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTTGTTTGTTTTGGTTTTTGAAAGTGTTGG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "AGAAAGTTTTTAGTAATAGGTTTATTATGT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTAGAGAGATTAGTTGTTTTATGGGATGTG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "GAGTTTTGAAATTATAGATGATAGTTTATG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "GATATTAAAGAGATATGGTTTGGGTTTAGG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "ATTTTGGTTAGTTGAGTATTATTAGGATAC"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "AAAATAATGTGGTATTTATGAGTTATTTTT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "GTTTTATTGTTTAGGTGATTATTTTGAATT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "AGGGTTTGTTTTGTGGAGAGGTTGTTGTTT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TGGAGTTTTTAGATAAATATTTCGATAGTG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "AAATGAGTAGTTAGGAAGTTTAGGGGAGGG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTTTTGAATATTTTTGGTTTATTTTTAGAA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTTGATTTGTTTGGTGATTGAATTATTAGT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTAGATTTGTTAGAGTGGGTTTATTTGGTT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTGTTATGGGTTGTTGTTTTGTTTGTTGTT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "GTGTTGAGTGGTAGTAGTGGGAGTTTTTTT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "GAGGTGGTTGAGTTTTTTTTTAGGATATTG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TTAGAGATGAAAAATGAGATTATTTGAAGG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "GTTTGGTTGATATTTTGAGTTTTTTTTGCA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TAAATTTATTGGAAATATGGTATTGTTGTT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TGTTGATATGGTTGTATAGGATTGTTTGAA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "AGATTAGTTTGAAGAATTATGATTTTTAGG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TAAGGAGGTGGTTATTGTTATTTGTGGGGT"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TGTATTTGTTTATATAGTTATGTGAATAGA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "GAATGTAAGGAAGTAAAGTAGGAGAAAGAA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "TATTGATGAAGTTTGTTTTAAGAAGTGTGG"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "GGGAATTTGTTATAAGTGTAGTGTATTCC"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "AGAAATTGTTGAAGAAGAAGAATTAGAGGA"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "reads = !zcat {out_fn}| head -256  | seqtk seq -A |grep -v '>'\n",
    "reads = ('\\n').join(reads)\n",
    "show_adapter(reads)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b28de976-ebee-4727-a44a-a3379cd8757e",
   "metadata": {},
   "source": [
    "What did `--rename='{{id}}_{{cut_prefix}}{{cut_suffix}} {{comment}}'` do?\n",
    "\n",
    "Looks like "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "183946b6-0828-455e-a317-a2a37261c6b9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@SRR23538290.50 50 length=121\n",
      "GGAAATTGTGGTAATTTTAGAGTTAATATATGTTGATGGGTGTTGATTTTTTTTGATGGTATCACGAGATCGGAAGAGCACACGTCTGAACTCCAGTCACCACTCAATCGCGTATTCCGTC\n",
      "\n",
      "gzip: stdout: Broken pipe\n"
     ]
    }
   ],
   "source": [
    "!zcat {in_fn}| head -2 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "97272d2b-9170-4935-8801-907e7ffcb1f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@SRR23538290.50_GGAAACCGTC 50 length=121\n",
      "TTGTGGTAATTTTAGAGTTAATATATGTTG\n",
      "\n",
      "gzip: stdout: Broken pipe\n"
     ]
    }
   ],
   "source": [
    "!zcat {out_fn}| head -2 "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22988563-df9e-471e-ad5c-e2b34d86f94f",
   "metadata": {},
   "source": [
    "Verify that adapters and all downstream elements of reads have been trimmed"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47d80e80-4f86-421a-bd13-f1b66cb48c48",
   "metadata": {},
   "source": [
    "No reads should have been deleted with cutadapt. Veryify that number of reads before and after cutadapt are the same."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "e03f142f-9d37-4d6c-965d-c17053979e68",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[4678947, 6830155, 5919881]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Untrimmed\n",
    "files = [fname(in_path,sample, \"fq.gz\") for sample in samples]\n",
    "ins = [nseqs(f) for f in files]\n",
    "ins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "172a1fac-5cf4-4d31-a7d1-25cc250689f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[4652729, 6780142, 5876778]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Adapter Trimmed Reads\n",
    "files = [fname(out_path,sample, \"fq.gz\") for sample in samples]\n",
    "ins = [nseqs(f) for f in files]\n",
    "ins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "f0a6990d-1e30-439d-b9dc-bba695c9d82c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[26154, 49936, 43004]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Too Short Reads\n",
    "files = [fname(out_path,sample, \"fastq_tooshort\") for sample in samples]\n",
    "ins = [nseqs(f) for f in files]\n",
    "ins"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba63237d-b9da-4e19-9b66-0c523efadc17",
   "metadata": {},
   "source": [
    "Compare the sum of read lengths in the original and adapter trimmed files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "9004da50-f983-496a-8ec4-7e8a45047d22",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[570831534, 833278910, 722225482]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "files = [fname(in_path,sample, \"fq.gz\") for sample in samples]\n",
    "res = []\n",
    "for f in files:\n",
    "    n = !seqtk seq -A {f}|grep -v '>'|wc -c\n",
    "    res.append(int(n[0]))\n",
    "ins = res\n",
    "ins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "fb5d33e9-c511-41f2-b8d4-3f7e606fa4ad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[143737536, 209183737, 181339895]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "files = [fname(out_path,sample, \"fq.gz\") for sample in samples]\n",
    "res = []\n",
    "for f in files:\n",
    "    n = !seqtk seq -A {f}|grep -v '>'|wc -c\n",
    "    res.append(int(n[0]))\n",
    "outs = res\n",
    "outs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "34da7817",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAioAAAGxCAYAAABMeZ2uAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8g+/7EAAAACXBIWXMAAA9hAAAPYQGoP6dpAABBvklEQVR4nO3dd3gU5f7+8XtJ2TQSWiCUEEDpVQ1gQAi9CAiiyNdCExQVpckREAIEA6HY9aDCoYo0PYAFBESaBTB0JHQSCNKkmFCTkMzvD3/Z45IAu5Bkh+T9uq69LubZZ2Y+u5kld56ZZ9ZiGIYhAAAAEyrg6gIAAABuhqACAABMi6ACAABMi6ACAABMi6ACAABMi6ACAABMi6ACAABMi6ACAABMi6ACAABMi6AC09u8ebMef/xxlS1bVlarVSVKlFBYWJhef/11V5d2x0aOHKmyZcvK3d1dhQoVyvR8fHy8LBaLQ4/4+Phb7uvKlSsaM2aM1q1bd8f1ZtQza9Ysh/plPAoUKKDChQurefPmWrVq1R3v/26VK1dOPXv2vOvtrFu3zu71ubm5qUSJEurSpYv27t1794U64F7+PPTs2VPlypVzdRm4x7i7ugDgVpYtW6bHHntMTZo00aRJk1SyZEmdPHlSW7Zs0YIFC/TOO++4ukSnff311xo3bpxGjBihtm3bymq1ZupTsmRJbdy40a7tlVdeUWJior744otMfW/lypUrioyMlCQ1adLk7op30GuvvaZnnnlGaWlp2rdvnyIjI/Xoo49qzZo1aty4ca7UkJPGjx+vpk2bKiUlRVu2bNHYsWP1448/avfu3SpdunSO7Tcvfh6A2yGowNQmTZqk8uXLa+XKlXJ3/9/h+n//93+aNGmSCyu7c7///rskqX///ipevHiWfaxWqx5++GG7Nn9/f6WkpGRqN6OyZcva6mzYsKEqVqyo8PBwTZ8+PU8ElYoVK9peX+PGjVWoUCH17t1bs2bN0ogRI+5q21euXJGPj0+Wz+XFzwNwO5z6gamdO3dOxYoVs/tPOUOBAvaHr8Vi0ZgxYzL1u3HYf9asWbJYLFqzZo1eeOEFFS1aVP7+/urevbsuX76sU6dO6amnnlKhQoVUsmRJDRkyRKmpqbetNT09XZMmTVKVKlVktVpVvHhxde/eXcePH7erZeTIkZKkEiVK3LRmRx07dkzPPfecihcvLqvVqqpVq+qdd95Renq6pL9PxQQGBkqSIiMjbacsMt6PQ4cOqVevXqpYsaJ8fHxUunRpdejQQbt3777jmrISGhoqSTp9+rRd+6lTp9S3b1+VKVNGnp6eKl++vCIjI3X9+nW7fpGRkapfv76KFCkif39/Pfjgg5o+fbpu/E7V1NRUvfHGGwoKCpKPj48eeeQR/fbbb5nquXLlioYMGaLy5cvLy8tLRYoUUWhoqObPn39Hry8jtBw9etTWtnDhQoWFhcnX11d+fn5q3bq1tm/fbrdez5495efnp927d6tVq1YqWLCgmjdvftP9OPN5WLhwoVq1aqWSJUvK29tbVatW1bBhw3T58uUsa9i3b59at24tX19flSxZUhMmTJAkbdq0SY888oh8fX1VqVIlzZ492279jM/TDz/8oF69eqlIkSLy9fVVhw4ddOTIkdu+d4ZhaMqUKapTp468vb1VuHBhPfnkkw6ti/yBoAJTCwsL0+bNm9W/f39t3rzZocDgqD59+iggIEALFizQyJEjNW/ePL3wwgtq166dateura+++ko9evTQO++8o48++ui223v55Zc1dOhQtWzZUt98843eeustrVixQg0aNNDZs2clSUuWLFHv3r0lSStWrNDGjRvVp0+fO6r/zz//VIMGDbRq1Sq99dZb+uabb9SiRQsNGTJEr776qqS/TwutWLFCktS7d29t3LhRGzduVEREhCTpxIkTKlq0qCZMmKAVK1bo3//+t9zd3VW/fn3t37//jurKSlxcnCSpUqVKtrZTp06pXr16WrlypUaNGqXvv/9evXv3VnR0tF544QW79ePj49W3b18tWrRIixcvVufOnfXaa6/prbfesuv3wgsv6O2331b37t319ddf64knnlDnzp114cIFu36DBw/WJ598ov79+2vFihX6/PPP1aVLF507d+6OXt+hQ4ckyRYKx48fr6efflrVqlXTokWL9Pnnn+vixYtq1KiRYmNj7dZNSUnRY489pmbNmunrr7+2nabLijOfh4MHD+rRRx/V9OnTtWLFCg0cOFCLFi1Shw4dMvVNTU1V586d1a5dO3399ddq27athg8frjfffFM9evTQ888/ryVLlqhy5crq2bOntm7dmmkbvXv3VoECBTRv3jy9//77+u2339SkSRP99ddft3zv+vbtq4EDB6pFixZaunSppkyZoj179qhBgwaZgi3yKQMwsbNnzxqPPPKIIcmQZHh4eBgNGjQwoqOjjYsXL9r1lWSMHj060zZCQkKMHj162JZnzpxpSDJee+01u36dOnUyJBnvvvuuXXudOnWMBx988JZ17t2715BkvPLKK3btmzdvNiQZb775pq1t9OjRhiTjzz//vOU2bxQeHm5Ur17dtjxs2DBDkrF582a7fi+//LJhsViM/fv3G4ZhGH/++edN35sbXb9+3UhJSTEqVqxoDBo0yNYeFxdnSDJmzpx5y/Uz+k2cONFITU01rl27ZuzYscMICwszSpYsacTFxdn69u3b1/Dz8zOOHj1qt423337bkGTs2bMny32kpaUZqampxtixY42iRYsa6enphmH872fwz7oNwzC++OILQ5LdMVCjRg2jU6dOt30/brR27VpDkrFw4UIjNTXVuHLlirFhwwbj/vvvN9zc3IydO3cax44dM9zd3TMdXxcvXjSCgoKMp556ytbWo0cPQ5IxY8YMh/bvzOfhn9LT043U1FRj/fr1hiRj586dmWr473//a2tLTU01AgMDDUnGtm3bbO3nzp0z3NzcjMGDB9vaMj5Pjz/+uN0+f/nlF0OSERUVZbevkJAQ2/LGjRsNScY777xjt25CQoLh7e1tvPHGGw69L8jb8syIyoYNG9ShQweVKlVKFotFS5cudXobK1eu1MMPP6yCBQsqMDBQTzzxhO0vQbhG0aJF9dNPPykmJkYTJkxQx44ddeDAAQ0fPlw1a9a0jVTcifbt29stV61aVZLUrl27TO3/HNLPytq1ayUp08ySevXqqWrVqvrxxx/vuM6bWbNmjapVq6Z69erZtffs2VOGYWjNmjW33cb169c1fvx4VatWTZ6ennJ3d5enp6cOHjx4V7NYhg4dKg8PD3l5ealOnTr6/fff9e2339rN+Pjuu+/UtGlTlSpVStevX7c92rZtK0lav3693Wtt0aKFAgIC5ObmJg8PD40aNUrnzp3TmTNnJP3vZ/Dss8/a1fLUU09lOlVSr149ff/99xo2bJjWrVunq1evOvX6unbtKg8PD/n4+Khx48ZKS0vTV199pVq1amnlypW6fv26unfvbve6vLy8FB4enuXsqyeeeMKh/TrzeThy5IieeeYZBQUF2d6z8PBwScr0s7VYLHr00Udty+7u7rr//vtVsmRJPfDAA7b2IkWKqHjx4ll+Hm583xs0aKCQkBDbzyUr3333nSwWi5577jm79yooKEi1a9e+q5lqyDvyTFC5fPmyateurY8//viO1j9y5Ig6duyoZs2aaceOHVq5cqXOnj2rzp07Z3OluBOhoaEaOnSovvzyS504cUKDBg1SfHz8XV1AWKRIEbtlT0/Pm7Zfu3btltvKOGWQ1QycUqVK3fEphdvt82b7+2dNtzJ48GBFRESoU6dO+vbbb7V582bFxMSodu3aTv/y/qcBAwYoJiZGP//8s95++22lpqaqY8eOdjWdPn1a3377rTw8POwe1atXlyTbL93ffvtNrVq1kiRNmzZNv/zyi2JiYmwXrWbUmbHtoKAgu1rc3d1VtGhRu7YPP/xQQ4cO1dKlS9W0aVMVKVJEnTp10sGDBx16fRMnTlRMTIy2bdumY8eO6ciRI+rUqZPtdUlS3bp1M722hQsXZgrXPj4+8vf3d2i/GW73ebh06ZIaNWqkzZs3KyoqSuvWrVNMTIwWL14sSZl+tj4+PvLy8rJr8/T0zPRZyGjP6vNw4/ue0Xar4/D06dMyDEMlSpTI9F5t2rTprv4QQd6RZ2b9tG3b1vaXWFZSUlI0cuRIffHFF/rrr79Uo0YNTZw40TZdc9u2bUpLS1NUVJTtorQhQ4aoY8eOSk1NlYeHR268DDjAw8NDo0eP1nvvvWebQSP9PVMmOTk5U/+cCAk3yvhFePLkSZUpU8buuRMnTqhYsWI5ss+TJ09maj9x4oQkObTPuXPnqnv37ho/frxd+9mzZ7O8v4ujypQpY7uAtmHDhgoKCtJzzz2n0aNH2/6YKFasmGrVqqVx48ZluY2MwLVgwQJ5eHjou+++s/tleuOoacbP4NSpU3ZThK9fv57pGPD19VVkZKQiIyN1+vRp2+hKhw4dtG/fvtu+vgoVKthe340y3vevvvpKISEht92WxWK5bZ9byerzsGbNGp04cULr1q2zjaJIuu31Infj1KlTWbbdf//9N12nWLFislgs+umnn7Kcpp9VG/KfPDOicju9evXSL7/8ogULFmjXrl3q0qWL2rRpY/sLKjQ0VG5ubpo5c6bS0tKUmJiozz//XK1atSKkuFBWv4il/w1dZ/wyk/6eUbNr1y67fmvWrNGlS5dyrsD/r1mzZpL+/sX/TzExMdq7d+8tZ3LcqebNmys2Nlbbtm2za58zZ44sFouaNm0q6X//2Wc1QmKxWDL9Mli2bJn++OOPbK312WefVZMmTTRt2jTbaYP27dvr999/13333afQ0NBMj4yfrcVikbu7u9zc3Gzbu3r1qj7//HO7fWT80XHjfWYWLVqUaRbRP5UoUUI9e/bU008/rf379+vKlSt39Vpbt24td3d3HT58OMvXdbOA4whHPw8Z4efGn+1nn312x/u+nRvf919//VVHjx695b172rdvL8Mw9Mcff2T5PtWsWTPH6sW9I8+MqNzK4cOHNX/+fB0/ftz2QR4yZIhWrFihmTNnavz48SpXrpxWrVqlLl26qG/fvkpLS1NYWJiWL1/u4urzt9atW6tMmTLq0KGDqlSpovT0dO3YsUPvvPOO/Pz8NGDAAFvfbt26KSIiQqNGjVJ4eLhiY2P18ccfKyAgIMfrrFy5sl588UV99NFHKlCggNq2bav4+HhFREQoODhYgwYNyvZ9Dho0SHPmzFG7du00duxYhYSEaNmyZZoyZYpefvll2wybggULKiQkRF9//bWaN2+uIkWKqFixYipXrpzat2+vWbNmqUqVKqpVq5a2bt2qyZMnZxoVyg4TJ05U/fr19dZbb+k///mPxo4dqx9++EENGjRQ//79VblyZV27dk3x8fFavny5Pv30U5UpU0bt2rXTu+++q2eeeUYvvviizp07p7fffjvTL+GqVavqueee0/vvvy8PDw+1aNFCv//+u95+++1Mp1bq16+v9u3bq1atWipcuLD27t2rzz//XGFhYTe9h4mjypUrp7Fjx2rEiBE6cuSI2rRpo8KFC+v06dP67bffbKM5d8LRz0ODBg1UuHBhvfTSSxo9erQ8PDz0xRdfaOfOnXf12m5ly5Yt6tOnj7p06aKEhASNGDFCpUuX1iuvvHLTdRo2bKgXX3xRvXr10pYtW9S4cWP5+vrq5MmT+vnnn1WzZk29/PLLOVYz7hEuvpg3R0gylixZYltetGiRIcnw9fW1e7i7u9uuwD958qRRsWJF41//+pexbds2Y/369UZ4eLjRvHlz26wC5L6FCxcazzzzjFGxYkXDz8/P8PDwMMqWLWt069bNiI2NteubnJxsvPHGG0ZwcLDh7e1thIeHGzt27LjprJ+YmBi79W82G6dHjx6Gr6/vbWtNS0szJk6caFSqVMnw8PAwihUrZjz33HNGQkKCQ/u5nRtn/RiGYRw9etR45plnjKJFixoeHh5G5cqVjcmTJxtpaWl2/VavXm088MADhtVqtZsBc+HCBaN3795G8eLFDR8fH+ORRx4xfvrpJyM8PNwIDw+3re/srJ/Jkydn+XyXLl0Md3d349ChQ4Zh/D0jqX///kb58uUNDw8Po0iRIsZDDz1kjBgxwrh06ZJtvRkzZhiVK1c2rFarUaFCBSM6OtqYPn26IcluJlFycrLx+uuvG8WLFze8vLyMhx9+2Ni4cWOmY2DYsGFGaGioUbhwYds2Bw0aZJw9e/aWry9j1s+XX355y36GYRhLly41mjZtavj7+xtWq9UICQkxnnzySWP16tW2Po4eWxmc+Tz8+uuvRlhYmOHj42MEBgYaffr0MbZt25bp53izGrI63gzj71l07dq1sy1nfJ5WrVpldOvWzShUqJDh7e1tPProo8bBgwft1r1x1k+GGTNmGPXr1zd8fX0Nb29v47777jO6d+9ubNmyxeH3BnmXxTBuuGNSHmCxWLRkyRLbxW0LFy7Us88+qz179tgNH0uSn5+fgoKCFBERoe+//15btmyxPXf8+HEFBwdr48aN98TdQAEgt82aNUu9evVSTEzMXZ3WAm4mX5z6eeCBB5SWlqYzZ86oUaNGWfa5cuVKphCTsZxxl08AAJC78szFtJcuXdKOHTu0Y8cOSX/fCXPHjh06duyYKlWqpGeffVbdu3fX4sWLFRcXp5iYGE2cONF2DUq7du0UExOjsWPH6uDBg9q2bZt69eqlkJAQu/sIAACA3JNnTv2sW7fONsvhn3r06KFZs2YpNTVVUVFRmjNnjv744w8VLVpUYWFhioyMtF1ZvmDBAk2aNEkHDhyQj4+PwsLCNHHiRFWpUiW3Xw4AAFAeCioAACDvyTOnfgAAQN5DUAEAAKZ1T8/6SU9P14kTJ1SwYMG7vg01AADIHYZh6OLFiypVqpTta2tu5p4OKidOnFBwcLCrywAAAHcgISHhtnfCvqeDSsGCBSX9/UKd/fZRAADgGklJSQoODrb9Hr+VezqoZJzu8ff3J6gAAHCPceSyDS6mBQAApkVQAQAApkVQAQAApnVPX6MCAMg/0tLSlJqa6uoy4AA3Nze5u7tny61DCCoAANO7dOmSjh8/Lr715d7h4+OjkiVLytPT8662Q1ABAJhaWlqajh8/Lh8fHwUGBnKDT5MzDEMpKSn6888/FRcXp4oVK972pm63QlABAJhaamqqDMNQYGCgvL29XV0OHODt7S0PDw8dPXpUKSkp8vLyuuNtcTEtAOCewEjKveVuRlHstpMtWwEAAMgBBBUAAGBaXKMCALgnlRu2LFf3Fz+hXe7uLz5e5cuX1/bt21WnTp1c3beZMKICAEAOSkhIUO/evVWqVCl5enoqJCREAwYM0Llz5265XnBwsE6ePKkaNWrkUqXmRFABACCHHDlyRKGhoTpw4IDmz5+vQ4cO6dNPP9WPP/6osLAwnT9/Psv1UlJS5ObmpqCgILm75++THwQVAABySL9+/eTp6alVq1YpPDxcZcuWVdu2bbV69Wr98ccfGjFihCSpXLlyioqKUs+ePRUQEKAXXnhB8fHxslgs2rFjh21733zzjSpWrChvb281bdpUs2fPlsVi0V9//SVJOnr0qDp06KDChQvL19dX1atX1/Lly13wyrNP/o5pwD0qt8/Nu0puXxMAZKfz589r5cqVGjduXKb7vwQFBenZZ5/VwoULNWXKFEnS5MmTFRERoZEjR2a5vfj4eD355JMaMGCA+vTpo+3bt2vIkCF2ffr166eUlBRt2LBBvr6+io2NlZ+fX868wFxCUAEAIAccPHhQhmGoatWqWT5ftWpVXbhwQX/++ackqVmzZnbBIz4+3q7/p59+qsqVK2vy5MmSpMqVK+v333/XuHHjbH2OHTumJ554QjVr1pQkVahQITtfkktw6gcAABfI+N6ijBvZhYaG3rL//v37VbduXbu2evXq2S33799fUVFRatiwoUaPHq1du3ZlY8WuQVABACAH3H///bJYLIqNjc3y+X379qlw4cIqVqyYJMnX1/eW2zMMI9PdeW/8ksY+ffroyJEj6tatm3bv3q3Q0FB99NFHd/EqXI+gAgBADihatKhatmypKVOm6OrVq3bPnTp1Sl988YW6du3q8FcDVKlSRTExMXZtW7ZsydQvODhYL730khYvXqzXX39d06ZNu/MXYQIEFQAAcsjHH3+s5ORktW7dWhs2bFBCQoJWrFihli1bqnTp0nbXl9xO3759tW/fPg0dOlQHDhzQokWLNGvWLEn/O300cOBArVy5UnFxcdq2bZvWrFlz02tk7hVcTAsAuCfdC7PCKlasqC1btmjMmDHq2rWrzp07p6CgIHXq1EmjR49WkSJFHN5W+fLl9dVXX+n111/XBx98oLCwMI0YMUIvv/yyrFarJCktLU39+vXT8ePH5e/vrzZt2ui9997LqZeXKwgqAADkoJCQEM2cOfOWfW6c4SP9fW+VG69Beeyxx/TYY4/ZlseNG6cyZcrIy8tLku7561GyQlABAOAeMWXKFNWtW1dFixbVL7/8osmTJ+vVV191dVk5iqACAMA94uDBg4qKitL58+dVtmxZvf766xo+fLiry8pRLr2Y9vr16xo5cqTKly8vb29vVahQQWPHjlV6erorywIAwJTee+89nThxQteuXdOBAwcUERGR578LyKWvbuLEifr00081e/ZsVa9eXVu2bFGvXr0UEBCgAQMGuLI0AABgAi4NKhs3blTHjh3Vrt3fV26XK1dO8+fPz3JeOAAAyH9ceurnkUce0Y8//qgDBw5Iknbu3Kmff/5Zjz76aJb9k5OTlZSUZPcAAAB5l0tHVIYOHarExERVqVJFbm5uSktL07hx4/T0009n2T86OlqRkZG5XCUAAHAVl46oLFy4UHPnztW8efO0bds2zZ49W2+//bZmz56dZf/hw4crMTHR9khISMjligEAQG5y6YjKv/71Lw0bNkz/93//J0mqWbOmjh49qujoaPXo0SNTf6vVarv7HgAAyPtcGlSuXLmiAgXsB3Xc3NyYngwAuL0xAbm8v8Qc3fysWbM0cOBA/fXXXzm6n+xksVi0ZMkSderUKcf24dJTPx06dNC4ceO0bNkyxcfHa8mSJXr33Xf1+OOPu7IsAADuisViueWjZ8+emdbp2rWrbXIJ/selIyofffSRIiIi9Morr+jMmTMqVaqU+vbtq1GjRrmyLAAA7srJkydt/164cKFGjRql/fv329q8vb3t+qempsrb2ztTO1w8olKwYEG9//77Onr0qK5evarDhw8rKipKnp6eriwLAIC7EhQUZHsEBATIYrHYlq9du6ZChQpp0aJFatKkiby8vDR37lzNmjVLhQoVsm1jzJgxqlOnjmbMmKGyZcvKz89PL7/8stLS0jRp0iQFBQWpePHiGjdunN2+LRaLPvvsM7Vv314+Pj6qWrWqNm7cqEOHDqlJkyby9fVVWFiYDh8+bLfet99+q4ceekheXl6qUKGCIiMjdf36ddvzBw8eVOPGjeXl5aVq1arphx9+yNH3MINLgwoAAPnV0KFD1b9/f+3du1etW7fOss/hw4f1/fffa8WKFZo/f75mzJihdu3a6fjx41q/fr0mTpyokSNHatOmTXbrvfXWW+revbt27NihKlWq6JlnnlHfvn01fPhw201V//llhitXrtRzzz2n/v37KzY2Vp999plmzZplC0Hp6enq3Lmz3NzctGnTJn366acaOnRoDr0z9vL2FwQAAGBSAwcOVOfOnW/ZJz09XTNmzFDBggVVrVo1NW3aVPv379fy5ctVoEABVa5cWRMnTtS6dev08MMP29br1auXnnrqKUl/B6KwsDBFRETYAtGAAQPUq1cvW/9x48Zp2LBhthm3FSpU0FtvvaU33nhDo0eP1urVq7V3717Fx8erTJkykqTx48erbdu22fqeZIWgAgCAC4SGht62T7ly5VSwYEHbcokSJeTm5mY3Y7ZEiRI6c+aM3Xq1atWye176+xYg/2y7du2akpKS5O/vr61btyomJsbuNFJaWpquXbumK1euaO/evSpbtqwtpEhSWFiYE6/2zhFUAABwAV9f39v28fDwsFu2WCxZtt14W49/9rFYLDdty1gvPT1dkZGRWY7weHl5yTCMTO0Z28hpBBUAAPK5Bx98UPv379f999+f5fPVqlXTsWPHdOLECZUqVUrS318snBsIKgAA5HOjRo1S+/btFRwcrC5duqhAgQLatWuXdu/eraioKLVo0UKVK1dW9+7d9c477ygpKUkjRozIldoIKgCAe1MO3yk2P2ndurW+++47jR07VpMmTZKHh4eqVKmiPn36SJIKFCigJUuWqHfv3qpXr57KlSunDz/8UG3atMnx2ixGViee7hFJSUkKCAhQYmKi/P39XV0OkGvKDVvm6hJyRfyEdq4uASZw7do1xcXFqXz58vLy8nJ1OXDQrX5uzvz+5j4qAADAtAgqAADAtAgqAADAtAgqAADAtAgqAIB7wj089yNfyq6fF0EFAGBqbm5ukqSUlBQXVwJnXLlyRVLmu+s6i/uoAABMzd3dXT4+Pvrzzz/l4eFh9z03MB/DMHTlyhWdOXNGhQoVsgXNO0VQAQCYmsViUcmSJRUXF6ejR4+6uhw4qFChQgoKCrrr7RBUAACm5+npqYoVK3L65x7h4eFx1yMpGQgqAIB7QoECBbgzbT7EiT4AAGBaBBUAAGBaBBUAAGBaBBUAAGBaBBUAAGBaBBUAAGBaBBUAAGBaBBUAAGBa3PANAJBtyg1b5uoSckX8hHauLiHfYEQFAACYFkEFAACYFkEFAACYFkEFAACYFkEFAACYlkuDSrly5WSxWDI9+vXr58qyAACASbh0enJMTIzS0tJsy7///rtatmypLl26uLAqAABgFi4NKoGBgXbLEyZM0H333afw8HAXVQQAAMzENDd8S0lJ0dy5czV48GBZLJYs+yQnJys5Odm2nJSUlFvlAQAAFzDNxbRLly7VX3/9pZ49e960T3R0tAICAmyP4ODg3CsQAADkOtMElenTp6tt27YqVarUTfsMHz5ciYmJtkdCQkIuVggAAHKbKU79HD16VKtXr9bixYtv2c9qtcpqteZSVQAAwNVMMaIyc+ZMFS9eXO3a8SVPAADgf1weVNLT0zVz5kz16NFD7u6mGOABAAAm4fKgsnr1ah07dkzPP/+8q0sBAAAm4/IhjFatWskwDFeXAQAATMjlIyoAAAA3Q1ABAACmRVABAACmRVABAACmRVABAACmRVABAACmRVABAACmRVABAACmRVABAACmRVABAACmRVABAACmRVABAACmRVABAACmdddBJS0tTTt27NCFCxeyox4AAAAbp4PKwIEDNX36dEl/h5Tw8HA9+OCDCg4O1rp167K7PgAAkI85HVS++uor1a5dW5L07bffKi4uTvv27dPAgQM1YsSIbC8QAADkX04HlbNnzyooKEiStHz5cnXp0kWVKlVS7969tXv37mwvEAAA5F9OB5USJUooNjZWaWlpWrFihVq0aCFJunLlitzc3LK9QAAAkH+5O7tCr1699NRTT6lkyZKyWCxq2bKlJGnz5s2qUqVKthcIAADyL6eDypgxY1SjRg0lJCSoS5cuslqtkiQ3NzcNGzYs2wsEAAD5l9NBRZKefPLJTG09evS462IAAAD+6Y6Cyo8//qgff/xRZ86cUXp6ut1zM2bMyJbCAAAAnA4qkZGRGjt2rEJDQ23XqQAAAOQEp4PKp59+qlmzZqlbt245UQ8AAICN09OTU1JS1KBBg5yoBQAAwI7TQaVPnz6aN29eTtQCAABgx6FTP4MHD7b9Oz09XVOnTtXq1atVq1YteXh42PV99913s7dCAACQbzkUVLZv3263XKdOHUnS77//nu0FAQAAZHAoqKxduzan6wAAAMjE6WtUnn/+eV28eDFT++XLl/X8889nS1EAAADSHQSV2bNn6+rVq5nar169qjlz5mRLUQAAAJITQSUpKUmJiYkyDEMXL15UUlKS7XHhwgUtX75cxYsXd7qAP/74Q88995yKFi0qHx8f1alTR1u3bnV6OwAAIO9x+IZvhQoVksVikcViUaVKlTI9b7FYFBkZ6dTOL1y4oIYNG6pp06b6/vvvVbx4cR0+fFiFChVyajsAACBvcjiorF27VoZhqFmzZvrvf/+rIkWK2J7z9PRUSEiISpUq5dTOJ06cqODgYM2cOdPWVq5cOae2AQAA8i6Hg0p4eLgkKS4uTmXLls2W7/j55ptv1Lp1a3Xp0kXr169X6dKl9corr+iFF17Isn9ycrKSk5Nty0lJSXddAwAAMC+nv+snMTFRu3fvztRusVjk5eWlsmXLymq1OrStI0eO6JNPPtHgwYP15ptv6rffflP//v1ltVrVvXv3TP2jo6OdPr0EAADuXRbDMAxnVihQoMAtR1M8PDzUtWtXffbZZ/Ly8rrltjw9PRUaGqpff/3V1ta/f3/FxMRo48aNmfpnNaISHBysxMRE+fv7O/MygHtauWHLXF1Croif0M7VJcBJHJtwRFJSkgICAhz6/e309OQlS5aoYsWKmjp1qnbs2KHt27dr6tSpqly5subNm6fp06drzZo1Gjly5G23VbJkSVWrVs2urWrVqjp27FiW/a1Wq/z9/e0eAAAg73L61M+4ceP0wQcfqHXr1ra2WrVqqUyZMoqIiNBvv/0mX19fvf7663r77bdvua2GDRtq//79dm0HDhxQSEiIs2UBAIA8yOkRld27d2cZJEJCQmzXrtSpU0cnT5687bYGDRqkTZs2afz48Tp06JDmzZunqVOnql+/fs6WBQAA8iCng0qVKlU0YcIEpaSk2NpSU1M1YcIEValSRdLfN3ErUaLEbbdVt25dLVmyRPPnz1eNGjX01ltv6f3339ezzz7rbFkAACAPcvrUz7///W899thjKlOmjGrVqiWLxaJdu3YpLS1N3333naS/Z/O88sorDm2vffv2at++vbNlAACAfMDpoNKgQQPFx8dr7ty5OnDggAzD0JNPPqlnnnlGBQsWlCR169Yt2wsFAAD5j9NBRZL8/Pz00ksvZXctAAAAdu4oqBw4cEDr1q3TmTNnlJ6ebvfcqFGjsqUwAAAAp4PKtGnT9PLLL6tYsWIKCgqyu/mbxWIhqAAAgGzjdFCJiorSuHHjNHTo0JyoBwAAwMbpoHLhwgV16dIlJ2oxHW4FDQCAazl9H5UuXbpo1apVOVELAACAHadHVO6//35FRERo06ZNqlmzpjw8POye79+/f7YVBwAA8jeng8rUqVPl5+en9evXa/369XbPWSwWggoAAMg2TgeVuLi4nKgDAAAgE6evUcmQkpKi/fv36/r169lZDwAAgI3TQeXKlSvq3bu3fHx8VL16dR07dkzS39emTJgwIdsLBAAA+ZfTQWX48OHauXOn1q1bJy8vL1t7ixYttHDhwmwtDgAA5G9OX6OydOlSLVy4UA8//LDdXWmrVaumw4cPZ2txAAAgf3N6ROXPP/9U8eLFM7VfvnzZLrgAAADcLaeDSt26dbVs2f/u2JoRTqZNm6awsLDsqwwAAOR7Tp/6iY6OVps2bRQbG6vr16/rgw8+0J49e7Rx48ZM91UBAAC4G06PqDRo0EC//PKLrly5ovvuu0+rVq1SiRIltHHjRj300EM5USMAAMinnB5RkaSaNWtq9uzZdm2nT5/W2LFjNWrUqGwpDAAA4I5v+HajU6dOKTIyMrs2BwAAkH1BBQAAILsRVAAAgGkRVAAAgGk5fDHt4MGDb/n8n3/+edfFAAAA/JPDQWX79u237dO4ceO7KgYAAOCfHA4qa9euzck6AAAAMuEaFQAAYFoEFQAAYFoEFQAAYFoEFQAAYFoEFQAAYFoOzfrZtWuXwxusVauWw33HjBmT6fuBSpQooVOnTjm8DQAAkHc5FFTq1Kkji8UiwzBksVhu2TctLc2pAqpXr67Vq1fblt3c3JxaHwAA5F0OBZW4uDjbv7dv364hQ4boX//6l8LCwiRJGzdu1DvvvKNJkyY5X4C7u4KCgpxeDwAA5H0OBZWQkBDbv7t06aIPP/xQjz76qK2tVq1aCg4OVkREhDp16uRUAQcPHlSpUqVktVpVv359jR8/XhUqVMiyb3JyspKTk23LSUlJTu0LAADcW5y+mHb37t0qX758pvby5csrNjbWqW3Vr19fc+bM0cqVKzVt2jSdOnVKDRo00Llz57LsHx0drYCAANsjODjY2fIBAMA9xOmgUrVqVUVFRenatWu2tuTkZEVFRalq1apObatt27Z64oknVLNmTbVo0ULLli2TJM2ePTvL/sOHD1diYqLtkZCQ4Gz5AADgHuLwd/1k+PTTT9WhQwcFBwerdu3akqSdO3fKYrHou+++u6tifH19VbNmTR08eDDL561Wq6xW613tAwAA3DucDir16tVTXFyc5s6dq3379skwDHXt2lXPPPOMfH1976qY5ORk7d27V40aNbqr7QAAgLzB6aAiST4+PnrxxRfveudDhgxRhw4dVLZsWZ05c0ZRUVFKSkpSjx497nrbAADg3ndHQUWSYmNjdezYMaWkpNi1P/bYYw5v4/jx43r66ad19uxZBQYG6uGHH9amTZvsZhkBAID8y+mgcuTIET3++OPavXu37SZwkmw3gnPmhm8LFixwdvcAACAfcXrWz4ABA1S+fHmdPn1aPj4+2rNnjzZs2KDQ0FCtW7cuB0oEAAD5ldMjKhs3btSaNWsUGBioAgUKqECBAnrkkUcUHR2t/v37a/v27TlRJwAAyIecHlFJS0uTn5+fJKlYsWI6ceKEpL/vXrt///7srQ4AAORrTo+o1KhRQ7t27VKFChVUv359TZo0SZ6enpo6depNb30PAABwJ5wOKiNHjtTly5clSVFRUWrfvr0aNWqkokWLauHChdleIAAAyL+cDiqtW7e2/btChQqKjY3V+fPnVbhwYdvMHwAAgOzg9DUqGQ4dOqSVK1fq6tWrKlKkSHbWBAAAIOkOgsq5c+fUvHlzVapUSY8++qhOnjwpSerTp49ef/31bC8QAADkX04HlUGDBsnDw0PHjh2Tj4+Prb1r165asWJFthYHAADyN6evUVm1apVWrlypMmXK2LVXrFhRR48ezbbCAAAAnB5RuXz5st1ISoazZ8/KarVmS1EAAADSHQSVxo0ba86cObZli8Wi9PR0TZ48WU2bNs3W4gAAQP7m9KmfyZMnq0mTJtqyZYtSUlL0xhtvaM+ePTp//rx++eWXnKgRAADkU06PqFSrVk27du1SvXr11LJlS12+fFmdO3fW9u3bdd999+VEjQAAIJ9yekRFkoKCghQZGWnXdu3aNb399tsaMmRIthQGAADg1IjK2bNntWzZMq1atUppaWmSpNTUVH3wwQcqV66cJkyYkCNFAgCA/MnhEZVff/1V7dq1U2JioiwWi0JDQzVz5kx16tRJ6enpGjlypJ5//vmcrBUAAOQzDo+oREREqHXr1tq1a5cGDBigmJgYtW/fXiNHjtTBgwf16quvZjltGQAA4E45HFR27typiIgI1ahRQ1FRUbJYLJo4caK6d+/OlxECAIAc4XBQOX/+vAIDAyVJPj4+8vHx0QMPPJBjhQEAADh8jYrFYtHFixfl5eUlwzBksVh05coVJSUl2fXz9/fP9iIBAED+5HBQMQxDlSpVslv+54hKRnjJmA0EAABwtxwOKmvXrs3JOgAAADJxOKiEh4fnZB0AAACZOH0LfQAAgNxCUAEAAKZFUAEAAKZFUAEAAKZFUAEAAKbl0Kyfzp07O7zBxYsX33ExAAAA/+RQUAkICMjpOgAAADJxKKjMnDkzp+tQdHS03nzzTQ0YMEDvv/9+ju8PAACYnymuUYmJidHUqVNVq1YtV5cCAABMxOE70/7TV199pUWLFunYsWNKSUmxe27btm1ObevSpUt69tlnNW3aNEVFRd1JOQAAII9yekTlww8/VK9evVS8eHFt375d9erVU9GiRXXkyBG1bdvW6QL69eundu3aqUWLFrftm5ycrKSkJLsHAADIu5wOKlOmTNHUqVP18ccfy9PTU2+88YZ++OEH9e/fX4mJiU5ta8GCBdq2bZuio6Md6h8dHa2AgADbIzg42NnyAQDAPcTpoHLs2DE1aNBAkuTt7a2LFy9Kkrp166b58+c7vJ2EhAQNGDBAc+fOlZeXl0PrDB8+XImJibZHQkKCs+UDAIB7iNNBJSgoSOfOnZMkhYSEaNOmTZKkuLg4GYbh8Ha2bt2qM2fO6KGHHpK7u7vc3d21fv16ffjhh3J3d1daWlqmdaxWq/z9/e0eAAAg73L6YtpmzZrp22+/1YMPPqjevXtr0KBB+uqrr7RlyxanbgzXvHlz7d69266tV69eqlKlioYOHSo3NzdnSwMAAHmM00Fl6tSpSk9PlyS99NJLKlKkiH7++Wd16NBBL730ksPbKViwoGrUqGHX5uvrq6JFi2ZqBwAA+ZPTQeX48eN2F7E+9dRTeuqpp2QYhhISElS2bNlsLRAAAORfTgeV8uXL6+TJkypevLhd+/nz51W+fPksry1x1Lp16+54XQAAkPc4fTGtYRiyWCyZ2i9duuTw7B0AAABHODyiMnjwYEmSxWJRRESEfHx8bM+lpaVp8+bNqlOnTrYXCAAA8i+Hg8r27dsl/T2isnv3bnl6etqe8/T0VO3atTVkyJDsrxAAAORbDgeVtWvXSvp7CvEHH3zAPUwAAECOc/pi2pkzZ9r+ffz4cVksFpUuXTpbiwIAAJDu4GLa9PR0jR07VgEBAQoJCVHZsmVVqFAhvfXWW7b7qwAAAGQHp0dURowYoenTp2vChAlq2LChDMPQL7/8ojFjxujatWsaN25cTtQJAADyIaeDyuzZs/Wf//xHjz32mK2tdu3aKl26tF555RWCCgAAyDZOn/o5f/68qlSpkqm9SpUqOn/+fLYUBQAAIN1BUKldu7Y+/vjjTO0ff/yxateunS1FAQAASE6c+qlQoYJiYmI0adIktWvXTqtXr1ZYWJgsFot+/fVXJSQkaPny5TlZKwAAyGccHlGJj49XWlqawsPDdeDAAT3++OP666+/dP78eXXu3Fn79+9Xo0aNcrJWAACQzzh9Ma0klSpViotmAQBAjnMqqMTGxurUqVO37FOrVq27KggAACCDU0GlefPmMgzjps9bLBalpaXddVEAAACSk0Fl8+bNCgwMzKlaAAAA7DgVVMqWLavixYvnVC0AAAB2nL6PCgAAQG5xOKiEh4fL09MzJ2sBAACw4/Cpn7Vr1+ZkHQAAAJlw6gcAAJgWQQUAAJiWQ0ElKSkpp+sAAADIxKGgUrhwYZ05c0aS1KxZM/311185WRMAAIAkB4OKn5+fzp07J0lat26dUlNTc7QoAAAAycFZPy1atFDTpk1VtWpVSdLjjz9+06nKa9asyb7qAABAvuZQUJk7d65mz56tw4cPa/369apevbp8fHxyujYAAJDPORRUvL299dJLL0mStmzZookTJ6pQoUI5WRcAAIBz3/Uj2d/4LeOblC0WS/ZVBAAA8P/d0X1U5syZo5o1a8rb21ve3t6qVauWPv/88+yuDQAA5HNOj6i8++67ioiI0KuvvqqGDRvKMAz98ssveumll3T27FkNGjQoJ+oEAAD5kNNB5aOPPtInn3yi7t2729o6duyo6tWra8yYMQQVAACQbZw+9XPy5Ek1aNAgU3uDBg108uRJp7b1ySefqFatWvL395e/v7/CwsL0/fffO1sSAADIo5wOKvfff78WLVqUqX3hwoWqWLGiU9sqU6aMJkyYoC1btmjLli1q1qyZOnbsqD179jhbFgAAyIOcPvUTGRmprl27asOGDWrYsKEsFot+/vln/fjjj1kGmFvp0KGD3fK4ceP0ySefaNOmTapevbqzpQEAgDzG6aDyxBNPaPPmzXrvvfe0dOlSGYahatWq6bffftMDDzxwx4WkpaXpyy+/1OXLlxUWFpZln+TkZCUnJ9uW+bJEAADyNqeDiiQ99NBDmjt3brYUsHv3boWFhenatWvy8/PTkiVLVK1atSz7RkdHKzIyMlv2CwAAzO+O7qOSnSpXrqwdO3Zo06ZNevnll9WjRw/FxsZm2Xf48OFKTEy0PRISEnK5WgAAkJvuaEQlO3l6eur++++XJIWGhiomJkYffPCBPvvss0x9rVarrFZrbpcIAABcxOUjKjcyDMPuOhQAAJB/uXRE5c0331Tbtm0VHBysixcvasGCBVq3bp1WrFjhyrIAAIBJuDSonD59Wt26ddPJkycVEBCgWrVqacWKFWrZsqUrywIAACbhdFC5du2aPvroI61du1ZnzpxRenq63fPbtm1zeFvTp093dvcAACAfcTqoPP/88/rhhx/05JNPql69erJYLDlRFwAAgPNBZdmyZVq+fLkaNmyYE/UAAADYOD3rp3Tp0ipYsGBO1AIAAGDH6aDyzjvvaOjQoTp69GhO1AMAAGDj9Kmf0NBQXbt2TRUqVJCPj488PDzsnj9//ny2FQcAAPI3p4PK008/rT/++EPjx49XiRIluJgWAADkGKeDyq+//qqNGzeqdu3aOVEPAACAjdPXqFSpUkVXr17NiVoAAADsOB1UJkyYoNdff13r1q3TuXPnlJSUZPcAAADILk6f+mnTpo0kqXnz5nbthmHIYrEoLS0teyoDAAD5ntNBZe3atTlRBwAAQCZOB5Xw8PCcqAMAACATp4PKhg0bbvl848aN77gYAACAf3I6qDRp0iRT2z/vpcI1KgAAILs4PevnwoULdo8zZ85oxYoVqlu3rlatWpUTNQIAgHzK6RGVgICATG0tW7aU1WrVoEGDtHXr1mwpDAAAwOkRlZsJDAzU/v37s2tzAAAAzo+o7Nq1y27ZMAydPHlSEyZM4Lb6AAAgWzkdVOrUqSOLxSLDMOzaH374Yc2YMSPbCgMAAHA6qMTFxdktFyhQQIGBgfLy8sq2ogAAAKQ7CCohISE5UQcAAEAmDl9Mu3nzZn3//fd2bXPmzFH58uVVvHhxvfjii0pOTs72AgEAQP7lcFAZM2aM3YW0u3fvVu/evdWiRQsNGzZM3377raKjo3OkSAAAkD85HFR27Nhh943JCxYsUP369TVt2jQNHjxYH374oRYtWpQjRQIAgPzJ4aBy4cIFlShRwra8fv16tWnTxrZct25dJSQkZG91AAAgX3M4qJQoUcI24yclJUXbtm1TWFiY7fmLFy/Kw8Mj+ysEAAD5lsNBpU2bNho2bJh++uknDR8+XD4+PmrUqJHt+V27dum+++7LkSIBAED+5PD05KioKHXu3Fnh4eHy8/PT7Nmz5enpaXt+xowZatWqVY4UCQAA8ieHg0pgYKB++uknJSYmys/PT25ubnbPf/nll/Lz88v2AgEAQP6VLd+eLElFihS562IAAAD+Kdu+PRkAACC7OT2ikp2io6O1ePFi7du3T97e3mrQoIEmTpyoypUru7IsAGYxJusR3DxpTKKrKwBMyaUjKuvXr1e/fv20adMm/fDDD7p+/bpatWqly5cvu7IsAABgEi4dUVmxYoXd8syZM1W8eHFt3bpVjRs3dlFVAADALFwaVG6UmPj30OfNLsxNTk62++LDpKSkXKkLAAC4hmkupjUMQ4MHD9YjjzyiGjVqZNknOjpaAQEBtkdwcHAuVwkAAHKTaYLKq6++ql27dmn+/Pk37TN8+HAlJibaHny3EAAAeZspTv289tpr+uabb7RhwwaVKVPmpv2sVqusVmsuVgYAAFzJpUHFMAy99tprWrJkidatW6fy5cu7shwAAGAyLg0q/fr107x58/T111+rYMGCOnXqlKS/737r7e3tytIAAIAJuPQalU8++USJiYlq0qSJSpYsaXssXLjQlWUBAACTcPmpHwAAgJsxzawfAACAG5li1g8AAPcUvocq1zCiAgAATIugAgAATIugAgAATIugAgAATIugAgAATIugAgAATIugAgAATIugAgAATIugAgAATIugAgAATIugAgAATIugAgAATIugAgAATIugAgAATIugAgAATMvd1QXABMYEuLqC3DMm0dUVAACcwIgKAAAwLYIKAAAwLYIKAAAwLYIKAAAwLYIKAAAwLYIKAAAwLYIKAAAwLYIKAAAwLYIKAAAwLYIKAAAwLYIKAAAwLYIKAAAwLYIKAAAwLZcGlQ0bNqhDhw4qVaqULBaLli5d6spyAACAybg0qFy+fFm1a9fWxx9/7MoyAACASbm7cudt27ZV27ZtXVkCAAAwMZcGFWclJycrOTnZtpyUlOTCagAAQE67py6mjY6OVkBAgO0RHBzs6pIAAEAOuqeCyvDhw5WYmGh7JCQkuLokAACQg+6pUz9Wq1VWq9XVZQAAgFxyT42oAACA/MWlIyqXLl3SoUOHbMtxcXHasWOHihQporJly7qwMgAAYAYuDSpbtmxR06ZNbcuDBw+WJPXo0UOzZs1yUVUAAMAsXBpUmjRpIsMwXFkCAAAwMa5RAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApuXyoDJlyhSVL19eXl5eeuihh/TTTz+5uiQAAGASLg0qCxcu1MCBAzVixAht375djRo1Utu2bXXs2DFXlgUAAEzCpUHl3XffVe/evdWnTx9VrVpV77//voKDg/XJJ5+4siwAAGAS7q7acUpKirZu3aphw4bZtbdq1Uq//vprluskJycrOTnZtpyYmChJSkpKypEa05Ov5Mh2zSbJYri6hNyTQ8dKbuPYzIM4Nu8pHJt3u8m/t2kYt38fXRZUzp49q7S0NJUoUcKuvUSJEjp16lSW60RHRysyMjJTe3BwcI7UmF8EuLqA3DQhX73ae16++mlxbN5T8tVPKwePzYsXLyog4Nbbd1lQyWCxWOyWDcPI1JZh+PDhGjx4sG05PT1d58+fV9GiRW+6Dm4tKSlJwcHBSkhIkL+/v6vLAWw4NmFWHJt3zzAMXbx4UaVKlbptX5cFlWLFisnNzS3T6MmZM2cyjbJksFqtslqtdm2FChXKqRLzFX9/fz5wMCWOTZgVx+bdud1ISgaXXUzr6emphx56SD/88INd+w8//KAGDRq4qCoAAGAmLj31M3jwYHXr1k2hoaEKCwvT1KlTdezYMb300kuuLAsAAJiES4NK165dde7cOY0dO1YnT55UjRo1tHz5coWEhLiyrHzFarVq9OjRmU6pAa7GsQmz4tjMXRbDkblBAAAALuDyW+gDAADcDEEFAACYFkEFAACYFkEFAACYFkEFAACYFkElH2nSpIkGDhxo1zZgwAA99NBDslqtqlOnjkvqAm48Nnfu3Kmnn35awcHB8vb2VtWqVfXBBx+4rkDkWzcem+fOnVObNm1UqlQpWa1WBQcH69VXX82xL8eFCb7rB65lGIaef/55bd68Wbt27XJ1OYAkaevWrQoMDNTcuXMVHBysX3/9VS+++KLc3Nz06quvuro85GMFChRQx44dFRUVpcDAQB06dEj9+vXT+fPnNW/ePFeXlydxH5V8omfPnpo9e7ZdW1xcnMqVKydJGjNmjJYuXaodO3bkfnHI1253bGbo16+f9u7dqzVr1uRidcjPHD02P/zwQ02ePFkJCQm5WF3+wYhKPvHBBx/owIEDqlGjhsaOHStJCgwMdHFVgOPHZmJioooUKZLb5SEfc+TYPHHihBYvXqzw8HBXlJgvcI1KPhEQECBPT0/5+PgoKChIQUFBcnNzc3VZgEPH5saNG7Vo0SL17dvXRVUiP7rVsfn000/Lx8dHpUuXlr+/v/7zn/+4uNq8i6ACwNT27Nmjjh07atSoUWrZsqWrywEkSe+99562bdumpUuX6vDhwxo8eLCrS8qzOPUDwLRiY2PVrFkzvfDCCxo5cqSrywFsMkZYqlSpoqJFi6pRo0aKiIhQyZIlXV1ansOISj7i6emptLQ0V5cBZJLVsblnzx41bdpUPXr00Lhx41xUGfI7R/7fzJiTkpycnBsl5TuMqOQj5cqV0+bNmxUfHy8/Pz8VKVJER44c0aVLl3Tq1CldvXrVNuunWrVq8vT0dG3ByDduPDZPnz6tpk2bqlWrVho8eLBOnTolSXJzc+MicOSqG4/N3377TadPn1bdunXl5+en2NhYvfHGG2rYsGGm2UDIHkxPzkcOHDigHj16aOfOnbp69ari4uLUs2dPrV+/PlPfrKbgATnlxmOzR48emaaFSlJISIji4+Nzv0DkWzcemzNmzNC0adMUGxur5ORkBQcHq3Pnzho2bJgKFSrk6nLzJIIKAAAwLa5RAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApkVQAQAApvX/ACQvsszvWl3YAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "make_table(ins, outs, \"Origs\", \"Trimmed\", \"Sum of Total Read Lengths\", samples, \"Sum of Total Reads Per Sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "728a7038-553c-4428-9112-becd7d1c87cf",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAjcAAAHFCAYAAAAOmtghAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjguMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8g+/7EAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAzs0lEQVR4nO3deVyVZf7/8feRHZXjlixqaOOGOmlhmduoGbilaTPfaBwDk2ayzZRs0jbL/IbZN7NF1MaU+uaULU5j6ZhYqKU5hYJWmvZTExeQNAOXRITr94cPz3eOoMIBOXD5ej4e9+PRfZ3rvu/PuaLO+3Hdm8MYYwQAAGCJOt4uAAAAoCoRbgAAgFUINwAAwCqEGwAAYBXCDQAAsArhBgAAWIVwAwAArEK4AQAAViHcAAAAqxBuAMs4HI5yLatXry5z+759+6pv377VWvOllpqaKofDoR9//PGC/Z566im3MfLz89OVV16pP//5z8rNza2eYs+xevXqC/77AlCar7cLAFC1vvzyS7f1Z555Runp6frss8/c2jt06FDm9ikpKZesttpixYoVcjqdOnbsmFauXKkXXnhB69evV1ZWlvz8/LxdHoCLINwAlrnhhhvc1q+44grVqVOnVPu5Tpw4oeDg4POGnstJdHS0mjRpIkm66aabdOjQIS1cuFBffPGF+vXr5+XqAFwMp6WAy1Dfvn3VqVMnrV27Vj169FBwcLDGjBnj+uw/T0v9+OOPcjgcev755/Xcc8+pZcuWCgoKUt++fbVjxw4VFRVp0qRJioiIkNPp1IgRI5SXl+d2vJYtW+rmm2/Wxx9/rGuuuUZBQUGKiorSxx9/LOnMaaOoqCjVrVtX119/vTIyMkrVnJGRoWHDhqlRo0YKDAzUNddco3fffbdUvw0bNqhnz54KDAxURESEJk+erKKiokqNV9euXSVJBw8edGtftWqV+vfvr5CQEAUHB6tnz5769NNP3fr8v//3/3TnnXeqTZs2Cg4OVrNmzTR06FB98803pY7z/fffa+DAgQoODlaTJk00duxYHT16tFS/zMxM3XzzzWratKkCAgIUERGhIUOGaN++fZX6noAtCDfAZSonJ0ejRo3SyJEjtXz5ct17770X7D979mytW7dOs2fP1vz58/X9999r6NChSkxM1E8//aQFCxZoxowZWrVqle66665S22/evFmTJ0/WI488oiVLlsjpdOrWW2/VlClTNH/+fD377LNatGiR8vPzdfPNN+vXX391bZuenq6ePXvql19+0dy5c/XPf/5TXbp0UVxcnFJTU139tm7dqv79++uXX35Ramqq5s6dq8zMTE2bNq1SY7V7925JUtu2bV1tb731lmJjYxUSEqI33nhD7777rho1aqQBAwa4BZwDBw6ocePGmj59ulasWKHZs2fL19dX3bp10/bt2139Dh48qD59+ujbb79VSkqK/vd//1fHjh3T/fff71bL8ePHFRMTo4MHD2r27NlKS0vTrFmzdOWVV5YZhIDLkgFgtYSEBFO3bl23tj59+hhJ5tNPPy3Vv0+fPqZPnz6u9d27dxtJpnPnzqa4uNjVPmvWLCPJDBs2zG378ePHG0kmPz/f1RYZGWmCgoLMvn37XG1ZWVlGkgkPDzfHjx93tX/44YdGklm6dKmrrX379uaaa64xRUVFbse6+eabTXh4uKuuuLg4ExQUZHJzc119Tp8+bdq3b28kmd27d19oqMyUKVOMJJObm2uKiorMkSNHzLvvvmvq1q1r/vjHP7r6HT9+3DRq1MgMHTrUbfvi4mLTuXNnc/3115/3GKdPnzanTp0ybdq0MRMmTHC1P/LII8bhcJisrCy3/jExMUaSSU9PN8YYk5GRYSSZDz/88ILfBbicMXMDXKYaNmyoG2+8sdz9Bw8erDp1/u9/GVFRUZKkIUOGuPU7256dne3W3qVLFzVr1qxUv759+yo4OLhU+549eySdOa3z/fff609/+pMk6fTp065l8ODBysnJcc2ApKenq3///goNDXXtz8fHR3FxceX+npIUFhYmPz8/NWzYULfddpuio6P1xhtvuD5fv369fv75ZyUkJLjVU1JSooEDB+rrr7/W8ePHXfU+++yz6tChg/z9/eXr6yt/f3/98MMP2rZtm2uf6enp6tixozp37uxWy8iRI93WW7durYYNG+qRRx7R3LlztXXr1gp9N+ByQLgBLlPh4eEV6t+oUSO3dX9//wu2nzx5skq2P3udy8SJE+Xn5+e2nD2VdujQIUnS4cOHFRYWVqr2stouZNWqVfr666/1ySef6Pe//73Wrl2rBx54wPX52Zr+8Ic/lKrpueeekzFGP//8syQpKSlJTzzxhIYPH66PPvpI//73v/X111+rc+fObqfeylu70+nUmjVr1KVLFz366KPq2LGjIiIiNGXKlEpfWwTYgrulgMuUw+HwdgnlcvaupcmTJ+vWW28ts0+7du0kSY0bNy7zeTQVfUZN586dXceNiYnRgAED9NprrykxMVHXXXed67NXXnnlvHehnZ09euuttxQfH69nn33W7fNDhw6pQYMGrvWK1P7b3/5W77zzjowx2rJli1JTUzV16lQFBQVp0qRJFfqugI0INwBqtHbt2qlNmzbavHlzqYBwrn79+mnp0qU6ePCgK1wUFxdr8eLFHh/f4XBo9uzZ6tChgx5//HF98skn6tmzpxo0aKCtW7eWuuC3rO0DAgLc2pYtW6b9+/erdevWbrXPmDFDmzdvdjs19fe///2C++7cubNefPFFpaamatOmTR5+S8AuhBsANd68efM0aNAgDRgwQKNHj1azZs30888/a9u2bdq0aZPee+89SdLjjz+upUuX6sYbb9STTz6p4OBgzZ4923X9i6fatGmjv/zlL0pJSdEXX3yhXr166ZVXXlFCQoJ+/vln/eEPf1DTpk31008/afPmzfrpp580Z84cSdLNN9+s1NRUtW/fXldffbU2btyo559/Xs2bN3c7xvjx47VgwQINGTJE06ZNU2hoqBYtWqTvv//erd/HH3+slJQUDR8+XFdddZWMMVqyZIl++eUXxcTEVOp7ArbgmhsANV6/fv301VdfqUGDBho/frxuuukm3XPPPVq1apVuuukmV79OnTpp1apVCgkJUUJCgv7yl7/o6quv1hNPPFHpGqZMmaJ69erpySeflCSNGjVK6enpOnbsmO6++27ddNNNevDBB7Vp0yb179/ftd1LL72kUaNGKTk5WUOHDtXSpUu1ZMkS/eY3v3Hbf1hYmNasWaMOHTronnvu0ahRoxQYGKhXX33VrV+bNm3UoEEDzZgxQ8OGDdN//dd/adOmTUpNTdWf//znSn9PwAYOY4zxdhEAAABVhZkbAABgFcINAACwCuEGAABYhXADAACsQrgBAABWIdwAAACrXHYP8SspKdGBAwdUv379WvP4eQAALnfGGB09elQRERFuL/Ety2UXbg4cOKAWLVp4uwwAAOCBvXv3lnrC97kuu3BTv359SWcGJyQkxMvVAACA8igoKFCLFi1cv+MXctmFm7OnokJCQgg3AADUMuW5pIQLigEAgFW8Gm7Wrl2roUOHKiIiQg6HQx9++OFFt1mzZo2io6MVGBioq666SnPnzr30hQIAgFrDq+Hm+PHj6ty5c6m33p7P7t27NXjwYPXu3VuZmZl69NFHNW7cOH3wwQeXuFIAAFBbePWam0GDBmnQoEHl7j937lxdeeWVmjVrliQpKipKGRkZ+p//+R/9/ve/v0RVAgCA2qRWXXPz5ZdfKjY21q1twIABysjIUFFRkZeqAgAANUmtulsqNzdXoaGhbm2hoaE6ffq0Dh06pPDw8FLbFBYWqrCw0LVeUFBwyesEAADeU6tmbqTSt4AZY8psPys5OVlOp9O18AA/AADsVqvCTVhYmHJzc93a8vLy5Ovrq8aNG5e5zeTJk5Wfn+9a9u7dWx2lAgAAL6lVp6W6d++ujz76yK1t5cqV6tq1q/z8/MrcJiAgQAEBAdVRHgAAqAG8OnNz7NgxZWVlKSsrS9KZW72zsrKUnZ0t6cysS3x8vKv/2LFjtWfPHiUlJWnbtm1asGCBXn/9dU2cONEb5QMAgBrIqzM3GRkZ6tevn2s9KSlJkpSQkKDU1FTl5OS4go4ktWrVSsuXL9eECRM0e/ZsRURE6OWXX+Y2cAAA4OIwZ6/IvUwUFBTI6XQqPz+fd0sBAFBLVOT3u1ZdUAwAAHAxhBsAAGAVwg0AALBKrboVHACAy03LScu8XUKF/Th9iFePz8wNAACwCuEGAABYhXADAACsQrgBAABWIdwAAACrEG4AAIBVCDcAAMAqhBsAAGAVwg0AALAK4QYAAFiFcAMAAKxCuAEAAFYh3AAAAKsQbgAAgFUINwAAwCqEGwAAYBXCDQAAsArhBgAAWIVwAwAArEK4AQAAViHcAAAAqxBuAACAVQg3AADAKoQbAABgFcINAACwCuEGAABYhXADAACsQrgBAABWIdwAAACrEG4AAIBVCDcAAMAqhBsAAGAVwg0AALAK4QYAAFiFcAMAAKxCuAEAAFYh3AAAAKsQbgAAgFUINwAAwCqEGwAAYBXCDQAAsArhBgAAWIVwAwAArEK4AQAAViHcAAAAqxBuAACAVQg3AADAKoQbAABgFcINAACwCuEGAABYhXADAACsQrgBAABWIdwAAACrEG4AAIBVCDcAAMAqhBsAAGAVwg0AALAK4QYAAFjF6+EmJSVFrVq1UmBgoKKjo/X5559fsP+iRYvUuXNnBQcHKzw8XHfeeacOHz5cTdUCAICazqvhZvHixRo/frwee+wxZWZmqnfv3ho0aJCys7PL7P/FF18oPj5eiYmJ+u677/Tee+/p66+/1l133VXNlQMAgJrKq+Fm5syZSkxM1F133aWoqCjNmjVLLVq00Jw5c8rsv2HDBrVs2VLjxo1Tq1at1KtXL919993KyMio5soBAEBN5bVwc+rUKW3cuFGxsbFu7bGxsVq/fn2Z2/To0UP79u3T8uXLZYzRwYMH9f7772vIkCHnPU5hYaEKCgrcFgAAYC+vhZtDhw6puLhYoaGhbu2hoaHKzc0tc5sePXpo0aJFiouLk7+/v8LCwtSgQQO98sor5z1OcnKynE6na2nRokWVfg8AAFCzeP2CYofD4bZujCnVdtbWrVs1btw4Pfnkk9q4caNWrFih3bt3a+zYsefd/+TJk5Wfn+9a9u7dW6X1AwCAmsXXWwdu0qSJfHx8Ss3S5OXllZrNOSs5OVk9e/bUww8/LEm6+uqrVbduXfXu3VvTpk1TeHh4qW0CAgIUEBBQ9V8AAADUSF6bufH391d0dLTS0tLc2tPS0tSjR48ytzlx4oTq1HEv2cfHR9KZGR8AAACvnpZKSkrS/PnztWDBAm3btk0TJkxQdna26zTT5MmTFR8f7+o/dOhQLVmyRHPmzNGuXbu0bt06jRs3Ttdff70iIiK89TUAAEAN4rXTUpIUFxenw4cPa+rUqcrJyVGnTp20fPlyRUZGSpJycnLcnnkzevRoHT16VK+++qoeeughNWjQQDfeeKOee+45b30FAABQwzjMZXY+p6CgQE6nU/n5+QoJCfF2OQAAXFDLScu8XUKF/Tj9/I9o8VRFfr+9frcUAABAVSLcAAAAqxBuAACAVQg3AADAKoQbAABgFcINAACwCuEGAABYhXADAACsQrgBAABWIdwAAACrEG4AAIBVCDcAAMAqhBsAAGAVwg0AALAK4QYAAFiFcAMAAKxCuAEAAFYh3AAAAKsQbgAAgFUINwAAwCqEGwAAYBXCDQAAsArhBgAAWIVwAwAArEK4AQAAViHcAAAAqxBuAACAVQg3AADAKoQbAABgFcINAACwCuEGAABYhXADAACsQrgBAABWIdwAAACrEG4AAIBVCDcAAMAqhBsAAGAVwg0AALAK4QYAAFiFcAMAAKxCuAEAAFYh3AAAAKsQbgAAgFUINwAAwCqEGwAAYBXCDQAAsArhBgAAWIVwAwAArEK4AQAAViHcAAAAqxBuAACAVQg3AADAKoQbAABgFcINAACwCuEGAABYhXADAACsQrgBAABWIdwAAACrEG4AAIBVCDcAAMAqhBsAAGAVr4eblJQUtWrVSoGBgYqOjtbnn39+wf6FhYV67LHHFBkZqYCAAP3mN7/RggULqqlaAABQ0/l68+CLFy/W+PHjlZKSop49e2revHkaNGiQtm7dqiuvvLLMbW677TYdPHhQr7/+ulq3bq28vDydPn26misHAAA1lcMYY7x18G7duunaa6/VnDlzXG1RUVEaPny4kpOTS/VfsWKFbr/9du3atUuNGjXy6JgFBQVyOp3Kz89XSEiIx7UDAFAdWk5a5u0SKuzH6UOqfJ8V+f326LTU7t27PSrsP506dUobN25UbGysW3tsbKzWr19f5jZLly5V165dNWPGDDVr1kxt27bVxIkT9euvv573OIWFhSooKHBbAACAvTwKN61bt1a/fv301ltv6eTJkx4d+NChQyouLlZoaKhbe2hoqHJzc8vcZteuXfriiy/07bff6h//+IdmzZql999/X/fdd995j5OcnCyn0+laWrRo4VG9AACgdvAo3GzevFnXXHONHnroIYWFhenuu+/WV1995VEBDofDbd0YU6rtrJKSEjkcDi1atEjXX3+9Bg8erJkzZyo1NfW8szeTJ09Wfn6+a9m7d69HdQIAgNrBo3DTqVMnzZw5U/v379fChQuVm5urXr16qWPHjpo5c6Z++umni+6jSZMm8vHxKTVLk5eXV2o256zw8HA1a9ZMTqfT1RYVFSVjjPbt21fmNgEBAQoJCXFbAACAvSp1K7ivr69GjBihd999V88995x27typiRMnqnnz5oqPj1dOTs55t/X391d0dLTS0tLc2tPS0tSjR48yt+nZs6cOHDigY8eOudp27NihOnXqqHnz5pX5KgAAwBKVCjcZGRm69957FR4erpkzZ2rixInauXOnPvvsM+3fv1+33HLLBbdPSkrS/PnztWDBAm3btk0TJkxQdna2xo4dK+nMKaX4+HhX/5EjR6px48a68847tXXrVq1du1YPP/ywxowZo6CgoMp8FQAAYAmPnnMzc+ZMLVy4UNu3b9fgwYP15ptvavDgwapT50xWatWqlebNm6f27dtfcD9xcXE6fPiwpk6dqpycHHXq1EnLly9XZGSkJCknJ0fZ2dmu/vXq1VNaWpoeeOABde3aVY0bN9Ztt92madOmefI1AACAhTx6zk2bNm00ZswY3XnnnQoLCyuzz6lTp/T2228rISGh0kVWJZ5zAwCoTXjOzRkV+f32aObmhx9+uGgff3//GhdsAACA/Ty65mbhwoV67733SrW/9957euONNypdFAAAgKc8CjfTp09XkyZNSrU3bdpUzz77bKWLAgAA8JRH4WbPnj1q1apVqfbIyEi3C4ABAACqm0fhpmnTptqyZUup9s2bN6tx48aVLgoAAMBTHoWb22+/XePGjVN6erqKi4tVXFyszz77TA8++KBuv/32qq4RAACg3Dy6W2ratGnas2eP+vfvL1/fM7soKSlRfHw819wAAACv8ijc+Pv7a/HixXrmmWe0efNmBQUF6be//a3r4XsAAADe4lG4Oatt27Zq27ZtVdUCAABQaR6Fm+LiYqWmpurTTz9VXl6eSkpK3D7/7LPPqqQ4AACAivIo3Dz44INKTU3VkCFD1KlTJzkcjqquCwAAwCMehZt33nlH7777rgYPHlzV9QAAAFSKR7eC+/v7q3Xr1lVdCwAAQKV5FG4eeughvfTSS/LgheIAAACXlEenpb744gulp6frX//6lzp27Cg/Pz+3z5csWVIlxQEAAFSUR+GmQYMGGjFiRFXXAgAAUGkehZuFCxdWdR0AAABVwqNrbiTp9OnTWrVqlebNm6ejR49Kkg4cOKBjx45VWXEAAAAV5dHMzZ49ezRw4EBlZ2ersLBQMTExql+/vmbMmKGTJ09q7ty5VV0nAABAuXg0c/Pggw+qa9euOnLkiIKCglztI0aM0KefflplxQEAAFSUx3dLrVu3Tv7+/m7tkZGR2r9/f5UUBgAA4AmPZm5KSkpUXFxcqn3fvn2qX79+pYsCAADwlEfhJiYmRrNmzXKtOxwOHTt2TFOmTOGVDAAAwKs8Oi314osvql+/furQoYNOnjypkSNH6ocfflCTJk309ttvV3WNAAAA5eZRuImIiFBWVpbefvttbdq0SSUlJUpMTNSf/vQntwuMAQAAqptH4UaSgoKCNGbMGI0ZM6Yq6wEAAKgUj8LNm2++ecHP4+PjPSoGAACgsjwKNw8++KDbelFRkU6cOCF/f38FBwcTbgAAgNd4dLfUkSNH3JZjx45p+/bt6tWrFxcUAwAAr/L43VLnatOmjaZPn15qVgcAAKA6VVm4kSQfHx8dOHCgKncJAABQIR5dc7N06VK3dWOMcnJy9Oqrr6pnz55VUhgAAIAnPAo3w4cPd1t3OBy64oordOONN+qFF16oiroAAAA84lG4KSkpqeo6AAAAqkSVXnMDAADgbR7N3CQlJZW778yZMz05BAAAgEc8CjeZmZnatGmTTp8+rXbt2kmSduzYIR8fH1177bWufg6Ho2qqBAAAKCePws3QoUNVv359vfHGG2rYsKGkMw/2u/POO9W7d2899NBDVVokAABAeXl0zc0LL7yg5ORkV7CRpIYNG2ratGncLQUAALzKo3BTUFCggwcPlmrPy8vT0aNHK10UAACApzwKNyNGjNCdd96p999/X/v27dO+ffv0/vvvKzExUbfeemtV1wgAAFBuHl1zM3fuXE2cOFGjRo1SUVHRmR35+ioxMVHPP/98lRYIAABQER6Fm+DgYKWkpOj555/Xzp07ZYxR69atVbdu3aquDwAAoEIq9RC/nJwc5eTkqG3btqpbt66MMVVVFwAAgEc8CjeHDx9W//791bZtWw0ePFg5OTmSpLvuuovbwAEAgFd5FG4mTJggPz8/ZWdnKzg42NUeFxenFStWVFlxAAAAFeXRNTcrV67UJ598oubNm7u1t2nTRnv27KmSwgAAADzh0czN8ePH3WZszjp06JACAgIqXRQAAICnPAo3v/vd7/Tmm2+61h0Oh0pKSvT888+rX79+VVYcAABARXl0Wur5559X3759lZGRoVOnTumvf/2rvvvuO/38889at25dVdcIAABQbh7N3HTo0EFbtmzR9ddfr5iYGB0/fly33nqrMjMz9Zvf/KaqawQAACi3Cs/cFBUVKTY2VvPmzdPTTz99KWoCAADwWIVnbvz8/PTtt9/K4XBcinoAAAAqxaPTUvHx8Xr99deruhYAAIBK8+iC4lOnTmn+/PlKS0tT165dS71TaubMmVVSHAAAQEVVKNzs2rVLLVu21Lfffqtrr71WkrRjxw63PpyuAgAA3lShcNOmTRvl5OQoPT1d0pnXLbz88ssKDQ29JMUBAABUVIWuuTn3rd//+te/dPz48SotCAAAoDI8uqD4rHPDDgAAgLdVKNw4HI5S19RwjQ0AAKhJKnTNjTFGo0ePdr0c8+TJkxo7dmypu6WWLFlSdRUCAABUQIVmbhISEtS0aVM5nU45nU6NGjVKERERrvWzS0WkpKSoVatWCgwMVHR0tD7//PNybbdu3Tr5+vqqS5cuFToeAACwW4VmbhYuXFilB1+8eLHGjx+vlJQU9ezZU/PmzdOgQYO0detWXXnllefdLj8/X/Hx8erfv78OHjxYpTUBAIDarVIXFFfWzJkzlZiYqLvuuktRUVGaNWuWWrRooTlz5lxwu7vvvlsjR45U9+7dq6lSAABQW3gt3Jw6dUobN25UbGysW3tsbKzWr19/3u0WLlyonTt3asqUKeU6TmFhoQoKCtwWAABgL6+Fm0OHDqm4uLjUAwBDQ0OVm5tb5jY//PCDJk2apEWLFsnXt3xn1JKTk92uB2rRokWlawcAADWXV09LSaVvJTfGlHl7eXFxsUaOHKmnn35abdu2Lff+J0+erPz8fNeyd+/eStcMAABqLo9enFkVmjRpIh8fn1KzNHl5eWW+zuHo0aPKyMhQZmam7r//fklSSUmJjDHy9fXVypUrdeONN5baLiAgwHXrOgAAsJ/XZm78/f0VHR2ttLQ0t/a0tDT16NGjVP+QkBB98803ysrKci1jx45Vu3btlJWVpW7dulVX6QAAoAbz2syNJCUlJemOO+5Q165d1b17d7322mvKzs7W2LFjJZ05pbR//369+eabqlOnjjp16uS2fdOmTRUYGFiqHQAAXL68Gm7i4uJ0+PBhTZ06VTk5OerUqZOWL1+uyMhISVJOTo6ys7O9WSIAAKhlHOYye/tlQUGBnE6n8vPzFRIS4u1yAAC4oJaTlnm7hAr7cfqQKt9nRX6/vX63FAAAQFUi3AAAAKsQbgAAgFUINwAAwCqEGwAAYBXCDQAAsArhBgAAWIVwAwAArEK4AQAAViHcAAAAqxBuAACAVQg3AADAKoQbAABgFcINAACwCuEGAABYhXADAACsQrgBAABWIdwAAACrEG4AAIBVCDcAAMAqhBsAAGAVwg0AALAK4QYAAFiFcAMAAKxCuAEAAFYh3AAAAKsQbgAAgFUINwAAwCqEGwAAYBXCDQAAsArhBgAAWIVwAwAArEK4AQAAViHcAAAAqxBuAACAVQg3AADAKoQbAABgFcINAACwCuEGAABYhXADAACsQrgBAABWIdwAAACrEG4AAIBVCDcAAMAqhBsAAGAVwg0AALAK4QYAAFiFcAMAAKxCuAEAAFYh3AAAAKsQbgAAgFUINwAAwCqEGwAAYBXCDQAAsArhBgAAWIVwAwAArEK4AQAAViHcAAAAqxBuAACAVQg3AADAKoQbAABgFa+Hm5SUFLVq1UqBgYGKjo7W559/ft6+S5YsUUxMjK644gqFhISoe/fu+uSTT6qxWgAAUNN5NdwsXrxY48eP12OPPabMzEz17t1bgwYNUnZ2dpn9165dq5iYGC1fvlwbN25Uv379NHToUGVmZlZz5QAAoKZyGGOMtw7erVs3XXvttZozZ46rLSoqSsOHD1dycnK59tGxY0fFxcXpySefLFf/goICOZ1O5efnKyQkxKO6AQCoLi0nLfN2CRX24/QhVb7Pivx+e23m5tSpU9q4caNiY2Pd2mNjY7V+/fpy7aOkpERHjx5Vo0aNztunsLBQBQUFbgsAALCX18LNoUOHVFxcrNDQULf20NBQ5ebmlmsfL7zwgo4fP67bbrvtvH2Sk5PldDpdS4sWLSpVNwAAqNm8fkGxw+FwWzfGlGory9tvv62nnnpKixcvVtOmTc/bb/LkycrPz3cte/furXTNAACg5vL11oGbNGkiHx+fUrM0eXl5pWZzzrV48WIlJibqvffe00033XTBvgEBAQoICKh0vQAAoHbw2syNv7+/oqOjlZaW5taelpamHj16nHe7t99+W6NHj9bf//53DRlS9RcsAQCA2s1rMzeSlJSUpDvuuENdu3ZV9+7d9dprryk7O1tjx46VdOaU0v79+/Xmm29KOhNs4uPj9dJLL+mGG25wzfoEBQXJ6XR67XsAAICaw6vhJi4uTocPH9bUqVOVk5OjTp06afny5YqMjJQk5eTkuD3zZt68eTp9+rTuu+8+3Xfffa72hIQEpaamVnf5AACgBvLqc268gefcAABqE55zc0ateM4NAADApUC4AQAAViHcAAAAqxBuAACAVQg3AADAKoQbAABgFcINAACwCuEGAABYhXADAACsQrgBAABWIdwAAACrEG4AAIBVCDcAAMAqhBsAAGAVwg0AALAK4QYAAFiFcAMAAKxCuAEAAFYh3AAAAKsQbgAAgFUINwAAwCqEGwAAYBXCDQAAsArhBgAAWIVwAwAArEK4AQAAViHcAAAAqxBuAACAVQg3AADAKoQbAABgFcINAACwCuEGAABYhXADAACsQrgBAABWIdwAAACrEG4AAIBVCDcAAMAqhBsAAGAVwg0AALAK4QYAAFiFcAMAAKxCuAEAAFYh3AAAAKsQbgAAgFUINwAAwCqEGwAAYBXCDQAAsArhBgAAWIVwAwAArEK4AQAAViHcAAAAqxBuAACAVQg3AADAKoQbAABgFcINAACwCuEGAABYhXADAACsQrgBAABWIdwAAACrEG4AAIBVCDcAAMAqXg83KSkpatWqlQIDAxUdHa3PP//8gv3XrFmj6OhoBQYG6qqrrtLcuXOrqVIAAFAbeDXcLF68WOPHj9djjz2mzMxM9e7dW4MGDVJ2dnaZ/Xfv3q3Bgwerd+/eyszM1KOPPqpx48bpgw8+qObKAQBATeUwxhhvHbxbt2669tprNWfOHFdbVFSUhg8fruTk5FL9H3nkES1dulTbtm1ztY0dO1abN2/Wl19+Wa5jFhQUyOl0Kj8/XyEhIZX/EgAAXEItJy3zdgkV9uP0IVW+z4r8fvtW+dHL6dSpU9q4caMmTZrk1h4bG6v169eXuc2XX36p2NhYt7YBAwbo9ddfV1FRkfz8/C5ZvQBwqdTGHy+gJvNauDl06JCKi4sVGhrq1h4aGqrc3Nwyt8nNzS2z/+nTp3Xo0CGFh4eX2qawsFCFhYWu9fz8fElnEuCl0GnKJ5dkvwAA1BaX4jf27D7Lc8LJa+HmLIfD4bZujCnVdrH+ZbWflZycrKeffrpUe4sWLSpaKgAAKAfnrEu376NHj8rpdF6wj9fCTZMmTeTj41NqliYvL6/U7MxZYWFhZfb39fVV48aNy9xm8uTJSkpKcq2XlJTo559/VuPGjS8YojxRUFCgFi1aaO/evVzPcwkxztWDca4ejHP1Yayrx6UaZ2OMjh49qoiIiIv29Vq48ff3V3R0tNLS0jRixAhXe1pamm655ZYyt+nevbs++ugjt7aVK1eqa9eu573eJiAgQAEBAW5tDRo0qFzxFxESEsJ/ONWAca4ejHP1YJyrD2NdPS7FOF9sxuYsr94KnpSUpPnz52vBggXatm2bJkyYoOzsbI0dO1bSmVmX+Ph4V/+xY8dqz549SkpK0rZt27RgwQK9/vrrmjhxore+AgAAqGG8es1NXFycDh8+rKlTpyonJ0edOnXS8uXLFRkZKUnKyclxe+ZNq1attHz5ck2YMEGzZ89WRESEXn75Zf3+97/31lcAAAA1jNcvKL733nt17733lvlZampqqbY+ffpo06ZNl7gqzwQEBGjKlCmlToOhajHO1YNxrh6Mc/VhrKtHTRhnrz7EDwAAoKp5/d1SAAAAVYlwAwAArEK4AQAAViHcAAAAqxBuKig5OVnXXXed6tevr6ZNm2r48OHavn27Wx9jjJ566ilFREQoKChIffv21Xfffeelimuni41zUVGRHnnkEf32t79V3bp1FRERofj4eB04cMCLVdc+5fl7/k933323HA6HZs2aVX1FWqK8Y71t2zYNGzZMTqdT9evX1w033OD2SAxcWHnG+dixY7r//vvVvHlzBQUFKSoqSnPmzPFSxbXTnDlzdPXVV7se1Ne9e3f961//cn3u7d9Bwk0FrVmzRvfdd582bNigtLQ0nT59WrGxsTp+/Lirz4wZMzRz5ky9+uqr+vrrrxUWFqaYmBgdPXrUi5XXLhcb5xMnTmjTpk164okntGnTJi1ZskQ7duzQsGHDvFx57VKev+ezPvzwQ/373/8u16PPUVp5xnrnzp3q1auX2rdvr9WrV2vz5s164oknFBgY6MXKa5fyjPOECRO0YsUKvfXWW64HyD7wwAP65z//6cXKa5fmzZtr+vTpysjIUEZGhm688UbdcsstrgDj9d9Bg0rJy8szksyaNWuMMcaUlJSYsLAwM336dFefkydPGqfTaebOneutMmu9c8e5LF999ZWRZPbs2VONldnlfOO8b98+06xZM/Ptt9+ayMhI8+KLL3qnQIuUNdZxcXFm1KhRXqzKPmWNc8eOHc3UqVPd+l177bXm8ccfr+7yrNKwYUMzf/78GvE7yMxNJeXn50uSGjVqJEnavXu3cnNzFRsb6+oTEBCgPn36aP369V6p0QbnjvP5+jgcjkv+7jCblTXOJSUluuOOO/Twww+rY8eO3irNOueOdUlJiZYtW6a2bdtqwIABatq0qbp166YPP/zQi1XWfmX9Tffq1UtLly7V/v37ZYxRenq6duzYoQEDBnirzFqtuLhY77zzjo4fP67u3bvXiN9Bwk0lGGOUlJSkXr16qVOnTpLkemv5uW82Dw0NLfVGc5RPWeN8rpMnT2rSpEkaOXIkL8Tz0PnG+bnnnpOvr6/GjRvnxersUtZY5+Xl6dixY5o+fboGDhyolStXasSIEbr11lu1Zs0aL1dcO53vb/rll19Whw4d1Lx5c/n7+2vgwIFKSUlRr169vFht7fPNN9+oXr16CggI0NixY/WPf/xDHTp0qBG/g15//UJtdv/992vLli364osvSn3mcDjc1o0xpdpQPhcaZ+nMxcW33367SkpKlJKSUs3V2aOscd64caNeeuklbdq0ib/fKlTWWJeUlEiSbrnlFk2YMEGS1KVLF61fv15z585Vnz59vFJrbXa+/3e8/PLL2rBhg5YuXarIyEitXbtW9957r8LDw3XTTTd5qdrap127dsrKytIvv/yiDz74QAkJCW5B3Ku/g9Vy8stC999/v2nevLnZtWuXW/vOnTuNJLNp0ya39mHDhpn4+PjqLNEK5xvns06dOmWGDx9urr76anPo0KFqrs4e5xvnF1980TgcDuPj4+NaJJk6deqYyMhI7xRby51vrAsLC42vr6955pln3Nr/+te/mh49elRniVY43zifOHHC+Pn5mY8//titPTEx0QwYMKA6S7RO//79zV/+8pca8TvIaakKMsbo/vvv15IlS/TZZ5+pVatWbp+3atVKYWFhSktLc7WdOnVKa9asUY8ePaq73FrrYuMsnZmxue222/TDDz9o1apVaty4sRcqrd0uNs533HGHtmzZoqysLNcSERGhhx9+WJ988omXqq6dLjbW/v7+uu6660rdtrxjxw5FRkZWZ6m12sXGuaioSEVFRapTx/3nz8fHxzV7Bs8YY1RYWFgzfgerJUJZ5J577jFOp9OsXr3a5OTkuJYTJ064+kyfPt04nU6zZMkS880335g//vGPJjw83BQUFHix8trlYuNcVFRkhg0bZpo3b26ysrLc+hQWFnq5+tqjPH/P5+JuKc+UZ6yXLFli/Pz8zGuvvWZ++OEH88orrxgfHx/z+eefe7Hy2qU849ynTx/TsWNHk56ebnbt2mUWLlxoAgMDTUpKihcrr10mT55s1q5da3bv3m22bNliHn30UVOnTh2zcuVKY4z3fwcJNxUkqcxl4cKFrj4lJSVmypQpJiwszAQEBJjf/e535ptvvvFe0bXQxcZ59+7d5+2Tnp7u1dprk/L8PZ+LcOOZ8o7166+/blq3bm0CAwNN586dzYcffuidgmup8oxzTk6OGT16tImIiDCBgYGmXbt25oUXXjAlJSXeK7yWGTNmjImMjDT+/v7miiuuMP3793cFG2O8/zvoMMaYSzs3BAAAUH245gYAAFiFcAMAAKxCuAEAAFYh3AAAAKsQbgAAgFUINwAAwCqEGwAAYBXCDYAa76mnnlKXLl28XYZHfvzxRzkcDmVlZXm7FOCyQbgBUC6jR4+Ww+GQw+GQr6+vrrzySt1zzz06cuSIt0urMQFi9OjRGj58uFdrAEC4AVABAwcOVE5Ojn788UfNnz9fH330ke69915vlwUAbgg3AMotICBAYWFhat68uWJjYxUXF6eVK1e69Vm4cKGioqIUGBio9u3bKyUlxe3zRx55RG3btlVwcLCuuuoqPfHEEyoqKnLrM336dIWGhqp+/fpKTEzUyZMnK1W3MUYzZszQVVddpaCgIHXu3Fnvv/++6/PVq1fL4XDo008/VdeuXRUcHKwePXqUekP3tGnT1LRpU9WvX1933XWXJk2a5Dpd9tRTT+mNN97QP//5T9cM1+rVq13b7tq1S/369VNwcLA6d+6sL7/8slLfCcAFVNtbrADUagkJCeaWW25xre/cudN06NDBhIaGutpee+01Ex4ebj744AOza9cu88EHH5hGjRqZ1NRUV59nnnnGrFu3zuzevdssXbrUhIaGmueee871+eLFi42/v7/529/+Zr7//nvz2GOPmfr165vOnTuft7azL1LNzMws8/NHH33UtG/f3qxYscLs3LnTLFy40AQEBJjVq1cbY4xJT083kky3bt3M6tWrzXfffWd69+5tevTo4drHW2+9ZQIDA82CBQvM9u3bzdNPP21CQkJcdR09etTcdtttZuDAgW5vqD9bW/v27c3HH39stm/fbv7whz+YyMhIU1RUVIF/AwDKi3ADoFwSEhKMj4+PqVu3rgkMDHS9bXnmzJmuPi1atDB///vf3bZ75plnTPfu3c+73xkzZpjo6GjXevfu3c3YsWPd+nTr1s3jcHPs2DETGBho1q9f79aemJho/vjHPxpj/i/crFq1yvX5smXLjCTz66+/umq477773PbRs2dPt7rODYD/Wdv8+fNdbd99952RZLZt23be7wTAc5yWAlBu/fr1U1ZWlv7973/rgQce0IABA/TAAw9Ikn766Sft3btXiYmJqlevnmuZNm2adu7c6drH+++/r169eiksLEz16tXTE088oezsbNfn27ZtU/fu3d2Oe+56RWzdulUnT55UTEyMW11vvvmmW12SdPXVV7v+OTw8XJKUl5cnSdq+fbuuv/56t/7nrl/IhfYNoGr5ersAALVH3bp11bp1a0nSyy+/rH79+unpp5/WM888o5KSEknS3/72N3Xr1s1tOx8fH0nShg0bdPvtt+vpp5/WgAED5HQ69c477+iFF164ZDWfrWvZsmVq1qyZ22cBAQFu635+fq5/djgcbtv/Z9tZxphy13GxfQOoOoQbAB6bMmWKBg0apHvuuUcRERFq1qyZdu3apT/96U9l9l+3bp0iIyP12GOPudr27Nnj1icqKkobNmxQfHy8q23Dhg0e19ihQwcFBAQoOztbffr08Xg/7dq101dffaU77rjD1ZaRkeHWx9/fX8XFxR4fA0DVINwA8Fjfvn3VsWNHPfvss3r11Vf11FNPady4cQoJCdGgQYNUWFiojIwMHTlyRElJSWrdurWys7P1zjvv6LrrrtOyZcv0j3/8w22fDz74oBISEtS1a1f16tVLixYt0nfffaerrrrqovWce3eTdCbcTJw4URMmTFBJSYl69eqlgoICrV+/XvXq1VNCQkK5vusDDzygP//5z+ratat69OihxYsXa8uWLW51tWzZUp988om2b9+uxo0by+l0lmvfAKqYty/6AVA7lHWxrDHGLFq0yPj7+5vs7GzXepcuXYy/v79p2LCh+d3vfmeWLFni6v/www+bxo0bm3r16pm4uDjz4osvGqfT6bbP//7v/zZNmjQx9erVMwkJCeavf/1ruS4oLmvZvXu3KSkpMS+99JJp166d8fPzM1dccYUZMGCAWbNmjTHm/y4oPnLkiGufmZmZru3Pmjp1qquuMWPGmHHjxpkbbrjB9XleXp6JiYkx9erVM5JMenp6mRc7HzlyxPU5gKrnMKYCJ40BAC4xMTEKCwvT//7v/3q7FAD/gdNSAFAOJ06c0Ny5czVgwAD5+Pjo7bff1qpVq5SWlubt0gCcg5kbACiHX3/9VUOHDtWmTZtUWFiodu3a6fHHH9ett97q7dIAnINwAwAArMJD/AAAgFUINwAAwCqEGwAAYBXCDQAAsArhBgAAWIVwAwAArEK4AQAAViHcAAAAqxBuAACAVf4/IhkZ31RIMS8AAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "reads = !xargs zcat {out_path}/*.gz | seqtk seq -A  |grep -v \">\" \n",
    "ds = [len(read) for read in reads]\n",
    "make_histogram(ds, \"Read Length\", \"Trimmed Reads\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf68d82c-0200-4613-b5b6-497fde519e16",
   "metadata": {},
   "source": [
    "## Step: fastqc_post"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8198a2d4-8f2f-46de-ba05-5b335a21bc2e",
   "metadata": {},
   "source": [
    "#### Post Trimming Quality Control"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "140ea0f2-cb1c-48a6-adbd-5264b92c8e1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_path = mkpath(\"trim\")\n",
    "out_path = mkpath(\"fastqc_post\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "aa72c0a9-9280-4255-b5d6-d0d0258181c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n",
      "Analysis complete for t1.fq.gz\n",
      "application/gzip\n",
      "Analysis complete for t2.fq.gz\n",
      "application/gzip\n",
      "Analysis complete for t3.fq.gz\n"
     ]
    }
   ],
   "source": [
    "for sample in samples:\n",
    "    ! fastqc {fname(in_path,sample,\"fq.gz\")} -o {out_path} 2> /dev/null"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69de1cf7-fa1b-49f1-ba6a-a14126989904",
   "metadata": {},
   "source": [
    "#### Consolidate fastqc Reports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "e751dc92-97ee-4f4e-af15-6d05c1fb49ce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  \u001b[91m///\u001b[0m \u001b]8;id=673316;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ 🔍 \u001b[2m| v1.19\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | Only using modules: fastqc\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /home/cdaniels/uofc_data/ubs_seq/UBS-seq_basic/workspace/fastqc_post\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m6/6\u001b[0m  0/6\u001b[0m  \n",
      "\u001b[?25h\u001b[34m|            fastqc\u001b[0m | Found 3 reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : fastqc_post/multiqc.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : fastqc_post/multiqc_data\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "!multiqc -f -fp -m fastqc -n multiqc -o {out_path} {out_path}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "bc5b1d72-84a7-4fd3-a8ae-e755b5015a52",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "multiqc_data  t1_fastqc.html  t2_fastqc.html  t3_fastqc.html\n",
      "multiqc.html  t1_fastqc.zip   t2_fastqc.zip   t3_fastqc.zip\n"
     ]
    }
   ],
   "source": [
    "! ls {out_path}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd222f8b-c131-4a5a-b865-d59c1c5879e7",
   "metadata": {},
   "source": [
    "## Step: Hisat3n_align"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa201443-78ec-4c4b-8e68-83f01a3b01f3",
   "metadata": {},
   "source": [
    "#### Align Samples to Genome with Hisat-3n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "80a6c377-5017-4865-aac4-eacc555b7345",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_path = mkpath(\"trim\")\n",
    "out_path = mkpath(\"hisat3n_align\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fba7ae0-92e6-4435-a5f9-bf6610ba7101",
   "metadata": {},
   "source": [
    "#### hisat-3n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41f584e2-8878-4cf9-91c9-49514d1a8ae7",
   "metadata": {},
   "source": [
    "http://daehwankimlab.github.io/hisat2/hisat-3n/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "98b74e19-5d09-491f-b605-8461a0b95ac4",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HISAT2 version 2.2.1-3n-0.0.3 by Daehwan Kim (infphilo@gmail.com, www.ccb.jhu.edu/people/infphilo)\n",
      "Usage: \n",
      "  hisat2 [options]* -x <ht2-idx> {-1 <m1> -2 <m2> | -U <r>} [-S <sam>]\n",
      "\n",
      "  <ht2-idx>  Index filename prefix (minus trailing .X.ht2).\n",
      "  <m1>       Files with #1 mates, paired with files in <m2>.\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\n",
      "  <m2>       Files with #2 mates, paired with files in <m1>.\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\n",
      "  <r>        Files with unpaired reads.\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\n",
      "  <sam>      File for SAM output (default: stdout)\n",
      "\n",
      "  <m1>, <m2>, <r> can be comma-separated lists (no whitespace) and can be\n",
      "  specified many times.  E.g. '-U file1.fq,file2.fq -U file3.fq'.\n",
      "\n",
      "Options (defaults in parentheses):\n",
      "\n",
      " Input:\n",
      "  -q                 query input files are FASTQ .fq/.fastq (default)\n",
      "  --qseq             query input files are in Illumina's qseq format\n",
      "  -f                 query input files are (multi-)FASTA .fa/.mfa\n",
      "  -r                 query input files are raw one-sequence-per-line\n",
      "  -c                 <m1>, <m2>, <r> are sequences themselves, not files\n",
      "  -s/--skip <int>    skip the first <int> reads/pairs in the input (none)\n",
      "  -u/--upto <int>    stop after first <int> reads/pairs (no limit)\n",
      "  -5/--trim5 <int>   trim <int> bases from 5'/left end of reads (0)\n",
      "  -3/--trim3 <int>   trim <int> bases from 3'/right end of reads (0)\n",
      "  --phred33          qualities are Phred+33 (default)\n",
      "  --phred64          qualities are Phred+64\n",
      "  --int-quals        qualities encoded as space-delimited integers\n",
      "\n",
      " Presets:                 Same as:\n",
      "   --fast                 --no-repeat-index\n",
      "   --sensitive            --bowtie2-dp 1 -k 30 --score-min L,0,-0.5\n",
      "   --very-sensitive       --bowtie2-dp 2 -k 50 --score-min L,0,-1\n",
      "\n",
      " Alignment:\n",
      "  --bowtie2-dp <int> use Bowtie2's dynamic programming alignment algorithm (0) - 0: no dynamic programming, 1: conditional dynamic programming, and 2: unconditional dynamic programming (slowest)\n",
      "  --n-ceil <func>    func for max # non-A/C/G/Ts permitted in aln (L,0,0.15)\n",
      "  --ignore-quals     treat all quality values as 30 on Phred scale (off)\n",
      "  --nofw             do not align forward (original) version of read (off)\n",
      "  --norc             do not align reverse-complement version of read (off)\n",
      "  --no-repeat-index  do not use repeat index\n",
      "\n",
      " 3N-Alignment:\n",
      "  --base-change <chr,chr>     the converted nucleotide and converted to nucleotide (C,T)\n",
      "  --directional-mapping       make directional mapping, please use this option only if your reads are prepared with a strand specific library (off)\n",
      "  --repeat-limit <int>        maximum number of repeat will be expanded for repeat alignment (1000)\n",
      "  --unique-only               only output the reads have unique alignment (off)\n",
      "\n",
      " Spliced Alignment:\n",
      "  --pen-cansplice <int>              penalty for a canonical splice site (0)\n",
      "  --pen-noncansplice <int>           penalty for a non-canonical splice site (12)\n",
      "  --pen-canintronlen <func>          penalty for long introns (G,-8,1) with canonical splice sites\n",
      "  --pen-noncanintronlen <func>       penalty for long introns (G,-8,1) with noncanonical splice sites\n",
      "  --min-intronlen <int>              minimum intron length (20)\n",
      "  --max-intronlen <int>              maximum intron length (500000)\n",
      "  --known-splicesite-infile <path>   provide a list of known splice sites\n",
      "  --novel-splicesite-outfile <path>  report a list of splice sites\n",
      "  --novel-splicesite-infile <path>   provide a list of novel splice sites\n",
      "  --no-temp-splicesite               disable the use of splice sites found\n",
      "  --no-spliced-alignment             disable spliced alignment\n",
      "  --rna-strandness <string>          specify strand-specific information (unstranded)\n",
      "  --tmo                              reports only those alignments within known transcriptome\n",
      "  --dta                              reports alignments tailored for transcript assemblers\n",
      "  --dta-cufflinks                    reports alignments tailored specifically for cufflinks\n",
      "  --avoid-pseudogene                 tries to avoid aligning reads to pseudogenes (experimental option)\n",
      "  --no-templatelen-adjustment        disables template length adjustment for RNA-seq reads\n",
      "\n",
      " Scoring:\n",
      "  --mp <int>,<int>   max and min penalties for mismatch; lower qual = lower penalty <6,2>\n",
      "  --sp <int>,<int>   max and min penalties for soft-clipping; lower qual = lower penalty <2,1>\n",
      "  --no-softclip      no soft-clipping\n",
      "  --np <int>         penalty for non-A/C/G/Ts in read/ref (1)\n",
      "  --rdg <int>,<int>  read gap open, extend penalties (5,3)\n",
      "  --rfg <int>,<int>  reference gap open, extend penalties (5,3)\n",
      "  --score-min <func> min acceptable alignment score w/r/t read length\n",
      "                     (L,0.0,-0.2)\n",
      "\n",
      " Reporting:\n",
      "  -k <int>           It searches for at most <int> distinct, primary alignments for each read. Primary alignments mean \n",
      "                     alignments whose alignment score is equal to or higher than any other alignments. The search terminates \n",
      "                     when it cannot find more distinct valid alignments, or when it finds <int>, whichever happens first. \n",
      "                     The alignment score for a paired-end alignment equals the sum of the alignment scores of \n",
      "                     the individual mates. Each reported read or pair alignment beyond the first has the SAM ‘secondary’ bit \n",
      "                     (which equals 256) set in its FLAGS field. For reads that have more than <int> distinct, \n",
      "                     valid alignments, hisat2 does not guarantee that the <int> alignments reported are the best possible \n",
      "                     in terms of alignment score. Default: 5 (linear index) or 10 (graph index).\n",
      "                     Note: HISAT2 is not designed with large values for -k in mind, and when aligning reads to long, \n",
      "                     repetitive genomes, large -k could make alignment much slower.\n",
      "  --max-seeds <int>  HISAT2, like other aligners, uses seed-and-extend approaches. HISAT2 tries to extend seeds to \n",
      "                     full-length alignments. In HISAT2, --max-seeds is used to control the maximum number of seeds that \n",
      "                     will be extended. For DNA-read alignment (--no-spliced-alignment), HISAT2 extends up to these many seeds\n",
      "                     and skips the rest of the seeds. For RNA-read alignment, HISAT2 skips extending seeds and reports \n",
      "                     no alignments if the number of seeds is larger than the number specified with the option, \n",
      "                     to be compatible with previous versions of HISAT2. Large values for --max-seeds may improve alignment \n",
      "                     sensitivity, but HISAT2 is not designed with large values for --max-seeds in mind, and when aligning \n",
      "                     reads to long, repetitive genomes, large --max-seeds could make alignment much slower. \n",
      "                     The default value is the maximum of 5 and the value that comes with -k times 2.\n",
      "  -a/--all           HISAT2 reports all alignments it can find. Using the option is equivalent to using both --max-seeds \n",
      "                     and -k with the maximum value that a 64-bit signed integer can represent (9,223,372,036,854,775,807).\n",
      "  --repeat           report alignments to repeat sequences directly\n",
      "\n",
      " Paired-end:\n",
      "  -I/--minins <int>  minimum fragment length (0), only valid with --no-spliced-alignment\n",
      "  -X/--maxins <int>  maximum fragment length (500), only valid with --no-spliced-alignment\n",
      "  --fr/--rf/--ff     -1, -2 mates align fw/rev, rev/fw, fw/fw (--fr)\n",
      "  --no-mixed         suppress unpaired alignments for paired reads\n",
      "  --no-discordant    suppress discordant alignments for paired reads\n",
      "\n",
      " Output:\n",
      "  -t/--time          print wall-clock time taken by search phases\n",
      "  --un <path>           write unpaired reads that didn't align to <path>\n",
      "  --al <path>           write unpaired reads that aligned at least once to <path>\n",
      "  --un-conc <path>      write pairs that didn't align concordantly to <path>\n",
      "  --al-conc <path>      write pairs that aligned concordantly at least once to <path>\n",
      "  (Note: for --un, --al, --un-conc, or --al-conc, add '-gz' to the option name, e.g.\n",
      "  --un-gz <path>, to gzip compress output, or add '-bz2' to bzip2 compress output.)\n",
      "  --summary-file <path> print alignment summary to this file.\n",
      "  --new-summary         print alignment summary in a new style, which is more machine-friendly.\n",
      "  --quiet               print nothing to stderr except serious errors\n",
      "  --met-file <path>     send metrics to file at <path> (off)\n",
      "  --met-stderr          send metrics to stderr (off)\n",
      "  --met <int>           report internal counters & metrics every <int> secs (1)\n",
      "  --no-head             suppress header lines, i.e. lines starting with @\n",
      "  --no-sq               suppress @SQ header lines\n",
      "  --rg-id <text>        set read group id, reflected in @RG line and RG:Z: opt field\n",
      "  --rg <text>           add <text> (\"lab:value\") to @RG line of SAM header.\n",
      "                        Note: @RG line only printed when --rg-id is set.\n",
      "  --omit-sec-seq        put '*' in SEQ and QUAL fields for secondary alignments.\n",
      "\n",
      " Performance:\n",
      "  -o/--offrate <int> override offrate of index; must be >= index's offrate\n",
      "  -p/--threads <int> number of alignment threads to launch (1)\n",
      "  --reorder          force SAM output order to match order of input reads\n",
      "  --mm               use memory-mapped I/O for index; many 'hisat2's can share\n",
      "\n",
      " Other:\n",
      "  --qc-filter        filter out reads that are bad according to QSEQ filter\n",
      "  --seed <int>       seed for random number generator (0)\n",
      "  --non-deterministic seed rand. gen. arbitrarily instead of using read attributes\n",
      "  --remove-chrname   remove 'chr' from reference names in alignment\n",
      "  --add-chrname      add 'chr' to reference names in alignment \n",
      "  --version          print version information and quit\n",
      "  -h/--help          print this usage message\n"
     ]
    }
   ],
   "source": [
    "! hisat-3n --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1833004-ab6e-43a1-93d9-ae7ce2b17aa3",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "```\n",
    "hisat3n --index {params.index}\n",
    "        -p {threads}\n",
    "        --summary-file {output.summary}\n",
    "        --new-summary\n",
    "        -q\n",
    "        -U {input}\n",
    "        --directional-mapping\n",
    "        --base-change C,T\n",
    "        --pen-noncansplice 20\n",
    "        --mp 4,1\n",
    "        --un {output.fq}\n",
    "        -S {output.sam}\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12ef71a2-1e8b-4c06-9953-bad3b25b49e0",
   "metadata": {},
   "source": [
    "##### paramaters:\n",
    "- --index {params.index}, Index filename prefix (minus trailing .X.ht2) \n",
    "- -p {nc}, number of threads\n",
    "- --summary-file, print alignment summary to this file.\n",
    "- --new-summary, print alignment summary in a new style, which is more machine-friendly.\n",
    "- -q, query input files are FASTQ .fq/.fastq (default)\n",
    "- -U {input}, \n",
    "- --directional-mapping, make directional mapping, please use this option only if your reads are prepared with a strand specific library (off)\n",
    "- --base-change C,T,  the converted nucleotide and converted to nucleotide (C,T)\n",
    "- --pen-noncansplice 20, penalty for a non-canonical splice site (12)\n",
    "- --mp 4,1,  max and min penalties for mismatch; lower qual = lower penalty <6,2>\n",
    "- --un {output.fq}, write unpaired reads that didn't align to <path> \n",
    "- -S {output.sam}, File for SAM output (default: stdout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "5cdd8492-fa94-4434-9881-9231dba08130",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HISAT2 summary stats:\n",
      "\tTotal reads: 4652729\n",
      "\t\tAligned 0 time: 813137 (17.48%)\n",
      "\t\tAligned 1 time: 2714806 (58.35%)\n",
      "\t\tAligned >1 times: 1124786 (24.17%)\n",
      "\tOverall alignment rate: 82.52%\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 6780142\n",
      "\t\tAligned 0 time: 1247996 (18.41%)\n",
      "\t\tAligned 1 time: 3917003 (57.77%)\n",
      "\t\tAligned >1 times: 1615143 (23.82%)\n",
      "\tOverall alignment rate: 81.59%\n",
      "HISAT2 summary stats:\n",
      "\tTotal reads: 5876778\n",
      "\t\tAligned 0 time: 1077999 (18.34%)\n",
      "\t\tAligned 1 time: 3405853 (57.95%)\n",
      "\t\tAligned >1 times: 1392926 (23.70%)\n",
      "\tOverall alignment rate: 81.66%\n"
     ]
    }
   ],
   "source": [
    "for sample in samples:\n",
    "    !hisat-3n --index {genome_idx}\\\n",
    "        -p {nc}\\\n",
    "        --summary-file {fname(out_path,sample,\"summary\")}\\\n",
    "        --new-summary\\\n",
    "        -q\\\n",
    "        -U {fname(in_path,sample,'fq.gz')}\\\n",
    "        --directional-mapping\\\n",
    "        --base-change C,T\\\n",
    "        --pen-noncansplice 20\\\n",
    "        --mp 4,1\\\n",
    "        --un {fname(out_path,sample,'unmapped.fq')}\\\n",
    "        -S {fname(out_path,sample,'sam')}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7674f153-02d5-47b8-aaf1-f577584fe698",
   "metadata": {},
   "source": [
    "#### Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "687fd6be-2a5b-49a0-b6a1-af382ed2f3ae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 6.9G\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 1.8G Feb  3 04:23 t1.sam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  182 Feb  3 04:23 t1.summary\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  91M Feb  3 04:23 t1.unmapped.fq\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 2.6G Feb  3 04:25 t2.sam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  183 Feb  3 04:25 t2.summary\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 139M Feb  3 04:25 t2.unmapped.fq\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 2.3G Feb  3 04:26 t3.sam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  183 Feb  3 04:26 t3.summary\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 120M Feb  3 04:26 t3.unmapped.fq\n"
     ]
    }
   ],
   "source": [
    "!ls -lh {out_path}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "0452ab50-d295-493d-912e-41cc7115943a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HISAT2 summary stats:\n",
      "\tTotal reads: 4652729\n",
      "\t\tAligned 0 time: 813137 (17.48%)\n",
      "\t\tAligned 1 time: 2714806 (58.35%)\n",
      "\t\tAligned >1 times: 1124786 (24.17%)\n",
      "\tOverall alignment rate: 82.52%\n"
     ]
    }
   ],
   "source": [
    "! cat {out_path}/t1.summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "a9a39eee-1fd7-4304-b980-bbb404d785df",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@SRR23538290.540_TTCAACACTC 540 length=121\n",
      "GTAATGAAGGGTAAGATGAAGTGAAAGGTA\n",
      "+\n",
      "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\n",
      "@SRR23538290.203_GGCTAGGGGG 203 length=121\n",
      "TGTTGAAAATGAGATTAAGATAGTGTA\n",
      "+\n",
      "FFFFFFFF,:FFFFFFFFFFFFFF,FF\n"
     ]
    }
   ],
   "source": [
    "! head -8 {out_path}/t1.unmapped.fq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "f7964202-9104-41f9-b23b-d6d841c66995",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR23538290.540_TTCAACACTC\t4\t*\t0\t0\t*\t*\t0\t0\tGTAATGAAGGGTAAGATGAAGTGAAAGGTA\tFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\tYT:Z:UU\n",
      "grep: write error: Broken pipe\n"
     ]
    }
   ],
   "source": [
    "! grep -v '@' {out_path}/t1.sam |head -1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a9a42a0-9bf2-44c4-b559-e40d5818bc6a",
   "metadata": {},
   "source": [
    "## Step: Hisat3n_sort"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "640c2ff5-df93-4afb-8b5d-35106e520ece",
   "metadata": {},
   "source": [
    "#### Sort and Index Hisat3n Sam Files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "2ae1a883-efc0-4773-b0c2-e7cfeaf9357e",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_path = mkpath(\"hisat3n_align\")\n",
    "out_path = mkpath(\"hisat3n_sort\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9c040e0f-93a0-49f3-a085-036a0c86c3c2",
   "metadata": {},
   "source": [
    "#### samtools"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2f487d0-1389-4878-a5d9-77bf3f1d66dd",
   "metadata": {},
   "source": [
    "http://www.htslib.org/doc/samtools.html"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b659402-522a-47c6-89f9-79c62a7f0e02",
   "metadata": {
    "scrolled": true
   },
   "source": [
    "```\n",
    "samtools view\n",
    "    -@ {threads}\n",
    "    -F4 -b {input} |\n",
    "samtools sort\n",
    "    -@ {threads}\n",
    "    --write-index\n",
    "    -m 4G\n",
    "    -O BAM\n",
    "    -o {output} -\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1431f54e-5483-4b0e-8c17-08957190476f",
   "metadata": {},
   "source": [
    "##### paramaters:\n",
    "```\n",
    "samtools view, SAM<->BAM<->CRAM conversion\n",
    "    -@ {nc}, number of threads \n",
    "    -F4, have none of the FLAGs present (-F 4 filters out unmapped reads) \n",
    "    -b, output a bam file\n",
    "    {input}  |\n",
    "samtools sort, sort alignment file\n",
    "    -@ {nc}, number of threads\n",
    "    --write-index, index the output files\n",
    "    -O BAM, output file format\n",
    "    -o {output} -\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "629253cc-154d-4d71-8448-a0bfea40dd18",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[bam_sort_core] merging from 0 files and 32 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 32 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 32 in-memory blocks...\n"
     ]
    }
   ],
   "source": [
    "for sample in samples:\n",
    "    ! samtools view -@ {nc} -F4 -b {fname(in_path, sample,'sam')}  | \\\n",
    "      samtools sort -@ {nc} --write-index -O BAM -o {fname(out_path,sample,'bam')} - "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "211047f3-5203-4e21-b259-7746b03a936b",
   "metadata": {},
   "source": [
    "#### Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "1140e2ed-32b3-42f7-a115-5e6cb3175653",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 784M\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 215M Feb  3 04:26 t1.bam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 975K Feb  3 04:26 t1.bam.csi\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 301M Feb  3 04:26 t2.bam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 897K Feb  3 04:26 t2.bam.csi\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 265M Feb  3 04:27 t3.bam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 927K Feb  3 04:27 t3.bam.csi\n"
     ]
    }
   ],
   "source": [
    "! ls -lh {out_path}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74ad4df6-1fa1-43dc-8302-859affe0690f",
   "metadata": {},
   "source": [
    "## Step: Hisat3n_dedup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad8d9bf1-95a8-4198-988c-4cc83db22bad",
   "metadata": {},
   "source": [
    "#### Remove Dulpicate Reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "4d6761e7-18a1-48a7-a9a7-17e32f5eec65",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_path = mkpath(\"hisat3n_sort\")\n",
    "out_path = mkpath(\"hisat3n_dedup\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e58c2685-80da-428c-bdfc-9447ebfc73fe",
   "metadata": {},
   "source": [
    "#### umicollapse"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d0f9db8-aaf5-4ee2-9c4f-97e79da12a0a",
   "metadata": {},
   "source": [
    "https://github.com/Daniel-Liu-c0deb0t/UMICollapse"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e09e2177-b96c-495b-b56c-c1b4689f0634",
   "metadata": {},
   "source": [
    "##### paramaters:\n",
    "```\n",
    "umicollapse bam, use a bam file\n",
    "    --two-pass, use a separate two-pass algorithm for SAM/BAM deduplication.\n",
    "    -i {input.bam}, indexed input bam file\n",
    "    -o {output.bam}, output bam file\n",
    "      > {output.log}\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "1af0f53c-3277-4745-8a64-63283360815c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Picked up _JAVA_OPTIONS: -Xmx8g\n",
      "INFO\t2024-02-03 04:27:24\tSortingCollection\tCreating merging iterator from 16 files\n",
      "Picked up _JAVA_OPTIONS: -Xmx8g\n",
      "INFO\t2024-02-03 04:33:11\tSortingCollection\tCreating merging iterator from 23 files\n",
      "Picked up _JAVA_OPTIONS: -Xmx8g\n",
      "INFO\t2024-02-03 04:39:56\tSortingCollection\tCreating merging iterator from 20 files\n"
     ]
    }
   ],
   "source": [
    "for sample in samples:\n",
    "    !umicollapse bam  \\\n",
    "    -i {fname(in_path,sample,'bam')} \\\n",
    "    -o {fname(out_path,sample,'bam')}\\\n",
    "    >  {fname(out_path,sample,'log')}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f6fe90e-c8b6-49c6-b21c-e0a0409f37de",
   "metadata": {},
   "source": [
    "#### Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "a52dc069-dd33-4659-9d68-6824b311f096",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 794M\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 219M Feb  3 04:27 t1.bam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  452 Feb  3 04:27 t1.log\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 306M Feb  3 04:39 t2.bam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  455 Feb  3 04:39 t2.log\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 270M Feb  3 04:40 t3.bam\n",
      "-rw-rw-r-- 1 cdaniels cdaniels  450 Feb  3 04:40 t3.log\n"
     ]
    }
   ],
   "source": [
    "! ls -lh {out_path}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "fd112fd3-6634-4648-9a0a-7a44a8efb7cf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Arguments\t[bam, -i, hisat3n_sort/t1.bam, -o, hisat3n_dedup/t1.bam]\n",
      "Done reading input file into memory!\n",
      "Number of input reads\t8319859\n",
      "Number of removed unmapped reads\t0\n",
      "Number of unremoved reads\t8319859\n",
      "Number of unique alignment positions\t4385853\n",
      "Average number of UMIs per alignment position\t1.8292589833722197\n",
      "Max number of UMIs over all alignment positions\t357\n",
      "Number of reads after deduplicating\t7946319\n",
      "UMI collapsing finished in 40.323 seconds!\n"
     ]
    }
   ],
   "source": [
    "! cat {out_path}/t1.log"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27f954cc-df3e-4048-ae0e-442526f0b5cf",
   "metadata": {},
   "source": [
    "## Step: Hisat3n_call"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c45dda76-3ea1-4749-a78e-792080774b53",
   "metadata": {},
   "source": [
    "#### Call Converted bases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "00cab6a3-dba0-42e6-8e9e-8d49894eac70",
   "metadata": {},
   "outputs": [],
   "source": [
    "in_path = mkpath(\"hisat3n_dedup\")\n",
    "out_path = mkpath(\"hisat3n_call\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ac3e272-df75-4b04-85b8-64b7df394d19",
   "metadata": {},
   "source": [
    "#### hisat-3n-table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d208fba5-fc27-4274-b416-92b00a6a22da",
   "metadata": {},
   "source": [
    "http://daehwankimlab.github.io/hisat2/hisat-3n/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa8ac256-7407-4efe-9011-3209695d1a87",
   "metadata": {},
   "source": [
    "##### paramaters:\n",
    "```\n",
    "samtools view -e \"rlen<100000\" -h {input} |\n",
    "hisat3ntable\n",
    "    -p {threads}\n",
    "    -m --alignments -\n",
    "    --ref {params.fa}\n",
    "    --output-name /dev/stdout\n",
    "    --base-change C,T                      |\n",
    "    bgzip -@ {threads} -c > {output}\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "7e03594a-6fb6-4129-a88c-666dddc9ebf0",
   "metadata": {},
   "outputs": [],
   "source": [
    "for sample in samples:\n",
    "    !samtools view -e \"rlen<100000\" -h {fname(in_path,sample,'bam')} |\\\n",
    "    hisat-3n-table\\\n",
    "        -p {nc}\\\n",
    "        -m --alignments -\\\n",
    "        --ref {genome_fa}\\\n",
    "        --output-name /dev/stdout\\\n",
    "        --base-change C,T                          |\\\n",
    "    bgzip \\\n",
    "        -@ {nc} \\\n",
    "        -c > {fname(out_path,sample,'tsv.gz')}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "119803f7-244e-4e46-aac7-d5be7f71e35d",
   "metadata": {},
   "source": [
    "#### Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "df266208-387b-49fa-b24b-43172430e981",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 91M\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 54K Feb  3 05:05 t1_filtered.csv\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 26M Feb  3 04:48 t1.tsv.gz\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 35M Feb  3 04:56 t2.tsv.gz\n",
      "-rw-rw-r-- 1 cdaniels cdaniels 31M Feb  3 05:05 t3.tsv.gz\n"
     ]
    }
   ],
   "source": [
    "! ls -lh {out_path}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f423b27-52c9-4d7f-b8fb-d9a5b272109b",
   "metadata": {},
   "source": [
    "There are 7 columns in the 3N-conversion-table:\n",
    "\n",
    "1. `ref:` the chromosome name.\n",
    "2. `pos:` 1-based position in ref.\n",
    "3. `strand:` ‘+’ for forward strand. ‘-‘ for reverse strand.\n",
    "4. `convertedBaseQualities:` the qualities for converted base in read-level measurement. Length of this string is equal to the number of converted Base in read-level measurement.\n",
    "5. `convertedBaseCount:` number of distinct read positions where converted base in read-level measurements were found. this number should equal to the length of convertedBaseQualities.\n",
    "6. `unconvertedBaseQualities:` the qualities for unconverted base in read-level measurement. Length of this string is equal to the number of unconverted Base in read-level measurement.\n",
    "7. `unconvertedBaseCount:` number of distinct read positions where unconverted base in read-level measurements were found. this number should equal to the length of unconvertedBaseQualities."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "d85c2597-99cb-4297-a05a-560482cf89a9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ref\tpos\tstrand\tconvertedBaseQualities\tconvertedBaseCount\tunconvertedBaseQualities\tunconvertedBaseCount\n",
      "1\t10567\t-\tF\t1\t\t0\n",
      "1\t10572\t-\tF\t1\t\t0\n",
      "1\t10578\t-\tF\t1\t\t0\n",
      "1\t10580\t-\tF\t1\t\t0\n",
      "1\t10581\t-\tF\t1\t\t0\n",
      "1\t10583\t-\tF\t1\t\t0\n",
      "1\t10590\t-\tF\t1\t\t0\n",
      "1\t10591\t-\tF\t1\t\t0\n",
      "1\t10592\t-\tF\t1\t\t0\n",
      "1\t10596\t-\tF\t1\t\t0\n",
      "1\t11491\t+\tF\t1\t\t0\n",
      "1\t11500\t+\tF\t1\t\t0\n",
      "1\t11502\t+\t:\t1\t\t0\n",
      "1\t11509\t+\t\t0\tF\t1\n",
      "1\t11511\t+\t\t0\tF\t1\n",
      "1\t12558\t+\tF\t1\t\t0\n",
      "1\t12563\t+\tF\t1\t\t0\n",
      "1\t12566\t+\tF\t1\t\t0\n",
      "1\t12581\t+\tF\t1\t\t0\n",
      "\n",
      "gzip: stdout: Broken pipe\n"
     ]
    }
   ],
   "source": [
    "! zcat {out_path}/t1.tsv.gz |head -20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "e7a7f5ce-b1a1-4efc-a283-2a22d320757d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(out_path/'t1.tsv.gz', sep='\\t', compression='gzip', low_memory=False, dtype={4: int, 6: int })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "id": "c13e256f-8f53-4c01-bb42-cf3897e33ef6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.017151172450944555"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p1 = len(df[df.unconvertedBaseCount == 1])/len(df);\n",
    "p1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "1e90d13a-6e1c-478c-9aec-8871eed9d944",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.002095434317852818"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p2 = len(df[df.unconvertedBaseCount == 2])/len(df);\n",
    "p2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "52e951dd-b811-4f81-9164-0fbec5e912d0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0007013542608583107"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p3 = len(df[df.unconvertedBaseCount == 3])/len(df);\n",
    "p3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "f96a8943-7119-4ba6-acb5-13605032b82c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.0002941627164420395"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p1*p1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "0a1a5886-0cfc-44c5-b694-f557598329c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2 = df[(df['unconvertedBaseCount'] > 1) &  (df['unconvertedBaseCount'] > df['convertedBaseCount'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "ecc3b924-2169-4742-bf55-8d4a16bc4745",
   "metadata": {},
   "outputs": [],
   "source": [
    "df2.to_csv(out_path/'t1_filtered.csv',index=False)  # Set index=False if you don't want to save row indices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "e04f6302-0013-4961-9aba-e5ee40bf5044",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref</th>\n",
       "      <th>pos</th>\n",
       "      <th>strand</th>\n",
       "      <th>convertedBaseQualities</th>\n",
       "      <th>convertedBaseCount</th>\n",
       "      <th>unconvertedBaseQualities</th>\n",
       "      <th>unconvertedBaseCount</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6386</th>\n",
       "      <td>1</td>\n",
       "      <td>629231</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>FF</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6388</th>\n",
       "      <td>1</td>\n",
       "      <td>629236</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>FFF</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6389</th>\n",
       "      <td>1</td>\n",
       "      <td>629238</td>\n",
       "      <td>+</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>FFF</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6390</th>\n",
       "      <td>1</td>\n",
       "      <td>629241</td>\n",
       "      <td>+</td>\n",
       "      <td>F</td>\n",
       "      <td>1</td>\n",
       "      <td>F:FFF</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6391</th>\n",
       "      <td>1</td>\n",
       "      <td>629243</td>\n",
       "      <td>+</td>\n",
       "      <td>F</td>\n",
       "      <td>1</td>\n",
       "      <td>FFFFF</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6392</th>\n",
       "      <td>1</td>\n",
       "      <td>629246</td>\n",
       "      <td>+</td>\n",
       "      <td>F</td>\n",
       "      <td>1</td>\n",
       "      <td>FFFFF</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6394</th>\n",
       "      <td>1</td>\n",
       "      <td>629256</td>\n",
       "      <td>+</td>\n",
       "      <td>F</td>\n",
       "      <td>1</td>\n",
       "      <td>:FFF</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6395</th>\n",
       "      <td>1</td>\n",
       "      <td>629258</td>\n",
       "      <td>+</td>\n",
       "      <td>F</td>\n",
       "      <td>1</td>\n",
       "      <td>FFFF</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6396</th>\n",
       "      <td>1</td>\n",
       "      <td>629259</td>\n",
       "      <td>+</td>\n",
       "      <td>F</td>\n",
       "      <td>1</td>\n",
       "      <td>,FFF</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     ref     pos strand convertedBaseQualities  convertedBaseCount  \\\n",
       "6386   1  629231      +                    NaN                   0   \n",
       "6388   1  629236      +                    NaN                   0   \n",
       "6389   1  629238      +                    NaN                   0   \n",
       "6390   1  629241      +                      F                   1   \n",
       "6391   1  629243      +                      F                   1   \n",
       "6392   1  629246      +                      F                   1   \n",
       "6394   1  629256      +                      F                   1   \n",
       "6395   1  629258      +                      F                   1   \n",
       "6396   1  629259      +                      F                   1   \n",
       "\n",
       "     unconvertedBaseQualities  unconvertedBaseCount  \n",
       "6386                       FF                     2  \n",
       "6388                      FFF                     3  \n",
       "6389                      FFF                     3  \n",
       "6390                    F:FFF                     5  \n",
       "6391                    FFFFF                     5  \n",
       "6392                    FFFFF                     5  \n",
       "6394                     :FFF                     4  \n",
       "6395                     FFFF                     4  \n",
       "6396                     ,FFF                     4  "
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2.iloc[0:9,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd33cd2b-b961-4884-927a-19e0fcffd5e5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "e2e06de0-54b7-409c-8183-880ad380ac2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "end = ! echo 'ubs_basic.ipnby | ubs_basic.py ENDED  :' `date`\n",
    "end = end[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "aabbe1e4-9117-4f77-ad88-275e25757124",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ubs_basic.ipnby | ubs_basic.py STARTED: Sat Feb 3 05:58:39 AM CST 2024\n",
      "ubs_basic.ipnby | ubs_basic.py ENDED  : Sat Feb 3 06:00:40 AM CST 2024\n"
     ]
    }
   ],
   "source": [
    "print(start)\n",
    "print(end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec8f3b4f-3c53-46f3-aa75-0df42384b7ee",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
